// @nordic/rendering-context/dist/cjs/getters-B8Gz35pU.js
'use strict';

var node_async_hooks = require('node:async_hooks');

const storage = new node_async_hooks.AsyncLocalStorage();
function getContext() {
    const context = storage.getStore();
    {
        return context;
    }
}

function getRenderingAssets() {
    return getContext().assets;
}
function getRenderingAppProps() {
    return getContext().appProps;
}
function getRenderingFlags() {
    return getContext().flags;
}
function getRenderingSupport() {
    return getContext().support;
}
function getRenderingTitle() {
    return getContext().title;
}
function getRenderingMainEntry() {
    return getContext().mainEntry;
}
function getRenderingNonce() {
    return getContext().nonce;
}
function getRenderingCsrfToken() {
    return getContext().csrfToken;
}
function getRenderingWrapperProps() {
    return getContext().wrapperProps;
}
function getRenderingFramework() {
    return getContext().framework;
}

exports.getContext = getContext;
exports.getRenderingAppProps = getRenderingAppProps;
exports.getRenderingAssets = getRenderingAssets;
exports.getRenderingCsrfToken = getRenderingCsrfToken;
exports.getRenderingFlags = getRenderingFlags;
exports.getRenderingFramework = getRenderingFramework;
exports.getRenderingMainEntry = getRenderingMainEntry;
exports.getRenderingNonce = getRenderingNonce;
exports.getRenderingSupport = getRenderingSupport;
exports.getRenderingTitle = getRenderingTitle;
exports.getRenderingWrapperProps = getRenderingWrapperProps;
exports.storage = storage;
