// @nordic/rendering-context/dist/cjs/index.browser.development.js
'use strict';

require('core-js/modules/es.regexp.flags.js');
var error = require('@nordic/utils/error');

class NordicRenderingContextError extends error.OwnerStackError {
  name = 'NordicRenderingContextError';
}
function throwContextNotInitializedError() {
  throw new NordicRenderingContextError('Nordic Rendering Context is not initialized.');
}

function getContext() {
  const context = window._n.ctx.r;
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- Can be undefined in case the render has not been executed into the rendering context
  if (!context) {
    throwContextNotInitializedError();
  }
  return context;
}

function getRenderingAssets() {
  return getContext().assets;
}
function getRenderingAppProps() {
  return getContext().appProps;
}
function getRenderingFlags() {
  return getContext().flags;
}
function getRenderingSupport() {
  return getContext().support;
}
function getRenderingTitle() {
  return getContext().title;
}
function getRenderingMainEntry() {
  return getContext().mainEntry;
}
function getRenderingNonce() {
  return getContext().nonce;
}
function getRenderingCsrfToken() {
  return getContext().csrfToken;
}
function getRenderingWrapperProps() {
  return getContext().wrapperProps;
}
function getRenderingFramework() {
  return getContext().framework;
}

function setRenderingAppProps(appProps) {
  getContext().appProps = appProps;
}

exports.getRenderingAppProps = getRenderingAppProps;
exports.getRenderingAssets = getRenderingAssets;
exports.getRenderingCsrfToken = getRenderingCsrfToken;
exports.getRenderingFlags = getRenderingFlags;
exports.getRenderingFramework = getRenderingFramework;
exports.getRenderingMainEntry = getRenderingMainEntry;
exports.getRenderingNonce = getRenderingNonce;
exports.getRenderingSupport = getRenderingSupport;
exports.getRenderingTitle = getRenderingTitle;
exports.getRenderingWrapperProps = getRenderingWrapperProps;
exports.setRenderingAppProps = setRenderingAppProps;
