// @nordic/rendering-context/dist/cjs/server.production.js
'use strict';

var getters = require('./getters-B8Gz35pU.js');
var env = require('@nordic/env');
var serverContext = require('@meli/server-context');
var frontendConfig = require('frontend-config');
var server = require('react-dom/server');
var server$1 = require('@nordic/core-context/server');
var coreContext = require('@nordic/core-context');
var fsCache = require('@nordic/fs-cache');
var serialize = require('@nordic/utils/serialize');
var nordicCspEngine = require('nordic-csp-engine');
var server$2 = require('@global-theme/context/server');
var assetsConfig = require('@nordic/assets-config');
require('node:async_hooks');

function getRenderingTrustedTypesRegex() {
    return getters.getContext().trustedTypesRegex;
}
function getRenderingCriticalCss() {
    return getters.getContext().criticalCss;
}
function getRenderingIslands() {
    return getters.getContext().islands;
}
function getRenderingMelidata() {
    return getters.getContext().melidata;
}
function getRenderingMetrics(name) {
    return getters.getContext().metrics[name];
}
function getRenderingLayoutView() {
    return getters.getContext().layoutView;
}

function setRenderingIslands(islands) {
    getters.getContext().islands = islands;
}

function validateIfHeadersSent(message) {
    if (!serverContext.getRes().headersSent) {
        return false;
    }
    return true;
}

const CRITICAL_COOKIE_DEFAULT_VALUE = '1';
const CRITICAL_COOKIE_SEPARATOR = '/';
const CRITICAL_COOKIE_MAX_LENGTH = 200;
const THIRTY_DAYS_IN_MS = 1000 * 60 * 60 * 24 * 30;
const REMOVE_LAST_SLASH_REGEX = /\/$/;

function getCookiePath() {
    const basePath = frontendConfig.config.ragnar?.basePath;
    if (!basePath) {
        return undefined;
    }
    // eslint-disable-next-line security/detect-non-literal-regexp
    return new RegExp(basePath.replace(REMOVE_LAST_SLASH_REGEX, '')).exec(serverContext.getReq().baseUrl)?.[0];
}
function setCookie({ name, path, value = CRITICAL_COOKIE_DEFAULT_VALUE, ttl = THIRTY_DAYS_IN_MS, }) {
    const req = serverContext.getReq();
    serverContext.getRes().cookie(name, value, {
        httpOnly: true,
        expires: new Date(Date.now() + ttl),
        sameSite: false,
        domain: req.hostname,
        secure: req.secure,
        path,
    });
}

function setRenderingCriticalCssCookie() {
    const { name, path, values } = getRenderingCriticalCss();
    if (!values.size) {
        return;
    }
    values.add(env.env.VERSION);
    const value = Array.from(values).join(CRITICAL_COOKIE_SEPARATOR);
    if (value.length <= CRITICAL_COOKIE_MAX_LENGTH) {
        setCookie({ name, value, path });
        return;
    }
}
function setRenderingCriticalCssValue(value) {
    if (validateIfHeadersSent()) {
        return false;
    }
    const { values } = getRenderingCriticalCss();
    const changed = !values.has(value);
    if (changed) {
        values.add(value);
    }
    return changed;
}

function getRenderingIIFESource(name) {
    const { iifeScripts } = getters.getContext();
    const src = iifeScripts.get(name);
    if (!src) {
        return null;
    }
    // Only the first one should inject the script
    iifeScripts.delete(name);
    return src;
}

const METRICS_ERRORS = {
    hotjar: 'The <Hotjar> component can only be used before the initial server response.\n' +
        'Please make sure the <Hotjar> component is not inside a suspended component.',
    googleTagManager: 'The <GTM> component can only be used before the initial server response.\n' +
        'Please make sure the <GTM> component is not inside a suspended component.',
    qualtricsInsights: 'The <QualtricsInsights> component can only be used before the initial server response.\n' +
        'Please make sure the <QualtricsInsights> component is not inside a suspended component.',
};

function setRenderingMetrics(name, props) {
    if (validateIfHeadersSent(METRICS_ERRORS[name])) {
        return;
    }
    const store = getRenderingMetrics(name);
    const storeProps = store.props;
    store.enabled = true;
    Object.entries(props).forEach(([key, value]) => {
        if (Array.isArray(storeProps[key]) && Array.isArray(value)) {
            storeProps[key].push(...value);
        }
        else {
            storeProps[key] = value;
        }
    });
}
function setHotjarProps(props) {
    setRenderingMetrics('hotjar', props);
}
function setGTMProps(props) {
    setRenderingMetrics('googleTagManager', props);
}
function setQualtricsInsightsProps(props) {
    setRenderingMetrics('qualtricsInsights', props);
}

function getRenderToString() {
    return getters.getRenderingFlags().staticMarkup
        ? server.renderToStaticMarkup
        : server.renderToString;
}

const APP_INFO_SCRIPT = {
    id: '__NORDIC_APP_INFO__',
    children: `_n.appInfo=${serialize.serializeJSON({
        name: env.env.APPLICATION,
        version: env.env.VERSION,
        scope: env.env.SCOPE,
        nordicVersion: env.env.NORDIC_VERSION,
        nordicWebkitVersion: env.env.NORDIC_WEBKIT_VERSION,
    })};`,
};
const NATIVE_ACTION_DEBUG_SCRIPT = {
    src: '@nordic/native-action/dist/iife/debug.js',
    inline: true,
};
const NATIVE_ACTION_SCRIPT = {
    src: '@nordic/native-action/dist/iife/index.js',
    inline: true,
};
const PIPEABLE_STREAM_SCRIPT = {
    src: '@nordic/streaming-context/dist/iife/readable.js',
    inline: true,
};

function getSerializableContext() {
    const context = getters.getContext();
    const { assets, flags } = context;
    let { appProps } = context;
    const result = {
        flags,
        assets: { cdn: assets.cdn, prefix: assets.prefix },
        support: context.support,
        title: context.title,
        framework: context.framework,
        nonce: context.nonce,
        csrfToken: context.csrfToken,
    };
    if (flags.staticMarkup) {
        return result;
    }
    appProps = flags.pipeableStream ? undefined : appProps;
    if (flags.islands) {
        appProps = appProps && { sharedState: appProps.sharedState };
    }
    else {
        result.wrapperProps = context.wrapperProps;
    }
    result.appProps = appProps;
    result.mainEntry = context.mainEntry;
    return result;
}
function getRenderingContextInitScript() {
    const { mainEntry, flags } = getters.getContext();
    let serializedManifest = '';
    let serializedMainScriptsNames = '';
    let serializedMainStylesNames = '';
    if (!flags.staticMarkup) {
        serializedManifest = fsCache.assetsCache.serializedManifest;
        if (mainEntry) {
            serializedMainScriptsNames =
                fsCache.assetsCache.getSerializedMainScriptsNames(mainEntry) ?? '';
            serializedMainStylesNames =
                fsCache.assetsCache.getSerializedMainStylesNames(mainEntry) ?? '';
        }
    }
    return {
        id: '__NORDIC_RENDERING_CTX__',
        children: `_n.ctx.r=${serialize.serializeJSON(getSerializableContext())};` +
            `_n.ctx.r.assets.manifest=new Map(${serializedManifest});` +
            `_n.ctx.r.assets.mainAssetsNames={scripts:new Set(${serializedMainScriptsNames}),styles:new Set(${serializedMainStylesNames})};`,
    };
}
function getNativeActionInitScripts() {
    const webkit = coreContext.getWebviewInfo()?.webkit;
    if (webkit?.engine !== 'v3') {
        return null;
    }
    if (webkit.debugMode) {
        return [APP_INFO_SCRIPT, NATIVE_ACTION_DEBUG_SCRIPT, NATIVE_ACTION_SCRIPT];
    }
    return [APP_INFO_SCRIPT, NATIVE_ACTION_SCRIPT];
}
function getPipeableStreamInitScript() {
    const { appProps, flags } = getters.getContext();
    if (!appProps ||
        !flags.pipeableStream ||
        flags.staticMarkup ||
        (flags.islands && !appProps.sharedState)) {
        return null;
    }
    return PIPEABLE_STREAM_SCRIPT;
}

function getRenderingContextInitScripts() {
    const { flags } = getters.getContext();
    return [
        server$1.getCoreContextInitScript(flags.staticMarkup),
        getRenderingContextInitScript(),
        getNativeActionInitScripts(),
        getPipeableStreamInitScript(),
    ];
}

function createCriticalCssStore(path) {
    const cookieName = `c_${env.env.APP_NAME_HASH}`;
    const cookieValue = serverContext.getReq().cookies[cookieName];
    const currentValues = cookieValue?.split(CRITICAL_COOKIE_SEPARATOR);
    return {
        name: cookieName,
        path: path || getCookiePath(),
        values: currentValues?.includes(env.env.VERSION)
            ? new Set(currentValues)
            : new Set(),
    };
}

const NoopWrapper = ({ children, }) => children;

const FONTS_CSS_COMMENT = '/*! * Web Fonts from Fontspring.com * * All OpenType features and all extended glyphs have been removed. *' +
    ' Fully installable fonts can be purchased at https://www.fontspring.com * *' +
    ' The fonts included in this stylesheet are subject to the End User License you purchased * from Fontspring.' +
    ' The fonts are protected under domestic and international trademark and * copyright law.' +
    ' You are prohibited from modifying, reverse engineering, duplicating, or * distributing this font software.' +
    ' * * (c) 2010-2023 Fontspring * * * * * The fonts included are copyrighted by the vendor listed below.' +
    ' * * Vendor: Mark Simonson Studio * License URL: https://www.fontspring.com/licenses/mark-simonson-studio/webfont * * */';
const FONTS_DATA = [
    { name: 'light', weight: 300 },
    { name: 'regular', weight: 400 },
    { name: 'semibold', weight: 600 },
];
async function getFontResources(cdn) {
    const links = [];
    const globalThemeFonts = await server$2.getFonts(cdn);
    if (globalThemeFonts) {
        globalThemeFonts.preload?.forEach((href) => {
            links.push({
                href,
                rel: 'preload',
                as: 'font',
                type: 'font/woff2',
                crossOrigin: 'anonymous',
            });
        });
        return { links, style: globalThemeFonts.css };
    }
    const css = FONTS_DATA.reduce((acc, { name, weight }) => {
        const url = `${cdn}/ui/webfonts/v3.0.0/proxima-nova/proximanova-${name}`;
        links.push({
            href: `${url}.woff2`,
            rel: 'preload',
            as: 'font',
            type: 'font/woff2',
            crossOrigin: 'anonymous',
        });
        return (`${acc}@font-face{font-family:'Proxima Nova';font-weight:${weight};font-display:swap;font-style:normal;` +
            `src:url(${url}.woff2) format("woff2"),url(${url}.woff) format("woff")}`);
    }, '');
    return {
        links,
        style: { id: '__FONTS__', children: `${FONTS_CSS_COMMENT} ${css}` },
    };
}

async function createLayoutViewStore(cdn) {
    const [globalThemeContextScript, globalThemeStyles, fontsResources] = await Promise.all([
        server$2.getGlobalThemeContextScript(),
        server$2.getStyles(),
        getFontResources(cdn),
    ]);
    return {
        lang: '',
        headSlots: {
            title: NoopWrapper,
            links: fontsResources.links,
            meta: [],
            styles: [fontsResources.style, globalThemeStyles],
            scripts: [globalThemeContextScript],
        },
        bodySlots: {
            start: [],
            main: NoopWrapper,
            end: [],
            scripts: [],
        },
    };
}

function getConfigProps(storeName) {
    const platform = coreContext.getPlatform();
    return ((platform
        ? frontendConfig.config.get(storeName, platform.id, platform.siteId)
        : frontendConfig.config[storeName]) ?? {});
}

async function createHotjarStore() {
    const req = serverContext.getReq();
    return {
        enabled: false,
        props: {},
        configProps: getConfigProps('hotjar'),
        debugMode: !!(req.header('x-hotjar-debug') || req.query.hotjar_debug),
        hasUserConsent: (await req.meliCookies?.userConsentCookieCategory('traceability')) ??
            false,
    };
}

async function getUserConsentCookieCategory(category) {
    const req = serverContext.getReq();
    return {
        [`user_consent_cookies_${category}`]: await req.meliCookies.userConsentCookieCategory(category),
    };
}
async function createCommonDataLayer() {
    const common = [];
    const req = serverContext.getReq();
    try {
        const cookiesPrefs = await req.meliCookies?.getCookiesUserPreferences();
        if (cookiesPrefs?.categories) {
            const consents = await Promise.all(Object.keys(cookiesPrefs.categories).map(getUserConsentCookieCategory));
            common.push(...consents);
        }
    }
    catch {
        // Should never fail, but just in case
    }
    if (typeof req.data_audience_3p_denied !== 'undefined') {
        common.push({ data_audience_3p_denied: req.data_audience_3p_denied });
    }
    return common;
}
async function createGTMStore() {
    return {
        enabled: false,
        debugMode: false,
        props: {},
        configProps: getConfigProps('googleTagManager'),
        commonDataLayer: await createCommonDataLayer(),
    };
}

function createQualtricsInsightsStore() {
    const req = serverContext.getReq();
    return {
        enabled: false,
        props: {},
        configProps: getConfigProps('qualtricsInsights'),
        debugMode: !!(req.header('x-qualtrics-insights-debug') ||
            req.query.qualtrics_insights_debug),
    };
}

async function createMetricsStore() {
    const [hotjar, googleTagManager] = await Promise.all([
        createHotjarStore(),
        createGTMStore(),
    ]);
    return {
        hotjar,
        googleTagManager,
        qualtricsInsights: createQualtricsInsightsStore(),
    };
}

function getAssetsData(mainEntry) {
    return {
        cdn: assetsConfig.getAssetsCDN(),
        prefix: assetsConfig.getAssetsPrefix(),
        manifest: fsCache.assetsCache.manifest,
        mainAssetsNames: mainEntry
            ? {
                scripts: fsCache.assetsCache.getMainScriptsNames(mainEntry),
                styles: fsCache.assetsCache.getMainStylesNames(mainEntry),
            }
            : { scripts: new Set(), styles: new Set() },
    };
}

function createIIFEScriptsStore(flags, support) {
    const entries = [
        ['n-safe-area', '@nordic/safe-area/dist/iife/index.js'],
        ['n-img-blur', '@nordic/image/dist/iife/blur.js'],
        ['n-img-fallback', '@nordic/image/dist/iife/fallback.js'],
        ['n-img-lazy', '@nordic/image/dist/iife/lazy.js'],
        [
            'n-img-handler',
            `@nordic/image/dist/iife/handler${support.customElements.builtIn ? '' : '-polyfill'}.js`,
        ],
    ];
    if (flags.islands) {
        entries.push(['n-islands-core', '@nordic/islands/dist/iife/core.js'], ['n-islands-visible', '@nordic/islands/dist/iife/visible.js']);
    }
    return new Map(entries);
}

function getWrapperProps() {
    const { id, className } = serverContext.getReq().nordicPages?.pageSettings ?? {};
    if (!className) {
        return undefined;
    }
    return { id, className };
}

async function runInRenderingContext(initialValues, callback) {
    const req = serverContext.getReq();
    const { browser } = req;
    const { support } = browser;
    const { appProps, flags, mainEntry } = initialValues;
    const assetsData = getAssetsData(mainEntry);
    const [securityConfig, metricsStore, layoutViewStore] = await Promise.all([
        nordicCspEngine.websecConfig.get('securityConfig'),
        createMetricsStore(),
        createLayoutViewStore(assetsData.cdn),
    ]);
    return getters.storage.run({
        mainEntry,
        appProps: !appProps ||
            mainEntry ||
            flags.staticMarkup ||
            flags.stream ||
            req.nordicPages
            ? appProps
            : appProps.sharedState && { sharedState: appProps.sharedState },
        melidata: initialValues.melidata,
        metrics: metricsStore,
        iifeScripts: createIIFEScriptsStore(flags, support),
        title: initialValues.title ?? '',
        nonce: req.csp?.nonce,
        csrfToken: req.csrfToken?.(),
        trustedTypesRegex: securityConfig?.trustedTypes.regex,
        assets: assetsData,
        criticalCss: createCriticalCssStore(initialValues.criticalCssPath),
        wrapperProps: getWrapperProps(),
        layoutView: layoutViewStore,
        framework: initialValues.framework,
        flags: {
            isBot: req.isBot,
            loadable: flags.loadable,
            islands: flags.islands,
            pipeableStream: flags.pipeableStream,
            stream: flags.stream,
            staticMarkup: req.isBot || browser.outdated || flags.staticMarkup,
        },
        support: {
            customElements: support.customElements,
            imgLoading: support.imgLoading,
            prefetch: support.prefetch,
            sameSite: support.sameSite,
            outdated: browser.outdated,
            share: support.share,
        },
    }, callback);
}

exports.getRenderToString = getRenderToString;
exports.getRenderingContextInitScripts = getRenderingContextInitScripts;
exports.getRenderingCriticalCss = getRenderingCriticalCss;
exports.getRenderingIIFESource = getRenderingIIFESource;
exports.getRenderingIslands = getRenderingIslands;
exports.getRenderingLayoutView = getRenderingLayoutView;
exports.getRenderingMelidata = getRenderingMelidata;
exports.getRenderingMetrics = getRenderingMetrics;
exports.getRenderingTrustedTypesRegex = getRenderingTrustedTypesRegex;
exports.runInRenderingContext = runInRenderingContext;
exports.setGTMProps = setGTMProps;
exports.setHotjarProps = setHotjarProps;
exports.setQualtricsInsightsProps = setQualtricsInsightsProps;
exports.setRenderingCriticalCssCookie = setRenderingCriticalCssCookie;
exports.setRenderingCriticalCssValue = setRenderingCriticalCssValue;
exports.setRenderingIslands = setRenderingIslands;
