// @nordic/fs-cache/src/assets/fs.ts
import path from 'node:path';

import { getReq } from '@meli/server-context';
import { ASSETS_CONFIG, getAssetsPrefix } from '@nordic/assets-config';
import { isTestEnv, isBlurableImg } from '@nordic/utils/validation';

import { readBuildFileContentSync, readFileContentSync } from '../utils/fs';
import { NordicBlurNotFoundError } from './error/blur-not-found';
import {
  SOURCE_MAP_COMMENT_REGEX,
  RELATIVE_PATH_REGEX,
  LAST_DOT_REGEX,
} from './regex';

function readAssetFile(filePath: string, fromBuild = false) {
  const readFile = fromBuild ? readBuildFileContentSync : readFileContentSync;
  const fileContent = readFile(filePath)
    ?.replace(SOURCE_MAP_COMMENT_REGEX, '')
    .trim();

  if (!__DEV__ || !fileContent) {
    return fileContent;
  }

  const req = getReq();
  // `@nordic-dev/pretty-errors` needs a full URL to parse sourcemaps from inline scripts.
  const hostURL = `${req.protocol}://${req.header('host')!}`;
  const relativeFilePath = path.relative(process.cwd(), path.resolve(filePath));

  return `${fileContent}\n//# sourceURL=${hostURL + getAssetsPrefix() + relativeFilePath}`;
}

function readImageAsset(assetName: string, fileName: string) {
  const content = readFileContentSync(
    path.join(
      ASSETS_CONFIG.IMAGES_BLUR_DIR,
      `${ASSETS_CONFIG.IMAGES_BLUR_PREFIX}${fileName.replace(LAST_DOT_REGEX, '-')}`,
    ),
  );

  // `__DEV__` is replaced by true or false in building, `isTestEnv()` is only for development code, to allow unit tests execution without app build
  if (__DEV__ && !content && !isTestEnv()) {
    throw new NordicBlurNotFoundError(assetName);
  }

  return content;
}

function readNodeModulesAsset(assetName: string, isSmartPOS?: boolean) {
  const nodeModulesPath = path.join('node_modules', assetName);
  const { dir, name, ext } = path.parse(nodeModulesPath);
  const envExt = __DEV__ ? '.development' : '.production';
  const smartPOSExt = isSmartPOS ? '.smartpos' : '';
  const basePath = path.join(dir, name);
  const envBasePath = `${basePath}${envExt}`;

  if (!smartPOSExt) {
    return (
      readAssetFile(`${envBasePath}${ext}`) ?? readAssetFile(nodeModulesPath)
    );
  }

  const smartPOSBasePath = `${basePath}${smartPOSExt}`;

  return (
    readAssetFile(`${smartPOSBasePath}${envExt}${ext}`) ??
    readAssetFile(`${smartPOSBasePath}${ext}`) ??
    readAssetFile(`${envBasePath}${ext}`) ??
    readAssetFile(nodeModulesPath)
  );
}

export function readAsset(
  assetName: string,
  fileName: string | undefined,
  isSmartPOS?: boolean,
) {
  if (RELATIVE_PATH_REGEX.test(assetName)) {
    return { content: readAssetFile(assetName) };
  }

  if (fileName) {
    return {
      content: isBlurableImg(assetName)
        ? readImageAsset(assetName, fileName)
        : readAssetFile(fileName, true),
    };
  }

  const content = readNodeModulesAsset(assetName, isSmartPOS);

  if (content) {
    return { content, fromNodeModules: true };
  }

  return { content: readAssetFile(assetName) };
}
