// @nordic/fs-cache/src/assets/index.ts
import type { MainAssets } from './manifest/types';

import { getAssetsPrefix } from '@nordic/assets-config';

import { readLoadableStats } from '../utils/webpack';
import { readAsset } from './fs';
import { createManifests } from './manifest';
import { LimitedStringMap } from './limited-map';

export type MainEntry = Array<string> | Set<string> | string;

class AssetsCache {
  // TODO: Remove when webpack support is dropped
  #loadableStats: Readonly<object> | null = null;

  #manifest = new Map<string, string>();

  #mainAssetsManifest = new Map<string, MainAssets>();

  #assetsContent = new LimitedStringMap();

  #serializedManifest = '';

  get loadableStats() {
    return this.#loadableStats;
  }

  get manifest() {
    return this.#manifest;
  }

  get mainAssetsManifest() {
    return this.#mainAssetsManifest;
  }

  get serializedManifest() {
    return this.#serializedManifest;
  }

  initialize() {
    ({
      manifest: this.#manifest,
      mainAssetsManifest: this.#mainAssetsManifest,
      serializedManifest: this.#serializedManifest,
    } = createManifests());

    this.#loadableStats = readLoadableStats();
  }

  getAssetContent(assetName: string, isSmartPOS?: boolean) {
    const cachedContent = this.#assetsContent.get(assetName);

    if (cachedContent) {
      return cachedContent;
    }

    const { content, fromNodeModules } = readAsset(
      assetName,
      this.#manifest.get(assetName),
      isSmartPOS,
    );

    if (content) {
      if (!__DEV__ || fromNodeModules) {
        this.#assetsContent.set(assetName, content);
      }

      return content;
    }

    return null;
  }

  getMainScripts(mainEntry: MainEntry) {
    return this.#getMainAssets(mainEntry, 'scripts', 'sources');
  }

  getMainScriptsNames(mainEntry: MainEntry) {
    return this.#getMainAssets(mainEntry, 'scripts', 'names');
  }

  getMainStyles(mainEntry: MainEntry) {
    return this.#getMainAssets(mainEntry, 'styles', 'sources');
  }

  getMainStylesNames(mainEntry: MainEntry) {
    return this.#getMainAssets(mainEntry, 'styles', 'names');
  }

  // TODO: Remove when webpack support is dropped
  getSerializedMainScriptsNames(mainEntry: string) {
    return this.#mainAssetsManifest.get(mainEntry)?.scripts.serializedNames;
  }

  // TODO: Remove when webpack support is dropped
  getSerializedMainStylesNames(mainEntry: string) {
    return this.#mainAssetsManifest.get(mainEntry)?.styles.serializedNames;
  }

  #getMainAssets(
    mainEntry: MainEntry,
    type: 'scripts' | 'styles',
    values: 'names' | 'sources',
  ) {
    const assets = new Set<string>();
    const prefix = values === 'sources' ? getAssetsPrefix() : '';

    const addAsset = (entry: string) => {
      this.#mainAssetsManifest
        .get(entry)
        ?.[type][values].forEach((asset) => assets.add(prefix + asset));
    };

    if (Array.isArray(mainEntry) || mainEntry instanceof Set) {
      mainEntry.forEach(addAsset);
    } else {
      addAsset(mainEntry);
    }

    return assets;
  }
}

export const assetsCache = new AssetsCache();
