// frontend-country_config/utils/update-data.js
const { RestClient, buildFlowStarterContext } = require('frontend-restclient');
const env = require('frontend-env');
const cloneDeep = require('lodash/cloneDeep');

const countriesJson = require('../config/countries.json');
const packageJson = require('../package.json');

const { PRODUCTION } = env;

/**
 * @typedef {import('../types/index').Country} Country
 * @typedef {import('../types/index').CountryId} CountryId
 * @typedef {Record<CountryId, Country>} Countries
 * @typedef {import('../types/index').Country['currencies'][string]} Currency
 * @typedef {import('../types/internal').Sites} Sites
 * @typedef {import('../types/internal').Site} Site
 * @typedef {import('frontend-restclient').RestClientConfig} RestClientConfig
 */

let countriesData = /** @type {Countries} */ (
  /** @type {unknown} */ (countriesJson)
);

/** @type {RestClientConfig} */
const restclientOptions = {
  timeout: PRODUCTION ? 3000 : 10000,
  headers: { 'User-Agent': `${packageJson.name}/${packageJson.version}` },
  retry: {
    maxRetries: 2,
    delay: 0,
  },
  logErrors: false,
};
const siteCoreRestclient = RestClient({
  ...restclientOptions,
  baseURL: env.FURY
    ? 'http://site-core.melisystems.com'
    : 'https://site-core.melioffice.com',
});
const countriesRestclient = RestClient({
  ...restclientOptions,
  baseURL: env.FURY
    ? 'http://addresses-masters.melisystems.com'
    : 'https://addresses-masters.melioffice.com',
});

/**
 * Gets the countries from the countries API
 * @returns {Promise<Array<Country>>}
 */
// eslint-disable-next-line @typescript-eslint/require-await
const getCountries = async () =>
  countriesRestclient
    .get('/bulk/countries', {
      context: buildFlowStarterContext(),
    })
    .then((res) => res.data);

/**
 * Fetches data from the requested endpoint
 * @param {string} endpoint
 */
async function getDataFromEndpoint(endpoint) {
  const config = {
    context: buildFlowStarterContext(),
  };

  const response = await siteCoreRestclient.get(endpoint, config);

  if (response.status !== 200) {
    throw new Error(
      `Request to ${endpoint} was not successful. Status code: ${response.status}`,
    );
  }

  // return the raw data instead of restclient structure response
  return response.data;
}

/**
 * Determine the currencies for a given country.
 *
 * @param {Country} country
 * @param {Currency[]} currencies
 * @param {Site[]} sites
 */
function determineCurrencies(country, currencies, sites) {
  /** @type {{[key:string]: Currency}} */
  const completeCurrenciesList = {};

  const currenciesFound = sites.find(
    (site) =>
      site.name === country.name ||
      // Fix name inconsitencies between Sites API and Countries API.
      (site.name === 'Dominicana' && country.name === 'República Dominicana') ||
      (site.name === 'Perú' && country.name === 'Peru'),
  );

  const partialCurrenciesList = currenciesFound
    ? currenciesFound.currencies
    : // Fix inconsitencies: 'Puerto Rico' extists in Countries API but not in Sites API.
      [{ id: country.currency_id }];

  partialCurrenciesList.forEach((siteCurrency) => {
    const found = currencies.find(
      (currency) => siteCurrency.id === currency.id,
    );

    if (found) {
      completeCurrenciesList[siteCurrency.id] = found;
    }
  });

  return completeCurrenciesList;
}

/**
 * Gets the data from the different APIs
 */
async function getData() {
  /**
   * @type {[Country[], Currency[], Sites]}
   */
  const [countries, currenciesRaw, sitesRaw] = await Promise.all([
    getCountries(),
    getDataFromEndpoint('currencies'),
    getDataFromEndpoint('sites'),
  ]);
  /**
   * @type {Promise<Site>[]}
   */
  const sitesPromises = sitesRaw.map(async ({ id }) =>
    getDataFromEndpoint(`sites/${id}`),
  );

  const sites = await Promise.all(sitesPromises);

  return countries.reduce((acc, country) => {
    country.currencies = determineCurrencies(country, currenciesRaw, sites);
    acc[country.id] = country;

    return acc;
  }, /** @type {Countries} */ ({}));
}

async function updateCountriesData() {
  return getData()
    .then((data) => {
      countriesData = data;

      return data;
    })
    .catch(() => {});
}

/**
 * This method looks for the process environment value, then from the
 * default specified as an argument.
 * @param {string | undefined} envValue
 * @param {boolean} defaultValue
 */
function initChecker(envValue, defaultValue) {
  return typeof envValue === 'undefined'
    ? defaultValue
    : Boolean(JSON.parse(envValue));
}

function getCountriesData() {
  return cloneDeep(countriesData);
}

module.exports = {
  getData,
  initChecker,
  getDataFromEndpoint,
  updateCountriesData,
  getCountriesData,
};
