// node-melitk-otel/lib/events/tracing.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.newEventsResourceBuilder = exports.newSpanProcessors = exports.newSampler = exports.newO11YEventsTracing = void 0;
const exporter_trace_otlp_grpc_1 = require("@opentelemetry/exporter-trace-otlp-grpc");
const frontend_env_1 = __importDefault(require("frontend-env"));
const api_1 = require("@opentelemetry/api");
const resources_1 = require("@opentelemetry/resources");
const semantic_conventions_1 = require("@opentelemetry/semantic-conventions");
const grpc_js_1 = require("@grpc/grpc-js");
const config_1 = require("./config/config");
const activator_1 = __importDefault(require("./activator"));
const alwaysRecord_1 = require("./sampler/alwaysRecord");
const batchAllSpanProcessor_1 = require("./processor/batch/batchAllSpanProcessor");
const noopActivator_1 = require("./activator/noopActivator");
const telemetry_1 = __importDefault(require("./sampler/telemetry"));
const configs_1 = require("../configs");
const isAdaptiveSamplingEnabled = () => frontend_env_1.default.ADAPTIVE_SAMPLING_ENABLED === 'true';
/**
 * Returns a new tracer exporter instance for the creation of Events, connecting to the Events OTLP Exporter.
 * @param {OTELConfig} config The O11y Events OTel configs.
 * @returns {OTLPTraceExporter} The tracer exporter for the creation of Events.
 */
const newTraceExporter = (config) => {
    api_1.diag.debug(`${configs_1.O11Y_EVENTS_LOG_PREFIX}Exporter configurations
    - endpoint: ${config.exporter.otlp.endpoint}
    - timeout: ${config.exporter.otlp.exportTimeout}`);
    const collectorOptions = {
        url: config.exporter.otlp.endpoint,
        metadata: new grpc_js_1.Metadata({ waitForReady: true }), // prevent UNAVAILABLE
        timeoutMillis: +config.exporter.otlp.exportTimeout,
    };
    return new exporter_trace_otlp_grpc_1.OTLPTraceExporter(collectorOptions);
};
/**
 * Creates a new {@link Sampler} instance.
 * If events are enabled, wrap the provided default sampler into an events sampler chain, the {@link AlwaysRecordSampler} instance.
 * @param {OTELConfig} config The O11y Events OTel configs.
 * @param {Sampler} defaultSampler The default sampler to wrap.
 * @param {Sampler} pingSampler The ping sampler to wrap.
 * @returns {Sampler} The sampler instance.
 */
const newSampler = (configs, defaultSampler, pingSampler) => new alwaysRecord_1.AlwaysRecordSampler(defaultSampler, pingSampler, configs.activator, new telemetry_1.default());
exports.newSampler = newSampler;
/**
 * Returns a new tracer provider instance for the creation of Events.
 * @param {OTLPTraceExporter} defaultTracerExporter The default tracing exporter, used to create the {@link BatchSpanProcessor}.
 * @param {OTLPTraceExporter} eventsTraceExporter The tracing exporter for the creation of Events, used to create the {@link BatchAllSpanProcessor}.
 * @param {OTELConfig} config The O11y Events OTel configs.
 * @param {Activator} eventsActivator The activator, to check whether the Events are enabled or not.
 * @param {MeterProvider} meterProvider The meter provider, used to export the metrics generated on Events.
 * @returns {NodeTracerProvider} The tracer provider for the creation of Events.
 */
const newSpanProcessors = (configs) => {
    api_1.diag.debug(`${configs_1.O11Y_EVENTS_LOG_PREFIX}Span processor configurations
    - scheduleDelay: ${configs.settings.o11yEvents.otel.processor.scheduleDelay}
    - exportTimeout: ${configs.settings.o11yEvents.otel.processor.exportTimeout}
    - maxQueueSize: ${configs.settings.o11yEvents.otel.processor.maxQueueSize}
    - maxExportBatchSize: ${configs.settings.o11yEvents.otel.processor.maxExportBatchSize}`);
    const eventsExporter = newTraceExporter(configs.settings.o11yEvents.otel);
    /**
     * TODO: The AttributePropagatingSpanProcessor is not working as expected.
     * Currently, it is not propagating the `x_request_id` attribute from the parent span to the child span.
     * We need to investigate why it is not working and fix it on the technical debt, as it follows:
     * https://mercadolibre.atlassian.net/browse/O11YEVENTS-183
     */
    // const attributePropagatingProcessor = new AttributePropagatingSpanProcessor(
    //   eventsActivator,
    // );
    // tracerProvider.addSpanProcessor(attributePropagatingProcessor);
    const batchAllSpanProcessor = new batchAllSpanProcessor_1.BatchAllSpanProcessor(eventsExporter, configs.activator, configs.settings.o11yEvents.otel.processor);
    return Array.of(batchAllSpanProcessor);
};
exports.newSpanProcessors = newSpanProcessors;
/**
 * Builds the resource for the tracer provider, containing the necessary attributes for Events creation.
 * @param {AttributesConfig} config The O11y Events configs.
 * @returns {Resource} The resource for the tracer provider.
 */
const newEventsResourceBuilder = (configs) => (0, resources_1.resourceFromAttributes)({
    [semantic_conventions_1.ATTR_SERVICE_NAME]: configs.settings.o11yEvents.otel.attributes.serviceName,
    [semantic_conventions_1.ATTR_SERVICE_VERSION]: configs.settings.o11yEvents.otel.attributes.serviceVersion,
    'fury.application': configs.settings.o11yEvents.otel.attributes.furyApplication,
    'fury.service': configs.settings.o11yEvents.otel.attributes.furyService,
    'fury.criticality': configs.settings.o11yEvents.otel.attributes.furyCriticality,
});
exports.newEventsResourceBuilder = newEventsResourceBuilder;
/**
 * Creates a new Builder to help instrument the
 * events Tracer Provider if it's enabled on the chain, and with all the
 * configurations needed to work. From it, you can get the Sampler and
 * the TracerProviderOptions to instrument a new TracerProvider that works with events.
 * @returns {O11YEventsTracing} The builder to help instrument the events Tracer Provider.
 */
const newO11YEventsTracing = () => {
    const settings = (0, config_1.loadConfigs)();
    const logger = api_1.diag.createComponentLogger({
        namespace: 'events',
    });
    let activator = new noopActivator_1.NoopActivator();
    if (!isAdaptiveSamplingEnabled()) {
        activator = (0, activator_1.default)({
            configs: settings.o11yEvents.activator,
            service: settings.o11yEvents.otel.attributes.serviceName,
        });
    }
    return {
        activator,
        settings,
        logger,
    };
};
exports.newO11YEventsTracing = newO11YEventsTracing;
