// @nordic/fs-cache/src/utils/fs.ts
/* eslint-disable security/detect-non-literal-fs-filename */
import path from 'node:path';
import fs from 'node:fs';
import fsPromises from 'node:fs/promises';

import { IS_VITE } from './constants';
import {
  readBuildFileContent as readWebpackBuildFileContent,
  readBuildFileContentSync as readWebpackBuildFileContentSync,
} from './webpack';
import {
  getBuildFilePath as getViteBuildFilePath,
  readBuildFileContent as readViteBuildFileContent,
  readBuildFileContentSync as readViteBuildFileContentSync,
} from './vite';

const TEXT_ENCODER = new TextEncoder();

export function getSizeInMemoryOfString(str: string) {
  return TEXT_ENCODER.encode(str).length;
}

/* v8 ignore next -- Temp condition while webpack support is maintained */
export const readBuildFileContent = IS_VITE
  ? readViteBuildFileContent
  : readWebpackBuildFileContent;

// TODO: Move Vite methods to this file when webpack support is dropped
export const getBuildFilePath = getViteBuildFilePath;

/* v8 ignore next -- Temp condition while webpack support is maintained */
export const readBuildFileContentSync = IS_VITE
  ? readViteBuildFileContentSync
  : readWebpackBuildFileContentSync;

export async function readFileContent(filePath: string) {
  try {
    return await fsPromises.readFile(path.resolve(filePath), 'utf8');
  } catch {
    return null;
  }
}

export function readFileContentSync(
  filePath: string,
  encoding: BufferEncoding = 'utf8',
) {
  try {
    return fs.readFileSync(path.resolve(filePath), encoding);
  } catch {
    return null;
  }
}

export function readJsonFileSync<T>(filePath: string) {
  const fileContent = readFileContentSync(filePath);

  if (!fileContent) {
    return null;
  }

  return JSON.parse(fileContent) as T;
}
