// @nordic/context/src/getters/index.ts
import {
  getDeviceType as getCoreDeviceType,
  getPlatformType as getCorePlatformType,
  getWebviewInfo as getCoreWebviewInfo,
} from '@nordic/core-context';
import {
  getRenderingAssets,
  getRenderingCsrfToken,
  getRenderingNonce,
} from '@nordic/rendering-context';

/**
 * Returns the prefix for the assets for the current page.
 *
 * **⚠️ WARNING**: This function only works within React components, rendering stage, or client-side.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getassetsprefix | Documentation}
 *
 * @example
 *
 * import { getAssetsPrefix } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const assetsPrefix = getAssetsPrefix();
 *
 *  return (
 *    <h1>
 *      Assets Prefix: {assetsPrefix}
 *    </h1>
 *  );
 * };
 */
export function getAssetsPrefix() {
  return getRenderingAssets().prefix;
}

/**
 * Returns the generated CSRF token for the current page.
 *
 * **⚠️ WARNING**: This function only works within React components, rendering stage, or client-side.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getcsrftoken | Documentation}
 *
 * @example
 *
 * import { getCsrfToken } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const csrfToken = getCsrfToken();
 *
 *  return (
 *    <h1>
 *      CSRF Token: {csrfToken}
 *    </h1>
 *  );
 * };
 */
export const getCsrfToken = getRenderingCsrfToken;

/**
 * Returns the generated nonce token for the current page.
 *
 * **⚠️ WARNING**: This function only works within React components, rendering stage, or client-side.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getnonce | Documentation}
 *
 * @example
 *
 * import { getNonce } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const nonce = getNonce();
 *
 *  return (
 *    <h1>
 *      Nonce: {nonce}
 *    </h1>
 *  );
 * };
 */
export const getNonce = getRenderingNonce;

/**
 * Provides the device type anywhere, including React Components.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getdevicetype | Documentation}
 *
 * @example
 *
 * import { getDeviceType } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const deviceType = getDeviceType();
 *
 *  return (
 *    <h1>
 *      Device Type: {deviceType}
 *    </h1>
 *  );
 * };
 */
export const getDeviceType = getCoreDeviceType;

/**
 * Provides the platform type anywhere, including React Components.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getplatformtype | Documentation}
 *
 * @example
 *
 * import { getPlatformType } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const platformType = getPlatformType();
 *
 *  return (
 *    <h1>
 *      Platform Type: {platformType}
 *    </h1>
 *  );
 * };
 */
export const getPlatformType = getCorePlatformType;

/**
 * Provides the webview information anywhere, including React Components.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getwebviewinfo | Documentation}
 *
 * @example
 *
 * import { getWebviewInfo } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const webviewInfo = getWebviewInfo();
 *
 *  return (
 *    <h1>
 *      Webview Info: {JSON.stringify(webviewInfo)}
 *    </h1>
 *  );
 * };
 */
export const getWebviewInfo = getCoreWebviewInfo;

export type {
  NordicDeviceType,
  NordicPlatformType,
  NordicNativeAppInfo,
  NordicWebkitEngine,
  NordicWebkitInfo,
  NordicWebkitScope,
  NordicWebviewInfo,
} from '@nordic/core-context';
