// @meli/input-validation/dist/utils/stringSecurityState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.INSECURE_CHECK_KINDS = exports.SECURE_CHECK_KINDS = void 0;
const metrics_1 = require("../metrics");
const environment_detector_1 = require("./environment-detector");
/** Zod check kinds whose regex/format restricts input to safe characters only. */
const SECURE_CHECK_KINDS = new Set([
    'uuid',
    'cuid2',
    'ulid',
    'nanoid',
    'datetime',
    'date',
    'time',
]);
exports.SECURE_CHECK_KINDS = SECURE_CHECK_KINDS;
/** Zod check kinds that don't restrict dangerous characters. */
const INSECURE_CHECK_KINDS = new Set([
    'email',
    'url',
    'cuid',
    'min',
    'max',
    'length',
    'includes',
    'startsWith',
    'endsWith',
]);
exports.INSECURE_CHECK_KINDS = INSECURE_CHECK_KINDS;
/** Insecure kinds where the developer has no actionable alternative. */
const SILENT_INSECURE_KINDS = new Set(['email']);
const METRIC = 'node_sdk_input_validation.type.secure';
const WARNING_SUFFIX = 'Add .secure(), .uuid(), .ulid(), .nanoid(), .cuid2(), .datetime(), .date(), or .time() to resolve this warning. ' +
    'Use .unsafe() only as a last resort when none of the above validations fit your use case — it does not add security protection and should not be used as a general bypass.';
function captureCallerFrame() {
    const { stack } = new Error();
    if (!stack) {
        return null;
    }
    const frame = stack
        .split('\n')
        .slice(1)
        .find((f) => f.trim().startsWith('at ') &&
        !f.includes('node_modules') &&
        !f.includes('input-validation/src/') &&
        !f.includes('input-validation/dist/') &&
        !f.trimEnd().endsWith('(<anonymous>)') &&
        !f.includes('node:'));
    return frame ?? null;
}
/**
 * Tracks the security posture of an ivString validation chain.
 *
 * Shared across all IvString instances in the same chain via `_addCheck`.
 * Metrics are deferred via `queueMicrotask` so the FINAL chain state is
 * evaluated before emitting — e.g. `ivString().max(10).secure()` → secure.
 *
 * A bare `ivString()` without any validation defaults to insecure.
 */
class StringSecurityState {
    hasSecure = false;
    hasUnsafe = false;
    hasInsecure = true;
    warnings = [];
    bareFrame;
    emissionScheduled = false;
    constructor() {
        this.bareFrame = (0, environment_detector_1.isNotProduction)() ? captureCallerFrame() : null;
        this.scheduleEmission();
    }
    markSecure() {
        this.hasSecure = true;
        this.scheduleEmission();
    }
    markUnsafe() {
        this.hasUnsafe = true;
        this.scheduleEmission();
    }
    markInsecure(method) {
        this.hasInsecure = true;
        if (!SILENT_INSECURE_KINDS.has(method)) {
            this.warnings.push({
                method,
                frame: (0, environment_detector_1.isNotProduction)() ? captureCallerFrame() : null,
            });
        }
        this.scheduleEmission();
    }
    scheduleEmission() {
        if (this.emissionScheduled) {
            return;
        }
        this.emissionScheduled = true;
        queueMicrotask(() => this.emitMetrics());
    }
    emitMetrics() {
        if (this.hasSecure) {
            (0, metrics_1.increment)(METRIC, 1, ['secure:true']);
            return;
        }
        if (!this.hasInsecure) {
            return;
        }
        if (this.hasUnsafe) {
            (0, metrics_1.increment)(METRIC, 1, ['secure:false', 'unsafe:explicit']);
            return;
        }
        (0, metrics_1.increment)(METRIC, 1, ['secure:false']);
        if (!(0, environment_detector_1.isNotProduction)()) {
            return;
        }
        if (this.warnings.length > 0) {
            this.warnings.forEach(({ method, frame }) => {
                const trace = frame ? `\n${frame}` : '';
                // eslint-disable-next-line no-console
                console.warn(`⚠️ [input-validation] Insecure usage: .${method}() does not restrict dangerous characters and could be vulnerable to injection attacks. ${WARNING_SUFFIX}${trace}`);
            });
        }
        else {
            const trace = this.bareFrame ? `\n${this.bareFrame}` : '';
            // eslint-disable-next-line no-console
            console.warn(`⚠️ [input-validation] Insecure usage: ivString() is being used without any security validation. ${WARNING_SUFFIX}${trace}`);
        }
    }
}
exports.default = StringSecurityState;
