// @andes/progress-indicator-circular/build/cjs/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _extends = require('@babel/runtime/helpers/extends');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var React = require('react');
var PropTypes = require('prop-types');
var classNames = require('classnames');
var visuallyHidden = require('@andes/visually-hidden');
var useSrLabel = require('@andes/common/build/hooks/useSrLabel');
var context = require('@andes/context');
var themes = require('@andes/themes');
var warning = require('@andes/common/build/utils/warning');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var _objectWithoutProperties__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutProperties);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var useSrLabel__default = /*#__PURE__*/_interopDefaultLegacy(useSrLabel);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
const namespace = 'andes-progress-indicator-circular';
const circleNamespace = `${namespace}__circle`;
const DEPRECATED_PROGRESS_CIRCULAR_COLORS = ['blue', 'white'];

/** @typedef {import('../../index').IndeterminateProps} IndeterminateProps */

/** @type {React.FunctionComponent<IndeterminateProps>} */
const Indeterminate = _ref => {
  let labelId = _ref.labelId,
    srLabel = _ref.srLabel;
  return /*#__PURE__*/React__default["default"].createElement("div", {
    "aria-label": srLabel,
    "aria-labelledby": srLabel ? undefined : labelId,
    role: "progressbar"
  }, /*#__PURE__*/React__default["default"].createElement("svg", {
    className: `${namespace}__svg`,
    "aria-hidden": "true",
    viewBox: "0 0 50 50",
    "data-andes-progress-indicator-circular-icon": true
  }, /*#__PURE__*/React__default["default"].createElement("circle", {
    className: circleNamespace,
    cx: "50%",
    cy: "50%",
    r: "20",
    fill: "none"
  })));
};
Indeterminate.propTypes = {
  labelId: PropTypes__default["default"].string,
  srLabel: PropTypes__default["default"].string
};
var Indeterminate$1 = Indeterminate;

/** @typedef {import('../../index').DeterminateProps} DeterminateProps */

/** @type {React.FunctionComponent<DeterminateProps>} */
const Determinate = _ref => {
  let _ref$continuous = _ref.continuous,
    continuous = _ref$continuous === void 0 ? false : _ref$continuous,
    id = _ref.id,
    labelId = _ref.labelId,
    srLabel = _ref.srLabel,
    _ref$value = _ref.value,
    value = _ref$value === void 0 ? 0 : _ref$value;
  const circunference = 124;
  const circlePercentage = circunference * value / 100;
  const circleClassnames = classNames__default["default"](circleNamespace, {
    [`${circleNamespace}--continuous`]: continuous
  });
  return /*#__PURE__*/React__default["default"].createElement("div", {
    role: "progressbar",
    "aria-valuenow": value,
    "aria-valuemin": 0,
    "aria-valuemax": 100,
    "aria-label": srLabel,
    "aria-labelledby": srLabel ? undefined : labelId
  }, /*#__PURE__*/React__default["default"].createElement("svg", {
    className: `${namespace}__svg`,
    "aria-hidden": "true",
    viewBox: "0 0 50 50",
    "data-andes-progress-indicator-circular-icon": true
  }, /*#__PURE__*/React__default["default"].createElement("circle", {
    className: `${circleNamespace}-bg`,
    cx: "25",
    cy: "25",
    r: "20",
    fill: "none"
  }), /*#__PURE__*/React__default["default"].createElement("circle", {
    key: id,
    className: circleClassnames,
    strokeDasharray: circunference,
    strokeDashoffset: `${circunference - Math.min(circlePercentage, circunference)}`,
    cx: "25",
    cy: "25",
    r: "20",
    fill: "none"
  })));
};
Determinate.propTypes = {
  continuous: PropTypes__default["default"].bool,
  id: PropTypes__default["default"].string,
  labelId: PropTypes__default["default"].string,
  srLabel: PropTypes__default["default"].string,
  value: PropTypes__default["default"].number
};
var Determinate$1 = Determinate;
const WARNING_COLORS = `This color will be deprecated in a future major version. 
Please use 'basic' | 'inverse' on <ProgressIndicatorCircular> instead`;

// This function attempts to keep backwards compatibility with old colors

/** @typedef {import('../..').ProgressIndicatorCircularProps['color']} ProgressCircularColors */

/**
 * @typedef {Object} Props
 * @prop {ProgressCircularColors} [color]
 */

/** @type {(props:Props) => ProgressCircularColors} */
const computeColorProps = _ref => {
  let color = _ref.color;
  // @deprecate [color] options: ['blue', 'white'] - Warning from: 9.0.0
  if (typeof color === 'string') {
    warning__default["default"](!DEPRECATED_PROGRESS_CIRCULAR_COLORS.includes(color), WARNING_COLORS);
  }
  switch (color) {
    case 'blue':
      return 'basic';
    case 'white':
      return 'inverse';
    default:
      return color || 'basic';
  }
};
var computeColorProps$1 = computeColorProps;
const _excluded = ["ref"];

/**
 * @typedef {import('../..').DeterminateSpecificProps &
 * import('../../').ProgressIndicatorCircularProps
 * } ProgressIndicatorCircularPublicProps
 */

/**
 * @typedef {Object} ProgressIndicatorCircularInternalProps
 * @prop {string} [ariaLabelledBy]
 */

/**
 * @typedef {ProgressIndicatorCircularPublicProps &
 * ProgressIndicatorCircularInternalProps} ProgressIndicatorCircularProps
 */

/** @type {React.FunctionComponent<ProgressIndicatorCircularProps>} */
const ProgressIndicatorCircular = _ref => {
  let ref = _ref.ref,
    props = _objectWithoutProperties__default["default"](_ref, _excluded);
  const id = props.id,
    _props$size = props.size,
    size = _props$size === void 0 ? 'medium' : _props$size,
    _props$color = props.color,
    color = _props$color === void 0 ? 'basic' : _props$color,
    _props$type = props.type,
    type = _props$type === void 0 ? 'indeterminate' : _props$type,
    className = props.className,
    label = props.label,
    value = props.value,
    children = props.children,
    _props$continuous = props.continuous,
    continuous = _props$continuous === void 0 ? false : _props$continuous,
    srLabel = props.srLabel,
    srAnnouncement = props.srAnnouncement,
    ariaLabelledBy = props.ariaLabelledBy,
    modifier = props.modifier;
  const componentId = context.useId(id);
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  const labelId = ariaLabelledBy || label && `${id}__label`;
  const srLabelVisibility = useSrLabel__default["default"]();
  const getSubcomponent = () => {
    const commonProps = {
      srLabel,
      labelId
    };
    if (type === 'determinate') {
      const determinateProps = _objectSpread__default["default"]({
        value,
        continuous,
        id: componentId
      }, commonProps);
      return /*#__PURE__*/React__default["default"].createElement(Determinate$1, determinateProps);
    }
    return /*#__PURE__*/React__default["default"].createElement(Indeterminate$1, commonProps);
  };
  const computedColor = computeColorProps$1({
    color
  });
  const progressIndicatorClassnames = classNames__default["default"](namespace, className, {
    [`${namespace}--${type}`]: type,
    [`${namespace}--${size}`]: size,
    [`${namespace}--${computedColor}`]: computedColor,
    [`${namespace}--${modifier}`]: modifier
  }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('progressIndicator'));
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    ref: ref,
    className: progressIndicatorClassnames,
    "data-andes-progress-indicator-circular": true,
    "data-andes-progress-indicator-circular-size": size,
    "data-andes-progress-indicator-circular-type": type
  }, themeHash), /*#__PURE__*/React__default["default"].createElement("div", {
    className: `${namespace}__container`
  }, /*#__PURE__*/React__default["default"].createElement("div", {
    className: `${namespace}__progress`
  }, srAnnouncement && /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, {
    role: "status",
    "aria-live": "polite"
  }, srLabelVisibility && srAnnouncement), getSubcomponent(), children && /*#__PURE__*/React__default["default"].createElement("div", {
    className: `${namespace}__content`
  }, children)), label && /*#__PURE__*/React__default["default"].createElement("span", {
    className: `${namespace}__label`,
    id: labelId
  }, label)));
};
ProgressIndicatorCircular.displayName = 'ProgressIndicatorCircular';
ProgressIndicatorCircular.propTypes = {
  id: PropTypes__default["default"].string,
  className: PropTypes__default["default"].string,
  label: PropTypes__default["default"].string,
  type: PropTypes__default["default"].oneOf(['indeterminate', 'determinate']),
  size: PropTypes__default["default"].oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge']),
  color: PropTypes__default["default"].oneOf(['blue', 'white', 'basic', 'inverse']),
  value: PropTypes__default["default"].number,
  continuous: PropTypes__default["default"].bool,
  children: PropTypes__default["default"].node,
  ariaLabelledBy: PropTypes__default["default"].string,
  srLabel: PropTypes__default["default"].string,
  srAnnouncement: PropTypes__default["default"].string,
  modifier: PropTypes__default["default"].oneOf(['block', 'fullscreen', 'inline'])
};
var ProgressIndicatorCircular$1 = ProgressIndicatorCircular;
exports.ProgressIndicatorCircular = ProgressIndicatorCircular$1;
