// @andes/badge/build/cjs/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var React = require('react');
var PropTypes = require('prop-types');
var classNames = require('classnames');
var visuallyHidden = require('@andes/visually-hidden');
var context = require('@andes/context');
var includes = require('@andes/common/build/utils/includes');
var warning = require('@andes/common/build/utils/warning');
var themes = require('@andes/themes');
var _extends = require('@babel/runtime/helpers/extends');
var FeedbackCaution24 = require('@andes/icons/FeedbackCaution24');
var FeedbackInfo24 = require('@andes/icons/FeedbackInfo24');
var FeedbackNegative24 = require('@andes/icons/FeedbackNegative24');
var FeedbackPositive24 = require('@andes/icons/FeedbackPositive24');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var _objectWithoutProperties__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutProperties);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var includes__default = /*#__PURE__*/_interopDefaultLegacy(includes);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var FeedbackCaution24__default = /*#__PURE__*/_interopDefaultLegacy(FeedbackCaution24);
var FeedbackInfo24__default = /*#__PURE__*/_interopDefaultLegacy(FeedbackInfo24);
var FeedbackNegative24__default = /*#__PURE__*/_interopDefaultLegacy(FeedbackNegative24);
var FeedbackPositive24__default = /*#__PURE__*/_interopDefaultLegacy(FeedbackPositive24);
const namespace = 'andes-badge';
const PILL_TYPE_ICON = 'icon';

/** @type {NonNullable<import('../../index').PillSpecificProps["roundedCorners"]>} */
const ROUNDED_CORNERS_VALUES = ['top-left', 'top-right', 'bottom-left', 'bottom-right'];
const CONTENT_TYPE = /** @type {const} */['icon', 'text'];
const DEPRECATED_BADGE_COLORS = ['accent', 'green', 'orange', 'red', 'gray'];
const WARNING_COLORS = `This color will be deprecated in a future major version,
 please use 'caution' | 'informative' | 'inverted' | 'negative' | 'neutral' | 'positive'`;
const WARNING_CONTENT_TYPE = `'contentType' will be deprecated in a future major version, 
 please use <PillIcon> instead of 'contentType="icon"' or remove it if you're using 'contentType="text"'.`;

// This function attempts to keep backwards compatibility with old colors

/** @typedef {import('../..').DeprecatedBadgeColors} DeprecatedBadgeColors */
/** @typedef {import('../..').BadgeColors} BadgeColors */

/**
 * @typedef {Object} Props
 * @prop {DeprecatedBadgeColors | BadgeColors} [color]
 */

/** @type {(props:Props) => DeprecatedBadgeColors | 'inverted'} */
const computeColorProps = _ref => {
  let color = _ref.color;
  // @deprecate [color] options: ['accent', 'gray', 'green', 'orange', 'red'] - Warning from: 9.0.0
  switch (color) {
    case 'informative':
      return 'accent';
    case 'neutral':
      return 'gray';
    case 'caution':
      return 'orange';
    case 'negative':
      return 'red';
    case 'positive':
      return 'green';
    default:
      return color || 'accent';
  }
};
var computeColorProps$1 = computeColorProps;
const _excluded$2 = ["className", "color", "id", "srLabel", "ref", "backgroundType", "customVariant"];

/**
 * Badges are used to indicate an status or to differentiate an element, block or content from others.
 */

/** @type {typeof import('../index').Dot} */
const Dot = props => {
  const className = props.className,
    color = props.color,
    id = props.id,
    srLabel = props.srLabel,
    ref = props.ref,
    backgroundType = props.backgroundType;
  props.customVariant;
  const otherProps = _objectWithoutProperties__default["default"](props, _excluded$2);
  const componentId = context.useId(id);
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;

  // @deprecate [color] options: ['accent', 'gray', 'green', 'orange', 'red'] - Warning from: 9.0.0
  if (typeof color === 'string') {
    warning__default["default"](!includes__default["default"](DEPRECATED_BADGE_COLORS, color), WARNING_COLORS);
  }
  const computedColor = computeColorProps$1({
    color
  });
  const classes = classNames__default["default"](namespace, className, `${namespace}--dot`, `${namespace}--${computedColor}`, {
    [`${namespace}--color-on-background`]: backgroundType === 'complexBackground'
  }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('badge', props));
  const sharedProps = _objectSpread__default["default"](_objectSpread__default["default"]({
    className: classes,
    id: componentId,
    ref
  }, otherProps), {}, {
    'data-andes-badge': true,
    'data-andes-badge-type': 'dot'
  }, themeHash);
  return /*#__PURE__*/React__default["default"].createElement("div", sharedProps, srLabel && /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, null, srLabel));
};
Dot.displayName = 'Dot';
Dot.propTypes = {
  className: PropTypes__default["default"].string,
  color: PropTypes__default["default"].oneOf(['caution', 'informative', 'inverted', 'negative', 'neutral', 'positive']),
  customVariant: PropTypes__default["default"].string,
  id: PropTypes__default["default"].string,
  srLabel: PropTypes__default["default"].string,
  backgroundType: PropTypes__default["default"].oneOf(['basic', 'complexBackground'])
};
var Dot$1 = Dot;
const _excluded$1 = ["className", "color", "id", "size", "srLabel", "ref", "backgroundType", "customVariant"];

/**
 * @param {'accent' | 'green' | 'orange' | 'red'} color
 */
const getSign = color => {
  const commonProps = {
    className: `${namespace}__icon-inner`,
    color: 'currentColor'
  };
  const colors = {
    accent: /*#__PURE__*/React__default["default"].createElement(FeedbackInfo24__default["default"], _extends__default["default"]({}, commonProps, {
      key: "badge-info"
    })),
    green: /*#__PURE__*/React__default["default"].createElement(FeedbackPositive24__default["default"], _extends__default["default"]({}, commonProps, {
      key: "badge-success"
    })),
    orange: /*#__PURE__*/React__default["default"].createElement(FeedbackCaution24__default["default"], _extends__default["default"]({}, commonProps, {
      key: "badge-warning"
    })),
    red: /*#__PURE__*/React__default["default"].createElement(FeedbackNegative24__default["default"], _extends__default["default"]({}, commonProps, {
      key: "badge-error"
    }))
  };
  return colors[color];
};

/** @type {typeof import('..').PillIcon} */
const PillIcon = props => {
  const className = props.className,
    color = props.color,
    id = props.id,
    _props$size = props.size,
    size = _props$size === void 0 ? 'large' : _props$size,
    srLabel = props.srLabel,
    ref = props.ref,
    backgroundType = props.backgroundType,
    customVariant = props.customVariant,
    otherProps = _objectWithoutProperties__default["default"](props, _excluded$1);
  const allProps = _objectSpread__default["default"]({
    customVariant,
    size
  }, props);
  const componentId = context.useId(id);
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;

  // @deprecate [color] options: ['accent', 'gray', 'green', 'orange', 'red'] - Warning from: 9.0.0
  if (typeof color === 'string') {
    warning__default["default"](!includes__default["default"](DEPRECATED_BADGE_COLORS, color), WARNING_COLORS);
  }
  const computedColor = computeColorProps$1({
    color
  });
  const andesDataAttributes = _objectSpread__default["default"]({
    'data-andes-badge': true,
    'data-andes-badge-type': 'icon',
    'data-andes-badge-size': size
  }, themeHash);
  const classes = classNames__default["default"](namespace, `${namespace}--pill`, `${namespace}--${computedColor}`, className, `${namespace}--pill-icon`, {
    [`${namespace}--${size}`]: size,
    [`${namespace}--color-on-background`]: backgroundType === 'complexBackground' // only allowed for loud hierarchy
  }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('badge', allProps));
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    className: classes,
    id: componentId,
    ref: ref
  }, otherProps, andesDataAttributes), srLabel && /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, null, srLabel), /*#__PURE__*/React__default["default"].createElement("div", {
    "aria-hidden": "true",
    className: `${namespace}__icon`
  }, getSign(
  // @ts-expect-error: PillIcon does not accept gray nor inverted
  computedColor)));
};
PillIcon.propTypes = {
  className: PropTypes__default["default"].string,
  color: PropTypes__default["default"].oneOf(['caution', 'informative', 'inverted', 'negative', 'neutral', 'positive']),
  customVariant: PropTypes__default["default"].string,
  id: PropTypes__default["default"].string,
  size: PropTypes__default["default"].oneOf(['xsmall', 'small', 'large']),
  srLabel: PropTypes__default["default"].string,
  backgroundType: PropTypes__default["default"].oneOf(['basic', 'complexBackground'])
};
PillIcon.displayName = 'PillIcon';
var PillIcon$1 = PillIcon;
const _excluded = ["children", "className", "color", "hierarchy", "contentType", "id", "roundedCorners", "secondaryColor", "size", "srLabel", "ref", "backgroundType", "customVariant"];

/** @typedef {import('..').PillSpecificProps} PillPublicProps */

/**
 * @typedef {Object} PillInternalProps
 * @prop {PillPublicProps['size'] | 'xsmall'} [size]
*/

/** @typedef {PillPublicProps & PillInternalProps} PillProps */

/**
 * Badges are used to indicate an status or to differentiate an element, block or content from others.
 */
/** @type {React.FunctionComponent<PillProps>} */
const Pill = props => {
  const children = props.children,
    className = props.className,
    color = props.color,
    _props$hierarchy = props.hierarchy,
    hierarchy = _props$hierarchy === void 0 ? 'loud' : _props$hierarchy,
    _willdeprecate$contentType = props.contentType,
    id = props.id,
    _props$roundedCorners = props.roundedCorners,
    roundedCorners = _props$roundedCorners === void 0 ? ROUNDED_CORNERS_VALUES : _props$roundedCorners,
    secondaryColor = props.secondaryColor,
    _props$size = props.size,
    size = _props$size === void 0 ? 'large' : _props$size,
    srLabel = props.srLabel,
    ref = props.ref,
    backgroundType = props.backgroundType,
    customVariant = props.customVariant,
    otherProps = _objectWithoutProperties__default["default"](props, _excluded);

  // @deprecate [contentType] - Warning from: 9.0.0
  warning__default["default"](typeof _willdeprecate$contentType === 'undefined', WARNING_CONTENT_TYPE);

  // @deprecate [color] options: ['accent', 'gray', 'green', 'orange', 'red'] - Warning from: 9.0.0
  if (typeof color === 'string') {
    warning__default["default"](!includes__default["default"](DEPRECATED_BADGE_COLORS, color), WARNING_COLORS);
  }
  const isIcon = _willdeprecate$contentType === PILL_TYPE_ICON;
  if (isIcon) {
    const iconProps = _objectSpread__default["default"]({
      className,
      color,
      id,
      size,
      srLabel,
      ref,
      backgroundType,
      customVariant
    }, otherProps);
    return /*#__PURE__*/React__default["default"].createElement(PillIcon$1, iconProps);
  }
  const allProps = _objectSpread__default["default"]({
    customVariant,
    contentType: _willdeprecate$contentType,
    hierarchy,
    roundedCorners,
    size
  }, props);
  const componentId = context.useId(id);
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  const computedColor = computeColorProps$1({
    color
  });
  const classes = classNames__default["default"](namespace, `${namespace}--pill`, `${namespace}--${computedColor}`, className, {
    [`${namespace}--${size}`]: size,
    [`${namespace}--${computedColor}--quiet`]: !secondaryColor && hierarchy === 'quiet',
    [`${namespace}--rounded-top-left`]: includes__default["default"](roundedCorners, 'top-left'),
    [`${namespace}--rounded-top-right`]: includes__default["default"](roundedCorners, 'top-right'),
    [`${namespace}--rounded-bottom-left`]: includes__default["default"](roundedCorners, 'bottom-left'),
    [`${namespace}--rounded-bottom-right`]: includes__default["default"](roundedCorners, 'bottom-right'),
    [`${namespace}--color-on-background`]: hierarchy === 'loud' && backgroundType === 'complexBackground'
  }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('badge', allProps));
  const andesDataAttributes = _objectSpread__default["default"]({
    'data-andes-badge': true,
    'data-andes-badge-type': 'pill',
    'data-andes-badge-hierarchy': hierarchy,
    'data-andes-badge-size': size
  }, themeHash);
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    className: classes,
    id: componentId,
    ref: ref
  }, otherProps, {
    style: {
      backgroundColor: secondaryColor === null || secondaryColor === void 0 ? void 0 : secondaryColor.background
    }
  }, andesDataAttributes), srLabel && /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, null, srLabel), children && /*#__PURE__*/React__default["default"].createElement("p", {
    "aria-hidden": !!srLabel || undefined,
    className: `${namespace}__content`,
    style: {
      color: secondaryColor === null || secondaryColor === void 0 ? void 0 : secondaryColor.text
    }
  }, children));
};
Pill.propTypes = {
  children: PropTypes__default["default"].node,
  className: PropTypes__default["default"].string,
  color: PropTypes__default["default"].oneOf(['caution', 'informative', 'inverted', 'negative', 'neutral', 'positive']),
  contentType: PropTypes__default["default"].oneOf(CONTENT_TYPE),
  customVariant: PropTypes__default["default"].string,
  hierarchy: PropTypes__default["default"].oneOf(['loud', 'quiet']),
  id: PropTypes__default["default"].string,
  roundedCorners: PropTypes__default["default"].arrayOf(PropTypes__default["default"].oneOf(ROUNDED_CORNERS_VALUES).isRequired),
  secondaryColor: PropTypes__default["default"].shape({
    background: PropTypes__default["default"].string.isRequired,
    text: PropTypes__default["default"].string
  }),
  size: PropTypes__default["default"].oneOf(['xsmall', 'small', 'large']),
  srLabel: PropTypes__default["default"].string,
  backgroundType: PropTypes__default["default"].oneOf(['basic', 'complexBackground'])
};
Pill.displayName = 'Pill';
var Pill$1 = Pill;
exports.Dot = Dot$1;
exports.Pill = Pill$1;
exports.PillIcon = PillIcon$1;
