// @andes/dropdown/build/cjs/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _extends = require('@babel/runtime/helpers/extends');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var React = require('react');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var PropTypes = require('prop-types');
var classNames = require('classnames');
var context = require('@andes/context');
var floatingMenu = require('@andes/floating-menu');
var form = require('@andes/form');
var ChevronDown20 = require('@andes/icons/ChevronDown20');
var visuallyHidden = require('@andes/visually-hidden');
var tag = require('@andes/tag');
var warning = require('@andes/common/build/utils/warning');
var themes = require('@andes/themes');
var computeDataAndesState = require('@andes/common/build/utils/computeDataAndesState');
var ChevronDown12 = require('@andes/icons/ChevronDown12');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _objectWithoutProperties__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutProperties);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var ChevronDown20__default = /*#__PURE__*/_interopDefaultLegacy(ChevronDown20);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var ChevronDown12__default = /*#__PURE__*/_interopDefaultLegacy(ChevronDown12);
const MODIFIER_ERROR = 'error';
const MODIFIER_READONLY = 'readonly';
const SIZE_LARGE = 'large';
const MENU_ALIGNMENT_CENTER = 'center';
const MENU_BOTTOM_OFFSET_Y = 6;
const MENU_STANDALONE_OFFSET_X = -9;
const MENU_BOUNDARIES = 8;

// Multiple variant
const DROPDOWN_ARROW_SPACING = 100;
const MAX_TAG_LIMIT = 1000;
const TAG_SMALL_LETTER_WIDTH = 8;
const TAG_SMALL_OFFSET = 30;

/**
 * @param {string | string[]} currentDisplayValues
 * @param {import('react').MutableRefObject<HTMLDivElement | null>} containerRef
 */
const calculateTagLimit = (currentDisplayValues, containerRef) => {
  var _containerRef$current;
  const containerWidth = /** @type {HTMLDivElement} */containerRef === null || containerRef === void 0 || (_containerRef$current = containerRef.current) === null || _containerRef$current === void 0 ? void 0 : _containerRef$current.offsetWidth;
  let totalLength = 0;
  let calculatedTagLimit = MAX_TAG_LIMIT;

  // eslint-disable-next-line no-restricted-syntax
  for (const _ref of Object.entries(currentDisplayValues)) {
    var _ref2 = _slicedToArray__default["default"](_ref, 2);
    const labelIndex = _ref2[0];
    const currentLabel = _ref2[1];
    const currentLabelLength = (currentLabel === null || currentLabel === void 0 ? void 0 : currentLabel.toString().length) * TAG_SMALL_LETTER_WIDTH + TAG_SMALL_OFFSET;
    if (totalLength + currentLabelLength + DROPDOWN_ARROW_SPACING > containerWidth) {
      calculatedTagLimit = parseInt(labelIndex, 10);
      break;
    }
    totalLength += currentLabelLength;
  }
  return calculatedTagLimit;
};

/**
 * @typedef {Object} TagContentProps
 * @prop {string | string[]} [currentDisplayValues]
 * @prop {string[]} [currentValues]
 * @prop {boolean} [disabled]
 * @prop {string} [namespace]
 * @prop {React.RefObject<HTMLDivElement | null>} containerRef
 * @prop {string} id
 * @prop {(
 *  event: React.KeyboardEvent | React.MouseEvent,
 *  value?: string | string[]
 * ) => void} onSelectionChange
 * @prop {import('../..').DropdownFormProps['renderMultipleTriggerContent']} [renderMultipleTriggerContent]
 */

/** @type {React.FunctionComponent<TagContentProps>} */
const TagContent = _ref => {
  let _ref$currentDisplayVa = _ref.currentDisplayValues,
    currentDisplayValues = _ref$currentDisplayVa === void 0 ? [] : _ref$currentDisplayVa,
    _ref$currentValues = _ref.currentValues,
    currentValues = _ref$currentValues === void 0 ? [] : _ref$currentValues,
    _ref$disabled = _ref.disabled,
    disabled = _ref$disabled === void 0 ? false : _ref$disabled,
    _ref$namespace = _ref.namespace,
    namespace = _ref$namespace === void 0 ? '' : _ref$namespace,
    containerRef = _ref.containerRef,
    id = _ref.id,
    onSelectionChange = _ref.onSelectionChange,
    renderMultipleTriggerContent = _ref.renderMultipleTriggerContent;
  const currentDisplayValuesHasItems = Array.isArray(currentDisplayValues) && currentDisplayValues.length > 0;
  const handleClose = React.useCallback(/** @param {number} index */index => /** @param {React.MouseEvent<HTMLButtonElement>} event */event => {
    if (disabled) return;
    const newValues = currentValues.filter((_, i) => i !== index);
    onSelectionChange(event, newValues);
  }, [currentValues, disabled, onSelectionChange]);

  /** @param {React.MouseEvent<HTMLDivElement>} event */
  const handleClick = event => {
    event.stopPropagation();
  };
  const getTagCollapsedProps = React.useCallback(/** @param {string | string[]} displayValues */displayValues => ({
    id: `${id}-display-values`,
    disabled,
    limit: calculateTagLimit(displayValues, containerRef)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }), [disabled, id]);
  const getTagProps = React.useCallback(
  /**
   * @param {string} displayValue
   * @param {number} index
   */
  (displayValue, index) => ({
    id: `${id}-tag-${index}`,
    size: (/** @type {'small' | 'large'} */'small'),
    disabled,
    label: displayValue,
    onClose: handleClose(index)
  }), [disabled, id, handleClose]);
  if (typeof renderMultipleTriggerContent === 'function') {
    return /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${namespace}__display-container`,
      id: `${id}-display-values`
    }, renderMultipleTriggerContent(currentValues, {
      getTagCollapsedProps,
      getTagProps
    }));
  }
  if (!currentDisplayValuesHasItems) return null;
  return (/*#__PURE__*/
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events
    React__default["default"].createElement("div", {
      className: `${namespace}__display-container`,
      id: `${id}-display-values`,
      onClick: handleClick
    }, /*#__PURE__*/React__default["default"].createElement(tag.TagCollapsed, getTagCollapsedProps(currentDisplayValues), currentDisplayValues.map((displayValue, index) => /*#__PURE__*/React__default["default"].createElement(tag.Tag, _extends__default["default"]({
      key: `${displayValue}_${index}`
    }, getTagProps(displayValue, index))))))
  );
};
TagContent.propTypes = {
  currentDisplayValues: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string.isRequired, PropTypes__default["default"].arrayOf(PropTypes__default["default"].string.isRequired)]),
  currentValues: PropTypes__default["default"].arrayOf(PropTypes__default["default"].string.isRequired),
  disabled: PropTypes__default["default"].bool,
  namespace: PropTypes__default["default"].string,
  // current could be HTMLDivElement or null, we can't use HTMLDivElement on prop types because it's undefined on SSR
  containerRef: PropTypes__default["default"].shape({
    current: PropTypes__default["default"].oneOfType([PropTypes__default["default"].node, PropTypes__default["default"].object])
  }).isRequired,
  id: PropTypes__default["default"].string.isRequired,
  onSelectionChange: PropTypes__default["default"].func.isRequired,
  renderMultipleTriggerContent: PropTypes__default["default"].func
};
var TagContent$1 = TagContent;
const namespace = 'andes-dropdown';
const triggerNamespace = `${namespace}__trigger`;
const formNamespace = 'andes-form-control';
const floatingMenuNamespace = 'andes-floating-menu';
const listNamespace = 'andes-list';
const listItemClassName = `${listNamespace}__item`;

/**
 * @typedef {Object} MenuOffsetProp
 * @prop {boolean} centeredMenu
 * @prop {import('react').RefObject<HTMLDivElement | null>} mainContainerRef
 */

/** @param {HTMLUListElement} menu */
const getItemToFocus = menu => {
  const optionsList = Array.prototype.slice.call(menu.querySelectorAll(`.${listItemClassName}`));
  if (optionsList.length) {
    // Search last navigated option
    let focusableOption = optionsList.find(option => option.tabIndex === 0);

    // Search selected option
    if (!focusableOption) {
      focusableOption = optionsList.find(option => option.className.indexOf(`.${listItemClassName}--selected`) !== -1);
    }

    // default to first option
    if (!focusableOption) {
      var _optionsList = _slicedToArray__default["default"](optionsList, 1);
      focusableOption = _optionsList[0];
    }
    return focusableOption;
  }
  return null;
};
const TRIGGER_MARGIN = 2;

/**
 * Fuction to calculate the position of the menu when centered over the trigger
 * @param {MenuOffsetProp} prop
 */
const getMenuOffset = _ref => {
  let centeredMenu = _ref.centeredMenu,
    mainContainerRef = _ref.mainContainerRef;
  if (!centeredMenu) {
    return {
      offset: undefined,
      calculated: true
    };
  }
  const main = mainContainerRef.current;
  const floatingMenu = main === null || main === void 0 ? void 0 : main.querySelector('ul');
  const scrollableContainer = main === null || main === void 0 ? void 0 : main.querySelector('.andes-popper');
  const trigger = main === null || main === void 0 ? void 0 : main.querySelector(`.${triggerNamespace}`);
  if (floatingMenu && trigger) {
    const itemToCenter = getItemToFocus(floatingMenu);
    if (itemToCenter) {
      const itemHeight = itemToCenter.offsetHeight;

      /* Calculates the position of the element related to the scrollable parent container */
      const distance = itemToCenter.offsetTop - /** @type {Element} */scrollableContainer.scrollTop;

      /* Positions the menu to center the selected element on top of the trigger */
      const offset = -itemHeight - distance - TRIGGER_MARGIN;
      return {
        offset,
        calculated: true
      };
    }
  }
  return {
    offset: undefined,
    calculated: false
  };
};
var getMenuOffset$1 = getMenuOffset;
const WARNING_LABEL = `<Dropdown> needs some plain text contents for accessibility reasons.
Please add a \`label\` or \`srLabel\` prop.`;
const WARNING_DUPLICATED_VALUE = `value {0} already exists.
Please, use a different value for each <DropdownItem>`;
const DROPDOWN_ITEM_DISPLAY_NAME = 'DropdownItem';
const DROPDOWN_GROUP_DISPLAY_NAME = 'DropdownGroup';
const DROPDOWN_NATIVE_DISPLAY_NAME = 'DropdownNative';
const DROPDOWN_NATIVE_ITEM_DISPLAY_NAME = 'DropdownNativeItem';

/**
 * @typedef {Object} ItemsPrimaryProps
 * @prop {import('react').ReactElement | import('react').ReactElement[]} children
 */

/** @param {ItemsPrimaryProps} props */
const getItemsPrimary = _ref => {
  var _firstOption$type;
  let children = _ref.children;
  const firstOption = Array.isArray(children) ? children[0] : children;
  // @ts-expect-error TS doesn't recognize displayName as a valid property
  const firstOptionType = firstOption === null || firstOption === void 0 || (_firstOption$type = firstOption.type) === null || _firstOption$type === void 0 ? void 0 : _firstOption$type.displayName;
  const reference = /** @type {Record<string, { isSlotted: Boolean; content: string }>} */{};

  /** @param {import('react').ReactElement} item */
  const setPrimaryReference = item => {
    const _item$props = item.props,
      itemChildren = _item$props.children,
      title = _item$props.title,
      itemValue = _item$props.value;
    warning__default["default"](!reference[itemValue], WARNING_DUPLICATED_VALUE.replace('{0}', itemValue));
    const defaultContent = !!title;
    const isSlotted = itemChildren && !defaultContent;
    reference[itemValue] = {
      isSlotted,
      content: title
    };
  };
  if (firstOptionType === DROPDOWN_ITEM_DISPLAY_NAME) {
    React.Children.forEach(children, setPrimaryReference);
  }
  if (firstOptionType === DROPDOWN_GROUP_DISPLAY_NAME) {
    React.Children.forEach(children, group => {
      React.Children.forEach(group.props.children, setPrimaryReference);
    });
  }
  return reference;
};
var getItemsPrimary$1 = getItemsPrimary;

/**
 * @typedef {Object} TriggerDisplayValuesProps
 * @prop {string | string[]} [values]
 * @prop {boolean} [multiple]
 * @prop {Record<string, { isSlotted: Boolean; content: string }>} itemsPrimaryReference
 */

/**
 * @param {TriggerDisplayValuesProps} props
 * @returns {string | string[] | undefined}
 */
const getTriggerDisplayValues = _ref => {
  let values = _ref.values,
    multiple = _ref.multiple,
    itemsPrimaryReference = _ref.itemsPrimaryReference;
  let displayValues;
  if (typeof values === 'undefined') return displayValues;
  if (Array.isArray(values)) {
    displayValues = values.map(selectedValue => {
      const _ref2 = itemsPrimaryReference[selectedValue] || {},
        isSlotted = _ref2.isSlotted,
        content = _ref2.content;
      return isSlotted ? selectedValue : content;
    });
  } else {
    const _ref3 = itemsPrimaryReference[values] || {},
      isSlotted = _ref3.isSlotted,
      content = _ref3.content;
    displayValues = isSlotted ? values : content;
  }
  if (displayValues && multiple && !Array.isArray(displayValues)) {
    displayValues = [displayValues];
  }
  return displayValues;
};
var getTriggerDisplayValues$1 = getTriggerDisplayValues;
const _excluded$5 = ["children", "defaultValue", "disabled", "helper", "label", "menuAlignment", "modifier", "multiple", "onChange", "onClose", "onOpen", "open", "placeholder", "searchConfig", "srLabel", "value"];

/**
 * @typedef {Object} SearchConfigProps
 * @prop {import('../utils/computeSearchProps').ComputeSearchProps} [searchConfig]
 */

/**
 * @param {import('react').RefObject<HTMLDivElement | null>} mainContainerRef
 * @param {import('../..').DropdownProps & SearchConfigProps} props
 */
const useDropdown = (mainContainerRef, _ref) => {
  let children = _ref.children,
    defaultValue = _ref.defaultValue;
  _ref.disabled;
  _ref.helper;
  _ref.label;
  let menuAlignment = _ref.menuAlignment;
  _ref.modifier;
  let multiple = _ref.multiple,
    onChange = _ref.onChange,
    onClose = _ref.onClose,
    onOpen = _ref.onOpen,
    open = _ref.open;
  _ref.placeholder;
  let searchConfig = _ref.searchConfig;
  _ref.srLabel;
  let value = _ref.value,
    other = _objectWithoutProperties__default["default"](_ref, _excluded$5);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const isUncontrolled = React.useMemo(() => value === undefined, []);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const isOpenUncontrolled = React.useMemo(() => open === undefined, []);
  const centeredMenu = React.useMemo(() => {
    if (multiple || searchConfig) return false;
    return menuAlignment === MENU_ALIGNMENT_CENTER;
  }, [menuAlignment, multiple, searchConfig]);
  const itemsPrimaryReference = React.useMemo(() => {
    if (!children) return {};
    return getItemsPrimary$1({
      children
    });
  }, [children]);
  const _useAndesContext = context.useAndesContext(),
    device = _useAndesContext.device;
  const isMobile = device === null || device === void 0 ? void 0 : device.mobile;
  const _useState = React.useState(value || defaultValue),
    _useState2 = _slicedToArray__default["default"](_useState, 2),
    currentValues = _useState2[0],
    setCurrentValues = _useState2[1];
  const _useState3 = React.useState(getTriggerDisplayValues$1({
      values: currentValues,
      multiple,
      itemsPrimaryReference
    })),
    _useState4 = _slicedToArray__default["default"](_useState3, 2),
    currentDisplayValues = _useState4[0],
    setCurrentDisplayValues = _useState4[1];
  const _useState5 = React.useState(false),
    _useState6 = _slicedToArray__default["default"](_useState5, 2),
    isMenuVisible = _useState6[0],
    setIsMenuVisible = _useState6[1];
  const offsetInitialValue = /** @type {number | undefined} */centeredMenu ? 0 : MENU_BOTTOM_OFFSET_Y;
  const _useState7 = React.useState(offsetInitialValue),
    _useState8 = _slicedToArray__default["default"](_useState7, 2),
    menuOffset = _useState8[0],
    setMenuOffset = _useState8[1];
  const _useState9 = React.useState(false),
    _useState0 = _slicedToArray__default["default"](_useState9, 2),
    menuOffsetCalculated = _useState0[0],
    setMenuOffsetCalculated = _useState0[1];
  const _useState1 = React.useState(/** @type {string | undefined} */undefined),
    _useState10 = _slicedToArray__default["default"](_useState1, 2),
    placement = _useState10[0],
    setPlacement = _useState10[1];

  /** @type {import('@andes/floating-menu').FloatingMenuProps['onChange']} */
  const handleOnChange = (event, values) => {
    if (isUncontrolled) {
      setCurrentValues(values);
      setCurrentDisplayValues(getTriggerDisplayValues$1({
        values,
        multiple,
        itemsPrimaryReference
      }));
    }
    if (typeof onChange === 'function') {
      // Don't modify event.target.value if the event comes from an input element
      // (e.g., searchbox) to prevent unwanted value changes in the input
      const isInputEvent = event.target instanceof HTMLInputElement;
      if (!isInputEvent) {
        // @ts-expect-error
        // value is not a valid prop for HTMLDivElement, we are adding them to the event
        event.target.value = values;
      }
      onChange(event, values);
    }
  };

  /** @type {import('@andes/floating-menu').FloatingMenuProps['onOpen']} */
  const handleOpen = event => {
    if (isOpenUncontrolled) {
      setIsMenuVisible(true);
      setTimeout(() => {
        const _getMenuOffset = getMenuOffset$1({
            centeredMenu,
            mainContainerRef
          }),
          offset = _getMenuOffset.offset,
          calculated = _getMenuOffset.calculated;
        setMenuOffset(offset);
        setMenuOffsetCalculated(calculated);
      }, 0);
    }
    if (typeof onOpen === 'function') onOpen(event);
  };

  /** @type {import('@andes/floating-menu').FloatingMenuProps['onClose']} */
  const handleClose = () => {
    if (isOpenUncontrolled) {
      setIsMenuVisible(false);
      if (centeredMenu) {
        setMenuOffset(0);
        setMenuOffsetCalculated(false);
      }
    }
    if (typeof onClose === 'function') onClose();
  };
  React.useEffect(() => {
    if (!isUncontrolled) {
      setCurrentValues(value);
      setCurrentDisplayValues(getTriggerDisplayValues$1({
        values: value,
        multiple,
        itemsPrimaryReference
      }));
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value]);
  React.useEffect(() => {
    if (!isOpenUncontrolled) {
      if (open) {
        setIsMenuVisible(true);
        setTimeout(() => {
          const _getMenuOffset2 = getMenuOffset$1({
              centeredMenu,
              mainContainerRef
            }),
            offset = _getMenuOffset2.offset,
            calculated = _getMenuOffset2.calculated;
          setMenuOffset(offset);
          setMenuOffsetCalculated(calculated);
        }, 0);
      } else {
        setIsMenuVisible(false);
        if (centeredMenu) {
          setMenuOffset(0);
          setMenuOffsetCalculated(false);
        }
      }
    }
  }, [open, isOpenUncontrolled, centeredMenu, mainContainerRef]);
  const getValuesAsArray = () => {
    if (Array.isArray(currentValues)) return currentValues;
    if (typeof currentValues === 'string' && !!currentValues || typeof currentValues === 'number') return [currentValues];
    return [];
  };
  const getPopperOptions = () => {
    if (menuAlignment === 'center' || isMobile && searchConfig) {
      const fallbackPlacements = isMobile && searchConfig ? ['top', 'bottom'] : [];
      return {
        modifiers: [{
          name: 'flip',
          options: {
            fallbackPlacements
          }
        }]
      };
    }
    return {};
  };
  const menuProps = _objectSpread__default["default"](_objectSpread__default["default"]({}, other), {}, {
    menuOffsetCalculated,
    menuOffsetY: menuOffset,
    onChange: handleOnChange,
    onClose: handleClose,
    onOpen: handleOpen,
    popperOptions: getPopperOptions(),
    onMount: (/** @type {import('tippy.js').Instance} */instance) => {
      var _instance$popperInsta;
      return setPlacement((_instance$popperInsta = instance.popperInstance) === null || _instance$popperInsta === void 0 || (_instance$popperInsta = _instance$popperInsta.state) === null || _instance$popperInsta === void 0 ? void 0 : _instance$popperInsta.placement);
    },
    placement: placement || 'bottom'
  });
  return {
    currentDisplayValues,
    currentValues: getValuesAsArray(),
    isMenuVisible,
    menuProps
  };
};
var useDropdown$1 = useDropdown;

/** @typedef {import('../../../floating-menu/lib/utils/computeSearchProps').ComputeSearchProps} ComputeSearchProps */

/** @param {ComputeSearchProps} props */
const computeSearchProps = _ref => {
  let searchProps = _ref.search;
  if (!searchProps) return undefined;
  if (typeof searchProps === 'boolean') {
    return {
      search: true
    };
  }
  return _objectSpread__default["default"](_objectSpread__default["default"]({}, searchProps), {}, {
    search: true
  });
};
var computeSearchProps$1 = computeSearchProps;

/* This function attempts to keep backwards compatibility with value and defaultValue props type */

/**
 * @typedef {Object} ComputeValueTypeProps
 * @prop {string | number | string[]} [defaultValueProp]
 * @prop {string | number | string[]} [valueProp]
 */

/** @param {ComputeValueTypeProps} props */
const computeValueType = _ref => {
  let defaultValueProp = _ref.defaultValueProp,
    valueProp = _ref.valueProp;
  const computeDefaultValue = typeof defaultValueProp === 'number' ? String(defaultValueProp) : defaultValueProp;
  const computeValue = typeof valueProp === 'number' ? String(valueProp) : valueProp;
  return {
    value: computeValue,
    defaultValue: computeDefaultValue
  };
};
var computeValueType$1 = computeValueType;

/**
 * @param {'form' | 'standalone' | 'native'} dropdownType
 * @param {Pick<import('../..').DropdownFormProps, "disabled" | "modifier">} state
 * @param {string} [dataAttributesComponent]
 */
const getDataAttributes = function (dropdownType, _ref) {
  let disabled = _ref.disabled,
    modifier = _ref.modifier;
  let dataAttributesComponent = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'dropdown';
  return {
    container: _objectSpread__default["default"]({
      [`data-andes-${dataAttributesComponent}`]: true,
      'data-andes-dropdown-type': dropdownType
    }, computeDataAndesState.computeDataAndesState({
      disabled,
      error: modifier === MODIFIER_ERROR,
      readonly: modifier === MODIFIER_READONLY
    })),
    label: {
      [`data-andes-${dataAttributesComponent}-label`]: true
    },
    placeholder: {
      [`data-andes-${dataAttributesComponent}-placeholder`]: true
    },
    trigger: {
      [`data-andes-${dataAttributesComponent}-trigger`]: true
    },
    triggerContent: {
      [`data-andes-${dataAttributesComponent}-trigger-content`]: true
    },
    value: {
      [`data-andes-${dataAttributesComponent}-value`]: true
    },
    chevron: {
      [`data-andes-${dataAttributesComponent}-chevron`]: true
    },
    helper: {
      [`data-andes-${dataAttributesComponent}-helper`]: true
    }
  };
};
const _excluded$4 = ["ref"],
  _excluded2$1 = ["menuOffsetCalculated", "menuOffsetY"];

/**
 * @typedef {Object} DropdownFormInternalProps
 * @prop {string} [dataAttributesComponent]
 */

/**
 * @typedef {import('../..').DropdownForm} DropdownFormPublicProps
 */

/**
 * @typedef {DropdownFormPublicProps & DropdownFormInternalProps} DropdownFormProps
 */

/** @type {React.FunctionComponent<React.ComponentProps<DropdownFormProps>>} */
const DropdownForm = _ref => {
  let ref = _ref.ref,
    props = _objectWithoutProperties__default["default"](_ref, _excluded$4);
  const _props$allowUnselect = props.allowUnselect,
    allowUnselect = _props$allowUnselect === void 0 ? false : _props$allowUnselect,
    children = props.children,
    className = props.className,
    _props$defaultOpen = props.defaultOpen,
    defaultOpen = _props$defaultOpen === void 0 ? false : _props$defaultOpen,
    defaultValueProp = props.defaultValue,
    _props$disabled = props.disabled,
    disabled = _props$disabled === void 0 ? false : _props$disabled,
    idProp = props.id,
    _props$label = props.label,
    label = _props$label === void 0 ? '' : _props$label,
    _props$menuAlignment = props.menuAlignment,
    menuAlignment = _props$menuAlignment === void 0 ? 'center' : _props$menuAlignment,
    menuMaxHeight = props.menuMaxHeight,
    modifier = props.modifier,
    _props$multiple = props.multiple,
    multiple = _props$multiple === void 0 ? false : _props$multiple,
    open = props.open,
    _props$placeholder = props.placeholder,
    placeholder = _props$placeholder === void 0 ? '' : _props$placeholder,
    renderCustomTrigger = props.renderCustomTrigger,
    renderMultipleTriggerContent = props.renderMultipleTriggerContent,
    _props$search = props.search,
    search = _props$search === void 0 ? false : _props$search,
    srLabel = props.srLabel,
    valueProp = props.value,
    dataAttributesComponent = props.dataAttributesComponent;
  const id = context.useId(idProp);
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  const dataAttributes = getDataAttributes('form', {
    disabled,
    modifier
  }, dataAttributesComponent);

  /** @type {React.RefObject<HTMLDivElement | null>} */
  const ownContainerRef = React.useRef(null);

  /** @type {React.RefObject<HTMLDivElement | null>} */
  const containerRef = /** @type {React.RefObject<HTMLDivElement>} */ref || ownContainerRef;
  const searchConfig = computeSearchProps$1({
    search
  });
  const _computeValueType = computeValueType$1({
      valueProp,
      defaultValueProp
    }),
    value = _computeValueType.value,
    defaultValue = _computeValueType.defaultValue;
  const defaultProps = {
    disabled,
    multiple,
    menuAlignment,
    search
  };
  const _useDropdown = useDropdown$1(containerRef, _objectSpread__default["default"](_objectSpread__default["default"](_objectSpread__default["default"]({}, props), defaultProps), {}, {
      id,
      searchConfig,
      value,
      defaultValue
    })),
    menuProps = _useDropdown.menuProps,
    isMenuVisible = _useDropdown.isMenuVisible,
    currentDisplayValues = _useDropdown.currentDisplayValues,
    currentValues = _useDropdown.currentValues;
  const currentDisplayValuesHasItems = Array.isArray(currentDisplayValues) && currentDisplayValues.length > 0;

  /**
   * Single selection
   * @param {(content: React.ReactNode) => void} [updateTriggerContent]
   * @returns {React.ReactNode}
   */
  const getTriggerContent = updateTriggerContent => {
    let newTriggerContent = null;
    if (!Array.isArray(currentDisplayValues) && currentDisplayValues) {
      newTriggerContent = /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({
        className: `${namespace}__display-values`,
        id: `${id}-display-values`
      }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.value), currentDisplayValues);
    } else if (currentDisplayValuesHasItems) {
      newTriggerContent = /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({
        className: `${namespace}__display-values`,
        id: `${id}-display-values`
      }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.value), [currentDisplayValues.join()]);
    } else if (placeholder) {
      newTriggerContent = /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({
        className: `${namespace}__placeholder`,
        id: `${id}-display-values`
      }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.placeholder), placeholder);
    }
    if (typeof renderCustomTrigger === 'function') {
      newTriggerContent = /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({
        className: `${namespace}__display-values`,
        id: `${id}-display-values`
      }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.value), renderCustomTrigger(currentDisplayValues, currentValues));
    }
    if (typeof updateTriggerContent === 'function') updateTriggerContent(newTriggerContent);
    return newTriggerContent;
  };

  /**
   * Multiple selection
   * @param {(content: React.ReactElement) => void} [updateTriggerContent]
   * @returns {React.ReactElement}
   */
  const getTagTriggerContent = updateTriggerContent => {
    const newTagContent = /*#__PURE__*/React__default["default"].createElement(TagContent$1, {
      currentDisplayValues: currentDisplayValues,
      currentValues: currentValues,
      disabled: disabled,
      namespace: namespace,
      containerRef: containerRef,
      onSelectionChange: menuProps.onChange,
      id: id,
      renderMultipleTriggerContent: renderMultipleTriggerContent
    });
    if (typeof updateTriggerContent === 'function') updateTriggerContent(newTagContent);
    return newTagContent;
  };
  const _useState = React.useState(getTriggerContent()),
    _useState2 = _slicedToArray__default["default"](_useState, 2),
    triggerContent = _useState2[0],
    setTriggerContent = _useState2[1];
  const _useState3 = React.useState(getTagTriggerContent()),
    _useState4 = _slicedToArray__default["default"](_useState3, 2),
    tagTriggerContent = _useState4[0],
    setTagTriggerContent = _useState4[1];
  React.useEffect(() => {
    if (!multiple || !currentDisplayValuesHasItems) {
      getTriggerContent(setTriggerContent);
    }
    if (multiple) {
      getTagTriggerContent(setTagTriggerContent);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentDisplayValues, disabled]);
  const getMessage = () => {
    const helper = props.helper;
    if (helper && !disabled) {
      return /*#__PURE__*/React__default["default"].createElement(form.Helper
      // @ts-expect-error: Internal prop
      , {
        andesDataAttributes: dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.helper,
        className: `${namespace}__helper`,
        id: `${id}-helper`,
        modifier: modifier === MODIFIER_ERROR ? MODIFIER_ERROR : undefined
      }, helper);
    }
    return null;
  };
  const getMenuWidth = () => {
    const main = containerRef.current;
    const trigger = main === null || main === void 0 ? void 0 : main.getElementsByClassName(triggerNamespace)[0];
    if (trigger) {
      return {
        width: trigger.clientWidth + MENU_BOUNDARIES
      };
    }
    return undefined;
  };
  const containerClassNames = classNames__default["default"](namespace, `${namespace}--form`, formNamespace, className, {
    [`${namespace}--open`]: isMenuVisible,
    [`${namespace}--multiple`]: multiple,
    [`${namespace}--bottom`]: menuAlignment === 'bottom' || search,
    [`${namespace}--error`]: modifier === MODIFIER_ERROR,
    [`${namespace}--disabled`]: disabled
  }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('dropdown'));
  const trigger = /*#__PURE__*/
  // eslint-disable-next-line jsx-a11y/role-supports-aria-props
  React__default["default"].createElement("button", _extends__default["default"]({
    "aria-labelledby": `${srLabel ? `${id}-sr-label` : `${id}-label`} ${id}-display-values`,
    "aria-describedby": getMessage() ? `${id}-helper` : undefined,
    "aria-invalid": modifier === MODIFIER_ERROR,
    className: triggerNamespace,
    disabled: disabled,
    type: "button",
    id: `${id}-trigger`
  }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.trigger), /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({
    className: `${triggerNamespace}-content`
  }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.triggerContent), (!multiple || !currentDisplayValuesHasItems) && triggerContent || /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({
    className: `${namespace}__display-values`
  }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.value)), /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    className: `${namespace}__chevron`,
    "aria-hidden": true
  }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.chevron), /*#__PURE__*/React__default["default"].createElement(ChevronDown20__default["default"], {
    color: "currentColor"
  }))));
  const menuOffsetCalculated = menuProps.menuOffsetCalculated,
    menuOffsetY = menuProps.menuOffsetY,
    other = _objectWithoutProperties__default["default"](menuProps, _excluded2$1);
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    className: containerClassNames,
    ref: containerRef,
    id: menuProps.id
  }, themeHash, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.container), srLabel && /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, {
    id: `${id}-sr-label`
  }, srLabel), label && /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({
    id: `${id}-label`,
    className: `${namespace}__label`,
    "aria-hidden": true
  }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.label), label), /*#__PURE__*/React__default["default"].createElement("div", {
    className: `${namespace}__trigger-wrapper`
  }, /*#__PURE__*/React__default["default"].createElement(floatingMenu.FloatingMenu, _extends__default["default"]({}, other, getMenuWidth(), {
    search: searchConfig,
    srLabel: srLabel || label,
    allowUnselect: allowUnselect,
    className: menuOffsetCalculated ? `${floatingMenuNamespace}--show` : '',
    defaultOpen: defaultOpen,
    defaultValue: defaultValue,
    id: `${menuProps.id}-menu`,
    menuMaxHeight: menuMaxHeight,
    multiple: multiple
    // @ts-expect-error
    // Offset prop doesn't exist in FloatingMenu, it's a popper prop
    ,

    offset: [0, menuOffsetY],
    open: open,
    trigger: trigger,
    type: "select",
    value: multiple ? currentValues : value
  }), children), multiple && tagTriggerContent), getMessage());
};

/** @type {import('prop-types').Validator<string>} */
const checkLabelProp = (_ref2, propName, componentName) => {
  let srLabel = _ref2.srLabel,
    label = _ref2.label;
  PropTypes__default["default"].checkPropTypes({
    srLabel: PropTypes__default["default"].string
  }, {
    srLabel
  }, propName, componentName);
  PropTypes__default["default"].checkPropTypes({
    label: PropTypes__default["default"].string
  }, {
    label
  }, propName, componentName);
  return !srLabel && !label ? new Error(WARNING_LABEL) : null;
};
DropdownForm.propTypes = {
  allowUnselect: PropTypes__default["default"].bool,
  children: PropTypes__default["default"].oneOfType([PropTypes__default["default"].element.isRequired, PropTypes__default["default"].arrayOf(PropTypes__default["default"].element.isRequired)]).isRequired,
  className: PropTypes__default["default"].string,
  defaultOpen: PropTypes__default["default"].bool,
  defaultValue: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].arrayOf(PropTypes__default["default"].string.isRequired)]),
  disabled: PropTypes__default["default"].bool,
  helper: PropTypes__default["default"].node,
  id: PropTypes__default["default"].string,
  label: checkLabelProp,
  menuAlignment: PropTypes__default["default"].oneOf(['bottom', 'center']),
  menuMaxHeight: PropTypes__default["default"].number,
  modifier: PropTypes__default["default"].oneOf(['error', 'readonly']),
  multiple: PropTypes__default["default"].bool,
  onChange: PropTypes__default["default"].func,
  onClose: PropTypes__default["default"].func,
  onOpen: PropTypes__default["default"].func,
  open: PropTypes__default["default"].bool,
  placeholder: PropTypes__default["default"].string,
  search: PropTypes__default["default"].oneOfType([PropTypes__default["default"].bool, PropTypes__default["default"].shape({
    defaultValue: PropTypes__default["default"].string,
    filterCondition: PropTypes__default["default"].func,
    onChange: PropTypes__default["default"].func,
    placeholder: PropTypes__default["default"].string
  })]),
  srLabel: checkLabelProp,
  renderCustomTrigger: PropTypes__default["default"].func,
  renderMultipleTriggerContent: PropTypes__default["default"].func,
  value: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string.isRequired, PropTypes__default["default"].arrayOf(PropTypes__default["default"].string.isRequired)])
};
var DropdownForm$1 = DropdownForm;
const _excluded$3 = ["ref"];

/** @type {typeof import('../..').Dropdown} */
const Dropdown = _ref => {
  let ref = _ref.ref,
    props = _objectWithoutProperties__default["default"](_ref, _excluded$3);
  return /*#__PURE__*/React__default["default"].createElement(DropdownForm$1, _extends__default["default"]({}, props, {
    ref: ref
  }));
};
Dropdown.displayName = 'Dropdown';
var Dropdown$1 = Dropdown;

/** @type {typeof import('../..').DropdownItem} */
const DropdownItem = props => {
  const children = props.children,
    description = props.description,
    rightContent = props.rightContent,
    title = props.title;
  return /*#__PURE__*/React__default["default"].createElement(floatingMenu.FloatingMenuItem, _extends__default["default"]({}, props, {
    title: title,
    description: description,
    rightContent: rightContent
  }), children);
};
DropdownItem.displayName = DROPDOWN_ITEM_DISPLAY_NAME;
DropdownItem.propTypes = {
  children: PropTypes__default["default"].node,
  description: PropTypes__default["default"].oneOfType([PropTypes__default["default"].node, PropTypes__default["default"].string]),
  disabled: PropTypes__default["default"].bool,
  rightContent: PropTypes__default["default"].oneOfType([PropTypes__default["default"].node, PropTypes__default["default"].string]),
  title: PropTypes__default["default"].string,
  value: PropTypes__default["default"].string.isRequired
};
var DropdownItem$1 = DropdownItem;
const _excluded$2 = ["children"];

/** @type {typeof import('../..').DropdownGroup} */
const DropdownGroup = _ref => {
  let children = _ref.children,
    props = _objectWithoutProperties__default["default"](_ref, _excluded$2);
  return /*#__PURE__*/React__default["default"].createElement(floatingMenu.FloatingMenuGroup, props, children);
};
DropdownGroup.displayName = DROPDOWN_GROUP_DISPLAY_NAME;
DropdownGroup.propTypes = {
  children: PropTypes__default["default"].oneOfType([PropTypes__default["default"].element.isRequired, PropTypes__default["default"].arrayOf(PropTypes__default["default"].element.isRequired)]).isRequired
};
var DropdownGroup$1 = DropdownGroup;
const _excluded$1 = ["children", "className", "defaultValue", "disabled", "helper", "id", "label", "modifier", "onChange", "value", "ref"];

/** @type {typeof import('../..').DropdownNative} */
const DropdownNative = _ref => {
  let children = _ref.children,
    className = _ref.className,
    defaultValueProp = _ref.defaultValue,
    disabled = _ref.disabled,
    helper = _ref.helper,
    id = _ref.id,
    label = _ref.label,
    modifier = _ref.modifier,
    onChange = _ref.onChange,
    valueProp = _ref.value,
    ref = _ref.ref,
    props = _objectWithoutProperties__default["default"](_ref, _excluded$1);
  const componentId = context.useId(id);
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  const dataAttributes = getDataAttributes('native', {
    disabled,
    modifier
  });
  const _computeValueType = computeValueType$1({
      valueProp,
      defaultValueProp
    }),
    value = _computeValueType.value,
    defaultValue = _computeValueType.defaultValue;

  /** @param {React.ChangeEvent<HTMLSelectElement>} event */
  const handleOnChange = event => {
    const newValue = event.target.value;
    if (typeof onChange === 'function') onChange(event, newValue);
  };
  const containerClassNames = classNames__default["default"](namespace, className, `${formNamespace}`, `${namespace}--form-native`, `${namespace}--default`, {
    [`${namespace}--error`]: modifier === MODIFIER_ERROR,
    [`${namespace}--disabled`]: disabled
  }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('dropdown'));
  const triggerClassNames = classNames__default["default"](`${namespace}__trigger`, `${namespace}__trigger-reset`);
  const options = React.Children.map(children, (/** @type {React.ReactElement} */item, index) => /*#__PURE__*/React.cloneElement(item, {
    index,
    type: 'native'
  }));
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    id: componentId,
    className: containerClassNames,
    ref: ref
  }, themeHash, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.container), /*#__PURE__*/React__default["default"].createElement("label", null, label && /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({
    className: `${namespace}__label`
  }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.label), label), /*#__PURE__*/React__default["default"].createElement("div", {
    className: `${namespace}__trigger-wrapper`
  }, /*#__PURE__*/React__default["default"].createElement("select", _extends__default["default"]({
    "aria-describedby": helper ? `${id}-helper` : undefined,
    "aria-invalid": modifier === MODIFIER_ERROR,
    className: triggerClassNames,
    defaultValue: defaultValue,
    disabled: disabled,
    id: id,
    key: "andes-select",
    onChange: handleOnChange,
    value: value
  }, props, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.trigger), options), /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    className: `${namespace}__chevron`,
    "aria-hidden": true
  }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.chevron), /*#__PURE__*/React__default["default"].createElement(ChevronDown20__default["default"], {
    color: "currentColor"
  })))), helper && !disabled && /*#__PURE__*/React__default["default"].createElement(form.Helper
  // @ts-expect-error: Internal prop
  , {
    andesDataAttributes: dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.helper,
    className: `${namespace}__helper`,
    id: `${id}-helper`,
    modifier: modifier === MODIFIER_ERROR ? MODIFIER_ERROR : undefined
  }, helper));
};
DropdownNative.displayName = DROPDOWN_NATIVE_DISPLAY_NAME;
DropdownNative.propTypes = {
  children: PropTypes__default["default"].oneOfType([PropTypes__default["default"].element.isRequired, PropTypes__default["default"].arrayOf(PropTypes__default["default"].element.isRequired)]).isRequired,
  className: PropTypes__default["default"].string,
  defaultValue: PropTypes__default["default"].string,
  disabled: PropTypes__default["default"].bool,
  helper: PropTypes__default["default"].node,
  id: PropTypes__default["default"].string,
  label: PropTypes__default["default"].string.isRequired,
  modifier: PropTypes__default["default"].oneOf([MODIFIER_ERROR]),
  onChange: PropTypes__default["default"].func,
  value: PropTypes__default["default"].string
};
var DropdownNative$1 = DropdownNative;

/** @type {typeof import('../..').DropdownNativeItem} */
const DropdownNativeItem = _ref => {
  let _ref$selected = _ref.selected,
    selected = _ref$selected === void 0 ? false : _ref$selected,
    _ref$disabled = _ref.disabled,
    disabled = _ref$disabled === void 0 ? false : _ref$disabled,
    value = _ref.value,
    title = _ref.title,
    children = _ref.children;
  return /*#__PURE__*/React__default["default"].createElement("option", {
    selected: selected || undefined,
    disabled: disabled,
    value: value
  }, title || children);
};
DropdownNativeItem.displayName = DROPDOWN_NATIVE_ITEM_DISPLAY_NAME;
DropdownNativeItem.propTypes = {
  children: PropTypes__default["default"].node,
  disabled: PropTypes__default["default"].bool,
  selected: PropTypes__default["default"].bool,
  title: PropTypes__default["default"].string,
  value: PropTypes__default["default"].string.isRequired
};
var DropdownNativeItem$1 = DropdownNativeItem;
const _excluded = ["ref"],
  _excluded2 = ["menuOffsetCalculated", "menuOffsetY"];

/** @type {typeof import('../..').DropdownStandalone} */
const DropdownStandalone = _ref => {
  let ref = _ref.ref,
    props = _objectWithoutProperties__default["default"](_ref, _excluded);
  const _props$allowUnselect = props.allowUnselect,
    allowUnselect = _props$allowUnselect === void 0 ? false : _props$allowUnselect,
    children = props.children,
    className = props.className,
    defaultOpen = props.defaultOpen,
    defaultValueProp = props.defaultValue,
    _props$disabled = props.disabled,
    disabled = _props$disabled === void 0 ? false : _props$disabled,
    idProp = props.id,
    _props$label = props.label,
    label = _props$label === void 0 ? '' : _props$label,
    _props$menuAlignment = props.menuAlignment,
    menuAlignment = _props$menuAlignment === void 0 ? 'center' : _props$menuAlignment,
    menuMaxHeight = props.menuMaxHeight,
    _props$multiple = props.multiple,
    multiple = _props$multiple === void 0 ? false : _props$multiple,
    open = props.open,
    _props$search = props.search,
    search = _props$search === void 0 ? false : _props$search,
    _props$size = props.size,
    size = _props$size === void 0 ? SIZE_LARGE : _props$size,
    srLabel = props.srLabel,
    valueProp = props.value,
    renderCustomTrigger = props.renderCustomTrigger;
  const id = context.useId(idProp);
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  const dataAttributes = getDataAttributes('standalone', {
    disabled
  });
  warning__default["default"](srLabel || label, WARNING_LABEL);

  /** @type {React.RefObject<HTMLDivElement | null>} */
  const ownContainerRef = React.useRef(null);

  /** @type {React.RefObject<HTMLDivElement | null>} */
  const containerRef = /** @type {React.RefObject<HTMLDivElement>} */ref || ownContainerRef;
  const searchConfig = computeSearchProps$1({
    search
  });
  const _computeValueType = computeValueType$1({
      valueProp,
      defaultValueProp
    }),
    value = _computeValueType.value,
    defaultValue = _computeValueType.defaultValue;
  const defaultProps = {
    disabled,
    menuAlignment,
    multiple,
    search,
    size
  };
  const _useDropdown = useDropdown$1(containerRef, _objectSpread__default["default"](_objectSpread__default["default"](_objectSpread__default["default"]({}, props), defaultProps), {}, {
      id,
      searchConfig,
      value,
      defaultValue
    })),
    menuProps = _useDropdown.menuProps,
    isMenuVisible = _useDropdown.isMenuVisible,
    currentDisplayValues = _useDropdown.currentDisplayValues,
    currentValues = _useDropdown.currentValues;

  /**
   * @param {(content: React.ReactNode) => void} [updateTriggerContent]
   * @returns {React.ReactNode}
   */
  const getTriggerContent = updateTriggerContent => {
    let newTriggerContent = /** @type {React.ReactNode} */'';
    if (typeof currentDisplayValues === 'string') {
      newTriggerContent = currentDisplayValues;
    } else if (Array.isArray(currentDisplayValues) && currentDisplayValues.length === 1) {
      var _currentDisplayValues = _slicedToArray__default["default"](currentDisplayValues, 1);
      newTriggerContent = _currentDisplayValues[0];
    }
    if (typeof renderCustomTrigger === 'function') {
      newTriggerContent = renderCustomTrigger(currentDisplayValues, currentValues);
    }
    if (typeof updateTriggerContent === 'function') updateTriggerContent(newTriggerContent);
    return newTriggerContent;
  };
  const _useState = React.useState(getTriggerContent()),
    _useState2 = _slicedToArray__default["default"](_useState, 2),
    triggerContent = _useState2[0],
    setTriggerContent = _useState2[1];
  React.useEffect(() => {
    getTriggerContent(setTriggerContent);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentDisplayValues]);
  const getMenuWidth = () => {
    const main = containerRef.current;
    const trigger = main === null || main === void 0 ? void 0 : main.getElementsByClassName(triggerNamespace)[0];
    if (trigger) {
      return {
        width: 'auto',
        minWidth: trigger.clientWidth + -MENU_STANDALONE_OFFSET_X
      };
    }
    return undefined;
  };
  const containerClassNames = classNames__default["default"](namespace, `${namespace}--standalone`, className, {
    [`${namespace}--open`]: isMenuVisible,
    [`${namespace}--${size}`]: size,
    [`${namespace}--disabled`]: disabled,
    [`${namespace}--bottom`]: menuAlignment === 'bottom' || searchConfig
  }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('dropdown'));
  const menuOffsetCalculated = menuProps.menuOffsetCalculated,
    menuOffsetY = menuProps.menuOffsetY,
    other = _objectWithoutProperties__default["default"](menuProps, _excluded2);
  const labelledBy = [label && `${id}-label`, triggerContent && `${id}-display-values`].filter(Boolean).join(' ') || undefined;
  const trigger = /*#__PURE__*/React__default["default"].createElement("button", _extends__default["default"]({
    "aria-label": srLabel,
    "aria-labelledby": !srLabel ? labelledBy : undefined,
    className: triggerNamespace,
    disabled: disabled,
    type: "button",
    id: `${id}-trigger`
  }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.trigger), label && /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({
    className: `${triggerContent ? 'andes-visually-hidden' : `${namespace}__display-values`}`,
    id: `${id}-label`
  }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.label), label), triggerContent && /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({
    className: `${namespace}__display-values`,
    id: `${id}-display-values`
  }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.value), triggerContent), /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({
    className: `${namespace}__arrow`,
    "aria-hidden": true
  }, dataAttributes === null || dataAttributes === void 0 ? void 0 : dataAttributes.chevron), /*#__PURE__*/React__default["default"].createElement(ChevronDown12__default["default"], {
    color: "currentColor"
  })));
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    id: menuProps.id,
    className: containerClassNames,
    ref: containerRef
  }, themeHash, dataAttributes.container), /*#__PURE__*/React__default["default"].createElement(floatingMenu.FloatingMenu, _extends__default["default"]({}, other, getMenuWidth(), {
    search: searchConfig,
    srLabel: srLabel || label,
    allowUnselect: allowUnselect,
    className: menuOffsetCalculated ? 'andes-floating-menu--show' : '',
    defaultOpen: defaultOpen,
    defaultValue: defaultValue,
    id: `${menuProps.id}-menu`,
    menuMaxHeight: menuMaxHeight,
    multiple: multiple
    // @ts-expect-error
    // Offset prop doesn't exist in FloatingMenu, it's a popper prop
    ,

    offset: [MENU_STANDALONE_OFFSET_X, menuOffsetY],
    open: open,
    placement: "bottom-start",
    trigger: trigger,
    type: "select",
    value: value
  }), children));
};
DropdownStandalone.propTypes = {
  allowUnselect: PropTypes__default["default"].bool,
  children: PropTypes__default["default"].oneOfType([PropTypes__default["default"].element.isRequired, PropTypes__default["default"].arrayOf(PropTypes__default["default"].element.isRequired)]).isRequired,
  className: PropTypes__default["default"].string,
  defaultOpen: PropTypes__default["default"].bool,
  defaultValue: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].arrayOf(PropTypes__default["default"].string.isRequired)]),
  disabled: PropTypes__default["default"].bool,
  id: PropTypes__default["default"].string,
  label: PropTypes__default["default"].string,
  menuAlignment: PropTypes__default["default"].oneOf(['bottom', 'center']),
  menuMaxHeight: PropTypes__default["default"].number,
  multiple: PropTypes__default["default"].bool,
  onChange: PropTypes__default["default"].func,
  onClose: PropTypes__default["default"].func,
  onOpen: PropTypes__default["default"].func,
  open: PropTypes__default["default"].bool,
  renderCustomTrigger: PropTypes__default["default"].func,
  search: PropTypes__default["default"].oneOfType([PropTypes__default["default"].bool, PropTypes__default["default"].shape({
    defaultValue: PropTypes__default["default"].string,
    placeholder: PropTypes__default["default"].string,
    onChange: PropTypes__default["default"].func,
    filterCondition: PropTypes__default["default"].func
  })]),
  size: PropTypes__default["default"].oneOf(['small', 'medium', 'large']),
  srLabel: PropTypes__default["default"].string,
  value: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string.isRequired, PropTypes__default["default"].arrayOf(PropTypes__default["default"].string.isRequired)])
};
var DropdownStandalone$1 = DropdownStandalone;
exports.Dropdown = Dropdown$1;
exports.DropdownForm = DropdownForm$1;
exports.DropdownGroup = DropdownGroup$1;
exports.DropdownItem = DropdownItem$1;
exports.DropdownNative = DropdownNative$1;
exports.DropdownNativeItem = DropdownNativeItem$1;
exports.DropdownStandalone = DropdownStandalone$1;
