// @nordic/rendering-context/src/server/storage/run.ts
import type { NordicRenderingContextFlags } from '../../common/types';
import type { NordicRenderingServerContext } from './types';

import { getReq } from '@meli/server-context';
import { websecConfig } from 'nordic-csp-engine';

import { createCriticalCssStore } from '../critical-css/store';
import { createLayoutViewStore } from '../layout/store';
import { createMetricsStore } from '../metrics/store';
import { getAssetsData } from '../utils/assets';
import { createIIFEScriptsStore } from '../iife/store';
import { getWrapperProps } from '../utils/wrapper';
import { storage } from '.';

export type NordicRenderingServerContextInitialValues = Pick<
  NordicRenderingServerContext,
  'appProps' | 'framework' | 'mainEntry' | 'melidata'
> & {
  flags: Omit<NordicRenderingContextFlags, 'isBot'>;
  criticalCssPath?: string;
  title?: string;
};

export async function runInRenderingContext<T>(
  initialValues: NordicRenderingServerContextInitialValues,
  callback: () => T,
): Promise<T> {
  const req = getReq();
  const { browser } = req;
  const { support } = browser;
  const { appProps, flags, mainEntry } = initialValues;
  const assetsData = getAssetsData(mainEntry);
  const [securityConfig, metricsStore, layoutViewStore] = await Promise.all([
    websecConfig.get('securityConfig'),
    createMetricsStore(),
    createLayoutViewStore(assetsData.cdn),
  ]);

  return storage.run(
    {
      mainEntry,
      appProps:
        !appProps ||
        mainEntry ||
        flags.staticMarkup ||
        flags.stream ||
        req.nordicPages
          ? appProps
          : appProps.sharedState && { sharedState: appProps.sharedState },
      melidata: initialValues.melidata,
      metrics: metricsStore,
      iifeScripts: createIIFEScriptsStore(flags, support),
      title: initialValues.title ?? '',
      nonce: req.csp?.nonce,
      csrfToken: req.csrfToken?.(),
      trustedTypesRegex: securityConfig?.trustedTypes.regex,
      assets: assetsData,
      criticalCss: createCriticalCssStore(initialValues.criticalCssPath),
      wrapperProps: getWrapperProps(),
      layoutView: layoutViewStore,
      framework: initialValues.framework,
      flags: {
        isBot: req.isBot,
        loadable: flags.loadable,
        islands: flags.islands,
        pipeableStream: flags.pipeableStream,
        stream: flags.stream,
        staticMarkup: req.isBot || browser.outdated || flags.staticMarkup,
      },
      support: {
        customElements: support.customElements,
        imgLoading: support.imgLoading,
        prefetch: support.prefetch,
        sameSite: support.sameSite,
        outdated: browser.outdated,
        share: support.share,
      },
    },
    callback,
  );
}
