// @andes/common/build/esm/metrics/node-index.production.js
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
const _excluded = ["type"];
const _require = require('frontend-statsd/internal'),
  statsd = _require.statsd;
const env = require('frontend-env');
const _require2 = require('@andes/common/build/utils/encodeBase32'),
  encodeStringToBase32 = _require2.encodeStringToBase32;
const transformToKebabCase = require('@andes/common/build/utils/transformToKebabCase');
const statName = 'andes.frontend_andes_ui.customization';
const initialValue = 1;
const sampleRate = 0.25;
const trackCustomization = _ref => {
  let type = _ref.type,
    other = _objectWithoutProperties(_ref, _excluded);
  statsd.increment(statName, initialValue, sampleRate, _objectSpread({
    application: env.APPLICATION,
    platform: env.PLATFORM || '',
    customizationType: type
  }, other), () => {});
};
const CUSTOMIZATION_TYPE = {
  CUSTOM: 'custom',
  HEADLESS: 'headless',
  STYLE_OVERRIDES: 'styleOverrides',
  TOKENS: 'tokens',
  SLOTS: 'slots'
};
const MAX_TAG_LENGTH = 200;
const shortenObjectToMaxLength = function () {
  let obj = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let maxLength = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  const shortenObject = {};
  const objectString = JSON.stringify(obj);
  const objectBase32 = encodeStringToBase32(objectString);
  if (objectBase32.length <= maxLength) {
    return obj;
  }
  let accumulatedCharacters = 0;
  for (const _ref2 of Object.entries(obj)) {
    var _ref3 = _slicedToArray(_ref2, 2);
    const key = _ref3[0];
    const value = _ref3[1];
    const stringProperty = `"${key}":${JSON.stringify(value)},`;
    if (accumulatedCharacters + stringProperty.length <= maxLength) {
      shortenObject[key] = value;
      accumulatedCharacters += stringProperty.length;
    } else {
      break;
    }
  }
  return shortenObject;
};
const parseObjectValuesToBase32 = function () {
  let obj = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return Object.entries(obj).reduce((acc, _ref4) => {
    let _ref5 = _slicedToArray(_ref4, 2),
      key = _ref5[0],
      value = _ref5[1];
    acc[key] = encodeStringToBase32(value);
    return acc;
  }, {});
};
const senderMetrics = tracker => {
  const sendCustom = _ref6 => {
    let component = _ref6.component,
      _ref6$details = _ref6.details,
      details = _ref6$details === void 0 ? {} : _ref6$details;
    const tagKey = 'custom-properties';
    const properties = shortenObjectToMaxLength(details, MAX_TAG_LENGTH - tagKey.length);
    const propertiesString = JSON.stringify(properties);
    const propertiesBase32 = encodeStringToBase32(propertiesString);
    const hasMoreProperties = Object.keys(properties).length < Object.keys(details).length;
    const data = {
      [tagKey]: propertiesBase32,
      'more-properties': hasMoreProperties
    };
    tracker === null || tracker === void 0 || tracker(_objectSpread({
      type: CUSTOMIZATION_TYPE.CUSTOM,
      component: typeof component === 'string' ? transformToKebabCase(component) : undefined
    }, data));
  };
  const sendHeadless = componentName => componentProps => {
    const shouldSendHeadlessComponentMetric = 'extend' in componentProps && componentProps.extend;
    if (shouldSendHeadlessComponentMetric) {
      tracker === null || tracker === void 0 || tracker({
        type: CUSTOMIZATION_TYPE.HEADLESS,
        component: componentName
      });
    }
  };
  const sendSlots = (componentName, slotConditionFn) => componentProps => {
    const shouldSlotComponentMetric = typeof slotConditionFn === 'function' ? slotConditionFn(componentProps) : false;
    if (shouldSlotComponentMetric) {
      tracker === null || tracker === void 0 || tracker({
        type: CUSTOMIZATION_TYPE.SLOTS,
        component: componentName
      });
    }
  };
  const sendStyleOverrides = _ref7 => {
    let component = _ref7.component;
    tracker === null || tracker === void 0 || tracker({
      type: CUSTOMIZATION_TYPE.STYLE_OVERRIDES,
      component: transformToKebabCase(component)
    });
  };
  const sendTokens = _ref8 => {
    let _ref8$details = _ref8.details,
      details = _ref8$details === void 0 ? {} : _ref8$details;
    const tokensStringToMaxLength = Object.entries(details).reduce((acc, _ref9) => {
      let _ref0 = _slicedToArray(_ref9, 2),
        key = _ref0[0],
        value = _ref0[1];
      acc[key] = typeof value === 'string' ? value.slice(0, Math.max(MAX_TAG_LENGTH - key.length, 0)) : value;
      return acc;
    }, {});
    const tokens = parseObjectValuesToBase32(tokensStringToMaxLength);
    tracker === null || tracker === void 0 || tracker(_objectSpread({
      type: CUSTOMIZATION_TYPE.TOKENS
    }, tokens));
  };
  return {
    sendCustom,
    sendHeadless,
    sendSlots,
    sendStyleOverrides,
    sendTokens
  };
};
module.exports = {
  trackCustomization,
  senderMetrics
};
