// @nordic/rendering-context/src/server/metrics/store/gtm.ts
import type { CookieCategories } from '@frontend-cookies-manager/cookies-helpers-middleware';
import type { GTMStore } from './types';

import { getReq } from '@meli/server-context';

import { getConfigProps } from './utils';

async function getUserConsentCookieCategory(category: CookieCategories) {
  const req = getReq();

  return {
    [`user_consent_cookies_${category}`]:
      await req.meliCookies!.userConsentCookieCategory(category),
  };
}

async function createCommonDataLayer(): Promise<GTMStore['commonDataLayer']> {
  const common: GTMStore['commonDataLayer'] = [];
  const req = getReq();

  try {
    const cookiesPrefs = await req.meliCookies?.getCookiesUserPreferences();

    if (cookiesPrefs?.categories) {
      const consents = await Promise.all(
        (Object.keys(cookiesPrefs.categories) as Array<CookieCategories>).map(
          getUserConsentCookieCategory,
        ),
      );

      common.push(...consents);
    }
  } catch {
    // Should never fail, but just in case
  }

  if (typeof req.data_audience_3p_denied !== 'undefined') {
    common.push({ data_audience_3p_denied: req.data_audience_3p_denied });
  }

  return common;
}

export async function createGTMStore(): Promise<GTMStore> {
  return {
    enabled: false,
    debugMode: false,
    props: {},
    configProps: getConfigProps('googleTagManager'),
    commonDataLayer: await createCommonDataLayer(),
  };
}
