// @nordic/rendering-context/src/server/metrics/store/types.ts
export type MetricStore<
  P extends Record<string, unknown> = Record<string, unknown>,
> = {
  /**
   * Whether this metric module is enabled.
   * To enable it, this metric React Component must be used.
   */
  enabled: boolean;

  /**
   * Props collected from this metric's React component
   */
  props: P;

  /**
   * Props set in frontend-config
   */
  configProps: P;

  /**
   * Whether the debug mode is enabled
   */
  debugMode: boolean;
};

export type HotjarStore<
  P extends Record<string, unknown> = Record<string, unknown>,
> = MetricStore<P> & {
  /**
   * Whether the user has consented to the use of traceability cookies
   */
  hasUserConsent: boolean;
};

export type GTMStore<
  P extends Record<string, unknown> = Record<string, unknown>,
> = MetricStore<P> & {
  /**
   * Data layer to inject to all GTM tags
   */
  commonDataLayer: Array<Record<string, boolean | number | string>>;
};

export type QualtricsInsightsStore<
  P extends Record<string, unknown> = Record<string, unknown>,
> = MetricStore<P>;

export type MetricsStore<
  HotjarProps extends Record<string, unknown> = Record<string, unknown>,
  GTMProps extends Record<string, unknown> = Record<string, unknown>,
  QualtricsInsightsProps extends Record<string, unknown> = Record<
    string,
    unknown
  >,
> = {
  hotjar: HotjarStore<HotjarProps>;
  googleTagManager: GTMStore<GTMProps>;
  qualtricsInsights: QualtricsInsightsStore<QualtricsInsightsProps>;
};

export type MetricStoreName = keyof MetricsStore;
