// @meli/input-validation/dist/iv/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.infer = exports.ivNullable = exports.ivOptional = exports.ivNullish = exports.createValidationMiddleware = exports.requestValidation = exports.ivDiscriminatedUnion = exports.ivIntersection = exports.ivUnion = exports.ivStringBoolean = exports.ivCoerce = exports.ivObject = exports.ivString = exports.ivArray = exports.ivNull = exports.ivUndefined = exports.ivNativeEnum = exports.ivEnum = exports.ivLiteral = exports.ivDate = exports.ivBigint = exports.ivBoolean = exports.ivNumber = void 0;
const zod_1 = require("zod");
const secureTypeMetric_1 = require("../utils/secureTypeMetric");
// Inherently secure types: all enhancements are applied via createSecureType
exports.ivNumber = (0, secureTypeMetric_1.createSecureType)(zod_1.z.number);
exports.ivBoolean = (0, secureTypeMetric_1.createSecureType)(zod_1.z.boolean);
exports.ivBigint = (0, secureTypeMetric_1.createSecureType)(zod_1.z.bigint);
exports.ivDate = (0, secureTypeMetric_1.createSecureType)(zod_1.z.date);
exports.ivLiteral = (0, secureTypeMetric_1.createSecureType)(zod_1.z.literal);
exports.ivEnum = (0, secureTypeMetric_1.createSecureType)(zod_1.z.enum);
exports.ivNativeEnum = (0, secureTypeMetric_1.createSecureType)(zod_1.z.nativeEnum);
exports.ivUndefined = (0, secureTypeMetric_1.createSecureType)(zod_1.z.undefined);
exports.ivNull = (0, secureTypeMetric_1.createSecureType)(zod_1.z.null);
var ivArray_1 = require("./entity/ivArray");
Object.defineProperty(exports, "ivArray", { enumerable: true, get: function () { return __importDefault(ivArray_1).default; } });
var ivString_1 = require("./entity/ivString");
Object.defineProperty(exports, "ivString", { enumerable: true, get: function () { return __importDefault(ivString_1).default; } });
var ivObject_1 = require("./entity/ivObject");
Object.defineProperty(exports, "ivObject", { enumerable: true, get: function () { return __importDefault(ivObject_1).default; } });
var ivCoerce_1 = require("./entity/ivCoerce");
Object.defineProperty(exports, "ivCoerce", { enumerable: true, get: function () { return __importDefault(ivCoerce_1).default; } });
var ivStringBoolean_1 = require("./entity/ivStringBoolean");
Object.defineProperty(exports, "ivStringBoolean", { enumerable: true, get: function () { return __importDefault(ivStringBoolean_1).default; } });
var ivUnion_1 = require("./operator/ivUnion");
Object.defineProperty(exports, "ivUnion", { enumerable: true, get: function () { return __importDefault(ivUnion_1).default; } });
var ivIntersection_1 = require("./operator/ivIntersection");
Object.defineProperty(exports, "ivIntersection", { enumerable: true, get: function () { return __importDefault(ivIntersection_1).default; } });
var ivDiscriminatedUnion_1 = require("./operator/ivDiscriminatedUnion");
Object.defineProperty(exports, "ivDiscriminatedUnion", { enumerable: true, get: function () { return __importDefault(ivDiscriminatedUnion_1).default; } });
var requestValidation_1 = require("./middleware/requestValidation");
Object.defineProperty(exports, "requestValidation", { enumerable: true, get: function () { return __importDefault(requestValidation_1).default; } });
Object.defineProperty(exports, "createValidationMiddleware", { enumerable: true, get: function () { return requestValidation_1.createValidationMiddleware; } });
var ivNullish_1 = require("./modifier/ivNullish");
Object.defineProperty(exports, "ivNullish", { enumerable: true, get: function () { return __importDefault(ivNullish_1).default; } });
var ivOptional_1 = require("./modifier/ivOptional");
Object.defineProperty(exports, "ivOptional", { enumerable: true, get: function () { return __importDefault(ivOptional_1).default; } });
var ivNullable_1 = require("./modifier/ivNullable");
Object.defineProperty(exports, "ivNullable", { enumerable: true, get: function () { return __importDefault(ivNullable_1).default; } });
// Función infer
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const infer = (schema) => ({});
exports.infer = infer;
