// @andes/textfield/build/cjs/masks/mx/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

/** @param {string} value */
const validateDate = value => {
  const _value$split = value.split('/'),
    _value$split2 = _slicedToArray__default["default"](_value$split, 3),
    day = _value$split2[0],
    month = _value$split2[1],
    year = _value$split2[2];
  const dayNumber = parseInt(day, 10);
  const monthNumber = parseInt(month, 10);
  const yearNumber = parseInt(year, 10);
  let newValue = value;

  // Validate if first day number greater than 3, auto adjust with 0 before
  if ((day === null || day === void 0 ? void 0 : day[0]) > '3' && !(day !== null && day !== void 0 && day[1])) {
    newValue = `0${day[0]}`;
  }

  // Validate no value is greater than max valid
  if (dayNumber > 31 || dayNumber === 0) {
    return day[0];
  }

  // Validate if first month number greater than 1, auto adjust with 0 before
  if ((month === null || month === void 0 ? void 0 : month[0]) > '1' && !(month !== null && month !== void 0 && month[1])) {
    newValue = `${day}/0${month[0]}`;
  }

  // Validate no value is greater than max valid
  if (monthNumber > 12 || monthNumber === 0) {
    return `${day}/${(month === null || month === void 0 ? void 0 : month[0]) === '1' || (month === null || month === void 0 ? void 0 : month[0]) === '0' ? month === null || month === void 0 ? void 0 : month[0] : ''}`;
  }

  // Validate day in month is valid
  const daysInMonth = [31, Number(year) % 4 === 0 ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  if (monthNumber && Number(String(monthNumber)[0]) !== 0 && dayNumber > daysInMonth[monthNumber - 1]) {
    return `${day}/${month[0]}`;
  }

  // Validates year digits
  if (yearNumber === 0) return `${day}/${month}`;
  if (yearNumber) {
    const firstDigit = year === null || year === void 0 ? void 0 : year[0];
    const secondDigit = year === null || year === void 0 ? void 0 : year[1];
    const thirdDigit = year === null || year === void 0 ? void 0 : year[2];
    const fourthDigit = year === null || year === void 0 ? void 0 : year[3];
    if (firstDigit && firstDigit !== '1' && firstDigit !== '2') {
      return `${day}/${month}/`;
    }
    if (secondDigit && !(firstDigit === '1' && secondDigit === '9' || firstDigit === '2' && (secondDigit === '1' || secondDigit === '0'))) {
      return `${day}/${month}/${firstDigit}`;
    }
    if (thirdDigit && !(secondDigit === '9' || secondDigit === '0' || secondDigit === '1' && thirdDigit === '0')) {
      return `${day}/${month}/${firstDigit}${secondDigit}`;
    }
    if (fourthDigit && !(secondDigit === '9' || secondDigit === '0' || secondDigit === '1' && fourthDigit === '0')) {
      return `${day}/${month}/${firstDigit}${secondDigit}${thirdDigit}`;
    }
  }
  return newValue;
};

/** @type {import('../../..').TextfieldMask} */
var date = {
  mask: 'DD/DD/DDDD',
  tokens: {
    D: {
      pattern: /\d/
    }
  },
  eager: true,
  preProcess: value => value.replace(/[^\d/:]/g, ''),
  postProcess: value => validateDate(value)
};

/** @type {import('../../..').TextfieldMask} */
var money = {
  mask: 'N.DD',
  tokens: {
    N: {
      pattern: /\d/,
      multiple: true
    },
    D: {
      pattern: /\d/,
      optional: true
    }
  },
  preProcess: val => val.replace(/,/g, ''),
  postProcess: val => {
    if (!val || val === '.') return '';
    const sub = 3 - (val.includes('.') ? val.length - val.indexOf('.') : 0);
    return Intl.NumberFormat('es-MX', {
      style: 'currency',
      currency: 'MXN'
    }).format(Number(val.replace(',', '.'))).replace('$', '').replace(/\s/g, '').slice(0, sub ? -sub : undefined);
  }
};
var phone = {
  mask: '(ZZ) DDD-DDDD',
  eager: true,
  tokens: {
    Z: {
      pattern: /[1-9]/
    },
    D: {
      pattern: /\d/
    }
  }
};
var zipcode = {
  mask: 'DDDDD',
  tokens: {
    D: {
      pattern: /\d/
    }
  }
};
var rfc = {
  mask: 'LLLLDDDDDDLLL',
  eager: true,
  tokens: {
    L: {
      pattern: /[A-ZÑ&]/
    },
    D: {
      pattern: /\d/
    }
  }
};
exports.PHONE = phone;
exports.RFC = rfc;
exports.ZIPCODE = zipcode;
exports.date = date;
exports.money = money;
