// @nordic/fs-cache/dist/esm/index.production.mjs
import { ASSETS_CONFIG, getAssetsPrefix } from '@nordic/assets-config';
import path from 'node:path';
import fs from 'node:fs';
import fsPromises from 'node:fs/promises';
import { freezeObject } from '@nordic/utils/freeze';
import { isBlurableImg } from '@nordic/utils/validation';
import { env } from '@nordic/env';
import { serializeJSON } from '@nordic/utils/serialize';
import { getDiscoveredNordicPages } from 'nordic-pages-discovery';
import { SourceMapConsumer } from 'source-map-js';
import { getOptionalModule } from '@nordic/utils';

async function readBuildFileContent$2(fileName) {
    const filePath = path.resolve('build', fileName);
    const readFile = fsPromises.readFile;
    try {
        const content = await readFile(filePath, 'utf8');
        return content;
    }
    catch {
        return null;
    }
}
function readBuildFileContentSync$2(fileName) {
    const filePath = path.resolve('build', fileName);
    const webpackDevFs = false;
    try {
        const readFile = webpackDevFs ? webpackDevFs.readFileSync : fs.readFileSync;
        return readFile(filePath, 'utf8');
    }
    catch {
        return null;
    }
}
function readLoadableStats() {
    try {
        return freezeObject(JSON.parse(fs.readFileSync('build/loadable-stats.json', 'utf8')));
    }
    catch {
        return null;
    }
}

const IS_VITE = env.BUNDLER === 'vite';

/* eslint-disable security/detect-non-literal-fs-filename */
function getBuildFilePath$1(fileName) {
    return path.resolve(ASSETS_CONFIG.CLIENT_OUTPUT_DIR, fileName);
}
async function readBuildFileContent$1(fileName) {
    const filePath = getBuildFilePath$1(fileName);
    try {
        const content = await fsPromises.readFile(filePath, 'utf8');
        return content;
    }
    catch {
        return null;
    }
}
function readBuildFileContentSync$1(fileName) {
    const filePath = getBuildFilePath$1(fileName);
    try {
        return fs.readFileSync(filePath, 'utf8');
    }
    catch {
        return null;
    }
}

/* eslint-disable security/detect-non-literal-fs-filename */
const TEXT_ENCODER = new TextEncoder();
function getSizeInMemoryOfString(str) {
    return TEXT_ENCODER.encode(str).length;
}
/* v8 ignore next -- Temp condition while webpack support is maintained */
const readBuildFileContent = IS_VITE
    ? readBuildFileContent$1
    : readBuildFileContent$2;
// TODO: Move Vite methods to this file when webpack support is dropped
const getBuildFilePath = getBuildFilePath$1;
/* v8 ignore next -- Temp condition while webpack support is maintained */
const readBuildFileContentSync = IS_VITE
    ? readBuildFileContentSync$1
    : readBuildFileContentSync$2;
async function readFileContent(filePath) {
    try {
        return await fsPromises.readFile(path.resolve(filePath), 'utf8');
    }
    catch {
        return null;
    }
}
function readFileContentSync(filePath, encoding = 'utf8') {
    try {
        return fs.readFileSync(path.resolve(filePath), encoding);
    }
    catch {
        return null;
    }
}
function readJsonFileSync(filePath) {
    const fileContent = readFileContentSync(filePath);
    if (!fileContent) {
        return null;
    }
    return JSON.parse(fileContent);
}

const MANIFEST_ENTRYPOINTS_EXTENSIONS_REGEX = /\.(?:js|css)$/;
const SOURCE_MAP_COMMENT_REGEX = /(?:\/\*|\/\/)#\ssourceMappingURL=.+?\.(?:cjs|mjs|js|css)\.map(?:\*\/)?/gs;
const RELATIVE_PATH_REGEX = /^[./]/;
const LAST_DOT_REGEX = /\.(?=[^.]*$)/;

function readAssetFile(filePath, fromBuild = false) {
    const readFile = fromBuild ? readBuildFileContentSync : readFileContentSync;
    const fileContent = readFile(filePath)
        ?.replace(SOURCE_MAP_COMMENT_REGEX, '')
        .trim();
    {
        return fileContent;
    }
}
function readImageAsset(assetName, fileName) {
    const content = readFileContentSync(path.join(ASSETS_CONFIG.IMAGES_BLUR_DIR, `${ASSETS_CONFIG.IMAGES_BLUR_PREFIX}${fileName.replace(LAST_DOT_REGEX, '-')}`));
    return content;
}
function readNodeModulesAsset(assetName, isSmartPOS) {
    const nodeModulesPath = path.join('node_modules', assetName);
    const { dir, name, ext } = path.parse(nodeModulesPath);
    const envExt = '.production';
    const smartPOSExt = isSmartPOS ? '.smartpos' : '';
    const basePath = path.join(dir, name);
    const envBasePath = `${basePath}${envExt}`;
    if (!smartPOSExt) {
        return (readAssetFile(`${envBasePath}${ext}`) ?? readAssetFile(nodeModulesPath));
    }
    const smartPOSBasePath = `${basePath}${smartPOSExt}`;
    return (readAssetFile(`${smartPOSBasePath}${envExt}${ext}`) ??
        readAssetFile(`${smartPOSBasePath}${ext}`) ??
        readAssetFile(`${envBasePath}${ext}`) ??
        readAssetFile(nodeModulesPath));
}
function readAsset(assetName, fileName, isSmartPOS) {
    if (RELATIVE_PATH_REGEX.test(assetName)) {
        return { content: readAssetFile(assetName) };
    }
    if (fileName) {
        return {
            content: isBlurableImg(assetName)
                ? readImageAsset(assetName, fileName)
                : readAssetFile(fileName, true),
        };
    }
    const content = readNodeModulesAsset(assetName, isSmartPOS);
    if (content) {
        return { content, fromNodeModules: true };
    }
    return { content: readAssetFile(assetName) };
}

function readDiscoveryEntrypoints() {
    const discoveryManifest = getDiscoveredNordicPages();
    const entrypoints = new Set();
    if (!discoveryManifest) {
        return entrypoints;
    }
    const addEntrypointsByPage = ({ paths }) => Object.values(paths).forEach(({ client }) => client && entrypoints.add(client));
    discoveryManifest.discoveredNavigablePages.forEach(addEntrypointsByPage);
    Object.values(discoveryManifest.discoveredHttpErrorPages).forEach(addEntrypointsByPage);
    return entrypoints;
}

const MANIFEST_EXCLUDE_EXTENSIONS_REGEX = /\.(?:map|json)$/;
const DEFAULT_STATS = {
    entrypoints: {},
    namedChunkGroups: {},
};
function createBaseManifestEntries$1() {
    const manifest = readJsonFileSync('build/manifest.json');
    if (!manifest) {
        return [];
    }
    return Object.entries(manifest).filter(([, value]) => !MANIFEST_EXCLUDE_EXTENSIONS_REGEX.test(value) &&
        !value.includes('node_modules'));
}
function createMainAssetsManifestEntries(namedChunkGroups, manifestEntries) {
    const invertedManifest = new Map(manifestEntries.map(([key, value]) => [value, key]));
    return Object.entries(namedChunkGroups).map(([key, value]) => {
        const scripts = {
            names: [],
            sources: []};
        const styles = {
            names: [],
            sources: []};
        value.assets?.forEach(({ name: fileName }) => {
            const { sources, names } = fileName.endsWith('.js') ? scripts : styles;
            const assetName = invertedManifest.get(fileName);
            sources.push(fileName);
            if (assetName) {
                names.push(assetName);
            }
        });
        return [
            key,
            {
                scripts: {
                    names: freezeObject(scripts.names),
                    sources: freezeObject(scripts.sources),
                    serializedNames: serializeJSON(scripts.names),
                },
                styles: {
                    names: freezeObject(styles.names),
                    sources: freezeObject(styles.sources),
                    serializedNames: serializeJSON(styles.names),
                },
            },
        ];
    });
}
function createManifestEntries$2() {
    const stats = readJsonFileSync('build/stats.json') ?? DEFAULT_STATS;
    const baseManifestEntries = createBaseManifestEntries$1();
    const mainAssetsManifestEntries = createMainAssetsManifestEntries(stats.namedChunkGroups, baseManifestEntries);
    const entrypointsValues = Object.values(stats.entrypoints);
    return {
        baseManifestEntries: baseManifestEntries.filter(([, value]) => !MANIFEST_ENTRYPOINTS_EXTENSIONS_REGEX.test(value) ||
            entrypointsValues.some(({ assets }) => assets?.some(({ name }) => name === value))),
        mainAssetsManifestEntries,
    };
}

function createBaseManifestEntries(clientStats) {
    const manifest = readJsonFileSync(ASSETS_CONFIG.MANIFEST_PATH) ?? {};
    const entries = Object.entries(manifest);
    if (!clientStats) {
        return entries;
    }
    return entries.filter(([key, value]) => {
        const entry = path.join(ASSETS_CONFIG.PREFIX, key);
        return (!MANIFEST_ENTRYPOINTS_EXTENSIONS_REGEX.test(value) ||
            clientStats[entry] ||
            clientStats[entry.replace(MANIFEST_ENTRYPOINTS_EXTENSIONS_REGEX, '.js')] ||
            // eslint-disable-next-line security/detect-non-literal-fs-filename
            (false));
    });
}
function createManifestEntries$1() {
    const clientStats = readJsonFileSync(ASSETS_CONFIG.CLIENT_STATS_PATH);
    const baseManifestEntries = createBaseManifestEntries(clientStats);
    const mainAssetsManifestEntries = [];
    return { baseManifestEntries, mainAssetsManifestEntries };
}

/* v8 ignore next -- Temp condition while webpack support is maintained */
const createManifestEntries = IS_VITE
    ? createManifestEntries$1
    : createManifestEntries$2;
function getSerializedManifest(manifestEntries) {
    const discoveryEntrypoints = readDiscoveryEntrypoints();
    return serializeJSON(manifestEntries.filter(([key, value]) => MANIFEST_ENTRYPOINTS_EXTENSIONS_REGEX.test(value) &&
        !discoveryEntrypoints.has(key.replace(MANIFEST_ENTRYPOINTS_EXTENSIONS_REGEX, ''))));
}
function createManifests() {
    const { baseManifestEntries, mainAssetsManifestEntries } = createManifestEntries();
    return {
        manifest: new Map(baseManifestEntries),
        mainAssetsManifest: new Map(mainAssetsManifestEntries),
        serializedManifest: getSerializedManifest(baseManifestEntries),
    };
}

const CACHE_MEMORY_LIMIT = 10000000; // Limit of 10MB in memory

class NordicAssetCacheLimitError extends Error {
    constructor(assetName) {
        super(`An error occurred while saving into file system cache: Asset "${assetName}" exceeds the memory limit of ` +
            `${CACHE_MEMORY_LIMIT / 1000000}MB defined for the cache.`);
        this.name = 'NordicAssetCacheLimitError';
    }
}

class LimitedStringMap extends Map {
    #mapMemorySize = 0;
    set(key, value) {
        const entrySize = getSizeInMemoryOfString(key + value);
        if (entrySize > CACHE_MEMORY_LIMIT) {
            throw new NordicAssetCacheLimitError(key);
        }
        // Delete first entry until cache is not exceeded anymore
        while (CACHE_MEMORY_LIMIT < this.#mapMemorySize + entrySize && this.size) {
            const [firstElementKey, firstElementValue] = this.entries().next().value;
            const firstElementSize = getSizeInMemoryOfString(firstElementKey + firstElementValue);
            this.delete(firstElementKey);
            this.#mapMemorySize -= firstElementSize;
        }
        this.#mapMemorySize += entrySize;
        return super.set(key, value);
    }
}

class AssetsCache {
    // TODO: Remove when webpack support is dropped
    #loadableStats = null;
    #manifest = new Map();
    #mainAssetsManifest = new Map();
    #assetsContent = new LimitedStringMap();
    #serializedManifest = '';
    get loadableStats() {
        return this.#loadableStats;
    }
    get manifest() {
        return this.#manifest;
    }
    get mainAssetsManifest() {
        return this.#mainAssetsManifest;
    }
    get serializedManifest() {
        return this.#serializedManifest;
    }
    initialize() {
        ({
            manifest: this.#manifest,
            mainAssetsManifest: this.#mainAssetsManifest,
            serializedManifest: this.#serializedManifest,
        } = createManifests());
        this.#loadableStats = readLoadableStats();
    }
    getAssetContent(assetName, isSmartPOS) {
        const cachedContent = this.#assetsContent.get(assetName);
        if (cachedContent) {
            return cachedContent;
        }
        const { content, fromNodeModules } = readAsset(assetName, this.#manifest.get(assetName), isSmartPOS);
        if (content) {
            {
                this.#assetsContent.set(assetName, content);
            }
            return content;
        }
        return null;
    }
    getMainScripts(mainEntry) {
        return this.#getMainAssets(mainEntry, 'scripts', 'sources');
    }
    getMainScriptsNames(mainEntry) {
        return this.#getMainAssets(mainEntry, 'scripts', 'names');
    }
    getMainStyles(mainEntry) {
        return this.#getMainAssets(mainEntry, 'styles', 'sources');
    }
    getMainStylesNames(mainEntry) {
        return this.#getMainAssets(mainEntry, 'styles', 'names');
    }
    // TODO: Remove when webpack support is dropped
    getSerializedMainScriptsNames(mainEntry) {
        return this.#mainAssetsManifest.get(mainEntry)?.scripts.serializedNames;
    }
    // TODO: Remove when webpack support is dropped
    getSerializedMainStylesNames(mainEntry) {
        return this.#mainAssetsManifest.get(mainEntry)?.styles.serializedNames;
    }
    #getMainAssets(mainEntry, type, values) {
        const assets = new Set();
        const prefix = values === 'sources' ? getAssetsPrefix() : '';
        const addAsset = (entry) => {
            this.#mainAssetsManifest
                .get(entry)?.[type][values].forEach((asset) => assets.add(prefix + asset));
        };
        if (Array.isArray(mainEntry) || mainEntry instanceof Set) {
            mainEntry.forEach(addAsset);
        }
        else {
            addAsset(mainEntry);
        }
        return assets;
    }
}
const assetsCache = new AssetsCache();

const CWD = process.cwd();
const APP_DIR_NAME = path.basename(CWD);
function getFileMappedPosition$1(fileContent, line, column) {
    const consumer = new SourceMapConsumer(JSON.parse(fileContent));
    const mappedPosition = consumer.originalPositionFor({ line, column });
    return mappedPosition.source ? mappedPosition : null;
}
async function getMappedPositionFromBuild({ file, line, column, }) {
    const mapFileContent = await readBuildFileContent(`${file}.map`);
    if (!mapFileContent) {
        return null;
    }
    const mappedPosition = getFileMappedPosition$1(mapFileContent, line, column);
    if (!mappedPosition) {
        return null;
    }
    return {
        source: mappedPosition.source.replace(`webpack://${APP_DIR_NAME}`, CWD),
        line: mappedPosition.line,
        column,
    };
}
async function getMappedPositionFromNodeModules(buildMappedPosition, nodeModulesCache) {
    if (!buildMappedPosition.source.startsWith(path.join(CWD, 'node_modules'))) {
        return null;
    }
    const mapFilePath = `${buildMappedPosition.source}.map`;
    const mapFileContent = nodeModulesCache.get(mapFilePath) ?? (await readFileContent(mapFilePath));
    if (!mapFileContent) {
        return null;
    }
    nodeModulesCache.set(mapFilePath, mapFileContent);
    const mappedPosition = getFileMappedPosition$1(mapFileContent, buildMappedPosition.line, buildMappedPosition.column);
    if (!mappedPosition) {
        return null;
    }
    return {
        source: path.join(path.dirname(buildMappedPosition.source), mappedPosition.source),
        line: mappedPosition.line,
        column: mappedPosition.column,
    };
}
async function getMappedPosition$2(nodeModulesCache, codeLocation) {
    const buildMappedPosition = await getMappedPositionFromBuild(codeLocation);
    if (!buildMappedPosition) {
        return null;
    }
    const nodeModulesMappedPosition = await getMappedPositionFromNodeModules(buildMappedPosition, nodeModulesCache);
    return nodeModulesMappedPosition ?? buildMappedPosition;
}

const NODE_MODULES_PATH = path.resolve('node_modules');
function parseURL(url) {
    try {
        return new URL(url);
    }
    catch {
        return null;
    }
}
async function transformFileToMapContent(filePath) {
    try {
        if (filePath.startsWith(NODE_MODULES_PATH)) {
            return null;
        }
        // TODO: Pending to replace until the Vite module is added
        // const { getViteDevServer } = await getOptionalModule(import('@nordic-dev/vite/server'));
        const { getViteDevServer } = await getOptionalModule(import('./vite-temp-w1Q8jG-9.js'));
        const transformedFile = await getViteDevServer?.()?.transformRequest(filePath);
        if (!transformedFile?.map) {
            return null;
        }
        return JSON.stringify(transformedFile.map);
    }
    catch {
        return null;
    }
}
async function readNodeModulesMapFileContent(nodeModulesCache, filePath) {
    if (!filePath.startsWith(NODE_MODULES_PATH)) {
        return null;
    }
    const cachedFileContent = nodeModulesCache.get(filePath);
    if (cachedFileContent) {
        return cachedFileContent;
    }
    const fileContent = await readFileContent(filePath);
    if (fileContent) {
        nodeModulesCache.set(filePath, fileContent);
    }
    return fileContent;
}
function getFileMappedPosition(fileContent, line, column) {
    const consumer = new SourceMapConsumer(JSON.parse(fileContent));
    const mappedPosition = consumer.originalPositionFor({ line, column });
    return mappedPosition.source
        ? {
            source: mappedPosition.source,
            line: mappedPosition.line,
            column: mappedPosition.column,
        }
        : null;
}
async function getFallbackMappedPosition(filePath, line, column) {
    const fileContent = await readFileContent(filePath);
    return fileContent ? { source: path.resolve(filePath), line, column } : null;
}

async function getMappedPosition$1(nodeModulesCache, { url, file, line, column }) {
    const pathname = parseURL(url)?.pathname;
    if (!pathname) {
        return null;
    }
    const fileName = pathname.slice(1, pathname.lastIndexOf(file) + file.length);
    const filePath = path.resolve(fileName);
    const [buildMapFileContent, transformedMapFileContent, nodeModulesMapFileContent,] = await Promise.all([
        readBuildFileContent(`${fileName}.map`),
        transformFileToMapContent(filePath),
        readNodeModulesMapFileContent(nodeModulesCache, `${filePath}.map`),
    ]);
    const fileMapContent = buildMapFileContent ||
        transformedMapFileContent ||
        nodeModulesMapFileContent;
    if (!fileMapContent) {
        const buildFileContent = await readBuildFileContent(fileName);
        return buildFileContent
            ? { source: getBuildFilePath(fileName), line, column }
            : getFallbackMappedPosition(filePath, line, column);
    }
    const mappedPosition = getFileMappedPosition(fileMapContent, line, column);
    if (!mappedPosition) {
        return getFallbackMappedPosition(filePath, line, column);
    }
    mappedPosition.source = path.join(path.dirname(buildMapFileContent ? getBuildFilePath(fileName) : filePath), mappedPosition.source);
    const deepNodeModulesMapFileContent = await readNodeModulesMapFileContent(nodeModulesCache, `${mappedPosition.source}.map`);
    if (!deepNodeModulesMapFileContent) {
        return mappedPosition;
    }
    const deepMappedPosition = getFileMappedPosition(deepNodeModulesMapFileContent, mappedPosition.line, mappedPosition.column);
    if (!deepMappedPosition) {
        return mappedPosition;
    }
    deepMappedPosition.source = path.join(path.dirname(mappedPosition.source), deepMappedPosition.source);
    return deepMappedPosition;
}

/* v8 ignore next -- Temp condition while webpack support is maintained */
const getMappedPosition = IS_VITE
    ? getMappedPosition$1
    : getMappedPosition$2;
class SourceMapCache {
    #nodeModulesCache = new Map();
    async getCodePositionPath({ url, file, line, column }) {
        const mappedPosition = await getMappedPosition(this.#nodeModulesCache, {
            url,
            file,
            line: line ? +line : 1,
            column: column ? +column : 0,
        });
        if (!mappedPosition) {
            return null;
        }
        return `${mappedPosition.source}:${mappedPosition.line}:${mappedPosition.column}`;
    }
}
const sourceMapCache = new SourceMapCache();

export { assetsCache, sourceMapCache };
