// @nordic/utils/dist/cjs/error.browser.development.js
'use strict';

require('core-js/modules/es.json.stringify.js');
var React = require('react');
var logger$1 = require('@meli/logger');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const logger = new logger$1.Logger('@nordic/utils');

function safeToString(value) {
  try {
    // Template literals are faster, but fail for Symbol values; String() keeps that fallback safe.
    return String(value);
  } catch {
    // Custom object coercion can still throw, so this is the final safe fallback.
    return 'Unable to stringify value';
  }
}
function stringifyUnknown(value) {
  if (typeof value === 'string') {
    return value;
  }
  // Prefer JSON for plain objects. Runtime can still return undefined for values such as undefined, functions or symbols.
  const result = JSON.stringify(value) || safeToString(value);
  return result;
}
function debugStringifyError(error) {
  try {
    logger.debug('Failed to stringify error.', error instanceof Error ? {
      message: error.message,
      stack: error.stack
    } : {
      message: safeToString(error)
    });
  } catch {
    // Avoid throwing from the error handling code.
  }
}
function stringifyError(error, selector) {
  try {
    if (!(error instanceof Error)) {
      return stringifyUnknown(error);
    }
    if (selector) {
      return stringifyUnknown(typeof selector === 'function' ? selector(error) : error[selector]);
    }
    return error.stack || error.message;
  } catch (err) {
    debugStringifyError(err);
    // Selectors or custom coercion can throw; always return a string for logging/reporting.
    return safeToString(error);
  }
}
class OwnerStackError extends Error {
  ownerStack;
  constructor(message) {
    super(message);
    this.ownerStack = React__namespace.captureOwnerStack() ;
  }
}

exports.OwnerStackError = OwnerStackError;
exports.stringifyError = stringifyError;
