// @nordic/utils/dist/cjs/validation.browser.development.js
'use strict';

/* eslint-disable @typescript-eslint/no-explicit-any */
const EXTERNAL_URL_REGEX = /^https?:|^\/\/|^data:image\//;
const IMG_EXTS_WITH_BLUR_AVAILABLE_REGEX = /\.(jpg|jpeg|png|webp|avif)$/;
function isObject(value) {
  return value !== null && typeof value === 'object' &&
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  value.constructor === Object;
}
function isClient() {
  return typeof window !== 'undefined' &&
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  typeof window.document?.createElement === 'function';
}
function isExternalUrl(value) {
  return EXTERNAL_URL_REGEX.test(value) || value.startsWith('//');
}
function isStorybook() {
  return isClient() && !!document.getElementById('storybook-root');
}
function isTestEnv() {
  if (typeof process === 'undefined') {
    return false;
  }
  const {
    NODE_ENV,
    NODE_TEST
  } = process.env;
  return NODE_ENV === 'test' && NODE_TEST !== 'e2e';
}
function isBlurableImg(src) {
  return !isExternalUrl(src) && IMG_EXTS_WITH_BLUR_AVAILABLE_REGEX.test(src);
}

exports.isBlurableImg = isBlurableImg;
exports.isClient = isClient;
exports.isExternalUrl = isExternalUrl;
exports.isObject = isObject;
exports.isStorybook = isStorybook;
exports.isTestEnv = isTestEnv;
