// @nordic/context/dist/esm/index.production.mjs
import { getDeviceType as getDeviceType$1, getPlatformType as getPlatformType$1, getWebviewInfo as getWebviewInfo$1 } from '@nordic/core-context';
import { getRenderingCsrfToken, getRenderingNonce, getRenderingAssets, getRenderingFlags, getRenderingAppProps } from '@nordic/rendering-context';
import { useMemo } from 'react';
import { freezeObject } from '@nordic/utils/freeze';
import { isExternalUrl } from '@nordic/utils/validation';

/**
 * Returns the prefix for the assets for the current page.
 *
 * **⚠️ WARNING**: This function only works within React components, rendering stage, or client-side.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getassetsprefix | Documentation}
 *
 * @example
 *
 * import { getAssetsPrefix } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const assetsPrefix = getAssetsPrefix();
 *
 *  return (
 *    <h1>
 *      Assets Prefix: {assetsPrefix}
 *    </h1>
 *  );
 * };
 */
function getAssetsPrefix() {
    return getRenderingAssets().prefix;
}
/**
 * Returns the generated CSRF token for the current page.
 *
 * **⚠️ WARNING**: This function only works within React components, rendering stage, or client-side.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getcsrftoken | Documentation}
 *
 * @example
 *
 * import { getCsrfToken } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const csrfToken = getCsrfToken();
 *
 *  return (
 *    <h1>
 *      CSRF Token: {csrfToken}
 *    </h1>
 *  );
 * };
 */
const getCsrfToken = getRenderingCsrfToken;
/**
 * Returns the generated nonce token for the current page.
 *
 * **⚠️ WARNING**: This function only works within React components, rendering stage, or client-side.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getnonce | Documentation}
 *
 * @example
 *
 * import { getNonce } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const nonce = getNonce();
 *
 *  return (
 *    <h1>
 *      Nonce: {nonce}
 *    </h1>
 *  );
 * };
 */
const getNonce = getRenderingNonce;
/**
 * Provides the device type anywhere, including React Components.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getdevicetype | Documentation}
 *
 * @example
 *
 * import { getDeviceType } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const deviceType = getDeviceType();
 *
 *  return (
 *    <h1>
 *      Device Type: {deviceType}
 *    </h1>
 *  );
 * };
 */
const getDeviceType = getDeviceType$1;
/**
 * Provides the platform type anywhere, including React Components.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getplatformtype | Documentation}
 *
 * @example
 *
 * import { getPlatformType } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const platformType = getPlatformType();
 *
 *  return (
 *    <h1>
 *      Platform Type: {platformType}
 *    </h1>
 *  );
 * };
 */
const getPlatformType = getPlatformType$1;
/**
 * Provides the webview information anywhere, including React Components.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getwebviewinfo | Documentation}
 *
 * @example
 *
 * import { getWebviewInfo } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const webviewInfo = getWebviewInfo();
 *
 *  return (
 *    <h1>
 *      Webview Info: {JSON.stringify(webviewInfo)}
 *    </h1>
 *  );
 * };
 */
const getWebviewInfo = getWebviewInfo$1;

/**
 * Returns the props of the current page, defined in the `getServerSideProps()` function.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#usepageprops | Documentation}
 *
 * @example
 *
 * import { usePageProps } from 'nordic/context';
 *
 * const FooComponent = () => {
 *  const { foo } = usePageProps();
 *
 *  return (
 *    <h1>
 *      Foo value: {foo}
 *    </h1>
 *  );
 * };
 */
function usePageProps() {
    getRenderingFlags();
    return (getRenderingAppProps()?.pageProps || {});
}
/**
 * Provides access to the configured assets with prefix.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#useassetsprefix | Documentation}
 *
 * @example
 *
 * import { useAssetsPrefix } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const assetsPrefix = useAssetsPrefix();
 *
 *  return (
 *    <h1>
 *      Asset with prefix: {assetsPrefix.join('some-asset.png')}
 *    </h1>
 *  );
 * };
 */
function useAssetsPrefix() {
    return useMemo(() => {
        const assets = getRenderingAssets();
        return freezeObject({
            value: assets.prefix,
            join: (assetName) => {
                if (isExternalUrl(assetName)) {
                    return assetName;
                }
                const fileName = assets.manifest.get(assetName);
                const fileNameWithPrefix = assets.prefix + (fileName || assetName);
                {
                    return fileNameWithPrefix;
                }
            },
        });
    }, []);
}

/**
 * Returns whether the current context is a webview.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#iswebview | Documentation}
 *
 * @example
 *
 * import { isWebview } from 'nordic/context';
 *
 * const SomeComponent = () => (
 *   <h1>
 *     Is Webview: {isWebview() ? 'Yes' : 'No'}
 *   </h1>
 * );
 */
const isWebview = () => !!getWebviewInfo();

export { getAssetsPrefix, getCsrfToken, getDeviceType, getNonce, getPlatformType, getWebviewInfo, isWebview, useAssetsPrefix, usePageProps };
