// @nordic/core-context/src/server/utils/webview/index.ts
import type { CookieDictionary } from '@nordic/utils/cookie';
import type {
  NordicPlatformType,
  NordicWebkitEngine,
  NordicWebkitInfo,
  NordicWebkitScope,
  NordicWebviewInfo,
} from '../../../common/types';

import { getReq } from '@meli/server-context';
import { parseCookieValue } from '@nordic/utils/cookie';

import {
  CONTEXT_COOKIE,
  LAST_COMMA_REGEX,
} from '../../device/utils/webview/constants';
import { concatCookieValue, setCookie } from './cookie';
import { parseEngineVersion, parseScopeValue } from './parser';

const WEBKIT_CONTEXT_DICTIONARY = {
  engine: { key: 'e', preparser: parseEngineVersion },
  version: { key: 'v' },
  debugMode: { key: 'd', type: 'binary' },
  fitToContent: { key: 'ftc', type: 'binary' },
  insetTop: { key: 'it', type: 'number' },
  insetRight: { key: 'ir', type: 'number' },
  insetBottom: { key: 'ib', type: 'number' },
  insetLeft: { key: 'il', type: 'number' },
  source: { key: 'src' },
  scope: { key: 's', preparser: parseScopeValue },
} as const satisfies CookieDictionary;

const CONTEXT_INFO_REGEX = /(e|v|ftc)=[^,]+,?/g;

function getWebkitInfoFromCookie(cookieValue: string | undefined) {
  if (!cookieValue) {
    return null;
  }

  return parseCookieValue(cookieValue, WEBKIT_CONTEXT_DICTIONARY);
}

type WebviewProfile = NordicWebviewInfo & {
  safeAreaInsets?: {
    top?: number;
    right?: number;
    bottom?: number;
    left?: number;
  };
};

export function getWebviewProfile(
  platformType: NordicPlatformType,
): WebviewProfile | undefined {
  if (platformType === 'web') {
    return undefined;
  }

  const req = getReq();
  const { nativeApp, webkit } = req.device;
  const { engineVersion, nativeVersion } = webkit ?? {};
  const engineValue = engineVersion === '2.0.0' ? 'v2' : 'v1';
  const fitToContentValue = req.header('x-wk-ftc');
  const appInfo = { name: nativeApp!.name, version: nativeApp!.version };
  const webkitInfo = getWebkitInfoFromCookie(req.cookies[CONTEXT_COOKIE]);

  if (!webkitInfo) {
    return {
      nativeApp: appInfo,
      webkit: {
        engine: engineValue,
        version: nativeVersion,
        fitToContent: fitToContentValue === '1',
      },
    };
  }

  const safeAreaInsets = {
    top: webkitInfo.insetTop,
    right: webkitInfo.insetRight,
    bottom: webkitInfo.insetBottom,
    left: webkitInfo.insetLeft,
  };

  if (req.ajax) {
    return {
      safeAreaInsets,
      nativeApp: appInfo,
      webkit: {
        debugMode: webkitInfo.debugMode,
        version: webkitInfo.version,
        engine: webkitInfo.engine as NordicWebkitEngine,
      },
    };
  }

  if (webkitInfo.source !== 'wk') {
    return {
      safeAreaInsets,
      nativeApp: appInfo,
      webkit: {
        debugMode: webkitInfo.debugMode,
        version: nativeVersion,
        fitToContent: fitToContentValue === '1',
        engine: engineValue,
      },
    };
  }

  return {
    safeAreaInsets,
    nativeApp: appInfo,
    webkit: {
      debugMode: webkitInfo.debugMode,
      version: webkitInfo.version,
      fitToContent: webkitInfo.fitToContent,
      engine: webkitInfo.engine as NordicWebkitEngine,
      scope: webkitInfo.scope as NordicWebkitScope,
    },
  };
}

export function setWebkitContextCookie(webkitInfo: NordicWebkitInfo) {
  const currentCookieValue = getReq().cookies[CONTEXT_COOKIE];
  const engineCookiePart = concatCookieValue('e', webkitInfo.engine);
  const versionCookiePart = concatCookieValue('v', webkitInfo.version);
  const fitToContentCookiePart = concatCookieValue(
    'ftc',
    webkitInfo.fitToContent ? '1' : '0',
  );
  const updatedCookieValue = currentCookieValue
    ? currentCookieValue.replace(CONTEXT_INFO_REGEX, (_, key) => {
        switch (key) {
          case 'e':
            return engineCookiePart;
          case 'v':
            return versionCookiePart;

          default:
            return fitToContentCookiePart;
        }
      })
    : engineCookiePart + versionCookiePart + fitToContentCookiePart;

  setCookie(CONTEXT_COOKIE, updatedCookieValue.replace(LAST_COMMA_REGEX, ''));
}
