// @nordic/core-context/src/server/device/utils/webview/constants.ts
import type { CookieDictionary } from '@nordic/utils/cookie';

export const WEBVIEW_USER_AGENT = 'x-webview-user-agent';

export const DC_COOKIE = 'x-nni-dc';

export const CONTEXT_COOKIE = 'nni-ctx';

export const DC_COOKIE_DICTIONARY = {
  back: {
    key: 'cam',
    type: 'binary',
  },
  perfCategory: {
    key: 'clss',
  },
  front: {
    key: 'fcam',
    type: 'binary',
  },
  flash: {
    key: 'flash',
    type: 'binary',
  },
  model: {
    key: 'mdl',
  },
  nfc: {
    key: 'nfc',
    type: 'binary',
  },
  rooted: {
    key: 'root',
    type: 'binary',
  },
  effectiveType: {
    key: 'etype',
  },
  connectionType: {
    key: 'ctype',
  },
  isConnectionSlow: {
    key: 'cslw',
    type: 'binary',
  },
  isConnectionExpensive: {
    key: 'cexp',
    type: 'binary',
  },
  osName: {
    key: 'os',
  },
  osVersion: {
    key: 'osv',
  },
  source: {
    key: 'src',
  },
} as const satisfies CookieDictionary;

export const DEVICE_CAPABILITIES_GROUPS_DICTIONARY: Record<
  string,
  {
    main: 'deviceInfo' | 'networkState';
    subs?: Array<'camera' | 'capabilities' | 'os'>;
    formatter?: (value: string) => string;
  }
> = {
  back: {
    main: 'deviceInfo',
    subs: ['capabilities', 'camera'],
  },
  perfCategory: {
    main: 'deviceInfo',
  },
  front: {
    main: 'deviceInfo',
    subs: ['capabilities', 'camera'],
  },
  flash: {
    main: 'deviceInfo',
    subs: ['capabilities', 'camera'],
  },
  model: {
    main: 'deviceInfo',
    formatter: (value: string) => value.replaceAll('+', ' '),
  },
  nfc: {
    main: 'deviceInfo',
    subs: ['capabilities'],
  },
  rooted: {
    main: 'deviceInfo',
    subs: ['os'],
  },
  effectiveType: {
    main: 'networkState',
  },
  connectionType: {
    main: 'networkState',
  },
  isConnectionSlow: {
    main: 'networkState',
  },
  isConnectionExpensive: {
    main: 'networkState',
  },
};

export const LAST_COMMA_REGEX = /,$/;
