// node-melitk-otel/lib/constants/otel.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_HTTP_CLIENT_REQUEST_DURATION_ATTRIBUTES = exports.DEFAULT_HTTP_SERVER_REQUEST_DURATION_ATTRIBUTES = exports.DEFAULT_HTTP_SERVER_DURATION_ATTRIBUTES = exports.DEFAULT_HTTP_CLIENT_DURATION_ATTRIBUTES = exports.HTTP_CLIENT_REQUEST_DURATION_ALLOW_KEYS_ENV = exports.HTTP_SERVER_REQUEST_DURATION_ALLOW_KEYS_ENV = exports.HTTP_SERVER_DURATION_ALLOW_KEYS_ENV = exports.HTTP_CLIENT_DURATION_ALLOW_KEYS_ENV = exports.EXPORT_TIMEOUT_MILLIS = exports.EXPORT_INTERVAL_MILLIS = exports.EXPONENTIAL_HISTOGRAM_RECORD_MIN_MAX = exports.EXPONENTIAL_HISTOGRAM_MAX_SIZE = exports.HISTOGRAM_BUCKETS_SEC = exports.HISTOGRAM_BUCKETS_MSEC = exports.OTEL_AGENT_PORT = exports.OTEL_AGENT_HOST = exports.VALID_LOG_LEVELS = exports.IGNORED_ROUTES = exports.HTTP_SERVER_DURATION_METRIC_NAME = exports.DEFAULT_EXPONENTIAL_HISTOGRAM_RECORD_MIN_MAX = exports.DEFAULT_EXPONENTIAL_HISTOGRAM_MAX_SIZE = exports.DEFAULT_EXPORT_TIMEOUT_MILLIS = exports.DEFAULT_EXPORT_INTERVAL_MILLIS = exports.MONITORING_PRECISION = exports.DEFAULT_TIMEOUT = exports.DEFAULT_OTEL_AGENT_PORT = exports.DEFAULT_OTEL_AGENT_HOST = void 0;
const semantic_conventions_1 = require("@opentelemetry/semantic-conventions");
const incubating_1 = require("@opentelemetry/semantic-conventions/incubating");
const env = __importStar(require("frontend-env"));
const DEFAULT_OTEL_AGENT_HOST = 'otel-agent';
exports.DEFAULT_OTEL_AGENT_HOST = DEFAULT_OTEL_AGENT_HOST;
const DEFAULT_OTEL_AGENT_PORT = '4317';
exports.DEFAULT_OTEL_AGENT_PORT = DEFAULT_OTEL_AGENT_PORT;
const DEFAULT_TIMEOUT = 100000;
exports.DEFAULT_TIMEOUT = DEFAULT_TIMEOUT;
const MONITORING_PRECISION = 5000;
exports.MONITORING_PRECISION = MONITORING_PRECISION;
const DEFAULT_EXPORT_INTERVAL_MILLIS = 10000;
exports.DEFAULT_EXPORT_INTERVAL_MILLIS = DEFAULT_EXPORT_INTERVAL_MILLIS;
const DEFAULT_EXPORT_TIMEOUT_MILLIS = 9000;
exports.DEFAULT_EXPORT_TIMEOUT_MILLIS = DEFAULT_EXPORT_TIMEOUT_MILLIS;
const DEFAULT_EXPONENTIAL_HISTOGRAM_MAX_SIZE = 120;
exports.DEFAULT_EXPONENTIAL_HISTOGRAM_MAX_SIZE = DEFAULT_EXPONENTIAL_HISTOGRAM_MAX_SIZE;
const DEFAULT_EXPONENTIAL_HISTOGRAM_RECORD_MIN_MAX = true;
exports.DEFAULT_EXPONENTIAL_HISTOGRAM_RECORD_MIN_MAX = DEFAULT_EXPONENTIAL_HISTOGRAM_RECORD_MIN_MAX;
/**
 * Histogram bucket boundaries in milliseconds (for old semconv metrics).
 * @constant
 */
const HISTOGRAM_BUCKETS_MSEC = [
    0, 1, 3, 5, 10, 25, 50, 75, 100, 250, 500, 750, 1000, 2500, 5000, 7500, 10000,
    25000, 50000, 100000,
];
exports.HISTOGRAM_BUCKETS_MSEC = HISTOGRAM_BUCKETS_MSEC;
/**
 * Histogram bucket boundaries in seconds (for new semconv metrics).
 * @constant
 */
const HISTOGRAM_BUCKETS_SEC = [
    0, 0.001, 0.003, 0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1,
    2.5, 5, 7.5, 10, 25, 50, 100,
];
exports.HISTOGRAM_BUCKETS_SEC = HISTOGRAM_BUCKETS_SEC;
const envConfig = env;
const OTEL_AGENT_HOST = envConfig.OTEL_AGENT_HOST || DEFAULT_OTEL_AGENT_HOST;
exports.OTEL_AGENT_HOST = OTEL_AGENT_HOST;
const OTEL_AGENT_PORT = envConfig.OTEL_AGENT_PORT || DEFAULT_OTEL_AGENT_PORT;
exports.OTEL_AGENT_PORT = OTEL_AGENT_PORT;
const EXPONENTIAL_HISTOGRAM_MAX_SIZE = Number(envConfig.MAX_SIZE_EXPONENTIAL_HISTOGRAM) ||
    DEFAULT_EXPONENTIAL_HISTOGRAM_MAX_SIZE;
exports.EXPONENTIAL_HISTOGRAM_MAX_SIZE = EXPONENTIAL_HISTOGRAM_MAX_SIZE;
const EXPONENTIAL_HISTOGRAM_RECORD_MIN_MAX = DEFAULT_EXPONENTIAL_HISTOGRAM_RECORD_MIN_MAX;
exports.EXPONENTIAL_HISTOGRAM_RECORD_MIN_MAX = EXPONENTIAL_HISTOGRAM_RECORD_MIN_MAX;
const EXPORT_INTERVAL_MILLIS = Number(envConfig.OTEL_METRIC_EXPORT_INTERVAL) ||
    DEFAULT_EXPORT_INTERVAL_MILLIS;
exports.EXPORT_INTERVAL_MILLIS = EXPORT_INTERVAL_MILLIS;
const EXPORT_TIMEOUT_MILLIS = Number(envConfig.OTEL_METRIC_EXPORT_TIMEOUT) || DEFAULT_EXPORT_TIMEOUT_MILLIS;
exports.EXPORT_TIMEOUT_MILLIS = EXPORT_TIMEOUT_MILLIS;
/**
 * @deprecated The `http.server.duration` metric is deprecated in favor of `http.server.request.duration`.
 * For now, we keep this metric for backwards compatibility, but it's highly recommended to
 * use `http.server.request.duration` instead.
 */
const HTTP_SERVER_DURATION_METRIC_NAME = 'http.server.duration';
exports.HTTP_SERVER_DURATION_METRIC_NAME = HTTP_SERVER_DURATION_METRIC_NAME;
/**
 * List of specific HTTP routes that always should be ignored independently of parent sampler
 */
const IGNORED_ROUTES = ['/ping'];
exports.IGNORED_ROUTES = IGNORED_ROUTES;
const HTTP_CLIENT_DURATION_ALLOW_KEYS_ENV = 'OTEL_METRIC_HTTP_CLIENT_DURATION_ALLOW_KEYS';
exports.HTTP_CLIENT_DURATION_ALLOW_KEYS_ENV = HTTP_CLIENT_DURATION_ALLOW_KEYS_ENV;
const HTTP_SERVER_DURATION_ALLOW_KEYS_ENV = 'OTEL_METRIC_HTTP_SERVER_DURATION_ALLOW_KEYS';
exports.HTTP_SERVER_DURATION_ALLOW_KEYS_ENV = HTTP_SERVER_DURATION_ALLOW_KEYS_ENV;
const HTTP_SERVER_REQUEST_DURATION_ALLOW_KEYS_ENV = 'OTEL_METRIC_HTTP_SERVER_REQUEST_DURATION_ALLOW_KEYS';
exports.HTTP_SERVER_REQUEST_DURATION_ALLOW_KEYS_ENV = HTTP_SERVER_REQUEST_DURATION_ALLOW_KEYS_ENV;
const HTTP_CLIENT_REQUEST_DURATION_ALLOW_KEYS_ENV = 'OTEL_METRIC_HTTP_CLIENT_REQUEST_DURATION_ALLOW_KEYS';
exports.HTTP_CLIENT_REQUEST_DURATION_ALLOW_KEYS_ENV = HTTP_CLIENT_REQUEST_DURATION_ALLOW_KEYS_ENV;
/**
 * Default attributes for the deprecated `http.client.duration` metric (old semconv, milliseconds).
 */
const DEFAULT_HTTP_CLIENT_DURATION_ATTRIBUTES = [
    incubating_1.ATTR_HTTP_METHOD,
    incubating_1.ATTR_HTTP_STATUS_CODE,
    incubating_1.ATTR_NET_PEER_NAME,
    incubating_1.ATTR_NET_PEER_PORT,
    semantic_conventions_1.ATTR_ERROR_TYPE,
];
exports.DEFAULT_HTTP_CLIENT_DURATION_ATTRIBUTES = DEFAULT_HTTP_CLIENT_DURATION_ATTRIBUTES;
/**
 * Default attributes for the deprecated `http.server.duration` metric (old semconv, milliseconds).
 */
const DEFAULT_HTTP_SERVER_DURATION_ATTRIBUTES = [
    semantic_conventions_1.ATTR_HTTP_ROUTE,
    incubating_1.ATTR_HTTP_SCHEME,
    incubating_1.ATTR_HTTP_METHOD,
    incubating_1.ATTR_HTTP_STATUS_CODE,
    incubating_1.ATTR_HTTP_FLAVOR,
    incubating_1.ATTR_NET_HOST_NAME,
    incubating_1.ATTR_NET_HOST_PORT,
    semantic_conventions_1.ATTR_ERROR_TYPE,
];
exports.DEFAULT_HTTP_SERVER_DURATION_ATTRIBUTES = DEFAULT_HTTP_SERVER_DURATION_ATTRIBUTES;
/**
 * Default attributes for the `http.server.request.duration` metric (new semconv, seconds).
 * @see https://opentelemetry.io/docs/specs/semconv/http/http-metrics/#metric-httpserverrequestduration
 */
const DEFAULT_HTTP_SERVER_REQUEST_DURATION_ATTRIBUTES = [
    semantic_conventions_1.ATTR_HTTP_ROUTE,
    semantic_conventions_1.ATTR_HTTP_REQUEST_METHOD,
    semantic_conventions_1.ATTR_HTTP_RESPONSE_STATUS_CODE,
    semantic_conventions_1.ATTR_ERROR_TYPE,
    semantic_conventions_1.ATTR_NETWORK_PROTOCOL_NAME,
    semantic_conventions_1.ATTR_NETWORK_PROTOCOL_VERSION,
    semantic_conventions_1.ATTR_URL_SCHEME,
    semantic_conventions_1.ATTR_SERVER_ADDRESS,
    semantic_conventions_1.ATTR_SERVER_PORT,
];
exports.DEFAULT_HTTP_SERVER_REQUEST_DURATION_ATTRIBUTES = DEFAULT_HTTP_SERVER_REQUEST_DURATION_ATTRIBUTES;
/**
 * Default attributes for the `http.client.request.duration` metric (new semconv, seconds).
 */
const DEFAULT_HTTP_CLIENT_REQUEST_DURATION_ATTRIBUTES = [
    semantic_conventions_1.ATTR_HTTP_REQUEST_METHOD,
    semantic_conventions_1.ATTR_HTTP_RESPONSE_STATUS_CODE,
    semantic_conventions_1.ATTR_ERROR_TYPE,
    semantic_conventions_1.ATTR_NETWORK_PROTOCOL_NAME,
    semantic_conventions_1.ATTR_NETWORK_PROTOCOL_VERSION,
    semantic_conventions_1.ATTR_URL_SCHEME,
    semantic_conventions_1.ATTR_SERVER_ADDRESS,
    semantic_conventions_1.ATTR_SERVER_PORT,
];
exports.DEFAULT_HTTP_CLIENT_REQUEST_DURATION_ATTRIBUTES = DEFAULT_HTTP_CLIENT_REQUEST_DURATION_ATTRIBUTES;
const VALID_LOG_LEVELS = [
    'NONE',
    'ERROR',
    'WARN',
    'INFO',
    'DEBUG',
    'VERBOSE',
];
exports.VALID_LOG_LEVELS = VALID_LOG_LEVELS;
