// node-melitk-otel/lib/events/activator/client/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivatorClient = void 0;
const frontend_restclient_1 = require("frontend-restclient");
class ActivatorClient {
    config;
    restclient;
    constructor(config) {
        this.config = config;
        this.restclient = (0, frontend_restclient_1.RestClient)({
            baseURL: this.config.url + this.config.servicesEndpoint,
            timeout: +this.config.timeout,
            retry: {
                maxRetries: +this.config.maxRetries,
            },
            headers: {
                'x-flow-starter': 'true',
            },
        });
    }
    /**
     * Checks the service status.
     * @param {ActivatorConfig} config The {@link ActivatorConfig} for the activator.
     * @param {string} service A {@link string} that represents the name of the service. Should be composed by Fury's scope and the service name.
     * @returns {Promise<ActivatorClientResponse>} A {@link Promise} that resolves to the {@link ActivatorClientResponse}.
     */
    async checkService(service) {
        const request = await this.restclient
            .get(service, {
            context: (0, frontend_restclient_1.buildFlowStarterContext)(),
        })
            .then((response) => response.data);
        return request;
    }
}
exports.ActivatorClient = ActivatorClient;
