// @meli/logger/dist/browser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const LEVEL_VALUES = {
    trace: 10,
    debug: 20,
    info: 30,
    warn: 40,
    error: 50,
    fatal: 60,
};
const CONSOLE_DICTIONARY = {
    trace: 'debug',
    debug: 'debug',
    info: 'info',
    warn: 'warn',
    error: 'error',
    fatal: 'error',
};
function LoggerImpl(name, options = {}) {
    this.name = name;
    this.level = options.level || 'info';
}
function log(level, msg, tags) {
    if (this.isLevelEnabled(level)) {
        // eslint-disable-next-line no-console
        console[CONSOLE_DICTIONARY[level]](msg, { ...tags, name: this.name });
    }
}
Object.assign(LoggerImpl.prototype, {
    isLevelEnabled(level) {
        return LEVEL_VALUES[level] >= LEVEL_VALUES[this.level];
    },
    trace(msg, tags) {
        log.call(this, 'trace', msg, tags);
    },
    debug(msg, tags) {
        log.call(this, 'debug', msg, tags);
    },
    info(msg, tags) {
        log.call(this, 'info', msg, tags);
    },
    warn(msg, tags) {
        log.call(this, 'warn', msg, tags);
    },
    error(msg, tags) {
        log.call(this, 'error', msg, tags);
    },
    fatal(msg, tags) {
        log.call(this, 'fatal', msg, tags);
    },
});
exports.Logger = LoggerImpl;
