// @meli/logger/dist/Logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const pino_1 = __importDefault(require("pino"));
const frontend_env_1 = __importDefault(require("frontend-env"));
const levels_1 = require("./internal/levels");
const logger_error_1 = require("./logger-error");
const context_1 = require("./context");
const formatters_1 = require("./internal/formatters");
const plain_text_stream_1 = require("./internal/plain-text-stream");
const pretty_1 = require("./internal/pretty");
const PINO_BASE_CONFIG = {
    formatters: {
        level: formatters_1.formatLevel,
        bindings: formatters_1.formatBindings,
    },
    mixin: context_1.contextProvider,
    timestamp: formatters_1.formatTimestamp,
};
// Cache root pino instances by format to avoid spawning duplicate transports/streams.
// Each Logger creates a cheap .child() from the cached root.
const rootInstances = new Map();
const canUsePretty = !frontend_env_1.default.FURY && (0, pretty_1.isPinoPrettyAvailable)();
function getRootLogger(format) {
    const cacheKey = canUsePretty ? 'pretty' : format;
    const currentRoot = rootInstances.get(cacheKey);
    if (currentRoot) {
        return currentRoot;
    }
    let destinationStream;
    if (cacheKey === 'text') {
        destinationStream = (0, plain_text_stream_1.createPlainTextStream)();
    }
    else if (cacheKey === 'pretty') {
        destinationStream = pino_1.default.transport({
            target: 'pino-pretty',
            options: pretty_1.PINO_PRETTY_OPTIONS,
        });
    }
    const root = (0, pino_1.default)(PINO_BASE_CONFIG, destinationStream);
    rootInstances.set(cacheKey, root);
    return root;
}
class Logger {
    #client;
    constructor(name, options = {}) {
        this.#client = getRootLogger(options.format ?? 'text').child({ name });
        this.#applyLevel((0, levels_1.getInitialLevel)(name, options.level));
    }
    get level() {
        return this.#client.level;
    }
    set level(level) {
        const logLevelEnv = frontend_env_1.default.LOG_LEVEL;
        if (!logLevelEnv || !(logLevelEnv in this.#client.levels.values)) {
            this.#applyLevel(level);
            return;
        }
        const { values } = this.#client.levels;
        if (values[level] >= values[logLevelEnv]) {
            this.#applyLevel(level);
            return;
        }
        if (!frontend_env_1.default.PRODUCTION) {
            throw new logger_error_1.LoggerError(`Cannot set level to '${level}'. LOG_LEVEL='${logLevelEnv}' enforces a minimum severity — only '${logLevelEnv}' and above are allowed.`);
        }
        this.#applyLevel(logLevelEnv);
    }
    isLevelEnabled(level) {
        return this.#client.isLevelEnabled(level);
    }
    trace(message, tags) {
        this.#log('trace', message, tags);
    }
    debug(message, tags) {
        this.#log('debug', message, tags);
    }
    info(message, tags) {
        this.#log('info', message, tags);
    }
    warn(message, tags) {
        this.#log('warn', message, tags);
    }
    error(message, tags) {
        this.#log('error', message, tags);
    }
    fatal(message, tags) {
        this.#log('fatal', message, tags);
    }
    #applyLevel(level) {
        try {
            this.#client.level = level;
        }
        catch (err) {
            throw new logger_error_1.LoggerError(`Invalid log level: '${level}'. Valid levels: trace, debug, info, warn, error, fatal.`, { cause: err });
        }
    }
    #log(method, message, tags = {}) {
        this.#client[method](tags, message);
    }
}
exports.Logger = Logger;
