// @andes/image/build/cjs/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var React = require('react');
var PropTypes = require('prop-types');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
const WARNING_ALT_ATTRIBUTE = "title attribute is used to indicate the image's alt attribute, it's required for accessibility purposes";

/** @type {typeof import('../..').Image} */
const Image = _ref => {
  let className = _ref.className,
    width = _ref.width,
    height = _ref.height,
    src = _ref.src,
    src2x = _ref.src2x,
    title = _ref.title;
  return (/*#__PURE__*/
    // eslint-disable-next-line nordic/no-img-tags
    React__default["default"].createElement("img", {
      className: className,
      width: width,
      height: height,
      src: src,
      alt: title,
      srcSet: src2x && `${src} 1x, ${src2x} 2x`
    })
  );
};

/** @type {import('prop-types').Validator<string>} */
const checkTitleProp = (_ref2, propName, componentName) => {
  let title = _ref2.title;
  PropTypes__default["default"].checkPropTypes({
    title: PropTypes__default["default"].string
  }, {
    title
  }, propName, componentName);
  return typeof title === 'undefined' ? new Error(WARNING_ALT_ATTRIBUTE) : null;
};
Image.propTypes = {
  title: checkTitleProp,
  src: PropTypes__default["default"].string,
  src2x: PropTypes__default["default"].string,
  width: PropTypes__default["default"].string,
  height: PropTypes__default["default"].string,
  className: PropTypes__default["default"].string
};
var Image$1 = Image;
exports.Image = Image$1;
