// @nordic/assets-config/dist/esm/index.production.mjs
import path from 'node:path';
import { getReq } from '@meli/server-context';
import { config } from 'frontend-config';
import { env } from '@nordic/env';

const FALLBACK_ASSETS_CDN = 'https://http2.mlstatic.com';
function getAssetsCDN() {
    try {
        const req = getReq();
        const { id = 'ML', siteId = 'MLA' } = req.platform || {};
        return (config.get('url.assets.link', id, siteId) || FALLBACK_ASSETS_CDN);
    }
    catch {
        return FALLBACK_ASSETS_CDN;
    }
}

function getAssetsPrefix() {
    // Do not use false here because while generating development assets the
    // bundler will resolve this module from the production entry, leaving the CDN
    // prefix inlined into dev assets. The runtime env is the source of truth for
    // deciding which prefix those generated assets must use.
    if (!env.PRODUCTION) {
        return env.ASSETS_BASE_PATH || '/';
    }
    return (config.assets?.prefix ||
        `${getAssetsCDN()}/frontend-assets/${env.APPLICATION}/`);
}

const NORDIC_DIR = path.resolve('.nordic');
const METADATA_DIR = path.join(NORDIC_DIR, '_metadata');
const ASSETS_CONFIG = {
    PUBLIC_DIR: path.resolve('public'),
    CLIENT_OUTPUT_DIR: path.resolve('build'),
    SERVER_OUTPUT_DIR: path.join(NORDIC_DIR, 'build'),
    METADATA_DIR,
    CLIENT_STATS_PATH: path.join(METADATA_DIR, 'client-stats.json'),
    SERVER_STATS_PATH: path.join(METADATA_DIR, 'server-stats.json'),
    MANIFEST_PATH: path.join(METADATA_DIR, 'manifest.json'),
    IMAGES_BLUR_DIR: path.join(NORDIC_DIR, 'images'),
    IMAGES_BLUR_PREFIX: 'blur-',
    MANIFEST_API_ROUTER_KEY: '_nordic_api-router.js',
    MANIFEST_APP_ROUTER_KEY: '_nordic_app-router.js',
    PREFIX: '_nordic/assets/',
};

export { ASSETS_CONFIG, getAssetsCDN, getAssetsPrefix };
