// frontend-restclient/src/helpers/ssrf-validations.js
const { AxiosError } = require('axios');
const env = require('frontend-env');

const {
  PATH_TRAVERSAL,
  ABSOLUTE_URL_NOT_ALLOWED,
  PARAMETER_POLLUTION,
} = require('../constants/errorMessages');
const {
  SSRF_PATH_TRAVERSAL,
  SSRF_ALLOW_ABSOLUTE_URLS,
  SSRF_PARAMETER_POLLUTION,
} = require('../constants/metrics');
const { statsd } = require('../utils/statsd');
const { logger } = require('../logger');
const { version } = require('../../package.json');

/**
 * @typedef {import('../../types/index').RestClientConfig} ClientConfig
 * @typedef {import('../../types/index').NordicRequestConfig} RequestConfig
 * @typedef {ClientConfig & RequestConfig} MergedConfig
 */

/**
 * Regular expression to detect path traversal sequences in URLs.
 * It matches a leading slash, backslash, or their URL-encoded forms (%2f, %5c),
 * followed by exactly two dots (either a literal '.' or '%2e'),
 * and concludes with a slash, backslash (or their URL-encoded equivalents) or the end of the string.
 * The case-insensitive flag ensures that encoded characters are recognized regardless of case.
 */
const PATH_TRAVERSAL_REGEX =
  /(?:\/|\\|%2f|%5c)(?:\.|%2e){2}(?:\/|\\|%2f|%5c|$)/i;

/**
 * Regular expression to detect absolute URLs.
 * It matches a protocol (optional) followed by a double slash.
 * The case-insensitive flag ensures that encoded characters are recognized regardless of case.
 */
// eslint-disable-next-line security/detect-unsafe-regex
const ABSOLUTE_URL_REGEX = /^([a-z][a-z\d+\-.]*:)?\/\//i;

/**
 * Regular expression to detect white spaces and control characters (0x00-0x1f).
 */
// eslint-disable-next-line no-control-regex
const WHITESPACE_AND_CONTROL_CHARS_REGEX = /[\s\x00-\x1f]/g;

/**
 * @param {string} url
 * @param {MergedConfig} config
 */
function checkPathTraversal(url, config) {
  const cleanUrl = url
    .split('?')[0]
    .split('#')[0]
    .replace(WHITESPACE_AND_CONTROL_CHARS_REGEX, '');

  if (PATH_TRAVERSAL_REGEX.test(cleanUrl)) {
    const message = `${PATH_TRAVERSAL}: ${url}`;

    statsd.increment(SSRF_PATH_TRAVERSAL, {
      technology: 'node',
      sdk_version: version,
    });

    logger.warn(message, { url });

    throw new AxiosError(message, AxiosError.ERR_BAD_REQUEST, config);
  }
}

/**
 * @param {MergedConfig} config
 */
function checkAbsoluteURL(config) {
  const { url, allowAbsoluteUrls } = config;

  if (!ABSOLUTE_URL_REGEX.test(url)) {
    return;
  }

  statsd.increment(SSRF_ALLOW_ABSOLUTE_URLS, {
    technology: 'node',
    action: allowAbsoluteUrls ? 'warn' : 'block',
    sdk_version: version,
  });

  if (allowAbsoluteUrls) {
    if (env.DEVELOPMENT) {
      logger.warn(
        '[restclient] "allowAbsoluteUrls" is deprecated and will be removed at the end of Q2R1 2026. Please configure a "baseURL" and use relative paths instead.',
      );
    }

    return;
  }

  throw new AxiosError(
    `${ABSOLUTE_URL_NOT_ALLOWED} Blocked URL: ${url}`,
    AxiosError.ERR_BAD_REQUEST,
    config,
  );
}

/**
 * @param {string} url
 * @param {MergedConfig} config
 */
function checkParameterPollution(url, config) {
  if (config.allowRepeatedParams) {
    return;
  }

  const queryString = url.split('?')[1];

  if (!queryString) {
    return;
  }

  const params = new URLSearchParams(queryString);

  const valueMap = new Map();
  const pollutedParams = new Set();

  params.forEach((value, rawName) => {
    const name = rawName.replace(WHITESPACE_AND_CONTROL_CHARS_REGEX, '');

    if (valueMap.has(name)) {
      valueMap.get(name).push(value);
      pollutedParams.add(name);
    } else {
      valueMap.set(name, [value]);
    }
  });

  if (pollutedParams.size > 0) {
    const message = `${PARAMETER_POLLUTION} "${url}"`;

    statsd.increment(SSRF_PARAMETER_POLLUTION, {
      technology: 'node',
      action: 'block',
      sdk_version: version,
    });

    logger.error('Parameter Pollution detected in URL query parameters.', {
      url,
      polluted_parameters: Array.from(pollutedParams),
    });

    throw new AxiosError(message, AxiosError.ERR_BAD_REQUEST, config);
  }
}

/**
 * Validate security-related parameters in the request.
 *
 * @param {string} url Full URL
 * @param {MergedConfig} mergedConfig
 */
function securityValidations(url, mergedConfig) {
  checkPathTraversal(url, mergedConfig);
  checkAbsoluteURL(mergedConfig);
  checkParameterPollution(url, mergedConfig);
}

module.exports = {
  securityValidations,
};
