// @nordic/utils/dist/esm/cookie.browser.development.mjs
import 'core-js/modules/es.iterator.constructor.js';
import 'core-js/modules/es.iterator.reduce.js';
import { filterXSS } from 'xss';

const FILTER_OPTIONS = {
  whiteList: {},
  stripIgnoreTag: true,
  stripIgnoreTagBody: ['*']
};
/**
 * Sanitizes the value received removing XSS and truncating at first quote
 */
function sanitizeValue(value) {
  const sanitized = filterXSS(value, FILTER_OPTIONS);
  if (!sanitized) {
    return '';
  }
  const singleQuoteIndex = sanitized.indexOf("'");
  const doubleQuoteIndex = sanitized.indexOf('"');
  if (singleQuoteIndex < 0 && doubleQuoteIndex < 0) {
    return sanitized;
  }
  let firstQuoteIndex = singleQuoteIndex < 0 ? doubleQuoteIndex : singleQuoteIndex;
  if (doubleQuoteIndex !== -1 && singleQuoteIndex !== doubleQuoteIndex) {
    firstQuoteIndex = Math.min(firstQuoteIndex, doubleQuoteIndex);
  }
  return sanitized.substring(0, firstQuoteIndex) || '';
}

function parseValue(value) {
  let type = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'string';
  switch (type) {
    case 'binary':
      return value === '1';
    case 'number':
      return +value;
    case 'string':
      return value;
    default:
      return undefined;
  }
}
/**
 * Parses a cookie value into a typed object based on a configuration dictionary.
 *
 * @example
 *
 * const result = parseCookieValue('v=1.0.0,m=1,mc=Samsung",user=Foo_Ipsum', {
 *   version: { key: 'v', type: 'number', parser: false },
 *   mobile: { key: 'm', type: 'binary' },
 *   user: { key: 'user', parser: (value) => value.replaceAll('_', ' ') },
 *   marca: { key: 'mc' },
 * } as const);
 *
 * // result = {
 * //   version: 1,
 * //   marca: 'Samsung',
 * //   mobile: true,
 * //   user: 'Foo Ipsum'
 * // }
 */
function parseCookieValue(cookieValue, dictionary) {
  const invertedDictionary = Object.entries(dictionary).reduce((acc, _ref) => {
    let [key, value] = _ref;
    return Object.assign(acc, {
      [value.key]: {
        key,
        type: value.type,
        preparser: value.preparser
      }
    });
  }, {});
  return cookieValue.split(',').reduce((acc, pair) => {
    const [key, value] = pair.split('=');
    if (!key || !value) {
      return acc;
    }
    const entry = invertedDictionary[key];
    if (!entry) {
      return acc;
    }
    const {
      key: entryKey,
      type,
      preparser = true
    } = entry;
    const parsedValue = typeof preparser === 'function' ? preparser(value) : parseValue(preparser ? sanitizeValue(value) : value, type);
    if (typeof parsedValue !== 'undefined') {
      acc[entryKey] = parsedValue;
    }
    return acc;
  }, {});
}

export { parseCookieValue };
