// @nordic/utils/dist/esm/error.browser.production.mjs
import 'core-js/modules/es.json.stringify.js';
import { Logger } from '@meli/logger';

const logger = new Logger('@nordic/utils');

function safeToString(value) {
  try {
    // Template literals are faster, but fail for Symbol values; String() keeps that fallback safe.
    return String(value);
  } catch {
    // Custom object coercion can still throw, so this is the final safe fallback.
    return 'Unable to stringify value';
  }
}
function stringifyUnknown(value) {
  if (typeof value === 'string') {
    return value;
  }
  // Prefer JSON for plain objects. Runtime can still return undefined for values such as undefined, functions or symbols.
  const result = JSON.stringify(value) || safeToString(value);
  return result;
}
function debugStringifyError(error) {
  try {
    logger.debug('Failed to stringify error.', error instanceof Error ? {
      message: error.message,
      stack: error.stack
    } : {
      message: safeToString(error)
    });
  } catch {
    // Avoid throwing from the error handling code.
  }
}
function stringifyError(error, selector) {
  try {
    if (!(error instanceof Error)) {
      return stringifyUnknown(error);
    }
    if (selector) {
      return stringifyUnknown(typeof selector === 'function' ? selector(error) : error[selector]);
    }
    return error.stack || error.message;
  } catch (err) {
    debugStringifyError(err);
    // Selectors or custom coercion can throw; always return a string for logging/reporting.
    return safeToString(error);
  }
}
class OwnerStackError extends Error {
  ownerStack;
  constructor(message) {
    super(message);
    this.ownerStack = null;
  }
}

export { OwnerStackError, stringifyError };
