// @andes/common/build/esm/constants/countries.development.js
const COUNTRY_AR = 'AR'; // Argentina
const COUNTRY_BO = 'BO'; // Bolivia
const COUNTRY_BR = 'BR'; // Brasil
const COUNTRY_CL = 'CL'; // Chile
const COUNTRY_CO = 'CO'; // Colombia
const COUNTRY_CR = 'CR'; // Costa Rica
const COUNTRY_EC = 'EC'; // Ecuador
const COUNTRY_SV = 'SV'; // El Salvador
const COUNTRY_GT = 'GT'; // Guatemala
const COUNTRY_HN = 'HN'; // Honduras
const COUNTRY_MX = 'MX'; // Mexico
const COUNTRY_NI = 'NI'; // Nicaragua
const COUNTRY_PA = 'PA'; // Panama
const COUNTRY_PY = 'PY'; // Paraguay
const COUNTRY_PE = 'PE'; // Perú
const COUNTRY_DO = 'DO'; // República Dominicana
const COUNTRY_UY = 'UY'; // Uruguay
const COUNTRY_VE = 'VE'; // Venezuela

const AVAILABLE_COUNTRIES = [COUNTRY_AR, COUNTRY_BO, COUNTRY_BR, COUNTRY_CL, COUNTRY_CO, COUNTRY_CR, COUNTRY_EC, COUNTRY_SV, COUNTRY_GT, COUNTRY_HN, COUNTRY_MX, COUNTRY_NI, COUNTRY_PA, COUNTRY_PY, COUNTRY_PE, COUNTRY_DO, COUNTRY_UY, COUNTRY_VE];
export { AVAILABLE_COUNTRIES, COUNTRY_AR, COUNTRY_BO, COUNTRY_BR, COUNTRY_CL, COUNTRY_CO, COUNTRY_CR, COUNTRY_DO, COUNTRY_EC, COUNTRY_GT, COUNTRY_HN, COUNTRY_MX, COUNTRY_NI, COUNTRY_PA, COUNTRY_PE, COUNTRY_PY, COUNTRY_SV, COUNTRY_UY, COUNTRY_VE };
