// @andes/common/build/esm/constants/currencies.development.js
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
const CURRENCY_ARS = 'ARS'; // Peso Argentino
const CURRENCY_BOB = 'BOB'; // Boliviano
const CURRENCY_BRL = 'BRL'; // Real
const CURRENCY_CLF = 'CLF'; // Unidad de Fomento
const CURRENCY_CLP = 'CLP'; // Peso Chileno
const CURRENCY_COP = 'COP'; // Peso Colombiano
const CURRENCY_CRC = 'CRC'; // Colon
const CURRENCY_CUC = 'CUC'; // Peso Cubano Convertible
const CURRENCY_CNY = 'CNY'; // Yuan Chino
const CURRENCY_CUP = 'CUP'; // Peso Cubano
const CURRENCY_DOP = 'DOP'; // Peso Dominicano
const CURRENCY_EUR = 'EUR'; // Euro
const CURRENCY_GTQ = 'GTQ'; // Quetzal
const CURRENCY_HNL = 'HNL'; // Lempira
const CURRENCY_MXN = 'MXN'; // Peso Mexicano
const CURRENCY_NIO = 'NIO'; // Córdoba
const CURRENCY_PAB = 'PAB'; // Balboa
const CURRENCY_PEN = 'PEN'; // Sol
const CURRENCY_PYG = 'PYG'; // Guarani
const CURRENCY_USD = 'USD'; // Dólar
const CURRENCY_UYU = 'UYU'; // Peso Uruguayo
const CURRENCY_VEF = 'VEF'; // Bolivar fuerte
const CURRENCY_VES = 'VES'; // Bolivar soberano
const CURRENCY_DCE = 'DCE'; // DCE
// Crypto
const CURRENCY_BTC = 'BTC'; // Bitcoin
const CURRENCY_ETH = 'ETH'; // Ethereum
const CURRENCY_USDC = 'USDC'; // USD Coin
const CURRENCY_USDP = 'USDP'; // Pax Dollar
const CURRENCY_MCN = 'MCN'; // Mercado Coin
const CURRENCY_MUSD = 'MUSD'; // Meli Dollar
const CURRENCY_LINK = 'LINK'; // Chainlink
const CURRENCY_UNI = 'UNI'; // Uniswap
const CURRENCY_LTC = 'LTC'; // Litecoin
const CURRENCY_XRP = 'XRP'; // XRP
const CURRENCY_SOL = 'SOL'; // Solana
const CURRENCY_DOT = 'DOT'; // Polkadot
const CURRENCY_AAVE = 'AAVE'; // Aave
const CURRENCY_SUI = 'SUI'; // Sui
const CURRENCY_ADA = 'ADA'; // Cardano
const CURRENCY_AVAX = 'AVAX'; // Avalanche
const CURRENCY_NEAR = 'NEAR'; // NEAR Protocol
const CURRENCY_HYPE = 'HYPE'; // Hyperliquid
const CURRENCY_XLM = 'XLM'; // Stellar
const CURRENCY_HBAR = 'HBAR'; // Hedera
const CURRENCY_PAXG = 'PAXG'; // PAX Gold

const DEFAULT_DECIMALS_AMOUNT = 2;
const DEFAULT_CURRENCY_SYMBOL = '$';
const DEFAULT_CURRENCY_SYMBOL_TEXT = 'pesos';
const CURRENCY_SYMBOLS = ['$', 'Bs', 'R$', 'UF', '₡', 'CUC', '€', 'Q', 'L', 'C$', 'B/.', 'S/', '₲', 'US$', 'Bs.', '¥'];
const CRYPTO_CURRENCY_IDS = {
  CURRENCY_BTC,
  CURRENCY_ETH,
  CURRENCY_USDP,
  CURRENCY_MCN,
  CURRENCY_LINK,
  CURRENCY_UNI,
  CURRENCY_LTC,
  CURRENCY_USDC,
  CURRENCY_MUSD,
  CURRENCY_XRP,
  CURRENCY_SOL,
  CURRENCY_DOT,
  CURRENCY_AAVE,
  CURRENCY_SUI,
  CURRENCY_ADA,
  CURRENCY_AVAX,
  CURRENCY_NEAR,
  CURRENCY_HYPE,
  CURRENCY_XLM,
  CURRENCY_HBAR,
  CURRENCY_PAXG
};
const DEFAULT_CURRENCY_IDS = {
  CURRENCY_ARS,
  CURRENCY_BOB,
  CURRENCY_BRL,
  CURRENCY_CLF,
  CURRENCY_CLP,
  CURRENCY_CNY,
  CURRENCY_COP,
  CURRENCY_CRC,
  CURRENCY_CUC,
  CURRENCY_CUP,
  CURRENCY_DOP,
  CURRENCY_EUR,
  CURRENCY_GTQ,
  CURRENCY_HNL,
  CURRENCY_MXN,
  CURRENCY_NIO,
  CURRENCY_PAB,
  CURRENCY_PEN,
  CURRENCY_PYG,
  CURRENCY_USD,
  CURRENCY_UYU,
  CURRENCY_VEF,
  CURRENCY_VES,
  CURRENCY_DCE
};
const AVAILABLE_CURRENCIES = _objectSpread(_objectSpread({}, DEFAULT_CURRENCY_IDS), CRYPTO_CURRENCY_IDS);
const DECIMALS_BY_CURRENCY_ID = {
  [CURRENCY_ARS]: 2,
  [CURRENCY_BOB]: 2,
  [CURRENCY_BRL]: 2,
  [CURRENCY_CLF]: 0,
  [CURRENCY_CLP]: 0,
  [CURRENCY_CNY]: 2,
  [CURRENCY_COP]: 0,
  [CURRENCY_CRC]: 2,
  [CURRENCY_CUC]: 2,
  [CURRENCY_CUP]: 2,
  [CURRENCY_DOP]: 2,
  [CURRENCY_EUR]: 2,
  [CURRENCY_GTQ]: 2,
  [CURRENCY_HNL]: 0,
  [CURRENCY_MXN]: 2,
  [CURRENCY_NIO]: 0,
  [CURRENCY_PAB]: 2,
  [CURRENCY_PEN]: 2,
  [CURRENCY_PYG]: 0,
  [CURRENCY_USD]: 2,
  [CURRENCY_UYU]: 2,
  [CURRENCY_VEF]: 2,
  [CURRENCY_VES]: 2,
  [CURRENCY_DCE]: 2,
  [CURRENCY_BTC]: 8,
  [CURRENCY_ETH]: 8,
  [CURRENCY_USDP]: 8,
  [CURRENCY_MCN]: 8,
  [CURRENCY_LINK]: 8,
  [CURRENCY_UNI]: 8,
  [CURRENCY_LTC]: 8,
  [CURRENCY_USDC]: 8,
  [CURRENCY_MUSD]: 8,
  [CURRENCY_XRP]: 8,
  [CURRENCY_SOL]: 8,
  [CURRENCY_DOT]: 8,
  [CURRENCY_AAVE]: 8,
  [CURRENCY_SUI]: 8,
  [CURRENCY_ADA]: 8,
  [CURRENCY_AVAX]: 8,
  [CURRENCY_NEAR]: 8,
  [CURRENCY_HYPE]: 8,
  [CURRENCY_XLM]: 8,
  [CURRENCY_HBAR]: 8,
  [CURRENCY_PAXG]: 8
};
const CURRENCY_SYMBOLS_BY_ID = {
  [CURRENCY_ARS]: '$',
  [CURRENCY_BOB]: 'Bs',
  [CURRENCY_BRL]: 'R$',
  [CURRENCY_CLF]: 'UF',
  [CURRENCY_CLP]: '$',
  [CURRENCY_CNY]: '¥',
  [CURRENCY_COP]: '$',
  [CURRENCY_CRC]: '₡',
  [CURRENCY_CUC]: 'CUC',
  [CURRENCY_CUP]: '$',
  [CURRENCY_DOP]: '$',
  [CURRENCY_EUR]: '€',
  [CURRENCY_GTQ]: 'Q',
  [CURRENCY_HNL]: 'L',
  [CURRENCY_MXN]: '$',
  [CURRENCY_NIO]: 'C$',
  [CURRENCY_PAB]: 'B/.',
  [CURRENCY_PEN]: 'S/',
  [CURRENCY_PYG]: '₲',
  [CURRENCY_USD]: 'US$',
  [CURRENCY_UYU]: '$',
  [CURRENCY_VEF]: 'Bs.',
  [CURRENCY_VES]: 'Bs.',
  [CURRENCY_DCE]: 'US$',
  [CURRENCY_BTC]: 'BTC',
  [CURRENCY_ETH]: 'ETH',
  [CURRENCY_USDP]: 'USDP',
  [CURRENCY_MCN]: 'MCN',
  [CURRENCY_LINK]: 'LINK',
  [CURRENCY_UNI]: 'UNI',
  [CURRENCY_LTC]: 'LTC',
  [CURRENCY_USDC]: 'USDC',
  [CURRENCY_MUSD]: 'MUSD',
  [CURRENCY_XRP]: 'XRP',
  [CURRENCY_SOL]: 'SOL',
  [CURRENCY_DOT]: 'DOT',
  [CURRENCY_AAVE]: 'AAVE',
  [CURRENCY_SUI]: 'SUI',
  [CURRENCY_ADA]: 'ADA',
  [CURRENCY_AVAX]: 'AVAX',
  [CURRENCY_NEAR]: 'NEAR',
  [CURRENCY_HYPE]: 'HYPE',
  [CURRENCY_XLM]: 'XLM',
  [CURRENCY_HBAR]: 'HBAR',
  [CURRENCY_PAXG]: 'PAXG'
};
const CURRENCY_TEXT_BY_ID = {
  [CURRENCY_ARS]: DEFAULT_CURRENCY_SYMBOL_TEXT,
  [CURRENCY_BOB]: 'bolivianos',
  [CURRENCY_BRL]: 'reales',
  [CURRENCY_CLF]: 'unidades de fomento',
  [CURRENCY_CLP]: DEFAULT_CURRENCY_SYMBOL_TEXT,
  [CURRENCY_CNY]: 'yuanes',
  [CURRENCY_COP]: DEFAULT_CURRENCY_SYMBOL_TEXT,
  [CURRENCY_CRC]: 'colones',
  [CURRENCY_CUC]: DEFAULT_CURRENCY_SYMBOL_TEXT,
  [CURRENCY_CUP]: DEFAULT_CURRENCY_SYMBOL_TEXT,
  [CURRENCY_DOP]: DEFAULT_CURRENCY_SYMBOL_TEXT,
  [CURRENCY_EUR]: 'euros',
  [CURRENCY_GTQ]: 'quetzales guatemaltecos',
  [CURRENCY_HNL]: 'lempiras',
  [CURRENCY_MXN]: DEFAULT_CURRENCY_SYMBOL_TEXT,
  [CURRENCY_NIO]: 'córdobas',
  [CURRENCY_PAB]: 'balboas',
  [CURRENCY_PEN]: 'soles',
  [CURRENCY_PYG]: 'guaraníes',
  [CURRENCY_USD]: 'dólares',
  [CURRENCY_UYU]: DEFAULT_CURRENCY_SYMBOL_TEXT,
  [CURRENCY_VEF]: 'bolívares',
  [CURRENCY_VES]: 'bolívares',
  [CURRENCY_DCE]: 'dólares',
  [CURRENCY_BTC]: 'bitcoin',
  [CURRENCY_ETH]: 'ethereum',
  [CURRENCY_USDP]: 'pax dólar',
  [CURRENCY_MCN]: 'mercado coin',
  [CURRENCY_MUSD]: 'meli dólar',
  [CURRENCY_LINK]: 'chainlink',
  [CURRENCY_UNI]: 'uniswap',
  [CURRENCY_LTC]: 'litecoin',
  [CURRENCY_USDC]: 'usd coin',
  [CURRENCY_XRP]: 'xrp',
  [CURRENCY_SOL]: 'solana',
  [CURRENCY_DOT]: 'polkadot',
  [CURRENCY_AAVE]: 'aave',
  [CURRENCY_SUI]: 'sui',
  [CURRENCY_ADA]: 'cardano',
  [CURRENCY_AVAX]: 'avalanche',
  [CURRENCY_NEAR]: 'near',
  [CURRENCY_HYPE]: 'hyperliquid',
  [CURRENCY_XLM]: 'stellar',
  [CURRENCY_HBAR]: 'hedera',
  [CURRENCY_PAXG]: 'pax gold'
};
export { AVAILABLE_CURRENCIES, CRYPTO_CURRENCY_IDS, CURRENCY_SYMBOLS, CURRENCY_SYMBOLS_BY_ID, CURRENCY_TEXT_BY_ID, DECIMALS_BY_CURRENCY_ID, DEFAULT_CURRENCY_IDS, DEFAULT_CURRENCY_SYMBOL, DEFAULT_CURRENCY_SYMBOL_TEXT, DEFAULT_DECIMALS_AMOUNT };
