// @nordic/fs-cache/src/source-map/mapped-position/index.ts
import path from 'node:path';

import { getBuildFilePath, readBuildFileContent } from '../../utils/fs';
import {
  parseURL,
  getFileMappedPosition,
  getFallbackMappedPosition,
  readNodeModulesMapFileContent,
  transformFileToMapContent,
} from './utils';

type CodeLocation = {
  url: string;
  file: string;
  line: number;
  column: number;
};

export async function getMappedPosition(
  nodeModulesCache: Map<string, string>,
  { url, file, line, column }: CodeLocation,
) {
  const pathname = parseURL(url)?.pathname;

  if (!pathname) {
    return null;
  }

  const fileName = pathname.slice(1, pathname.lastIndexOf(file) + file.length);
  const filePath = path.resolve(fileName);

  const [
    buildMapFileContent,
    transformedMapFileContent,
    nodeModulesMapFileContent,
  ] = await Promise.all([
    readBuildFileContent(`${fileName}.map`),
    transformFileToMapContent(filePath),
    readNodeModulesMapFileContent(nodeModulesCache, `${filePath}.map`),
  ]);

  const fileMapContent =
    buildMapFileContent ||
    transformedMapFileContent ||
    nodeModulesMapFileContent;

  if (!fileMapContent) {
    const buildFileContent = await readBuildFileContent(fileName);

    return buildFileContent
      ? { source: getBuildFilePath(fileName), line, column }
      : getFallbackMappedPosition(filePath, line, column);
  }

  const mappedPosition = getFileMappedPosition(fileMapContent, line, column);

  if (!mappedPosition) {
    return getFallbackMappedPosition(filePath, line, column);
  }

  mappedPosition.source = path.join(
    path.dirname(buildMapFileContent ? getBuildFilePath(fileName) : filePath),
    mappedPosition.source,
  );

  const deepNodeModulesMapFileContent = await readNodeModulesMapFileContent(
    nodeModulesCache,
    `${mappedPosition.source}.map`,
  );

  if (!deepNodeModulesMapFileContent) {
    return mappedPosition;
  }

  const deepMappedPosition = getFileMappedPosition(
    deepNodeModulesMapFileContent,
    mappedPosition.line,
    mappedPosition.column,
  );

  if (!deepMappedPosition) {
    return mappedPosition;
  }

  deepMappedPosition.source = path.join(
    path.dirname(mappedPosition.source),
    deepMappedPosition.source,
  );

  return deepMappedPosition;
}
