// @nordic/fs-cache/src/source-map/mapped-position/utils.ts
import path from 'node:path';

import { SourceMapConsumer } from 'source-map-js';
import { getOptionalModule } from '@nordic/utils';

import { readFileContent } from '../../utils/fs';

const NODE_MODULES_PATH = path.resolve('node_modules');

export function parseURL(url: string) {
  try {
    return new URL(url);
  } catch {
    return null;
  }
}

export async function transformFileToMapContent(filePath: string) {
  try {
    if (filePath.startsWith(NODE_MODULES_PATH)) {
      return null;
    }

    // TODO: Pending to replace until the Vite module is added
    // const { getViteDevServer } = await getOptionalModule(import('@nordic-dev/vite/server'));
    const { getViteDevServer } = await getOptionalModule(import('./vite-temp'));

    const transformedFile =
      await getViteDevServer?.()?.transformRequest(filePath);

    if (!transformedFile?.map) {
      return null;
    }

    return JSON.stringify(transformedFile.map);
  } catch {
    return null;
  }
}

export async function readNodeModulesMapFileContent(
  nodeModulesCache: Map<string, string>,
  filePath: string,
) {
  if (!filePath.startsWith(NODE_MODULES_PATH)) {
    return null;
  }

  const cachedFileContent = nodeModulesCache.get(filePath);

  if (cachedFileContent) {
    return cachedFileContent;
  }

  const fileContent = await readFileContent(filePath);

  if (fileContent) {
    nodeModulesCache.set(filePath, fileContent);
  }

  return fileContent;
}

export function getFileMappedPosition(
  fileContent: string,
  line: number,
  column: number,
) {
  const consumer = new SourceMapConsumer(JSON.parse(fileContent));
  const mappedPosition = consumer.originalPositionFor({ line, column });

  return mappedPosition.source
    ? {
        source: mappedPosition.source,
        line: mappedPosition.line,
        column: mappedPosition.column,
      }
    : null;
}

export async function getFallbackMappedPosition(
  filePath: string,
  line: number,
  column: number,
) {
  const fileContent = await readFileContent(filePath);

  return fileContent ? { source: path.resolve(filePath), line, column } : null;
}
