// @nordic/fs-cache/src/source-map/mapped-position/webpack.ts
// TODO: Remove file when webpack support is dropped
import type { MappedPosition, RawSourceMap } from 'source-map-js';

import path from 'node:path';

import { SourceMapConsumer } from 'source-map-js';

import { readBuildFileContent, readFileContent } from '../../utils/fs';

type CodeLocation = {
  file: string;
  line: number;
  column: number;
};

const CWD = process.cwd();

const APP_DIR_NAME = path.basename(CWD);

function getFileMappedPosition(
  fileContent: string,
  line: number,
  column: number,
) {
  const consumer = new SourceMapConsumer(
    JSON.parse(fileContent) as RawSourceMap,
  );
  const mappedPosition = consumer.originalPositionFor({ line, column });

  return mappedPosition.source ? mappedPosition : null;
}

async function getMappedPositionFromBuild({
  file,
  line,
  column,
}: CodeLocation) {
  const mapFileContent = await readBuildFileContent(`${file}.map`);

  if (!mapFileContent) {
    return null;
  }

  const mappedPosition = getFileMappedPosition(mapFileContent, line, column);

  if (!mappedPosition) {
    return null;
  }

  return {
    source: mappedPosition.source.replace(`webpack://${APP_DIR_NAME}`, CWD),
    line: mappedPosition.line,
    column,
  };
}

async function getMappedPositionFromNodeModules(
  buildMappedPosition: MappedPosition,
  nodeModulesCache: Map<string, string>,
) {
  if (!buildMappedPosition.source.startsWith(path.join(CWD, 'node_modules'))) {
    return null;
  }

  const mapFilePath = `${buildMappedPosition.source}.map`;
  const mapFileContent =
    nodeModulesCache.get(mapFilePath) ?? (await readFileContent(mapFilePath));

  if (!mapFileContent) {
    return null;
  }

  nodeModulesCache.set(mapFilePath, mapFileContent);

  const mappedPosition = getFileMappedPosition(
    mapFileContent,
    buildMappedPosition.line,
    buildMappedPosition.column,
  );

  if (!mappedPosition) {
    return null;
  }

  return {
    source: path.join(
      path.dirname(buildMappedPosition.source),
      mappedPosition.source,
    ),
    line: mappedPosition.line,
    column: mappedPosition.column,
  };
}

export async function getMappedPosition(
  nodeModulesCache: Map<string, string>,
  codeLocation: CodeLocation,
) {
  const buildMappedPosition = await getMappedPositionFromBuild(codeLocation);

  if (!buildMappedPosition) {
    return null;
  }

  const nodeModulesMappedPosition = await getMappedPositionFromNodeModules(
    buildMappedPosition,
    nodeModulesCache,
  );

  return nodeModulesMappedPosition ?? buildMappedPosition;
}
