// @andes/close-listener/build/cjs/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var React = require('react');
var PropTypes = require('prop-types');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);

/* global document */

/** @typedef {import('../..').CloseListenerProps} CloseListenerProps */

/** @extends React.Component<CloseListenerProps> */
class CloseListener extends React__default["default"].Component {
  /** @param {CloseListenerProps} props */
  constructor(props) {
    super(props);
    this.handleRef = this.handleRef.bind(this);
    this.handleDocumentClick = this.handleDocumentClick.bind(this);
    this.pressEscToClose = this.pressEscToClose.bind(this);
  }
  componentDidMount() {
    document.addEventListener('click', this.handleDocumentClick);
    document.addEventListener('keydown', this.pressEscToClose, false);
  }
  componentWillUnmount() {
    document.removeEventListener('click', this.handleDocumentClick);
    document.removeEventListener('keydown', this.pressEscToClose, false);
  }

  /** @param {KeyboardEvent} event */
  pressEscToClose(event) {
    const _this$props = this.props,
      _this$props$handleEsc = _this$props.handleEsc,
      handleEsc = _this$props$handleEsc === void 0 ? true : _this$props$handleEsc,
      onCloseOutside = _this$props.onCloseOutside;
    if (handleEsc) {
      if (event.code === 'Escape') {
        // esc key
        onCloseOutside(event);
      }
    }
  }

  /** @param {MouseEvent} event */
  handleDocumentClick(event) {
    const _this$props2 = this.props,
      _this$props2$handleCl = _this$props2.handleClick,
      handleClick = _this$props2$handleCl === void 0 ? true : _this$props2$handleCl,
      onCloseOutside = _this$props2.onCloseOutside;
    if (handleClick) {
      if (this.element && !this.element.contains(/** @type {Node} */event.target)) {
        onCloseOutside(event);
      }
    }
  }

  /** @param {HTMLElement} el */
  handleRef(el) {
    const children = this.props.children;
    this.element = el;
    // @ts-expect-error ref prop doesn't exist on ReactElement
    if (children.ref && typeof children.ref === 'function') {
      // @ts-expect-error ref prop doesn't exist on ReactElement
      children.ref(el);
    }
  }
  render() {
    const children = this.props.children;
    return /*#__PURE__*/React__default["default"].cloneElement(/** @type {React.ReactElement} */children, {
      ref: this.handleRef
    });
  }
}
CloseListener.propTypes = {
  children: PropTypes__default["default"].element.isRequired,
  onCloseOutside: PropTypes__default["default"].func.isRequired,
  handleClick: PropTypes__default["default"].bool,
  handleEsc: PropTypes__default["default"].bool
};
var CloseListener$1 = CloseListener;
exports.CloseListener = CloseListener$1;
