// @andes/close-listener/build/cjs/index.production.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var React = require('react');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
class CloseListener extends React__default["default"].Component {
  constructor(props) {
    super(props);
    this.handleRef = this.handleRef.bind(this);
    this.handleDocumentClick = this.handleDocumentClick.bind(this);
    this.pressEscToClose = this.pressEscToClose.bind(this);
  }
  componentDidMount() {
    document.addEventListener('click', this.handleDocumentClick);
    document.addEventListener('keydown', this.pressEscToClose, false);
  }
  componentWillUnmount() {
    document.removeEventListener('click', this.handleDocumentClick);
    document.removeEventListener('keydown', this.pressEscToClose, false);
  }
  pressEscToClose(event) {
    const _this$props = this.props,
      _this$props$handleEsc = _this$props.handleEsc,
      handleEsc = _this$props$handleEsc === void 0 ? true : _this$props$handleEsc,
      onCloseOutside = _this$props.onCloseOutside;
    if (handleEsc) {
      if (event.code === 'Escape') {
        onCloseOutside(event);
      }
    }
  }
  handleDocumentClick(event) {
    const _this$props2 = this.props,
      _this$props2$handleCl = _this$props2.handleClick,
      handleClick = _this$props2$handleCl === void 0 ? true : _this$props2$handleCl,
      onCloseOutside = _this$props2.onCloseOutside;
    if (handleClick) {
      if (this.element && !this.element.contains(event.target)) {
        onCloseOutside(event);
      }
    }
  }
  handleRef(el) {
    const children = this.props.children;
    this.element = el;
    if (children.ref && typeof children.ref === 'function') {
      children.ref(el);
    }
  }
  render() {
    const children = this.props.children;
    return /*#__PURE__*/React__default["default"].cloneElement(children, {
      ref: this.handleRef
    });
  }
}
var CloseListener$1 = CloseListener;
exports.CloseListener = CloseListener$1;
