// nordic-pages-discovery/lib/islands/get-islands-from-file/jsx-elements/elements-of-ast.js
const traverse = require('@babel/traverse').default;
const babelTypes = require('@babel/types');

const { ISLAND_WRAPPER } = require('../../../constants/islands');
const { APP_FILE_MEMBERS } = require('../../../constants/file');
const { NordicIslandsError } = require('../../../error/nordic-islands-error');
const {
  getMoreThanOneIslandChildError,
  getComponentImportedDynamicallyInsideIslandWrapperError,
  getIslandChildNotExportedError,
} = require('../../../error/get-discovery-error');

/**
 * @typedef {import('../../../helpers/islands/component').Component} JSXElementInstance
 * @typedef {Record<string, JSXElementInstance>} JSXElementsManifest
 * @typedef {import('../../../helpers/file-system/islands-manifest').Island} Island
 *
 * @typedef {import('../../../helpers/islands/component').Component & { property: undefined | string? }} Component
 * @typedef {Object} ImportDynamicCaller
 * @property {string} type
 * @property {string | undefined} file
 * @property {string | undefined} module
 *
 * @typedef {Object} ComponentContext
 * @property {string} name
 * @property {Component} component
 * @property {{
 *   isIsland: boolean,
 *   islands: null | undefined,
 *   property: undefined | string?,
 *   importDynamicCaller: ImportDynamicCaller | undefined }} info
 *
 * @typedef {Object} FileContextData
 * @property {string} filePath
 * @property {Record<string, import('./react-component-nodes').ComponentNode | undefined>} nodes
 * @property {import('../../../helpers/islands/component').ComponentManifest} importedComponents
 * @property {string | null} islandWrapper
 */

/**
 * @param {babelTypes.JSXElement} node
 * @param {string} islandWrapper
 */
const isIslandWrapper = (node, islandWrapper) =>
  (babelTypes.isJSXIdentifier(node.openingElement.name) &&
    node.openingElement.name.name === islandWrapper) ||
  (babelTypes.isJSXMemberExpression(node.openingElement.name) &&
    node.openingElement.name.object.name === islandWrapper &&
    node.openingElement.name.property.name === ISLAND_WRAPPER);

/**
 * @param {JSXElementsManifest | {}} sortedElements
 * @param {ComponentContext} componentContext
 */
const addElementToSortedElements = (
  sortedElements,
  { name, component, info },
) => {
  if (!sortedElements[name]) {
    sortedElements[name] = { ...component };
  }

  Object.assign(sortedElements[name], info);
};

/**
 * @param {JSXElementsManifest | {}} destination
 * @param {babelTypes.JSXElement} nodePath
 * @param {FileContextData} fileContextData
 */
const treatComponentWithIslandWrapperForDestinationAddition = (
  destination,
  nodePath,
  { filePath, nodes, importedComponents, islandWrapper },
) => {
  // is Island wrapper child
  if (
    babelTypes.isJSXElement(nodePath.parent) &&
    isIslandWrapper(nodePath.parent, islandWrapper)
  ) {
    return true;
  }

  if (!isIslandWrapper(nodePath.node, islandWrapper)) {
    return false;
  }

  // example: <Island><MyComponent /></Island>
  const jsxChildren = nodePath.node.children.filter((child) =>
    babelTypes.isJSXElement(child),
  );

  if (jsxChildren.length !== 1) {
    throw new NordicIslandsError(
      getMoreThanOneIslandChildError(jsxChildren.length),
      {
        type: 'file',
        path: filePath,
      },
    );
  }

  const islandChildName = jsxChildren[0].openingElement.name?.name;
  const islandChildComponent = importedComponents[islandChildName];

  if (!islandChildComponent) {
    // example: div wrap with Island wrapper
    if (!nodes[islandChildName]) {
      throw new NordicIslandsError(
        `The <Island> component should be used to wrap a custom React Component but the <${islandChildName}> component was used instead`,
        {
          type: `${APP_FILE_MEMBERS.PAGE} or ${APP_FILE_MEMBERS.COMPONENT}`,
          path: filePath,
        },
      );
    }

    if (typeof nodes[islandChildName].namedExport !== 'boolean') {
      throw new NordicIslandsError(
        getIslandChildNotExportedError(islandChildName),
        { type: 'file', path: filePath },
      );
    }

    addElementToSortedElements(destination, {
      name: islandChildName,
      component: {
        importName: nodes[islandChildName].importName,
        path: filePath,
        namedExport: nodes[islandChildName].namedExport,
      },
      info: {
        islands: null,
        isIsland: true,
      },
    });

    delete nodes[islandChildName];

    return true;
  }

  // Catch components imported dynamically with loadable used inside Island wrapper
  if (
    islandChildComponent.importDynamicCaller?.file &&
    islandChildComponent.importDynamicCaller?.module !== 'react' &&
    islandChildComponent.importDynamicCaller?.type !== 'lazy'
  ) {
    throw new NordicIslandsError(
      getComponentImportedDynamicallyInsideIslandWrapperError(islandChildName),
      {
        type: `${APP_FILE_MEMBERS.PAGE} or ${APP_FILE_MEMBERS.COMPONENT}`,
        path: filePath,
      },
    );
  }

  addElementToSortedElements(destination, {
    name: islandChildName,
    component: islandChildComponent,
    info: {
      islands: null,
      isIsland: true,
    },
  });

  return true;
};

/**
 *
 * @param {import('../../../cache/ast-cache').BabelAST} astToAnalize
 * @param {FileContextData} fileContextData
 * @returns {JSXElementsManifest | {}}
 */
const getSortedJSXElementsFromAST = (
  astToAnalize,
  { filePath, nodes, importedComponents, islandWrapper },
) => {
  const sortedElements = {};

  traverse(astToAnalize, {
    noScope: true,
    JSXElement(path) {
      if (
        islandWrapper &&
        treatComponentWithIslandWrapperForDestinationAddition(
          sortedElements,
          path,
          {
            importedComponents,
            nodes,
            islandWrapper,
            filePath,
          },
        )
      ) {
        return;
      }

      const elementName = path.node.openingElement.name?.name;

      if (sortedElements[elementName]) {
        return;
      }

      const component = importedComponents[elementName];

      // Not part of page components (ex. div)
      if (!component) {
        if (!nodes[elementName]) {
          return;
        }

        const elementAsts = Array.from(nodes[elementName].node);

        delete nodes[elementName];

        elementAsts.forEach((elementAst) => {
          Object.assign(
            sortedElements,
            getSortedJSXElementsFromAST(elementAst, {
              filePath,
              nodes,
              islandWrapper,
              importedComponents,
            }),
          );
        });

        return;
      }

      // Component without islands analized previously
      if (component.islands === null) {
        return;
      }

      addElementToSortedElements(sortedElements, {
        name: elementName,
        component,
        info: { isIsland: false },
      });
    },
    Identifier(path) {
      if (
        babelTypes.isVariableDeclarator(path.parent) ||
        babelTypes.isJSXElement(path.parent)
      ) {
        return;
      }

      const potentialElement = {
        name: path.node.name,
        property: null,
      };

      // is "... = variable.prop" (ex. "component = Components.MyComponent;")
      if (
        babelTypes.isMemberExpression(path.parent) &&
        babelTypes.isIdentifier(path.parent.property) &&
        babelTypes.isIdentifier(path.parent.object)
      ) {
        potentialElement.property = path.parent.property.name;
        potentialElement.name = `${path.parent.object.name}-${potentialElement.property}`;
      }

      if (sortedElements[potentialElement.name]) {
        return;
      }

      let component;

      // Avoid errors when looking for properties of Object.prototype (ex .toString())
      if (
        Object.prototype.hasOwnProperty.call(
          importedComponents,
          potentialElement.name,
        )
      ) {
        component = importedComponents[potentialElement.name];
      }

      if (
        Object.prototype.hasOwnProperty.call(importedComponents, path.node.name)
      ) {
        component = importedComponents[path.node.name];
      }

      // Not part of imported page components (ex. function callee)
      if (!component) {
        if (!nodes[potentialElement.name]?.node) {
          return;
        }

        const elementAsts = Array.from(nodes[potentialElement.name].node);

        delete nodes[potentialElement.name];

        elementAsts.forEach((elementAst) => {
          Object.assign(
            sortedElements,
            getSortedJSXElementsFromAST(elementAst, {
              filePath,
              nodes,
              islandWrapper,
              importedComponents,
            }),
          );
        });

        return;
      }

      // Component without islands analized previously
      if (component.islands === null) {
        return;
      }

      // example 1: component = Components.MyComponent;
      // example 2: component = Components[i];
      // example 3: component = MyComponent;
      addElementToSortedElements(sortedElements, {
        name: potentialElement.name,
        component,
        info: {
          property:
            potentialElement.property &&
            typeof component.property === 'undefined'
              ? potentialElement.property
              : null,
          isIsland: false,
        },
      });
    },
  });

  return sortedElements;
};

module.exports = {
  getSortedJSXElementsFromAST,
};
