// frontend-restclient/src/RestClient.js
const axios = require('axios');
const retry = require('retry');
const LRUCache = require('stale-lru-cache');

const { MemoryCache } = require('./MemoryCache');
const defaults = require('../config/defaults');
const { logger, logErrorIfApply } = require('./logger');
const { getPoolName } = require('./tracking/get/appMetrics');
const {
  getETagCacheConfig,
  getRequestCacheData,
  getFromCache,
} = require('./cache-utils');
const mergeContextHeaders = require('./merge-context-headers');
const { buildAutomaticContext } = require('./context');
const { GENERIC_BAD_RETRY_POLICY_IMPLEMENTATION, BAD_HOOK_TYPE } = require('./constants/errorMessages');
const retryConfigBuilder = require('./helpers/retryConfigBuilder');
const isRetryPolicyValid = require('./helpers/isRetryPolicyValid');
const { resolveBaseURL } = require('./helpers/resolveBaseURL');
const { OpenTelemetryClient } = require('./telemetry');
const {
  RETRY_COUNT_HEADER,
  SOCKET_TIMEOUT,
  POOL_NAME,
} = require('./constants/headers');
const { securityValidations } = require('./helpers/ssrf-validations');

/**
 * @typedef {import('../types/index').RestClientConfig} ClientConfig
 * @typedef {import('../types/index').NordicRequestConfig} RequestConfig
 */

/**
 * RestClient class to create a new instance of axios with a custom config.
 */
class RestClient {
  /**
   * @param {ClientConfig} config
   */
  constructor(config = {}) {
    /** @type {ClientConfig} */
    this.config = { ...defaults, ...config };
    this.config.baseURL = resolveBaseURL(config);
    this.config.eTagCache = getETagCacheConfig(config);
    this.client = axios.create(this.config);
    this.eTagCache = new LRUCache(this.config.eTagCache || defaults.eTagCache);
    this.poolNameIdentifier = getPoolName();
    this.eTagPendingRequests = {};
    if (this.config.cacheClient) {
      this.cache = this.config.cacheClient();
    } else if (this.config.cache) {
      this.cache = new MemoryCache(this.config.cache);
    }

    this._injectHooks(this.client);
  }

  /**
   * Makes a new request with a same config from doRequest and returns a promise.
   * @param {RequestConfig} config
   */
  makeClientRequest(config) {
    const fullUrl = this.client.getUri(config);
    const mergedConfig = axios.mergeConfig(this.config, config);

    try {
      securityValidations(fullUrl, mergedConfig);
    } catch (error) {
      return Promise.reject(error);
    }

    // Prepare a new request
    const request = (reqConfig) => {
      const {
        handleConcurrentRequests,
        processResponse,
        configWithCacheData,
        injectXCacheHeader,
      } = getRequestCacheData(reqConfig, this);

      const requestPromise = this.client.request(configWithCacheData)
        .then(injectXCacheHeader)
        .then(processResponse)
        // Catch the error to log and throw an error, which triggers the next
        // reject handler in the chain to be called.
        .catch((error) => {
          logErrorIfApply(error, reqConfig, this.config);
          throw error;
        });

      return handleConcurrentRequests(requestPromise);
    };

    // Make request with retries
    if (config.retry) {
      if (isRetryPolicyValid(config.retry.policy)) {
        return this.retry(request, config);
      }

      return Promise.reject(GENERIC_BAD_RETRY_POLICY_IMPLEMENTATION);
    }
    // Make request
    return request(config);
  }

  /**
   * Makes a request with a given config and returns a response.
   * Besides, adds traceability with OTel.
   * @param {Object} resolve
   * @param {Object} config
   * @param {string} config.method
   * @param {string} config.url
   * @param {string} config.config
   * @param {Object} OTel
   * @see https://github.com/mzabriskie/axios#request-config
   */
  doCacheRequest(resolve, config, OTel) {
    let isCached = false;
    return getFromCache({
      reqConfig: config,
      cache: this.cache,
      clientConfig: this.config,
    })
      .then((cachedResponse) => {
        // Return cached response
        if (cachedResponse) {
          isCached = true;
          OTel.endSpan();
          resolve(cachedResponse);
        }
        return undefined;
      })
      .catch((error) => logErrorIfApply(error, config, this.config))
      .finally(() => {
        if (!isCached) {
          resolve(this.makeClientRequest(config)
            .then((resp) => {
              OTel.addAttributesToSpan(config, resp);
              return resp;
            })
            .catch((error) => {
              OTel.recordSpanError();
              throw error;
            })
            .finally(() => OTel.endSpan()));
        }
      });
  }

  /**
   * Makes a request with a given method, url, and config and returns a promise.
   * @param {RequestConfig} config
   */
  doRequest(config) {
    /**
     * Ignore certain props from request config to prevent SSRF attacks that try to override these props.
     * This ensures these props can only be configured when instantiating the client.
     */
    const { baseURL, allowAbsoluteUrls, allowRepeatedParams, ...reqConfig } = config;
    const OTel = new OpenTelemetryClient();

    reqConfig.retry = retryConfigBuilder(reqConfig, this.config);

    /**
     * Set more headers
     */
    reqConfig.headers = {
      ...reqConfig.headers,
      [POOL_NAME]: this.poolNameIdentifier,
      [SOCKET_TIMEOUT]: this.config.timeout ? this.config.timeout.toString() : 'undefined',
      [RETRY_COUNT_HEADER]: '0',
    };

    /**
     * Adds context headers into request headers when available.
     */
    if (reqConfig.context) {
      reqConfig.headers = mergeContextHeaders(reqConfig.context, reqConfig.headers);
    } else {
      const context = buildAutomaticContext();

      if (context) {
        reqConfig.headers = mergeContextHeaders(context, reqConfig.headers);
      }
    }

    logger.debug(`The headers for ${reqConfig.url} are:\n${JSON.stringify(reqConfig.headers, null, 2)}`);

    /**
     * Make request first adding traceability support with OTel
     */
    return new Promise((resolve) => OTel.asyncBuildActiveSpan(
      () => this.doCacheRequest(resolve, reqConfig, OTel),
      reqConfig,
    ));
  }

  /**
   * Alias to make a GET request and returns a promise.
   * @param url String
   * @param config Object
   * @see https://github.com/mzabriskie/axios#request-config
   */
  get(url, config = {}) {
    config.method = 'get';
    config.url = url;
    return this.doRequest(config);
  }

  /**
   * Alias to make a POST request and returns a promise.
   * @param url String
   * @param config Object
   * @see https://github.com/mzabriskie/axios#request-config
   */
  post(url, config = {}) {
    config.method = 'post';
    config.url = url;
    return this.doRequest(config);
  }

  /**
   * Alias to make a PUT request and returns a promise.
   * @param url String
   * @param config Object
   * @see https://github.com/mzabriskie/axios#request-config
   */
  put(url, config = {}) {
    config.method = 'put';
    config.url = url;
    return this.doRequest(config);
  }

  /**
   * Alias to make a PATCH request and returns a promise.
   * @param url String
   * @param config Object
   * @see https://github.com/mzabriskie/axios#request-config
   */
  patch(url, config = {}) {
    config.method = 'patch';
    config.url = url;
    return this.doRequest(config);
  }

  /**
   * Alias to make a DELETE request and returns a promise.
   * @param url String
   * @param config Object
   * @see https://github.com/mzabriskie/axios#request-config
   */
  delete(url, config = {}) {
    config.method = 'delete';
    config.url = url;
    return this.doRequest(config);
  }

  /**
   * Alias to make a HEAD request and returns a promise.
   * @param url String
   * @param config Object
   * @see https://github.com/mzabriskie/axios#request-config
   */
  head(url, config = {}) {
    config.method = 'head';
    config.url = url;
    return this.doRequest(config);
  }

  /**
   * Alias to make an OPTIONS request and returns a promise.
   * @param url String
   * @param config Object
   * @see https://github.com/mzabriskie/axios#request-config
   */
  options(url, config = {}) {
    config.method = 'options';
    config.url = url;
    return this.doRequest(config);
  }

  /**
   * Alias to make a PURGE request and returns a promise.
   * @param url String
   * @param config Object
   * @see https://github.com/mzabriskie/axios#request-config
   */
  purge(url, config = {}) {
    config.method = 'purge';
    config.url = url;
    return this.doRequest(config);
  }

  /**
   * It makes concurrent requests that combine multiple promises and returns
   * a collection of responses.
   * @param requests Object
   * @see https://github.com/mzabriskie/axios#concurrency
   */
  all(requests) {
    return axios.all(requests)
      .then(axios.spread((...response) => response))
      .catch((error) => {
        logErrorIfApply(error, {}, this.config);
        throw error;
      });
  }

  /**
   * Retry Strategy.
   * It supports both strategies: simple and backoff (using factor option).
   * @See: https://github.com/tim-kos/node-retry
   */
  retry(request, config) {
    const { baseURL } = this.config;
    const queryParams = config.params
      ? `?${new URLSearchParams(config.params).toString()}`
      : '';
    const fullUrl = `${baseURL?.replace(/\/$/, '') || ''}${config.url}${queryParams}`;

    // Return a new promise
    return new Promise((resolve, reject) => {
      // Configure the operation to retry
      const operation = retry.operation({
        retries: config.retry.maxRetries,
        minTimeout: config.retry.delay,
        maxTimeout: config.retry.maxDelay,
        factor: config.retry.factor,
      });

      // For each attempt make a new request.
      operation.attempt((currentAttempt) => {
        Promise.resolve(currentAttempt)
          .then(() => {
            const retryCount = (currentAttempt - 1); // Since first attempt isn't a retry, its just the first op call

            logger.debug(`Request attempt Nº ${currentAttempt}. Retry Nº ${retryCount}.`, { url: fullUrl });

            const updatedHeaders = {
              ...config.headers,
              [RETRY_COUNT_HEADER]: retryCount.toString(),
            };
            return request({ ...config, headers: updatedHeaders });
          })
          // If the request has not an error, resolve the promise with the response.
          .then((response) => resolve(response))
          /**
           * The operation will be retried when:
           * - the request's method is allowed;
           * - the error code is allowed;
           * - the maximum amount of retries has not been reached.
           */
          .catch((error) => {
            // Retry only if error contain a config object and matches the retry policy
            if (
              !(error.config
                && config.retry.policy.hasToRetry(error)
                && operation.retry(error)) // note this will execute a new retry when necessary: https://www.npmjs.com/package/retry#retryoperationretryerror
            ) {
              // Reject the promise with the given error.
              reject(error);
            }
          });
      });
    });
  }

  /**
  * Inject hooks if they were passed through the instance config
  * @return {undefined} Does not return anything useful
  */
  _injectHooks(client) {
    const { requestHook, responseHook } = this.config;

    // If the requestHook or the responseHook are defined they need to be functions
    if ((requestHook && typeof requestHook !== 'function') || (responseHook && typeof responseHook !== 'function')) throw new Error(BAD_HOOK_TYPE);

    // The user should not be able to modify the request configs and response data
    // If we directly set the interceptor with the user hook the return value will overwrite the response / request config.
    // Due to this reason, we need to wrap the user hook
    const userHookWrapper = (hook) => (data) => {
      hook(data);
      return data;
    };

    if (requestHook) {
      client.interceptors.request.use(userHookWrapper(requestHook));
    }

    if (responseHook) {
      client.interceptors.response.use(userHookWrapper(responseHook));
    }
  }
}

/**
 * Expose RestClient
 */
module.exports = RestClient;
