// @nordic/assets-config/dist/cjs/index.production.js
'use strict';

var path = require('node:path');
var serverContext = require('@meli/server-context');
var frontendConfig = require('frontend-config');
var env = require('@nordic/env');

const FALLBACK_ASSETS_CDN = 'https://http2.mlstatic.com';
function getAssetsCDN() {
    try {
        const req = serverContext.getReq();
        const { id = 'ML', siteId = 'MLA' } = req.platform || {};
        return (frontendConfig.config.get('url.assets.link', id, siteId) || FALLBACK_ASSETS_CDN);
    }
    catch {
        return FALLBACK_ASSETS_CDN;
    }
}

function getAssetsPrefix() {
    // Do not use false here because while generating development assets the
    // bundler will resolve this module from the production entry, leaving the CDN
    // prefix inlined into dev assets. The runtime env is the source of truth for
    // deciding which prefix those generated assets must use.
    if (!env.env.PRODUCTION) {
        return env.env.ASSETS_BASE_PATH || '/';
    }
    return (frontendConfig.config.assets?.prefix ||
        `${getAssetsCDN()}/frontend-assets/${env.env.APPLICATION}/`);
}

const NORDIC_DIR = path.resolve('.nordic');
const METADATA_DIR = path.join(NORDIC_DIR, '_metadata');
const ASSETS_CONFIG = {
    PUBLIC_DIR: path.resolve('public'),
    CLIENT_OUTPUT_DIR: path.resolve('build'),
    SERVER_OUTPUT_DIR: path.join(NORDIC_DIR, 'build'),
    METADATA_DIR,
    CLIENT_STATS_PATH: path.join(METADATA_DIR, 'client-stats.json'),
    SERVER_STATS_PATH: path.join(METADATA_DIR, 'server-stats.json'),
    MANIFEST_PATH: path.join(METADATA_DIR, 'manifest.json'),
    IMAGES_BLUR_DIR: path.join(NORDIC_DIR, 'images'),
    IMAGES_BLUR_PREFIX: 'blur-',
    MANIFEST_API_ROUTER_KEY: '_nordic_api-router.js',
    MANIFEST_APP_ROUTER_KEY: '_nordic_app-router.js',
    PREFIX: '_nordic/assets/',
};

exports.ASSETS_CONFIG = ASSETS_CONFIG;
exports.getAssetsCDN = getAssetsCDN;
exports.getAssetsPrefix = getAssetsPrefix;
