// @andes/thumbnail/build/cjs/index.production.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _extends = require('@babel/runtime/helpers/extends');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var React = require('react');
var classNames = require('classnames');
var warning = require('@andes/common/build/utils/warning');
require('@andes/common/build/utils/includes');
var context = require('@andes/context');
var visuallyHidden = require('@andes/visually-hidden');
var themes = require('@andes/themes');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
const namespace$1 = 'andes-thumbnail';
const WARNING_HIERARCHY_DEPRECATED = `hierarchy prop value "neutral" will be deprecated in a future major version.
Please, use hierarchy="mute" prop on <Thumbnail> instead`;
const getBadgeSizeByThumbnailSize = size => {
  if (size === '80' || size === '72' || size === '64') return 'large';
  return 'small';
};
function isIcon(children) {
  if (/*#__PURE__*/React.isValidElement(children) && typeof children.type === 'function') {
    return children.type(children.props).type === 'svg';
  }
  return false;
}
const isImage = children => {
  if (/*#__PURE__*/React.isValidElement(children)) {
    return children.type === 'img' || children.props.src;
  }
  return false;
};
function isText(children) {
  if (/*#__PURE__*/React.isValidElement(children)) {
    return !children.props.src && !isIcon(children);
  }
  return false;
}
const verifyBadgeType = (size, badge) => {
  const isDot = badge.type.displayName === 'Dot';
  if (isDot && (size === '24' || size === '32') || !isDot && size !== '24' && size !== '32') {
    return badge;
  }
  warning__default["default"](!isDot, `${size} should use Pill Badges`);
  return false;
};
const getContainerCustomColorStyle = (color, hierarchy) => {
  if (!color) return;
  const backgroundColor = color.backgroundColor,
    borderColor = color.borderColor,
    textColor = color.color;
  return {
    style: {
      [`--andes-thumbnail-${hierarchy}-color-text-default`]: textColor,
      [`--andes-thumbnail-${hierarchy}-color-fill-default`]: backgroundColor,
      '--andes-thumbnail-color-border-default': borderColor,
      '--andes-color-icon-primary': textColor
    }
  };
};
const Thumbnail = /*#__PURE__*/function () {
  const Thumbnail = _ref => {
    var _badge$props;
    let badge = _ref.badge,
      children = _ref.children,
      className = _ref.className,
      customColor = _ref.color,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      _ref$hierarchy = _ref.hierarchy,
      hierarchy = _ref$hierarchy === void 0 ? 'mute' : _ref$hierarchy,
      id = _ref.id,
      _ref$modifier = _ref.modifier,
      modifier = _ref$modifier === void 0 ? 'circle' : _ref$modifier,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? '48' : _ref$size,
      srLabel = _ref.srLabel,
      ref = _ref.ref,
      backgroundType = _ref.backgroundType,
      _ref$andesDataAttribu = _ref.andesDataAttributes,
      andesDataAttributes = _ref$andesDataAttribu === void 0 ? {} : _ref$andesDataAttribu;
    const _useThemeContext = themes.useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const componentId = context.useId(id);
    if (typeof hierarchy === 'string') {
      warning__default["default"](hierarchy !== 'neutral', WARNING_HIERARCHY_DEPRECATED);
    }
    const badgeColor = badge ? ((_badge$props = badge.props) === null || _badge$props === void 0 ? void 0 : _badge$props.color) || 'accent' : undefined;
    const computedHierarchy = badge ? 'mute' : hierarchy;
    const classes = classNames__default["default"](namespace$1, `${namespace$1}--${modifier}`, `${namespace$1}--${size}`, {
      [`${namespace$1}__badge`]: badge,
      [`${namespace$1}__badge-${badgeColor}`]: badgeColor,
      [`${namespace$1}__image`]: isImage(children),
      [`${namespace$1}__image--disabled`]: isImage(children) && disabled,
      [`${namespace$1}__icon`]: isIcon(children),
      [`${namespace$1}--${computedHierarchy}`]: isIcon(children) || isText(children),
      [`${namespace$1}--${computedHierarchy}-disabled`]: (isIcon(children) || isText(children)) && disabled
    }, className);
    const wrapperClasses = classNames__default["default"](`${namespace$1}-container`, {
      [`${namespace$1}--color-on-background`]: backgroundType === 'complexBackground'
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('thumbnail'));
    const a11yProps = _objectSpread__default["default"](_objectSpread__default["default"]({}, !isImage(children) && {
      'aria-hidden': !!srLabel
    }), {}, {
      title: srLabel
    });
    const getColor = () => {
      var _children$props;
      const color = (_children$props = children.props) === null || _children$props === void 0 ? void 0 : _children$props.color;
      return {
        color: color !== 'rgba(0, 0, 0, 0.9)' ? color : 'currentColor'
      };
    };
    const sharedProps = _objectSpread__default["default"](_objectSpread__default["default"](_objectSpread__default["default"]({}, themeHash), andesDataAttributes), {}, {
      'data-andes-thumbnail': true,
      'data-andes-thumbnail-hierarchy': computedHierarchy,
      'data-andes-thumbnail-size': size
    });
    return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
      className: wrapperClasses,
      ref: ref
    }, sharedProps), /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
      className: classes,
      "data-andes-thumbnail-content": true
    }, getContainerCustomColorStyle(customColor, computedHierarchy)), React__default["default"].Children.only(isIcon(children) ? /*#__PURE__*/React__default["default"].cloneElement(children, _objectSpread__default["default"](_objectSpread__default["default"]({}, getColor()), a11yProps)) : /*#__PURE__*/React__default["default"].cloneElement(children, _objectSpread__default["default"]({}, a11yProps))), srLabel && /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, null, srLabel)), badge && verifyBadgeType(size, badge) && /*#__PURE__*/React__default["default"].cloneElement(badge, {
      size: getBadgeSizeByThumbnailSize(size),
      id: badge.props.id || `${componentId}-thumbnail-badge`,
      backgroundType
    }));
  };
  Thumbnail.displayName = 'Thumbnail';
  return Thumbnail;
}();
var Thumbnail$1 = Thumbnail;
const TYPE_GRID = 'grid';
const SIZE_16 = '16';
const SIZE_24 = '24';
const SIZE_32 = '32';
const SIZE_40 = '40';
const SIZE_48 = '48';
const SIZE_72 = '72';
const SIZE_80 = '80';
const namespace = 'andes-thumbnail-multiple';
const getChildrenSize = (type, containerSize) => {
  let childrenSize;
  const isLargeContainer = containerSize === SIZE_72 || containerSize === SIZE_80;
  const isSmallContainer = containerSize === SIZE_40 || containerSize === SIZE_48;
  if (type === TYPE_GRID) {
    if (isLargeContainer) {
      childrenSize = SIZE_32;
    } else if (isSmallContainer) {
      childrenSize = SIZE_16;
    } else {
      childrenSize = SIZE_24;
    }
  } else {
    childrenSize = containerSize;
  }
  return childrenSize;
};
const ThumbnailMultiple = /*#__PURE__*/function () {
  const ThumbnailMultiple = _ref => {
    let children = _ref.children,
      className = _ref.className,
      _ref$containerSize = _ref.containerSize,
      containerSize = _ref$containerSize === void 0 ? '56' : _ref$containerSize,
      id = _ref.id,
      _ref$modifier = _ref.modifier,
      modifier = _ref$modifier === void 0 ? 'circle' : _ref$modifier,
      overflow = _ref.overflow,
      _ref$type = _ref.type,
      type = _ref$type === void 0 ? 'grid' : _ref$type,
      ref = _ref.ref;
    const componentId = context.useId(id);
    const _useThemeContext = themes.useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const classes = classNames__default["default"](namespace, `${namespace}--${type}`, `${namespace}--${type}-${containerSize}`, className, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('thumbnail'));
    const childSize = getChildrenSize(type, containerSize);
    return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
      className: classes,
      ref: ref,
      "data-andes-thumbnail-multiple": true,
      "data-andes-thumbnail-multiple-type": type
    }, themeHash), React__default["default"].Children.map(children, (child, index) => /*#__PURE__*/React__default["default"].cloneElement(child, {
      size: childSize,
      modifier,
      id: child.props.id || `${componentId}-${index}`
    })), overflow && /*#__PURE__*/React__default["default"].createElement(Thumbnail$1, {
      id: `${componentId}-overflow`,
      className: `${namespace}--overflow`,
      size: childSize,
      modifier: modifier,
      andesDataAttributes: {
        'data-andes-thumbnail-multiple-overflow': true
      }
    }, /*#__PURE__*/React__default["default"].createElement("span", null, "+", overflow)));
  };
  ThumbnailMultiple.displayName = 'ThumbnailMultiple';
  return ThumbnailMultiple;
}();
var ThumbnailMultiple$1 = ThumbnailMultiple;
exports.Thumbnail = Thumbnail$1;
exports.ThumbnailMultiple = ThumbnailMultiple$1;
