// @andes/thumbnail/build/cjs/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _extends = require('@babel/runtime/helpers/extends');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var React = require('react');
var PropTypes = require('prop-types');
var classNames = require('classnames');
var warning = require('@andes/common/build/utils/warning');
var includes = require('@andes/common/build/utils/includes');
var context = require('@andes/context');
var visuallyHidden = require('@andes/visually-hidden');
var themes = require('@andes/themes');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var includes__default = /*#__PURE__*/_interopDefaultLegacy(includes);
const namespace$1 = 'andes-thumbnail';
const WARNING_HIERARCHY_DEPRECATED = `hierarchy prop value "neutral" will be deprecated in a future major version.
Please, use hierarchy="mute" prop on <Thumbnail> instead`;
const allowedBadgeTypes = ['Dot', 'Pill'];

/** @typedef {import('../../index').ThumbnailSize} Size */
/** @typedef {import('../../index').Color} Color */
/** @typedef {import('../../index').ThumbnailHierarchy} Hierarchy */

/**
 * @param {Size} size
 * @return {'large' | 'small'}
 */
const getBadgeSizeByThumbnailSize = size => {
  if (size === '80' || size === '72' || size === '64') return 'large';
  return 'small';
};

/**
 * @param {React.ReactElement} children
 * @return {boolean}
 */
function isIcon(children) {
  if (/*#__PURE__*/React.isValidElement(children) && typeof children.type === 'function') {
    // @ts-expect-error - The type of children.type is callable, was previously checked with typeof
    return children.type(children.props).type === 'svg';
  }
  return false;
}

/**
 * @param {React.ReactElement} children
 * @return {boolean}
 */
const isImage = children => {
  if (/*#__PURE__*/React.isValidElement(children)) {
    return children.type === 'img' || children.props.src;
  }
  return false;
};

/**
 * @param {React.ReactElement} children
 * @return {boolean}
 */
function isText(children) {
  if (/*#__PURE__*/React.isValidElement(children)) {
    return !children.props.src && !isIcon(children);
  }
  return false;
}

/**
 * @param {Size} size
 * @param {React.ReactElement} badge
 * @return {React.ReactElement | boolean}
 */
const verifyBadgeType = (size, badge) => {
  // @ts-expect-error: we add displayName
  const isDot = badge.type.displayName === 'Dot';
  if (isDot && (size === '24' || size === '32') || !isDot && size !== '24' && size !== '32') {
    return badge;
  }
  warning__default["default"](!isDot, `${size} should use Pill Badges`);
  return false;
};

/**
 * @param {Color | undefined} color
 * @param {Hierarchy} hierarchy
 * @return {Record<string, unknown> | undefined}
 */
const getContainerCustomColorStyle = (color, hierarchy) => {
  if (!color) return;
  const backgroundColor = color.backgroundColor,
    borderColor = color.borderColor,
    textColor = color.color;
  return {
    style: {
      [`--andes-thumbnail-${hierarchy}-color-text-default`]: textColor,
      [`--andes-thumbnail-${hierarchy}-color-fill-default`]: backgroundColor,
      '--andes-thumbnail-color-border-default': borderColor,
      '--andes-color-icon-primary': textColor
    }
  };
};

/**
 * @typedef {Object} ThumbnailInternalProps
 * @prop {Record<string, unknown>} [andesDataAttributes]
 */

/** @typedef {import('../../index').ThumbnailProps} ThumbnailPublicProps */

/** @typedef {ThumbnailInternalProps & ThumbnailPublicProps} ThumbnailProps */

/** @type {React.FunctionComponent<ThumbnailProps>} */
const Thumbnail = _ref => {
  var _badge$props;
  let badge = _ref.badge,
    children = _ref.children,
    className = _ref.className,
    customColor = _ref.color,
    _ref$disabled = _ref.disabled,
    disabled = _ref$disabled === void 0 ? false : _ref$disabled,
    _ref$hierarchy = _ref.hierarchy,
    hierarchy = _ref$hierarchy === void 0 ? 'mute' : _ref$hierarchy,
    id = _ref.id,
    _ref$modifier = _ref.modifier,
    modifier = _ref$modifier === void 0 ? 'circle' : _ref$modifier,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? '48' : _ref$size,
    srLabel = _ref.srLabel,
    ref = _ref.ref,
    backgroundType = _ref.backgroundType,
    _ref$andesDataAttribu = _ref.andesDataAttributes,
    andesDataAttributes = _ref$andesDataAttribu === void 0 ? {} : _ref$andesDataAttribu;
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  const componentId = context.useId(id);

  // @deprecate [hierarchy="neutral"] - Warning from: 9.0.0
  if (typeof hierarchy === 'string') {
    warning__default["default"](hierarchy !== 'neutral', WARNING_HIERARCHY_DEPRECATED);
  }
  const badgeColor = badge ? ((_badge$props = badge.props) === null || _badge$props === void 0 ? void 0 : _badge$props.color) || 'accent' : undefined;
  const computedHierarchy = badge ? 'mute' : hierarchy;
  const classes = classNames__default["default"](namespace$1, `${namespace$1}--${modifier}`, `${namespace$1}--${size}`, {
    [`${namespace$1}__badge`]: badge,
    [`${namespace$1}__badge-${badgeColor}`]: badgeColor,
    [`${namespace$1}__image`]: isImage(children),
    [`${namespace$1}__image--disabled`]: isImage(children) && disabled,
    [`${namespace$1}__icon`]: isIcon(children),
    [`${namespace$1}--${computedHierarchy}`]: isIcon(children) || isText(children),
    [`${namespace$1}--${computedHierarchy}-disabled`]: (isIcon(children) || isText(children)) && disabled
  }, className);
  const wrapperClasses = classNames__default["default"](`${namespace$1}-container`, {
    [`${namespace$1}--color-on-background`]: backgroundType === 'complexBackground'
  }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('thumbnail'));
  const a11yProps = _objectSpread__default["default"](_objectSpread__default["default"]({}, !isImage(children) && {
    'aria-hidden': !!srLabel
  }), {}, {
    title: srLabel
  });
  const getColor = () => {
    var _children$props;
    // @ts-expect-error - It is previously verified that it is an SVG which can contain the color prop.
    const color = (_children$props = children.props) === null || _children$props === void 0 ? void 0 : _children$props.color;
    return {
      color: color !== 'rgba(0, 0, 0, 0.9)' ? color : 'currentColor'
    };
  };
  const sharedProps = _objectSpread__default["default"](_objectSpread__default["default"](_objectSpread__default["default"]({}, themeHash), andesDataAttributes), {}, {
    'data-andes-thumbnail': true,
    'data-andes-thumbnail-hierarchy': computedHierarchy,
    'data-andes-thumbnail-size': size
  });
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    className: wrapperClasses,
    ref: ref
  }, sharedProps), /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    className: classes,
    "data-andes-thumbnail-content": true
  }, getContainerCustomColorStyle(customColor, computedHierarchy)), React__default["default"].Children.only(isIcon(children) ? /*#__PURE__*/React__default["default"].cloneElement(children, _objectSpread__default["default"](_objectSpread__default["default"]({}, getColor()), a11yProps)) : /*#__PURE__*/React__default["default"].cloneElement(children, _objectSpread__default["default"]({}, a11yProps))), srLabel && /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, null, srLabel)), badge && verifyBadgeType(size, badge) && /*#__PURE__*/React__default["default"].cloneElement(badge, {
    size: getBadgeSizeByThumbnailSize(size),
    id: badge.props.id || `${componentId}-thumbnail-badge`,
    backgroundType
  }));
};

/** @type {import('prop-types').Validator<React.ReactElement>} */
const checkBadgeType = (_ref2, propName, componentName) => {
  var _badge$type;
  let badge = _ref2.badge;
  PropTypes__default["default"].checkPropTypes({
    badge: PropTypes__default["default"].element
  }, {
    badge
  }, propName, componentName);
  const badgeType = badge === null || badge === void 0 || (_badge$type = badge.type) === null || _badge$type === void 0 ? void 0 : _badge$type.displayName;
  if (!badgeType) return null;
  if (!includes__default["default"](allowedBadgeTypes, badgeType)) {
    throw new TypeError(`${badgeType} is not allowed`);
  }
  return null;
};

/** @type {import('prop-types').Validator<Size>} */
const checkThumbnailSize = (_ref3, propName, componentName) => {
  let size = _ref3.size,
    children = _ref3.children,
    badge = _ref3.badge;
  PropTypes__default["default"].checkPropTypes({
    size: PropTypes__default["default"].oneOf(['96', '80', '72', '64', '56', '48', '40', '32', '24', '16'])
  }, {
    size
  }, propName, componentName);
  if (size === '16' && isIcon(children)) {
    return new Error('Thumbnail of size 16px can\'t be used with icon.');
  }
  if (size === '16' && typeof badge !== 'undefined') {
    return new Error('Thumbnail of size 16px can\'t be used with badge.');
  }
  return null;
};
Thumbnail.displayName = 'Thumbnail';
Thumbnail.propTypes = {
  badge: checkBadgeType,
  children: PropTypes__default["default"].element.isRequired,
  className: PropTypes__default["default"].string,
  color: PropTypes__default["default"].shape({
    backgroundColor: PropTypes__default["default"].string,
    borderColor: PropTypes__default["default"].string,
    color: PropTypes__default["default"].string
  }),
  disabled: PropTypes__default["default"].bool,
  hierarchy: PropTypes__default["default"].oneOf(['loud', 'quiet', 'mute']),
  id: PropTypes__default["default"].string,
  modifier: PropTypes__default["default"].oneOf(['circle', 'square']),
  size: checkThumbnailSize,
  srLabel: PropTypes__default["default"].string,
  backgroundType: PropTypes__default["default"].oneOf(['default', 'complexBackground']),
  andesDataAttributes: PropTypes__default["default"].shape({})
};
var Thumbnail$1 = Thumbnail;
const TYPE_GRID = 'grid';
const TYPE_STACKED = 'stacked';
const AVAILABLE_TYPES = [TYPE_GRID, TYPE_STACKED];
const SIZE_16 = '16';
const SIZE_24 = '24';
const SIZE_32 = '32';
const SIZE_40 = '40';
const SIZE_48 = '48';
const SIZE_56 = '56';
const SIZE_64 = '64';
const SIZE_72 = '72';
const SIZE_80 = '80';
const GRID_SIZES = [SIZE_40, SIZE_48, SIZE_56, SIZE_64, SIZE_72, SIZE_80];
const STACKED_SIZES = [SIZE_24, SIZE_32, SIZE_40, SIZE_48, SIZE_56, SIZE_64, SIZE_72, SIZE_80];
const namespace = 'andes-thumbnail-multiple';

/** @typedef {import('../../index').ThumbnailType} Type */
/** @typedef {Exclude<import('../../index').ThumbnailSize, '96'>} Size */

/**
 * @param {Type} type
 * @param {Size} containerSize
 * @return {Size}
 */
const getChildrenSize = (type, containerSize) => {
  /** @type {Size} */
  let childrenSize;
  const isLargeContainer = containerSize === SIZE_72 || containerSize === SIZE_80;
  const isSmallContainer = containerSize === SIZE_40 || containerSize === SIZE_48;
  if (type === TYPE_GRID) {
    if (isLargeContainer) {
      childrenSize = SIZE_32;
    } else if (isSmallContainer) {
      childrenSize = SIZE_16;
    } else {
      childrenSize = SIZE_24;
    }
  } else {
    childrenSize = containerSize;
  }
  return childrenSize;
};

/** @type {import('prop-types').Validator<React.ReactElement>} */
const checkCount = (_ref, propName, componentName) => {
  let children = _ref.children;
  PropTypes__default["default"].checkPropTypes({
    children: PropTypes__default["default"].element.isRequired
  }, {
    children
  }, propName, componentName);
  const childrenQuantity = React__default["default"].Children.count(children);
  if (childrenQuantity < 2 || childrenQuantity > 4) {
    return new Error('Thumbnail Multiple is meant for 2, 3 or 4 thumbnails. Please use the appropiate amount of children');
  }
  return null;
};

/** @type {import('prop-types').Validator<Type>} */
const checkThumbnailType = (_ref2, propName, componentName) => {
  let type = _ref2.type,
    containerSize = _ref2.containerSize;
  PropTypes__default["default"].checkPropTypes({
    type: PropTypes__default["default"].oneOf(AVAILABLE_TYPES)
  }, {
    type
  }, propName, componentName);
  if (type === TYPE_GRID && !includes__default["default"](GRID_SIZES, containerSize)) {
    return new Error(`Grid type only accepts sizes '40', '48', '56', '64', '72', 80'.
    Please select one of these options`);
  }
  if (type === TYPE_STACKED && !includes__default["default"](STACKED_SIZES, containerSize)) {
    return new Error(`Stacked type only accepts sizes '24', '32', '40', '48', '56', '64', '72', '80'.
    Please select one of these options`);
  }
  return null;
};

/** @type {typeof import('../../index').ThumbnailMultiple} */
const ThumbnailMultiple = _ref3 => {
  let children = _ref3.children,
    className = _ref3.className,
    _ref3$containerSize = _ref3.containerSize,
    containerSize = _ref3$containerSize === void 0 ? '56' : _ref3$containerSize,
    id = _ref3.id,
    _ref3$modifier = _ref3.modifier,
    modifier = _ref3$modifier === void 0 ? 'circle' : _ref3$modifier,
    overflow = _ref3.overflow,
    _ref3$type = _ref3.type,
    type = _ref3$type === void 0 ? 'grid' : _ref3$type,
    ref = _ref3.ref;
  const componentId = context.useId(id);
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  const classes = classNames__default["default"](namespace, `${namespace}--${type}`, `${namespace}--${type}-${containerSize}`, className, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('thumbnail'));
  const childSize = getChildrenSize(type, containerSize);
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    className: classes,
    ref: ref,
    "data-andes-thumbnail-multiple": true,
    "data-andes-thumbnail-multiple-type": type
  }, themeHash), React__default["default"].Children.map(children, (/** @type {React.ReactElement} */child, index) => /*#__PURE__*/React__default["default"].cloneElement(child, {
    size: childSize,
    modifier,
    id: child.props.id || `${componentId}-${index}`
  })), overflow && /*#__PURE__*/React__default["default"].createElement(Thumbnail$1, {
    id: `${componentId}-overflow`,
    className: `${namespace}--overflow`,
    size: childSize,
    modifier: modifier,
    andesDataAttributes: {
      'data-andes-thumbnail-multiple-overflow': true
    }
  }, /*#__PURE__*/React__default["default"].createElement("span", null, "+", overflow)));
};
ThumbnailMultiple.propTypes = {
  children: checkCount,
  className: PropTypes__default["default"].string,
  containerSize: PropTypes__default["default"].oneOf(['24', '32', '40', '48', '56', '64', '72', '80']),
  id: PropTypes__default["default"].string,
  modifier: PropTypes__default["default"].oneOf(['circle', 'square']),
  overflow: PropTypes__default["default"].number,
  type: checkThumbnailType
};
ThumbnailMultiple.displayName = 'ThumbnailMultiple';
var ThumbnailMultiple$1 = ThumbnailMultiple;
exports.Thumbnail = Thumbnail$1;
exports.ThumbnailMultiple = ThumbnailMultiple$1;
