// @nordic/rendering-context/src/server/scripts/getters.ts
import type {
  NordicRenderingContext,
  NordicRenderingContextAssets,
} from '../../common/types';

import { getWebviewInfo } from '@nordic/core-context';
import { assetsCache } from '@nordic/fs-cache';
import { serializeJSON } from '@nordic/utils/serialize';

import { getContext } from '../storage';
import {
  APP_INFO_SCRIPT,
  NATIVE_ACTION_DEBUG_SCRIPT,
  NATIVE_ACTION_SCRIPT,
  PIPEABLE_STREAM_SCRIPT,
} from './constants';

function getSerializableContext() {
  const context = getContext();
  const { assets, flags } = context;

  let { appProps } = context;

  const result: Omit<NordicRenderingContext, 'assets'> & {
    // Pick only serializable properties
    assets: Pick<NordicRenderingContextAssets, 'cdn' | 'prefix'>;
  } = {
    flags,
    assets: { cdn: assets.cdn, prefix: assets.prefix },
    support: context.support,
    title: context.title,
    framework: context.framework,
    nonce: context.nonce,
    csrfToken: context.csrfToken,
  };

  if (flags.staticMarkup) {
    return result;
  }

  appProps = flags.pipeableStream ? undefined : appProps;

  if (flags.islands) {
    appProps = appProps && { sharedState: appProps.sharedState };
  } else {
    result.wrapperProps = context.wrapperProps;
  }

  result.appProps = appProps;
  result.mainEntry = context.mainEntry;

  return result;
}

export function getRenderingContextInitScript() {
  const { mainEntry, flags } = getContext();

  let serializedManifest = '';
  let serializedMainScriptsNames = '';
  let serializedMainStylesNames = '';

  if (!flags.staticMarkup) {
    serializedManifest = assetsCache.serializedManifest;

    if (mainEntry) {
      serializedMainScriptsNames =
        assetsCache.getSerializedMainScriptsNames(mainEntry) ?? '';
      serializedMainStylesNames =
        assetsCache.getSerializedMainStylesNames(mainEntry) ?? '';
    }
  }

  return {
    id: '__NORDIC_RENDERING_CTX__',
    children:
      `_n.ctx.r=${serializeJSON(getSerializableContext())};` +
      `_n.ctx.r.assets.manifest=new Map(${serializedManifest});` +
      `_n.ctx.r.assets.mainAssetsNames={scripts:new Set(${serializedMainScriptsNames}),styles:new Set(${serializedMainStylesNames})};`,
  };
}

export function getNativeActionInitScripts() {
  const webkit = getWebviewInfo()?.webkit;

  if (webkit?.engine !== 'v3') {
    return null;
  }

  if (__DEV__ || webkit.debugMode) {
    return [APP_INFO_SCRIPT, NATIVE_ACTION_DEBUG_SCRIPT, NATIVE_ACTION_SCRIPT];
  }

  return [APP_INFO_SCRIPT, NATIVE_ACTION_SCRIPT];
}

export function getPipeableStreamInitScript() {
  const { appProps, flags } = getContext();

  if (
    !appProps ||
    !flags.pipeableStream ||
    flags.staticMarkup ||
    (flags.islands && !appProps.sharedState)
  ) {
    return null;
  }

  return PIPEABLE_STREAM_SCRIPT;
}
