// frontend-config/config/platforms.js
/**
 * List of available platforms on Mercado Libre
 *
 * Platforms description:
 *   ML - Mercado Libre
 *   MP - Mercado Pago
 *   TC - TuCarro
 *   PI - Portalinmobiliario: Real estate in CL only
 *   MS - Mercado Shops
 *   BO - BackOffice
 *
 * The next platforms are exists but still not configured:
 *   MClics - Mercado Libre Publicidad
 *   MEnvios - Mercado Envios
 *   Developrs - http://developers.mercadolibre.com/ mostly is a ML but with a different config, out of scope
 *
 * Platforms that are currently down
 *   GI - Guiadinmuebles: Real estate in MX only, currently is down to not compete with MT
 *   AP - AutoPlaza: is in state of shutting down, all data is migrating to ML
 *   TA - TuAvion: currently is alive but only in maintenance mode, will be merged with TC
 *
 * @type {string[]}
 */
const platforms = ['ML', 'MP', 'TC', 'PI', 'MS', 'BO'];

const defaultPlatform = 'ML';

/**
 *
 * @see https://api.mercadolibre.com/sites
 * @const {Object} sites
 */
const sites = {
  ML: [
    'MSV', 'MNI', 'MGT', 'MLC', 'MEC', 'MHN', 'MPE', 'MLU', 'MRD',
    'MCR', 'MLA', 'MBO', 'MCO', 'MLV', 'MPA', 'MLM', 'MPY', 'MLB', 'CBT',
  ],
  MP: ['MLA', 'MLB', 'MLC', 'MCO', 'MLM', 'MLV', 'MPE', 'MLU', 'MEC', 'CBT', 'MUS'],
  TC: ['MCO', 'MLV'],
  PI: ['MLC'],
  MS: ['MLA', 'MLC', 'MCO', 'MLB', 'MLV', 'MLM'],
  BO: ['ALL'],
};

const defaultSite = 'MLA';

exports = module.exports = { // eslint-disable-line no-multi-assign
  platforms,
  defaultPlatform,
  defaultSite,
  sites,
};
