// node-melitk-otel/lib/events/processor/batch/batchAllSpanProcessor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchAllSpanProcessor = void 0;
const api_1 = require("@opentelemetry/api");
const core_1 = require("@opentelemetry/core");
const telemetry_1 = __importDefault(require("./telemetry"));
const configs_1 = require("../../../configs");
/**
 * Represents the BatchAllSpanProcessor class which extends SpanProcessor.
 * This class is responsible for processing spans in batches.
 * @class
 * @extends {SpanProcessor}
 */
class BatchAllSpanProcessor {
    _exporter;
    _activator;
    _maxExportBatchSize;
    _maxQueueSize;
    _scheduledDelayMillis;
    _exportTimeoutMillis;
    _shutdownOnce;
    _finishedSpans = [];
    _isExporting = false;
    _exportingSize = 0;
    _droppedSpansCount = 0;
    _timer;
    _telemetry;
    /**
     * Creates a new instance of the BatchAllSpanProcessor class.
     * @param {SpanExporter} _exporter The span exporter.
     * @param {Activator} _activator The activator.
     * @param {ProcessorConfig} _config The processor config.
     */
    constructor(_exporter, _activator, _config) {
        this._exporter = _exporter;
        this._activator = _activator;
        this._maxExportBatchSize = +_config.maxExportBatchSize;
        this._maxQueueSize = +_config.maxQueueSize;
        this._scheduledDelayMillis = +_config.scheduleDelay;
        this._exportTimeoutMillis = +_config.exportTimeout;
        this._shutdownOnce = new core_1.BindOnceFuture(async () => this._shutdown(), this);
        if (this._maxExportBatchSize > this._maxQueueSize) {
            this._maxExportBatchSize = this._maxQueueSize;
        }
        this._telemetry = new telemetry_1.default();
        this._telemetry.observeQueueSizeMetric(() => this._finishedSpans.length);
        this._telemetry.observeQueueCapacityMetric(this._maxQueueSize);
        this._telemetry.observeBatchSizeMetric(() => this._exportingSize);
        this._telemetry.observeBatchCapacityMetric(this._maxExportBatchSize);
    }
    async forceFlush() {
        if (this._shutdownOnce.isCalled) {
            return this._shutdownOnce.promise;
        }
        return this._flushAll();
    }
    /** Method {@link onStart} is called when a span is started. */
    onStart() { }
    /**
     * Method {@link onEnd} is called when a span is finished. It enqueues activator ReadableSpan for later processing.
     *
     * @returns void.
     */
    onEnd(span) {
        // Do not enqueue spans if the activator is not active.
        if (!this._activator.isActive()) {
            return;
        }
        // Do not enqueue spans after shutdown.
        if (this._shutdownOnce.isCalled) {
            return;
        }
        this._addToBuffer(span);
    }
    /**
     * Method {@link shutdown} flushes the queue and waits until all spans are processed.
     * It only executes once. Subsequent call does nothing.
     *
     * @returns Promise<void>.
     */
    async shutdown() {
        return this._shutdownOnce.call();
    }
    /**
     * Method {@link _shutdown} flushes the queue and shuts down the exporter.
     *
     * @returns Promise<void>.
     */
    async _shutdown() {
        return Promise.resolve()
            .then(async () => this._flushAll())
            .then(async () => this._exporter.shutdown());
    }
    /**
     * Method {@link _addToBuffer} adds a span in the buffer.
     * @param span {@link ReadableSpan} to be added to the buffer.
     * @returns void
     */
    _addToBuffer(span) {
        if (this._finishedSpans.length >= this._maxQueueSize) {
            // limit reached, drop span
            this._droppedSpansCount += 1;
            return;
        }
        if (this._droppedSpansCount > 0) {
            this._droppedSpansCount = 0;
        }
        this._finishedSpans.push(span);
        this._maybeStartTimer();
    }
    /**
     * Method {@link _flushAll} send all spans to the exporter respecting the batch size limit.
     * This function is used only on forceFlush or shutdown,
     * for all other cases _flush should be used
     * */
    async _flushAll() {
        return new Promise((resolve, reject) => {
            const startTime = new Date();
            let isSuccess = true;
            const promises = [];
            // calculate the number of batches
            const count = Math.ceil(this._finishedSpans.length / this._maxExportBatchSize);
            for (let i = 0, j = count; i < j; i += 1) {
                promises.push(this._flushOneBatch());
            }
            Promise.all(promises)
                .then(() => {
                resolve();
            })
                .catch((value) => {
                isSuccess = false;
                reject(value);
            })
                .finally(() => {
                const endTime = new Date();
                const timeElapsed = startTime.getTime() - endTime.getTime();
                this._telemetry.recordExportQueueTime(timeElapsed, isSuccess);
            });
        });
    }
    /**
     * Method {@link _flushOneBatch} sends one batch of spans to the exporter.
     * This function is used only on forceFlush or shutdown,
     * for all other cases {@link _flushAll} should be used.
     * @returns {Promise<void>}
     */
    async _flushOneBatch() {
        this._clearTimer();
        if (this._finishedSpans.length === 0) {
            return Promise.resolve();
        }
        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                // don't wait anymore for export, this way the next batch can start
                reject(new Error('export timed out'));
            }, this._exportTimeoutMillis);
            // prevent downstream exporter calls from generating spans
            api_1.context.with((0, core_1.suppressTracing)(api_1.context.active()), () => {
                // reset the finished spans buffer here because the next invocations of the _flush method
                // could pass the same finished spans to the exporter if the buffer is cleared
                // outside the execution of this callback.
                let spans;
                if (this._finishedSpans.length <= this._maxExportBatchSize) {
                    spans = this._finishedSpans;
                    this._finishedSpans = [];
                }
                else {
                    spans = this._finishedSpans.splice(0, this._maxExportBatchSize);
                }
                const doExport = () => this._exporter.export(spans, (result) => {
                    this._exportingSize += spans.length;
                    clearTimeout(timer);
                    this._telemetry.incrementDroppedSpansMetric(this._droppedSpansCount);
                    this._telemetry.incrementProcessedSpansMetric(spans.length);
                    if (result.code === core_1.ExportResultCode.SUCCESS) {
                        this._exportingSize -= spans.length;
                        this._telemetry.incrementExportSpansSentMetric(spans.length);
                        resolve();
                    }
                    else {
                        this._exportingSize -= spans.length;
                        this._telemetry.incrementExportSpansFailedMetric(spans.length);
                        api_1.diag.error(`${configs_1.O11Y_EVENTS_LOG_PREFIX}Export failed - code: ${result.code} - error: ${result.error}`);
                        reject(result.error ?? new Error('span export failed'));
                    }
                });
                let pendingResources = null;
                for (let i = 0, len = spans.length; i < len; i += 1) {
                    const span = spans[i];
                    if (span.resource.asyncAttributesPending &&
                        span.resource.waitForAsyncAttributes) {
                        if (pendingResources === null) {
                            pendingResources = [];
                        }
                        pendingResources.push(span.resource.waitForAsyncAttributes());
                    }
                }
                // avoid scheduling a promise to make the behaviour more predictable and easier to test
                if (pendingResources === null) {
                    doExport();
                }
                else {
                    Promise.all(pendingResources)
                        .then(doExport)
                        .catch((err) => {
                        api_1.diag.error(`${configs_1.O11Y_EVENTS_LOG_PREFIX}Export threw an exception`, err);
                        (0, core_1.globalErrorHandler)(err);
                        reject(err);
                    });
                }
            });
        });
    }
    /**
     * Method {@link _maybeStartTimer} starts the timer if the processor is not exporting.
     * If the processor is exporting, it does nothing.
     * @returns {void}
     */
    _maybeStartTimer() {
        if (this._isExporting) {
            return;
        }
        const flush = () => {
            this._isExporting = true;
            this._flushOneBatch()
                .catch((e) => {
                (0, core_1.globalErrorHandler)(e);
            })
                .finally(() => {
                this._isExporting = false;
                if (this._finishedSpans.length > 0) {
                    this._clearTimer();
                    this._maybeStartTimer();
                }
            });
        };
        // we only wait if the queue doesn't have enough elements yet
        if (this._finishedSpans.length >= this._maxExportBatchSize) {
            flush();
            return;
        }
        if (this._timer !== undefined) {
            return;
        }
        this._timer = setTimeout(() => flush(), this._scheduledDelayMillis);
        (0, core_1.unrefTimer)(this._timer);
    }
    _clearTimer() {
        if (this._timer !== undefined) {
            clearTimeout(this._timer);
            this._timer = undefined;
        }
    }
}
exports.BatchAllSpanProcessor = BatchAllSpanProcessor;
