// node-melitk-otel/lib/events/processor/batch/telemetry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const metric_1 = require("../../common/metric");
const config_1 = require("../../config/config");
/**
 * Represents the BatchAllSpanProcessorTelemetry class which extends O11yMetrics.
 * This class provides metrics related to the Batch All Span Processor's behavior.
 * @class
 * @extends {O11yMetrics}
 */
class BatchAllSpanProcessorTelemetry extends metric_1.O11yMetrics {
    /**
     * Represents the metric `bsp.spans.dropped`.
     * This metric represents the number of spans that couldn't be enqueued by the Batch All Span Processor because the queue was full.
     * @type {Counter}
     * @private
     * @readonly
     */
    _droppedSpansMetric;
    /**
     * Represents the metric `bsp.spans.processed`.
     * This metric represents the number of spans attempted to be exported by the Batch All Span Processor.
     * @type {Counter}
     * @private
     * @readonly
     */
    _processedSpansMetric;
    /**
     * Represents the metric `bsp.spans.processed.export.failed`.
     * This metric represents the number of processed spans by the Batch All Span Processor that failed to be exported.
     * @type {Counter}
     * @private
     * @readonly
     */
    _exportSpansFailedMetric;
    /**
     * Represents the metric `bsp.spans.processed.export.sent`.
     * This metric represents the number of spans successfully exported by the Batch All Span Processor.
     * @type {Counter}
     * @private
     * @readonly
     */
    _exportSpansSentMetric;
    /**
     * Represents the metric `bsp.queue.size`.
     * This metric represents the estimated number of spans enqueued by the Batch All Span Processor that are waiting to be exported.
     * @type {ObservableGauge}
     * @private
     * @readonly
     */
    _queueSizeMetric;
    /**
     * Represents the metric `bsp.queue.capacity`.
     * This metric represents the maximum number of spans that the Batch All Span Processor can enqueue at once.
     * @type {ObservableGauge}
     * @private
     * @readonly
     */
    _queueCapacityMetric;
    /**
     * Represents the metric `bsp.queue.batch.export.time`.
     * This metric represents the time taken to export a batch of spans queued.
     * @type {Histogram}
     * @private
     * @readonly
     */
    _queueExportTime;
    /**
     * Represents the metric `bsp.batch.size`.
     * This metric represents the estimated number of spans of the Batch All Span Processor that will be attempted to be exported at once.
     * @type {ObservableGauge}
     * @private
     * @readonly
     */
    _batchSizeMetric;
    /**
     * Represents the metric `bsp.batch.capacity`.
     * This metric represents the maximum number of spans that the Batch All Span Processor will attempt to export at once.
     * @type {ObservableGauge}
     * @private
     * @readonly
     */
    _batchCapacityMetric;
    /**
     * Creates an instance of BatchAllSpanProcessorTelemetry.
     */
    constructor() {
        super();
        this._droppedSpansMetric = this.newCounter('bsp.spans.dropped', {
            description: "This metric represents the number of spans that couldn't be enqueued by the Batch All Span Processor because the queue was full.",
            unit: '1',
        });
        this._processedSpansMetric = this.newCounter('bsp.spans.processed', {
            description: 'This metric represents the number of spans attempted to be exported by the Batch All Span Processor.',
            unit: '1',
        });
        this._exportSpansFailedMetric = this.newCounter('bsp.spans.processed.export.failed', {
            description: 'This metric represents the number of processed spans by the Batch All Span Processor that failed to be exported.',
            unit: '1',
        });
        this._exportSpansSentMetric = this.newCounter('bsp.spans.processed.export.sent', {
            description: 'This metric represents the number of spans successfully exported by the Batch All Span Processor.',
            unit: '1',
        });
        this._queueSizeMetric = this.newObservableGauge('bsp.queue.size', {
            description: 'This metric represents the estimated number of spans enqueued by the Batch All Span Processor that are waiting to be exported.',
            unit: '1',
        });
        this._queueCapacityMetric = this.newObservableGauge('bsp.queue.capacity', {
            description: 'This metric represents the maximum number of spans that the Batch All Span Processor can enqueue at once.',
            unit: '1',
        });
        this._queueExportTime = this.newHistogram('bsp.queue.batch.export.time', {
            description: 'This metric represents the time taken to export a batch of spans queued.',
            unit: 'ms',
        });
        this._batchSizeMetric = this.newObservableGauge('bsp.batch.size', {
            description: 'This metric represents the estimated number of spans of the Batch All Span Processor that will be attempted to be exported at once.',
            unit: '1',
        });
        this._batchCapacityMetric = this.newObservableGauge('bsp.batch.capacity', {
            description: 'This metric represents the maximum number of spans that the Batch All Span Processor will attempt to export at once.',
            unit: '1',
        });
    }
    /**
     * Increments the metric `bsp.spans.dropped` by the specified amount.
     * @param total The amount to increment the metric by.
     */
    incrementDroppedSpansMetric = (total) => {
        this._droppedSpansMetric.add(total, config_1.metricDefaultAttributes);
    };
    /**
     * Increments the metric `bsp.spans.processed` by the specified amount.
     * @param total The amount to increment the metric by.
     */
    incrementProcessedSpansMetric = (total) => {
        this._processedSpansMetric.add(total, config_1.metricDefaultAttributes);
    };
    /**
     * Increments the metric `bsp.spans.processed.export.failed` by the specified amount.
     * @param total The amount to increment the metric by.
     */
    incrementExportSpansFailedMetric = (total) => {
        this._exportSpansFailedMetric.add(total, config_1.metricDefaultAttributes);
    };
    /**
     * Increments the metric `bsp.spans.processed.export.sent` by the specified amount.
     * @param total The amount to increment the metric by.
     */
    incrementExportSpansSentMetric = (total) => {
        this._exportSpansSentMetric.add(total, config_1.metricDefaultAttributes);
    };
    /**
     * Observes the metric `bsp.queue.size` with the specified callback.
     * @param queueSize The callback function returning the queue size.
     */
    observeQueueSizeMetric = (queueSize) => {
        this._queueSizeMetric.addCallback((observer) => {
            observer.observe(queueSize(), config_1.metricDefaultAttributes);
        });
    };
    /**
     * Observes the metric `bsp.queue.capacity` with the specified callback.
     * @param queueCapacity The callback function returning the queue capacity.
     */
    observeQueueCapacityMetric = (queueCapacity) => {
        this._queueCapacityMetric.addCallback((observer) => {
            observer.observe(queueCapacity, config_1.metricDefaultAttributes);
        });
    };
    /**
     * Records the metric `bsp.queue.batch.export.time` with the specified time spent and success.
     * @param timeSpent The time spent to export a batch of spans queued.
     * @param success A boolean indicating whether the export was successful or not.
     */
    recordExportQueueTime = (timeSpent, success) => {
        this._queueExportTime.record(timeSpent, {
            success,
            ...config_1.metricDefaultAttributes,
        });
    };
    /**
     * Observes the metric `bsp.batch.size` with the specified callback.
     * @param batchSize The callback function returning the batch size.
     */
    observeBatchSizeMetric = (batchSize) => {
        this._batchSizeMetric.addCallback((observer) => {
            observer.observe(batchSize(), config_1.metricDefaultAttributes);
        });
    };
    /**
     * Observes the metric `bsp.batch.capacity` with the specified callback.
     * @param batchCapacity The callback function returning the batch capacity.
     */
    observeBatchCapacityMetric = (batchCapacity) => {
        this._batchCapacityMetric.addCallback((observer) => {
            observer.observe(batchCapacity, config_1.metricDefaultAttributes);
        });
    };
}
exports.default = BatchAllSpanProcessorTelemetry;
