// nordic-pages-discovery/lib/discovery/files-discovery.js
const { existsSync } = require('fs');

const { glob } = require('glob');

const { FILES_IGNORED_IN_DISCOVERY } = require('../constants/files-discovery');
const { FILE_EXTENSIONS_TO_DISCOVER_REGEX } = require('../constants/regexs');
const { getFileOwner } = require('../helpers/file/owner');
const {
  getFileFullPath,
  checkIfFileIsInRoot,
} = require('../helpers/file/path-detail');
const { FILE_OWNERS } = require('../constants/file');
const { getTranspiledPath } = require('../helpers/build');
const {
  compareToSortFilesByOwnerAndPage,
} = require('../helpers/file/comparators');

/**
 * Discovers all the files located in the received directory and classifies them
 * by owner (one of {@link FILE_OWNERS})
 */
class FilesDiscovery {
  /**
   * @private
   * @type { string } - Full path of the directory where to discover files
   */
  baseDir;

  /**
   * @private
   * @type { string } - Full path of the directory where to discover build files
   */
  buildDir;

  /**
   * @param { string } baseDir - Path of the directory where to discover files
   * @param { string | undefined } buildDir - Path of the directory where to discover build files
   */
  constructor(baseDir, buildDir) {
    this.baseDir = baseDir;
    this.buildDir = buildDir;

    /**
     * @type { string[] }
     */
    this.navigablePageFiles = [];
    /**
     * @type { string[] }
     */
    this.httpErrorPageFiles = [];
    /**
     * @type { string | undefined }
     */
    this.routerConfigFile = undefined;
    /**
     * @type { string | undefined }
     */
    this.baseHooksFile = undefined;

    this.discoverFiles();
  }

  /**
   * @private
   */
  discoverFiles() {
    /**
     * @type { string[] }
     */
    let discoveredFiles = [];

    // eslint-disable-next-line security/detect-non-literal-fs-filename
    if (existsSync(this.baseDir)) {
      discoveredFiles = glob
        .sync(FILE_EXTENSIONS_TO_DISCOVER_REGEX, {
          cwd: this.baseDir,
          ignore: FILES_IGNORED_IN_DISCOVERY,
        })
        .sort(compareToSortFilesByOwnerAndPage);
    }

    if (!discoveredFiles.length) {
      return;
    }

    this.classifyDiscoveredFilesByOwner(discoveredFiles);
  }

  /**
   * @private
   * Set the routerConfigFile property only when no other router config files
   * were found previously
   * @param { string[] } routerConfigInRoot - List of routerConfig files discovered
   */
  setRouterConfigFile(routerConfigInRoot) {
    if (
      routerConfigInRoot.length > 1 ||
      (routerConfigInRoot.length && !routerConfigInRoot[0].endsWith('.js'))
    ) {
      throw new Error(
        "Expected to find a single Nordic Page JS '_router.config.js' file " +
          `in the root of ${this.baseDir}.\n` +
          `Router config files found:\n- ${routerConfigInRoot.join('\n- ')}`,
      );
    }

    const [routerConfigPath] = routerConfigInRoot;

    if (!routerConfigPath) {
      return;
    }

    this.routerConfigFile = routerConfigPath;
  }

  /**
   * @private
   * Set the baseHooksFile property only when no other base hooks file
   * were found previously
   * @param { string[] } baseHooksInRoot - List of base hooks files discovered
   */
  setBaseHooksFile(baseHooksInRoot) {
    if (baseHooksInRoot.length <= 1) {
      const [baseHooksPath] = baseHooksInRoot;

      this.baseHooksFile = baseHooksPath
        ? getTranspiledPath(baseHooksPath, this.baseDir, this.buildDir) ||
          baseHooksPath
        : undefined;

      return;
    }

    throw new Error(
      'Expected to find a single Nordic Pages base server hooks file ' +
        `in the root of ${this.baseDir}.\n` +
        `Base server hooks files found:\n- ${baseHooksInRoot.join('\n- ')}`,
    );
  }

  /**
   * @private
   * @param { string[] } discoveredFiles
   */
  classifyDiscoveredFilesByOwner(discoveredFiles) {
    /**
     * @type { string[] }
     */
    const routerConfigInRoot = [];
    /**
     * @type { string[] }
     */
    const baseHooksInRoot = [];

    for (let index = 0; index < discoveredFiles.length; index += 1) {
      const filePath = discoveredFiles[index];
      const fileOwner = getFileOwner(filePath);

      if (fileOwner === FILE_OWNERS.navigablePage) {
        this.navigablePageFiles = discoveredFiles.slice(index);

        break;
      }

      switch (fileOwner) {
        case FILE_OWNERS.httpErrorPage: {
          this.httpErrorPageFiles.push(filePath);

          break;
        }
        case FILE_OWNERS.routerConfig: {
          if (checkIfFileIsInRoot(filePath)) {
            routerConfigInRoot.push(getFileFullPath(filePath, this.baseDir));
          }

          break;
        }
        case FILE_OWNERS.baseHooks: {
          if (checkIfFileIsInRoot(filePath)) {
            baseHooksInRoot.push(getFileFullPath(filePath, this.baseDir));
          }

          break;
        }

        default:
          break;
      }
    }

    this.setRouterConfigFile(routerConfigInRoot);
    this.setBaseHooksFile(baseHooksInRoot);
  }
}

module.exports = { FilesDiscovery };
