// nordic-pages-discovery/lib/discovery/nordic-pages-discovery.js
const path = require('path');
const { existsSync } = require('fs');

const { getDuplicatedPagesLog } = require('../helpers/duplicate');
const {
  classifyFilesByRoute,
  classifyFilesByStatusError,
} = require('../helpers/file/classifier');
const { informInvalidPaths } = require('../helpers/file/inconsistency');
const { getInfoFromViewPath } = require('../helpers/page-info');
const { FilesDiscovery } = require('./files-discovery');
const { NavigablePage } = require('../page/navigable-page');
const { HttpErrorPage } = require('../page/http-error-page');
const {
  NORDIC_PAGES_PATH,
  APP_BUILD_PATH,
} = require('../constants/files-discovery');
const { PAGE_MEMBERS, NAVIGABLE_PAGE_ID } = require('../constants/page');
const { getFileFullPath } = require('../helpers/file/path-detail');
const { getIslandsFromPage } = require('../islands/get-islands-from-file');
const { normalizePathToPageDataId } = require('../helpers/normalizers');

/**
 * @typedef { typeof import('../constants/file').FILE_OWNERS } FILE_OWNERS
 * @typedef { keyof typeof import('../constants/device').DEVICE_TYPES } DeviceType
 * @typedef { import('../page/paths').Paths } Paths
 * @typedef { { [device in DeviceType]?: Paths } } PagePathsByDevice
 * @typedef { { [pageId: string]: string[] | undefined } } ClassifiedFilesByPageId
 * @typedef { Record<string, {hydrate: import('../helpers/page-info').HydrateValidValues, canUseBootstrapPage: boolean, islands?: string[]}> } PageInfoByView
 * @typedef { import('../page/navigable-page').CatchAllMetadata? } CatchAllMetadata
 * @typedef { Array<{ source: Array<string | RegExp> | string | RegExp, destination: string }> } RouterRewrites
 * @typedef { { rewrites?: RouterRewrites } } RouterConfig
 * @typedef { import('../helpers/file-system/islands-manifest').Island } Island
 *
 * @typedef { object } PageTypeData
 * @property { typeof NavigablePage | typeof HttpErrorPage } PageConstructor
 * @property { 'route' | 'statusError' } pageId
 *
 * @typedef { object } ClassifierData
 * @property { (files: string[]) => ({} | ClassifiedFilesByPageId) } classifier
 * @property { string[] } filesById
 *
 * @typedef { object } BuildPagesParams
 * @property { PageTypeData } pageTypeData
 * @property { ClassifierData } classifierData
 * @property { string } discoveredPagesProp
 */

/**
 * Discovers the Nordic Pages of an application and the information they need to work properly
 */
class NordicPagesDiscovery {
  /**
   * @private
   * @type { string } Full path of the project root directory
   */
  #rootDir;

  /**
   * @private
   * @type { string | undefined } Full path of the directory of build files
   */
  #buildDir;

  /**
   * @private
   * @type { string } Full path of the directory where to discover the Nordic Pages
   */
  nordicPagesDir;

  /**
   * @private
   * @type { string | undefined } Full path of the directory where to discover Nordic Pages build files
   */
  nordicPagesBuildDir;

  /**
   * @private
   * @type { FilesDiscovery } Files discovered classified by owner (one of {@link FILE_OWNERS})
   */
  filesDiscovery;

  /**
   * @private
   * @type { Array<{ source: { route: string | RegExp, catchAllMetadata: CatchAllMetadata }, destination: string }> } Discovered Nordic Page navigable routes
   */
  discoveredRoutes;

  /**
   * @private
   * @type { { [route: string]: NavigablePage } }
   */
  navigablePagesData;

  /**
   * @param { string } dir - Relative path of the directory where to discover the Nordic Pages
   */
  constructor(dir) {
    this.#rootDir = process.cwd();

    this.nordicPagesDir = path.join(this.#rootDir, NORDIC_PAGES_PATH);

    this.#buildDir = path.join(APP_BUILD_PATH, dir);

    // eslint-disable-next-line security/detect-non-literal-fs-filename
    this.#buildDir = existsSync(this.#buildDir) ? this.#buildDir : undefined;

    this.nordicPagesBuildDir = this.#buildDir
      ? path.join(this.#buildDir, NORDIC_PAGES_PATH)
      : undefined;

    this.filesDiscovery = new FilesDiscovery(
      this.nordicPagesDir,
      this.nordicPagesBuildDir,
    );

    // Initialization discovered Nordic Pages props
    this.discoveredRoutes = [];
    this.navigablePagesData = {};
    /**
     * @type { NavigablePage[] }
     */
    this.discoveredNavigablePages = [];
    /**
     * @type { { [statusError: string]: { paths: PagePathsByDevice } } }
     */
    this.discoveredHttpErrorPages = {};
    /**
     * @type { string | undefined }
     */
    this.baseHooksFile = undefined;
    /**
     * @type { import('../helpers/file-system/islands-manifest').IslandsManifest }
     */
    this.clientIslands = {};
    /**
     * @type {RouterConfig | null}
     */
    this.routerConfig = this.readRouterConfig();

    this.discoverNordicPages();
  }

  /**
   * @private
   * Discovers the Nordic Pages and the information they need to work properly,
   * based on the filesDiscovery property value.
   * If duplicated pages are found in the process, a duplicated pages error is thrown.
   */
  discoverNordicPages() {
    const duplicatedNavigablePages = this.buildNavigablePagesData();
    const duplicatedHttpErrorPages = this.buildDiscoveredHttpErrorPages();

    this.baseHooksFile = this.filesDiscovery.baseHooksFile;

    if (duplicatedNavigablePages || duplicatedHttpErrorPages) {
      throw new Error(
        getDuplicatedPagesLog(
          duplicatedNavigablePages,
          duplicatedHttpErrorPages,
        ),
      );
    }

    this.setRoutesBasedOnRewrites();
    this.buildDiscoveredNavigablePages();
  }

  buildDiscoveredNavigablePages() {
    const routesAdded = new Set();

    this.discoveredRoutes.forEach(({ source, destination }) => {
      const navigablePage = this.navigablePagesData[destination];

      if (!navigablePage) {
        throw new Error(
          `No Nordic Page definition found for the rewrite destination "${destination}" defined in ${this.filesDiscovery.routerConfigFile}. ` +
            'Please check the router config and fix it following the router rewrites documentation: https://nordic.adminml.com/docs/nordic-page/router-rewrites',
        );
      }

      // Do not add routes already added with previous discovered route object
      if (routesAdded.has(source.route)) {
        return;
      }

      routesAdded.add(source.route);

      this.discoveredNavigablePages.push(
        new NavigablePage({
          ...navigablePage,
          route: source.route,
          catchAllMetadata: source.catchAllMetadata,
        }),
      );
    });
  }

  /**
   * Returns the Nordic Pages discovered and the information they need to work properly
   * @returns Discovered Nordic Pages and information they need
   */
  getDiscoveredNordicPages() {
    return {
      baseHooksFile: this.baseHooksFile,
      discoveredNavigablePages: this.discoveredNavigablePages,
      discoveredHttpErrorPages: this.discoveredHttpErrorPages,
      hasBuild: !!this.nordicPagesBuildDir,
    };
  }

  /**
   * @private
   * @param {BuildPagesParams} buildPagesParams
   * @returns { { [pageId: string]: string[] } | undefined } Duplicated page files by page id (if any was found) or empty object
   */
  buildDiscoveredPagesByType({
    pageTypeData,
    classifierData,
    discoveredPagesProp,
  }) {
    const { classifier, filesById } = classifierData;
    const classifiedFiles = classifier(filesById);

    const { PageConstructor, pageId } = pageTypeData;
    const duplicatedPages = Object.entries(classifiedFiles).reduce(
      (duplicatedAcc, [id, files]) => {
        const { catchAllMetadata, [PAGE_MEMBERS.view]: pageViewFiles } = files;

        const page = new PageConstructor({
          [pageId]: id,
          catchAllMetadata,
        });

        const pageInfoByViewFile = this.getPageInfoByViewFiles(pageViewFiles);

        page.addPageFiles({
          files,
          pageInfoByViewFile,
          pagesDir: this.nordicPagesDir,
          pagesBuildDir: this.nordicPagesBuildDir,
        });

        page.completePaths(this.nordicPagesDir, this.nordicPagesBuildDir);

        if (!Object.keys(page.paths).length) {
          return duplicatedAcc;
        }

        let dataId = null;

        if (pageId === NAVIGABLE_PAGE_ID) {
          dataId = normalizePathToPageDataId(page.paths.base.view);
        }

        const duplicatedFiles = page.checkIfDuplicated(
          this[discoveredPagesProp],
          dataId,
        );

        if (page.validate()) {
          if (pageId === NAVIGABLE_PAGE_ID) {
            this[discoveredPagesProp][dataId] = page;
            this.discoveredRoutes.push({
              source: { route: id, catchAllMetadata },
              destination: dataId,
            });
          } else {
            this[discoveredPagesProp][id] = {
              paths: page.paths,
            };
          }
        } else {
          informInvalidPaths(page);
        }

        if (duplicatedFiles.length) {
          duplicatedAcc[id] = duplicatedFiles;
        }

        return duplicatedAcc;
      },
      {},
    );

    return Object.keys(duplicatedPages).length ? duplicatedPages : undefined;
  }

  /**
   * @param {string[]} files
   * @returns {PageInfoByView}
   */
  getPageInfoByViewFiles(files) {
    const pageInfoByView = Object.fromEntries(
      files.map((filePath) => {
        const fullPath = getFileFullPath(filePath, this.nordicPagesDir);
        const { hydrate, canUseBootstrapPage } = getInfoFromViewPath(fullPath);

        if (hydrate !== 'islands') {
          return [filePath, { hydrate, canUseBootstrapPage }];
        }

        const islands = getIslandsFromPage(fullPath);

        Object.assign(this.clientIslands, islands);

        return [
          filePath,
          { hydrate, canUseBootstrapPage, islands: Object.keys(islands) },
        ];
      }),
    );

    return pageInfoByView;
  }

  /**
   * @private
   * Builds the navigable pages data of the application based on the navigable pages files
   * of the filesDiscovery property.
   * When a page is built and is valid, it is added to the navigablePagesData property
   * @returns { { [route: string]: string[] } | undefined } Duplicated page files by route (if any was found) or empty object
   */
  buildNavigablePagesData() {
    /** @type { BuildPagesParams['pageTypeData'] } */
    const pageTypeData = {
      PageConstructor: NavigablePage,
      pageId: NAVIGABLE_PAGE_ID,
    };
    const classifierData = {
      classifier: classifyFilesByRoute,
      filesById: this.filesDiscovery.navigablePageFiles,
    };

    return this.buildDiscoveredPagesByType({
      pageTypeData,
      classifierData,
      discoveredPagesProp: 'navigablePagesData',
    });
  }

  /**
   * @private
   * Builds the error pages of the application based on the http error pages files
   * of the filesDiscovery property.
   * When a page is built and is valid, it is added to the discoveredHttpErrorPages property
   * @returns { { [statusError: string]: string[] } | undefined } Duplicated page files by status error (if any was found) or empty object
   */
  buildDiscoveredHttpErrorPages() {
    /** @type { BuildPagesParams['pageTypeData'] } */
    const pageTypeData = {
      PageConstructor: HttpErrorPage,
      pageId: 'statusError',
    };
    const classifierData = {
      classifier: classifyFilesByStatusError,
      filesById: this.filesDiscovery.httpErrorPageFiles,
    };

    return this.buildDiscoveredPagesByType({
      pageTypeData,
      classifierData,
      discoveredPagesProp: 'discoveredHttpErrorPages',
    });
  }

  /**
   * @returns {RouterConfig}
   */
  readRouterConfig() {
    const { routerConfigFile } = this.filesDiscovery;

    if (!routerConfigFile) {
      return null;
    }

    try {
      // eslint-disable-next-line n/global-require, import/no-dynamic-require
      return require(routerConfigFile);
    } catch (e) {
      return null;
    }
  }

  setRoutesBasedOnRewrites() {
    const rewritesDiscoveredRoutes = [];

    this.routerConfig?.rewrites?.forEach(({ source, destination }, index) => {
      if (typeof destination !== 'string') {
        throw new Error(
          'Rewrite destinations defined in the router config file should be of type string ' +
            `but destination of type '${typeof destination}' was found instead in the position ${index} of the config array.\nFile: ${this.filesDiscovery.routerConfigFile}\n` +
            `Please check the router rewrites API documentation: https://nordic.adminml.com/docs/api/Nordic%20Features/nordic-page/router-rewrites#destination`,
        );
      }

      const sources = Array.isArray(source) ? source : [source];

      sources.forEach((route) => {
        if (typeof route !== 'string' && !(route instanceof RegExp)) {
          throw new Error(
            'Rewrite sources defined in the router config file should be of type string or a regular expression ' +
              `but source of type '${typeof route}' was found in the position ${index} of the router config array.\nFile: ${this.filesDiscovery.routerConfigFile}\n` +
              'Please check the router rewrites API documentation: https://nordic.adminml.com/docs/api/Nordic%20Features/nordic-page/router-rewrites#source',
          );
        }

        rewritesDiscoveredRoutes.push({
          source: {
            route,
            catchAllMetadata: null,
          },
          destination,
        });
      });
    });

    this.discoveredRoutes.unshift(...rewritesDiscoveredRoutes);
  }
}

module.exports = { NordicPagesDiscovery };
