// @meli/input-validation/dist/iv/entity/ivArray.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IvArray = void 0;
const zod_1 = require("zod");
const input_validation_storage_1 = require("@meli/input-validation-storage");
const metrics_1 = require("../../metrics");
const types_1 = require("../../utils/types");
const ModifierChaining_1 = __importDefault(require("../wrapper/ModifierChaining"));
const PropertyDisabler_1 = __importDefault(require("../wrapper/PropertyDisabler"));
const validationLogger_1 = require("../../utils/validationLogger");
const constants_1 = require("../../utils/constants");
const chooseLengthConfig = (value) => value !== null && value !== undefined ? { value } : null;
class IvArray extends zod_1.z.ZodArray {
    constructor(schema, options) {
        super({
            type: schema,
            typeName: zod_1.z.ZodFirstPartyTypeKind.ZodArray,
            exactLength: chooseLengthConfig(options?.exactLength),
            minLength: chooseLengthConfig(options?.minLength),
            maxLength: chooseLengthConfig(options?.maxLength),
        });
        ModifierChaining_1.default.applyModifiers(this);
        ModifierChaining_1.default.applyArrayModifiers(this);
        PropertyDisabler_1.default.disableProperties(this);
    }
    validate(input, options = {}) {
        const store = input_validation_storage_1.ValidationState.current();
        const { traceRequestId, disableLogs } = options;
        try {
            if (!store) {
                const validation = this.safeParse(input);
                if (validation.success) {
                    (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                        'result:valid',
                        'method:validate',
                        `sdk_version:${constants_1.SDK_VERSION}`,
                    ]);
                }
                else {
                    if (!disableLogs) {
                        (0, validationLogger_1.logValidationError)(validation.error.errors, traceRequestId);
                    }
                    (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                        'result:invalid',
                        'method:validate',
                        `sdk_version:${constants_1.SDK_VERSION}`,
                    ]);
                }
                return validation.success;
            }
            const scope = input_validation_storage_1.ValidationState.getCurrentScope(input, options);
            const inputValidationState = store.get(input) ?? {
                state: false,
                fullyValidated: false,
                validatedProps: new Set(),
                scopes: new Map(),
            };
            inputValidationState.state = true;
            store.set(input, inputValidationState);
            const validation = this.safeParse(input);
            if (validation.success) {
                input_validation_storage_1.ValidationState.markFullyValidated(input, scope);
                (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                    'result:valid',
                    'method:validate',
                    `sdk_version:${constants_1.SDK_VERSION}`,
                ]);
            }
            else {
                if (!disableLogs) {
                    (0, validationLogger_1.logValidationError)(validation.error.errors, traceRequestId);
                }
                (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                    'result:invalid',
                    'method:validate',
                    `sdk_version:${constants_1.SDK_VERSION}`,
                ]);
            }
            return validation.success;
        }
        finally {
            if (store) {
                const inputValidationState = store.get(input) ?? {
                    state: false,
                    fullyValidated: false,
                    validatedProps: new Set(),
                    scopes: new Map(),
                };
                inputValidationState.state = false;
                store.set(input, inputValidationState);
            }
        }
    }
}
exports.IvArray = IvArray;
const ivArray = (schema, options) => {
    const zodSchema = (0, types_1.resolveSchema)(schema, 'Array');
    return new IvArray(zodSchema, options);
};
exports.default = ivArray;
