// @meli/input-validation/dist/iv/entity/ivObject.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IvObject = void 0;
const zod_1 = require("zod");
const input_validation_storage_1 = require("@meli/input-validation-storage");
const metrics_1 = require("../../metrics");
const config_1 = require("../../config");
const PropertyDisabler_1 = __importDefault(require("../wrapper/PropertyDisabler"));
const ModifierChaining_1 = __importDefault(require("../wrapper/ModifierChaining"));
const validationLogger_1 = require("../../utils/validationLogger");
const constants_1 = require("../../utils/constants");
const environment_detector_1 = require("../../utils/environment-detector");
class IvObject {
    schema;
    constructor(schema) {
        this.schema = schema;
        PropertyDisabler_1.default.disableProperties(this);
        ModifierChaining_1.default.applyModifiers(this);
    }
    unwrapObjectSchema() {
        let schema = this.schema;
        // Unwraps modifiers like .optional() or .nullable()
        if (schema instanceof zod_1.z.ZodOptional || schema instanceof zod_1.z.ZodNullable) {
            schema = schema.unwrap();
        }
        if (schema instanceof zod_1.ZodObject) {
            return schema;
        }
        throw new Error('This method can only be used on an object-based schema.');
    }
    get shape() {
        return this.unwrapObjectSchema().shape;
    }
    get rawShape() {
        return this.unwrapObjectSchema().shape;
    }
    validate(object, options = {}) {
        const store = input_validation_storage_1.ValidationState.current();
        const { traceRequestId, disableLogs } = options;
        try {
            if (!store) {
                const validation = this.schema.safeParse(object);
                if (validation.success) {
                    (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                        'result:valid',
                        'method:validate',
                        `sdk_version:${constants_1.SDK_VERSION}`,
                    ]);
                }
                else {
                    if (!disableLogs) {
                        (0, validationLogger_1.logValidationError)(validation.error.errors, traceRequestId);
                    }
                    (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                        'result:invalid',
                        'method:validate',
                        `sdk_version:${constants_1.SDK_VERSION}`,
                    ]);
                }
                return validation.success;
            }
            const scope = input_validation_storage_1.ValidationState.getCurrentScope(object, options);
            const inputValidationState = store.get(object) ?? {
                state: false,
                fullyValidated: false,
                validatedProps: new Set(),
                scopes: new Map(),
            };
            inputValidationState.state = true;
            store.set(object, inputValidationState);
            if (this.schema instanceof zod_1.z.ZodObject) {
                const newProps = new Set(Object.keys(this.schema.shape));
                const scopedProps = scope === 'GLOBAL'
                    ? inputValidationState.validatedProps
                    : inputValidationState.scopes.get(scope)?.validatedProps ||
                        new Set();
                newProps.forEach((prop) => scopedProps.add(prop));
                if (scope !== 'GLOBAL') {
                    if (!inputValidationState.scopes.has(scope)) {
                        inputValidationState.scopes.set(scope, {
                            validatedProps: scopedProps,
                        });
                    }
                }
            }
            const validation = this.schema.safeParse(object);
            if (validation.success) {
                (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                    'result:valid',
                    'method:validate',
                    `sdk_version:${constants_1.SDK_VERSION}`,
                ]);
            }
            else {
                if (!disableLogs) {
                    (0, validationLogger_1.logValidationError)(validation.error.errors, traceRequestId);
                }
                (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                    'result:invalid',
                    'method:validate',
                    `sdk_version:${constants_1.SDK_VERSION}`,
                ]);
                if (store) {
                    if (scope === 'GLOBAL') {
                        inputValidationState.validatedProps.clear();
                    }
                    else {
                        inputValidationState.scopes.get(scope)?.validatedProps?.clear();
                    }
                }
            }
            return validation.success;
        }
        finally {
            if (store) {
                const inputValidationState = store.get(object) ?? {
                    state: false,
                    fullyValidated: false,
                    validatedProps: new Set(),
                    scopes: new Map(),
                };
                inputValidationState.state = false;
                store.set(object, inputValidationState);
            }
        }
    }
    safeParse(input, options = {}) {
        const store = input_validation_storage_1.ValidationState.current();
        const { traceRequestId, disableLogs } = options;
        const isTrackable = input !== null && typeof input === 'object';
        const trackedInput = input;
        try {
            if (!store || !isTrackable) {
                const validation = this.schema.safeParse(input);
                if (validation.success) {
                    (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                        'result:valid',
                        'method:safeParse',
                        `sdk_version:${constants_1.SDK_VERSION}`,
                    ]);
                }
                else {
                    if (!disableLogs) {
                        (0, validationLogger_1.logValidationError)(validation.error.errors, traceRequestId);
                    }
                    (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                        'result:invalid',
                        'method:safeParse',
                        `sdk_version:${constants_1.SDK_VERSION}`,
                    ]);
                }
                return validation;
            }
            const scope = input_validation_storage_1.ValidationState.getCurrentScope(trackedInput, options);
            const inputValidationState = store.get(trackedInput) ?? {
                state: false,
                fullyValidated: false,
                validatedProps: new Set(),
                scopes: new Map(),
            };
            inputValidationState.state = true;
            store.set(trackedInput, inputValidationState);
            if (this.schema instanceof zod_1.z.ZodObject) {
                const newProps = new Set(Object.keys(this.schema.shape));
                const scopedProps = scope === 'GLOBAL'
                    ? inputValidationState.validatedProps
                    : inputValidationState.scopes.get(scope)?.validatedProps ||
                        new Set();
                newProps.forEach((prop) => scopedProps.add(prop));
                if (scope !== 'GLOBAL') {
                    if (!inputValidationState.scopes.has(scope)) {
                        inputValidationState.scopes.set(scope, {
                            validatedProps: scopedProps,
                        });
                    }
                }
            }
            const validation = this.schema.safeParse(input);
            if (validation.success) {
                (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                    'result:valid',
                    'method:safeParse',
                    `sdk_version:${constants_1.SDK_VERSION}`,
                ]);
            }
            else {
                if (!disableLogs) {
                    (0, validationLogger_1.logValidationError)(validation.error.errors, traceRequestId);
                }
                (0, metrics_1.increment)('node_sdk_input_validation.consume', 1, [
                    'result:invalid',
                    'method:safeParse',
                    `sdk_version:${constants_1.SDK_VERSION}`,
                ]);
                if (scope === 'GLOBAL') {
                    inputValidationState.validatedProps.clear();
                }
                else {
                    inputValidationState.scopes.get(scope)?.validatedProps?.clear();
                }
            }
            return validation;
        }
        finally {
            if (store && isTrackable) {
                const inputValidationState = store.get(trackedInput) ?? {
                    state: false,
                    fullyValidated: false,
                    validatedProps: new Set(),
                    scopes: new Map(),
                };
                inputValidationState.state = false;
                store.set(trackedInput, inputValidationState);
            }
        }
    }
    and(merger) {
        const mergedShape = {};
        Object.assign(mergedShape, this.unwrapObjectSchema().shape);
        if (merger instanceof IvObject) {
            Object.assign(mergedShape, merger.rawShape);
        }
        else {
            Object.assign(mergedShape, merger);
        }
        return new IvObject(zod_1.z.object(mergedShape));
    }
    merge(merger) {
        const baseSchema = this.unwrapObjectSchema();
        const mergedSchema = baseSchema.merge(zod_1.z.object(merger));
        return new IvObject(mergedSchema);
    }
    strip() {
        if ((0, environment_detector_1.isNotProduction)()) {
            console.warn('⚠️ DEPRECATION WARNING: .strip() is deprecated and will be removed in a future version. ' +
                'Use .strict() instead. Note: .strip() removes unknown properties silently, ' +
                'while .strict() rejects objects with unknown properties and throws validation errors.');
        }
        const baseSchema = this.unwrapObjectSchema();
        return new IvObject(baseSchema.strip());
    }
    strict() {
        const baseSchema = this.unwrapObjectSchema();
        return new IvObject(baseSchema.strict());
    }
    required() {
        const baseSchema = this.unwrapObjectSchema();
        return new IvObject(baseSchema.required());
    }
    extends(extension) {
        const baseSchema = this.unwrapObjectSchema();
        const extendedSchema = baseSchema.extend(extension);
        return new IvObject(extendedSchema);
    }
    or(input) {
        const rightSchema = input instanceof IvObject ? input.getSchema() : input;
        const unionSchema = this.schema.or(rightSchema);
        return new IvObject(unionSchema);
    }
    optional() {
        const optionalSchema = this.schema.optional();
        return new IvObject(optionalSchema);
    }
    nullable() {
        const nullableSchema = this.schema.nullable();
        return new IvObject(nullableSchema);
    }
    nullish() {
        const nullishSchema = this.schema.nullish();
        return new IvObject(nullishSchema);
    }
    getSchema() {
        return this.schema;
    }
    /**
     * @deprecated Use `createValidationMiddleware` for Express routes, or `.validate()` / `.safeParse()` for direct validation.
     * Will be removed in a future major version. In non-production environments a runtime warning is emitted.
     */
    parse(input) {
        if ((0, environment_detector_1.isNotProduction)()) {
            console.warn('⚠️ DEPRECATION WARNING: .parse() is deprecated and will be removed in a future version. ' +
                'Use createValidationMiddleware for Express routes, or .validate() / .safeParse() for direct validation.');
        }
        return this.schema.parse(input);
    }
    // ❌ MÉTODOS NO UTILIZADOS - Comentados para referencia
    // Estos métodos no están documentados en README y no se usan en tests
    // Se mantienen comentados por si se necesitan en el futuro
    // transform<U>(transformFn: (input: any) => U): IvObject<ZodEffects<T>> {
    //   const transformedSchema = this.schema.transform(transformFn);
    //   return new IvObject(transformedSchema);
    // }
    refine(refinement, refinementOptions) {
        const refinedSchema = this.schema.refine(refinement, refinementOptions);
        return new IvObject(refinedSchema);
    }
}
exports.IvObject = IvObject;
const ivObject = (shape) => {
    if (!shape || typeof shape !== 'object') {
        throw new Error('Schema must be a valid object');
    }
    const processedShape = {};
    Object.entries(shape).forEach(([key, value]) => {
        if (value instanceof IvObject) {
            processedShape[key] = value.getSchema();
        }
        else if (value instanceof zod_1.ZodType) {
            processedShape[key] = value;
        }
        else {
            throw new Error(`Invalid schema for key "${key}". Expected IvObject.`);
        }
    });
    const schema = config_1.STRICT_MODE
        ? zod_1.z.object(processedShape).strict()
        : zod_1.z.object(processedShape);
    return new IvObject(schema);
};
exports.default = ivObject;
