// @meli/input-validation/dist/iv/entity/ivString.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IvString = void 0;
/* eslint-disable @typescript-eslint/method-signature-style */
const zod_1 = require("zod");
const metrics_1 = require("../../metrics");
const config_1 = require("../../config");
const stringSecurityState_1 = __importStar(require("../../utils/stringSecurityState"));
const ModifierChaining_1 = __importDefault(require("../wrapper/ModifierChaining"));
const PropertyDisabler_1 = __importDefault(require("../wrapper/PropertyDisabler"));
const constants_1 = require("../../utils/constants");
class IvString extends zod_1.z.ZodString {
    /**
     * Shared security state across the entire chain. All IvString instances
     * created via _addCheck share the same reference, so the final security
     * posture is evaluated after the full chain is built.
     */
    securityState;
    constructor(props, securityState) {
        super(props);
        this.securityState = securityState ?? new stringSecurityState_1.default();
        // eslint-disable-next-line no-underscore-dangle
        this._addCheck = (check) => {
            const checks = Array.isArray(check) ? check : [check];
            // Detect security implications of each new check
            checks.forEach((c) => {
                const { kind } = c;
                if (kind === 'regex') {
                    // SECURE_REGEX (from .secure()) is the only known-safe regex
                    const isSecure = 'regex' in c &&
                        c.regex.source === config_1.SECURE_REGEX.source &&
                        c.regex.flags === config_1.SECURE_REGEX.flags;
                    if (isSecure) {
                        this.securityState.markSecure();
                    }
                    else {
                        this.securityState.markInsecure(kind);
                    }
                }
                else if (stringSecurityState_1.SECURE_CHECK_KINDS.has(kind)) {
                    this.securityState.markSecure();
                }
                else if (stringSecurityState_1.INSECURE_CHECK_KINDS.has(kind)) {
                    this.securityState.markInsecure(kind);
                }
                // Other checks (e.g. base64) are intentionally ignored — no metric
            });
            return new IvString({
                // eslint-disable-next-line no-underscore-dangle
                ...this._def,
                checks: [
                    // eslint-disable-next-line no-underscore-dangle
                    ...this._def.checks,
                    ...checks,
                ],
            }, this.securityState);
        };
        /**
         * Applies SECURE_REGEX to protect against malicious character combinations.
         * Emits the legacy string.secure metric immediately for backward compatibility.
         * Security classification (type.secure) is handled by _addCheck via StringSecurityState.
         */
        this.secure = () => {
            (0, metrics_1.increment)('node_sdk_input_validation.string.secure', 1, [
                'method:secure',
                `sdk_version:${constants_1.SDK_VERSION}`,
                'secure:true',
            ]);
            return this.regex(config_1.SECURE_REGEX);
        };
        this.unsafe = () => {
            this.securityState.markUnsafe();
            return this;
        };
        ModifierChaining_1.default.applyModifiers(this);
        PropertyDisabler_1.default.disableProperties(this);
    }
    regex(regex, message) {
        return super.regex(regex, message);
    }
}
exports.IvString = IvString;
const ivString = () => new IvString({
    checks: [],
    typeName: zod_1.z.ZodFirstPartyTypeKind.ZodString,
    coerce: false,
});
exports.default = ivString;
