// @andes/common/build/hooks/useTransition.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var react = require('react');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
const PRE_ENTER = 0;
const ENTERING = 1;
const ENTERED = 2;
const PRE_EXIT = 3;
const EXITING = 4;
const EXITED = 5;
const UNMOUNTED = 6;
const STATUS = ['preEnter', 'entering', 'entered', 'preExit', 'exiting', 'exited', 'unmounted'];
const getState = status => ({
  _s: status,
  status: STATUS[status],
  isEnter: status < PRE_EXIT,
  isMounted: status !== UNMOUNTED,
  isResolved: status === ENTERED || status > EXITING
});
const startOrEnd = unmounted => unmounted ? UNMOUNTED : EXITED;
const getEndStatus = (status, unmountOnExit) => {
  switch (status) {
    case ENTERING:
    case PRE_ENTER:
      return ENTERED;
    case EXITING:
    case PRE_EXIT:
      return startOrEnd(unmountOnExit);
  }
};
const getTimeout = timeout => typeof timeout === 'object' ? [timeout.enter, timeout.exit] : [timeout, timeout];
const nextTick = (transitState, status) => setTimeout(() => {
  Number.isNaN(document.body.offsetTop) || transitState(status + 1);
}, 0);
const updateState = (status, setState, latestState, timeoutId, onChange) => {
  clearTimeout(timeoutId.current);
  const state = getState(status);
  setState(state);
  latestState.current = state;
  if (typeof onChange === 'function') {
    onChange({
      current: state
    });
  }
};
const useTransition = function () {
  let _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
    _ref$enter = _ref.enter,
    enter = _ref$enter === void 0 ? true : _ref$enter,
    _ref$exit = _ref.exit,
    exit = _ref$exit === void 0 ? true : _ref$exit,
    preEnter = _ref.preEnter,
    preExit = _ref.preExit,
    timeout = _ref.timeout,
    initialEntered = _ref.initialEntered,
    mountOnEnter = _ref.mountOnEnter,
    unmountOnExit = _ref.unmountOnExit,
    onChange = _ref.onStateChange;
  const _useState = react.useState(() => getState(initialEntered ? ENTERED : startOrEnd(mountOnEnter))),
    _useState2 = _slicedToArray__default["default"](_useState, 2),
    state = _useState2[0],
    setState = _useState2[1];
  const latestState = react.useRef(state);
  const timeoutId = react.useRef(undefined);
  const _getTimeout = getTimeout(timeout),
    _getTimeout2 = _slicedToArray__default["default"](_getTimeout, 2),
    enterTimeout = _getTimeout2[0],
    exitTimeout = _getTimeout2[1];
  const endTransition = react.useCallback(() => {
    const status = getEndStatus(latestState.current._s, unmountOnExit);
    if (status) {
      updateState(status, setState, latestState, timeoutId, onChange);
    }
  }, [onChange, unmountOnExit]);
  const toggle = react.useCallback(toEnter => {
    const transitState = status => {
      updateState(status, setState, latestState, timeoutId, onChange);
      switch (status) {
        case ENTERING:
          if (enterTimeout && enterTimeout >= 0) timeoutId.current = setTimeout(endTransition, enterTimeout);
          break;
        case EXITING:
          if (exitTimeout && exitTimeout >= 0) timeoutId.current = setTimeout(endTransition, exitTimeout);
          break;
        case PRE_ENTER:
        case PRE_EXIT:
          timeoutId.current = nextTick(transitState, status);
          break;
      }
    };
    const enterStage = latestState.current.isEnter;
    if (typeof toEnter !== 'boolean') toEnter = !enterStage;
    if (toEnter) {
      if (!enterStage) {
        transitState(enter ? preEnter ? PRE_ENTER : ENTERING : ENTERED);
      }
    } else if (enterStage) {
      transitState(exit ? preExit ? PRE_EXIT : EXITING : startOrEnd(unmountOnExit));
    }
  }, [endTransition, onChange, enter, exit, preEnter, preExit, enterTimeout, exitTimeout, unmountOnExit]);
  react.useEffect(() => () => clearTimeout(timeoutId.current), []);
  return [state, toggle, endTransition];
};
exports.useTransition = useTransition;
