// @andes/snackbar/build/cjs/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _extends = require('@babel/runtime/helpers/extends');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var React = require('react');
var utils = require('@andes/common/build/webkit/utils');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var PropTypes = require('prop-types');
var classNames = require('classnames');
var visuallyHidden = require('@andes/visually-hidden');
var includes = require('@andes/common/build/utils/includes');
var warning = require('@andes/common/build/utils/warning');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _objectWithoutProperties__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutProperties);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var includes__default = /*#__PURE__*/_interopDefaultLegacy(includes);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
const namespace = 'andes-snackbar';
const DEPRECATED_COLOR_GREEN = 'green';
const DEPRECATED_COLOR_RED = 'red';
const AVAILABLE_COLOR_POSITIVE = 'positive';
const AVAILABLE_COLOR_NEGATIVE = 'negative';
const DEPRECATED_COLOR = [DEPRECATED_COLOR_GREEN, DEPRECATED_COLOR_RED];
const AVAILABLE_COLOR = [AVAILABLE_COLOR_POSITIVE, AVAILABLE_COLOR_NEGATIVE];
const WARNING_COLOR_GREEN = `
  '${DEPRECATED_COLOR_GREEN}' option will be removed from 'color' options in a future major version.
  Please, use ${AVAILABLE_COLOR_POSITIVE}.`;
const WARNING_COLOR_RED = `
  '${DEPRECATED_COLOR_RED}' option will be removed from 'color' options in a future major version.
  Please, use ${AVAILABLE_COLOR_NEGATIVE}.`;
const DEPRECATED_COLOR_WARNINGS = {
  [`${DEPRECATED_COLOR_GREEN}`]: WARNING_COLOR_GREEN,
  [`${DEPRECATED_COLOR_RED}`]: WARNING_COLOR_RED
};

// This function attempts to keep backwards compatibility with type prop

/** @typedef {Pick<import('../..').SnackbarProps, 'color'>} ComputeColorProp */
/** @typedef {'negative' | 'positive' } ComputeColorResult */

/** @type {(props: ComputeColorProp) => ComputeColorResult | undefined} */
const computeTypeColor = _ref => {
  let color = _ref.color;
  if (includes__default["default"](['red', 'negative'], color || '')) return 'negative';
  if (includes__default["default"](['green', 'positive'], color || '')) return 'positive';
  return undefined;
};
var computeTypeColor$1 = computeTypeColor;

/** @typedef {import('..').ActionConfigType} ActionConfig */

/**
 * @typedef SnackbarState
 * @prop {boolean} display
 * @prop {boolean} show
 * @prop {boolean} prevShow
 * @prop {boolean} srLabelVisibility
 * @prop {boolean} timeoutExecuted
 */

/** @typedef {import('..').SnackbarProps} SnackbarPublicProps */

/**
 * @typedef {Object} InternalProps
 * @prop {React.Ref<HTMLDivElement>} [innerRef]
 * @prop {React.ReactNode} [children]
 * @prop {{ root?: string }} [classNames]
 * @prop {Record<string, unknown>} [andesDataAttributes]
 */

/** @typedef {SnackbarPublicProps & InternalProps} SnackbarProps */

/** @extends React.Component<SnackbarProps, SnackbarState> */
class Snackbar extends React__default["default"].Component {
  /** @param {SnackbarProps} props */
  constructor(props) {
    super(props);
    const _this$props$show = this.props.show,
      show = _this$props$show === void 0 ? false : _this$props$show;
    this.onActionClick = this.onActionClick.bind(this);
    this.onAnimationEnd = this.onAnimationEnd.bind(this);
    this.setMainContainerRef = this.setMainContainerRef.bind(this);
    /** @type {SnackbarState} */
    this.state = {
      display: show,
      show,
      prevShow: show,
      timeoutExecuted: false,
      srLabelVisibility: false
    };
    this.style = {};
  }
  componentDidMount() {
    const _this$props = this.props,
      _this$props$show2 = _this$props.show,
      show = _this$props$show2 === void 0 ? false : _this$props$show2,
      color = _this$props.color,
      delay = _this$props.delay,
      onOpen = _this$props.onOpen;

    // @deprecate [color] - Warning from: 9.0.0
    if (typeof color === 'string') {
      warning__default["default"](!includes__default["default"](DEPRECATED_COLOR, color), DEPRECATED_COLOR_WARNINGS[(/** @type {keyof DEPRECATED_COLOR_WARNINGS} */color)]);
    }
    const fixedElement = /** @type {HTMLElement} */document.querySelector('.andes-snackbar-fixed-element');
    if (fixedElement) {
      this.style = {
        bottom: `${fixedElement.offsetHeight + 16}px`
      };
    }
    if (show && delay) {
      this.delayTimeout = setTimeout(() => {
        this.setState({
          show: false
        }, () => {
          if (this.isReduceMotion()) this.onClose();
        });
      }, delay);
    }
    this.labelTimeout = setTimeout(() => {
      this.setState({
        srLabelVisibility: true
      });
    }, 30);
    if (show && typeof onOpen === 'function') {
      onOpen();
    }
  }
  componentWillUnmount() {
    clearTimeout(this.delayTimeout);
    clearTimeout(this.labelTimeout);
  }

  /**
   * @param {SnackbarState} prevState
   * @param {SnackbarProps} nextProps
   */
  static getDerivedStateFromProps(nextProps, prevState) {
    const _nextProps$show = nextProps.show,
      show = _nextProps$show === void 0 ? false : _nextProps$show;
    const stateShow = prevState.show,
      timeoutExecuted = prevState.timeoutExecuted,
      prevShow = prevState.prevShow;
    const showChangedProp = show !== prevShow;
    if (showChangedProp) {
      if (show && !timeoutExecuted) {
        return {
          display: true,
          show,
          prevShow: show
        };
      }
      if (show !== stateShow && !timeoutExecuted) {
        return {
          show: show || false,
          prevShow: show
        };
      }
      if (timeoutExecuted) {
        return {
          timeoutExecuted: false,
          prevShow: show
        };
      }
    }
    return null;
  }

  /**
   * @param {SnackbarProps} _
   * @param {SnackbarState} prevState
   */
  componentDidUpdate(_, prevState) {
    const stateShow = this.state.show;
    const prevStateShow = prevState.show;
    const _this$props2 = this.props,
      onOpen = _this$props2.onOpen,
      _this$props2$show = _this$props2.show,
      show = _this$props2$show === void 0 ? false : _this$props2$show,
      delay = _this$props2.delay;
    const startedShowing = !prevStateShow && stateShow;
    const hasDelayProps = delay;
    if (stateShow !== prevStateShow) {
      if (show && onOpen) onOpen();
      if (this.isReduceMotion() && !show) this.onClose();
    }
    if (startedShowing && hasDelayProps) {
      this.delayTimeout = setTimeout(() => {
        this.setState({
          show: false,
          timeoutExecuted: true
        }, () => {
          if (this.isReduceMotion()) this.onClose();
        });
      }, delay);
    }
  }

  /**
   * @param {React.SyntheticEvent<HTMLAnchorElement | HTMLButtonElement>} event
   * @param {ActionConfig['onClick']} [onClickProp]
   */
  onActionClick(event, onClickProp) {
    if (typeof onClickProp === 'function') {
      clearTimeout(this.delayTimeout);
      onClickProp(event);
    }
  }
  onClose() {
    const onClose = this.props.onClose;
    clearTimeout(this.delayTimeout);
    if (typeof onClose === 'function') onClose();
  }
  onAnimationEnd() {
    const show = this.state.show;
    if (!show) {
      this.onClose();
      this.setState({
        display: false
      });
    }
  }
  isReduceMotion() {
    var _window;
    return typeof window !== 'undefined' && ((_window = window) === null || _window === void 0 ? void 0 : _window.matchMedia('(prefers-reduced-motion: reduce)').matches) === true;
  }

  /** @param {HTMLDivElement} ref */
  setMainContainerRef(ref) {
    const innerRef = this.props.innerRef;
    if (innerRef) {
      if (typeof innerRef === 'function') {
        innerRef(ref);
      } else {
        innerRef.current = ref;
      }
    }
  }

  /**
   * @typedef {Object} Props
   * @prop {ActionConfig['text']} [text]
   * @prop {ActionConfig['onClick']} [onClick]
   */
  /** @param {Props} props */
  renderAction(_ref) {
    let text = _ref.text,
      onClick = _ref.onClick;
    const common = {
      'data-andes-snackbar-action': true,
      className: `${namespace}__action`,
      onClick: (/** @type {ActionConfig['onClick']} */event => this.onActionClick(event, onClick))
    };
    return text && /*#__PURE__*/React__default["default"].createElement("button", common, text);
  }
  render() {
    const _this$props3 = this.props,
      action = _this$props3.action,
      className = _this$props3.className,
      color = _this$props3.color,
      children = _this$props3.children,
      message = _this$props3.message,
      classes = _this$props3.classNames,
      andesDataAttributes = _this$props3.andesDataAttributes;
    const _this$state = this.state,
      display = _this$state.display,
      show = _this$state.show,
      srLabelVisibility = _this$state.srLabelVisibility;
    const computedAction = action || {};
    const computedColor = computeTypeColor$1({
      color
    });
    const containerClassNames = classNames__default["default"](namespace, className, `${namespace}--${computedAction.position || 'right'}`, {
      [`${namespace}--${computedColor}`]: computedColor,
      [`${namespace}--with-action`]: computedAction || children,
      [`${namespace}--animate-show`]: show,
      [`${namespace}--animate-hide`]: show === false
    }, classes === null || classes === void 0 ? void 0 : classes.root);
    const isReduced = this.isReduceMotion();
    if (!display || show === null || isReduced && show === false) {
      return null;
    }
    return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
      className: containerClassNames,
      onAnimationEnd: this.onAnimationEnd,
      role: "presentation",
      ref: this.setMainContainerRef,
      style: this.style,
      "data-andes-snackbar": true,
      "data-andes-snackbar-color": color
    }, andesDataAttributes), /*#__PURE__*/React__default["default"].createElement("div", null, /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, {
      role: "alert",
      "aria-live": "assertive"
    }, srLabelVisibility && message), /*#__PURE__*/React__default["default"].createElement("p", {
      className: `${namespace}__message`,
      "data-andes-snackbar-message": true
    }, message), children && /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${namespace}__children`,
      "data-andes-snackbar-content": true
    }, children)), this.renderAction(computedAction));
  }
}
Snackbar.displayName = 'Snackbar';
Snackbar.propTypes = {
  action: PropTypes__default["default"].shape({
    text: PropTypes__default["default"].string.isRequired,
    onClick: PropTypes__default["default"].func,
    position: PropTypes__default["default"].oneOf(['right', 'bottom'])
  }),
  className: PropTypes__default["default"].string,
  color: PropTypes__default["default"].oneOf(AVAILABLE_COLOR),
  children: PropTypes__default["default"].node,
  delay: PropTypes__default["default"].number,
  innerRef: PropTypes__default["default"].shape({
    current: PropTypes__default["default"].oneOfType([PropTypes__default["default"].node, PropTypes__default["default"].object])
  }),
  message: PropTypes__default["default"].string.isRequired,
  onClose: PropTypes__default["default"].func,
  onOpen: PropTypes__default["default"].func,
  show: PropTypes__default["default"].bool,
  classNames: PropTypes__default["default"].shape({}),
  andesDataAttributes: PropTypes__default["default"].shape({})
};
var Snackbar$1 = Snackbar;

// @ts-ignore
const SnackbarWebkit = props => {
  const _React$useState = React__default["default"].useState(false),
    _React$useState2 = _slicedToArray__default["default"](_React$useState, 2),
    error = _React$useState2[0],
    setError = _React$useState2[1];
  const _props$show = props.show,
    show = _props$show === void 0 ? false : _props$show,
    message = props.message,
    action = props.action,
    delay = props.delay,
    onChange = props.onChange,
    onOpen = props.onOpen,
    onClose = props.onClose,
    color = props.color;
  function getColor() {
    let realColor = 'basic';
    if (color === 'green' || color === 'positive') {
      realColor = 'positive';
    } else if (color === 'red' || color === 'negative') {
      realColor = 'negative';
    }
    return realColor;
  }
  async function onOpenSnackbar() {
    // Webkit expects duration in seconds
    const durationInSeconds = delay / 1000;
    const options = {
      message,
      duration: durationInSeconds,
      action: action === null || action === void 0 ? void 0 : action.text,
      onDismiss: typeof onClose === 'function',
      color: getColor()
    };

    // @ts-ignore
    function getEventName(name, id) {
      return `nordic:${name}${id ? `:${id}` : ''}`;
    }

    // @ts-ignore
    const _await$MobileWebKit$e = await MobileWebKit.executeNative({
        method: 'wk:snackbar-show',
        args: options
      }),
      id = _await$MobileWebKit$e.id;
    const dismissEvent = getEventName('onSnackbarDismiss', id);

    // @ts-ignore
    MobileWebKit.on(dismissEvent, () => {
      if (typeof onClose === 'function') {
        onClose();
        if (typeof onChange === 'function') onChange(false);
      }
      // @ts-ignore
      MobileWebKit.off(dismissEvent);
    });
    if (typeof onOpen === 'function') {
      onOpen();
      if (typeof onChange === 'function') onChange(true);
    }
  }
  React.useEffect(() => {
    if (show && !error) {
      onOpenSnackbar().catch(() => {
        setError(true);
      });
    }
  }, [show, error]);
  if (show && error) {
    return /*#__PURE__*/React__default["default"].createElement(Snackbar$1, props);
  }
  return null;
};
var SnackbarWebkit$1 = SnackbarWebkit;
const _excluded = ["ref"];

/**
 * @typedef {Object} SnackbarContainerInternalProps
 * @prop {Record<string, unknown>} [andesDataAttributes]
 */

/** @typedef {import('../..').SnackbarProps} SnackbarContainerPublicProps */

/** @type {React.FunctionComponent<SnackbarContainerPublicProps & SnackbarContainerInternalProps>} */
const SnackbarContainer = _ref => {
  var _props$action;
  let ref = _ref.ref,
    props = _objectWithoutProperties__default["default"](_ref, _excluded);
  if (utils.isWebkit2()
  // @ts-expect-error - children is not a prop of SnackbarProps
  && !(props !== null && props !== void 0 && props.children) && !ref && !(props !== null && props !== void 0 && (_props$action = props.action) !== null && _props$action !== void 0 && _props$action.onClick) && props.delay) {
    return /*#__PURE__*/React__default["default"].createElement(SnackbarWebkit$1, props);
  }
  return /*#__PURE__*/React__default["default"].createElement(Snackbar$1, _extends__default["default"]({
    innerRef: ref
  }, props));
};
SnackbarContainer.displayName = 'Snackbar';
var SnackbarContainer$1 = SnackbarContainer;
exports.Snackbar = SnackbarContainer$1;
