// frontend-restclient/config/defaults.js
const crypto = require('node:crypto');
const http = require('node:http');
const https = require('node:https');

const { AxiosError } = require('axios');
const env = require('frontend-env');

const { TransportRetryPolicy } = require('../src/policies');
const { version } = require('../package.json');
const {
  throwIfJsonExpected,
} = require('../src/utils/json-response-validation');

/**
 * Default configuration
 * @see https://github.com/axios/axios/tree/v0.x#request-config
 * @type {import('../types/index').RestClientConfig}
 */
const defaults = {
  baseURL:
    (env.FURY || env.SANDBOX) && !env.IS_CI
      ? 'http://internal.mercadolibre.com'
      : 'https://internal-api.mercadolibre.com',
  method: 'get',
  timeout: 3000,
  responseType: 'json',
  allowAbsoluteUrls: false,
  allowRepeatedParams: false,
  logErrors: true,
  transformResponse: [
    function transformResponse(data, headers) {
      const transitional = this.transitional || defaults.transitional;
      const forcedJSONParsing = transitional?.forcedJSONParsing;
      const JSONRequested = this.responseType === 'json';

      if (
        data &&
        typeof data === 'string' &&
        ((forcedJSONParsing && !this.responseType) || JSONRequested)
      ) {
        const silentJSONParsing = transitional?.silentJSONParsing;
        const strictJSONParsing = !silentJSONParsing && JSONRequested;

        try {
          return JSON.parse(data);
        } catch (e) {
          if (!strictJSONParsing) {
            throwIfJsonExpected(e, headers, this, data);

            return data;
          }

          if (e.name === 'SyntaxError') {
            throw AxiosError.from(
              e,
              AxiosError.ERR_BAD_RESPONSE,
              this,
              null,
              this.response,
            );
          }

          throw e;
        }
      }

      return data;
    },
  ],
  maxRedirects: 0,
  httpAgent: new http.Agent({
    keepAlive: true,
    keepAliveMsecs: 1000,
    maxSockets: 100,
    maxFreeSockets: 50,
  }),
  httpsAgent: new https.Agent({
    keepAlive: true,
    keepAliveMsecs: 2000,
    maxSockets: 100,
    maxFreeSockets: 60,
    // TODO: Remove when all MeLi Infrastructure supports Secure Renegotiation
    secureOptions: crypto.constants.SSL_OP_LEGACY_SERVER_CONNECT,
  }),
  headers: {
    'user-agent': `restclient-node/${version}`,
    'accept-encoding': 'gzip',
  },
  retry: {
    maxRetries: 1,
    delay: 0,
    maxDelay: Infinity,
    factor: 1,
    policy: TransportRetryPolicy,
  },
  cache: {
    maxSize: 500,
    maxAge: Infinity,
    staleWhileRevalidate: 0,
  },
  eTagCache: {
    maxSize: 100,
  },
};

module.exports = defaults;
