// @andes/checkbox/build/esm/module.js
import _extends from '@babel/runtime/helpers/extends';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import React, { useRef, useState, useEffect } from 'react';
import classNames from 'classnames';
import { useId } from '@andes/context';
import { VisuallyHidden } from '@andes/visually-hidden';
import { useThemeContext } from '@andes/themes';
import FeedbackPositive16 from '@andes/icons/FeedbackPositive16';
import { computeDataAndesState } from '@andes/common/build/utils/computeDataAndesState';
import warning from '@andes/common/build/utils/warning';
const IndeterminateIcon = () => /*#__PURE__*/React.createElement("svg", {
  "aria-hidden": "true",
  width: "16",
  height: "16",
  viewBox: "0 0 16 16",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/React.createElement("rect", {
  width: "10",
  height: "1.6",
  x: "3",
  y: "7.2",
  fill: "currentColor"
}));
var IndeterminateIcon$1 = IndeterminateIcon;
const namespace = 'andes-checkbox';
const WARNING_LABEL = `<Checkbox> needs some plain text contents for accessibility reasons.
Please add a \`label\` or \`srLabel\` prop.`;
const LABEL_POSITION_LEFT = 'left';
const Checkbox = /*#__PURE__*/function () {
  const Checkbox = props => {
    const checked = props.checked,
      className = props.className,
      defaultChecked = props.defaultChecked,
      _props$disabled = props.disabled,
      disabled = _props$disabled === void 0 ? false : _props$disabled,
      _props$error = props.error,
      error = _props$error === void 0 ? false : _props$error,
      _props$highlight = props.highlight,
      highlight = _props$highlight === void 0 ? false : _props$highlight,
      id = props.id,
      _props$indeterminate = props.indeterminate,
      indeterminate = _props$indeterminate === void 0 ? false : _props$indeterminate,
      _props$inputProps = props.inputProps,
      inputProps = _props$inputProps === void 0 ? {} : _props$inputProps,
      label = props.label,
      _props$labelPosition = props.labelPosition,
      labelPosition = _props$labelPosition === void 0 ? 'right' : _props$labelPosition,
      onChange = props.onChange,
      _props$srHidden = props.srHidden,
      srHidden = _props$srHidden === void 0 ? false : _props$srHidden,
      srLabel = props.srLabel,
      ref = props.ref;
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const _useRef = useRef(typeof checked !== 'undefined'),
      isControlled = _useRef.current;
    const _useState = useState(defaultChecked || checked),
      _useState2 = _slicedToArray(_useState, 2),
      isChecked = _useState2[0],
      setIsChecked = _useState2[1];
    warning(Boolean(label || srLabel || srHidden), WARNING_LABEL);
    useEffect(() => {
      if (isControlled && checked !== undefined) {
        setIsChecked(checked);
      }
    }, [checked, isControlled]);
    const componentId = useId(id);
    const srLabelId = `${componentId}-srLabel`;
    const inputRef = useRef(null);
    const handleChange = event => {
      if (!isControlled) setIsChecked(event.target.checked);
      if (typeof onChange === 'function') onChange(event);
    };
    const checkboxClassnames = classNames(namespace, className, {
      [`${namespace}--error`]: error,
      [`${namespace}--highlight`]: highlight,
      [`${namespace}--label-left`]: labelPosition === LABEL_POSITION_LEFT,
      [`${namespace}--indeterminate`]: indeterminate && isChecked,
      [`${namespace}--disabled`]: disabled
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('checkbox'));
    const labelClassnames = classNames(`${namespace}__label`, {
      [`${namespace}__label-text`]: label
    });
    const Wrapper = label ? 'label' : 'div';
    useEffect(() => {
      if (inputRef !== null && inputRef !== void 0 && inputRef.current) {
        inputRef.current.indeterminate = indeterminate && isChecked || false;
      }
    }, [inputRef, indeterminate, isChecked]);
    return /*#__PURE__*/React.createElement(Wrapper, _extends({
      className: checkboxClassnames,
      ref: ref,
      "data-andes-checkbox": true,
      "data-andes-checkbox-variant": highlight ? 'highlight' : undefined
    }, computeDataAndesState({
      checked: isChecked,
      disabled,
      indeterminate: indeterminate && isChecked,
      error: error && !isChecked
    }), themeHash), /*#__PURE__*/React.createElement("span", {
      className: `${namespace}__checkbox`,
      "data-andes-checkbox-container": true
    }, /*#__PURE__*/React.createElement("input", _extends({}, inputProps, {
      ref: inputRef,
      checked: checked,
      className: `${namespace}__input`,
      defaultChecked: defaultChecked,
      disabled: disabled,
      id: componentId,
      onChange: handleChange,
      type: "checkbox"
    }, srLabel ? {
      'aria-labelledby': srLabelId
    } : {}, {
      "data-andes-checkbox-input": true
    })), /*#__PURE__*/React.createElement("span", {
      className: `${namespace}__icon`
    }, indeterminate && isChecked ? /*#__PURE__*/React.createElement(IndeterminateIcon$1, null) : /*#__PURE__*/React.createElement(FeedbackPositive16, {
      color: "currentColor"
    }))), label && /*#__PURE__*/React.createElement("span", {
      "aria-hidden": !!srLabel || undefined,
      className: labelClassnames,
      "data-andes-checkbox-label": true
    }, label), srLabel && /*#__PURE__*/React.createElement(VisuallyHidden, {
      id: srLabelId
    }, srLabel));
  };
  Checkbox.displayName = 'Checkbox';
  return Checkbox;
}();
var Checkbox$1 = Checkbox;
export { Checkbox$1 as Checkbox };
