// @nordic/utils/src/cookie.ts
import { sanitizeValue } from './xss-parser';

export type CookieValueType = 'binary' | 'number' | 'string';

export type CookieDictionaryEntry = {
  /**
   * The key to get the cookie value from the resulting object.
   */
  key: string;

  /**
   * The type of the cookie value.
   *
   * @default 'string'
   */
  type?: CookieValueType;

  /**
   * The function to pre-parse the cookie value before parsing its type (for example, to prevent XSS attacks).
   *
   * It can be:
   * - A Boolean value to indicate whether or not to use the default parser of the XSS attack function.
   * - Or a Function to use as the parser.
   *
   * @default true
   */
  preparser?: boolean | ((value: string) => unknown);
};

export type CookieDictionary = Record<string, CookieDictionaryEntry>;

export type CookieValueTypeMap = {
  binary: boolean;
  number: number;
  string: string;
};

export type CookieParseResult<T extends CookieDictionary> = {
  -readonly [K in keyof T]?: T[K] extends {
    type: infer U extends keyof CookieValueTypeMap;
  }
    ? CookieValueTypeMap[U]
    : string;
};

function parseValue(
  value: string,
  type: CookieValueType = 'string',
): boolean | number | string | undefined {
  switch (type) {
    case 'binary':
      return value === '1';
    case 'number':
      return +value;
    case 'string':
      return value;

    default:
      return undefined;
  }
}

/**
 * Parses a cookie value into a typed object based on a configuration dictionary.
 *
 * @example
 *
 * const result = parseCookieValue('v=1.0.0,m=1,mc=Samsung",user=Foo_Ipsum', {
 *   version: { key: 'v', type: 'number', parser: false },
 *   mobile: { key: 'm', type: 'binary' },
 *   user: { key: 'user', parser: (value) => value.replaceAll('_', ' ') },
 *   marca: { key: 'mc' },
 * } as const);
 *
 * // result = {
 * //   version: 1,
 * //   marca: 'Samsung',
 * //   mobile: true,
 * //   user: 'Foo Ipsum'
 * // }
 */
export function parseCookieValue<T extends CookieDictionary>(
  cookieValue: string,
  dictionary: T,
) {
  const invertedDictionary = Object.entries(
    dictionary,
  ).reduce<CookieDictionary>(
    (acc, [key, value]) =>
      Object.assign(acc, {
        [value.key]: { key, type: value.type, preparser: value.preparser },
      }),
    {},
  );

  return cookieValue.split(',').reduce<Record<string, unknown>>((acc, pair) => {
    const [key, value] = pair.split('=');

    if (!key || !value) {
      return acc;
    }

    const entry = invertedDictionary[key];

    if (!entry) {
      return acc;
    }

    const { key: entryKey, type, preparser = true } = entry;

    const parsedValue =
      typeof preparser === 'function'
        ? preparser(value)
        : parseValue(preparser ? sanitizeValue(value) : value, type);

    if (typeof parsedValue !== 'undefined') {
      acc[entryKey] = parsedValue;
    }

    return acc;
  }, {}) as CookieParseResult<T>;
}
