// @nordic/utils/src/error.ts
import * as React from 'react';

import { logger } from './logger';

function safeToString(value: unknown) {
  try {
    // Template literals are faster, but fail for Symbol values; String() keeps that fallback safe.
    return String(value);
  } catch {
    // Custom object coercion can still throw, so this is the final safe fallback.
    return 'Unable to stringify value';
  }
}

function stringifyUnknown(value: unknown) {
  if (typeof value === 'string') {
    return value;
  }

  // Prefer JSON for plain objects. Runtime can still return undefined for values such as undefined, functions or symbols.
  const result = JSON.stringify(value) || safeToString(value);

  return result;
}

function debugStringifyError(error: unknown) {
  try {
    logger.debug(
      'Failed to stringify error.',
      error instanceof Error
        ? { message: error.message, stack: error.stack }
        : { message: safeToString(error) },
    );
  } catch {
    // Avoid throwing from the error handling code.
  }
}

export function stringifyError(
  error: unknown,
  selector?: 'message' | 'stack' | ((error: Error) => string),
) {
  try {
    if (!(error instanceof Error)) {
      return stringifyUnknown(error);
    }

    if (selector) {
      return stringifyUnknown(
        typeof selector === 'function' ? selector(error) : error[selector],
      );
    }

    return error.stack || error.message;
  } catch (err) {
    debugStringifyError(err);

    // Selectors or custom coercion can throw; always return a string for logging/reporting.
    return safeToString(error);
  }
}

export class OwnerStackError extends Error {
  ownerStack: string | null;

  constructor(message: string) {
    super(message);
    this.ownerStack = __DEV__ ? React.captureOwnerStack() : null;
  }
}
