// @nordic/utils/src/validation.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
const EXTERNAL_URL_REGEX = /^https?:|^\/\/|^data:image\//;

const IMG_EXTS_WITH_BLUR_AVAILABLE_REGEX = /\.(jpg|jpeg|png|webp|avif)$/;

export function isObject(value: any): value is object {
  return (
    value !== null &&
    typeof value === 'object' &&
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    value.constructor === Object
  );
}

export function isClient() {
  return (
    typeof window !== 'undefined' &&
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    typeof window.document?.createElement === 'function'
  );
}

export function isExternalUrl(value: string) {
  return EXTERNAL_URL_REGEX.test(value) || value.startsWith('//');
}

export function isStorybook() {
  return isClient() && !!document.getElementById('storybook-root');
}

export function isTestEnv() {
  if (typeof process === 'undefined') {
    return false;
  }

  const { NODE_ENV, NODE_TEST } = process.env;

  return NODE_ENV === 'test' && NODE_TEST !== 'e2e';
}

export function isBlurableImg(src: string) {
  return !isExternalUrl(src) && IMG_EXTS_WITH_BLUR_AVAILABLE_REGEX.test(src);
}
