// @andes/badge/build/esm/module.js
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React from 'react';
import classNames from 'classnames';
import { VisuallyHidden } from '@andes/visually-hidden';
import { useId } from '@andes/context';
import includes from '@andes/common/build/utils/includes';
import warning from '@andes/common/build/utils/warning';
import { useThemeContext } from '@andes/themes';
import _extends from '@babel/runtime/helpers/extends';
import FeedbackCaution24 from '@andes/icons/FeedbackCaution24';
import FeedbackInfo24 from '@andes/icons/FeedbackInfo24';
import FeedbackNegative24 from '@andes/icons/FeedbackNegative24';
import FeedbackPositive24 from '@andes/icons/FeedbackPositive24';
const namespace = 'andes-badge';
const PILL_TYPE_ICON = 'icon';
const ROUNDED_CORNERS_VALUES = ['top-left', 'top-right', 'bottom-left', 'bottom-right'];
const DEPRECATED_BADGE_COLORS = ['accent', 'green', 'orange', 'red', 'gray'];
const WARNING_COLORS = `This color will be deprecated in a future major version,
 please use 'caution' | 'informative' | 'inverted' | 'negative' | 'neutral' | 'positive'`;
const WARNING_CONTENT_TYPE = `'contentType' will be deprecated in a future major version, 
 please use <PillIcon> instead of 'contentType="icon"' or remove it if you're using 'contentType="text"'.`;
const computeColorProps = _ref => {
  let color = _ref.color;
  switch (color) {
    case 'informative':
      return 'accent';
    case 'neutral':
      return 'gray';
    case 'caution':
      return 'orange';
    case 'negative':
      return 'red';
    case 'positive':
      return 'green';
    default:
      return color || 'accent';
  }
};
var computeColorProps$1 = computeColorProps;
const _excluded$2 = ["className", "color", "id", "srLabel", "ref", "backgroundType", "customVariant"];
const Dot = /*#__PURE__*/function () {
  const Dot = props => {
    const className = props.className,
      color = props.color,
      id = props.id,
      srLabel = props.srLabel,
      ref = props.ref,
      backgroundType = props.backgroundType;
    props.customVariant;
    const otherProps = _objectWithoutProperties(props, _excluded$2);
    const componentId = useId(id);
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    if (typeof color === 'string') {
      warning(!includes(DEPRECATED_BADGE_COLORS, color), WARNING_COLORS);
    }
    const computedColor = computeColorProps$1({
      color
    });
    const classes = classNames(namespace, className, `${namespace}--dot`, `${namespace}--${computedColor}`, {
      [`${namespace}--color-on-background`]: backgroundType === 'complexBackground'
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('badge', props));
    const sharedProps = _objectSpread(_objectSpread({
      className: classes,
      id: componentId,
      ref
    }, otherProps), {}, {
      'data-andes-badge': true,
      'data-andes-badge-type': 'dot'
    }, themeHash);
    return /*#__PURE__*/React.createElement("div", sharedProps, srLabel && /*#__PURE__*/React.createElement(VisuallyHidden, null, srLabel));
  };
  Dot.displayName = 'Dot';
  return Dot;
}();
var Dot$1 = Dot;
const _excluded$1 = ["className", "color", "id", "size", "srLabel", "ref", "backgroundType", "customVariant"];
const getSign = color => {
  const commonProps = {
    className: `${namespace}__icon-inner`,
    color: 'currentColor'
  };
  const colors = {
    accent: /*#__PURE__*/React.createElement(FeedbackInfo24, _extends({}, commonProps, {
      key: "badge-info"
    })),
    green: /*#__PURE__*/React.createElement(FeedbackPositive24, _extends({}, commonProps, {
      key: "badge-success"
    })),
    orange: /*#__PURE__*/React.createElement(FeedbackCaution24, _extends({}, commonProps, {
      key: "badge-warning"
    })),
    red: /*#__PURE__*/React.createElement(FeedbackNegative24, _extends({}, commonProps, {
      key: "badge-error"
    }))
  };
  return colors[color];
};
const PillIcon = /*#__PURE__*/function () {
  const PillIcon = props => {
    const className = props.className,
      color = props.color,
      id = props.id,
      _props$size = props.size,
      size = _props$size === void 0 ? 'large' : _props$size,
      srLabel = props.srLabel,
      ref = props.ref,
      backgroundType = props.backgroundType,
      customVariant = props.customVariant,
      otherProps = _objectWithoutProperties(props, _excluded$1);
    const allProps = _objectSpread({
      customVariant,
      size
    }, props);
    const componentId = useId(id);
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    if (typeof color === 'string') {
      warning(!includes(DEPRECATED_BADGE_COLORS, color), WARNING_COLORS);
    }
    const computedColor = computeColorProps$1({
      color
    });
    const andesDataAttributes = _objectSpread({
      'data-andes-badge': true,
      'data-andes-badge-type': 'icon',
      'data-andes-badge-size': size
    }, themeHash);
    const classes = classNames(namespace, `${namespace}--pill`, `${namespace}--${computedColor}`, className, `${namespace}--pill-icon`, {
      [`${namespace}--${size}`]: size,
      [`${namespace}--color-on-background`]: backgroundType === 'complexBackground'
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('badge', allProps));
    return /*#__PURE__*/React.createElement("div", _extends({
      className: classes,
      id: componentId,
      ref: ref
    }, otherProps, andesDataAttributes), srLabel && /*#__PURE__*/React.createElement(VisuallyHidden, null, srLabel), /*#__PURE__*/React.createElement("div", {
      "aria-hidden": "true",
      className: `${namespace}__icon`
    }, getSign(computedColor)));
  };
  PillIcon.displayName = 'PillIcon';
  return PillIcon;
}();
var PillIcon$1 = PillIcon;
const _excluded = ["children", "className", "color", "hierarchy", "contentType", "id", "roundedCorners", "secondaryColor", "size", "srLabel", "ref", "backgroundType", "customVariant"];
const Pill = /*#__PURE__*/function () {
  const Pill = props => {
    const children = props.children,
      className = props.className,
      color = props.color,
      _props$hierarchy = props.hierarchy,
      hierarchy = _props$hierarchy === void 0 ? 'loud' : _props$hierarchy,
      _willdeprecate$contentType = props.contentType,
      id = props.id,
      _props$roundedCorners = props.roundedCorners,
      roundedCorners = _props$roundedCorners === void 0 ? ROUNDED_CORNERS_VALUES : _props$roundedCorners,
      secondaryColor = props.secondaryColor,
      _props$size = props.size,
      size = _props$size === void 0 ? 'large' : _props$size,
      srLabel = props.srLabel,
      ref = props.ref,
      backgroundType = props.backgroundType,
      customVariant = props.customVariant,
      otherProps = _objectWithoutProperties(props, _excluded);
    warning(typeof _willdeprecate$contentType === 'undefined', WARNING_CONTENT_TYPE);
    if (typeof color === 'string') {
      warning(!includes(DEPRECATED_BADGE_COLORS, color), WARNING_COLORS);
    }
    const isIcon = _willdeprecate$contentType === PILL_TYPE_ICON;
    if (isIcon) {
      const iconProps = _objectSpread({
        className,
        color,
        id,
        size,
        srLabel,
        ref,
        backgroundType,
        customVariant
      }, otherProps);
      return /*#__PURE__*/React.createElement(PillIcon$1, iconProps);
    }
    const allProps = _objectSpread({
      customVariant,
      contentType: _willdeprecate$contentType,
      hierarchy,
      roundedCorners,
      size
    }, props);
    const componentId = useId(id);
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const computedColor = computeColorProps$1({
      color
    });
    const classes = classNames(namespace, `${namespace}--pill`, `${namespace}--${computedColor}`, className, {
      [`${namespace}--${size}`]: size,
      [`${namespace}--${computedColor}--quiet`]: !secondaryColor && hierarchy === 'quiet',
      [`${namespace}--rounded-top-left`]: includes(roundedCorners, 'top-left'),
      [`${namespace}--rounded-top-right`]: includes(roundedCorners, 'top-right'),
      [`${namespace}--rounded-bottom-left`]: includes(roundedCorners, 'bottom-left'),
      [`${namespace}--rounded-bottom-right`]: includes(roundedCorners, 'bottom-right'),
      [`${namespace}--color-on-background`]: hierarchy === 'loud' && backgroundType === 'complexBackground'
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('badge', allProps));
    const andesDataAttributes = _objectSpread({
      'data-andes-badge': true,
      'data-andes-badge-type': 'pill',
      'data-andes-badge-hierarchy': hierarchy,
      'data-andes-badge-size': size
    }, themeHash);
    return /*#__PURE__*/React.createElement("div", _extends({
      className: classes,
      id: componentId,
      ref: ref
    }, otherProps, {
      style: {
        backgroundColor: secondaryColor === null || secondaryColor === void 0 ? void 0 : secondaryColor.background
      }
    }, andesDataAttributes), srLabel && /*#__PURE__*/React.createElement(VisuallyHidden, null, srLabel), children && /*#__PURE__*/React.createElement("p", {
      "aria-hidden": !!srLabel || undefined,
      className: `${namespace}__content`,
      style: {
        color: secondaryColor === null || secondaryColor === void 0 ? void 0 : secondaryColor.text
      }
    }, children));
  };
  Pill.displayName = 'Pill';
  return Pill;
}();
var Pill$1 = Pill;
export { Dot$1 as Dot, Pill$1 as Pill, PillIcon$1 as PillIcon };
