// @andes/progress-indicator-circular/build/esm/module.js
import _extends from '@babel/runtime/helpers/extends';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React from 'react';
import classNames from 'classnames';
import { VisuallyHidden } from '@andes/visually-hidden';
import useSrLabel from '@andes/common/build/hooks/useSrLabel';
import { useId } from '@andes/context';
import { useThemeContext } from '@andes/themes';
import warning from '@andes/common/build/utils/warning';
const namespace = 'andes-progress-indicator-circular';
const circleNamespace = `${namespace}__circle`;
const DEPRECATED_PROGRESS_CIRCULAR_COLORS = ['blue', 'white'];
const Indeterminate = _ref => {
  let labelId = _ref.labelId,
    srLabel = _ref.srLabel;
  return /*#__PURE__*/React.createElement("div", {
    "aria-label": srLabel,
    "aria-labelledby": srLabel ? undefined : labelId,
    role: "progressbar"
  }, /*#__PURE__*/React.createElement("svg", {
    className: `${namespace}__svg`,
    "aria-hidden": "true",
    viewBox: "0 0 50 50",
    "data-andes-progress-indicator-circular-icon": true
  }, /*#__PURE__*/React.createElement("circle", {
    className: circleNamespace,
    cx: "50%",
    cy: "50%",
    r: "20",
    fill: "none"
  })));
};
var Indeterminate$1 = Indeterminate;
const Determinate = _ref => {
  let _ref$continuous = _ref.continuous,
    continuous = _ref$continuous === void 0 ? false : _ref$continuous,
    id = _ref.id,
    labelId = _ref.labelId,
    srLabel = _ref.srLabel,
    _ref$value = _ref.value,
    value = _ref$value === void 0 ? 0 : _ref$value;
  const circunference = 124;
  const circlePercentage = circunference * value / 100;
  const circleClassnames = classNames(circleNamespace, {
    [`${circleNamespace}--continuous`]: continuous
  });
  return /*#__PURE__*/React.createElement("div", {
    role: "progressbar",
    "aria-valuenow": value,
    "aria-valuemin": 0,
    "aria-valuemax": 100,
    "aria-label": srLabel,
    "aria-labelledby": srLabel ? undefined : labelId
  }, /*#__PURE__*/React.createElement("svg", {
    className: `${namespace}__svg`,
    "aria-hidden": "true",
    viewBox: "0 0 50 50",
    "data-andes-progress-indicator-circular-icon": true
  }, /*#__PURE__*/React.createElement("circle", {
    className: `${circleNamespace}-bg`,
    cx: "25",
    cy: "25",
    r: "20",
    fill: "none"
  }), /*#__PURE__*/React.createElement("circle", {
    key: id,
    className: circleClassnames,
    strokeDasharray: circunference,
    strokeDashoffset: `${circunference - Math.min(circlePercentage, circunference)}`,
    cx: "25",
    cy: "25",
    r: "20",
    fill: "none"
  })));
};
var Determinate$1 = Determinate;
const WARNING_COLORS = `This color will be deprecated in a future major version. 
Please use 'basic' | 'inverse' on <ProgressIndicatorCircular> instead`;
const computeColorProps = _ref => {
  let color = _ref.color;
  if (typeof color === 'string') {
    warning(!DEPRECATED_PROGRESS_CIRCULAR_COLORS.includes(color), WARNING_COLORS);
  }
  switch (color) {
    case 'blue':
      return 'basic';
    case 'white':
      return 'inverse';
    default:
      return color || 'basic';
  }
};
var computeColorProps$1 = computeColorProps;
const _excluded = ["ref"];
const ProgressIndicatorCircular = /*#__PURE__*/function () {
  const ProgressIndicatorCircular = _ref => {
    let ref = _ref.ref,
      props = _objectWithoutProperties(_ref, _excluded);
    const id = props.id,
      _props$size = props.size,
      size = _props$size === void 0 ? 'medium' : _props$size,
      _props$color = props.color,
      color = _props$color === void 0 ? 'basic' : _props$color,
      _props$type = props.type,
      type = _props$type === void 0 ? 'indeterminate' : _props$type,
      className = props.className,
      label = props.label,
      value = props.value,
      children = props.children,
      _props$continuous = props.continuous,
      continuous = _props$continuous === void 0 ? false : _props$continuous,
      srLabel = props.srLabel,
      srAnnouncement = props.srAnnouncement,
      ariaLabelledBy = props.ariaLabelledBy,
      modifier = props.modifier;
    const componentId = useId(id);
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const labelId = ariaLabelledBy || label && `${id}__label`;
    const srLabelVisibility = useSrLabel();
    const getSubcomponent = () => {
      const commonProps = {
        srLabel,
        labelId
      };
      if (type === 'determinate') {
        const determinateProps = _objectSpread({
          value,
          continuous,
          id: componentId
        }, commonProps);
        return /*#__PURE__*/React.createElement(Determinate$1, determinateProps);
      }
      return /*#__PURE__*/React.createElement(Indeterminate$1, commonProps);
    };
    const computedColor = computeColorProps$1({
      color
    });
    const progressIndicatorClassnames = classNames(namespace, className, {
      [`${namespace}--${type}`]: type,
      [`${namespace}--${size}`]: size,
      [`${namespace}--${computedColor}`]: computedColor,
      [`${namespace}--${modifier}`]: modifier
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('progressIndicator'));
    return /*#__PURE__*/React.createElement("div", _extends({
      ref: ref,
      className: progressIndicatorClassnames,
      "data-andes-progress-indicator-circular": true,
      "data-andes-progress-indicator-circular-size": size,
      "data-andes-progress-indicator-circular-type": type
    }, themeHash), /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__container`
    }, /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__progress`
    }, srAnnouncement && /*#__PURE__*/React.createElement(VisuallyHidden, {
      role: "status",
      "aria-live": "polite"
    }, srLabelVisibility && srAnnouncement), getSubcomponent(), children && /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__content`
    }, children)), label && /*#__PURE__*/React.createElement("span", {
      className: `${namespace}__label`,
      id: labelId
    }, label)));
  };
  ProgressIndicatorCircular.displayName = 'ProgressIndicatorCircular';
  return ProgressIndicatorCircular;
}();
var ProgressIndicatorCircular$1 = ProgressIndicatorCircular;
export { ProgressIndicatorCircular$1 as ProgressIndicatorCircular };
