// @global-theme/core/dist/cjs/index.development.js
'use strict';

var serverContext = require('@meli/server-context');
var frontendConfig = require('frontend-config');
var frontendRestclient = require('frontend-restclient');
var internal = require('frontend-statsd/internal');
var frontendLogger = require('frontend-logger');
var deepmerge = require('deepmerge');

/**
 * Metrics utility for tracking global theme performance and behavior.
 * Provides consistent metric naming and tags across the package.
 */
const METRICS_PREFIX = 'global_theme_web';
/**
 * Essential metrics only - minimal set for monitoring performance and errors
 */
const METRICS = {
    THEME_LOAD_DURATION: `${METRICS_PREFIX}.load.duration`,
    THEME_LOAD_ERROR: `${METRICS_PREFIX}.load.error`,
    THEME_FALLBACK_USED: `${METRICS_PREFIX}.fallback.used`,
    THEME_CACHE_MAXAGE: `${METRICS_PREFIX}.cache.maxage`,
    THEME_CACHE_NO_MAXAGE: `${METRICS_PREFIX}.cache.no_maxage`,
    THEME_RESOLVED: `${METRICS_PREFIX}.theme_resolved`,
    THEME_COOKIE_MULTIPLE: `${METRICS_PREFIX}.cookie.multiple_detected`,
};
/**
 * Send increment counter metric
 * @param metric - Metric name
 * @param tags - Optional tags object
 */
const increment = (metric, tags = {}) => {
    internal.statsd.increment(metric, tags);
};
/**
 * Send timing metric in milliseconds
 * @param metric - Metric name
 * @param duration - Duration in milliseconds
 * @param tags - Optional tags object
 */
const timing = (metric, duration, tags = {}) => {
    internal.statsd.timing(metric, duration, tags);
};
/**
 * Send histogram metric
 * @param metric - Metric name
 * @param value - Value to record
 * @param tags - Optional tags object
 */
const histogram = (metric, value, tags = {}) => {
    internal.statsd.histogram(metric, value, tags);
};
/**
 * Creates a timer function that can be stopped to measure duration
 * @param metric - Metric name
 * @param tags - Optional tags object
 * @returns Stop function that records the timing
 */
const startTimer = (metric, tags = {}) => {
    const startTime = Date.now();
    return () => {
        const duration = Date.now() - startTime;
        timing(metric, duration, tags);
        return duration;
    };
};

const ANDES_PREFIX = '--andes';
const BASE_FONT_SIZE = 16;
const GLOBAL_THEME_ID = '__GLOBAL_THEME__';
const COMPONENT_TOKENS_KEY = 'components';
/**
 * Package name constant for global-theme-web
 */
const PACKAGE_NAME = 'global-theme-web';
const FONT_CONFIG = {
    inter: {
        host: 'frontend-assets/mp3-web-lib/fonts',
        extensions: ['woff2'],
        familyName: 'Inter',
        fileName: 'Inter',
        version: '1',
    },
    'proxima nova': {
        host: 'ui/webfonts/v3.0.0/proxima-nova',
        extensions: ['woff2', 'woff'],
        familyName: 'Proxima Nova',
        fileName: 'proximanova',
        weightConverter: (weight) => weight.toLowerCase(),
        defaultWeights: [300, 400, 600],
    },
};

/**
 * Base logger instance for global-theme-web package.
 */
const baseLogger = frontendLogger.LoggerFactory(PACKAGE_NAME);
/**
 * Log level hierarchy for filtering
 */
const LOG_LEVELS = {
    debug: 0,
    info: 1,
    warn: 2,
    error: 3,
    none: 4,
};
/**
 * Singleton filtered logger class
 */
class FilteredLogger {
    currentLogLevel = 'none';
    minLevel = LOG_LEVELS.none;
    /**
     * Configure the log level filter
     * @param logLevel - Minimum log level to display ('none', 'debug', 'info', 'warn', 'error')
     */
    configure(logLevel) {
        this.currentLogLevel = logLevel || 'none';
        this.minLevel = LOG_LEVELS[this.currentLogLevel] ?? LOG_LEVELS.none;
    }
    debug(message, meta) {
        if (LOG_LEVELS.debug >= this.minLevel) {
            baseLogger.debug(message, meta);
        }
    }
    info(message, meta) {
        if (LOG_LEVELS.info >= this.minLevel) {
            baseLogger.info(message, meta);
        }
    }
    warn(message, meta) {
        if (LOG_LEVELS.warn >= this.minLevel) {
            baseLogger.warn(message, meta);
        }
    }
    error(message, meta) {
        if (LOG_LEVELS.error >= this.minLevel) {
            baseLogger.error(message, meta);
        }
    }
}
/**
 * Singleton instance of the filtered logger
 */
const logger = new FilteredLogger();

/**
 * Headers to ignore when logging requests
 */
const IGNORED_REQUEST_HEADERS = [
    'common',
    'delete',
    'get',
    'head',
    'post',
    'put',
    'patch',
    'accept-encoding',
];
/**
 * Headers to ignore when logging responses
 */
const IGNORED_RESPONSE_HEADERS = [
    'connection',
    'keep-alive',
    'transfer-encoding',
    'server',
];
/**
 * Generate curl command from request config
 * @param config - Request configuration
 * @returns Formatted curl command
 */
const generateCurl = (config) => {
    const { method = 'GET', url = '', baseURL = '', params = {}, headers = {}, data, } = config;
    // Build query string
    const hasParams = Object.keys(params).length > 0;
    let queryString = '';
    if (hasParams) {
        const queryParams = Object.keys(params)
            .sort()
            .map((key) => `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`)
            .join('&');
        queryString = `?${queryParams}`;
    }
    // Build full URL
    const fullUrl = url.startsWith('http')
        ? `${url}${queryString}`
        : `${baseURL}${url}${queryString}`;
    // Format headers
    const headersObj = headers;
    const headerLines = Object.keys(headersObj)
        .filter((key) => headersObj[key] && !IGNORED_REQUEST_HEADERS.includes(key))
        .sort()
        .map((key) => {
        const value = headersObj[key];
        const stringValue = typeof value === 'object' ? JSON.stringify(value) : String(value);
        return `  --header '${key}: ${stringValue}'`;
    });
    // Format data
    const dataLine = data ? `  --data-raw '${JSON.stringify(data)}'` : null;
    // Build curl command
    return [
        `[REQ] curl --location --request ${method.toUpperCase()} '${fullUrl}'`,
        ...headerLines,
        dataLine,
    ]
        .filter(Boolean)
        .join(' \\\n');
};
/**
 * Get first level keys from response data
 * @param data - Response data
 * @returns Object with first level keys only
 */
const getFirstLevelKeys = (data) => {
    if (!data || typeof data !== 'object') {
        return data;
    }
    if (Array.isArray(data)) {
        return `[Array with ${data.length} items]`;
    }
    const result = {};
    Object.keys(data).forEach((key) => {
        const value = data[key];
        if (value === null || value === undefined) {
            result[key] = value;
        }
        else if (typeof value === 'object') {
            result[key] = Array.isArray(value)
                ? `[Array with ${value.length} items]`
                : '[Object]';
        }
        else {
            result[key] = value;
        }
    });
    return result;
};
/**
 * Format response for logging
 * @param response - Axios response object
 * @returns Formatted response log
 */
const formatResponse = (response) => {
    const { status, statusText, headers, data } = response;
    // Format response headers
    const headersObj = headers;
    const responseHeaders = Object.keys(headersObj)
        .filter((key) => headersObj[key] && !IGNORED_RESPONSE_HEADERS.includes(key))
        .sort()
        .reduce((acc, key) => {
        acc[key] = headersObj[key];
        return acc;
    }, {});
    // Get first level keys from data
    const responseData = getFirstLevelKeys(data);
    return [
        `[RES] Status: ${status} ${statusText}`,
        `Headers: ${JSON.stringify(responseHeaders, null, 2)}`,
        `Data: ${JSON.stringify(responseData, null, 2)}`,
    ].join('\n');
};
/**
 * Request interceptor function for logging HTTP requests
 * @param config - Request configuration from axios/RestClient
 * @returns The same config object (for chaining)
 */
const requestLoggerInterceptor = (config) => {
    try {
        const curl = generateCurl(config);
        logger.debug(`\n\n${curl}\n\n`, {
            component: PACKAGE_NAME,
            type: 'request-log',
        });
    }
    catch (error) {
        logger.error('Failed to log request', { error: error.message });
    }
    return config;
};
/**
 * Response interceptor function for logging HTTP responses
 * @param response - Axios response object
 * @returns The same response object (for chaining)
 */
const responseLoggerInterceptor = (response) => {
    try {
        const formattedResponse = formatResponse(response);
        logger.debug(`\n\n${formattedResponse}\n\n`, {
            component: PACKAGE_NAME,
            type: 'response-log',
        });
    }
    catch (error) {
        logger.error('Failed to log response', { error: error.message });
    }
    return response;
};

/**
 * Enhanced RestClient with additional features:
 * - Request and Response logging (uses singleton logger configured externally)
 * - Future: metrics, retries, caching, authentication, etc.
 *
 * @param config - RestClient configuration
 * @returns RestClient instance with enhanced features
 */
const EnhancedRestClient = (config) => {
    const restClient = frontendRestclient.RestClient(config);
    // Feature 1: Request and Response logging (only in development)
    // Based on the reference implementation, frontend-restclient exposes axios through .client
    if (process.env.NODE_ENV === 'development') {
        if (restClient.client?.interceptors?.request) {
            restClient.client.interceptors.request.use(requestLoggerInterceptor);
        }
        if (restClient.client?.interceptors?.response) {
            restClient.client.interceptors.response.use(responseLoggerInterceptor);
        }
    }
    // Future features can be added here:
    // - Automatic metrics collection
    // - Smart retry logic
    // - Response caching
    // - Authentication handling
    // - Rate limiting
    // - Request/response transformation
    return restClient;
};

const IS_DEV = process.env.NODE_ENV === 'development';
const URL_SUFFIX = '/themes/resource';
const BASE_URL_PROD = `http://themes.melisystems.com${URL_SUFFIX}`;
const BASE_URL_DEV = `https://themes.melioffice.com${URL_SUFFIX}`;
const TIMEOUT_PROD = 50;
const TIMEOUT_DEV = 1000;
const restClient = EnhancedRestClient({
    baseURL: IS_DEV ? BASE_URL_DEV : BASE_URL_PROD,
    timeout: IS_DEV ? TIMEOUT_DEV : TIMEOUT_PROD,
    retry: false,
    headers: {
        'user-agent': PACKAGE_NAME, // Needed because 'restclient-node/5.1.1' is filtered by CDN
    },
});
/**
 * Normalize the raw API response into an internal Theme object.
 * Extracts `max-age` from the `Cache-Control` header when present.
 *
 * @param response - API response object
 * @returns Normalized theme object
 */
const parseResponse = (response) => {
    const { data: { name, version, value }, headers: { 'cache-control': cacheControl }, } = response;
    const maxAge = cacheControl?.match(/max-age=(\d+)/)?.[1];
    // Track cache maxAge observability for successful API responses
    if (maxAge) {
        histogram(METRICS.THEME_CACHE_MAXAGE, parseInt(maxAge, 10));
    }
    else {
        increment(METRICS.THEME_CACHE_NO_MAXAGE);
    }
    return { id: `${name}:${version}`, name, version, maxAge, value };
};
/**
 * Base method for fetching themes with shared error handling and response parsing.
 *
 * @param params - Query parameters for the theme request.
 * @param context - Optional request context for tracing/instrumentation.
 * @returns Promise resolving to normalized theme object
 * @throws Propagates errors from the underlying REST client.
 */
const fetchTheme = async (params, context) => {
    try {
        const cache = frontendConfig.config.globalTheme
            ?.cache ?? true;
        const response = await restClient.get('', {
            params,
            context,
            cache,
        });
        return parseResponse(response);
    }
    catch (error) {
        increment(METRICS.THEME_LOAD_ERROR, {
            theme: params.id ?? params.name ?? 'unknown',
        });
        throw error;
    }
};
/**
 * Fetch a theme by its full identifier (e.g., "brand:1.2.3").
 *
 * @param id - Theme identifier in the form "<name>:<version>".
 * @param context - Optional request context for tracing/instrumentation.
 * @returns Promise resolving to normalized theme object
 * @throws Propagates errors from the underlying REST client.
 */
const getThemeById = async (id, context) => fetchTheme({ id }, context);
/**
 * Fetch the latest available version of a theme by name.
 *
 * @param name - Theme name.
 * @param context - Optional request context for tracing/instrumentation.
 * @returns Promise resolving to normalized theme object
 * @throws Propagates errors from the underlying REST client.
 */
const getLatestThemeByName = async (name, context) => fetchTheme({ name, version: 'latest' }, context);

/**
 * Maps numeric font weight values to OpenType standard names.
 *
 * @param weight - Numeric font weight value
 * @returns OpenType standard weight name
 */
const getOpenTypeWeightName = (weight) => {
    const weightMap = {
        100: 'Thin',
        200: 'ExtraLight',
        300: 'Light',
        400: 'Regular',
        500: 'Medium',
        600: 'SemiBold',
        700: 'Bold',
        800: 'ExtraBold',
        900: 'Black',
        950: 'ExtraBlack',
    };
    return weightMap[weight] || 'Regular';
};
/**
 * Generates the font file URL based on font family, weight name, and optional CDN URL.
 *
 * @param fontFamily - Font family name from theme
 * @param weightName - OpenType weight name
 * @param format - Font file format
 * @param cdnUrl - Optional CDN base URL
 * @returns Font file URL
 */
const generateFontUrl = (fontFamily, weightName, fontHost, format, version, cdnUrl) => {
    const baseUrl = cdnUrl || 'https://http2.mlstatic.com';
    return `${baseUrl}/${fontHost}/${fontFamily}-${weightName}.${format}${version ? `?v=${version}` : ''}`;
};
/**
 * Generates @font-face CSS rule for a font weight.
 *
 * @param fontFamily - Font family name
 * @param weight - Numeric font weight value
 * @param urls - Array of font file URLs with their formats
 * @returns @font-face CSS rule
 */
const generateFontFaceStyle = (fontFamily, weight, urls) => {
    const srcUrls = urls.map(({ url, format }) => `url('${url}') format('${format}')`);
    // keep this line minified
    return `@font-face{font-family:'${fontFamily}';font-weight:${weight};font-display:swap;font-style:normal;src:${srcUrls.join(',')};}`;
};

/**
 * Extracts font information from theme object and formats it into a structured array.
 * Collects font families and weights from text.body, text.heading, and text.display.
 * Font weight names follow OpenType standard naming convention.
 *
 * @param theme - Theme object containing design tokens
 * @returns Array of font objects with weights
 */
const extractFontsFromTheme = (theme) => {
    const fontsMap = new Map();
    try {
        // theme.default.text is the default font config for the theme
        const textFontsConfig = theme?.default?.text;
        if (!textFontsConfig) {
            return [];
        }
        const sections = Object.keys(textFontsConfig);
        sections.forEach((section) => {
            const family = textFontsConfig[section].family?.default;
            const weights = textFontsConfig[section].weight;
            if (family && weights && typeof weights === 'object') {
                if (!fontsMap.has(family)) {
                    fontsMap.set(family, new Set());
                }
                const fontWeights = fontsMap.get(family);
                Object.values(weights).forEach((weightValue) => {
                    const parsedWeight = Number(weightValue);
                    if (!Number.isNaN(parsedWeight)) {
                        fontWeights.add(parsedWeight);
                    }
                });
            }
        });
        fontsMap.forEach((weights, family) => {
            const fontConfig = FONT_CONFIG[family?.toLowerCase()];
            const defaultWeights = fontConfig?.defaultWeights;
            if (defaultWeights && Array.isArray(defaultWeights)) {
                defaultWeights.forEach((weight) => {
                    weights.add(weight);
                });
            }
        });
        const fonts = Array.from(fontsMap.entries(), ([name, weightsSet]) => ({
            name,
            weights: Array.from(weightsSet)
                .sort((a, b) => a - b)
                .map((weight) => {
                const weightName = getOpenTypeWeightName(weight);
                const fontConfig = FONT_CONFIG[name?.toLowerCase()];
                const extensions = fontConfig?.extensions || [];
                const fontFileName = fontConfig?.fileName || name;
                const fontFamilyName = fontConfig?.familyName || name;
                const fontHost = fontConfig?.host || '';
                const version = fontConfig?.version || '';
                const weightConverter = fontConfig?.weightConverter;
                return {
                    name: weightName,
                    weight,
                    extensions,
                    getUrl: (cdnUrl, format, fontVersion) => generateFontUrl(fontFileName, weightConverter ? weightConverter(weightName) : weightName, fontHost, format || extensions[0], fontVersion || version, cdnUrl),
                    getFontFaceStyle: (cdnUrl, fontVersion) => {
                        const urls = extensions.map((format) => ({
                            url: generateFontUrl(fontFileName, weightConverter ? weightConverter(weightName) : weightName, fontHost, format, fontVersion || version, cdnUrl),
                            format,
                        }));
                        return generateFontFaceStyle(fontFamilyName, weight, urls);
                    },
                };
            }),
        }));
        return fonts;
    }
    catch (error) {
        logger.error('Failed to extract fonts from theme', {
            error: error.message,
        });
        return [];
    }
};

/**
 * Converts pixel values to rem units based on BASE_FONT_SIZE
 */
const pxToRem = (pxValue) => `${pxValue / BASE_FONT_SIZE}rem`;
const addUnit = (value, propertyPath) => {
    // Determine the unit based on the property path
    const path = propertyPath.toLowerCase();
    // Cases that return value without modification
    if (typeof value === 'string' || // Already a string (like rgba colors)
        path.includes('weight') || // Font weights without unit
        path.includes('family') // Font families without modification
    ) {
        return value;
    }
    // Cases that return pxToRem(value)
    if (path.includes('letter-spacing') || // Letter spacing in rem (negative values too)
        (path.includes('size') && path.includes('text')) || // Font sizes in rem
        path.includes('line-height') // Line heights in rem
    ) {
        return pxToRem(value);
    }
    // This includes paragraph-spacing, blur, and all other spacing/sizing values
    return `${value}px`;
};
const toCssEntries = (themeObject, vars = {}, prefix = ANDES_PREFIX, propertyPath = '') => {
    Object.keys(themeObject).forEach((key) => {
        const value = themeObject[key];
        const currentPath = propertyPath ? `${propertyPath}.${key}` : key;
        // Skip 'components' in the first iteration of the loop
        const shouldSkipInCssProperty = key === COMPONENT_TOKENS_KEY && !propertyPath;
        const cssProperty = shouldSkipInCssProperty ? prefix : `${prefix}-${key}`;
        if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
            toCssEntries(value, vars, cssProperty, currentPath);
        }
        else {
            vars[cssProperty] = addUnit(value, currentPath);
        }
    });
    return vars;
};
const varsToString = (vars) => Object.entries(vars)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([key, value]) => `${key}:${value};`)
    .join('');
const themeToVarsString = (themeObject) => {
    const vars = toCssEntries(themeObject);
    return varsToString(vars);
};
const toCss = (theme, options = {}) => {
    const { mode = 'light' } = options;
    if (mode === 'light' || !theme.dark) {
        return `:root{${themeToVarsString(theme.default)}}`;
    }
    const mergedTheme = deepmerge(theme.default, theme.dark);
    const darkVars = themeToVarsString(mergedTheme);
    return `:root{${darkVars}}`;
};

/** Cookie key for storing theme information in format: [name]-[version]-[mode] */
const THEME_COOKIE_KEY = 'x-theme';
const PLATFORM_FALLBACK_THEME_NAME = {
    MP: 'andes_x',
    ML: 'andes_legacy',
};

const getSameSiteValue = (req) => {
    const hasSupport = req.browser?.support?.sameSite || false;
    return hasSupport ? 'none' : false;
};
/**
 * Generates secure cookie configuration based on the request platform.
 *
 * @param req - Express request object with platform information
 * @returns Cookie configuration with domain, security, and HTTP-only settings
 */
const getCookieConfig = (req) => {
    const config = {
        httpOnly: true,
        secure: true,
        sameSite: getSameSiteValue(req),
    };
    return config;
};
/**
 * Parses cookie value in format [name]-[version]-[mode]
 * Example: andes_legacy-1.0.0-light
 * Name uses underscores, version uses dots (X.X.X), separated by dashes
 *
 * @param cookieValue - Cookie value to parse
 * @returns Parsed cookie data or undefined if invalid
 */
const parseCookieValue = (cookieValue) => {
    if (!cookieValue || typeof cookieValue !== 'string') {
        return undefined;
    }
    const parts = cookieValue.split('-');
    if (parts.length !== 3) {
        return undefined;
    }
    const [name, version, mode] = parts;
    if (!name || !version || !mode) {
        return undefined;
    }
    return { name, version, mode };
};
/**
 * Gets the value of the theme cookie from the request.
 *
 * @param req - Express request object
 * @returns Parsed cookie data or undefined if invalid
 */
const getParsedCookieValue = (req) => {
    const cookieValue = req.cookies[THEME_COOKIE_KEY];
    return parseCookieValue(cookieValue);
};
/**
 * Removes the theme cookie from the response.
 *
 * @param req - Express request object with platform information
 * @param res - Express response object for clearing the cookie
 */
const removeCookie = (req, res) => {
    if (res.headersSent) {
        logger.warn('Cannot clear theme cookie - headers already sent.');
        return;
    }
    const cookieConfig = getCookieConfig(req);
    res.clearCookie(THEME_COOKIE_KEY, cookieConfig);
};

var id$1 = "andes_x:1.4.0";
var name$1 = "andes_x";
var version$1 = "1.4.0";
var value$1 = {
	"default": {
		border: {
			radius: {
				full: 9999,
				none: 0,
				tiny: 4,
				xsmall: 6,
				small: 8,
				medium: 12,
				large: 16,
				xlarge: 20,
				huge: 24,
				containers: {
					"default": 20,
					compact: 12
				},
				"actions-inputs": {
					"default": 12,
					compact: 8
				},
				"square-to-circle": {
					large: 12,
					small: 8
				}
			},
			width: {
				none: 0,
				small: 1,
				medium: 2,
				large: 3,
				xlarge: 4
			}
		},
		ai: {
			gradient: {
				"start-early": "rgba(18, 177, 238, 1.000)",
				early: "rgba(0, 214, 241, 1.000)",
				"early-middle": "rgba(68, 225, 242, 1.000)",
				middle: "rgba(180, 244, 244, 1.000)",
				"middle-late": "rgba(226, 248, 222, 1.000)",
				late: "rgba(236, 247, 198, 1.000)",
				start: "rgba(67, 76, 228, 1.000)",
				"late-end": "rgba(255, 243, 148, 1.000)",
				end: "rgba(255, 230, 0, 1.000)"
			},
			text: {
				body: {
					family: {
						"default": "Inter"
					},
					size: {
						medium: 14,
						large: 16
					},
					"line-height": {
						medium: 20,
						large: 22
					},
					weight: {
						"default": 400,
						emphasis: 600
					},
					"letter-spacing": {
						"default": 0
					},
					"paragraph-spacing": {
						xsmall: 10,
						small: 12,
						medium: 14,
						large: 16
					}
				},
				heading: {
					family: {
						"default": "Inter"
					},
					size: {
						large: 18,
						xlarge: 20,
						huge: 24
					},
					"line-height": {
						large: 22,
						xlarge: 24,
						huge: 28
					},
					weight: {
						lean: 400,
						narrow: 600,
						"default": 700
					},
					"letter-spacing": {
						"default": 0
					}
				}
			}
		},
		color: {
			gradient: {
				overlay: {
					start: "rgba(255, 255, 255, 0.000)",
					end: "rgba(255, 255, 255, 0.800)"
				},
				skeleton: {
					fill: {
						"default": "rgba(231, 233, 243, 1.000)",
						start: "rgba(255, 255, 255, 0.000)",
						early: "rgba(255, 255, 255, 0.200)",
						middle: "rgba(255, 255, 255, 0.800)",
						late: "rgba(255, 255, 255, 0.200)",
						end: "rgba(255, 255, 255, 0.000)"
					}
				}
			},
			background: {
				primary: "rgba(255, 255, 255, 1.000)",
				secondary: "rgba(231, 233, 243, 1.000)",
				overlay: "rgba(40, 40, 52, 0.600)",
				"static": "rgba(255, 255, 255, 1.000)"
			},
			surface: {
				elevated: "rgba(255, 255, 255, 1.000)",
				primary: {
					idle: "rgba(255, 255, 255, 1.000)",
					hover: "rgba(244, 245, 249, 1.000)",
					active: "rgba(231, 233, 243, 1.000)",
					disabled: "rgba(255, 255, 255, 1.000)"
				},
				secondary: {
					idle: "rgba(244, 245, 249, 1.000)",
					hover: "rgba(231, 233, 243, 1.000)",
					active: "rgba(208, 212, 230, 1.000)",
					disabled: "rgba(244, 245, 249, 1.000)"
				}
			},
			fill: {
				primary: "rgba(255, 255, 255, 1.000)",
				"primary-on-fill": "rgba(40, 40, 52, 1.000)",
				"primary-on-fill-inverse": "rgba(255, 255, 255, 1.000)",
				secondary: "rgba(208, 212, 230, 1.000)",
				"secondary-on-fill": "rgba(40, 40, 52, 0.300)",
				"secondary-on-fill-inverse": "rgba(255, 255, 255, 0.500)",
				inverse: "rgba(40, 40, 52, 1.000)",
				disabled: "rgba(208, 212, 230, 1.000)",
				"disabled-on-fill": "rgba(40, 40, 52, 0.200)",
				"disabled-on-fill-inverse": "rgba(255, 255, 255, 0.200)",
				"accent-loud": "rgba(67, 76, 228, 1.000)",
				"accent-loud-on-fill": "rgba(40, 40, 52, 1.000)",
				"accent-loud-on-fill-inverse": "rgba(255, 255, 255, 1.000)",
				"accent-quiet": "rgba(233, 241, 255, 1.000)"
			},
			border: {
				primary: "rgba(208, 212, 230, 1.000)",
				"primary-on-fill": "rgba(40, 40, 52, 0.300)",
				"primary-on-fill-inverse": "rgba(255, 255, 255, 0.500)",
				accent: "rgba(67, 76, 228, 1.000)",
				"accent-on-fill": "rgba(40, 40, 52, 1.000)",
				"accent-on-fill-inverse": "rgba(255, 255, 255, 1.000)",
				inverse: "rgba(255, 255, 255, 1.000)",
				disabled: "rgba(181, 185, 212, 1.000)",
				elevated: "rgba(255, 255, 255, 0.000)"
			},
			icon: {
				primary: "rgba(40, 40, 52, 1.000)",
				secondary: "rgba(100, 101, 135, 1.000)",
				accent: "rgba(67, 76, 228, 1.000)",
				"accent-loud-on-fill": "rgba(255, 255, 255, 1.000)",
				"accent-quiet-on-fill": "rgba(67, 76, 228, 1.000)",
				inverse: "rgba(255, 255, 255, 1.000)",
				disabled: "rgba(156, 158, 191, 1.000)",
				"disabled-on-fill": "rgba(255, 255, 255, 0.400)",
				"disabled-on-fill-inverse": "rgba(40, 40, 52, 0.400)"
			},
			text: {
				primary: "rgba(40, 40, 52, 1.000)",
				secondary: "rgba(100, 101, 135, 1.000)",
				accent: "rgba(67, 76, 228, 1.000)",
				"accent-loud-on-fill": "rgba(255, 255, 255, 1.000)",
				"accent-quiet-on-fill": "rgba(67, 76, 228, 1.000)",
				inverse: "rgba(255, 255, 255, 1.000)",
				disabled: "rgba(156, 158, 191, 1.000)",
				"disabled-on-fill": "rgba(255, 255, 255, 0.400)",
				"disabled-on-fill-inverse": "rgba(40, 40, 52, 0.400)",
				link: {
					idle: "rgba(67, 76, 228, 1.000)",
					hover: "rgba(53, 58, 197, 1.000)",
					active: "rgba(39, 44, 150, 1.000)",
					"idle-inverse": "rgba(255, 255, 255, 1.000)",
					"hover-inverse": "rgba(255, 255, 255, 0.900)",
					"active-inverse": "rgba(255, 255, 255, 0.800)",
					"idle-neutral": "rgba(100, 101, 135, 1.000)",
					"hover-neutral": "rgba(67, 69, 101, 1.000)",
					"active-neutral": "rgba(40, 40, 52, 1.000)"
				}
			},
			brand: {
				fill: {
					loud: "rgba(255, 230, 0, 1.000)",
					quiet: "rgba(255, 243, 148, 1.000)"
				},
				gradient: {
					start: "rgba(249, 194, 0, 1.000)",
					end: "rgba(255, 230, 0, 1.000)"
				},
				text: {
					loud: "rgba(255, 230, 0, 1.000)",
					"loud-on-fill": "rgba(40, 40, 52, 1.000)",
					"quiet-on-fill": "rgba(40, 40, 52, 1.000)"
				},
				icon: {
					loud: "rgba(255, 230, 0, 1.000)",
					"loud-on-fill": "rgba(40, 40, 52, 1.000)",
					"quiet-on-fill": "rgba(40, 40, 52, 1.000)"
				}
			},
			feedback: {
				fill: {
					"informative-loud": "rgba(67, 76, 228, 1.000)",
					"informative-quiet": "rgba(233, 241, 255, 1.000)",
					"positive-loud": "rgba(31, 137, 35, 1.000)",
					"positive-quiet": "rgba(222, 250, 222, 1.000)",
					"caution-loud": "rgba(215, 64, 9, 1.000)",
					"caution-quiet": "rgba(255, 237, 199, 1.000)",
					"negative-loud": "rgba(218, 22, 48, 1.000)",
					"negative-quiet": "rgba(255, 229, 233, 1.000)",
					"neutral-loud": "rgba(100, 101, 135, 1.000)",
					"neutral-quiet": "rgba(231, 233, 243, 1.000)"
				},
				text: {
					"informative-loud": "rgba(67, 76, 228, 1.000)",
					"informative-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"informative-quiet-on-fill": "rgba(39, 44, 150, 1.000)",
					"positive-loud": "rgba(31, 137, 35, 1.000)",
					"positive-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"positive-quiet-on-fill": "rgba(18, 69, 21, 1.000)",
					"caution-loud": "rgba(215, 64, 9, 1.000)",
					"caution-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"caution-quiet-on-fill": "rgba(148, 39, 0, 1.000)",
					"negative-loud": "rgba(158, 0, 21, 1.000)",
					"negative-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"negative-quiet-on-fill": "rgba(158, 0, 21, 1.000)",
					"neutral-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"neutral-quiet-on-fill": "rgba(40, 40, 52, 1.000)"
				},
				border: {
					"informative-loud": "rgba(67, 76, 228, 1.000)",
					"positive-loud": "rgba(31, 137, 35, 1.000)",
					"caution-loud": "rgba(215, 64, 9, 1.000)",
					"negative-loud": "rgba(218, 22, 48, 1.000)"
				},
				icon: {
					"informative-loud": "rgba(67, 76, 228, 1.000)",
					"informative-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"informative-quiet-on-fill": "rgba(39, 44, 150, 1.000)",
					"positive-loud": "rgba(31, 137, 35, 1.000)",
					"positive-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"positive-quiet-on-fill": "rgba(18, 69, 21, 1.000)",
					"caution-loud": "rgba(215, 64, 9, 1.000)",
					"caution-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"caution-quiet-on-fill": "rgba(148, 39, 0, 1.000)",
					"negative-loud": "rgba(158, 0, 21, 1.000)",
					"negative-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"negative-quiet-on-fill": "rgba(158, 0, 21, 1.000)",
					"neutral-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"neutral-quiet-on-fill": "rgba(40, 40, 52, 1.000)"
				}
			},
			interactive: {
				fill: {
					loud: {
						idle: "rgba(67, 76, 228, 1.000)",
						"idle-on-fill": "rgba(255, 255, 255, 1.000)",
						"idle-on-fill-inverse": "rgba(255, 255, 255, 1.000)",
						hover: "rgba(53, 58, 197, 1.000)",
						"hover-on-fill": "rgba(255, 255, 255, 0.900)",
						"hover-on-fill-inverse": "rgba(255, 255, 255, 0.900)",
						active: "rgba(39, 44, 150, 1.000)",
						"active-on-fill": "rgba(255, 255, 255, 0.800)",
						"active-on-fill-inverse": "rgba(255, 255, 255, 0.800)"
					},
					quiet: {
						idle: "rgba(233, 241, 255, 1.000)",
						"idle-on-fill": "rgba(40, 40, 52, 0.700)",
						"idle-on-fill-inverse": "rgba(255, 255, 255, 0.700)",
						hover: "rgba(222, 233, 255, 1.000)",
						"hover-on-fill": "rgba(40, 40, 52, 0.800)",
						"hover-on-fill-inverse": "rgba(255, 255, 255, 0.800)",
						active: "rgba(198, 216, 255, 1.000)",
						"active-on-fill": "rgba(40, 40, 52, 0.900)",
						"active-on-fill-inverse": "rgba(255, 255, 255, 0.900)"
					},
					mute: {
						idle: "rgba(255, 255, 255, 0.000)",
						"idle-on-fill": "rgba(255, 255, 255, 0.000)",
						"idle-on-fill-inverse": "rgba(255, 255, 255, 0.000)",
						hover: "rgba(233, 241, 255, 1.000)",
						"hover-on-fill": "rgba(40, 40, 52, 0.300)",
						"hover-on-fill-inverse": "rgba(255, 255, 255, 0.300)",
						active: "rgba(222, 233, 255, 1.000)",
						"active-on-fill": "rgba(40, 40, 52, 0.400)",
						"active-on-fill-inverse": "rgba(255, 255, 255, 0.400)"
					}
				},
				border: {
					idle: "rgba(135, 136, 171, 1.000)",
					hover: "rgba(100, 101, 135, 1.000)",
					active: "rgba(67, 76, 228, 1.000)"
				},
				icon: {
					idle: "rgba(100, 101, 135, 1.000)",
					hover: "rgba(67, 69, 101, 1.000)",
					active: "rgba(40, 40, 52, 1.000)",
					"idle-accent": "rgba(67, 76, 228, 1.000)",
					"hover-accent": "rgba(53, 58, 197, 1.000)",
					"active-accent": "rgba(39, 44, 150, 1.000)",
					"idle-inverse": "rgba(255, 255, 255, 1.000)",
					"hover-inverse": "rgba(255, 255, 255, 0.900)",
					"active-inverse": "rgba(255, 255, 255, 0.800)"
				}
			},
			selected: {
				fill: {
					idle: "rgba(67, 76, 228, 1.000)",
					hover: "rgba(53, 58, 197, 1.000)",
					active: "rgba(39, 44, 150, 1.000)"
				},
				text: {
					idle: "rgba(67, 76, 228, 1.000)",
					hover: "rgba(53, 58, 197, 1.000)",
					active: "rgba(39, 44, 150, 1.000)"
				},
				border: {
					idle: "rgba(67, 76, 228, 1.000)",
					hover: "rgba(53, 58, 197, 1.000)",
					active: "rgba(39, 44, 150, 1.000)"
				},
				icon: {
					idle: "rgba(67, 76, 228, 1.000)",
					hover: "rgba(53, 58, 197, 1.000)",
					active: "rgba(39, 44, 150, 1.000)"
				}
			},
			secondary: {
				bluebell: {
					subtle: "rgba(222, 241, 255, 1.000)",
					soft: "rgba(166, 213, 255, 1.000)",
					vivid: "rgba(88, 176, 252, 1.000)",
					vibrant: "rgba(38, 133, 235, 1.000)",
					intense: "rgba(0, 67, 130, 1.000)"
				},
				lavender: {
					subtle: "rgba(236, 232, 255, 1.000)",
					soft: "rgba(208, 196, 255, 1.000)",
					vivid: "rgba(166, 143, 247, 1.000)",
					vibrant: "rgba(111, 78, 204, 1.000)",
					intense: "rgba(68, 41, 135, 1.000)"
				},
				lemon: {
					subtle: "rgba(240, 250, 200, 1.000)",
					soft: "rgba(217, 250, 140, 1.000)",
					vivid: "rgba(181, 232, 79, 1.000)",
					vibrant: "rgba(147, 204, 53, 1.000)",
					intense: "rgba(79, 115, 23, 1.000)"
				},
				mint: {
					subtle: "rgba(217, 250, 245, 1.000)",
					soft: "rgba(149, 231, 219, 1.000)",
					vivid: "rgba(82, 209, 190, 1.000)",
					vibrant: "rgba(13, 163, 143, 1.000)",
					intense: "rgba(0, 107, 95, 1.000)"
				},
				peach: {
					subtle: "rgba(255, 222, 204, 1.000)",
					soft: "rgba(251, 174, 141, 1.000)",
					vivid: "rgba(232, 112, 84, 1.000)",
					vibrant: "rgba(207, 74, 48, 1.000)",
					intense: "rgba(135, 34, 18, 1.000)"
				},
				rose: {
					subtle: "rgba(255, 224, 242, 1.000)",
					soft: "rgba(255, 171, 209, 1.000)",
					vivid: "rgba(250, 105, 167, 1.000)",
					vibrant: "rgba(219, 64, 132, 1.000)",
					intense: "rgba(145, 17, 76, 1.000)"
				},
				sunflower: {
					subtle: "rgba(252, 245, 199, 1.000)",
					soft: "rgba(235, 215, 129, 1.000)",
					vivid: "rgba(212, 174, 51, 1.000)",
					vibrant: "rgba(168, 133, 17, 1.000)",
					intense: "rgba(92, 72, 0, 1.000)"
				},
				violet: {
					subtle: "rgba(250, 232, 255, 1.000)",
					soft: "rgba(243, 189, 249, 1.000)",
					vivid: "rgba(233, 134, 241, 1.000)",
					vibrant: "rgba(185, 62, 189, 1.000)",
					intense: "rgba(108, 3, 130, 1.000)"
				}
			}
		},
		glass: {
			blur: {
				"default": 16
			},
			progressive: {
				blur: {
					start: 0,
					end: 4
				}
			}
		},
		shadow: {
			low: {
				top: {
					position: {
						x: 0,
						y: -2
					},
					blur: {
						"default": 4
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(40, 40, 52, 0.100)"
					}
				},
				center: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 4
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(40, 40, 52, 0.100)"
					}
				},
				bottom: {
					position: {
						x: 0,
						y: 2
					},
					blur: {
						"default": 4
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(40, 40, 52, 0.100)"
					}
				}
			},
			high: {
				top: {
					position: {
						x: 0,
						y: -8
					},
					blur: {
						"default": 12
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(40, 40, 52, 0.100)"
					}
				},
				center: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 12
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(40, 40, 52, 0.100)"
					}
				},
				bottom: {
					position: {
						x: 0,
						y: 8
					},
					blur: {
						"default": 12
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(40, 40, 52, 0.100)"
					}
				}
			},
			highlight: {
				accent: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 0
					},
					spread: {
						"default": 2
					},
					color: {
						"default": "rgba(222, 233, 255, 1.000)"
					}
				},
				positive: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 0
					},
					spread: {
						"default": 2
					},
					color: {
						"default": "rgba(222, 250, 222, 1.000)"
					}
				},
				caution: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 0
					},
					spread: {
						"default": 2
					},
					color: {
						"default": "rgba(255, 237, 199, 1.000)"
					}
				},
				negative: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 0
					},
					spread: {
						"default": 2
					},
					color: {
						"default": "rgba(255, 229, 233, 1.000)"
					}
				}
			}
		},
		size: {
			height: {
				xnano: 4,
				nano: 6,
				xmicro: 8,
				micro: 12,
				xtiny: 16,
				tiny: 20,
				xsmall: 24,
				small: 32,
				medium: 40,
				large: 48,
				xlarge: 56,
				huge: 64,
				xhuge: 72,
				mega: 80,
				xmega: 88,
				actions: {
					xsmall: 24,
					small: 32,
					medium: 40,
					large: 48
				},
				inputs: {
					small: 32,
					medium: 40,
					large: 48
				},
				"selection-controls": {
					tiny: 20
				},
				icons: {
					micro: 12,
					xtiny: 16,
					tiny: 20,
					xsmall: 24,
					small: 32,
					medium: 40,
					large: 48
				},
				media: {
					xsmall: 88,
					small: 120,
					medium: 128,
					large: 176,
					xlarge: 240,
					huge: 320
				}
			},
			width: {
				xnano: 4,
				nano: 6,
				xmicro: 8,
				micro: 12,
				xtiny: 16,
				tiny: 20,
				xsmall: 24,
				small: 32,
				medium: 40,
				large: 48,
				xlarge: 56,
				huge: 64,
				xhuge: 72,
				mega: 80,
				actions: {
					xsmall: 24,
					small: 32,
					medium: 40,
					large: 48
				},
				inputs: {
					small: 32,
					medium: 40,
					large: 48
				},
				"selection-controls": {
					tiny: 20
				},
				icons: {
					micro: 12,
					xtiny: 16,
					tiny: 20,
					xsmall: 24,
					small: 32,
					medium: 40,
					large: 48
				},
				media: {
					xsmall: 88,
					small: 120,
					medium: 128,
					large: 176,
					xlarge: 240,
					huge: 320
				},
				xmega: 88
			}
		},
		spacing: {
			padding: {
				none: 0,
				pico: 2,
				xnano: 4,
				nano: 6,
				xmicro: 8,
				micro: 12,
				xtiny: 16,
				tiny: 20,
				xsmall: 24,
				small: 32,
				medium: 40,
				large: 48,
				xlarge: 56,
				huge: 64,
				xhuge: 72,
				mega: 80,
				xmega: 88
			},
			gap: {
				none: 0,
				pico: 2,
				xnano: 4,
				nano: 6,
				xmicro: 8,
				micro: 12,
				xtiny: 16,
				tiny: 20,
				xsmall: 24,
				small: 32,
				medium: 40,
				large: 48,
				xlarge: 56,
				huge: 64,
				xhuge: 72,
				mega: 80,
				xmega: 88,
				forms: {
					horizontal: {
						near: 12,
						mid: 16,
						wide: 24,
						related: 8
					},
					vertical: {
						near: 24,
						mid: 32,
						wide: 48
					}
				},
				negative: {
					pico: -2,
					xnano: -4,
					nano: -6,
					xmicro: -8,
					micro: -12,
					xtiny: -16
				}
			}
		},
		text: {
			body: {
				family: {
					"default": "Inter"
				},
				size: {
					xsmall: 10,
					small: 12,
					medium: 14,
					large: 16
				},
				"line-height": {
					xsmall: 12,
					small: 16,
					medium: 18,
					large: 20
				},
				weight: {
					"default": 400,
					emphasis: 600
				},
				"letter-spacing": {
					"default": 0
				},
				"paragraph-spacing": {
					xsmall: 10,
					small: 12,
					medium: 14,
					large: 16
				}
			},
			heading: {
				family: {
					"default": "Inter"
				},
				size: {
					tiny: 10,
					xsmall: 12,
					small: 14,
					medium: 16,
					large: 18,
					xlarge: 20,
					huge: 24,
					xhuge: 28,
					mega: 32
				},
				"line-height": {
					tiny: 12,
					xsmall: 16,
					small: 18,
					medium: 20,
					large: 22,
					xlarge: 24,
					huge: 28,
					xhuge: 34,
					mega: 40
				},
				weight: {
					lean: 400,
					narrow: 600,
					"default": 700
				},
				"letter-spacing": {
					"default": 0
				}
			},
			display: {
				family: {
					"default": "Inter"
				},
				size: {
					xsmall: 40,
					small: 48,
					medium: 56
				},
				"line-height": {
					xsmall: 48,
					small: 56,
					medium: 66
				},
				weight: {
					lean: 400,
					narrow: 600,
					"default": 700
				},
				"letter-spacing": {
					"default": -1
				}
			}
		},
		charts: {
			color: {
				axis: "rgba(208, 212, 230, 1.000)",
				gridline: "rgba(208, 212, 230, 1.000)",
				"reference-line": "rgba(135, 136, 171, 1.000)",
				data: {
					reference: "rgba(208, 212, 230, 1.000)",
					loud: "rgba(67, 76, 228, 1.000)",
					quiet: "rgba(198, 216, 255, 1.000)"
				},
				interactive: {
					data: {
						loud: {
							idle: "rgba(67, 76, 228, 1.000)",
							hover: "rgba(53, 58, 197, 1.000)",
							active: "rgba(39, 44, 150, 1.000)"
						},
						quiet: {
							idle: "rgba(198, 216, 255, 1.000)",
							hover: "rgba(163, 188, 254, 1.000)",
							active: "rgba(122, 149, 255, 1.000)"
						}
					}
				},
				selected: {
					data: {
						idle: "rgba(67, 76, 228, 1.000)",
						hover: "rgba(53, 58, 197, 1.000)",
						active: "rgba(39, 44, 150, 1.000)"
					}
				}
			}
		},
		components: {
			badge: {
				border: {
					radius: {
						"default": 9999,
						"corner-small": 8,
						"corner-large": 12
					}
				}
			},
			button: {
				loud: {
					color: {
						fill: {
							loading: "rgba(53, 58, 197, 1.000)",
							"loading-on-fill": "rgba(255, 255, 255, 0.800)",
							"loading-on-fill-inverse": "rgba(255, 255, 255, 0.800)"
						}
					},
					bar: {
						color: {
							fill: {
								loading: "rgba(39, 44, 150, 1.000)",
								"loading-on-fill": "rgba(255, 255, 255, 1.000)",
								"loading-on-fill-inverse": "rgba(255, 255, 255, 1.000)"
							}
						}
					}
				},
				quiet: {
					color: {
						fill: {
							loading: "rgba(222, 233, 255, 1.000)",
							"loading-on-fill": "rgba(40, 40, 52, 0.300)",
							"loading-on-fill-inverse": "rgba(255, 255, 255, 0.500)"
						}
					},
					bar: {
						color: {
							fill: {
								loading: "rgba(198, 216, 255, 1.000)",
								"loading-on-fill": "rgba(40, 40, 52, 0.500)",
								"loading-on-fill-inverse": "rgba(255, 255, 255, 0.700)"
							}
						}
					}
				},
				mute: {
					color: {
						fill: {
							loading: "rgba(233, 241, 255, 1.000)",
							"loading-on-fill": "rgba(40, 40, 52, 0.100)",
							"loading-on-fill-inverse": "rgba(255, 255, 255, 0.300)"
						}
					},
					bar: {
						color: {
							fill: {
								loading: "rgba(222, 233, 255, 1.000)",
								"loading-on-fill": "rgba(40, 40, 52, 0.200)",
								"loading-on-fill-inverse": "rgba(255, 255, 255, 0.400)"
							}
						}
					}
				},
				split: {
					loud: {
						color: {
							border: "rgba(39, 44, 150, 1.000)",
							"border-on-fill": "rgba(40, 40, 52, 0.100)",
							"border-on-fill-inverse": "rgba(40, 40, 52, 0.100)"
						}
					},
					quiet: {
						color: {
							border: "rgba(198, 216, 255, 1.000)",
							"border-on-fill": "rgba(40, 40, 52, 0.200)",
							"border-on-fill-inverse": "rgba(40, 40, 52, 0.100)"
						}
					}
				}
			},
			card: {
				primary: {
					border: {
						width: {
							"default": 1
						}
					},
					shadow: {
						position: {
							x: 0,
							y: 0
						},
						blur: {
							"default": 0
						},
						spread: {
							"default": 0
						},
						color: {
							"default": "rgba(255, 255, 255, 0.000)"
						}
					}
				}
			},
			"coach-mark": {
				highlighter: {
					border: {
						width: {
							"default": 12
						}
					},
					color: {
						"default": "rgba(255, 255, 255, 0.700)"
					},
					shadow: {
						blur: {
							"default": 0
						},
						color: {
							"default": "rgba(255, 255, 255, 0.300)"
						},
						position: {
							x: 0,
							y: 0
						},
						spread: {
							"default": 28
						}
					}
				}
			},
			"date-picker": {
				range: {
					day: {
						"start-end": {
							color: {
								selected: {
									fill: {
										disabled: "rgba(181, 185, 212, 1.000)"
									}
								}
							}
						}
					}
				}
			},
			header: {
				color: {
					fill: {
						"default": "rgba(255, 255, 255, 0.000)",
						"default-on-scroll": "rgba(255, 255, 255, 0.800)",
						"on-fill": "rgba(255, 255, 255, 0.000)",
						"on-fill-on-scroll": "rgba(40, 40, 52, 0.400)"
					},
					text: {
						"default": "rgba(40, 40, 52, 1.000)",
						"on-fill": "rgba(255, 255, 255, 1.000)"
					}
				},
				text: {
					heading: {
						size: 24,
						weight: 700
					}
				},
				button: {
					color: {
						icon: {
							idle: "rgba(67, 76, 228, 1.000)",
							"idle-on-fill": "rgba(255, 255, 255, 1.000)",
							"idle-on-fill-inverse": "rgba(40, 40, 52, 1.000)"
						},
						fill: {
							idle: "rgba(233, 241, 255, 1.000)",
							active: "rgba(198, 216, 255, 1.000)",
							"idle-on-fill": "rgba(40, 40, 52, 0.700)",
							"active-on-fill": "rgba(40, 40, 52, 0.900)",
							"idle-on-fill-inverse": "rgba(255, 255, 255, 0.700)",
							"active-on-fill-inverse": "rgba(255, 255, 255, 0.900)"
						},
						text: {
							idle: "rgba(67, 76, 228, 1.000)",
							"idle-on-fill": "rgba(255, 255, 255, 1.000)",
							"idle-on-fill-inverse": "rgba(40, 40, 52, 1.000)"
						}
					}
				},
				"status-bar": {
					color: {
						content: {
							"on-fill": "rgba(255, 255, 255, 1.000)"
						}
					}
				}
			},
			"progress-indicator": {
				circular: {
					xsmall: {
						icon: {
							size: {
								height: {
									"default": 10
								},
								width: {
									"default": 10
								}
							}
						}
					}
				}
			},
			"segmented-control": {
				color: {
					fill: "rgba(255, 255, 255, 1.000)"
				},
				segment: {
					shadow: {
						highlight: {
							accent: {
								position: {
									x: 0,
									y: 0
								},
								blur: {
									"default": 0
								},
								spread: {
									"default": 2
								},
								color: {
									accent: "rgba(222, 233, 255, 1.000)"
								}
							}
						}
					},
					border: {
						width: {
							"default": 2
						}
					}
				}
			},
			tabs: {
				pipe: {
					size: {
						height: {
							"default": 3
						}
					}
				}
			},
			thumbnail: {
				spacing: {
					padding: {
						large: 3,
						small: 1
					}
				}
			},
			"view-indicator": {
				color: {
					fill: "rgba(255, 255, 255, 0.400)",
					border: "rgba(255, 255, 255, 0.300)"
				},
				view: {
					color: {
						selected: {
							fill: "rgba(40, 40, 52, 1.000)"
						}
					},
					size: {
						height: {
							selected: {
								small: 8,
								large: 24
							},
							unselected: {
								small: 4,
								medium: 6,
								large: 8
							}
						},
						width: {
							selected: {
								small: 8,
								large: 24
							},
							unselected: {
								small: 4,
								medium: 6,
								large: 8
							}
						}
					}
				}
			},
			tag: {
				color: {
					fill: {
						positive: {
							hover: "rgba(222, 250, 222, 1.000)",
							active: "rgba(189, 245, 190, 1.000)"
						},
						caution: {
							hover: "rgba(255, 237, 199, 1.000)",
							active: "rgba(255, 218, 163, 1.000)"
						},
						negative: {
							hover: "rgba(255, 229, 233, 1.000)",
							active: "rgba(254, 205, 212, 1.000)"
						}
					}
				}
			},
			fab: {
				quiet: {
					color: {
						fill: {
							"default": "rgba(233, 241, 255, 1.000)",
							hover: "rgba(222, 233, 255, 1.000)",
							active: "rgba(198, 216, 255, 1.000)"
						}
					}
				}
			}
		},
		table: {
			cell: {
				color: {
					surface: {
						disabled: "rgba(244, 245, 249, 1.000)"
					}
				}
			}
		}
	},
	dark: {
		color: {
			gradient: {
				overlay: {
					start: "rgba(4, 4, 7, 0.000)",
					end: "rgba(4, 4, 7, 0.800)"
				},
				skeleton: {
					fill: {
						"default": "rgba(25, 26, 46, 1.000)",
						start: "rgba(4, 4, 7, 0.100)",
						early: "rgba(4, 4, 7, 0.900)",
						middle: "rgba(4, 4, 7, 1.000)",
						late: "rgba(4, 4, 7, 0.900)",
						end: "rgba(4, 4, 7, 0.100)"
					}
				}
			},
			background: {
				primary: "rgba(4, 4, 7, 1.000)",
				secondary: "rgba(40, 42, 67, 1.000)",
				overlay: "rgba(4, 4, 7, 0.800)",
				"static": "rgba(4, 4, 7, 1.000)"
			},
			surface: {
				elevated: "rgba(13, 13, 27, 1.000)",
				primary: {
					idle: "rgba(4, 4, 7, 1.000)",
					hover: "rgba(13, 13, 27, 1.000)",
					active: "rgba(25, 26, 46, 1.000)",
					disabled: "rgba(4, 4, 7, 1.000)"
				},
				secondary: {
					idle: "rgba(25, 26, 46, 1.000)",
					hover: "rgba(40, 42, 67, 1.000)",
					active: "rgba(63, 65, 95, 1.000)",
					disabled: "rgba(25, 26, 46, 1.000)"
				}
			},
			fill: {
				primary: "rgba(4, 4, 7, 1.000)",
				"primary-on-fill": "rgba(255, 255, 255, 1.000)",
				"primary-on-fill-inverse": "rgba(40, 40, 52, 1.000)",
				secondary: "rgba(63, 65, 95, 1.000)",
				"secondary-on-fill": "rgba(255, 255, 255, 0.500)",
				"secondary-on-fill-inverse": "rgba(40, 40, 52, 0.300)",
				inverse: "rgba(255, 255, 255, 1.000)",
				disabled: "rgba(40, 42, 67, 1.000)",
				"disabled-on-fill": "rgba(255, 255, 255, 0.200)",
				"disabled-on-fill-inverse": "rgba(4, 4, 7, 0.200)",
				"accent-loud-on-fill": "rgba(255, 255, 255, 1.000)",
				"accent-loud-on-fill-inverse": "rgba(40, 40, 52, 1.000)",
				"accent-quiet": "rgba(27, 39, 85, 1.000)"
			},
			border: {
				primary: "rgba(63, 65, 95, 1.000)",
				"primary-on-fill": "rgba(255, 255, 255, 0.500)",
				"primary-on-fill-inverse": "rgba(40, 40, 52, 0.300)",
				"accent-on-fill": "rgba(255, 255, 255, 1.000)",
				"accent-on-fill-inverse": "rgba(40, 40, 52, 1.000)",
				inverse: "rgba(4, 4, 7, 1.000)",
				disabled: "rgba(63, 65, 95, 1.000)",
				elevated: "rgba(40, 42, 67, 1.000)"
			},
			icon: {
				primary: "rgba(255, 255, 255, 1.000)",
				secondary: "rgba(146, 146, 176, 1.000)",
				accent: "rgba(113, 140, 249, 1.000)",
				"accent-quiet-on-fill": "rgba(255, 255, 255, 1.000)",
				inverse: "rgba(4, 4, 7, 1.000)",
				disabled: "rgba(101, 102, 134, 1.000)",
				"disabled-on-fill": "rgba(4, 4, 7, 0.400)",
				"disabled-on-fill-inverse": "rgba(255, 255, 255, 0.400)"
			},
			text: {
				primary: "rgba(255, 255, 255, 1.000)",
				secondary: "rgba(146, 146, 176, 1.000)",
				accent: "rgba(113, 140, 249, 1.000)",
				"accent-quiet-on-fill": "rgba(255, 255, 255, 1.000)",
				inverse: "rgba(4, 4, 7, 1.000)",
				disabled: "rgba(101, 102, 134, 1.000)",
				"disabled-on-fill": "rgba(4, 4, 7, 0.400)",
				"disabled-on-fill-inverse": "rgba(255, 255, 255, 0.400)",
				link: {
					idle: "rgba(113, 140, 249, 1.000)",
					hover: "rgba(141, 163, 252, 1.000)",
					active: "rgba(163, 182, 255, 1.000)",
					"idle-inverse": "rgba(4, 4, 7, 1.000)",
					"hover-inverse": "rgba(13, 13, 27, 1.000)",
					"active-inverse": "rgba(25, 26, 46, 1.000)",
					"idle-neutral": "rgba(146, 146, 176, 1.000)",
					"hover-neutral": "rgba(183, 181, 207, 1.000)",
					"active-neutral": "rgba(216, 214, 229, 1.000)"
				}
			},
			brand: {
				fill: {
					quiet: "rgba(78, 64, 29, 1.000)"
				},
				gradient: {
					start: "rgba(13, 13, 27, 1.000)",
					end: "rgba(25, 26, 46, 1.000)"
				},
				text: {
					loud: "rgba(250, 231, 107, 1.000)",
					"loud-on-fill": "rgba(4, 4, 7, 1.000)",
					"quiet-on-fill": "rgba(250, 231, 107, 1.000)"
				},
				icon: {
					loud: "rgba(250, 231, 107, 1.000)",
					"loud-on-fill": "rgba(4, 4, 7, 1.000)",
					"quiet-on-fill": "rgba(250, 231, 107, 1.000)"
				}
			},
			feedback: {
				fill: {
					"informative-quiet": "rgba(27, 39, 85, 1.000)",
					"positive-loud": "rgba(35, 198, 94, 1.000)",
					"positive-quiet": "rgba(18, 49, 30, 1.000)",
					"caution-loud": "rgba(255, 145, 36, 1.000)",
					"caution-quiet": "rgba(61, 40, 0, 1.000)",
					"negative-loud": "rgba(245, 77, 99, 1.000)",
					"negative-quiet": "rgba(66, 10, 18, 1.000)",
					"neutral-loud": "rgba(216, 214, 229, 1.000)",
					"neutral-quiet": "rgba(63, 65, 95, 1.000)"
				},
				text: {
					"informative-loud": "rgba(113, 140, 249, 1.000)",
					"informative-quiet-on-fill": "rgba(163, 182, 255, 1.000)",
					"positive-loud": "rgba(35, 198, 94, 1.000)",
					"positive-loud-on-fill": "rgba(18, 49, 30, 1.000)",
					"positive-quiet-on-fill": "rgba(136, 236, 172, 1.000)",
					"caution-loud": "rgba(255, 145, 36, 1.000)",
					"caution-loud-on-fill": "rgba(61, 40, 0, 1.000)",
					"caution-quiet-on-fill": "rgba(255, 200, 128, 1.000)",
					"negative-loud": "rgba(245, 77, 99, 1.000)",
					"negative-loud-on-fill": "rgba(66, 10, 18, 1.000)",
					"negative-quiet-on-fill": "rgba(255, 163, 175, 1.000)",
					"neutral-loud-on-fill": "rgba(13, 13, 27, 1.000)",
					"neutral-quiet-on-fill": "rgba(216, 214, 229, 1.000)"
				},
				border: {
					"informative-loud": "rgba(113, 140, 249, 1.000)",
					"positive-loud": "rgba(35, 198, 94, 1.000)",
					"caution-loud": "rgba(255, 145, 36, 1.000)",
					"negative-loud": "rgba(245, 77, 99, 1.000)"
				},
				icon: {
					"informative-loud": "rgba(113, 140, 249, 1.000)",
					"informative-quiet-on-fill": "rgba(163, 182, 255, 1.000)",
					"positive-loud": "rgba(35, 198, 94, 1.000)",
					"positive-loud-on-fill": "rgba(18, 49, 30, 1.000)",
					"positive-quiet-on-fill": "rgba(136, 236, 172, 1.000)",
					"caution-loud": "rgba(255, 145, 36, 1.000)",
					"caution-loud-on-fill": "rgba(61, 40, 0, 1.000)",
					"caution-quiet-on-fill": "rgba(255, 200, 128, 1.000)",
					"negative-loud": "rgba(245, 77, 99, 1.000)",
					"negative-loud-on-fill": "rgba(66, 10, 18, 1.000)",
					"negative-quiet-on-fill": "rgba(255, 163, 175, 1.000)",
					"neutral-loud-on-fill": "rgba(13, 13, 27, 1.000)",
					"neutral-quiet-on-fill": "rgba(216, 214, 229, 1.000)"
				}
			},
			interactive: {
				fill: {
					loud: {
						"idle-on-fill": "rgba(4, 4, 7, 1.000)",
						"idle-on-fill-inverse": "rgba(4, 4, 7, 1.000)",
						"hover-on-fill": "rgba(4, 4, 7, 0.900)",
						"hover-on-fill-inverse": "rgba(4, 4, 7, 0.900)",
						"active-on-fill": "rgba(4, 4, 7, 0.800)",
						"active-on-fill-inverse": "rgba(4, 4, 7, 0.800)"
					},
					quiet: {
						idle: "rgba(27, 39, 85, 1.000)",
						"idle-on-fill": "rgba(255, 255, 255, 0.700)",
						"idle-on-fill-inverse": "rgba(4, 4, 7, 0.700)",
						hover: "rgba(41, 57, 121, 1.000)",
						"hover-on-fill": "rgba(255, 255, 255, 0.800)",
						"hover-on-fill-inverse": "rgba(4, 4, 7, 0.800)",
						active: "rgba(56, 76, 157, 1.000)",
						"active-on-fill": "rgba(255, 255, 255, 0.900)",
						"active-on-fill-inverse": "rgba(4, 4, 7, 0.900)"
					},
					mute: {
						idle: "rgba(4, 4, 7, 0.000)",
						"idle-on-fill": "rgba(4, 4, 7, 0.000)",
						"idle-on-fill-inverse": "rgba(4, 4, 7, 0.000)",
						hover: "rgba(27, 39, 85, 1.000)",
						"hover-on-fill": "rgba(255, 255, 255, 0.300)",
						"hover-on-fill-inverse": "rgba(4, 4, 7, 0.300)",
						active: "rgba(41, 57, 121, 1.000)",
						"active-on-fill": "rgba(255, 255, 255, 0.400)",
						"active-on-fill-inverse": "rgba(4, 4, 7, 0.400)"
					}
				},
				border: {
					idle: "rgba(101, 102, 134, 1.000)",
					hover: "rgba(146, 146, 176, 1.000)",
					active: "rgba(113, 140, 249, 1.000)"
				},
				icon: {
					idle: "rgba(146, 146, 176, 1.000)",
					hover: "rgba(183, 181, 207, 1.000)",
					active: "rgba(216, 214, 229, 1.000)",
					"idle-accent": "rgba(113, 140, 249, 1.000)",
					"hover-accent": "rgba(141, 163, 252, 1.000)",
					"active-accent": "rgba(163, 182, 255, 1.000)",
					"idle-inverse": "rgba(4, 4, 7, 1.000)",
					"hover-inverse": "rgba(13, 13, 27, 1.000)",
					"active-inverse": "rgba(25, 26, 46, 1.000)"
				}
			},
			selected: {
				text: {
					idle: "rgba(113, 140, 249, 1.000)",
					hover: "rgba(141, 163, 252, 1.000)",
					active: "rgba(163, 182, 255, 1.000)"
				},
				border: {
					idle: "rgba(113, 140, 249, 1.000)",
					hover: "rgba(141, 163, 252, 1.000)",
					active: "rgba(163, 182, 255, 1.000)"
				},
				icon: {
					idle: "rgba(113, 140, 249, 1.000)",
					hover: "rgba(141, 163, 252, 1.000)",
					active: "rgba(163, 182, 255, 1.000)"
				}
			},
			secondary: {
				bluebell: {
					subtle: "rgba(27, 55, 84, 1.000)",
					soft: "rgba(24, 81, 144, 1.000)",
					vivid: "rgba(54, 140, 214, 1.000)",
					vibrant: "rgba(100, 180, 250, 1.000)",
					intense: "rgba(154, 211, 255, 1.000)"
				},
				lavender: {
					subtle: "rgba(47, 37, 92, 1.000)",
					soft: "rgba(68, 48, 125, 1.000)",
					vivid: "rgba(125, 100, 209, 1.000)",
					vibrant: "rgba(166, 143, 255, 1.000)",
					intense: "rgba(203, 190, 255, 1.000)"
				},
				lemon: {
					subtle: "rgba(41, 53, 25, 1.000)",
					soft: "rgba(90, 124, 35, 1.000)",
					vivid: "rgba(137, 178, 54, 1.000)",
					vibrant: "rgba(175, 217, 76, 1.000)",
					intense: "rgba(221, 239, 150, 1.000)"
				},
				mint: {
					subtle: "rgba(26, 62, 63, 1.000)",
					soft: "rgba(9, 100, 89, 1.000)",
					vivid: "rgba(44, 148, 133, 1.000)",
					vibrant: "rgba(84, 199, 182, 1.000)",
					intense: "rgba(152, 227, 215, 1.000)"
				},
				peach: {
					subtle: "rgba(82, 33, 25, 1.000)",
					soft: "rgba(126, 46, 32, 1.000)",
					vivid: "rgba(189, 76, 51, 1.000)",
					vibrant: "rgba(229, 111, 62, 1.000)",
					intense: "rgba(255, 174, 128, 1.000)"
				},
				rose: {
					subtle: "rgba(77, 31, 54, 1.000)",
					soft: "rgba(133, 40, 82, 1.000)",
					vivid: "rgba(194, 78, 128, 1.000)",
					vibrant: "rgba(255, 128, 186, 1.000)",
					intense: "rgba(255, 181, 224, 1.000)"
				},
				sunflower: {
					subtle: "rgba(67, 39, 13, 1.000)",
					soft: "rgba(103, 70, 13, 1.000)",
					vivid: "rgba(163, 113, 20, 1.000)",
					vibrant: "rgba(229, 174, 23, 1.000)",
					intense: "rgba(240, 212, 134, 1.000)"
				},
				violet: {
					subtle: "rgba(69, 31, 77, 1.000)",
					soft: "rgba(113, 39, 116, 1.000)",
					vivid: "rgba(169, 88, 176, 1.000)",
					vibrant: "rgba(224, 127, 235, 1.000)",
					intense: "rgba(240, 184, 255, 1.000)"
				}
			}
		},
		shadow: {
			low: {
				top: {
					color: {
						"default": "rgba(4, 4, 7, 0.900)"
					}
				},
				center: {
					color: {
						"default": "rgba(4, 4, 7, 0.900)"
					}
				},
				bottom: {
					color: {
						"default": "rgba(4, 4, 7, 0.900)"
					}
				}
			},
			high: {
				top: {
					color: {
						"default": "rgba(4, 4, 7, 0.900)"
					}
				},
				center: {
					color: {
						"default": "rgba(4, 4, 7, 0.900)"
					}
				},
				bottom: {
					color: {
						"default": "rgba(4, 4, 7, 0.900)"
					}
				}
			},
			highlight: {
				accent: {
					color: {
						"default": "rgba(27, 39, 85, 1.000)"
					}
				},
				positive: {
					color: {
						"default": "rgba(18, 49, 30, 1.000)"
					}
				},
				caution: {
					color: {
						"default": "rgba(61, 40, 0, 1.000)"
					}
				},
				negative: {
					color: {
						"default": "rgba(66, 10, 18, 1.000)"
					}
				}
			}
		},
		charts: {
			color: {
				axis: "rgba(63, 65, 95, 1.000)",
				gridline: "rgba(63, 65, 95, 1.000)",
				"reference-line": "rgba(183, 181, 207, 1.000)",
				data: {
					reference: "rgba(63, 65, 95, 1.000)",
					quiet: "rgba(27, 39, 85, 1.000)"
				}
			}
		},
		components: {
			button: {
				loud: {
					color: {
						fill: {
							"loading-on-fill": "rgba(4, 4, 7, 0.800)",
							"loading-on-fill-inverse": "rgba(4, 4, 7, 0.800)"
						}
					},
					bar: {
						color: {
							fill: {
								"loading-on-fill": "rgba(4, 4, 7, 1.000)",
								"loading-on-fill-inverse": "rgba(4, 4, 7, 1.000)"
							}
						}
					}
				},
				quiet: {
					color: {
						fill: {
							loading: "rgba(41, 57, 121, 1.000)",
							"loading-on-fill": "rgba(255, 255, 255, 0.500)",
							"loading-on-fill-inverse": "rgba(4, 4, 7, 0.300)"
						}
					},
					bar: {
						color: {
							fill: {
								loading: "rgba(56, 76, 157, 1.000)",
								"loading-on-fill": "rgba(255, 255, 255, 0.700)",
								"loading-on-fill-inverse": "rgba(4, 4, 7, 0.500)"
							}
						}
					}
				},
				mute: {
					color: {
						fill: {
							loading: "rgba(27, 39, 85, 1.000)",
							"loading-on-fill": "rgba(255, 255, 255, 0.300)",
							"loading-on-fill-inverse": "rgba(4, 4, 7, 0.100)"
						}
					},
					bar: {
						color: {
							fill: {
								loading: "rgba(41, 57, 121, 1.000)",
								"loading-on-fill": "rgba(255, 255, 255, 0.400)",
								"loading-on-fill-inverse": "rgba(4, 4, 7, 0.200)"
							}
						}
					}
				},
				split: {
					loud: {
						color: {
							"border-on-fill": "rgba(255, 255, 255, 0.200)",
							"border-on-fill-inverse": "rgba(255, 255, 255, 0.200)"
						}
					},
					quiet: {
						color: {
							border: "rgba(56, 76, 157, 1.000)",
							"border-on-fill": "rgba(255, 255, 255, 0.300)",
							"border-on-fill-inverse": "rgba(255, 255, 255, 0.200)"
						}
					}
				}
			},
			card: {
				primary: {
					shadow: {
						color: {
							"default": "rgba(4, 4, 7, 0.000)"
						}
					}
				}
			},
			"date-picker": {
				range: {
					day: {
						"start-end": {
							color: {
								selected: {
									fill: {
										disabled: "rgba(63, 65, 95, 1.000)"
									}
								}
							}
						}
					}
				}
			},
			header: {
				color: {
					fill: {
						"default": "rgba(4, 4, 7, 0.000)",
						"default-on-scroll": "rgba(4, 4, 7, 0.800)",
						"on-fill": "rgba(4, 4, 7, 0.000)",
						"on-fill-on-scroll": "rgba(255, 255, 255, 0.400)"
					},
					text: {
						"default": "rgba(255, 255, 255, 1.000)",
						"on-fill": "rgba(4, 4, 7, 1.000)"
					}
				},
				button: {
					color: {
						icon: {
							idle: "rgba(255, 255, 255, 1.000)",
							"idle-on-fill": "rgba(4, 4, 7, 1.000)",
							"idle-on-fill-inverse": "rgba(255, 255, 255, 1.000)"
						},
						fill: {
							idle: "rgba(27, 39, 85, 1.000)",
							active: "rgba(56, 76, 157, 1.000)",
							"idle-on-fill": "rgba(255, 255, 255, 0.700)",
							"active-on-fill": "rgba(255, 255, 255, 0.900)",
							"idle-on-fill-inverse": "rgba(4, 4, 7, 0.700)",
							"active-on-fill-inverse": "rgba(4, 4, 7, 0.900)"
						},
						text: {
							idle: "rgba(255, 255, 255, 1.000)",
							"idle-on-fill": "rgba(4, 4, 7, 1.000)",
							"idle-on-fill-inverse": "rgba(255, 255, 255, 1.000)"
						}
					}
				},
				"status-bar": {
					color: {
						content: {
							"on-fill": "rgba(4, 4, 7, 1.000)"
						}
					}
				}
			},
			"segmented-control": {
				color: {
					fill: "rgba(4, 4, 7, 1.000)"
				},
				segment: {
					shadow: {
						highlight: {
							accent: {
								color: {
									accent: "rgba(255, 255, 255, 0.100)"
								}
							}
						}
					}
				}
			},
			"view-indicator": {
				color: {
					fill: "rgba(4, 4, 7, 0.400)",
					border: "rgba(4, 4, 7, 0.300)"
				},
				view: {
					color: {
						selected: {
							fill: "rgba(216, 214, 229, 1.000)"
						}
					}
				}
			},
			tag: {
				color: {
					fill: {
						positive: {
							hover: "rgba(18, 49, 30, 1.000)",
							active: "rgba(22, 76, 42, 1.000)"
						},
						caution: {
							hover: "rgba(61, 40, 0, 1.000)",
							active: "rgba(107, 60, 0, 1.000)"
						},
						negative: {
							hover: "rgba(66, 10, 18, 1.000)",
							active: "rgba(99, 14, 26, 1.000)"
						}
					}
				}
			},
			fab: {
				quiet: {
					color: {
						fill: {
							"default": "rgba(27, 39, 85, 1.000)",
							hover: "rgba(41, 57, 121, 1.000)",
							active: "rgba(56, 76, 157, 1.000)"
						}
					}
				}
			}
		},
		table: {
			cell: {
				color: {
					surface: {
						disabled: "rgba(25, 26, 46, 1.000)"
					}
				}
			}
		}
	}
};
var andesXFallback = {
	id: id$1,
	name: name$1,
	version: version$1,
	value: value$1
};

var id = "andes_legacy:1.4.0";
var name = "andes_legacy";
var version = "1.4.0";
var value = {
	"default": {
		border: {
			radius: {
				full: 9999,
				none: 0,
				tiny: 4,
				xsmall: 6,
				small: 8,
				medium: 12,
				large: 16,
				xlarge: 20,
				huge: 24,
				containers: {
					"default": 6,
					compact: 4
				},
				"actions-inputs": {
					"default": 6,
					compact: 4
				},
				"square-to-circle": {
					large: 9999,
					small: 9999
				}
			},
			width: {
				none: 0,
				small: 1,
				medium: 2,
				large: 3,
				xlarge: 4
			}
		},
		color: {
			gradient: {
				skeleton: {
					fill: {
						start: "rgba(0, 0, 0, 0.000)",
						early: "rgba(0, 0, 0, 0.050)",
						middle: "rgba(0, 0, 0, 0.150)",
						late: "rgba(0, 0, 0, 0.050)",
						end: "rgba(0, 0, 0, 0.000)",
						"default": "rgba(0, 0, 0, 0.070)"
					}
				},
				overlay: {
					start: "rgba(255, 255, 255, 0.000)",
					end: "rgba(255, 255, 255, 0.800)"
				}
			},
			background: {
				primary: "rgba(237, 237, 237, 1.000)",
				secondary: "rgba(245, 245, 245, 1.000)",
				overlay: "rgba(0, 0, 0, 0.800)",
				"static": "rgba(255, 255, 255, 1.000)"
			},
			surface: {
				elevated: "rgba(255, 255, 255, 1.000)",
				primary: {
					idle: "rgba(255, 255, 255, 1.000)",
					hover: "rgba(0, 0, 0, 0.040)",
					active: "rgba(0, 0, 0, 0.070)",
					disabled: "rgba(255, 255, 255, 1.000)"
				},
				secondary: {
					idle: "rgba(245, 245, 245, 1.000)",
					hover: "rgba(0, 0, 0, 0.040)",
					active: "rgba(0, 0, 0, 0.070)",
					disabled: "rgba(245, 245, 245, 1.000)"
				}
			},
			fill: {
				primary: "rgba(255, 255, 255, 1.000)",
				"primary-on-fill": "rgba(26, 26, 26, 1.000)",
				"primary-on-fill-inverse": "rgba(255, 255, 255, 1.000)",
				secondary: "rgba(0, 0, 0, 0.100)",
				"secondary-on-fill": "rgba(0, 0, 0, 0.150)",
				"secondary-on-fill-inverse": "rgba(255, 255, 255, 0.250)",
				inverse: "rgba(0, 0, 0, 0.900)",
				disabled: "rgba(0, 0, 0, 0.100)",
				"disabled-on-fill": "rgba(0, 0, 0, 0.150)",
				"disabled-on-fill-inverse": "rgba(255, 255, 255, 0.250)",
				"accent-loud": "rgba(52, 131, 250, 1.000)",
				"accent-loud-on-fill": "rgba(26, 26, 26, 1.000)",
				"accent-loud-on-fill-inverse": "rgba(255, 255, 255, 1.000)",
				"accent-quiet": "rgba(65, 137, 230, 0.100)"
			},
			border: {
				primary: "rgba(0, 0, 0, 0.100)",
				"primary-on-fill": "rgba(0, 0, 0, 0.150)",
				"primary-on-fill-inverse": "rgba(255, 255, 255, 0.250)",
				accent: "rgba(52, 131, 250, 1.000)",
				"accent-on-fill": "rgba(26, 26, 26, 1.000)",
				"accent-on-fill-inverse": "rgba(255, 255, 255, 1.000)",
				inverse: "rgba(255, 255, 255, 1.000)",
				disabled: "rgba(0, 0, 0, 0.250)",
				elevated: "rgba(255, 255, 255, 0.000)"
			},
			icon: {
				primary: "rgba(0, 0, 0, 0.900)",
				secondary: "rgba(0, 0, 0, 0.550)",
				accent: "rgba(52, 131, 250, 1.000)",
				"accent-loud-on-fill": "rgba(255, 255, 255, 1.000)",
				"accent-quiet-on-fill": "rgba(52, 131, 250, 1.000)",
				inverse: "rgba(255, 255, 255, 1.000)",
				disabled: "rgba(0, 0, 0, 0.250)",
				"disabled-on-fill": "rgba(255, 255, 255, 0.250)",
				"disabled-on-fill-inverse": "rgba(0, 0, 0, 0.250)"
			},
			text: {
				primary: "rgba(0, 0, 0, 0.900)",
				secondary: "rgba(0, 0, 0, 0.550)",
				accent: "rgba(52, 131, 250, 1.000)",
				"accent-loud-on-fill": "rgba(255, 255, 255, 1.000)",
				"accent-quiet-on-fill": "rgba(52, 131, 250, 1.000)",
				inverse: "rgba(255, 255, 255, 1.000)",
				disabled: "rgba(0, 0, 0, 0.250)",
				"disabled-on-fill": "rgba(255, 255, 255, 0.250)",
				"disabled-on-fill-inverse": "rgba(0, 0, 0, 0.250)",
				link: {
					idle: "rgba(52, 131, 250, 1.000)",
					hover: "rgba(41, 104, 200, 1.000)",
					active: "rgba(31, 78, 150, 1.000)",
					"idle-inverse": "rgba(255, 255, 255, 1.000)",
					"hover-inverse": "rgba(255, 255, 255, 1.000)",
					"active-inverse": "rgba(255, 255, 255, 1.000)",
					"idle-neutral": "rgba(0, 0, 0, 0.550)",
					"hover-neutral": "rgba(0, 0, 0, 0.800)",
					"active-neutral": "rgba(0, 0, 0, 0.900)"
				}
			},
			brand: {
				fill: {
					loud: "rgba(255, 230, 0, 1.000)",
					quiet: "rgba(255, 243, 148, 1.000)"
				},
				gradient: {
					start: "rgba(249, 194, 0, 1.000)",
					end: "rgba(255, 230, 0, 1.000)"
				},
				text: {
					loud: "rgba(255, 230, 0, 1.000)",
					"loud-on-fill": "rgba(0, 0, 0, 0.900)",
					"quiet-on-fill": "rgba(0, 0, 0, 0.900)"
				},
				icon: {
					loud: "rgba(255, 230, 0, 1.000)",
					"loud-on-fill": "rgba(0, 0, 0, 0.900)",
					"quiet-on-fill": "rgba(0, 0, 0, 0.900)"
				}
			},
			feedback: {
				fill: {
					"informative-loud": "rgba(52, 131, 250, 1.000)",
					"informative-quiet": "rgba(65, 137, 230, 0.100)",
					"positive-loud": "rgba(0, 166, 80, 1.000)",
					"positive-quiet": "rgba(0, 166, 80, 0.100)",
					"caution-loud": "rgba(255, 119, 51, 1.000)",
					"caution-quiet": "rgba(255, 119, 51, 0.100)",
					"negative-loud": "rgba(242, 61, 79, 1.000)",
					"negative-quiet": "rgba(242, 61, 79, 0.100)",
					"neutral-loud": "rgba(115, 115, 115, 1.000)",
					"neutral-quiet": "rgba(0, 0, 0, 0.100)"
				},
				text: {
					"informative-loud": "rgba(52, 131, 250, 1.000)",
					"informative-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"informative-quiet-on-fill": "rgba(41, 104, 200, 1.000)",
					"positive-loud": "rgba(0, 166, 80, 1.000)",
					"positive-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"positive-quiet-on-fill": "rgba(0, 135, 68, 1.000)",
					"caution-loud": "rgba(255, 119, 51, 1.000)",
					"caution-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"caution-quiet-on-fill": "rgba(230, 84, 11, 1.000)",
					"negative-loud": "rgba(242, 61, 79, 1.000)",
					"negative-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"negative-quiet-on-fill": "rgba(209, 36, 64, 1.000)",
					"neutral-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"neutral-quiet-on-fill": "rgba(0, 0, 0, 0.550)"
				},
				border: {
					"informative-loud": "rgba(52, 131, 250, 1.000)",
					"positive-loud": "rgba(0, 166, 80, 1.000)",
					"caution-loud": "rgba(255, 119, 51, 1.000)",
					"negative-loud": "rgba(242, 61, 79, 1.000)"
				},
				icon: {
					"informative-loud": "rgba(52, 131, 250, 1.000)",
					"informative-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"informative-quiet-on-fill": "rgba(41, 104, 200, 1.000)",
					"positive-loud": "rgba(0, 166, 80, 1.000)",
					"positive-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"positive-quiet-on-fill": "rgba(0, 135, 68, 1.000)",
					"caution-loud": "rgba(255, 119, 51, 1.000)",
					"caution-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"caution-quiet-on-fill": "rgba(230, 84, 11, 1.000)",
					"negative-loud": "rgba(242, 61, 79, 1.000)",
					"negative-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"negative-quiet-on-fill": "rgba(209, 36, 64, 1.000)",
					"neutral-loud-on-fill": "rgba(255, 255, 255, 1.000)",
					"neutral-quiet-on-fill": "rgba(0, 0, 0, 0.550)"
				}
			},
			interactive: {
				fill: {
					loud: {
						idle: "rgba(52, 131, 250, 1.000)",
						"idle-on-fill": "rgba(255, 255, 255, 1.000)",
						"idle-on-fill-inverse": "rgba(255, 255, 255, 1.000)",
						hover: "rgba(41, 104, 200, 1.000)",
						"hover-on-fill": "rgba(255, 255, 255, 0.800)",
						"hover-on-fill-inverse": "rgba(255, 255, 255, 0.800)",
						active: "rgba(31, 78, 150, 1.000)",
						"active-on-fill": "rgba(255, 255, 255, 0.550)",
						"active-on-fill-inverse": "rgba(255, 255, 255, 0.550)"
					},
					quiet: {
						idle: "rgba(65, 137, 230, 0.150)",
						"idle-on-fill": "rgba(0, 0, 0, 0.150)",
						"idle-on-fill-inverse": "rgba(255, 255, 255, 0.250)",
						hover: "rgba(65, 137, 230, 0.200)",
						"hover-on-fill": "rgba(0, 0, 0, 0.250)",
						"hover-on-fill-inverse": "rgba(255, 255, 255, 0.450)",
						active: "rgba(65, 137, 230, 0.300)",
						"active-on-fill": "rgba(0, 0, 0, 0.450)",
						"active-on-fill-inverse": "rgba(255, 255, 255, 0.800)"
					},
					mute: {
						idle: "rgba(255, 255, 255, 0.000)",
						"idle-on-fill": "rgba(255, 255, 255, 0.000)",
						"idle-on-fill-inverse": "rgba(255, 255, 255, 0.000)",
						hover: "rgba(65, 137, 230, 0.100)",
						"hover-on-fill": "rgba(0, 0, 0, 0.070)",
						"hover-on-fill-inverse": "rgba(255, 255, 255, 0.250)",
						active: "rgba(65, 137, 230, 0.200)",
						"active-on-fill": "rgba(0, 0, 0, 0.100)",
						"active-on-fill-inverse": "rgba(255, 255, 255, 0.450)"
					}
				},
				border: {
					idle: "rgba(0, 0, 0, 0.250)",
					hover: "rgba(0, 0, 0, 0.550)",
					active: "rgba(52, 131, 250, 1.000)"
				},
				icon: {
					idle: "rgba(0, 0, 0, 0.550)",
					hover: "rgba(0, 0, 0, 0.800)",
					active: "rgba(0, 0, 0, 0.900)",
					"idle-accent": "rgba(52, 131, 250, 1.000)",
					"hover-accent": "rgba(41, 104, 200, 1.000)",
					"active-accent": "rgba(31, 78, 150, 1.000)",
					"idle-inverse": "rgba(255, 255, 255, 1.000)",
					"hover-inverse": "rgba(255, 255, 255, 1.000)",
					"active-inverse": "rgba(255, 255, 255, 1.000)"
				}
			},
			selected: {
				fill: {
					idle: "rgba(52, 131, 250, 1.000)",
					hover: "rgba(41, 104, 200, 1.000)",
					active: "rgba(31, 78, 150, 1.000)"
				},
				text: {
					idle: "rgba(52, 131, 250, 1.000)",
					hover: "rgba(41, 104, 200, 1.000)",
					active: "rgba(31, 78, 150, 1.000)"
				},
				border: {
					idle: "rgba(52, 131, 250, 1.000)",
					hover: "rgba(41, 104, 200, 1.000)",
					active: "rgba(31, 78, 150, 1.000)"
				},
				icon: {
					idle: "rgba(52, 131, 250, 1.000)",
					hover: "rgba(41, 104, 200, 1.000)",
					active: "rgba(31, 78, 150, 1.000)"
				}
			},
			secondary: {
				bluebell: {
					subtle: "rgba(198, 231, 253, 1.000)",
					soft: "rgba(139, 206, 250, 1.000)",
					vivid: "rgba(62, 173, 247, 1.000)",
					vibrant: "rgba(40, 115, 187, 1.000)",
					intense: "rgba(30, 83, 127, 1.000)"
				},
				lavender: {
					subtle: "rgba(223, 204, 251, 1.000)",
					soft: "rgba(190, 153, 246, 1.000)",
					vivid: "rgba(147, 85, 240, 1.000)",
					vibrant: "rgba(94, 54, 179, 1.000)",
					intense: "rgba(70, 41, 122, 1.000)"
				},
				lemon: {
					subtle: "rgba(221, 247, 204, 1.000)",
					soft: "rgba(185, 239, 152, 1.000)",
					vivid: "rgba(139, 228, 83, 1.000)",
					vibrant: "rgba(89, 167, 53, 1.000)",
					intense: "rgba(66, 113, 40, 1.000)"
				},
				mint: {
					subtle: "rgba(194, 246, 242, 1.000)",
					soft: "rgba(131, 235, 228, 1.000)",
					vivid: "rgba(49, 222, 210, 1.000)",
					vibrant: "rgba(31, 163, 151, 1.000)",
					intense: "rgba(23, 109, 101, 1.000)"
				},
				peach: {
					subtle: "rgba(255, 233, 203, 1.000)",
					soft: "rgba(255, 210, 151, 1.000)",
					vivid: "rgba(255, 180, 82, 1.000)",
					vibrant: "rgba(194, 122, 53, 1.000)",
					intense: "rgba(133, 86, 39, 1.000)"
				},
				rose: {
					subtle: "rgba(253, 213, 237, 1.000)",
					soft: "rgba(250, 171, 218, 1.000)",
					vivid: "rgba(247, 115, 193, 1.000)",
					vibrant: "rgba(187, 74, 134, 1.000)",
					intense: "rgba(127, 55, 93, 1.000)"
				},
				sunflower: {
					subtle: "rgba(255, 241, 203, 1.000)",
					soft: "rgba(255, 226, 151, 1.000)",
					vivid: "rgba(255, 206, 82, 1.000)",
					vibrant: "rgba(194, 147, 53, 1.000)",
					intense: "rgba(133, 99, 39, 1.000)"
				},
				violet: {
					subtle: "rgba(246, 208, 244, 1.000)",
					soft: "rgba(236, 160, 233, 1.000)",
					vivid: "rgba(224, 96, 218, 1.000)",
					vibrant: "rgba(164, 62, 158, 1.000)",
					intense: "rgba(110, 46, 106, 1.000)"
				}
			}
		},
		ai: {
			gradient: {
				"start-early": "rgba(18, 177, 238, 1.000)",
				early: "rgba(0, 214, 241, 1.000)",
				"early-middle": "rgba(68, 225, 242, 1.000)",
				middle: "rgba(180, 244, 244, 1.000)",
				"middle-late": "rgba(226, 248, 222, 1.000)",
				late: "rgba(236, 247, 198, 1.000)",
				start: "rgba(52, 131, 250, 1.000)",
				"late-end": "rgba(255, 243, 148, 1.000)",
				end: "rgba(255, 230, 0, 1.000)"
			},
			text: {
				body: {
					family: {
						"default": "Proxima Nova"
					},
					size: {
						medium: 14,
						large: 16
					},
					"line-height": {
						medium: 20,
						large: 22
					},
					weight: {
						"default": 400,
						emphasis: 600
					},
					"letter-spacing": {
						"default": 0
					},
					"paragraph-spacing": {
						xsmall: 9,
						small: 10,
						medium: 12,
						large: 14
					}
				},
				heading: {
					family: {
						"default": "Proxima Nova"
					},
					size: {
						large: 18,
						xlarge: 20,
						huge: 24
					},
					"line-height": {
						large: 22,
						xlarge: 25,
						huge: 30
					},
					weight: {
						lean: 400,
						narrow: 600,
						"default": 600
					},
					"letter-spacing": {
						"default": 0
					}
				}
			}
		},
		components: {
			fab: {
				quiet: {
					color: {
						fill: {
							hover: "rgba(217, 231, 250, 1.000)",
							active: "rgba(198, 220, 247, 1.000)",
							"default": "rgba(255, 255, 255, 1.000)"
						}
					}
				}
			},
			badge: {
				border: {
					radius: {
						"default": 9999,
						"corner-small": 8,
						"corner-large": 12
					}
				}
			},
			button: {
				loud: {
					color: {
						fill: {
							loading: "rgba(41, 104, 200, 1.000)",
							"loading-on-fill": "rgba(255, 255, 255, 0.800)",
							"loading-on-fill-inverse": "rgba(255, 255, 255, 0.800)"
						}
					},
					bar: {
						color: {
							fill: {
								loading: "rgba(31, 78, 150, 1.000)",
								"loading-on-fill": "rgba(255, 255, 255, 0.550)",
								"loading-on-fill-inverse": "rgba(255, 255, 255, 0.550)"
							}
						}
					}
				},
				quiet: {
					color: {
						fill: {
							loading: "rgba(65, 137, 230, 0.200)",
							"loading-on-fill": "rgba(0, 0, 0, 0.250)",
							"loading-on-fill-inverse": "rgba(255, 255, 255, 0.550)"
						}
					},
					bar: {
						color: {
							fill: {
								loading: "rgba(65, 137, 230, 0.300)",
								"loading-on-fill": "rgba(0, 0, 0, 0.450)",
								"loading-on-fill-inverse": "rgba(255, 255, 255, 0.800)"
							}
						}
					}
				},
				mute: {
					color: {
						fill: {
							loading: "rgba(65, 137, 230, 0.100)",
							"loading-on-fill": "rgba(0, 0, 0, 0.070)",
							"loading-on-fill-inverse": "rgba(255, 255, 255, 0.250)"
						}
					},
					bar: {
						color: {
							fill: {
								loading: "rgba(65, 137, 230, 0.200)",
								"loading-on-fill": "rgba(0, 0, 0, 0.100)",
								"loading-on-fill-inverse": "rgba(255, 255, 255, 0.450)"
							}
						}
					}
				},
				split: {
					loud: {
						color: {
							border: "rgba(0, 0, 0, 0.100)",
							"border-on-fill": "rgba(0, 0, 0, 0.100)",
							"border-on-fill-inverse": "rgba(0, 0, 0, 0.100)"
						}
					},
					quiet: {
						color: {
							border: "rgba(0, 0, 0, 0.100)",
							"border-on-fill": "rgba(0, 0, 0, 0.100)",
							"border-on-fill-inverse": "rgba(0, 0, 0, 0.100)"
						}
					}
				}
			},
			card: {
				primary: {
					border: {
						width: {
							"default": 0
						}
					},
					shadow: {
						position: {
							x: 0,
							y: 6
						},
						blur: {
							"default": 16
						},
						spread: {
							"default": 0
						},
						color: {
							"default": "rgba(0, 0, 0, 0.100)"
						}
					}
				}
			},
			"coach-mark": {
				highlighter: {
					border: {
						width: {
							"default": 12
						}
					},
					color: {
						"default": "rgba(255, 255, 255, 0.800)"
					},
					shadow: {
						blur: {
							"default": 0
						},
						color: {
							"default": "rgba(255, 255, 255, 0.040)"
						},
						position: {
							x: 0,
							y: 0
						},
						spread: {
							"default": 28
						}
					}
				}
			},
			"date-picker": {
				range: {
					day: {
						"start-end": {
							color: {
								selected: {
									fill: {
										disabled: "rgba(0, 0, 0, 0.100)"
									}
								}
							}
						}
					}
				}
			},
			header: {
				color: {
					fill: {
						"default": "rgba(255, 230, 0, 1.000)",
						"default-on-scroll": "rgba(255, 230, 0, 1.000)",
						"on-fill": "rgba(255, 230, 0, 1.000)",
						"on-fill-on-scroll": "rgba(255, 230, 0, 1.000)"
					},
					text: {
						"default": "rgba(26, 26, 26, 1.000)",
						"on-fill": "rgba(26, 26, 26, 1.000)"
					}
				},
				text: {
					heading: {
						size: 20,
						weight: 400
					}
				},
				button: {
					color: {
						icon: {
							idle: "rgba(0, 0, 0, 0.900)",
							"idle-on-fill": "rgba(0, 0, 0, 0.900)",
							"idle-on-fill-inverse": "rgba(0, 0, 0, 0.900)"
						},
						fill: {
							idle: "rgba(255, 255, 255, 0.250)",
							active: "rgba(255, 255, 255, 0.800)",
							"idle-on-fill": "rgba(255, 255, 255, 0.250)",
							"active-on-fill": "rgba(255, 255, 255, 0.800)",
							"idle-on-fill-inverse": "rgba(255, 255, 255, 0.250)",
							"active-on-fill-inverse": "rgba(255, 255, 255, 0.800)"
						},
						text: {
							idle: "rgba(0, 0, 0, 0.900)",
							"idle-on-fill": "rgba(0, 0, 0, 0.900)",
							"idle-on-fill-inverse": "rgba(0, 0, 0, 0.900)"
						}
					}
				},
				"status-bar": {
					color: {
						content: {
							"on-fill": "rgba(0, 0, 0, 0.900)"
						}
					}
				}
			},
			"progress-indicator": {
				circular: {
					xsmall: {
						icon: {
							size: {
								height: {
									"default": 10
								},
								width: {
									"default": 10
								}
							}
						}
					}
				}
			},
			"segmented-control": {
				color: {
					fill: "rgba(0, 0, 0, 0.040)"
				},
				segment: {
					shadow: {
						highlight: {
							accent: {
								position: {
									x: 0,
									y: 1
								},
								blur: {
									"default": 2
								},
								spread: {
									"default": 0
								},
								color: {
									accent: "rgba(0, 0, 0, 0.100)"
								}
							}
						}
					},
					border: {
						width: {
							"default": 0
						}
					}
				}
			},
			tabs: {
				pipe: {
					size: {
						height: {
							"default": 3
						}
					}
				}
			},
			thumbnail: {
				spacing: {
					padding: {
						large: 3,
						small: 1
					}
				}
			},
			"view-indicator": {
				color: {
					fill: "rgba(255, 255, 255, 0.000)",
					border: "rgba(255, 255, 255, 0.000)"
				},
				view: {
					color: {
						selected: {
							fill: "rgba(52, 131, 250, 1.000)"
						}
					},
					size: {
						height: {
							selected: {
								small: 6,
								large: 6
							},
							unselected: {
								small: 2,
								medium: 4,
								large: 6
							}
						},
						width: {
							selected: {
								small: 6,
								large: 6
							},
							unselected: {
								small: 2,
								medium: 4,
								large: 6
							}
						}
					}
				}
			},
			tag: {
				color: {
					fill: {
						positive: {
							hover: "rgba(0, 166, 80, 0.100)",
							active: "rgba(0, 166, 80, 0.200)"
						},
						caution: {
							hover: "rgba(255, 119, 51, 0.100)",
							active: "rgba(255, 119, 51, 0.200)"
						},
						negative: {
							hover: "rgba(242, 61, 79, 0.100)",
							active: "rgba(242, 61, 79, 0.200)"
						}
					}
				}
			}
		},
		glass: {
			blur: {
				"default": 8
			},
			progressive: {
				blur: {
					start: 0,
					end: 4
				}
			}
		},
		shadow: {
			low: {
				top: {
					position: {
						x: 0,
						y: -2
					},
					blur: {
						"default": 2
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(0, 0, 0, 0.100)"
					}
				},
				center: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 2
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(0, 0, 0, 0.100)"
					}
				},
				bottom: {
					position: {
						x: 0,
						y: 1
					},
					blur: {
						"default": 2
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(0, 0, 0, 0.100)"
					}
				}
			},
			high: {
				top: {
					position: {
						x: 0,
						y: -6
					},
					blur: {
						"default": 16
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(0, 0, 0, 0.100)"
					}
				},
				center: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 16
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(0, 0, 0, 0.100)"
					}
				},
				bottom: {
					position: {
						x: 0,
						y: 6
					},
					blur: {
						"default": 16
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(0, 0, 0, 0.100)"
					}
				}
			},
			highlight: {
				accent: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 0
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(255, 255, 255, 0.000)"
					}
				},
				positive: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 0
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(255, 255, 255, 0.000)"
					}
				},
				caution: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 0
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(255, 255, 255, 0.000)"
					}
				},
				negative: {
					position: {
						x: 0,
						y: 0
					},
					blur: {
						"default": 0
					},
					spread: {
						"default": 0
					},
					color: {
						"default": "rgba(255, 255, 255, 0.000)"
					}
				}
			}
		},
		size: {
			height: {
				xnano: 4,
				nano: 6,
				xmicro: 8,
				micro: 12,
				xtiny: 16,
				tiny: 20,
				xsmall: 24,
				small: 32,
				medium: 40,
				large: 48,
				xlarge: 56,
				huge: 64,
				xhuge: 72,
				mega: 80,
				xmega: 88,
				actions: {
					xsmall: 24,
					small: 24,
					medium: 32,
					large: 48
				},
				inputs: {
					small: 32,
					medium: 40,
					large: 48
				},
				"selection-controls": {
					tiny: 20
				},
				icons: {
					micro: 12,
					xtiny: 16,
					tiny: 20,
					xsmall: 24,
					small: 32,
					medium: 40,
					large: 48
				},
				media: {
					xsmall: 88,
					small: 120,
					medium: 128,
					large: 176,
					xlarge: 240,
					huge: 320
				}
			},
			width: {
				xnano: 4,
				nano: 6,
				xmicro: 8,
				micro: 12,
				xtiny: 16,
				tiny: 20,
				xsmall: 24,
				small: 32,
				medium: 40,
				large: 48,
				xlarge: 56,
				huge: 64,
				xhuge: 72,
				mega: 80,
				actions: {
					xsmall: 24,
					small: 24,
					medium: 32,
					large: 48
				},
				inputs: {
					small: 32,
					medium: 40,
					large: 48
				},
				"selection-controls": {
					tiny: 20
				},
				icons: {
					micro: 12,
					xtiny: 16,
					tiny: 20,
					xsmall: 24,
					small: 32,
					medium: 40,
					large: 48
				},
				media: {
					xsmall: 88,
					small: 120,
					medium: 128,
					large: 176,
					xlarge: 240,
					huge: 320
				},
				xmega: 88
			}
		},
		spacing: {
			padding: {
				none: 0,
				pico: 2,
				xnano: 4,
				nano: 6,
				xmicro: 8,
				micro: 12,
				xtiny: 16,
				tiny: 20,
				xsmall: 24,
				small: 32,
				medium: 40,
				large: 48,
				xlarge: 56,
				huge: 64,
				xhuge: 72,
				mega: 80,
				xmega: 88
			},
			gap: {
				none: 0,
				pico: 2,
				xnano: 4,
				nano: 6,
				xmicro: 8,
				micro: 12,
				xtiny: 16,
				tiny: 20,
				xsmall: 24,
				small: 32,
				medium: 40,
				large: 48,
				xlarge: 56,
				huge: 64,
				xhuge: 72,
				mega: 80,
				xmega: 88,
				forms: {
					horizontal: {
						near: 16,
						mid: 24,
						wide: 32,
						related: 8
					},
					vertical: {
						near: 16,
						mid: 24,
						wide: 32
					}
				},
				negative: {
					pico: -2,
					xnano: -4,
					nano: -6,
					xmicro: -8,
					micro: -12,
					xtiny: -16
				}
			}
		},
		text: {
			body: {
				family: {
					"default": "Proxima Nova"
				},
				size: {
					xsmall: 12,
					small: 14,
					medium: 16,
					large: 18
				},
				"line-height": {
					xsmall: 15,
					small: 18,
					medium: 20,
					large: 22
				},
				weight: {
					"default": 400,
					emphasis: 600
				},
				"letter-spacing": {
					"default": 0
				},
				"paragraph-spacing": {
					xsmall: 9,
					small: 10,
					medium: 12,
					large: 14
				}
			},
			heading: {
				family: {
					"default": "Proxima Nova"
				},
				size: {
					tiny: 12,
					xsmall: 14,
					small: 16,
					medium: 18,
					large: 20,
					xlarge: 24,
					huge: 28,
					xhuge: 32,
					mega: 40
				},
				"line-height": {
					tiny: 15,
					xsmall: 18,
					small: 20,
					medium: 22,
					large: 25,
					xlarge: 30,
					huge: 35,
					xhuge: 40,
					mega: 48
				},
				weight: {
					lean: 400,
					narrow: 600,
					"default": 600
				},
				"letter-spacing": {
					"default": 0
				}
			},
			display: {
				family: {
					"default": "Proxima Nova"
				},
				size: {
					xsmall: 48,
					small: 56,
					medium: 64
				},
				"line-height": {
					xsmall: 56,
					small: 66,
					medium: 72
				},
				weight: {
					lean: 400,
					narrow: 600,
					"default": 600
				},
				"letter-spacing": {
					"default": 0
				}
			}
		},
		charts: {
			color: {
				axis: "rgba(191, 191, 191, 1.000)",
				gridline: "rgba(140, 140, 140, 1.000)",
				"reference-line": "rgba(26, 26, 26, 1.000)",
				data: {
					reference: "rgba(217, 217, 217, 1.000)",
					loud: "rgba(52, 131, 250, 1.000)",
					quiet: "rgba(65, 137, 230, 0.150)"
				},
				interactive: {
					data: {
						loud: {
							idle: "rgba(52, 131, 250, 1.000)",
							hover: "rgba(41, 104, 200, 1.000)",
							active: "rgba(31, 78, 150, 1.000)"
						},
						quiet: {
							idle: "rgba(65, 137, 230, 0.150)",
							hover: "rgba(65, 137, 230, 0.200)",
							active: "rgba(65, 137, 230, 0.300)"
						}
					}
				},
				selected: {
					data: {
						idle: "rgba(52, 131, 250, 1.000)",
						hover: "rgba(41, 104, 200, 1.000)",
						active: "rgba(31, 78, 150, 1.000)"
					}
				}
			}
		},
		table: {
			cell: {
				color: {
					surface: {
						disabled: "rgba(245, 245, 245, 1.000)"
					}
				}
			}
		}
	}
};
var andesLegacyFallback = {
	id: id,
	name: name,
	version: version,
	value: value
};

const fallbackThemes = {
    andes_x: andesXFallback,
    andes_legacy: andesLegacyFallback,
};
/**
 * Gets the default BU theme name based on platform.
 *
 * @param platformId - Optional platform identifier (e.g. 'MP', 'ML')
 * @returns Default BU theme name for the given platform, falls back to ML
 */
const getBUDefaultThemeName = (platformId) => PLATFORM_FALLBACK_THEME_NAME[platformId] ?? PLATFORM_FALLBACK_THEME_NAME.ML;
/**
 * Type guard for ThemeMode. Useful for narrowing values coming from
 * untrusted sources (cookies, query params, etc.).
 */
const isThemeMode = (value) => value === 'light' || value === 'dark';
/**
 * Determines the theme mode from options.
 * Priority: explicit mode option > default 'light'
 *
 * @param options - Options object
 * @returns Theme mode
 */
const getThemeMode = (options = {}) => {
    if (isThemeMode(options.mode)) {
        return options.mode;
    }
    return 'light';
};
/**
 * Normalizes version to ensure it defaults to 'latest' if not provided.
 * If version is provided but not valid semver format, defaults to 'latest'.
 *
 * @param version - Theme version
 * @returns Normalized version ('latest' if not provided or invalid semver)
 */
const normalizeVersion = (version) => {
    if (!version || version === 'latest') {
        return 'latest';
    }
    // Validate semver format (X.Y.Z where X, Y, Z are numbers)
    const semverRegex = /^\d+\.\d+\.\d+$/;
    if (!semverRegex.test(version)) {
        return 'latest';
    }
    return version;
};
/**
 * Checks if the theme supports dark mode.
 *
 * @param theme - Theme object
 * @returns True if the theme supports dark mode, false otherwise
 */
const themeSupportsDarkMode = (theme) => !!(theme.dark && typeof theme.dark === 'object');
/**
 * Gets the validated theme mode, falling back to 'light' if dark mode is requested
 * but not supported by the theme.
 *
 * @param theme - Theme object
 * @param mode - Theme mode
 * @returns Validated theme mode
 */
const getValidatedMode = (theme, mode) => {
    if (mode === 'dark' && !themeSupportsDarkMode(theme)) {
        return 'light';
    }
    return mode;
};

/**
 * Helper function to format theme response with consistent structure.
 * Transforms a theme object into the standardized response format with CSS variables.
 *
 * @param theme - Theme object containing design tokens to be converted to CSS
 * @param name - Theme name
 * @param version - Theme version
 * @param options - Options for theme formatting
 * @returns Formatted theme response with styles object, name, version, mode, and fonts
 */
const formatThemeResponse = (theme, name, version, options = {}) => {
    const requestedMode = getThemeMode(options);
    try {
        const fonts = extractFontsFromTheme(theme);
        return {
            styles: {
                id: GLOBAL_THEME_ID,
                children: toCss(theme, { mode: requestedMode }),
            },
            name,
            version,
            mode: requestedMode,
            fonts,
        };
    }
    catch (error) {
        logger.error('Failed to convert theme to CSS', {
            error: error.message,
        });
        return {
            styles: {
                id: GLOBAL_THEME_ID,
                children: '',
            },
            name,
            version,
            mode: requestedMode,
            fonts: [],
        };
    }
};
/**
 * Loads fallback theme based on theme name.
 * Priority: frontend-config > fallback JSON file.
 *
 * @param themeName - Theme name to determine fallback type
 * @param tags - Request tags for metrics
 * @param options - Options for theme formatting
 * @returns Formatted fallback theme
 */
const loadFallbackTheme = (themeName, tags = { platform: 'ML', webview: 'false' }, options = {}) => {
    try {
        // Get theme from frontend-config fallbacks
        const themeConfig = frontendConfig.config.globalTheme?.fallbacks?.[themeName];
        if (!themeConfig?.name || !themeConfig.version || !themeConfig.value) {
            throw new Error(`Theme configuration not found for ${themeName}`);
        }
        logger.info(`Theme loaded from ${themeConfig.name} configuration fallback`);
        increment(METRICS.THEME_FALLBACK_USED, {
            ...tags,
            type: 'config_fallback',
        });
        const themeValue = themeConfig.value;
        const requestedMode = getThemeMode(options);
        const validatedMode = getValidatedMode(themeValue, requestedMode);
        return formatThemeResponse(themeValue, themeConfig.name, themeConfig.version, {
            ...options,
            mode: validatedMode,
        });
    }
    catch (error) {
        logger.info('Configuration fallback theme loading failed', {
            error: error.message,
            themeName,
        });
        let fallbackTheme = fallbackThemes[themeName];
        let fallbackThemeName = themeName;
        if (!fallbackTheme) {
            const BU = getBUDefaultThemeName();
            fallbackTheme = fallbackThemes[BU];
            fallbackThemeName = BU;
            logger.info(`Fallback theme not found for ${themeName}, using ${fallbackThemeName} fallback`);
        }
        logger.info(`Theme loaded from ${fallbackTheme.name} fallback file`);
        increment(METRICS.THEME_FALLBACK_USED, {
            ...tags,
            type: 'fallback_theme_file',
        });
        const fallbackValue = fallbackTheme.value;
        const requestedMode = getThemeMode(options);
        const validatedMode = getValidatedMode(fallbackValue, requestedMode);
        return formatThemeResponse(fallbackValue, fallbackTheme.name, fallbackTheme.version, {
            ...options,
            mode: validatedMode,
        });
    }
};
const getErrorFallbackTheme = (useOptions, optionsName, parsedCookie, buTheme, mode) => {
    if (useOptions) {
        return {
            fallbackName: optionsName || buTheme,
            fallbackMode: mode,
        };
    }
    const cookieMode = parsedCookie?.mode;
    return {
        fallbackName: parsedCookie?.name || buTheme,
        fallbackMode: isThemeMode(cookieMode) ? cookieMode : mode,
    };
};
/**
 * Retrieves the global theme based on the provided options and request context.
 *
 * Handles theme resolution with the following priority:
 * 1. Theme name and version from options (if name is provided in options)
 * 2. Theme info from cookie (if present)
 * 3. Default BU theme from API
 * 4. Fallback theme from frontend-config (if API fails)
 * 5. Fallback theme from fallback-themes directory (if config not available)
 *
 * The function also manages theme cookies for caching and sets request metadata
 * for observability purposes.
 *
 * @param options - Options for theme retrieval and formatting
 * @returns Theme response with styles object (id and children), name, version, mode, and fonts
 */
const getGlobalTheme = async (options = {}) => {
    const { disableGlobalTheme = false } = options;
    const req = serverContext.getReq();
    const res = serverContext.getRes();
    if (disableGlobalTheme) {
        removeCookie(req, res);
        const buTheme = getBUDefaultThemeName();
        const themeData = fallbackThemes[buTheme];
        return {
            styles: {
                id: GLOBAL_THEME_ID,
                children: '',
            },
            fonts: extractFontsFromTheme(themeData.value),
        };
    }
    // Configure singleton logger for this request
    const logLevel = frontendConfig.config.globalTheme
        ?.logLevel || 'none';
    logger.configure(logLevel);
    const isWebview = req.device.webview;
    const tags = {
        platform: req.platform?.id ?? 'unknown',
        webview: isWebview ? 'true' : 'false',
    };
    const context = frontendRestclient.buildContext(req);
    const stopTimer = startTimer(METRICS.THEME_LOAD_DURATION, tags);
    const parsedCookie = getParsedCookieValue(req);
    // Multiple x-theme cookies in the jar indicate cookie accumulation across apps — a known
    // ecosystem health problem. Track it to measure recovery once contaminating apps are fixed.
    if (Array.isArray(req.cookies[THEME_COOKIE_KEY])) {
        increment(METRICS.THEME_COOKIE_MULTIPLE, tags);
    }
    const buTheme = getBUDefaultThemeName(req.platform?.id);
    const { name: optionsName, version: optionsVersion, strategy = 'overrideCookie', } = options;
    const mode = getThemeMode(options);
    const useOptions = !!(optionsName &&
        (strategy === 'overrideCookie' || !parsedCookie));
    try {
        // Return theme from options if name is provided
        if (useOptions) {
            const version = normalizeVersion(optionsVersion);
            let themeData;
            if (version === 'latest') {
                themeData = await getLatestThemeByName(optionsName, context);
                logger.info(`Theme loaded from API by name (from options) - name: ${themeData.name}, version: ${themeData.version}, mode: ${mode}`);
            }
            else {
                themeData = await getThemeById(`${optionsName}:${version}`, context);
                logger.info(`Theme loaded from API by ID (from options) - name: ${themeData.name}, version: ${themeData.version}, mode: ${mode}`);
            }
            const themeValue = themeData.value;
            const validatedMode = getValidatedMode(themeValue, mode);
            const response = formatThemeResponse(themeValue, themeData.name, themeData.version, {
                ...options,
                mode: validatedMode,
            });
            increment(METRICS.THEME_RESOLVED, {
                ...tags,
                source: 'options',
                theme_name: themeData.name,
            });
            return response;
        }
        // Return theme from cookie if present
        if (parsedCookie) {
            const { name: cookieThemeName, version: cookieThemeVersion, mode: cookieThemeMode, } = parsedCookie;
            const themeData = await getThemeById(`${cookieThemeName}:${cookieThemeVersion}`, context);
            logger.info(`Theme loaded from API by ID (from cookie) - name: ${cookieThemeName}, version: ${cookieThemeVersion}, mode: ${cookieThemeMode}`);
            const finalMode = isThemeMode(cookieThemeMode)
                ? cookieThemeMode
                : mode;
            const response = formatThemeResponse(themeData.value, cookieThemeName, cookieThemeVersion, {
                ...options,
                mode: finalMode,
            });
            increment(METRICS.THEME_RESOLVED, {
                ...tags,
                source: 'cookie',
                theme_name: cookieThemeName,
            });
            return response;
        }
        const themeData = fallbackThemes[buTheme];
        const themeValue = themeData.value;
        const validatedMode = getValidatedMode(themeValue, mode);
        logger.info(`Theme loaded with default ML theme - name: ${buTheme}, version: ${themeData.version}`);
        const response = formatThemeResponse(themeValue, themeData.name, themeData.version, {
            ...options,
            mode: validatedMode,
        });
        increment(METRICS.THEME_RESOLVED, {
            ...tags,
            source: 'default',
            theme_name: themeData.name,
        });
        return response;
    }
    catch (error) {
        // Return fallback theme if API fails
        logger.error('Failed to load theme from API', {
            error: error.message,
            name: optionsName,
            version: optionsVersion,
            isWebview,
            themeId: req.cookies[THEME_COOKIE_KEY],
        });
        const { fallbackName, fallbackMode } = getErrorFallbackTheme(useOptions, optionsName, parsedCookie, buTheme, mode);
        increment(METRICS.THEME_RESOLVED, {
            ...tags,
            source: 'fallback',
            theme_name: fallbackName,
        });
        return loadFallbackTheme(fallbackName, tags, {
            ...options,
            mode: fallbackMode,
        });
    }
    finally {
        stopTimer();
    }
};

exports.getGlobalTheme = getGlobalTheme;
exports.isThemeMode = isThemeMode;
