// @andes/listbox/build/cjs/index.production.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _extends = require('@babel/runtime/helpers/extends');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var React = require('react');
var listbox = require('@react-aria/listbox');
var list$1 = require('@react-stately/list');
var utils = require('@react-aria/utils');
var list = require('@andes/list');
var itemNodeGenerator = require('@andes/list/build/generators/itemNodeGenerator');
var includes = require('@andes/common/build/utils/includes');
var isMobile = require('@andes/common/build/utils/isMobile');
var classNames = require('classnames');
var warning = require('@andes/common/build/utils/warning');
var hasImage = require('@andes/list/build/utils/hasImage');
var hasSlot = require('@andes/list/build/utils/hasSlot');
var isLegacyLeftContent = require('@andes/list/build/utils/isLegacyLeftContent');
var keyboardCodes = require('@andes/common/build/constants/keyboardCodes');
var ListContext = require('@andes/list/build/context/ListContext');
var getAttributesByListType = require('@andes/list/build/utils/getAttributesByListType');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _objectWithoutProperties__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutProperties);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var itemNodeGenerator__default = /*#__PURE__*/_interopDefaultLegacy(itemNodeGenerator);
var includes__default = /*#__PURE__*/_interopDefaultLegacy(includes);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
const ListOption = () => null;
ListOption.displayName = 'ListOption';
ListOption.getCollectionNode = itemNodeGenerator__default["default"];
var ListOption$1 = ListOption;
const SUPPORTED_ITEM_TYPES = ['ListItem', 'DropdownItem'];
const SUPPORTED_GROUP_TYPES = ['ListGroup', 'DropdownGroup'];
const isSupportedType = function (child) {
  var _child$type, _child$type2, _child$type3, _child$type4;
  let types = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  if (!child) return false;
  return includes__default["default"](types, ((_child$type = child.type) === null || _child$type === void 0 ? void 0 : _child$type.displayName) || ((_child$type2 = child.type) === null || _child$type2 === void 0 ? void 0 : _child$type2.name)) || includes__default["default"](types, ((_child$type3 = child.type) === null || _child$type3 === void 0 || (_child$type3 = _child$type3.wraps) === null || _child$type3 === void 0 ? void 0 : _child$type3.displayName) || ((_child$type4 = child.type) === null || _child$type4 === void 0 ? void 0 : _child$type4.name));
};
var isSupportedType$1 = isSupportedType;
const isSupportedGroupType = child => isSupportedType$1(child, SUPPORTED_GROUP_TYPES);
var isSupportedGroupType$1 = isSupportedGroupType;
const isSupportedItemType = child => isSupportedType$1(child, SUPPORTED_ITEM_TYPES);
var isSupportedItemType$1 = isSupportedItemType;
const useChildrenAdapter = children => React.Children.map(children, child => {
  if (isSupportedItemType$1(child)) return /*#__PURE__*/React__default["default"].createElement(ListOption$1, child.props);
  if (isSupportedGroupType$1(child)) {
    const groupChildren = React.Children.map(child.props.children, item => {
      if (isSupportedItemType$1(item)) return /*#__PURE__*/React__default["default"].createElement(ListOption$1, item.props);
      return item;
    });
    return /*#__PURE__*/React__default["default"].createElement(list.ListGroup, child.props, groupChildren);
  }
  return child;
});
var useChildrenAdapter$1 = useChildrenAdapter;
const WARNING_LIST_ITEM_CHILDREN = 'Children can\'t be used in conjunction with props title, description, leftContent and rightContent, in future versions children will have precedence over these properties';
const WARNING_LIST_ITEM_LEFT_CONTENT = 'The current object format \'{ title: string, src: string, alt: string, src2x: string }\' will be deprecated from the \'leftContent\' options. Please use a \'node\' instead.';
const _excluded$2 = ["item", "state"],
  _excluded2$1 = ["onClick", "onKeyDown"],
  _excluded3 = ["onSelect"];
const ListBoxItem = _ref => {
  let item = _ref.item,
    state = _ref.state,
    otherProps = _objectWithoutProperties__default["default"](_ref, _excluded$2);
  const key = item.key,
    listItemProps = item.props;
  const _useListContext = ListContext.useListContext(),
    listType = _useListContext.listType;
  const _useMemo = React.useMemo(() => getAttributesByListType.getAttributesByListType(listType)(), []),
    selection = _useMemo.selection;
  const itemSelectionRef = React.useRef(null);
  const _useOption = listbox.useOption({
      key
    }, state, itemSelectionRef),
    optionProps = _useOption.optionProps,
    selected = _useOption.isSelected;
  const inheritedProps = _objectSpread__default["default"](_objectSpread__default["default"]({}, listItemProps), {}, {
    key,
    selected
  });
  const itemClassName = classNames__default["default"](classNames__default["default"], {
    'andes-list__item-with-selection-component': !!otherProps.selectionComponent
  });
  const onClick = optionProps.onClick,
    onKeyDown = optionProps.onKeyDown,
    otherOptionProps = _objectWithoutProperties__default["default"](optionProps, _excluded2$1);
  const onSelect = otherProps.onSelect,
    otherOtherProps = _objectWithoutProperties__default["default"](otherProps, _excluded3);
  const onClickHandler = e => {
    const disabled = inheritedProps.disabled,
      value = inheritedProps.value,
      inheritedOnSelect = inheritedProps.onSelect;
    const inheritedOnClick = otherProps.onClick;
    if (!disabled) {
      if (typeof inheritedOnClick === 'function') inheritedOnClick(e, value);
      if (typeof onClick === 'function') onClick(e);
      if (typeof inheritedOnSelect === 'function') inheritedOnSelect(e, {
        value
      });
      if (typeof onSelect === 'function') onSelect(e, {
        value
      });
    }
  };
  const handleOnKeyDown = event => {
    if (typeof onKeyDown === 'function') onKeyDown(event);
    if (event.code === keyboardCodes.ENTER || event.code === keyboardCodes.SPACE) {
      onClickHandler(event);
    }
  };
  const getCheckbox = () => {
    const Check = otherProps.selectionComponent;
    return /*#__PURE__*/React.isValidElement(Check) ? /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({}, selection, {
      className: "andes-list__item-selection-control andes-listbox__item-selection-control"
    }), /*#__PURE__*/React.cloneElement(Check, {
      checked: inheritedProps.selected,
      disabled: inheritedProps.disabled
    })) : null;
  };
  const itemProps = _objectSpread__default["default"]({}, utils.mergeProps(inheritedProps, otherOtherProps));
  const children = itemProps.children,
    description = itemProps.description,
    disabled = itemProps.disabled,
    id = itemProps.id,
    leftContent = itemProps.leftContent,
    maxLines = itemProps.maxLines,
    paddingSize = itemProps.paddingSize,
    rightContent = itemProps.rightContent,
    rightContentPosition = itemProps.rightContentPosition,
    size = itemProps.size,
    title = itemProps.title;
  const hasDescription = !!description;
  const hasLeftContentImage = hasImage.hasImage(leftContent);
  const isSlotted = hasSlot.hasSlot(children, {
    title,
    description,
    rightContent
  }, {
    legacyBehavior: true
  });
  const itemCustomEventsHandlers = {
    onClick: onClickHandler,
    onKeyDown: handleOnKeyDown
  };
  warning__default["default"](!(children && (title || description || rightContent)), WARNING_LIST_ITEM_CHILDREN);
  const legacyLeftContent = isLegacyLeftContent.isLegacyLeftContent(leftContent);
  warning__default["default"](!legacyLeftContent, WARNING_LIST_ITEM_LEFT_CONTENT);
  return /*#__PURE__*/React__default["default"].createElement(list.ListBaseItem, _extends__default["default"]({
    className: itemClassName,
    id: id,
    ref: itemSelectionRef,
    disabled: disabled,
    selected: selected,
    hasDescription: hasDescription,
    hasLeftContentImage: hasLeftContentImage,
    paddingSize: paddingSize,
    size: size,
    slotted: isSlotted
  }, otherOptionProps, itemCustomEventsHandlers), getCheckbox(), /*#__PURE__*/React__default["default"].createElement(list.ListItemContent, _extends__default["default"]({}, itemProps, {
    key: key,
    title: title,
    description: description,
    leftContent: leftContent,
    rightContent: rightContent,
    rightContentPosition: rightContentPosition,
    maxLines: maxLines,
    size: size,
    id: id || (otherOptionProps === null || otherOptionProps === void 0 ? void 0 : otherOptionProps.id),
    legacyBehavior: true
  }), children));
};
var ListBoxItem$1 = ListBoxItem;
const _excluded$1 = ["group", "state", "listId"];
const ListBoxGroup = _ref => {
  let group = _ref.group,
    state = _ref.state,
    listId = _ref.listId,
    otherProps = _objectWithoutProperties__default["default"](_ref, _excluded$1);
  const heading = group.rendered,
    ariaLabel = group['aria-label'],
    listGroupProps = group.props,
    index = group.index;
  const _useListBoxSection = listbox.useListBoxSection({
      heading,
      'aria-label': ariaLabel
    }),
    itemProps = _useListBoxSection.itemProps,
    headingProps = _useListBoxSection.headingProps,
    groupProps = _useListBoxSection.groupProps;
  headingProps.id = `${listId}__group-heading-${index}`;
  groupProps['aria-labelledby'] = headingProps.id;
  return /*#__PURE__*/React__default["default"].createElement(list.ListGroup, _extends__default["default"]({
    label: "",
    itemProps: itemProps,
    headingProps: headingProps,
    groupProps: groupProps
  }, listGroupProps), [...group.childNodes].map(groupNode => /*#__PURE__*/React__default["default"].createElement(ListBoxItem$1, _extends__default["default"]({
    key: groupNode.key,
    item: groupNode,
    state: state
  }, otherProps))));
};
var ListBoxGroup$1 = ListBoxGroup;
const findItemInChildren = (children, itemKey) => {
  let findItem;
  children.forEach(child => {
    var _child$props, _child$props2;
    if (isSupportedType$1(child, ['ListOption', 'FloatingMenuItem']) && ((child === null || child === void 0 || (_child$props = child.props) === null || _child$props === void 0 ? void 0 : _child$props.value) === itemKey || (child === null || child === void 0 ? void 0 : child.key) === itemKey)) {
      findItem = child;
      return;
    }
    if (isSupportedType$1(child, ['ListGroup']) && child !== null && child !== void 0 && (_child$props2 = child.props) !== null && _child$props2 !== void 0 && _child$props2.children) {
      const foundItem = findItemInChildren(child.props.children, itemKey);
      if (foundItem) {
        findItem = foundItem;
      }
    }
  });
  return findItem || null;
};
const createMouseEventForElement = element => {
  const target = element ? Object.assign(element, {
    value: undefined
  }) : {};
  const clickEvent = new MouseEvent('click', {
    bubbles: true,
    cancelable: true,
    view: window
  });
  Object.defineProperty(clickEvent, 'target', {
    get: () => target
  });
  return clickEvent;
};
const _excluded = ["ref"],
  _excluded2 = ["children", "id", "selectionComponent", "selectionMode", "isRadioList", "defaultSelectedKeys", "selectedKeys", "onClick", "onSelect", "shouldUseVirtualFocus", "srLabel"];
const ListBox = /*#__PURE__*/function () {
  const ListBox = _ref => {
    let ref = _ref.ref,
      props = _objectWithoutProperties__default["default"](_ref, _excluded);
    const children = props.children,
      id = props.id,
      selectionComponent = props.selectionComponent,
      _props$selectionMode = props.selectionMode,
      selectionMode = _props$selectionMode === void 0 ? 'single' : _props$selectionMode,
      _props$isRadioList = props.isRadioList,
      isRadioList = _props$isRadioList === void 0 ? false : _props$isRadioList,
      defaultSelectedKeys = props.defaultSelectedKeys,
      selectedKeys = props.selectedKeys,
      onClick = props.onClick,
      onSelect = props.onSelect,
      shouldUseVirtualFocus = props.shouldUseVirtualFocus,
      srLabel = props.srLabel,
      other = _objectWithoutProperties__default["default"](props, _excluded2);
    const listSelectionRef = React.useRef(null);
    const selectedManager = React.useRef({});
    const prevState = React.useRef(selectedKeys || defaultSelectedKeys || []);
    const lastPointerType = React.useRef('mouse');
    const handlePointerDown = React.useCallback(e => {
      lastPointerType.current = e.pointerType || 'mouse';
    }, []);
    React.useEffect(() => {
      const resolvedRef = typeof ref === 'function' ? listSelectionRef : ref || listSelectionRef;
      const el = resolvedRef.current;
      if (!el) return undefined;
      el.addEventListener('pointerdown', handlePointerDown);
      return () => el.removeEventListener('pointerdown', handlePointerDown);
    }, [handlePointerDown, ref]);
    const callOnSelectCallback = () => {
      const _selectedManager$curr = selectedManager.current,
        event = _selectedManager$curr.event,
        values = _selectedManager$curr.values,
        itemDetails = _selectedManager$curr.itemDetails;
      if (event && values && itemDetails) {
        if (typeof onSelect === 'function') onSelect(event, values, itemDetails);
        selectedManager.current = {};
      }
    };
    const onItemSelect = (event, itemDetails) => {
      selectedManager.current.event = event;
      selectedManager.current.itemDetails = itemDetails;
      callOnSelectCallback();
    };
    const getIdsAndDisabledKeys = () => {
      const disabledKeys = [];
      const listItemsIds = [];
      React.Children.forEach(children, child => {
        if (isSupportedGroupType$1(child)) {
          React.Children.forEach(child.props.children, item => {
            var _item$props, _item$props2;
            listItemsIds.push(`${id}-option-${item.props.value}`);
            if ((_item$props = item.props) !== null && _item$props !== void 0 && _item$props.disabled || (_item$props2 = item.props) !== null && _item$props2 !== void 0 && _item$props2.disableSelectionComponent) disabledKeys.push(item.props.value);
          });
        }
        if (isSupportedItemType$1(child)) {
          var _child$props, _child$props2;
          if ((_child$props = child.props) !== null && _child$props !== void 0 && _child$props.disabled || (_child$props2 = child.props) !== null && _child$props2 !== void 0 && _child$props2.disableSelectionComponent) disabledKeys.push(child.props.value);
        }
      });
      return {
        disabledKeys,
        listItemsIds: listItemsIds.join(' ')
      };
    };
    const _useMemo = React.useMemo(() => getIdsAndDisabledKeys(), [children]),
      disabledKeys = _useMemo.disabledKeys,
      listItemsIds = _useMemo.listItemsIds;
    const adaptedChildren = useChildrenAdapter$1(children);
    const handleSelectionChangeForTouchDevices = values => {
      var _ref2, _ref2$querySelector, _itemObject$props, _itemObject$props2;
      const prevKeys = prevState.current;
      const newKeys = values ? [...values] : [];
      let itemKey = '';
      if (selectionMode === 'single') {
        if (newKeys.length > 0) {
          prevState.current = newKeys;
          itemKey = newKeys[0];
        } else {
          var _prevKeys = _slicedToArray__default["default"](prevKeys, 1);
          itemKey = _prevKeys[0];
        }
      } else {
        const addedKeys = newKeys.filter(key => !prevKeys.includes(key));
        const removeKeys = prevKeys.filter(key => !newKeys.includes(key));
        itemKey = addedKeys.length > 0 ? addedKeys[0] : removeKeys[0];
        prevState.current = newKeys;
      }
      const itemElement = (_ref2 = ref || listSelectionRef) === null || _ref2 === void 0 || (_ref2 = _ref2.current) === null || _ref2 === void 0 || (_ref2$querySelector = _ref2.querySelector) === null || _ref2$querySelector === void 0 ? void 0 : _ref2$querySelector.call(_ref2, `li[data-key="${itemKey}"]`);
      if (!itemElement) return;
      const itemObject = findItemInChildren(adaptedChildren, itemKey);
      const itemValue = itemObject === null || itemObject === void 0 || (_itemObject$props = itemObject.props) === null || _itemObject$props === void 0 ? void 0 : _itemObject$props.value;
      const clickEvent = createMouseEventForElement(itemElement);
      if (typeof onClick === 'function') {
        onClick(clickEvent, itemValue);
      }
      if (typeof onSelect === 'function') {
        onSelect(clickEvent, newKeys, {
          value: itemValue
        });
      }
      if (typeof (itemObject === null || itemObject === void 0 || (_itemObject$props2 = itemObject.props) === null || _itemObject$props2 === void 0 ? void 0 : _itemObject$props2.onSelect) === 'function') {
        itemObject.props.onSelect(clickEvent, {
          value: itemValue
        });
      }
    };
    const onSelectionChanged = values => {
      if (isMobile.hasTouchSupport() && lastPointerType.current === 'touch') {
        handleSelectionChangeForTouchDevices(values);
        return;
      }
      selectedManager.current.values = Array.from(values || []);
      callOnSelectCallback();
    };
    const listStateHook = isRadioList ? list$1.useSingleSelectListState : list$1.useListState;
    const state = listStateHook(_objectSpread__default["default"](_objectSpread__default["default"]({
      disabledKeys,
      selectionMode
    }, props), {}, {
      onSelectionChange: onSelectionChanged,
      children: adaptedChildren
    }));
    const _useListBox = listbox.useListBox(_objectSpread__default["default"](_objectSpread__default["default"]({
        'aria-label': srLabel,
        disallowEmptySelection: true,
        disabledKeys,
        selectionMode
      }, props), {}, {
        shouldUseVirtualFocus
      }), state, ref || listSelectionRef),
      listBoxProps = _useListBox.listBoxProps;
    return /*#__PURE__*/React__default["default"].createElement(list.List, _extends__default["default"]({
      "aria-owns": listItemsIds.length ? listItemsIds : undefined,
      srLabel: srLabel,
      component: "ul",
      listType: "listbox"
    }, utils.mergeProps(other, listBoxProps), {
      ref: ref || listSelectionRef,
      "aria-label": undefined
    }), [...state.collection].map(listSelectNode => {
      const commonProps = _objectSpread__default["default"]({
        key: listSelectNode.key,
        state,
        selectionComponent
      }, isMobile.hasTouchSupport() && lastPointerType.current === 'touch' ? {} : {
        onClick,
        onSelect: (event, itemDetails) => {
          if (typeof onItemSelect === 'function') onItemSelect(event, itemDetails);
          if (typeof listSelectNode.onSelect === 'function') listSelectNode.onSelect(event, itemDetails);
        }
      });
      return listSelectNode.type === 'item' ? /*#__PURE__*/React__default["default"].createElement(ListBoxItem$1, _extends__default["default"]({
        item: listSelectNode
      }, commonProps, {
        key: listSelectNode.key
      })) : /*#__PURE__*/React__default["default"].createElement(ListBoxGroup$1, _extends__default["default"]({
        group: listSelectNode
      }, commonProps, {
        listId: id,
        key: listSelectNode.key
      }));
    }));
  };
  ListBox.displayName = 'ListBox';
  return ListBox;
}();
var ListBox$1 = ListBox;
exports.ListBox = ListBox$1;
