// node-melitk-otel/lib/events/processor/attribute/attributeProcessor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttributePropagatingSpanProcessor = void 0;
const api_1 = require("@opentelemetry/api");
const types_1 = require("./types");
/**
 * Represents the AttributePropagatingSpanProcessor class which implements the SpanProcessor interface.
 * This class provides the functionality to propagate attributes from the parent span to the child span.
 * @class
 * @implements {SpanProcessor}
 */
class AttributePropagatingSpanProcessor {
    activator;
    /**
     * Creates a new instance of the AttributePropagatingSpanProcessor class.
     * @param {Activator} activator The activator instance.
     */
    constructor(activator) {
        this.activator = activator;
    }
    /**
     * Method {@link onStart} propagates the `x_request_id` attribute from the parent span to the child span.
     * @param {Span} span The span instance.
     * @param {Context} parentContext The parent context.
     * @returns {void}
     */
    onStart(span, parentContext) {
        if (!this.activator.isActive()) {
            return;
        }
        const parentSpan = api_1.trace.getSpan(parentContext);
        if (parentSpan !== undefined &&
            (0, api_1.isSpanContextValid)(parentSpan.spanContext()) &&
            !parentSpan.isRecording()) {
            const { attributes } = parentSpan;
            const requestId = this.findAttr(types_1.requestIdKey, attributes);
            if (requestId !== undefined) {
                span.setAttribute(types_1.requestIdKey, requestId);
            }
        }
    }
    /**
     * Method {@link findAttr} finds the attribute with the given key in the attributes object.
     * @param {string} key The key of the attribute to find.
     * @param {Attributes} attributes The attributes object to search in.
     * @returns {AttributeValue | undefined} The attribute value if found, otherwise undefined.
     */
    findAttr(key, attributes) {
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
        if (attributes === undefined) {
            return undefined;
        }
        const attr = attributes[key];
        if (attr) {
            return attr;
        }
        return undefined;
    }
    /** Method {@link onEnd} does nothing. */
    onEnd() { }
    /** Method {@link shutdown} does nothing. */
    shutdown() { }
    /** Method {@link forceFlush} does nothing. */
    forceFlush() { }
}
exports.AttributePropagatingSpanProcessor = AttributePropagatingSpanProcessor;
