// @andes/common/build/esm/index.development.js
import React from 'react';
import PropTypes from 'prop-types';
import { Image } from '@andes/image';
import { CloseListener } from '@andes/close-listener';
import { Widther } from '@andes/widther';
import { VisuallyHidden, VisuallyHiddenFocusable } from '@andes/visually-hidden';
export { canUseDOM } from 'exenv';
const WARNING_PORTAL = 'The <Portal> component will be removed from this package in the next major release. Please use the "@andes/portal" package instead.';
const WARNING_VISUALLY_HIDDEN = 'The <VisuallyHidden> component will be deprecated from this package in the next major release. Please use the "@andes/visually-hidden" package instead.';
const WARNING_VISUALLY_HIDDEN_FOCUSABLE = 'The <VisuallyHiddenFocusable> component will be deprecated from this package in the next major release. Please use the "@andes/visually-hidden" package instead.';
const WARNING_CLOSE_LISTENER = 'The <CloseListener> component will be deprecated from this package in the next major release. Please use the "@andes/close-listener" package instead.';
const WARNING_WIDTHER = 'The <Widther> component will be deprecated from this package in the next major release. Please use the "@andes/widther" package instead.';
const WARNING_IMAGE = 'The <Image> component will be deprecate from this package in the next major release. Please use the "@andes/image" package instead.';

/* eslint-disable quotes */
/* eslint-disable no-console */

/**
 * @param {boolean} condition
 * @param {string} message
 * @param {string} [type]
 */
const print = function (condition, message) {
  let type = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'error';
  if (condition) {
    return;
  }
  if (typeof console !== 'undefined') {
    let logFunction = /** @type {function | undefined} */console[(/** @type {keyof typeof console} */type)];
    if (typeof logFunction !== 'function') logFunction = console.error;
    logFunction(message);
  }
  try {
    throw new Error(message);
    // eslint-disable-next-line no-empty
  } catch (x) {}
};
const warning = process.env.NODE_ENV !== "'production'" && process.env.NODE_ENV !== "production" ? print : () => {};
var warning$1 = warning;

/** @type {typeof import('../../..').Image} */
const ImageDeprecate = props => {
  warning$1(false, WARNING_IMAGE);
  return /*#__PURE__*/React.createElement(Image, props);
};
ImageDeprecate.propTypes = {
  title: PropTypes.string,
  src: PropTypes.string,
  src2x: PropTypes.string,
  width: PropTypes.string,
  height: PropTypes.string,
  className: PropTypes.string
};
var ImageDeprecate$1 = ImageDeprecate;

/** @type {typeof import('../../..').CloseListener} */
const CloseListenerDeprecate = props => {
  warning$1(false, WARNING_CLOSE_LISTENER);
  return /*#__PURE__*/React.createElement(CloseListener, props);
};
CloseListenerDeprecate.propTypes = {
  children: PropTypes.element.isRequired,
  onCloseOutside: PropTypes.func.isRequired,
  handleClick: PropTypes.bool,
  handleEsc: PropTypes.bool
};
var CloseListenerDeprecate$1 = CloseListenerDeprecate;
const PortalDeprecate = () => {
  warning$1(false, WARNING_PORTAL);
  return null;
};
var PortalDeprecate$1 = PortalDeprecate;

/** @type {typeof import('../../..').Widther<any>} */
const WidtherDeprecate = WrappedComponent => {
  warning$1(false, WARNING_WIDTHER);
  return Widther(WrappedComponent);
};
var WidtherDeprecate$1 = WidtherDeprecate;

/** @type {typeof import('../../..').VisuallyHidden} */
const VisuallyHiddenDeprecate = props => {
  warning$1(false, WARNING_VISUALLY_HIDDEN);
  return /*#__PURE__*/React.createElement(VisuallyHidden, props);
};
VisuallyHiddenDeprecate.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.arrayOf(PropTypes.node)]).isRequired,
  component: PropTypes.oneOf(['div', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'span'])
};
var VisuallyHiddenDeprecate$1 = VisuallyHiddenDeprecate;

/** @type {typeof import('../../..').VisuallyHiddenFocusable} */
const VisuallyHiddenFocusableDeprecate = props => {
  warning$1(false, WARNING_VISUALLY_HIDDEN_FOCUSABLE);
  return /*#__PURE__*/React.createElement(VisuallyHiddenFocusable, props);
};
VisuallyHiddenFocusableDeprecate.propTypes = {
  className: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.arrayOf(PropTypes.node)]),
  component: PropTypes.oneOf(['input', 'textarea', 'button', 'select', 'option', 'a', 'details', 'summary'])
};
var VisuallyHiddenFocusableDeprecate$1 = VisuallyHiddenFocusableDeprecate;
export { CloseListenerDeprecate$1 as CloseListener, ImageDeprecate$1 as Image, PortalDeprecate$1 as Portal, VisuallyHiddenDeprecate$1 as VisuallyHidden, VisuallyHiddenFocusableDeprecate$1 as VisuallyHiddenFocusable, WidtherDeprecate$1 as Widther };
