// @andes/common/build/esm/module.js
import React from 'react';
import { Image } from '@andes/image';
import { CloseListener } from '@andes/close-listener';
import { Widther } from '@andes/widther';
import { VisuallyHidden, VisuallyHiddenFocusable } from '@andes/visually-hidden';
export { canUseDOM } from 'exenv';
const WARNING_PORTAL = 'The <Portal> component will be removed from this package in the next major release. Please use the "@andes/portal" package instead.';
const WARNING_VISUALLY_HIDDEN = 'The <VisuallyHidden> component will be deprecated from this package in the next major release. Please use the "@andes/visually-hidden" package instead.';
const WARNING_VISUALLY_HIDDEN_FOCUSABLE = 'The <VisuallyHiddenFocusable> component will be deprecated from this package in the next major release. Please use the "@andes/visually-hidden" package instead.';
const WARNING_CLOSE_LISTENER = 'The <CloseListener> component will be deprecated from this package in the next major release. Please use the "@andes/close-listener" package instead.';
const WARNING_WIDTHER = 'The <Widther> component will be deprecated from this package in the next major release. Please use the "@andes/widther" package instead.';
const WARNING_IMAGE = 'The <Image> component will be deprecate from this package in the next major release. Please use the "@andes/image" package instead.';
const print = function (condition, message) {
  let type = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'error';
  if (condition) {
    return;
  }
  if (typeof console !== 'undefined') {
    let logFunction = console[type];
    if (typeof logFunction !== 'function') logFunction = console.error;
    logFunction(message);
  }
  try {
    throw new Error(message);
  } catch (x) {}
};
const warning = process.env.NODE_ENV !== "'production'" && process.env.NODE_ENV !== "production" ? print : () => {};
var warning$1 = warning;
const ImageDeprecate = props => {
  warning$1(false, WARNING_IMAGE);
  return /*#__PURE__*/React.createElement(Image, props);
};
var ImageDeprecate$1 = ImageDeprecate;
const CloseListenerDeprecate = props => {
  warning$1(false, WARNING_CLOSE_LISTENER);
  return /*#__PURE__*/React.createElement(CloseListener, props);
};
var CloseListenerDeprecate$1 = CloseListenerDeprecate;
const PortalDeprecate = () => {
  warning$1(false, WARNING_PORTAL);
  return null;
};
var PortalDeprecate$1 = PortalDeprecate;
const WidtherDeprecate = WrappedComponent => {
  warning$1(false, WARNING_WIDTHER);
  return Widther(WrappedComponent);
};
var WidtherDeprecate$1 = WidtherDeprecate;
const VisuallyHiddenDeprecate = props => {
  warning$1(false, WARNING_VISUALLY_HIDDEN);
  return /*#__PURE__*/React.createElement(VisuallyHidden, props);
};
var VisuallyHiddenDeprecate$1 = VisuallyHiddenDeprecate;
const VisuallyHiddenFocusableDeprecate = props => {
  warning$1(false, WARNING_VISUALLY_HIDDEN_FOCUSABLE);
  return /*#__PURE__*/React.createElement(VisuallyHiddenFocusable, props);
};
var VisuallyHiddenFocusableDeprecate$1 = VisuallyHiddenFocusableDeprecate;
export { CloseListenerDeprecate$1 as CloseListener, ImageDeprecate$1 as Image, PortalDeprecate$1 as Portal, VisuallyHiddenDeprecate$1 as VisuallyHidden, VisuallyHiddenFocusableDeprecate$1 as VisuallyHiddenFocusable, WidtherDeprecate$1 as Widther };
