// @andes/textfield/build/esm/index.development.js
import _extends from '@babel/runtime/helpers/extends';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { useState, useRef, useEffect } from 'react';
import classNames from 'classnames';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import PropTypes from 'prop-types';
import { useFocusRing, FocusRing } from '@react-aria/focus';
import { useAndesContext, useId } from '@andes/context';
import { BACKSPACE, TAB, KEY_V } from '@andes/common/build/constants/keyboardCodes';
import { autofocus } from '@andes/common/build/utils/autofocus';
import { isWebView } from '@andes/common/build/webkit/utils';
import { hasTouchSupport } from '@andes/common/build/utils/isMobile';
import includes from '@andes/common/build/utils/includes';
import useSrLabel from '@andes/common/build/hooks/useSrLabel';
import { VisuallyHidden } from '@andes/visually-hidden';
import FeedbackPositive12 from '@andes/icons/FeedbackPositive12';
import FeedbackPositive20 from '@andes/icons/FeedbackPositive20';
import { Helper } from '@andes/form';
import { useThemeContext } from '@andes/themes';
import { computeDataAndesState } from '@andes/common/build/utils/computeDataAndesState';
import { i18NPropTypeChecker } from '@andes/common/build/utils/propTypesValidators';
import VisibilityON20 from '@andes/icons/VisibilityON20';
import VisibilityOFF20 from '@andes/icons/VisibilityOFF20';
import transformToKebabCase from '@andes/common/build/utils/transformToKebabCase';
import { DropdownStandalone, DropdownItem } from '@andes/dropdown';
import { ProgressIndicatorCircular } from '@andes/progress-indicator-circular';
import autosize from 'autosize';
import Close12 from '@andes/icons/Close12';
import Close20 from '@andes/icons/Close20';
import warning from '@andes/common/build/utils/warning';
var enTranslations = {
  "andes-textfield\u0004{0} caracteres restantes": [null, "{0} remaining characters"],
  "andes-textfield\u0004Maximo {0} caracteres": [null, "Maximum of {0} characters"],
  "andes-textfield\u0004Campo de texto valido": [null, "Valid text field"],
  "andes-textfield\u0004Campo de texto procesando": [null, "Text field processing"],
  "andes-textfield\u0004Borrar": [null, "Remove"],
  "andes-textfield\u0004Dígito {0}": [null, "Digit {0}"],
  "andes-textfield\u0004Mostrar contraseña": [null, "Show Password"]
};
var zhCnTranslations = {
  "andes-textfield\u0004{0} caracteres restantes": [null, "剩余 {0} 个字符"],
  "andes-textfield\u0004Maximo {0} caracteres": [null, "最多可输入 {0} 个字符"],
  "andes-textfield\u0004Campo de texto valido": [null, "有效文本字段"],
  "andes-textfield\u0004Campo de texto procesando": [null, "文本字段处理中"],
  "andes-textfield\u0004Borrar": [null, "移除"],
  "andes-textfield\u0004Dígito {0}": [null, "数字 {0}"],
  "andes-textfield\u0004Mostrar contraseña": [null, "显示密码"]
};
var esArTranslations = {
  "andes-textfield\u0004{0} caracteres restantes": [null, "{0} caracteres restantes"],
  "andes-textfield\u0004Maximo {0} caracteres": [null, "Máximo {0} caracteres"],
  "andes-textfield\u0004Campo de texto valido": [null, "Campo de texto valido"],
  "andes-textfield\u0004Campo de texto procesando": [null, "Campo de texto procesando"],
  "andes-textfield\u0004Borrar": [null, "Borrar"],
  "andes-textfield\u0004Dígito {0}": [null, "Dígito {0}"],
  "andes-textfield\u0004Mostrar contraseña": [null, "Mostrar Contraseña"]
};
var ptBrTranslations = {
  "andes-textfield\u0004{0} caracteres restantes": [null, "{0} caracteres restantes"],
  "andes-textfield\u0004Maximo {0} caracteres": [null, "Máximo de {0} caracteres"],
  "andes-textfield\u0004Campo de texto valido": [null, "Campo de texto válido"],
  "andes-textfield\u0004Campo de texto procesando": [null, "Processamento de campo de texto"],
  "andes-textfield\u0004Borrar": [null, "Excluir"],
  "andes-textfield\u0004Dígito {0}": [null, "Dígito {0}"],
  "andes-textfield\u0004Mostrar contraseña": [null, "Mostrar Senha"]
};
const formControlNamespace = 'andes-form-control';
const splitButtonNamespace = `${formControlNamespace}__split-button`;
const namespace$3 = 'andes-textfield';

/**
 * @typedef {Object} ValidatedIconProps
 * @prop {string} notificationsLabel
 * @prop {string} [size]
 */

/** @type {React.FunctionComponent<ValidatedIconProps>} */
const ValidatedIcon = _ref => {
  let notificationsLabel = _ref.notificationsLabel,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'large' : _ref$size;
  const srLabelVisibility = useSrLabel();
  return /*#__PURE__*/React.createElement("span", {
    role: "status",
    "aria-live": "polite"
  }, /*#__PURE__*/React.createElement(VisuallyHidden, null, srLabelVisibility && notificationsLabel), /*#__PURE__*/React.createElement("span", {
    className: `${formControlNamespace}__validated-icon`
  }, size === 'large' ? /*#__PURE__*/React.createElement(FeedbackPositive20, {
    color: "currentColor"
  }) : /*#__PURE__*/React.createElement(FeedbackPositive12, {
    color: "currentColor"
  })));
};
ValidatedIcon.propTypes = {
  notificationsLabel: PropTypes.string.isRequired,
  size: PropTypes.string
};
var ValidatedIcon$1 = ValidatedIcon;
const WARNING_LABEL = `Some plain text contents are needed for accessibility reasons.
Please add a \`label\` or \`srLabel\` prop.`;
const WARNING_CHILDREN_NOT_SUPPORTED = `This child is not supported, use instead a text, 'tooltip', 'clear icon' 
or other component as recommended by the documentation.`;
const WARNING_DEPRECATED_MODIFIER_WARNING = `\`warning\` modifier will be deprecated in a future major version.
Please use \`caution\` instead.`;

/** @type {import('prop-types').Validator<string>} */
const checkLabelProp = (_ref, propName, componentName) => {
  let label = _ref.label,
    srLabel = _ref.srLabel;
  PropTypes.checkPropTypes({
    label: PropTypes.string
  }, {
    label
  }, propName, componentName);
  PropTypes.checkPropTypes({
    srLabel: PropTypes.string
  }, {
    label
  }, propName, componentName);
  return !label && !srLabel ? new Error(WARNING_LABEL) : null;
};

/** @type {import('prop-types').Validator<PropTypes.node>} */
const checkChildrenProp = (_ref2, propName, componentName) => {
  var _children$type;
  let children = _ref2.children;
  PropTypes.checkPropTypes({
    children: PropTypes.node
  }, {
    children
  }, propName, componentName);
  return includes(['Button', 'Checkbox'], children === null || children === void 0 || (_children$type = children.type) === null || _children$type === void 0 ? void 0 : _children$type.displayName) ? new Error(WARNING_CHILDREN_NOT_SUPPORTED) : null;
};
const numericNamespace = 'andes-form-control--numeric';
const namespace$2 = 'andes-form-control';
const translationsNamespace = 'andes-textfield';
const onlyNumberRegex = /^[0-9]*$/;
const DEFAULT_FOCUS_INDEX = 0;
const NOT_FOCUSABLE_INDEX = -1;
const numbersRegex = /[^0-9]/gi;
const specialCharacters = ['Backspace', 'Tab', 'Enter', 'Control', 'Shift', 'Escape', 'Meta', 'Alt', 'CapsLock', 'ArrowLeft', 'ArrowRight'];

/** @typedef {import('../..').NumericProps} NumericPublicProps */

/** @param {boolean | number} focus */
const getFocusIndexOrFalsy = focus => focus === false ? NOT_FOCUSABLE_INDEX : Number(focus === true ? DEFAULT_FOCUS_INDEX : focus);

/**
 * @param {number} digits
 * @returns {string[]}
 */
const getDefaultValue = digits => {
  if (!digits || typeof digits !== 'number') return [];
  return new Array(digits).fill('');
};

/** @param {string | string []} value */
const isEmpty = value => {
  const array = typeof value === 'string' ? value.split('') : value;
  return array.every(item => !item);
};

/**
 * @param {string | Array<string>} [value]
 * @param {number} [digits]
 * @returns {Array<string> | undefined}
 */
const parsedValues = (value, digits) => {
  if (typeof value === 'string') {
    let parsedValue = value.replace(numbersRegex, '').split('');
    if (digits && parsedValue.length < digits) {
      const diff = digits - parsedValue.length;
      const arrayEmpty = new Array(diff).fill('');
      parsedValue = parsedValue.concat(arrayEmpty);
    }
    return parsedValue;
  }
  if (Array.isArray(value)) return value;
};

/**
 * @typedef {Object} NumericInternalProps
 * @prop {Record<string, unknown>} [andesDataAttributes]
 */

/**
 * @typedef {NumericPublicProps & NumericInternalProps} NumericProps
 */

/** @type {React.FunctionComponent<NumericProps>} */
const Numeric = _ref => {
  let _ref$centered = _ref.centered,
    centered = _ref$centered === void 0 ? false : _ref$centered,
    className = _ref.className,
    defaultValue = _ref.defaultValue,
    _ref$digits = _ref.digits,
    digits = _ref$digits === void 0 ? 1 : _ref$digits,
    _ref$disabled = _ref.disabled,
    disabled = _ref$disabled === void 0 ? false : _ref$disabled,
    _ref$focus = _ref.focus,
    focus = _ref$focus === void 0 ? DEFAULT_FOCUS_INDEX : _ref$focus,
    helper = _ref.helper,
    id = _ref.id,
    label = _ref.label,
    modifier = _ref.modifier,
    onChange = _ref.onChange,
    onPaste = _ref.onPaste,
    srLabel = _ref.srLabel,
    _ref$strict = _ref.strict,
    strict = _ref$strict === void 0 ? false : _ref$strict,
    value = _ref.value,
    ref = _ref.ref,
    andesDataAttributes = _ref.andesDataAttributes;
  const _useState = useState(/** @type {Array<string>} */
    parsedValues(defaultValue, digits) || getDefaultValue(digits)),
    _useState2 = _slicedToArray(_useState, 2),
    inputValue = _useState2[0],
    setInputValue = _useState2[1];
  const _useState3 = useState(getFocusIndexOrFalsy(focus)),
    _useState4 = _slicedToArray(_useState3, 2),
    focusedItem = _useState4[0],
    setFocusedItem = _useState4[1];
  const _useRef = useRef(typeof value !== 'undefined'),
    isControlled = _useRef.current;
  const currentValue = isControlled ? parsedValues(value, digits) || [] : inputValue;

  /** @type {React.RefObject<Array<HTMLInputElement | null> | []>} */
  const inputsRef = useRef([]);
  const firstRender = useRef(true);
  const _useFocusRing = useFocusRing(),
    isFocusVisible = _useFocusRing.isFocusVisible,
    focusProps = _useFocusRing.focusProps;
  const _useAndesContext = useAndesContext('@andes/textfield', {
      en: enTranslations,
      'es-AR': esArTranslations,
      'pt-BR': ptBrTranslations,
      'zh-CN': zhCnTranslations
    }),
    i18n = _useAndesContext.i18n,
    device = _useAndesContext.device;
  const isMobile = device === null || device === void 0 ? void 0 : device.mobile;
  const componentId = useId(id);

  /**
   * @param {React.KeyboardEvent<HTMLInputElement>} event
   * @param {number} index
   * */
  const moveFocusBackwards = (event, index) => {
    const newIndex = index - 1;
    const newValue = [...currentValue];
    if (newIndex >= 0) newValue[newIndex] = '';
    setFocusedItem(state => newIndex > -1 ? newIndex : state);
    if (!isControlled) {
      setInputValue(newValue);
    }
    if (typeof onChange === 'function') {
      onChange(/** @type {React.ChangeEvent<HTMLInputElement>} */ /** @type {unknown} */event, newValue);
    }
  };
  const moveFocusForward = () => {
    setFocusedItem(state => {
      const newIndex = state + 1;
      return newIndex < digits ? newIndex : state;
    });
  };

  /**
   * @param {React.FormEvent<HTMLInputElement> | React.ChangeEvent<HTMLInputElement>} event
   * @param {string | string[]} values
   */
  const assignPastedValues = (event, values) => {
    const newValues = currentValue;
    const newFocus = values.length > digits ? digits - 1 : values.length - 1;
    for (let i = 0; i < currentValue.length; i += 1) {
      var _values$i;
      newValues[i] = (_values$i = values[i]) !== null && _values$i !== void 0 ? _values$i : '';
    }
    setInputValue(newValues);
    setFocusedItem(newFocus);
    if (typeof onPaste === 'function') onPaste(/** @type {React.ClipboardEvent<HTMLInputElement>} */event, newValues);
  };

  /** @param {React.ClipboardEvent<HTMLInputElement>} event */
  const handlePaste = event => {
    var _event$clipboardData;
    const textPlainItems = Array.from((event === null || event === void 0 || (_event$clipboardData = event.clipboardData) === null || _event$clipboardData === void 0 ? void 0 : _event$clipboardData.items) || []).filter(item => item.type === 'text/plain' || item.type === 'text');
    if (textPlainItems.length > 0) {
      var _textPlainItems$;
      (_textPlainItems$ = textPlainItems[0]) === null || _textPlainItems$ === void 0 || _textPlainItems$.getAsString(text => {
        const pastedValues = text.replace(numbersRegex, '').split('');
        assignPastedValues(event, pastedValues);
      });
    }
  };

  /** @param {React.FormEvent<HTMLInputElement>} event */
  const onInputHandler = event => {
    const eventValue = (/** @type {React.ChangeEvent<HTMLInputElement>} */event).target.value;
    if ((eventValue === null || eventValue === void 0 ? void 0 : eventValue.length) === digits) {
      setFocusedItem(digits - 1);
      return;
    }
    if (!onlyNumberRegex.test(eventValue) || eventValue.length > 1 && eventValue.length !== digits) {
      return;
    }
    if ((eventValue === null || eventValue === void 0 ? void 0 : eventValue.trim()) !== '') {
      moveFocusForward();
    }
  };

  /**
   * @param {React.ChangeEvent<HTMLInputElement>} event
   * @param {number} index
   */
  const onChangeHandler = (event, index) => {
    const eventValue = event.target.value;
    if (!onlyNumberRegex.test(eventValue) || eventValue.length > 1 && eventValue.length !== digits) {
      return;
    }
    const newValue = [...currentValue];
    if (eventValue.length === digits) {
      const values = eventValue.split('');
      for (let i = 0; i < currentValue.length; i += 1) {
        var _values$i2;
        newValue[i] = (_values$i2 = values[i]) !== null && _values$i2 !== void 0 ? _values$i2 : '';
      }
    } else {
      newValue[index] = eventValue;
    }
    if (!isControlled) {
      setInputValue(newValue);
    }
    if (typeof onChange === 'function') {
      onChange(event, newValue);
    }
    return /** @type {string | string[]} */newValue;
  };

  /**
   * @param {React.KeyboardEvent<HTMLInputElement>} event
   * @param {number} index
   */
  const onKeyDownHandler = (event, index) => {
    const eventValue = /** @type {HTMLInputElement} */event.target.value;
    if (event.code === BACKSPACE || event.key === 'Backspace') {
      if (eventValue === '') {
        moveFocusBackwards(event, index);
        return;
      }
    }
    if (event.code === TAB || event.key === 'Tab') {
      if (index === digits - 1 || index === 0) {
        setFocusedItem(NOT_FOCUSABLE_INDEX);
        return;
      }
    }
    const isPasting = (event.metaKey || event.ctrlKey) && event.code === KEY_V;
    if (isPasting) return;
    if (includes(specialCharacters, event.key)) return;
    const isInvalid = event.key.match(numbersRegex);
    if (isInvalid) {
      event.preventDefault();
    }
  };

  /** @param {number} index */
  const focusValidItem = index => {
    const item = inputsRef.current[index];
    if (item) {
      if (!firstRender.current) {
        item.focus();
      } else {
        autofocus(item, firstRender.current);
        firstRender.current = false;
      }
    }
  };

  /** @param {React.MouseEvent<HTMLInputElement>} event */
  const onClickHandler = event => {
    if (isEmpty(currentValue) && (isWebView() || !hasTouchSupport() && !isMobile)) {
      event.preventDefault();
      if (focusedItem === DEFAULT_FOCUS_INDEX) {
        focusValidItem(DEFAULT_FOCUS_INDEX);
      } else {
        setFocusedItem(DEFAULT_FOCUS_INDEX);
      }
    }
  };

  /**
   * @param {React.FocusEvent<HTMLInputElement>} event
   * @param {number} index
   */
  const onFocusHandler = (event, index) => {
    setFocusedItem(index);
    if (typeof focusProps.onFocus === 'function') focusProps.onFocus(event);
  };

  /** @param {React.FocusEvent<HTMLInputElement>} event */
  const handleBlur = event => {
    setFocusedItem(NOT_FOCUSABLE_INDEX);
    if (typeof focusProps.onBlur === 'function') focusProps.onBlur(event);
  };
  useEffect(() => {
    if (focusedItem !== NOT_FOCUSABLE_INDEX) {
      focusValidItem(focusedItem);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [focusedItem]);
  const containerClassNames = classNames(numericNamespace, className, {
    [`${namespace$2}--${modifier}`]: modifier,
    [`${namespace$2}--centered`]: centered
  });
  return /*#__PURE__*/React.createElement("div", _extends({
    className: containerClassNames,
    role: "group",
    "aria-label": srLabel,
    "aria-labelledby": label ? `${componentId}-label` : undefined,
    "aria-describedby": helper ? `${componentId}-message` : undefined,
    onPaste: handlePaste,
    ref: ref
  }, andesDataAttributes), label && /*#__PURE__*/React.createElement("span", {
    id: `${componentId}-label`,
    className: `${namespace$2}__label`,
    "data-andes-codeinput-label": true
  }, label), /*#__PURE__*/React.createElement("div", {
    className: `${numericNamespace}-wrapper`
  }, [...Array(digits)].map((_, index) => {
    const textFieldPadding = classNames(`${namespace$2}__textfield-space`, {
      [`${namespace$2}__textfield-space--big`]: digits === 6 && index === 2
    });
    const inputWrapperClasses = classNames(namespace$2, `${namespace$2}--textfield`, {
      [`${namespace$2}--centered`]: centered,
      [`${namespace$2}--focused`]: focusedItem === index,
      [`${namespace$2}--visible-focused`]: isFocusVisible && focusedItem === index,
      [`${namespace$2}--${modifier}`]: modifier,
      [`${namespace$2}--disabled`]: disabled
    });
    return /*#__PURE__*/React.createElement("div", {
      className: textFieldPadding,
      key: `${componentId}-${index}`
    }, /*#__PURE__*/React.createElement("div", {
      className: inputWrapperClasses
    }, /*#__PURE__*/React.createElement("div", {
      className: `${namespace$2}__control`
    }, /*#__PURE__*/React.createElement("input", _extends({
      className: `${namespace$2}__field`,
      "aria-label": i18n.pgettext(translationsNamespace, 'Dígito {0}', index + 1),
      type: strict ? 'tel' : 'number',
      pattern: "[0-9]*",
      value: currentValue[index],
      id: `${componentId}-digit-${index}`,
      onChange: evt => onChangeHandler(evt, index),
      onInput: onInputHandler,
      onKeyDown: evt => onKeyDownHandler(evt, index),
      onFocus: evt => onFocusHandler(evt, index),
      onBlur: handleBlur,
      onMouseDown: onClickHandler,
      disabled: disabled,
      ref: el => {
        inputsRef.current[index] = el;
      },
      "aria-invalid": modifier === 'error'
    }, index === 0 && {
      autoComplete: 'one-time-code',
      autoFocus: focus !== false
    }, {
      "data-andes-codeinput-input": true
    })), modifier === 'completed' && /*#__PURE__*/React.createElement(ValidatedIcon$1, {
      notificationsLabel: i18n.pgettext(translationsNamespace, 'Campo de texto valido')
    }))));
  })), helper && /*#__PURE__*/React.createElement("div", {
    className: `${namespace$2}__bottom`
  }, /*#__PURE__*/React.createElement(Helper, {
    id: `${id}-message`,
    modifier: modifier === 'error' ? modifier : undefined
    // @ts-expect-error: Internal prop
    ,

    andesDataAttributes: {
      'data-andes-codeinput-helper': true
    }
  }, helper)));
};
Numeric.displayName = 'Numeric';
Numeric.propTypes = {
  centered: PropTypes.bool,
  className: PropTypes.string,
  defaultValue: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string]),
  digits: PropTypes.number,
  disabled: PropTypes.bool,
  focus: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]),
  helper: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  id: PropTypes.string,
  label: checkLabelProp,
  modifier: PropTypes.oneOf(['completed', 'error']),
  onChange: PropTypes.func,
  onPaste: PropTypes.func,
  srLabel: checkLabelProp,
  strict: PropTypes.bool,
  value: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string]),
  andesDataAttributes: PropTypes.shape({})
};
var Numeric$1 = Numeric;
const _excluded$5 = ["className", "id", "ref"];
const namespace$1 = 'andes-code-input';

/**
 * @typedef {import('../..').CodeInput} CodeInputProps
 */

/** @type {CodeInputProps} */
const CodeInput = _ref => {
  let className = _ref.className,
    id = _ref.id,
    ref = _ref.ref,
    otherProps = _objectWithoutProperties(_ref, _excluded$5);
  const _useThemeContext = useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  const codeInputClassNames = classNames(namespace$1, className, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('textfield'));
  const componentId = useId(id);
  const disabled = otherProps.disabled,
    modifier = otherProps.modifier;
  return /*#__PURE__*/React.createElement(Numeric$1, _extends({
    className: codeInputClassNames,
    id: componentId,
    ref: ref,
    andesDataAttributes: _objectSpread(_objectSpread({
      'data-andes-codeinput': true,
      'data-andes-codeinput-modifier': modifier
    }, computeDataAndesState({
      disabled,
      error: modifier === 'error'
    })), themeHash)
  }, otherProps));
};
CodeInput.displayName = 'CodeInput';
CodeInput.propTypes = {
  centered: PropTypes.bool,
  className: PropTypes.string,
  defaultValue: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string]),
  digits: PropTypes.number,
  disabled: PropTypes.bool,
  focus: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]),
  helper: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  id: PropTypes.string,
  label: checkLabelProp,
  modifier: PropTypes.oneOf(['completed', 'error']),
  onChange: PropTypes.func,
  onPaste: PropTypes.func,
  strict: PropTypes.bool,
  srLabel: checkLabelProp,
  value: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string])
};
var CodeInput$1 = CodeInput;

/**
 * @typedef {Object} LabelDataset
 * @prop {string} showLabel
 * @prop {string} hideLabel
 * @prop {string} show
 */

var handlers = {
  /**
   * @param {HTMLDivElement | undefined | null} element
   */
  onFocus: function onFocus(element) {
    element === null || element === void 0 || element.classList.add('andes-form-control--focused');
  },
  /**
   * @param {HTMLElement | undefined | null} element
   * @param {string} value
   */
  onBlur: function onBlur(element, value) {
    element === null || element === void 0 || element.classList.remove('andes-form-control--focused');
    if (value && value.length) {
      element === null || element === void 0 || element.classList.add('andes-form-control--floated');
    } else {
      element === null || element === void 0 || element.classList.remove('andes-form-control--floated');
    }
  },
  /**
   * @param {string} value
   * @param {HTMLInputElement | undefined} inputHiddenRef
   */
  onChange: function onChange(value, inputHiddenRef) {
    if (inputHiddenRef) {
      inputHiddenRef.value = value;
    }
  }
};
const _excluded$4 = ["autoFocus", "className", "defaultValue", "helper", "i18n", "id", "innerRef", "inputProps", "label", "maxLength", "modifier", "onClickShowPassword", "showHide", "srLabel", "value", "andesDataAttributes"];
const namespace = 'andes-form-control';
const iconAttributes = {
  color: 'currentColor'
};

/** @typedef {import('@meli/i18n').I18n} I18n */

/**
 * @typedef PasswordState
 * @prop {boolean} focused
 * @prop {boolean} showPassword
 */

/** @typedef {import('../..').PasswordProps} PasswordPublicProps */

/**
 * @typedef {Object} PasswordInternalProps
 * @prop {React.RefObject<HTMLDivElement | null> | ((refElement: HTMLDivElement | null) => void)} [innerRef]
 * @prop {import('@meli/i18n').I18n} i18n
 * @prop {Record<string, unknown>} [andesDataAttributes]
 */

/** @typedef {PasswordPublicProps & PasswordInternalProps} PasswordProps */

/** @extends {React.Component<PasswordProps, PasswordState>} */
class Password extends React.Component {
  /** @param {PasswordProps} props */
  constructor(props) {
    super(props);

    /** @type {PasswordState} */
    this.state = {
      focused: props.autoFocus || false,
      showPassword: false
    };
    this.isUncontrolledValue = props.value === undefined;
    this.setField = this.setField.bind(this);
    this.setContainer = this.setContainer.bind(this);
    this.onBlurHandler = this.onBlurHandler.bind(this);
    this.onFocusHandler = this.onFocusHandler.bind(this);
    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.onChangeIconHandler = this.onChangeIconHandler.bind(this);
    this.renderShowHidePasswordButton = this.renderShowHidePasswordButton.bind(this);
    this.setHiddenPass = this.setHiddenPass.bind(this);
  }

  /** @param {React.FocusEvent<HTMLInputElement>} event */
  onBlurHandler(event) {
    const container = this.container,
      field = this.field;
    const value = (/** @type {HTMLInputElement} */field).value;
    const onBlur = this.props.onBlur;
    handlers.onBlur(container, value);
    this.setState({
      focused: false
    });
    if (onBlur) onBlur(event);
  }

  /** @param {React.FocusEvent<HTMLInputElement>} event */
  onFocusHandler(event) {
    const container = this.container;
    const onFocus = this.props.onFocus;
    handlers.onFocus(container);
    this.setState({
      focused: true
    });
    if (onFocus) onFocus(event);
  }

  /** @param {React.ChangeEvent<HTMLInputElement>} event */
  onChangeHandler(event) {
    const value = event.target.value;
    const _this$props = this.props,
      _this$props$showHide = _this$props.showHide,
      showHide = _this$props$showHide === void 0 ? false : _this$props$showHide,
      onChange = _this$props.onChange;
    const hiddenPass = this.hiddenPass;
    if (showHide && this.isUncontrolledValue) handlers.onChange(value, hiddenPass);
    if (onChange) onChange(event);
  }
  onChangeIconHandler() {
    const onClickShowPassword = this.props.onClickShowPassword;
    this.setState(prevState => _objectSpread(_objectSpread({}, prevState), {}, {
      showPassword: !prevState.showPassword
    }));
    if (onClickShowPassword) onClickShowPassword();
  }

  /** @param {HTMLInputElement} fieldRef */
  setField(fieldRef) {
    this.field = fieldRef;
  }

  /** @param {HTMLDivElement} containerRef */
  setContainer(containerRef) {
    const innerRef = this.props.innerRef;
    this.container = containerRef;
    if (innerRef) /** @type {React.RefObject<HTMLDivElement | null>} */innerRef.current = containerRef;
  }

  /** @param {HTMLInputElement} passRef */
  setHiddenPass(passRef) {
    this.hiddenPass = passRef;
  }

  /** @param {I18n} i18n */
  renderShowHidePasswordButton(i18n) {
    const id = this.props.id;
    const showPassword = this.state.showPassword;
    return /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__password-embedded`
    }, /*#__PURE__*/React.createElement("input", {
      checked: showPassword,
      className: `${namespace}__password-checkbox`,
      id: `${id}-checkbox`,
      type: "checkbox",
      onChange: this.onChangeIconHandler,
      "aria-label": i18n.pgettext('andes-textfield', 'Mostrar contraseña')
    }), /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__password-icon`
    }, showPassword ? /*#__PURE__*/React.createElement(VisibilityON20, iconAttributes) : /*#__PURE__*/React.createElement(VisibilityOFF20, iconAttributes)));
  }
  render() {
    const _this$props2 = this.props,
      _this$props2$autoFocu = _this$props2.autoFocus,
      autoFocus = _this$props2$autoFocu === void 0 ? false : _this$props2$autoFocu,
      className = _this$props2.className,
      defaultValue = _this$props2.defaultValue,
      helper = _this$props2.helper,
      i18n = _this$props2.i18n,
      id = _this$props2.id;
    _this$props2.innerRef;
    const _this$props2$inputPro = _this$props2.inputProps,
      inputProps = _this$props2$inputPro === void 0 ? {} : _this$props2$inputPro,
      label = _this$props2.label,
      _this$props2$maxLengt = _this$props2.maxLength,
      maxLength = _this$props2$maxLengt === void 0 ? 120 : _this$props2$maxLengt,
      modifier = _this$props2.modifier;
    _this$props2.onClickShowPassword;
    const _this$props2$showHide = _this$props2.showHide,
      showHide = _this$props2$showHide === void 0 ? false : _this$props2$showHide,
      srLabel = _this$props2.srLabel,
      value = _this$props2.value,
      andesDataAttributes = _this$props2.andesDataAttributes,
      other = _objectWithoutProperties(_this$props2, _excluded$4);
    const _this$state = this.state,
      focused = _this$state.focused,
      showPassword = _this$state.showPassword;
    const fieldProps = _objectSpread(_objectSpread(_objectSpread({}, other), inputProps), {}, {
      'aria-describedby': helper ? `${id}-message` : undefined,
      'aria-invalid': modifier === 'error',
      'aria-label': srLabel,
      autoFocus,
      className: `${namespace}__field`,
      id,
      // eslint-disable-next-line object-shorthand
      maxLength: (/** @type {number} */maxLength),
      onBlur: this.onBlurHandler,
      onChange: this.onChangeHandler,
      onFocus: this.onFocusHandler,
      ref: this.setField,
      type: showPassword ? 'text' : 'password',
      value,
      defaultValue,
      autoComplete: inputProps.autoComplete || 'current-password'
    });
    const fieldHiddenProps = {
      style: {
        display: 'none'
      },
      type: 'password',
      id: 'password-hidden',
      value,
      defaultValue,
      ref: this.setHiddenPass,
      autoComplete: inputProps.autoComplete || 'current-password'
    };
    const classes = classNames(namespace, className, `${namespace}--textfield`, {
      [`${namespace}--password`]: showHide,
      [`${namespace}--focused`]: autoFocus,
      [`${namespace}--${modifier}`]: modifier
    });
    return /*#__PURE__*/React.createElement("div", _extends({
      className: classes,
      ref: this.setContainer,
      "data-andes-password": true
    }, andesDataAttributes), label && /*#__PURE__*/React.createElement("label", {
      htmlFor: id,
      className: `${namespace}__label`
    }, label), /*#__PURE__*/React.createElement(FocusRing, {
      within: true,
      isTextInput: true,
      focusRingClass: focused ? 'visible-focused' : ''
    }, /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__control`
    }, /*#__PURE__*/React.createElement("input", _extends({
      key: "field"
    }, fieldProps, {
      "data-andes-password-input": true
    })), /*#__PURE__*/React.createElement("input", _extends({
      key: "field-password-hidden"
    }, fieldHiddenProps)), showHide && this.renderShowHidePasswordButton(i18n))), (helper || showHide) && /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__bottom`
    }, /*#__PURE__*/React.createElement(Helper, {
      id: `${id}-message`,
      modifier: modifier === 'error' ? modifier : undefined
      // @ts-expect-error: Internal prop
      ,

      andesDataAttributes: {
        'data-andes-password-helper': true
      }
    }, helper)));
  }
}
Password.propTypes = {
  autoFocus: PropTypes.bool,
  className: PropTypes.string,
  defaultValue: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  helper: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  id: PropTypes.string,
  innerRef: PropTypes.shape({
    current: PropTypes.oneOfType([PropTypes.node, PropTypes.object])
  }),
  inputProps: PropTypes.shape({}),
  i18n: i18NPropTypeChecker,
  label: checkLabelProp,
  maxLength: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  modifier: PropTypes.oneOf(['indeterminate', 'completed', 'error']),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onClickShowPassword: PropTypes.func,
  onFocus: PropTypes.func,
  placeholder: PropTypes.string,
  showHide: PropTypes.bool,
  srLabel: checkLabelProp,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
};
var Password$1 = Password;
const _excluded$3 = ["ref", "id", "className"];

/**  @type {typeof import('../..').Password} */
const PasswordContainer = _ref => {
  let ref = _ref.ref,
    id = _ref.id,
    className = _ref.className,
    otherProps = _objectWithoutProperties(_ref, _excluded$3);
  const componentId = useId(id);
  const _useThemeContext = useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  const _useAndesContext = useAndesContext('@andes/textfield', {
      en: enTranslations,
      'es-AR': esArTranslations,
      'pt-BR': ptBrTranslations,
      'zh-CN': zhCnTranslations
    }),
    i18n = _useAndesContext.i18n;
  return /*#__PURE__*/React.createElement(Password$1, _extends({
    innerRef: ref
  }, otherProps, {
    i18n: i18n,
    id: componentId,
    className: classNames(className, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('textfield')),
    andesDataAttributes: themeHash
  }));
};
PasswordContainer.displayName = 'PasswordContainer';
PasswordContainer.propTypes = {
  id: PropTypes.string
};
var PasswordContainer$1 = PasswordContainer;

/* eslint-disable max-len */

/**
 * @typedef {Object} ClearButtonProps
 * @prop {string} [`aria-label`]
 * @prop {string} namespace
 * @prop {() => void} onClick
 * @prop {string} [size]
 */

/** @type {React.FunctionComponent<ClearButtonProps>} */
const ClearButton = _ref => {
  let namespace = _ref.namespace,
    onClick = _ref.onClick,
    _ref$ariaLabel = _ref['aria-label'],
    ariaLabel = _ref$ariaLabel === void 0 ? '' : _ref$ariaLabel,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'large' : _ref$size;
  return /*#__PURE__*/React.createElement("button", {
    "aria-label": ariaLabel,
    className: `${namespace}__clear-icon`,
    onClick: onClick,
    onMouseDown: e => e.preventDefault(),
    type: "button"
  }, size === 'large' ? /*#__PURE__*/React.createElement(Close20, {
    color: "currentColor"
  }) : /*#__PURE__*/React.createElement(Close12, {
    color: "currentColor"
  }));
};
ClearButton.propTypes = {
  'aria-label': PropTypes.string,
  namespace: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  size: PropTypes.string
};
var ClearButton$1 = ClearButton;

/** @typedef {import('../..').TextFieldProps["modifier"] | "caution"} ModifierProp */

const DEPRECATED_MODIFIERS = ['warning'];

/** @param {ModifierProp} [modifier] */
const computeModifier = modifier => {
  warning(!modifier || typeof modifier === 'string' && !DEPRECATED_MODIFIERS.includes(modifier), WARNING_DEPRECATED_MODIFIER_WARNING);
  switch (modifier) {
    case 'warning':
    case 'caution':
      return 'caution';
    case 'error':
      return 'error';
    default:
      return undefined;
  }
};
const _excluded$2 = ["centered", "children", "className", "clear", "countdown", "disabled", "defaultValue", "embeddedChildren", "focus", "helper", "id", "innerRef", "i18n", "inputProps", "label", "maxLength", "modifier", "multiline", "numeric", "onChange", "placeholder", "prefix", "rows", "select", "size", "split", "srLabel", "suffix", "value", "autoFocus", "andesDataAttributes"];

/**
 * @typedef TextFieldState
 * @prop {string} height
 * @prop {number} countdown
 * @prop {boolean} focused
 * @prop {HTMLInputElement | HTMLTextAreaElement | null} fieldRef
 * @prop {TextFieldProps} prevProps
 */

/**
 * @typedef {Object} TextFieldInternalProps
 * @prop {React.RefObject<HTMLDivElement | null> | ((refElement: HTMLDivElement | null) => void)} [innerRef]
 * @prop {boolean} [embeddedChildren]
 * @prop {import('@meli/i18n').I18n} i18n
 * @prop {boolean} [numeric]
 * @prop {boolean} [split]
 * @prop {Record<string, unknown>} [andesDataAttributes]
 */

/** @typedef {import('../..').TextFieldProps} TextFieldPublicProps */
/** @typedef {TextFieldPublicProps & TextFieldInternalProps} TextFieldProps */

const emptyEvent = /** @type {React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>} */{
  target: {
    value: ''
  }
};

/**
 * Textfield allows enter data. It can be for enter numbers, text and passwords as well.
 */

/** @extends {React.Component<TextFieldProps, TextFieldState>} */
class TextField extends React.Component {
  /** @param {TextFieldProps} props */
  constructor(props) {
    super(props);
    const value = String(props.defaultValue || props.value || '');

    /** @type TextFieldState */
    this.state = {
      height: 'auto',
      fieldRef: null,
      countdown: value.length,
      focused: props.autoFocus || false,
      prevProps: props
    };
    this.isUncontrolledValue = props.value === undefined;
    this.setField = this.setField.bind(this);
    this.setRef = this.setRef.bind(this);
    this.onBlurHandler = this.onBlurHandler.bind(this);
    this.onFocusHandler = this.onFocusHandler.bind(this);
    this.onInputHandler = this.onInputHandler.bind(this);
    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.onClearHandler = this.onClearHandler.bind(this);
  }
  componentDidMount() {
    const _this$props = this.props,
      _this$props$multiline = _this$props.multiline,
      multiline = _this$props$multiline === void 0 ? false : _this$props$multiline,
      _this$props$focus = _this$props.focus,
      focus = _this$props$focus === void 0 ? false : _this$props$focus,
      _this$props$select = _this$props.select,
      select = _this$props$select === void 0 ? false : _this$props$select,
      _this$props$autoFocus = _this$props.autoFocus,
      autoFocus = _this$props$autoFocus === void 0 ? false : _this$props$autoFocus;
    const field = this.field;
    if (multiline && field) autosize(field);
    if (focus || autoFocus) {
      autofocus(field);
    }
    if (select) field === null || field === void 0 || field.select();
  }

  /**
   * @param {TextFieldState} prevState
   * @param {TextFieldProps} nextProps
   */
  static getDerivedStateFromProps(nextProps, prevState) {
    const fieldRef = prevState.fieldRef,
      prevProps = prevState.prevProps;
    const _nextProps$multiline = nextProps.multiline,
      multiline = _nextProps$multiline === void 0 ? false : _nextProps$multiline,
      nextValue = nextProps.value,
      _nextProps$focus = nextProps.focus,
      focus = _nextProps$focus === void 0 ? false : _nextProps$focus,
      _nextProps$select = nextProps.select,
      select = _nextProps$select === void 0 ? false : _nextProps$select,
      _nextProps$countdown = nextProps.countdown,
      countdown = _nextProps$countdown === void 0 ? false : _nextProps$countdown;
    // @ts-expect-error
    const isUncontrolledValue = prevState.isUncontrolledValue;
    const propsToCheck = ['multiline', 'focus', 'select', 'value', 'countdown'];
    const changedProps = propsToCheck.some(prop => nextProps[prop] !== prevProps[prop]);
    if (changedProps) {
      if (multiline && fieldRef) autosize.update(fieldRef);
      if (focus) fieldRef === null || fieldRef === void 0 || fieldRef.focus();
      if (select) fieldRef === null || fieldRef === void 0 || fieldRef.select();

      // Count characters
      if (countdown && !isUncontrolledValue && nextValue !== (fieldRef === null || fieldRef === void 0 ? void 0 : fieldRef.value)) {
        return {
          countdown: String(nextValue).length,
          prevProps: nextProps
        };
      }
    }
    return changedProps ? {
      prevProps: nextProps
    } : null;
  }

  /** @param {React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>} event */
  onBlurHandler(event) {
    const onBlur = this.props.onBlur;
    this.setState({
      focused: false
    });
    if (onBlur) onBlur(event);
  }

  /** @param {React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>} event */
  onFocusHandler(event) {
    const onFocus = this.props.onFocus;
    this.setState({
      focused: true
    });
    if (onFocus) onFocus(event);
  }

  /** @param {React.FormEvent<HTMLInputElement | HTMLTextAreaElement>} event */
  onInputHandler(event) {
    const onInput = this.props.onInput;
    const value = (/** @type {React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>} */event).target.value;
    this.setState({
      countdown: value.length
    });
    if (typeof onInput === 'function') onInput(/** @type {React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>} */event);
  }

  /** @param {React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>} event */
  onChangeHandler(event) {
    const onChange = this.props.onChange;
    if (typeof onChange === 'function') onChange(event);
  }
  onClearHandler() {
    const onChange = this.props.onChange;
    const field = this.field,
      isUncontrolledValue = this.isUncontrolledValue,
      onInputHandler = this.onInputHandler,
      onChangeHandler = this.onChangeHandler;
    if (isUncontrolledValue) /** @type {HTMLInputElement | HTMLTextAreaElement} */field.value = '';
    onInputHandler(emptyEvent);
    if (onChange) onChangeHandler(emptyEvent);
  }

  /** @param {HTMLInputElement | HTMLTextAreaElement | null} field */
  setField(field) {
    this.field = field;
    this.setState({
      fieldRef: field
    });
  }

  /** @param {HTMLDivElement} ref */
  setRef(ref) {
    const innerRef = this.props.innerRef;
    if (typeof innerRef === 'function') {
      innerRef(ref);
      return;
    }
    if (innerRef) innerRef.current = ref;
  }
  renderChildren() {
    const _this$props2 = this.props,
      children = _this$props2.children,
      _this$props2$embedded = _this$props2.embeddedChildren,
      embeddedChildren = _this$props2$embedded === void 0 ? true : _this$props2$embedded;
    if (embeddedChildren) {
      return /*#__PURE__*/React.createElement("div", {
        className: `${formControlNamespace}__embedded`,
        "data-andes-textfield-right-content": true
      }, children);
    }
    return children;
  }
  renderHelper() {
    const _this$props3 = this.props,
      helper = _this$props3.helper,
      id = _this$props3.id,
      modifier = _this$props3.modifier;
    return /*#__PURE__*/React.createElement(React.Fragment, null, helper && /*#__PURE__*/React.createElement(Helper, {
      id: `${id}-message`,
      modifier: computeModifier(modifier),
      className: `${formControlNamespace}--internal-helper`
      // @ts-expect-error: Internal prop
      ,

      andesDataAttributes: {
        'data-andes-textfield-helper': true
      }
    }, helper));
  }
  render() {
    const _this$props4 = this.props,
      _this$props4$centered = _this$props4.centered,
      centered = _this$props4$centered === void 0 ? false : _this$props4$centered,
      children = _this$props4.children,
      className = _this$props4.className,
      _this$props4$clear = _this$props4.clear,
      clear = _this$props4$clear === void 0 ? false : _this$props4$clear,
      _this$props4$countdow = _this$props4.countdown,
      countdown = _this$props4$countdow === void 0 ? false : _this$props4$countdow,
      _this$props4$disabled = _this$props4.disabled,
      disabled = _this$props4$disabled === void 0 ? false : _this$props4$disabled,
      defaultValue = _this$props4.defaultValue;
    _this$props4.embeddedChildren;
    const focus = _this$props4.focus,
      helper = _this$props4.helper,
      id = _this$props4.id;
    _this$props4.innerRef;
    const i18n = _this$props4.i18n,
      _this$props4$inputPro = _this$props4.inputProps,
      inputProps = _this$props4$inputPro === void 0 ? {} : _this$props4$inputPro,
      label = _this$props4.label,
      _this$props4$maxLengt = _this$props4.maxLength,
      maxLength = _this$props4$maxLengt === void 0 ? 120 : _this$props4$maxLengt,
      modifier = _this$props4.modifier,
      _this$props4$multilin = _this$props4.multiline,
      multiline = _this$props4$multilin === void 0 ? false : _this$props4$multilin,
      _this$props4$numeric = _this$props4.numeric,
      numeric = _this$props4$numeric === void 0 ? false : _this$props4$numeric;
    _this$props4.onChange;
    const placeholder = _this$props4.placeholder,
      _this$props4$prefix = _this$props4.prefix,
      prefix = _this$props4$prefix === void 0 ? '' : _this$props4$prefix,
      _this$props4$rows = _this$props4.rows,
      rows = _this$props4$rows === void 0 ? 1 : _this$props4$rows;
    _this$props4.select;
    const _this$props4$size = _this$props4.size,
      size = _this$props4$size === void 0 ? 'large' : _this$props4$size,
      _this$props4$split = _this$props4.split,
      split = _this$props4$split === void 0 ? false : _this$props4$split,
      srLabel = _this$props4.srLabel,
      _this$props4$suffix = _this$props4.suffix,
      suffix = _this$props4$suffix === void 0 ? '' : _this$props4$suffix,
      value = _this$props4.value,
      autoFocus = _this$props4.autoFocus,
      andesDataAttributes = _this$props4.andesDataAttributes,
      otherProps = _objectWithoutProperties(_this$props4, _excluded$2);
    const _this$state = this.state,
      focused = _this$state.focused,
      height = _this$state.height,
      count = _this$state.countdown;
    const showBottomInfo = !numeric && !split && modifier !== 'readonly' && !disabled && (countdown || helper);
    const classes = classNames(formControlNamespace, `${formControlNamespace}--textfield`, className, {
      [`${formControlNamespace}--multiline`]: multiline,
      [`${formControlNamespace}--countdown`]: countdown,
      [`${formControlNamespace}--centered`]: centered,
      [`${formControlNamespace}--focused`]: focused,
      [`${formControlNamespace}--${modifier}`]: modifier,
      [`${formControlNamespace}--disabled`]: disabled,
      [`${formControlNamespace}--medium`]: size === 'medium',
      [`${formControlNamespace}--textfield-with-prefix`]: !!prefix,
      [`${formControlNamespace}--textfield-with-suffix`]: !!suffix
    });
    const getAriaDescribedby = () => {
      const ariaDescribedby = inputProps['aria-describedby'];
      if (split) return ariaDescribedby;
      if (helper && countdown) return `${id}-message ${id}-count`;
      if (countdown) return `${id}-count`;
      if (helper) return `${id}-message`;
    };
    const suffixClasses = classNames(`${formControlNamespace}__suffix`);
    const prefixClasses = classNames(`${formControlNamespace}__prefix`);
    const fieldClasses = classNames(`${formControlNamespace}__field`, {
      [`${formControlNamespace}__field--multiline`]: multiline,
      [`${formControlNamespace}--medium`]: size === 'medium'
    });
    const bottomInfoClasses = classNames(`${formControlNamespace}__bottom`, {
      [`${formControlNamespace}--medium`]: size === 'medium'
    });
    const fieldProps = _objectSpread(_objectSpread(_objectSpread({}, otherProps), inputProps), {}, {
      id,
      defaultValue,
      value,
      className: fieldClasses,
      disabled: disabled || modifier === 'readonly',
      // eslint-disable-next-line object-shorthand
      maxLength: (/** @type {number} */maxLength),
      style: multiline ? {
        height
      } : {},
      placeholder,
      onBlur: this.onBlurHandler,
      onFocus: this.onFocusHandler,
      onChange: this.onChangeHandler,
      ref: this.setField,
      'aria-invalid': modifier === 'error' || modifier === 'warning' || modifier === 'caution' || (inputProps === null || inputProps === void 0 ? void 0 : inputProps['aria-invalid']) || (/** @type {boolean} */otherProps['aria-invalid']),
      'aria-label': srLabel,
      'aria-describedby': getAriaDescribedby(),
      // eslint-disable-next-line object-shorthand
      rows: (/** @type {number} */rows),
      autoFocus: focus || autoFocus
    });
    const Input = multiline ? 'textarea' : 'input';
    return /*#__PURE__*/React.createElement("div", _extends({
      className: classes,
      ref: this.setRef
    }, andesDataAttributes, {
      "data-andes-textfield": true,
      "data-andes-textfield-modifier": modifier
    }, computeDataAndesState({
      disabled,
      error: modifier === 'error',
      warning: modifier === 'warning'
    })), label && /*#__PURE__*/React.createElement("label", {
      htmlFor: id,
      "data-andes-textfield-label": true
    }, /*#__PURE__*/React.createElement("span", {
      className: `${formControlNamespace}__label`
    }, label)), /*#__PURE__*/React.createElement(FocusRing, {
      within: true,
      isTextInput: true,
      focusRingClass: focused ? 'visible-focused' : ''
    }, /*#__PURE__*/React.createElement("div", {
      className: `${formControlNamespace}__control`
    }, prefix && /*#__PURE__*/React.createElement("span", {
      className: prefixClasses,
      "data-andes-textfield-prefix": true
    }, prefix), /*#__PURE__*/React.createElement(Input, _extends({
      key: "field"
    }, fieldProps, {
      onInput: this.onInputHandler,
      "data-andes-textfield-input": true,
      "data-andes-textfield-input-type": Input
    })), suffix && /*#__PURE__*/React.createElement("span", {
      className: suffixClasses,
      "data-andes-textfield-suffix": true
    }, suffix), modifier === 'completed' && /*#__PURE__*/React.createElement(ValidatedIcon$1, {
      notificationsLabel: i18n.pgettext(namespace$3, 'Campo de texto valido'),
      size: size
    }), modifier === 'indeterminate' && /*#__PURE__*/React.createElement(ProgressIndicatorCircular, {
      className: `${formControlNamespace}--textfield-progress`,
      size: "xsmall",
      srAnnouncement: i18n.pgettext(namespace$3, 'Campo de texto procesando'),
      srLabel: i18n.pgettext(namespace$3, 'Campo de texto procesando')
    }), clear && /*#__PURE__*/React.createElement(ClearButton$1, {
      "aria-label": i18n.pgettext(namespace$3, 'Borrar'),
      namespace: formControlNamespace,
      onClick: this.onClearHandler,
      size: size
    }), children && this.renderChildren())), showBottomInfo && /*#__PURE__*/React.createElement("div", {
      className: bottomInfoClasses,
      "data-andes-textfield-bottom-info": true
    }, this.renderHelper(), countdown && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("span", {
      id: `${id}-countdown-progress`,
      className: `${formControlNamespace}__countdown`,
      role: "progressbar",
      "aria-valuenow": count,
      "aria-valuemin": 0,
      "aria-valuemax": Number(maxLength),
      "aria-label": i18n.pgettext(namespace$3, 'Maximo {0} caracteres', maxLength),
      "data-andes-textfield-countdown": true
    }, `${count} / ${maxLength}`), /*#__PURE__*/React.createElement(VisuallyHidden, {
      id: `${id}-count`,
      "aria-live": "polite",
      "aria-atomic": "true"
    }, i18n.pgettext(namespace$3, '{0} caracteres restantes', Number(maxLength) - count)))));
  }
}
TextField.displayName = 'TextField';
TextField.propTypes = {
  andesDataAttributes: PropTypes.shape({}),
  autoFocus: PropTypes.bool,
  centered: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
  clear: PropTypes.bool,
  countdown: PropTypes.bool,
  defaultValue: PropTypes.string,
  disabled: PropTypes.bool,
  embeddedChildren: PropTypes.bool,
  focus: PropTypes.bool,
  helper: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  id: PropTypes.string.isRequired,
  innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.shape({
    current: PropTypes.oneOfType([PropTypes.node, PropTypes.object])
  })]),
  inputProps: PropTypes.shape({}),
  i18n: i18NPropTypeChecker,
  label: checkLabelProp,
  maxLength: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  modifier: PropTypes.oneOf(['indeterminate', 'completed', 'error', 'readonly', 'caution']),
  multiline: PropTypes.bool,
  numeric: PropTypes.bool,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onInput: PropTypes.func,
  placeholder: PropTypes.string,
  prefix: PropTypes.string,
  rows: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  select: PropTypes.bool,
  size: PropTypes.oneOf(['medium', 'large']),
  split: PropTypes.bool,
  srLabel: checkLabelProp,
  suffix: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
};
var TextField$1 = TextField;
const _excluded$1 = ["ref"],
  _excluded2 = ["autoFocus", "children", "className", "defaultSelection", "defaultValue", "disabled", "helper", "inputProps", "id", "label", "list", "modifier", "placeholder", "onBlur", "onChange", "onDropdownOpen", "onFocus", "splitIcon", "splitLabel", "splitPosition", "srLabel", "value"];

/** @type {typeof import('../..').SplitTextfield} */
const SplitTextfield = _ref => {
  let ref = _ref.ref,
    props = _objectWithoutProperties(_ref, _excluded$1);
  const _props$autoFocus = (/** @type {import('../..').SplitTextfieldProps} */props).autoFocus,
    autoFocus = _props$autoFocus === void 0 ? false : _props$autoFocus,
    children = (/** @type {import('../..').SplitTextfieldProps} */props).children,
    className = (/** @type {import('../..').SplitTextfieldProps} */props).className,
    _props$defaultSelecti = (/** @type {import('../..').SplitTextfieldProps} */props).defaultSelection,
    defaultSelection = _props$defaultSelecti === void 0 ? 0 : _props$defaultSelecti,
    defaultValue = (/** @type {import('../..').SplitTextfieldProps} */props).defaultValue,
    _props$disabled = (/** @type {import('../..').SplitTextfieldProps} */props).disabled,
    disabled = _props$disabled === void 0 ? false : _props$disabled,
    helper = (/** @type {import('../..').SplitTextfieldProps} */props).helper,
    inputProps = (/** @type {import('../..').SplitTextfieldProps} */props).inputProps,
    id = (/** @type {import('../..').SplitTextfieldProps} */props).id,
    label = (/** @type {import('../..').SplitTextfieldProps} */props).label,
    list = (/** @type {import('../..').SplitTextfieldProps} */props).list,
    modifier = (/** @type {import('../..').SplitTextfieldProps} */props).modifier,
    placeholder = (/** @type {import('../..').SplitTextfieldProps} */props).placeholder,
    onBlur = (/** @type {import('../..').SplitTextfieldProps} */props).onBlur,
    onChange = (/** @type {import('../..').SplitTextfieldProps} */props).onChange,
    onDropdownOpen = (/** @type {import('../..').SplitTextfieldProps} */props).onDropdownOpen,
    onFocus = (/** @type {import('../..').SplitTextfieldProps} */props).onFocus,
    splitIcon = (/** @type {import('../..').SplitTextfieldProps} */props).splitIcon,
    splitLabel = (/** @type {import('../..').SplitTextfieldProps} */props).splitLabel,
    _props$splitPosition = (/** @type {import('../..').SplitTextfieldProps} */props).splitPosition,
    splitPosition = _props$splitPosition === void 0 ? 'left' : _props$splitPosition,
    srLabel = (/** @type {import('../..').SplitTextfieldProps} */props).srLabel,
    value = (/** @type {import('../..').SplitTextfieldProps} */props).value,
    other = _objectWithoutProperties(/** @type {import('../..').SplitTextfieldProps} */props, _excluded2);
  const getDefaultValue = () => {
    var _list$content$default;
    return list === null || list === void 0 || (_list$content$default = list.content[defaultSelection]) === null || _list$content$default === void 0 ? void 0 : _list$content$default.title;
  };

  /** @type {React.RefObject<HTMLDivElement | null>} */
  const ownContainerRef = useRef(null);

  /** @type {React.RefObject<HTMLDivElement | null>} */
  const containerRef = /** @type {React.RefObject<HTMLDivElement | null>} */ref || ownContainerRef;

  /**
   * This is a ref to the TextField component
   * @type {React.RefObject<any>}
   */
  const fieldRef = useRef(undefined);
  const componentId = useId(id);
  const _useFocusRing = useFocusRing({
      isTextInput: true
    }),
    isFocusVisible = _useFocusRing.isFocusVisible,
    focusProps = _useFocusRing.focusProps;
  const isReadonly = modifier === 'readonly';
  const _useThemeContext = useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  const _useAndesContext = useAndesContext('@andes/textfield', {
      en: enTranslations,
      'es-AR': esArTranslations,
      'pt-BR': ptBrTranslations,
      'zh-CN': zhCnTranslations
    }),
    i18n = _useAndesContext.i18n;
  const containerClassNames = classNames(formControlNamespace, className, `${formControlNamespace}--textfield`, `${formControlNamespace}--split`, {
    [`${formControlNamespace}--split__embedded`]: children,
    [`${formControlNamespace}--${modifier}`]: modifier,
    [`${formControlNamespace}--focused`]: autoFocus,
    [`${formControlNamespace}--disabled`]: disabled
  }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('textfield'));
  const splitControlClassNames = classNames(`${formControlNamespace}__control`, {
    [`${formControlNamespace}__control--visible-focused`]: isFocusVisible
  });
  const splitButtonClassNames = classNames(splitButtonNamespace, {
    [`${splitButtonNamespace}--right`]: splitPosition === 'right',
    [`${splitButtonNamespace}--readonly`]: isReadonly
  });
  const fieldClassNames = classNames(`${formControlNamespace}__field ${formControlNamespace}__split-field`, {
    [`${formControlNamespace}__split-field--readonly`]: isReadonly
  });
  const iconClassNames = classNames(`${splitButtonNamespace}-content--icon`, {
    [`${splitButtonNamespace}-content--icon-disabled`]: disabled
  });
  const splitItemClassNames = `${formControlNamespace}--split-item`;

  /** @type {import('@andes/dropdown').DropdownCommonProps['onChange']} */
  const customOnChange = (_, itemValue) => {
    if (!list) return;
    const content = list.content,
      onClick = list.onClick;
    if (typeof onClick === 'function' && itemValue) {
      // @ts-expect-error: The selected item will always exist
      onClick(content.find(_ref2 => {
        let title = _ref2.title;
        return title === itemValue;
      }));
    }
  };

  /** @param {React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>} event */
  const customOnBlur = event => {
    var _fieldRef$current;
    handlers.onBlur(containerRef.current, (_fieldRef$current = fieldRef.current) === null || _fieldRef$current === void 0 ? void 0 : _fieldRef$current.field.value);
    if (typeof focusProps.onBlur === 'function') focusProps.onBlur(event);
    if (typeof onBlur === 'function') onBlur(event);
  };

  /** @param {React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>} event */
  const customOnFocus = event => {
    handlers.onFocus(containerRef.current);
    if (typeof focusProps.onFocus === 'function') focusProps.onFocus(event);
    if (typeof onFocus === 'function') onFocus(event);
  };

  /** @param {React.ReactNode} triggerContent */
  const renderCustomTrigger = triggerContent => {
    let adaptedTriggerContent = triggerContent;
    if (splitLabel) adaptedTriggerContent = splitLabel;
    if (splitIcon) {
      adaptedTriggerContent = /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("span", {
        className: iconClassNames
      }, splitIcon), adaptedTriggerContent);
    }
    return adaptedTriggerContent;
  };
  const fieldProps = _objectSpread(_objectSpread({}, other), {}, {
    inputProps: _objectSpread({
      'aria-describedby': helper ? `${componentId}-message` : undefined
    }, inputProps),
    'aria-invalid': modifier === 'error',
    'aria-labelledby': `${componentId}-dropdown-trigger`,
    srLabel: srLabel || label,
    className: fieldClassNames,
    children,
    defaultValue,
    disabled,
    id: componentId,
    modifier: modifier !== 'error' ? modifier : undefined,
    onBlur: customOnBlur,
    onChange,
    onFocus: customOnFocus,
    placeholder,
    ref: fieldRef,
    value,
    i18n
  });
  return /*#__PURE__*/React.createElement("div", _extends({
    className: containerClassNames,
    ref: containerRef
  }, themeHash, {
    "data-andes-split-textfield": true
  }), label && /*#__PURE__*/React.createElement("label", {
    htmlFor: componentId
  }, /*#__PURE__*/React.createElement("span", {
    className: `${formControlNamespace}__label`
  }, label)), /*#__PURE__*/React.createElement("div", {
    className: splitControlClassNames
  }, splitPosition === 'right' && /*#__PURE__*/React.createElement(TextField$1, _extends({}, fieldProps, {
    split: true
  })), /*#__PURE__*/React.createElement(DropdownStandalone, {
    className: splitButtonClassNames,
    label: srLabel || label,
    menuAlignment: "bottom",
    defaultValue: getDefaultValue(),
    onOpen: onDropdownOpen,
    disabled: disabled || isReadonly,
    id: `${componentId}-dropdown`,
    renderCustomTrigger: renderCustomTrigger,
    onChange: customOnChange
  }, list === null || list === void 0 ? void 0 : list.content.map(item => /*#__PURE__*/
  /** @type {React.ReactElement} */
  React.createElement(DropdownItem, {
    value: item.title,
    className: splitItemClassNames,
    key: transformToKebabCase(item.title),
    title: item.title,
    leftContent: item.icon && /*#__PURE__*/React.createElement(item.icon, null)
  }))), splitPosition === 'left' && /*#__PURE__*/React.createElement(TextField$1, _extends({}, fieldProps, {
    split: true
  }))), helper && !disabled && !isReadonly && /*#__PURE__*/React.createElement("div", {
    className: `${formControlNamespace}__bottom`
  }, /*#__PURE__*/React.createElement(Helper, {
    id: `${componentId}-message`,
    modifier: computeModifier(modifier)
    // @ts-expect-error: Internal prop
    ,

    andesDataAttributes: {
      'data-andes-split-textfield-helper': true
    }
  }, helper)));
};
SplitTextfield.displayName = 'SplitTextfield';
SplitTextfield.propTypes = {
  autoFocus: PropTypes.bool,
  children: checkChildrenProp,
  className: PropTypes.string,
  defaultSelection: PropTypes.number,
  defaultValue: PropTypes.string,
  disabled: PropTypes.bool,
  helper: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  id: PropTypes.string,
  inputProps: PropTypes.shape({}),
  label: checkLabelProp,
  list: PropTypes.shape({
    onClick: PropTypes.func,
    content: PropTypes.arrayOf(PropTypes.shape({
      title: PropTypes.string,
      icon: PropTypes.func
    }))
  }),
  modifier: PropTypes.oneOf(['error', 'readonly', 'indeterminate', 'completed']),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onDropdownOpen: PropTypes.func,
  onFocus: PropTypes.func,
  placeholder: PropTypes.string,
  splitIcon: PropTypes.oneOfType([PropTypes.node, PropTypes.shape({
    title: PropTypes.string,
    src: PropTypes.string,
    alt: PropTypes.string,
    src2x: PropTypes.string
  })]),
  splitLabel: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  splitPosition: PropTypes.oneOf(['left', 'right']),
  srLabel: checkLabelProp,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
};
var SplitTextfield$1 = SplitTextfield;
const _excluded = ["ref", "id", "className"];

/**  @type {typeof import('../..').TextField} */
const TextfieldContainer = _ref => {
  let ref = _ref.ref,
    id = _ref.id,
    className = _ref.className,
    otherProps = _objectWithoutProperties(_ref, _excluded);
  const componentId = useId(id);
  const _useAndesContext = useAndesContext('@andes/textfield', {
      en: enTranslations,
      'es-AR': esArTranslations,
      'pt-BR': ptBrTranslations,
      'zh-CN': zhCnTranslations
    }),
    i18n = _useAndesContext.i18n;
  const _useThemeContext = useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  return /*#__PURE__*/React.createElement(TextField$1, _extends({
    i18n: i18n,
    innerRef: ref
  }, otherProps, {
    id: componentId,
    className: classNames(className, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('textfield')),
    andesDataAttributes: themeHash
  }));
};
TextfieldContainer.displayName = 'TextfieldContainer';
TextfieldContainer.propTypes = {
  id: PropTypes.string
};
var TextfieldContainer$1 = TextfieldContainer;
export { CodeInput$1 as CodeInput, PasswordContainer$1 as Password, SplitTextfield$1 as SplitTextfield, TextfieldContainer$1 as TextField };
