// @andes/snackbar/build/esm/module.js
import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { useEffect } from 'react';
import { isWebkit2 } from '@andes/common/build/webkit/utils';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import classNames from 'classnames';
import { VisuallyHidden } from '@andes/visually-hidden';
import includes from '@andes/common/build/utils/includes';
import warning from '@andes/common/build/utils/warning';
const namespace = 'andes-snackbar';
const DEPRECATED_COLOR_GREEN = 'green';
const DEPRECATED_COLOR_RED = 'red';
const AVAILABLE_COLOR_POSITIVE = 'positive';
const AVAILABLE_COLOR_NEGATIVE = 'negative';
const DEPRECATED_COLOR = [DEPRECATED_COLOR_GREEN, DEPRECATED_COLOR_RED];
const WARNING_COLOR_GREEN = `
  '${DEPRECATED_COLOR_GREEN}' option will be removed from 'color' options in a future major version.
  Please, use ${AVAILABLE_COLOR_POSITIVE}.`;
const WARNING_COLOR_RED = `
  '${DEPRECATED_COLOR_RED}' option will be removed from 'color' options in a future major version.
  Please, use ${AVAILABLE_COLOR_NEGATIVE}.`;
const DEPRECATED_COLOR_WARNINGS = {
  [`${DEPRECATED_COLOR_GREEN}`]: WARNING_COLOR_GREEN,
  [`${DEPRECATED_COLOR_RED}`]: WARNING_COLOR_RED
};
const computeTypeColor = _ref => {
  let color = _ref.color;
  if (includes(['red', 'negative'], color || '')) return 'negative';
  if (includes(['green', 'positive'], color || '')) return 'positive';
  return undefined;
};
var computeTypeColor$1 = computeTypeColor;
class Snackbar extends React.Component {
  constructor(props) {
    super(props);
    const _this$props$show = this.props.show,
      show = _this$props$show === void 0 ? false : _this$props$show;
    this.onActionClick = this.onActionClick.bind(this);
    this.onAnimationEnd = this.onAnimationEnd.bind(this);
    this.setMainContainerRef = this.setMainContainerRef.bind(this);
    this.state = {
      display: show,
      show,
      prevShow: show,
      timeoutExecuted: false,
      srLabelVisibility: false
    };
    this.style = {};
  }
  componentDidMount() {
    const _this$props = this.props,
      _this$props$show2 = _this$props.show,
      show = _this$props$show2 === void 0 ? false : _this$props$show2,
      color = _this$props.color,
      delay = _this$props.delay,
      onOpen = _this$props.onOpen;
    if (typeof color === 'string') {
      warning(!includes(DEPRECATED_COLOR, color), DEPRECATED_COLOR_WARNINGS[color]);
    }
    const fixedElement = document.querySelector('.andes-snackbar-fixed-element');
    if (fixedElement) {
      this.style = {
        bottom: `${fixedElement.offsetHeight + 16}px`
      };
    }
    if (show && delay) {
      this.delayTimeout = setTimeout(() => {
        this.setState({
          show: false
        }, () => {
          if (this.isReduceMotion()) this.onClose();
        });
      }, delay);
    }
    this.labelTimeout = setTimeout(() => {
      this.setState({
        srLabelVisibility: true
      });
    }, 30);
    if (show && typeof onOpen === 'function') {
      onOpen();
    }
  }
  componentWillUnmount() {
    clearTimeout(this.delayTimeout);
    clearTimeout(this.labelTimeout);
  }
  static getDerivedStateFromProps(nextProps, prevState) {
    const _nextProps$show = nextProps.show,
      show = _nextProps$show === void 0 ? false : _nextProps$show;
    const stateShow = prevState.show,
      timeoutExecuted = prevState.timeoutExecuted,
      prevShow = prevState.prevShow;
    const showChangedProp = show !== prevShow;
    if (showChangedProp) {
      if (show && !timeoutExecuted) {
        return {
          display: true,
          show,
          prevShow: show
        };
      }
      if (show !== stateShow && !timeoutExecuted) {
        return {
          show: show || false,
          prevShow: show
        };
      }
      if (timeoutExecuted) {
        return {
          timeoutExecuted: false,
          prevShow: show
        };
      }
    }
    return null;
  }
  componentDidUpdate(_, prevState) {
    const stateShow = this.state.show;
    const prevStateShow = prevState.show;
    const _this$props2 = this.props,
      onOpen = _this$props2.onOpen,
      _this$props2$show = _this$props2.show,
      show = _this$props2$show === void 0 ? false : _this$props2$show,
      delay = _this$props2.delay;
    const startedShowing = !prevStateShow && stateShow;
    const hasDelayProps = delay;
    if (stateShow !== prevStateShow) {
      if (show && onOpen) onOpen();
      if (this.isReduceMotion() && !show) this.onClose();
    }
    if (startedShowing && hasDelayProps) {
      this.delayTimeout = setTimeout(() => {
        this.setState({
          show: false,
          timeoutExecuted: true
        }, () => {
          if (this.isReduceMotion()) this.onClose();
        });
      }, delay);
    }
  }
  onActionClick(event, onClickProp) {
    if (typeof onClickProp === 'function') {
      clearTimeout(this.delayTimeout);
      onClickProp(event);
    }
  }
  onClose() {
    const onClose = this.props.onClose;
    clearTimeout(this.delayTimeout);
    if (typeof onClose === 'function') onClose();
  }
  onAnimationEnd() {
    const show = this.state.show;
    if (!show) {
      this.onClose();
      this.setState({
        display: false
      });
    }
  }
  isReduceMotion() {
    var _window;
    return typeof window !== 'undefined' && ((_window = window) === null || _window === void 0 ? void 0 : _window.matchMedia('(prefers-reduced-motion: reduce)').matches) === true;
  }
  setMainContainerRef(ref) {
    const innerRef = this.props.innerRef;
    if (innerRef) {
      if (typeof innerRef === 'function') {
        innerRef(ref);
      } else {
        innerRef.current = ref;
      }
    }
  }
  renderAction(_ref) {
    let text = _ref.text,
      onClick = _ref.onClick;
    const common = {
      'data-andes-snackbar-action': true,
      className: `${namespace}__action`,
      onClick: event => this.onActionClick(event, onClick)
    };
    return text && /*#__PURE__*/React.createElement("button", common, text);
  }
  render() {
    const _this$props3 = this.props,
      action = _this$props3.action,
      className = _this$props3.className,
      color = _this$props3.color,
      children = _this$props3.children,
      message = _this$props3.message,
      classes = _this$props3.classNames,
      andesDataAttributes = _this$props3.andesDataAttributes;
    const _this$state = this.state,
      display = _this$state.display,
      show = _this$state.show,
      srLabelVisibility = _this$state.srLabelVisibility;
    const computedAction = action || {};
    const computedColor = computeTypeColor$1({
      color
    });
    const containerClassNames = classNames(namespace, className, `${namespace}--${computedAction.position || 'right'}`, {
      [`${namespace}--${computedColor}`]: computedColor,
      [`${namespace}--with-action`]: computedAction || children,
      [`${namespace}--animate-show`]: show,
      [`${namespace}--animate-hide`]: show === false
    }, classes === null || classes === void 0 ? void 0 : classes.root);
    const isReduced = this.isReduceMotion();
    if (!display || show === null || isReduced && show === false) {
      return null;
    }
    return /*#__PURE__*/React.createElement("div", _extends({
      className: containerClassNames,
      onAnimationEnd: this.onAnimationEnd,
      role: "presentation",
      ref: this.setMainContainerRef,
      style: this.style,
      "data-andes-snackbar": true,
      "data-andes-snackbar-color": color
    }, andesDataAttributes), /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(VisuallyHidden, {
      role: "alert",
      "aria-live": "assertive"
    }, srLabelVisibility && message), /*#__PURE__*/React.createElement("p", {
      className: `${namespace}__message`,
      "data-andes-snackbar-message": true
    }, message), children && /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__children`,
      "data-andes-snackbar-content": true
    }, children)), this.renderAction(computedAction));
  }
}
Snackbar.displayName = 'Snackbar';
var Snackbar$1 = Snackbar;
const SnackbarWebkit = props => {
  const _React$useState = React.useState(false),
    _React$useState2 = _slicedToArray(_React$useState, 2),
    error = _React$useState2[0],
    setError = _React$useState2[1];
  const _props$show = props.show,
    show = _props$show === void 0 ? false : _props$show,
    message = props.message,
    action = props.action,
    delay = props.delay,
    onChange = props.onChange,
    onOpen = props.onOpen,
    onClose = props.onClose,
    color = props.color;
  function getColor() {
    let realColor = 'basic';
    if (color === 'green' || color === 'positive') {
      realColor = 'positive';
    } else if (color === 'red' || color === 'negative') {
      realColor = 'negative';
    }
    return realColor;
  }
  async function onOpenSnackbar() {
    const durationInSeconds = delay / 1000;
    const options = {
      message,
      duration: durationInSeconds,
      action: action === null || action === void 0 ? void 0 : action.text,
      onDismiss: typeof onClose === 'function',
      color: getColor()
    };
    function getEventName(name, id) {
      return `nordic:${name}${id ? `:${id}` : ''}`;
    }
    const _await$MobileWebKit$e = await MobileWebKit.executeNative({
        method: 'wk:snackbar-show',
        args: options
      }),
      id = _await$MobileWebKit$e.id;
    const dismissEvent = getEventName('onSnackbarDismiss', id);
    MobileWebKit.on(dismissEvent, () => {
      if (typeof onClose === 'function') {
        onClose();
        if (typeof onChange === 'function') onChange(false);
      }
      MobileWebKit.off(dismissEvent);
    });
    if (typeof onOpen === 'function') {
      onOpen();
      if (typeof onChange === 'function') onChange(true);
    }
  }
  useEffect(() => {
    if (show && !error) {
      onOpenSnackbar().catch(() => {
        setError(true);
      });
    }
  }, [show, error]);
  if (show && error) {
    return /*#__PURE__*/React.createElement(Snackbar$1, props);
  }
  return null;
};
var SnackbarWebkit$1 = SnackbarWebkit;
const _excluded = ["ref"];
const SnackbarContainer = /*#__PURE__*/function () {
  const SnackbarContainer = _ref => {
    var _props$action;
    let ref = _ref.ref,
      props = _objectWithoutProperties(_ref, _excluded);
    if (isWebkit2() && !(props !== null && props !== void 0 && props.children) && !ref && !(props !== null && props !== void 0 && (_props$action = props.action) !== null && _props$action !== void 0 && _props$action.onClick) && props.delay) {
      return /*#__PURE__*/React.createElement(SnackbarWebkit$1, props);
    }
    return /*#__PURE__*/React.createElement(Snackbar$1, _extends({
      innerRef: ref
    }, props));
  };
  SnackbarContainer.displayName = 'Snackbar';
  return SnackbarContainer;
}();
var SnackbarContainer$1 = SnackbarContainer;
export { SnackbarContainer$1 as Snackbar };
