// @nordic/core-context/src/common/types.ts
export type NordicDeviceType = 'desktop' | 'mobile' | 'tablet';

export type NordicPlatformType = 'android' | 'ios' | 'smartpos' | 'web';

export type NordicWebkitEngine = 'v1' | 'v2' | 'v3';

export type NordicWebkitScope = 'bottomsheet' | 'component' | 'page' | 'view';

export type NordicNativeAppInfo = {
  /**
   * Native app name.
   *
   * @example 'MercadoLibre'
   */
  name: string;

  /**
   * Native app version.
   *
   * @example '2.94.1'
   */
  version: string;
};

export type NordicWebkitInfo = {
  /**
   * Webkit engine name.
   *
   * @example 'v2'
   */
  engine: NordicWebkitEngine;

  /**
   * Webkit library version.
   *
   * @example '7.10.0'
   */
  version?: string;

  /**
   * Indicates whether the debug mode is enabled.
   *
   * **⚠️ WARNING**: This value is only defined when the engine is `v3`.
   */
  debugMode?: boolean;

  /**
   * Indicates whether the webview is set to fit the content height.
   */
  fitToContent?: boolean;

  /**
   * Webview rendering scope.
   */
  scope?: NordicWebkitScope;
};

export type NordicWebviewInfo = {
  /**
   * Native app information where the webview is displayed.
   */
  nativeApp: NordicNativeAppInfo;

  /**
   * Webkit information used to render the webview.
   */
  webkit: NordicWebkitInfo;
};

export type NordicCoreContext = {
  deviceType: NordicDeviceType;
  platformType: NordicPlatformType;
  webviewInfo?: NordicWebviewInfo;
  safeAreaInsets?: unknown;
  deviceInfo?: unknown;
  networkState?: unknown;
  i18n?: unknown;
  platform?: unknown;
};

declare global {
  interface NordicWindowContext {
    c?: NordicCoreContext;
  }
}
