// @andes/form/build/esm/module.js
import _extends from '@babel/runtime/helpers/extends';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { useState, useEffect } from 'react';
import { useAndesContext, useId } from '@andes/context';
import warning from '@andes/common/build/utils/warning';
import classNames from 'classnames';
import includes from '@andes/common/build/utils/includes';
import { PillIcon } from '@andes/badge';
import { useThemeContext } from '@andes/themes';
import { computeDataAndesState } from '@andes/common/build/utils/computeDataAndesState';
const WARNING_CSRF = `To prevent CSRF attacks, <Form> needs a csrf token. Please, provide one in <AndesProvider> \
via <csrfToken> prop.
Refer to https://github.com/melisource/fury_frontend-andes-ui/tree/master/packages/form#csrf-protection-by-default \
for more details.`;
const _excluded$1 = ["children", "ref"];
const Form = /*#__PURE__*/function () {
  const Form = _ref => {
    let children = _ref.children,
      ref = _ref.ref,
      props = _objectWithoutProperties(_ref, _excluded$1);
    const _useAndesContext = useAndesContext(),
      csrfToken = _useAndesContext.csrfToken;
    const _useState = useState(null),
      _useState2 = _slicedToArray(_useState, 2),
      metaCsrfToken = _useState2[0],
      setMetaCsrfToken = _useState2[1];
    const inputValue = csrfToken || metaCsrfToken;
    useEffect(() => {
      if (!csrfToken) {
        const metaTag = document.querySelector('meta[name=csrf-token]');
        if (metaTag) {
          setMetaCsrfToken(metaTag.getAttribute('content'));
        }
      }
    }, [csrfToken]);
    warning(csrfToken, WARNING_CSRF);
    return /*#__PURE__*/React.createElement("form", _extends({}, props, {
      ref: ref
    }), inputValue && /*#__PURE__*/React.createElement("input", {
      type: "hidden",
      name: "_csrf",
      value: inputValue
    }), children);
  };
  Form.displayName = 'Form';
  return Form;
}();
var Form$1 = Form;
const namespace = 'andes-helper';
const MODIFIER_CAUTION = 'caution';
const MODIFIER_ERROR = 'error';
const COLOR_CAUTION = 'caution';
const COLOR_NEGATIVE = 'negative';
const MODIFIER = [MODIFIER_CAUTION, MODIFIER_ERROR];
const COLOR = {
  [`${MODIFIER_CAUTION}`]: COLOR_CAUTION,
  [`${MODIFIER_ERROR}`]: COLOR_NEGATIVE
};
var enTranslations = {
  "andes-helper\u0004Precaución": [null, "Warning"],
  "andes-helper\u0004Error": [null, "Error"]
};
var esArTranslations = {
  "andes-helper\u0004Precaución": [null, "Precaución"],
  "andes-helper\u0004Error": [null, "Error"]
};
var ptBrTranslations = {
  "andes-helper\u0004Precaución": [null, "Alerta"],
  "andes-helper\u0004Error": [null, "Erro"]
};
const _excluded = ["andesDataAttributes", "children", "className", "id", "modifier"];
const Helper = /*#__PURE__*/function () {
  const Helper = props => {
    const _props$andesDataAttri = props.andesDataAttributes,
      andesDataAttributes = _props$andesDataAttri === void 0 ? {} : _props$andesDataAttri,
      children = props.children,
      className = props.className,
      id = props.id,
      modifier = props.modifier,
      otherProps = _objectWithoutProperties(props, _excluded);
    const componentId = useId(id);
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const _useAndesContext = useAndesContext('@andes/form', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations
      }),
      i18n = _useAndesContext.i18n;
    const showBadgeIcon = includes(MODIFIER, modifier || '');
    const helperClassName = classNames(namespace, className, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('helper'), {
      [`${namespace}--${modifier}`]: modifier
    });
    return /*#__PURE__*/React.createElement("div", _extends({
      className: helperClassName,
      id: componentId
    }, otherProps, themeHash, andesDataAttributes, {
      "data-andes-helper": true
    }, computeDataAndesState({
      error: modifier === 'error',
      caution: modifier === 'caution'
    })), showBadgeIcon && /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__badge`,
      "data-andes-helper-badge": true
    }, /*#__PURE__*/React.createElement(PillIcon, {
      color: modifier ? COLOR[modifier] : undefined,
      size: "xsmall",
      srLabel: i18n.pgettext(namespace, modifier === 'error' ? 'Error' : 'Precaución')
    })), /*#__PURE__*/React.createElement("span", {
      className: `${namespace}__label`,
      "data-andes-helper-label": true
    }, children));
  };
  Helper.displayName = 'Helper';
  return Helper;
}();
var Helper$1 = Helper;
export { Form$1 as Form, Helper$1 as Helper };
