// babel-plugin-nordic-module-resolver/src/index.js
/* eslint-disable n/no-exports-assign */
const normalizeOptions = require('./normalize-options');
const resolvePath = require('./resolve-path');
const transformCall = require('./transformers/call');
const transformImport = require('./transformers/import');

const importVisitors = {
  CallExpression: transformCall,
  'ImportDeclaration|ExportDeclaration': transformImport,
};

const visitor = {
  Program: {
    enter(programPath, state) {
      programPath.traverse(importVisitors, state);
    },
    exit(programPath, state) {
      programPath.traverse(importVisitors, state);
    },
  },
};

const moduleResolver = ({ types }) => ({
  name: 'nordic-module-resolver',

  manipulateOptions(opts) {
    if (typeof opts.filename === 'undefined') {
      opts.filename = 'unknown';
    }
  },

  pre(file) {
    this.types = types;

    const { filename } = file.opts;

    const subpath = this.opts.subpathImportsResolver?.(filename);

    if (subpath) {
      this.opts = { ...this.opts, alias: { ...subpath, ...this.opts.alias } };
    }

    this.normalizedOpts = normalizeOptions(filename, this.opts);
    // We need to keep track of all handled nodes so we do not try to transform them twice,
    // because we run before (enter) and after (exit) all nodes are handled
    this.moduleResolverVisited = new Set();
  },

  visitor,

  post() {
    this.moduleResolverVisited.clear();
  },
});

module.exports = moduleResolver;
exports = module.exports;
exports.resolvePath = resolvePath;
