// babel-plugin-nordic-module-resolver/src/utils.js
const fs = require('fs');
const path = require('path');
const { fileURLToPath } = require('url');

const resolve = require('resolve');

function nodeResolvePath(modulePath, basedir, extensions) {
  try {
    return resolve.sync(modulePath, { basedir, extensions });
  } catch (e) {
    return null;
  }
}

function isRelativePath(nodePath) {
  return nodePath.match(/^\.?\.\//);
}

function toPosixPath(modulePath) {
  return modulePath.replace(/\\/g, '/');
}

function toLocalPath(modulePath) {
  let localPath = modulePath.replace(/\/index$/, ''); // remove trailing /index

  if (!isRelativePath(localPath)) {
    localPath = `./${localPath}`; // insert `./` to make it a relative path
  }

  return localPath;
}

function stripExtension(modulePath, stripExtensions) {
  let name = path.basename(modulePath);

  stripExtensions.some((extension) => {
    if (name.endsWith(extension)) {
      name = name.slice(0, name.length - extension.length);

      return true;
    }

    return false;
  });

  return name;
}

function replaceExtension(modulePath, opts) {
  const filename = stripExtension(modulePath, opts.stripExtensions);

  return path.join(path.dirname(modulePath), filename);
}

function matchesPattern(types, calleePath, pattern) {
  const { node } = calleePath;

  if (types.isMemberExpression(node)) {
    return calleePath.matchesPattern(pattern);
  }

  if (!types.isIdentifier(node) || pattern.includes('.')) {
    return false;
  }

  const name = pattern.split('.')[0];

  return node.name === name;
}

function isImportCall(types, calleePath) {
  return types.isImport(calleePath.node.callee);
}

function escapeRegExp(string) {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

const toPath = (urlOrPath) =>
  urlOrPath instanceof URL ? fileURLToPath(urlOrPath) : urlOrPath;

const findPackageJson = ({ cwd = process.cwd() } = {}) => {
  const name = 'package.json';

  let directory = path.resolve(toPath(cwd) ?? '');

  const { root } = path.parse(directory);

  const stopAt = path.resolve(root);

  while (directory && directory !== stopAt && directory !== root) {
    const filePath = path.isAbsolute(name) ? name : path.join(directory, name);

    try {
      const stats = fs.statSync(filePath, { throwIfNoEntry: false });

      if (stats?.isFile()) {
        return filePath;
      }
    } catch {
      // eslint-disable no-empty
    }

    directory = path.dirname(directory);
  }

  return null;
};

module.exports = {
  nodeResolvePath,
  isRelativePath,
  toPosixPath,
  toLocalPath,
  stripExtension,
  replaceExtension,
  matchesPattern,
  isImportCall,
  escapeRegExp,
  findPackageJson,
};
