// @andes/form/build/cjs/index.production.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _extends = require('@babel/runtime/helpers/extends');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var React = require('react');
var context = require('@andes/context');
var warning = require('@andes/common/build/utils/warning');
var classNames = require('classnames');
var includes = require('@andes/common/build/utils/includes');
var badge = require('@andes/badge');
var themes = require('@andes/themes');
var computeDataAndesState = require('@andes/common/build/utils/computeDataAndesState');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _objectWithoutProperties__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutProperties);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var includes__default = /*#__PURE__*/_interopDefaultLegacy(includes);
const WARNING_CSRF = `To prevent CSRF attacks, <Form> needs a csrf token. Please, provide one in <AndesProvider> \
via <csrfToken> prop.
Refer to https://github.com/melisource/fury_frontend-andes-ui/tree/master/packages/form#csrf-protection-by-default \
for more details.`;
const _excluded$1 = ["children", "ref"];
const Form = /*#__PURE__*/function () {
  const Form = _ref => {
    let children = _ref.children,
      ref = _ref.ref,
      props = _objectWithoutProperties__default["default"](_ref, _excluded$1);
    const _useAndesContext = context.useAndesContext(),
      csrfToken = _useAndesContext.csrfToken;
    const _useState = React.useState(null),
      _useState2 = _slicedToArray__default["default"](_useState, 2),
      metaCsrfToken = _useState2[0],
      setMetaCsrfToken = _useState2[1];
    const inputValue = csrfToken || metaCsrfToken;
    React.useEffect(() => {
      if (!csrfToken) {
        const metaTag = document.querySelector('meta[name=csrf-token]');
        if (metaTag) {
          setMetaCsrfToken(metaTag.getAttribute('content'));
        }
      }
    }, [csrfToken]);
    warning__default["default"](csrfToken, WARNING_CSRF);
    return /*#__PURE__*/React__default["default"].createElement("form", _extends__default["default"]({}, props, {
      ref: ref
    }), inputValue && /*#__PURE__*/React__default["default"].createElement("input", {
      type: "hidden",
      name: "_csrf",
      value: inputValue
    }), children);
  };
  Form.displayName = 'Form';
  return Form;
}();
var Form$1 = Form;
const namespace = 'andes-helper';
const MODIFIER_CAUTION = 'caution';
const MODIFIER_ERROR = 'error';
const COLOR_CAUTION = 'caution';
const COLOR_NEGATIVE = 'negative';
const MODIFIER = [MODIFIER_CAUTION, MODIFIER_ERROR];
const COLOR = {
  [`${MODIFIER_CAUTION}`]: COLOR_CAUTION,
  [`${MODIFIER_ERROR}`]: COLOR_NEGATIVE
};
var enTranslations = {
  "andes-helper\u0004Precaución": [null, "Warning"],
  "andes-helper\u0004Error": [null, "Error"]
};
var esArTranslations = {
  "andes-helper\u0004Precaución": [null, "Precaución"],
  "andes-helper\u0004Error": [null, "Error"]
};
var ptBrTranslations = {
  "andes-helper\u0004Precaución": [null, "Alerta"],
  "andes-helper\u0004Error": [null, "Erro"]
};
const _excluded = ["andesDataAttributes", "children", "className", "id", "modifier"];
const Helper = /*#__PURE__*/function () {
  const Helper = props => {
    const _props$andesDataAttri = props.andesDataAttributes,
      andesDataAttributes = _props$andesDataAttri === void 0 ? {} : _props$andesDataAttri,
      children = props.children,
      className = props.className,
      id = props.id,
      modifier = props.modifier,
      otherProps = _objectWithoutProperties__default["default"](props, _excluded);
    const componentId = context.useId(id);
    const _useThemeContext = themes.useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const _useAndesContext = context.useAndesContext('@andes/form', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations
      }),
      i18n = _useAndesContext.i18n;
    const showBadgeIcon = includes__default["default"](MODIFIER, modifier || '');
    const helperClassName = classNames__default["default"](namespace, className, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('helper'), {
      [`${namespace}--${modifier}`]: modifier
    });
    return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
      className: helperClassName,
      id: componentId
    }, otherProps, themeHash, andesDataAttributes, {
      "data-andes-helper": true
    }, computeDataAndesState.computeDataAndesState({
      error: modifier === 'error',
      caution: modifier === 'caution'
    })), showBadgeIcon && /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${namespace}__badge`,
      "data-andes-helper-badge": true
    }, /*#__PURE__*/React__default["default"].createElement(badge.PillIcon, {
      color: modifier ? COLOR[modifier] : undefined,
      size: "xsmall",
      srLabel: i18n.pgettext(namespace, modifier === 'error' ? 'Error' : 'Precaución')
    })), /*#__PURE__*/React__default["default"].createElement("span", {
      className: `${namespace}__label`,
      "data-andes-helper-label": true
    }, children));
  };
  Helper.displayName = 'Helper';
  return Helper;
}();
var Helper$1 = Helper;
exports.Form = Form$1;
exports.Helper = Helper$1;
