// @nordic/core-context/dist/cjs/getters-D0aaOTRR.js
'use strict';

var node_async_hooks = require('node:async_hooks');
var validation = require('@nordic/utils/validation');
var error = require('@nordic/utils/error');

class NordicCoreContextError extends error.OwnerStackError {
    name = 'NordicCoreContextError';
}
function throwContextNotInitializedError() {
    throw new NordicCoreContextError('Nordic Core Context is not initialized.');
}

function getTestContextScope() {
    return global._n.ctx;
}

const storage = new node_async_hooks.AsyncLocalStorage();
function getContext() {
    const context = storage.getStore();
    if (validation.isTestEnv()) {
        return getTestContextScope().c;
    }
    if (!context) {
        throwContextNotInitializedError();
    }
    return context;
}

function getDeviceType() {
    return getContext().deviceType;
}
function getPlatformType() {
    return getContext().platformType;
}
function getWebviewInfo() {
    return getContext().webviewInfo;
}
function getDeviceInfo() {
    return getContext().deviceInfo;
}
function getSafeAreaInsets() {
    return getContext().safeAreaInsets;
}
function getI18n() {
    return getContext().i18n;
}
function getNetworkState() {
    return getContext().networkState;
}
function getPlatform() {
    return getContext().platform;
}

exports.getContext = getContext;
exports.getDeviceInfo = getDeviceInfo;
exports.getDeviceType = getDeviceType;
exports.getI18n = getI18n;
exports.getNetworkState = getNetworkState;
exports.getPlatform = getPlatform;
exports.getPlatformType = getPlatformType;
exports.getSafeAreaInsets = getSafeAreaInsets;
exports.getTestContextScope = getTestContextScope;
exports.getWebviewInfo = getWebviewInfo;
exports.storage = storage;
