// @nordic/core-context/dist/cjs/getters-fsHJdWAH.js
'use strict';

var node_async_hooks = require('node:async_hooks');

const storage = new node_async_hooks.AsyncLocalStorage();
function getContext() {
    const context = storage.getStore();
    {
        return context;
    }
}

function getDeviceType() {
    return getContext().deviceType;
}
function getPlatformType() {
    return getContext().platformType;
}
function getWebviewInfo() {
    return getContext().webviewInfo;
}
function getDeviceInfo() {
    return getContext().deviceInfo;
}
function getSafeAreaInsets() {
    return getContext().safeAreaInsets;
}
function getI18n() {
    return getContext().i18n;
}
function getNetworkState() {
    return getContext().networkState;
}
function getPlatform() {
    return getContext().platform;
}

exports.getContext = getContext;
exports.getDeviceInfo = getDeviceInfo;
exports.getDeviceType = getDeviceType;
exports.getI18n = getI18n;
exports.getNetworkState = getNetworkState;
exports.getPlatform = getPlatform;
exports.getPlatformType = getPlatformType;
exports.getSafeAreaInsets = getSafeAreaInsets;
exports.getWebviewInfo = getWebviewInfo;
exports.storage = storage;
