// @nordic/core-context/dist/cjs/index.browser.production.js
'use strict';

function getContext() {
  const context = window._n.ctx.c;
  return context;
}

function getDeviceType() {
  return getContext().deviceType;
}
function getPlatformType() {
  return getContext().platformType;
}
function getWebviewInfo() {
  return getContext().webviewInfo;
}
function getDeviceInfo() {
  return getContext().deviceInfo;
}
function getSafeAreaInsets() {
  return getContext().safeAreaInsets;
}
function getI18n() {
  return getContext().i18n;
}
function getNetworkState() {
  return getContext().networkState;
}
function getPlatform() {
  return getContext().platform;
}

function setI18n(i18n) {
  getContext().i18n = i18n;
}
function setPlatform(platform) {
  getContext().platform = platform;
}

function getTestContextScope() {
  return global._n.ctx;
}

exports.getDeviceInfo = getDeviceInfo;
exports.getDeviceType = getDeviceType;
exports.getI18n = getI18n;
exports.getNetworkState = getNetworkState;
exports.getPlatform = getPlatform;
exports.getPlatformType = getPlatformType;
exports.getSafeAreaInsets = getSafeAreaInsets;
exports.getTestContextScope = getTestContextScope;
exports.getWebviewInfo = getWebviewInfo;
exports.setI18n = setI18n;
exports.setPlatform = setPlatform;
