// @andes/image/build/esm/index.development.js
import React from 'react';
import PropTypes from 'prop-types';
const WARNING_ALT_ATTRIBUTE = "title attribute is used to indicate the image's alt attribute, it's required for accessibility purposes";

/** @type {typeof import('../..').Image} */
const Image = _ref => {
  let className = _ref.className,
    width = _ref.width,
    height = _ref.height,
    src = _ref.src,
    src2x = _ref.src2x,
    title = _ref.title;
  return (/*#__PURE__*/
    // eslint-disable-next-line nordic/no-img-tags
    React.createElement("img", {
      className: className,
      width: width,
      height: height,
      src: src,
      alt: title,
      srcSet: src2x && `${src} 1x, ${src2x} 2x`
    })
  );
};

/** @type {import('prop-types').Validator<string>} */
const checkTitleProp = (_ref2, propName, componentName) => {
  let title = _ref2.title;
  PropTypes.checkPropTypes({
    title: PropTypes.string
  }, {
    title
  }, propName, componentName);
  return typeof title === 'undefined' ? new Error(WARNING_ALT_ATTRIBUTE) : null;
};
Image.propTypes = {
  title: checkTitleProp,
  src: PropTypes.string,
  src2x: PropTypes.string,
  width: PropTypes.string,
  height: PropTypes.string,
  className: PropTypes.string
};
var Image$1 = Image;
export { Image$1 as Image };
