// @andes/listbox/build/esm/index.development.js
import _extends from '@babel/runtime/helpers/extends';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { Children, useMemo, useRef, isValidElement, cloneElement, useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useOption, useListBoxSection, useListBox } from '@react-aria/listbox';
import { useSingleSelectListState, useListState } from '@react-stately/list';
import { mergeProps } from '@react-aria/utils';
import { ListGroup, ListBaseItem, ListItemContent, List } from '@andes/list';
import itemNodeGenerator from '@andes/list/build/generators/itemNodeGenerator';
import includes from '@andes/common/build/utils/includes';
import { hasTouchSupport } from '@andes/common/build/utils/isMobile';
import classNames from 'classnames';
import warning from '@andes/common/build/utils/warning';
import { hasImage } from '@andes/list/build/utils/hasImage';
import { hasSlot } from '@andes/list/build/utils/hasSlot';
import { isLegacyLeftContent } from '@andes/list/build/utils/isLegacyLeftContent';
import { ENTER, SPACE } from '@andes/common/build/constants/keyboardCodes';
import { useListContext } from '@andes/list/build/context/ListContext';
import { getAttributesByListType } from '@andes/list/build/utils/getAttributesByListType';

/**
 * @typedef {Object} ListOptionProps
 * @prop {Record<string, unknown>} Props
 */

/**
 * @type {import('react').FunctionComponent<ListOptionProps>
 * & { getCollectionNode: (props:ListOptionProps) => Generator } }
 */
const ListOption = () => null;
ListOption.displayName = 'ListOption';
ListOption.getCollectionNode = itemNodeGenerator;
var ListOption$1 = ListOption;
const SUPPORTED_ITEM_TYPES = ['ListItem', 'DropdownItem'];
const SUPPORTED_GROUP_TYPES = ['ListGroup', 'DropdownGroup'];

/**
 * @param {import('react').ReactElement} child
 * @param {string[]} [types]
 * @return {boolean}
 */
const isSupportedType = function (child) {
  var _child$type, _child$type2, _child$type3, _child$type4;
  let types = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  if (!child) return false;
  return (
    // @ts-expect-error TS doesn't recognize that displayName is a prop of type
    includes(types, ((_child$type = child.type) === null || _child$type === void 0 ? void 0 : _child$type.displayName) || ((_child$type2 = child.type) === null || _child$type2 === void 0 ? void 0 : _child$type2.name))
    // @ts-expect-error TS doesn't recognize that displayName is a prop of type
    || includes(types, ((_child$type3 = child.type) === null || _child$type3 === void 0 || (_child$type3 = _child$type3.wraps) === null || _child$type3 === void 0 ? void 0 : _child$type3.displayName) || ((_child$type4 = child.type) === null || _child$type4 === void 0 ? void 0 : _child$type4.name))
  );
};
var isSupportedType$1 = isSupportedType;

/** @param {import('react').ReactElement} child */
const isSupportedGroupType = child => isSupportedType$1(child, SUPPORTED_GROUP_TYPES);
var isSupportedGroupType$1 = isSupportedGroupType;

/** @param { import('react').ReactElement } child */
const isSupportedItemType = child => isSupportedType$1(child, SUPPORTED_ITEM_TYPES);
var isSupportedItemType$1 = isSupportedItemType;

/** @param {React.ReactElement | React.ReactElement[]} children */
const useChildrenAdapter = children => Children.map(children, child => {
  if (isSupportedItemType$1(child)) return /*#__PURE__*/React.createElement(ListOption$1, child.props);
  if (isSupportedGroupType$1(child)) {
    const groupChildren = Children.map(child.props.children, item => {
      if (isSupportedItemType$1(item)) return /*#__PURE__*/React.createElement(ListOption$1, item.props);
      return item;
    });
    return /*#__PURE__*/React.createElement(ListGroup, child.props, groupChildren);
  }
  return child;
});
var useChildrenAdapter$1 = useChildrenAdapter;
const WARNING_LIST_ITEM_CHILDREN = 'Children can\'t be used in conjunction with props title, description, leftContent and rightContent, in future versions children will have precedence over these properties';
const WARNING_LIST_ITEM_LEFT_CONTENT = 'The current object format \'{ title: string, src: string, alt: string, src2x: string }\' will be deprecated from the \'leftContent\' options. Please use a \'node\' instead.';
const _excluded$2 = ["item", "state"],
  _excluded2$1 = ["onClick", "onKeyDown"],
  _excluded3 = ["onSelect"];

/**
 * @typedef {Object} ListItemInternalProps
 * @prop {number} [index]
 * @prop {number} [groupIndex]
 * @prop {(
 *  event: React.KeyboardEvent<HTMLLIElement> | React.MouseEvent<HTMLLIElement>,
 *  selectedValue: { value: string }
 * ) => void} [onSelect]
 */

/**
 * @typedef {Object} ListBoxState
 * @prop {import('@react-stately/list').ListState<object>
 * | import('@react-stately/list').SingleSelectListState<object>} state
 */

/**
 * @typedef {Object} ListBoxItemChildrenItem
 * @prop {import('../../../list/lib/components/ListItem').ListItemProps & { value: string }} props
 * @prop {number | string} key
 * @prop {string | null} [aria-label]
 */

/**
 * @typedef {Object} ListBoxItemProps
 * @prop {ListBoxItemChildrenItem} item
 * @prop {React.ReactNode} [selectionComponent]
 * @prop {'large' | 'medium' | 'small'} [size]
 * @prop {(
 *  event: React.MouseEvent<HTMLLIElement>,
 *  itemValue?: string
 * ) => void} [onClick]
 * @prop {(
 *  event: React.KeyboardEvent<HTMLLIElement> | React.MouseEvent<HTMLLIElement>,
 *  selectedValue: { value: string }
 * ) => void} [onSelect]
 */

/** @type {React.FunctionComponent<ListBoxItemProps & ListBoxState>} */
const ListBoxItem = _ref => {
  let item = _ref.item,
    state = _ref.state,
    otherProps = _objectWithoutProperties(_ref, _excluded$2);
  const key = item.key,
    listItemProps = item.props;
  const _useListContext = useListContext(),
    listType = _useListContext.listType;
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const _useMemo = useMemo(() => getAttributesByListType(listType)(), []),
    selection = _useMemo.selection;
  const itemSelectionRef = useRef(null);
  const _useOption = useOption({
      key
    }, state, itemSelectionRef),
    optionProps = _useOption.optionProps,
    selected = _useOption.isSelected;
  const inheritedProps = _objectSpread(_objectSpread({}, listItemProps), {}, {
    key,
    selected
  });
  const itemClassName = classNames(classNames, {
    'andes-list__item-with-selection-component': !!otherProps.selectionComponent
  });
  const onClick = optionProps.onClick,
    onKeyDown = optionProps.onKeyDown,
    otherOptionProps = _objectWithoutProperties(optionProps, _excluded2$1);
  const onSelect = otherProps.onSelect,
    otherOtherProps = _objectWithoutProperties(otherProps, _excluded3);

  /** @param {React.MouseEvent<HTMLLIElement>} e */
  const onClickHandler = e => {
    const disabled = inheritedProps.disabled,
      value = inheritedProps.value,
      inheritedOnSelect = inheritedProps.onSelect;
    const inheritedOnClick = otherProps.onClick;
    if (!disabled) {
      if (typeof inheritedOnClick === 'function') inheritedOnClick(e, value);
      if (typeof onClick === 'function') onClick(e);
      if (typeof inheritedOnSelect === 'function') inheritedOnSelect(e, {
        value
      });
      if (typeof onSelect === 'function') onSelect(e, {
        value
      });
    }
  };

  /** @type {React.KeyboardEventHandler<HTMLLIElement>} */
  const handleOnKeyDown = event => {
    if (typeof onKeyDown === 'function') onKeyDown(event);
    if (event.code === ENTER || event.code === SPACE) {
      // @ts-expect-error We are assigning a KeyboardEvent to a MouseEvent
      onClickHandler(event);
    }
  };
  const getCheckbox = () => {
    const Check = otherProps.selectionComponent;
    return /*#__PURE__*/isValidElement(Check) ? /*#__PURE__*/React.createElement("div", _extends({}, selection, {
      className: "andes-list__item-selection-control andes-listbox__item-selection-control"
    }), /*#__PURE__*/cloneElement(Check, {
      // @ts-expect-error: Themes integration. Check typing.
      checked: inheritedProps.selected,
      disabled: inheritedProps.disabled
    })) : null;
  };
  const itemProps = _objectSpread({}, mergeProps(inheritedProps, otherOtherProps));
  const children = itemProps.children,
    description = itemProps.description,
    disabled = itemProps.disabled,
    id = itemProps.id,
    leftContent = itemProps.leftContent,
    maxLines = itemProps.maxLines,
    paddingSize = itemProps.paddingSize,
    rightContent = itemProps.rightContent,
    rightContentPosition = itemProps.rightContentPosition,
    size = itemProps.size,
    title = itemProps.title;
  const hasDescription = !!description;
  const hasLeftContentImage = hasImage(leftContent);
  const isSlotted = hasSlot(children, {
    title,
    description,
    rightContent
  }, {
    legacyBehavior: true
  });
  const itemCustomEventsHandlers = {
    onClick: onClickHandler,
    onKeyDown: handleOnKeyDown
  };

  // @deprecate Warning from: 9.0.0
  warning(!(children && (title || description || rightContent)), WARNING_LIST_ITEM_CHILDREN);

  // @deprecate [leftContent] - Warning from: 9.0.0
  const legacyLeftContent = isLegacyLeftContent(leftContent);
  warning(!legacyLeftContent, WARNING_LIST_ITEM_LEFT_CONTENT);
  return /*#__PURE__*/React.createElement(ListBaseItem, _extends({
    className: itemClassName,
    id: id,
    ref: itemSelectionRef,
    disabled: disabled,
    selected: selected,
    hasDescription: hasDescription,
    hasLeftContentImage: hasLeftContentImage,
    paddingSize: paddingSize,
    size: size,
    slotted: isSlotted
  }, otherOptionProps, itemCustomEventsHandlers), getCheckbox(), /*#__PURE__*/React.createElement(ListItemContent, _extends({}, itemProps, {
    key: key,
    title: title,
    description: description,
    leftContent: leftContent,
    rightContent: rightContent,
    rightContentPosition: rightContentPosition,
    maxLines: maxLines,
    size: size,
    id: id || (otherOptionProps === null || otherOptionProps === void 0 ? void 0 : otherOptionProps.id)
    // @ts-expect-error: internal props
    ,

    legacyBehavior: true
  }), children));
};
ListBoxItem.propTypes = {
  item: PropTypes.shape({
    'aria-label': PropTypes.string,
    props: PropTypes.shape({
      value: PropTypes.string.isRequired,
      children: PropTypes.node,
      leftContent: PropTypes.node
    }).isRequired,
    index: PropTypes.number.isRequired,
    key: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
    type: PropTypes.string,
    value: PropTypes.string
  }).isRequired,
  selectionComponent: PropTypes.element,
  // This was added because state is produced by a lib,
  // so the props can be different depending on the hook that is invoked.
  state: PropTypes.shape({
    collection: PropTypes.shape({}).isRequired,
    disabledKeys: PropTypes.shape({}).isRequired,
    selectionManager: PropTypes.shape({}).isRequired,
    selectedKey: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    setSelectedKey: PropTypes.func,
    selectedItem: PropTypes.shape({})
  }).isRequired,
  side: PropTypes.oneOf(['left', 'right']),
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  onClick: PropTypes.func,
  onSelect: PropTypes.func
};
var ListBoxItem$1 = ListBoxItem;
const _excluded$1 = ["group", "state", "listId"];

/**
 * @typedef {Object} ListBoxGroupProp
 * @prop {string} [`aria-label`]
 * @prop {Object} [childNodes]
 * @prop {React.ReactNode} [rendered]
 * @prop {number} [index]
 * @prop {string} [textValue]
 * @prop {string} [type]
 * @prop {Omit<import('../../../list/lib/components/ListItem').ListItemProps, 'size' | 'component' | 'id'> &
 *  { children?: React.ReactElement | React.ReactElement[] } } [props]
 * @prop {object | string} [value]
 */

/**
 * @typedef {Object} ListGroupProps
 * @prop {{ role?: string }} [itemProps]
 * @prop {{ id?: string, 'aria-hidden'?: boolean }} [headingProps]
 * @prop {{ role?: string }} [groupProps]
 * @prop {React.ElementType} [component]
 * @prop {ListBoxGroupProp} group
 * @prop {string} listId
 * @prop {(
 *  event: React.MouseEvent<HTMLLIElement>
 * ) => void} [onClick]
 * @prop {(
 *  event: React.KeyboardEvent<HTMLLIElement> | React.MouseEvent<HTMLLIElement>,
 *  selectedValue: { value: string }
 * ) => void} [onSelect]
 */

/** @type {React.FunctionComponent<ListGroupProps & import('./ListBoxItem').ListBoxState>} */
const ListBoxGroup = _ref => {
  let group = _ref.group,
    state = _ref.state,
    listId = _ref.listId,
    otherProps = _objectWithoutProperties(_ref, _excluded$1);
  const heading = group.rendered,
    ariaLabel = group['aria-label'],
    listGroupProps = group.props,
    index = group.index;
  const _useListBoxSection = useListBoxSection({
      heading,
      'aria-label': ariaLabel
    }),
    itemProps = _useListBoxSection.itemProps,
    headingProps = _useListBoxSection.headingProps,
    groupProps = _useListBoxSection.groupProps;

  // override react-aria unique/random generated ids
  headingProps.id = `${listId}__group-heading-${index}`;
  groupProps['aria-labelledby'] = headingProps.id;
  return /*#__PURE__*/React.createElement(ListGroup, _extends({
    label: "",
    itemProps: itemProps,
    headingProps: headingProps,
    groupProps: groupProps
  }, listGroupProps),
  // @ts-expect-error
  [...group.childNodes].map(groupNode => /*#__PURE__*/React.createElement(ListBoxItem$1, _extends({
    key: groupNode.key,
    item: groupNode,
    state: state
  }, otherProps))));
};
ListBoxGroup.propTypes = {
  // group could be null according to how prop-types handles optional params
  group: PropTypes.shape({
    'aria-label': PropTypes.string.isRequired,
    childNodes: PropTypes.shape({
      'aria-label': PropTypes.string,
      props: PropTypes.shape({
        value: PropTypes.string
      }),
      index: PropTypes.number,
      key: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      type: PropTypes.string,
      value: PropTypes.string
    }),
    rendered: PropTypes.string,
    index: PropTypes.number,
    textValue: PropTypes.string,
    type: PropTypes.string,
    props: PropTypes.shape({
      children: PropTypes.arrayOf(PropTypes.element.isRequired),
      component: PropTypes.string,
      id: PropTypes.string,
      label: PropTypes.string,
      size: PropTypes.oneOf(['small', 'medium', 'large'])
    }).isRequired
  }).isRequired,
  // This was added because state is produced by a lib,
  // so the props can be different depending on the hook that is invoked.
  state: PropTypes.shape({
    collection: PropTypes.shape({}).isRequired,
    disabledKeys: PropTypes.shape({}).isRequired,
    selectionManager: PropTypes.shape({}).isRequired,
    selectedKey: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    setSelectedKey: PropTypes.func,
    selectedItem: PropTypes.shape({})
  }).isRequired,
  listId: PropTypes.string.isRequired
};
var ListBoxGroup$1 = ListBoxGroup;

/**
 * @param {Array<React.ReactElement>} children
 * @param {string} itemKey
 * @return {React.ReactElement | null}
 */
const findItemInChildren = (children, itemKey) => {
  let findItem;
  children.forEach(child => {
    var _child$props, _child$props2;
    // itemNodeGenerator uses value as the key (see more: https://github.com/melisource/fury_frontend-andes-ui/blob/master/packages/list/lib/generators/itemNodeGenerator.js#L29)
    if (isSupportedType$1(child, ['ListOption', 'FloatingMenuItem']) && ((child === null || child === void 0 || (_child$props = child.props) === null || _child$props === void 0 ? void 0 : _child$props.value) === itemKey || (child === null || child === void 0 ? void 0 : child.key) === itemKey)) {
      findItem = child;
      return;
    }
    if (isSupportedType$1(child, ['ListGroup']) && child !== null && child !== void 0 && (_child$props2 = child.props) !== null && _child$props2 !== void 0 && _child$props2.children) {
      const foundItem = findItemInChildren(child.props.children, itemKey);
      if (foundItem) {
        findItem = foundItem;
      }
    }
  });
  return findItem || null;
};

/** @param {HTMLLIElement} element */
const createMouseEventForElement = element => {
  const target = element ? Object.assign(element, {
    value: undefined
  }) : {};
  const clickEvent = new MouseEvent('click', {
    bubbles: true,
    cancelable: true,
    view: window
  });
  Object.defineProperty(clickEvent, 'target', {
    get: () => target
  });
  return clickEvent;
};
const _excluded = ["ref"],
  _excluded2 = ["children", "id", "selectionComponent", "selectionMode", "isRadioList", "defaultSelectedKeys", "selectedKeys", "onClick", "onSelect", "shouldUseVirtualFocus", "srLabel"];

/**
 * @typedef {React.RefObject<{
 * event?: React.MouseEvent<HTMLLIElement> | React.KeyboardEvent<HTMLLIElement>,
 * values?: string[],
 * itemDetails?: ItemDetail}>
 * } SelectedManagerRef
 */

/**
 * @typedef {Object} ItemDetail
 * @prop {string} value
 */

/** @typedef {import('./ListBoxItem').ListBoxItemChildrenItem} ListBoxItemProp */

/** @typedef {import('./ListBoxGroup').ListBoxGroupProp} ListBoxGroupProp */

/** @type {typeof import ('../..').ListBox} */
const ListBox = _ref => {
  let ref = _ref.ref,
    props = _objectWithoutProperties(_ref, _excluded);
  const children = props.children,
    id = props.id,
    selectionComponent = props.selectionComponent,
    _props$selectionMode = props.selectionMode,
    selectionMode = _props$selectionMode === void 0 ? 'single' : _props$selectionMode,
    _props$isRadioList = props.isRadioList,
    isRadioList = _props$isRadioList === void 0 ? false : _props$isRadioList,
    defaultSelectedKeys = props.defaultSelectedKeys,
    selectedKeys = props.selectedKeys,
    onClick = props.onClick,
    onSelect = props.onSelect,
    shouldUseVirtualFocus = props.shouldUseVirtualFocus,
    srLabel = props.srLabel,
    other = _objectWithoutProperties(props, _excluded2);

  /** @type {React.RefObject<HTMLUListElement | null>} */
  const listSelectionRef = useRef(null);

  /** @type {SelectedManagerRef} */
  const selectedManager = useRef({});

  /** @type {React.MutableRefObject<Array<string>>} */
  const prevState = useRef(selectedKeys || defaultSelectedKeys || []);

  /**
   * Tracks the actual pointer type of the last interaction ('mouse', 'touch', 'pen').
   * Using navigator.maxTouchPoints to detect touch is incorrect — it reflects hardware
   * capability (static OS property), not the current input method. On Windows 10/11
   * machines with touchscreens, maxTouchPoints > 0 even when using a mouse, causing
   * mouse clicks to be treated as touch events and breaking selection in Dropdown.
   * @type {React.MutableRefObject<string>}
   */
  const lastPointerType = useRef('mouse');
  const handlePointerDown = useCallback((/** @type {PointerEvent} */e) => {
    lastPointerType.current = e.pointerType || 'mouse';
  }, []);
  useEffect(() => {
    const resolvedRef = typeof ref === 'function' ? listSelectionRef : ref || listSelectionRef;
    const el = resolvedRef.current;
    if (!el) return undefined;
    el.addEventListener('pointerdown', handlePointerDown);
    return () => el.removeEventListener('pointerdown', handlePointerDown);
  }, [handlePointerDown, ref]);
  const callOnSelectCallback = () => {
    const _selectedManager$curr = selectedManager.current,
      event = _selectedManager$curr.event,
      values = _selectedManager$curr.values,
      itemDetails = _selectedManager$curr.itemDetails;
    if (event && values && itemDetails) {
      if (typeof onSelect === 'function') onSelect(event, values, itemDetails);
      selectedManager.current = {};
    }
  };

  /**
   * @param {React.MouseEvent<HTMLLIElement> | React.KeyboardEvent<HTMLLIElement>} event
   * @param {ItemDetail} itemDetails
   */
  const onItemSelect = (event, itemDetails) => {
    selectedManager.current.event = event;
    selectedManager.current.itemDetails = itemDetails;
    callOnSelectCallback();
  };
  const getIdsAndDisabledKeys = () => {
    const disabledKeys = /** @type {string[]} */[];
    const listItemsIds = /** @type {string[]} */[];
    Children.forEach(children, (/** @type {React.ReactElement} */child) => {
      if (isSupportedGroupType$1(child)) {
        Children.forEach(child.props.children, item => {
          var _item$props, _item$props2;
          listItemsIds.push(`${id}-option-${item.props.value}`);
          if ((_item$props = item.props) !== null && _item$props !== void 0 && _item$props.disabled || (_item$props2 = item.props) !== null && _item$props2 !== void 0 && _item$props2.disableSelectionComponent) disabledKeys.push(item.props.value);
        });
      }
      if (isSupportedItemType$1(child)) {
        var _child$props, _child$props2;
        if ((_child$props = child.props) !== null && _child$props !== void 0 && _child$props.disabled || (_child$props2 = child.props) !== null && _child$props2 !== void 0 && _child$props2.disableSelectionComponent) disabledKeys.push(child.props.value);
      }
    });
    return {
      disabledKeys,
      listItemsIds: listItemsIds.join(' ')
    };
  };

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const _useMemo = useMemo(() => getIdsAndDisabledKeys(), [children]),
    disabledKeys = _useMemo.disabledKeys,
    listItemsIds = _useMemo.listItemsIds;

  // Replaces ListItem with ListOption to prevent passing a generator and allows using forwardRef
  const adaptedChildren = useChildrenAdapter$1(children);

  /** @param {string | Set<string> | null} values */
  const handleSelectionChangeForTouchDevices = values => {
    var _ref2, _ref2$querySelector, _itemObject$props, _itemObject$props2;
    const prevKeys = /** @type {Array<string>} */prevState.current;
    const newKeys = values ? [...values] : [];
    let itemKey = '';
    if (selectionMode === 'single') {
      if (newKeys.length > 0) {
        prevState.current = newKeys;
        itemKey = newKeys[0];
      } else {
        var _prevKeys = _slicedToArray(prevKeys, 1);
        itemKey = _prevKeys[0];
      }
    } else {
      const addedKeys = newKeys.filter(key => !prevKeys.includes(key));
      const removeKeys = prevKeys.filter(key => !newKeys.includes(key));
      itemKey = addedKeys.length > 0 ? addedKeys[0] : removeKeys[0];
      prevState.current = newKeys;
    }
    const itemElement = /** @type {React.RefObject<HTMLUListElement>} */(_ref2 = ref || listSelectionRef) === null || _ref2 === void 0 || (_ref2 = _ref2.current) === null || _ref2 === void 0 || (_ref2$querySelector = _ref2.querySelector) === null || _ref2$querySelector === void 0 ? void 0 : _ref2$querySelector.call(_ref2, `li[data-key="${itemKey}"]`);
    if (!itemElement) return;
    const itemObject = findItemInChildren(adaptedChildren, itemKey);
    const itemValue = itemObject === null || itemObject === void 0 || (_itemObject$props = itemObject.props) === null || _itemObject$props === void 0 ? void 0 : _itemObject$props.value;
    const clickEvent = createMouseEventForElement(/** @type {HTMLLIElement} */itemElement);
    if (typeof onClick === 'function') {
      onClick(/** @type {React.MouseEvent<HTMLLIElement>} */ /** @type {unknown} */clickEvent, itemValue);
    }
    if (typeof onSelect === 'function') {
      onSelect(/** @type {React.MouseEvent<HTMLLIElement>} */ /** @type {unknown} */clickEvent, newKeys, {
        value: itemValue
      });
    }
    if (typeof (itemObject === null || itemObject === void 0 || (_itemObject$props2 = itemObject.props) === null || _itemObject$props2 === void 0 ? void 0 : _itemObject$props2.onSelect) === 'function') {
      itemObject.props.onSelect(clickEvent, {
        value: itemValue
      });
    }
  };

  /** @param {string | Set<string> | null} values */
  const onSelectionChanged = values => {
    if (hasTouchSupport() && lastPointerType.current === 'touch') {
      handleSelectionChangeForTouchDevices(values);
      return;
    }
    selectedManager.current.values = Array.from(values || []);
    callOnSelectCallback();
  };
  const listStateHook = isRadioList ? useSingleSelectListState : useListState;
  const state = listStateHook(_objectSpread(_objectSpread({
    disabledKeys,
    selectionMode
  }, props), {}, {
    onSelectionChange: (/** @type {(values: React.Key | Set<React.Key> | null) => void} */onSelectionChanged),
    children: adaptedChildren
  }));
  const _useListBox = useListBox(_objectSpread(_objectSpread({
      'aria-label': srLabel,
      disallowEmptySelection: true,
      disabledKeys,
      selectionMode
    }, props), {}, {
      shouldUseVirtualFocus
    }), state, /** @type {React.RefObject<HTMLElement>} */ref || listSelectionRef),
    listBoxProps = _useListBox.listBoxProps;
  return /*#__PURE__*/React.createElement(List, _extends({
    "aria-owns": listItemsIds.length ? listItemsIds : undefined,
    srLabel: srLabel,
    component: "ul"
    // @ts-expect-error: internal props
    ,

    listType: "listbox"
  }, mergeProps(other, listBoxProps), {
    ref: ref || listSelectionRef
    /* FIX-ME: when aria-label is removed from List. This is needed if not an error is thrown for using the prop,
    and useListBox needs this prop or else, throws a warning */,
    "aria-label": undefined
  }), [...state.collection].map(listSelectNode => {
    const commonProps = _objectSpread({
      key: listSelectNode.key,
      state,
      selectionComponent
    }, hasTouchSupport() && lastPointerType.current === 'touch' ? {} : {
      onClick,
      /**
       * @param {React.MouseEvent<HTMLLIElement> | React.KeyboardEvent<HTMLLIElement>} event
       * @param {ItemDetail} itemDetails
       */
      onSelect: (event, itemDetails) => {
        if (typeof onItemSelect === 'function') onItemSelect(event, itemDetails);
        // @ts-expect-error
        // listSelectNode is an object according to the definition of the lib's hook, TS doesn't recognizes onSelect
        if (typeof listSelectNode.onSelect === 'function') listSelectNode.onSelect(event, itemDetails);
      }
    });
    return listSelectNode.type === 'item' ? /*#__PURE__*/React.createElement(ListBoxItem$1, _extends({
      item: (/** @type {ListBoxItemProp} */listSelectNode)
    }, commonProps, {
      key: listSelectNode.key
    })) : /*#__PURE__*/React.createElement(ListBoxGroup$1, _extends({
      group: (/** @type {ListBoxGroupProp} */listSelectNode)
    }, commonProps, {
      listId: id,
      key: listSelectNode.key
    }));
  }));
};
ListBox.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element.isRequired)]).isRequired,
  defaultSelectedKeys: PropTypes.arrayOf(PropTypes.string.isRequired),
  id: PropTypes.string.isRequired,
  isRadioList: PropTypes.bool,
  onClick: PropTypes.func,
  onSelect: PropTypes.func,
  selectedKeys: PropTypes.arrayOf(PropTypes.string.isRequired),
  selectionComponent: PropTypes.element,
  selectionMode: PropTypes.oneOf(['single', 'multiple']),
  shouldUseVirtualFocus: PropTypes.bool,
  srLabel: PropTypes.string
};
ListBox.displayName = 'ListBox';
var ListBox$1 = ListBox;
export { ListBox$1 as ListBox };
