// @andes/common/build/cjs/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var React = require('react');
var PropTypes = require('prop-types');
var image = require('@andes/image');
var closeListener = require('@andes/close-listener');
var widther = require('@andes/widther');
var visuallyHidden = require('@andes/visually-hidden');
var exenv = require('exenv');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
const WARNING_PORTAL = 'The <Portal> component will be removed from this package in the next major release. Please use the "@andes/portal" package instead.';
const WARNING_VISUALLY_HIDDEN = 'The <VisuallyHidden> component will be deprecated from this package in the next major release. Please use the "@andes/visually-hidden" package instead.';
const WARNING_VISUALLY_HIDDEN_FOCUSABLE = 'The <VisuallyHiddenFocusable> component will be deprecated from this package in the next major release. Please use the "@andes/visually-hidden" package instead.';
const WARNING_CLOSE_LISTENER = 'The <CloseListener> component will be deprecated from this package in the next major release. Please use the "@andes/close-listener" package instead.';
const WARNING_WIDTHER = 'The <Widther> component will be deprecated from this package in the next major release. Please use the "@andes/widther" package instead.';
const WARNING_IMAGE = 'The <Image> component will be deprecate from this package in the next major release. Please use the "@andes/image" package instead.';

/* eslint-disable quotes */
/* eslint-disable no-console */

/**
 * @param {boolean} condition
 * @param {string} message
 * @param {string} [type]
 */
const print = function (condition, message) {
  let type = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'error';
  if (condition) {
    return;
  }
  if (typeof console !== 'undefined') {
    let logFunction = /** @type {function | undefined} */console[(/** @type {keyof typeof console} */type)];
    if (typeof logFunction !== 'function') logFunction = console.error;
    logFunction(message);
  }
  try {
    throw new Error(message);
    // eslint-disable-next-line no-empty
  } catch (x) {}
};
const warning = process.env.NODE_ENV !== "'production'" && process.env.NODE_ENV !== "production" ? print : () => {};
var warning$1 = warning;

/** @type {typeof import('../../..').Image} */
const ImageDeprecate = props => {
  warning$1(false, WARNING_IMAGE);
  return /*#__PURE__*/React__default["default"].createElement(image.Image, props);
};
ImageDeprecate.propTypes = {
  title: PropTypes__default["default"].string,
  src: PropTypes__default["default"].string,
  src2x: PropTypes__default["default"].string,
  width: PropTypes__default["default"].string,
  height: PropTypes__default["default"].string,
  className: PropTypes__default["default"].string
};
var ImageDeprecate$1 = ImageDeprecate;

/** @type {typeof import('../../..').CloseListener} */
const CloseListenerDeprecate = props => {
  warning$1(false, WARNING_CLOSE_LISTENER);
  return /*#__PURE__*/React__default["default"].createElement(closeListener.CloseListener, props);
};
CloseListenerDeprecate.propTypes = {
  children: PropTypes__default["default"].element.isRequired,
  onCloseOutside: PropTypes__default["default"].func.isRequired,
  handleClick: PropTypes__default["default"].bool,
  handleEsc: PropTypes__default["default"].bool
};
var CloseListenerDeprecate$1 = CloseListenerDeprecate;
const PortalDeprecate = () => {
  warning$1(false, WARNING_PORTAL);
  return null;
};
var PortalDeprecate$1 = PortalDeprecate;

/** @type {typeof import('../../..').Widther<any>} */
const WidtherDeprecate = WrappedComponent => {
  warning$1(false, WARNING_WIDTHER);
  return widther.Widther(WrappedComponent);
};
var WidtherDeprecate$1 = WidtherDeprecate;

/** @type {typeof import('../../..').VisuallyHidden} */
const VisuallyHiddenDeprecate = props => {
  warning$1(false, WARNING_VISUALLY_HIDDEN);
  return /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, props);
};
VisuallyHiddenDeprecate.propTypes = {
  children: PropTypes__default["default"].oneOfType([PropTypes__default["default"].node, PropTypes__default["default"].arrayOf(PropTypes__default["default"].node)]).isRequired,
  component: PropTypes__default["default"].oneOf(['div', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'span'])
};
var VisuallyHiddenDeprecate$1 = VisuallyHiddenDeprecate;

/** @type {typeof import('../../..').VisuallyHiddenFocusable} */
const VisuallyHiddenFocusableDeprecate = props => {
  warning$1(false, WARNING_VISUALLY_HIDDEN_FOCUSABLE);
  return /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHiddenFocusable, props);
};
VisuallyHiddenFocusableDeprecate.propTypes = {
  className: PropTypes__default["default"].string,
  children: PropTypes__default["default"].oneOfType([PropTypes__default["default"].node, PropTypes__default["default"].arrayOf(PropTypes__default["default"].node)]),
  component: PropTypes__default["default"].oneOf(['input', 'textarea', 'button', 'select', 'option', 'a', 'details', 'summary'])
};
var VisuallyHiddenFocusableDeprecate$1 = VisuallyHiddenFocusableDeprecate;
Object.defineProperty(exports, 'canUseDOM', {
  enumerable: true,
  get: function () {
    return exenv.canUseDOM;
  }
});
exports.CloseListener = CloseListenerDeprecate$1;
exports.Image = ImageDeprecate$1;
exports.Portal = PortalDeprecate$1;
exports.VisuallyHidden = VisuallyHiddenDeprecate$1;
exports.VisuallyHiddenFocusable = VisuallyHiddenFocusableDeprecate$1;
exports.Widther = WidtherDeprecate$1;
