// frontend-restclient/src/telemetry/OpenTelemetryClient.js
const OpenTelemetry = require('@opentelemetry/api');
const { SemanticAttributes } = require('@opentelemetry/semantic-conventions');
const { SpanKind } = require('@opentelemetry/api');
const pkg = require('../../package.json');

const { RETRY_COUNT_HEADER } = require('../constants/headers');
const {
  POOL_NAME,
  RETRY_STRATEGY,
  RETRY_POLICY,
  RETRIES,
  VERSION,
  RESTCLIENT_SPAN_NAME,
  LINEAR_BACKOFF,
  EXPONENTIAL_BACKOFF,
} = require('../constants/otel');

class OpenTelemetryClient {
  /**
   * Class constructor
   */
  constructor() {
    this.context = OpenTelemetry.context.active();
    this.spanKind = SpanKind.INTERNAL;
    this.span = {};
  }

  /**
   * Allows to create a span config used for start traces
   * @param {object} config - Config related to span creation
   * @param {string} config.method - The operations verb related to request
   * @param {string} config.url - The url related to request
   * @param {object} config.headers - Headers to be sent
   * @param {object} config.retry - Retry configs for a given request
   * @param {object} config.retry.policy - The retry policy to be used
   * @param {string} config.retry.policy.name - The name of the retry policy
   */
  spanConfig(config) {
    return {
      kind: this.spanKind,
      attributes: {
        [SemanticAttributes.HTTP_METHOD]: config.method,
        [SemanticAttributes.HTTP_URL]: config.url,
        [POOL_NAME]: config.headers?.['X-Rest-Pool-Name'],
        [RETRY_POLICY]: config.retry?.policy?.name,
        [VERSION]: pkg.version,
      },
    };
  }

  /**
   * Build active span for RestClient traces in async callbacks
   * @param {() => Promise} asyncCallback - The operations verb related to request
   * @param {object} config - Config related to span creation
   * @param {string} config.method - The operations verb related to request
   * @returns {Promise<{ response: * }>} - If is a response, will be returned as original definition
   */
  async asyncBuildActiveSpan(asyncCallback, config) {
    const asyncValues = {};
    const tracer = OpenTelemetry.trace.getTracer(pkg.name, pkg.version);

    await tracer.startActiveSpan(
      `${RESTCLIENT_SPAN_NAME} ${config.method}`,
      this.spanConfig(config),
      this.context,
      async (span) => {
        this.span = span;
        asyncValues.response = await asyncCallback();
      },
    );
    return asyncValues;
  }

  /**
   * Set span to error. Status code is not necessary cause are controlled by rest client
   */
  recordSpanError() {
    this.span.setStatus({ code: OpenTelemetry.SpanStatusCode.ERROR });
  }

  /**
   * Add more attributes to active span
   * @param {object} config - Config related to span creation
   * @param {object} response - Request response
   * @param {object} response.headers - Headers to be response
   * @param {object} response.status - Status to be response
   */
  addAttributesToSpan(config, response) {
    if (response?.request?.getHeader(RETRY_COUNT_HEADER) > 0) {
      this.span.setAttribute(RETRIES, response?.request?.getHeader(RETRY_COUNT_HEADER));
      this.span.setAttribute(RETRY_STRATEGY, this._getRetryStrategy(config));
    }
    this.span.setAttribute(SemanticAttributes.HTTP_STATUS_CODE, response?.status);
    this.span.setStatus({ code: OpenTelemetry.SpanStatusCode.OK });
  }

  /**
   * Build active span for RestClient traces in async callbacks
   * @param {object} config - Config related to span creation
   * @param {string} config.retry - Retry configs for a given request
   * @param {string} config.retry.factor - The type of retries factor to use (disabled by default).
   * @returns {string} - Return the retry strategy (LinearBackoff or ExponentialBackoff)
   */
  _getRetryStrategy(config) {
    if (config?.retry?.factor === 2) {
      return EXPONENTIAL_BACKOFF;
    }

    return LINEAR_BACKOFF;
  }

  /**
   * Finishes the span's propagation
   */
  endSpan() {
    this.span.end();
  }
}

module.exports = OpenTelemetryClient;
