// @andes/common/build/cjs/utils/propTypesValidators.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var PropTypes = require('prop-types');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
const i18NPropTypeChecker = PropTypes__default["default"].shape({
  pgettext: PropTypes__default["default"].func.isRequired,
  npgettext: PropTypes__default["default"].func.isRequired,
  gettext: PropTypes__default["default"].func.isRequired,
  ngettext: PropTypes__default["default"].func.isRequired,
  translations: PropTypes__default["default"].shape({}).isRequired,
  jsx: PropTypes__default["default"].shape({
    pgettext: PropTypes__default["default"].func.isRequired,
    npgettext: PropTypes__default["default"].func.isRequired,
    gettext: PropTypes__default["default"].func.isRequired,
    ngettext: PropTypes__default["default"].func.isRequired
  }).isRequired
});
exports.i18NPropTypeChecker = i18NPropTypeChecker;
