// @andes/common/build/cjs/utils/transitionHelpers.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
const PRE_ENTER = 0;
const ENTERING = 1;
const ENTERED = 2;
const PRE_EXIT = 3;
const EXITING = 4;
const EXITED = 5;
const UNMOUNTED = 6;
const STATUS = /** @type {const} */['preEnter', 'entering', 'entered', 'preExit', 'exiting', 'exited', 'unmounted'];

/** @typedef {PRE_ENTER | ENTERING | ENTERED | PRE_EXIT | EXITING | EXITED |  UNMOUNTED} Status */

/**
 * @typedef {Object} TimeoutObject
 * @property {number} [enter]
 * @property {number} [exit]
 */

/** @typedef {TimeoutObject | number } Timeout */

/**
 * @typedef {Object} State
 * @prop {number} _s
 * @prop {typeof STATUS[number]} status
 * @prop {boolean} isEnter
 * @prop {boolean} isMounted
 * @prop {boolean} isResolved
 */

/**
 * @param {import('../constants/transitionsStatuses').Status} status
 */
const getState = status => ({
  _s: status,
  status: STATUS[status],
  isEnter: status < PRE_EXIT,
  isMounted: status !== UNMOUNTED,
  isResolved: status === ENTERED || status > EXITING
});

/** @param {boolean} [unmounted] */
const startOrEnd = unmounted => unmounted ? UNMOUNTED : EXITED;

/**
 * @param {number} status
 * @param {boolean} [unmountOnExit]
 */
const getEndStatus = (status, unmountOnExit) => {
  switch (status) {
    case ENTERING:
    case PRE_ENTER:
      return ENTERED;
    case EXITING:
    case PRE_EXIT:
      return startOrEnd(unmountOnExit);
  }
};

/** @param {Timeout} [timeout] */
const getTimeout = timeout => typeof timeout === 'object' ? [timeout.enter, timeout.exit] : [timeout, timeout];

/**
 * @param {Function} transitState
 * @param {import('../constants/transitionsStatuses').Status} status
 */
const nextTick = (transitState, status) => setTimeout(() => {
  // Reading document.body.offsetTop can force browser to repaint before transition to the next state
  // eslint-disable-next-line no-unused-expressions
  Number.isNaN(document.body.offsetTop) || transitState(status + 1);
}, 0);
exports.getEndStatus = getEndStatus;
exports.getState = getState;
exports.getTimeout = getTimeout;
exports.nextTick = nextTick;
exports.startOrEnd = startOrEnd;
