// nordic-pages-discovery/lib/helpers/babel/component-info.js
const nodePath = require('path');

const traverse = require('@babel/traverse').default;
const babelTypes = require('@babel/types');

const { APP_FILE_MEMBERS } = require('../../constants/file');
const { NordicIslandsError } = require('../../error/nordic-islands-error');
const { astCache } = require('../../cache/ast-cache');
const { getScriptFullPathWithExtension } = require('../file/path-detail');
const {
  getLookingForComponentPathError,
  getIslandChildExportedWithCallExpressionError,
  getLookingForComponentImportError,
  getForwardRefIslandWithoutNameError,
} = require('../../error/get-discovery-error');
const {
  isObjectExport,
  isIdentifierDefaultExport,
  isObjectDefaultExport,
  isVariableNamedExport,
} = require('./export-statement');
const { isEnumDeclaration, isExportType } = require('./typescript');
const { IMPORT_TYPES } = require('../../constants/islands');
const { getFnInfoIfForwardRef } = require('../forward-ref');
const { getFileDirname } = require('../file/path-detail');
const {
  isPotentialDynamicImportDeclarationUsage,
  getPotentialDynamicImportDeclarationUsage,
} = require('../../islands/get-islands-from-file/imported-components/dynamic-import-declaration');

/**
 * @typedef {import('../../cache/ast-cache').BabelAST} BabelAST
 *
 * @typedef {Object} ComponentInfo
 * @property {string | null} name
 * @property {string | null} exportedName
 * @property {string | null} path
 * @property {boolean | null} namedExport
 * @property {boolean} shouldUseDisplayName
 * @property {import('../forward-ref').ForwardRefInfo} forwardRefInnerFn
 *
 * @typedef {Object} ComponentImportInfo
 * @property {boolean} shouldUseDisplayName
 * @property {boolean} namedExport
 * @property {import('../forward-ref').ForwardRefInfo} forwardRefInnerFn
 * @property {string?} name
 * @property {string?} path
 *
 * @typedef {Object} PartialComponentInfo
 * @property {string} name
 * @property {boolean} namedExport
 * @property {boolean} isIsland
 */

/**
 * @param { parser.ParseResult<babelTypes.File> } ast
 * @param { string } componentName
 * @param { string } filePath
 * @returns {ComponentImportInfo}
 */
const getComponentImportInfoFromAst = (ast, componentName, filePath) => {
  try {
    // TODO: Remove forwardRefInnerFn prop use when "forwardRef" support is removed
    const importInfo = {
      name: null,
      path: null,
      namedExport: null,
      shouldUseDisplayName: false,
      forwardRefInnerFn: { name: null, isLocal: false },
    };

    const originDir = getFileDirname(filePath);

    traverse(ast, {
      AssignmentExpression(path) {
        if (
          !babelTypes.isProgram(path.parentPath.parent) ||
          !babelTypes.isIdentifier(path.node.left) ||
          path.node.left.name !== componentName
        ) {
          return;
        }

        // TODO: Set shouldUseDisplayName based on isCallExpression result when "forwardRef" support is removed
        if (!babelTypes.isCallExpression(path.node.right)) {
          importInfo.shouldUseDisplayName = false;

          return;
        }

        importInfo.forwardRefInnerFn = getFnInfoIfForwardRef(
          ast,
          path.node.right,
        );
        importInfo.shouldUseDisplayName =
          !importInfo.forwardRefInnerFn.name &&
          importInfo.forwardRefInnerFn.isLocal;

        if (
          !isPotentialDynamicImportDeclarationUsage(
            path,
            path.node.left,
            path.node.right,
          )
        ) {
          return;
        }

        // Loadable cases
        // example: const MyComponent = loadable(() => import('./my-component'));
        const appComponentUsage = getPotentialDynamicImportDeclarationUsage(
          path,
          { file: filePath, dir: originDir },
        );

        if (appComponentUsage[componentName]) {
          importInfo.name = appComponentUsage[componentName].importName;
          importInfo.path = appComponentUsage[componentName].path;
          importInfo.namedExport = appComponentUsage[componentName].namedExport;
          path.stop();
        }
      },
      VariableDeclarator(path) {
        if (
          !(
            babelTypes.isProgram(path.parentPath.parent) ||
            babelTypes.isProgram(path.parentPath.parentPath.parent)
          ) ||
          !babelTypes.isIdentifier(path.node.id) ||
          path.node.id.name !== componentName
        ) {
          return;
        }

        // TODO: Set shouldUseDisplayName based on isCallExpression result when "forwardRef" support is removed
        if (!babelTypes.isCallExpression(path.node.init)) {
          importInfo.shouldUseDisplayName = false;

          return;
        }

        importInfo.forwardRefInnerFn = getFnInfoIfForwardRef(
          ast,
          path.node.init,
        );
        importInfo.shouldUseDisplayName =
          !importInfo.forwardRefInnerFn.name &&
          importInfo.forwardRefInnerFn.isLocal;

        if (
          !isPotentialDynamicImportDeclarationUsage(
            path,
            path.node.id,
            path.node.init,
          )
        ) {
          return;
        }

        // Loadable cases
        // example: const MyComponent = loadable(() => import('./my-component'));
        const appComponentUsage = getPotentialDynamicImportDeclarationUsage(
          path,
          { file: filePath, dir: originDir },
        );

        if (appComponentUsage[componentName]) {
          importInfo.name = appComponentUsage[componentName].importName;
          importInfo.path = appComponentUsage[componentName].path;
          importInfo.namedExport = appComponentUsage[componentName].namedExport;
          path.stop();
        }
      },
      ImportDeclaration(path) {
        path.node.specifiers?.forEach((specifier) => {
          if (specifier.local.name !== componentName) {
            return;
          }

          const name =
            babelTypes.isImportSpecifier(specifier) &&
            babelTypes.isIdentifier(specifier.imported)
              ? specifier.imported.name
              : specifier.local.name;

          importInfo.name = name;
          importInfo.path = getScriptFullPathWithExtension(
            path.node.source.value,
            nodePath.dirname(filePath),
          );
          importInfo.namedExport = babelTypes.isImportSpecifier(specifier);

          path.stop();
        });
      },
    });

    return importInfo;
  } catch (error) {
    throw new NordicIslandsError(
      getLookingForComponentImportError(componentName, error.message),
      {
        type: APP_FILE_MEMBERS.COMPONENT,
        path: filePath,
      },
      error,
    );
  }
};

/**
 * @param {ComponentInfo} componentInfo
 * @param {{ast: BabelAST, usedName: string, filePath: string, isIsland: boolean }} componentExtraInfo
 * @param {babel.NodePath<babelTypes.ExportDefaultDeclaration>} babelPath
 */
const setInfoOfComponentWrappedWithFn = (
  componentInfo,
  { ast, usedName, filePath, isIsland },
  babelPath,
) => {
  const { arguments: args = [] } = babelPath.node.declaration;

  // TODO: Remove forwardRefInnerFn prop use when "forwardRef" support is removed
  componentInfo.forwardRefInnerFn = getFnInfoIfForwardRef(
    ast,
    babelPath.node.declaration,
  );

  // TODO: Delete "!componentInfo.forwardRefInnerFn" condition when "forwardRef" support is removed
  if (isIsland && !componentInfo.forwardRefInnerFn.name) {
    throw new NordicIslandsError(
      componentInfo.forwardRefInnerFn.name === ''
        ? getForwardRefIslandWithoutNameError()
        : getIslandChildExportedWithCallExpressionError(),
      {
        type: APP_FILE_MEMBERS.COMPONENT,
        path: filePath,
      },
    );
  }

  // TODO: Delete isForwardRef use and condition when "forwardRef" support is removed
  const isForwardRef = typeof componentInfo.forwardRefInnerFn.name === 'string';

  if (!args.length || isForwardRef) {
    componentInfo.name = IMPORT_TYPES.DEFAULT;
    componentInfo.exportedName =
      isIsland && !isForwardRef ? null : IMPORT_TYPES.DEFAULT;
    componentInfo.namedExport = false;
    componentInfo.path = filePath;

    return;
  }

  args.forEach((arg) => {
    if (!babelTypes.isIdentifier(arg)) {
      return;
    }

    const importInfo = getComponentImportInfoFromAst(ast, arg.name, filePath);

    componentInfo.name = IMPORT_TYPES.DEFAULT;
    componentInfo.exportedName = isIsland ? null : usedName;
    componentInfo.namedExport = false;
    componentInfo.forwardRefInnerFn = importInfo.forwardRefInnerFn;
    componentInfo.shouldUseDisplayName = componentInfo.shouldUseDisplayName
      ? true
      : importInfo.shouldUseDisplayName;

    if (!importInfo.path) {
      return;
    }

    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    const component = getCompleteComponentInfoByFilePath(importInfo.path, {
      name: importInfo.name,
      namedExport: importInfo.namedExport,
      isIsland,
    });

    if (!component.name) {
      return;
    }

    Object.assign(componentInfo, component);
  });
};

/**
 * @param {{ ast: BabelAST, path: string }} fileInfo
 * @param {PartialComponentInfo & { importName: string, path: string }} componentAnalized
 */
const assignComponentInfo = (
  { ast, path },
  { name, importName, namedExport, isIsland, path: componentPath },
) => {
  const importInfo = getComponentImportInfoFromAst(ast, name, path);
  // TODO: Remove forwardRefInnerFn prop use when "forwardRef" support is removed
  const componentInfo = {
    name,
    exportedName: importName,
    namedExport,
    shouldUseDisplayName: importInfo.shouldUseDisplayName,
    forwardRefInnerFn: importInfo.forwardRefInnerFn,
    path: componentPath,
  };

  if (!importInfo.path) {
    return componentInfo;
  }

  // eslint-disable-next-line @typescript-eslint/no-use-before-define
  const component = getCompleteComponentInfoByFilePath(importInfo.path, {
    name: importInfo.namedExport ? importInfo.name : IMPORT_TYPES.DEFAULT,
    namedExport: importInfo.namedExport,
    isIsland,
  });

  if (!component.name) {
    return componentInfo;
  }

  Object.assign(componentInfo, component);

  return componentInfo;
};

/**
 * Returns the component information (name, export name, path and wether is named export or not) obtained from the file from which the component is exported
 * @param {string} filePath
 * @param {PartialComponentInfo} componentInfoFromFile - Component information obtained from the file path received
 */
const getCompleteComponentInfoByFilePath = (
  filePath,
  { name: componentUseName, namedExport: componentUsedAsNamed, isIsland },
) => {
  try {
    const ast = astCache.getAstByFilePath(filePath);

    // TODO: Remove forwardRefInnerFn prop use when "forwardRef" support is removed
    /** @type {ComponentInfo} */
    const componentInfo = {
      name: null,
      exportedName: null,
      path: filePath,
      namedExport: null,
      shouldUseDisplayName: false,
      forwardRefInnerFn: { name: null, isLocal: false },
    };

    traverse(ast, {
      ExportNamedDeclaration(babelPath) {
        if (!componentUsedAsNamed) {
          return;
        }

        if (isExportType(babelPath.node) || isEnumDeclaration(babelPath.node)) {
          if (
            babelPath.node.declaration &&
            babelTypes.isIdentifier(babelPath.node.declaration.id) &&
            babelPath.node.declaration.id.name === componentUseName
          ) {
            componentInfo.path = null;
          }

          babelPath.skip();

          return;
        }

        if (
          isVariableNamedExport(babelPath) &&
          babelTypes.isIdentifier(
            babelPath.node.declaration.declarations[0].id,
          ) &&
          babelPath.node.declaration.declarations[0].id.name ===
            componentUseName
        ) {
          const { id } = babelPath.node.declaration.declarations[0];

          Object.assign(
            componentInfo,
            assignComponentInfo(
              { ast, path: filePath },
              {
                name: id.name,
                importName: componentUseName,
                namedExport: true,
                isIsland,
                path: componentInfo.path || filePath,
              },
            ),
          );

          babelPath.stop();

          return;
        }

        // export class Component{}
        if (
          babelTypes.isClassDeclaration(babelPath.node.declaration) &&
          babelTypes.isIdentifier(babelPath.node.declaration.id) &&
          babelPath.node.declaration.id.name === componentUseName
        ) {
          const { id } = babelPath.node.declaration;

          componentInfo.name = id.name;
          componentInfo.exportedName = id.name;
          componentInfo.path = filePath;
          componentInfo.namedExport = true;

          babelPath.stop();

          return;
        }

        if (!isObjectExport(babelPath)) {
          return;
        }

        babelPath.node.specifiers.forEach((specifier) => {
          if (
            !babelTypes.isExportSpecifier(specifier) ||
            !babelTypes.isIdentifier(specifier.exported) ||
            specifier.component ||
            specifier.exported.name !== componentUseName
          ) {
            return;
          }

          // is "export {} from 'path';"
          if (babelTypes.isStringLiteral(babelPath.node.source)) {
            componentInfo.exportedName = componentUseName;
            componentInfo.namedExport =
              specifier.local.name !== IMPORT_TYPES.DEFAULT;

            const component = getCompleteComponentInfoByFilePath(
              getScriptFullPathWithExtension(
                babelPath.node.source.value,
                nodePath.dirname(filePath),
              ),
              {
                name: specifier.local.name,
                namedExport: componentInfo.namedExport,
                isIsland,
              },
            );

            if (!component.name) {
              return;
            }

            Object.assign(componentInfo, component);

            babelPath.stop();

            return;
          }

          const previousShouldUseDisplayName =
            componentInfo.shouldUseDisplayName;

          // is "export {};"
          Object.assign(
            componentInfo,
            assignComponentInfo(
              { ast, path: filePath },
              {
                name: specifier.local.name,
                importName: componentUseName,
                namedExport: true,
                isIsland,
                path: componentInfo.path || filePath,
              },
            ),
          );

          if (previousShouldUseDisplayName) {
            componentInfo.shouldUseDisplayName = true;
          }

          babelPath.stop();
        });
      },
      ExportDefaultDeclaration(babelPath) {
        // is "export default Component;"
        if (!componentUsedAsNamed && isIdentifierDefaultExport(babelPath)) {
          const previousShouldUseDisplayName =
            componentInfo.shouldUseDisplayName;

          Object.assign(
            componentInfo,
            assignComponentInfo(
              { ast, path: filePath },
              {
                name: babelPath.node.declaration.name,
                importName: componentUseName,
                namedExport: false,
                isIsland,
                path: componentInfo.path || filePath,
              },
            ),
          );

          if (previousShouldUseDisplayName) {
            componentInfo.shouldUseDisplayName = true;
          }

          babelPath.stop();

          return;
        }

        // is "export default Component = () => {};"
        if (
          !componentUsedAsNamed &&
          babelTypes.isAssignmentExpression(babelPath.node.declaration) &&
          babelTypes.isIdentifier(babelPath.node.declaration.left)
        ) {
          Object.assign(
            componentInfo,
            assignComponentInfo(
              { ast, path: filePath },
              {
                name: babelPath.node.declaration.left.name,
                importName: IMPORT_TYPES.DEFAULT,
                namedExport: false,
                isIsland,
                path: componentInfo.path || filePath,
              },
            ),
          );

          babelPath.stop();

          return;
        }

        // is "export default () => {...};"
        if (
          !componentUsedAsNamed &&
          babelTypes.isArrowFunctionExpression(babelPath.node.declaration)
        ) {
          if (isIsland) {
            throw new NordicIslandsError(
              getIslandChildExportedWithCallExpressionError(),
              {
                type: APP_FILE_MEMBERS.COMPONENT,
                path: filePath,
              },
            );
          }

          componentInfo.name = IMPORT_TYPES.DEFAULT;
          componentInfo.exportedName = IMPORT_TYPES.DEFAULT;
          componentInfo.namedExport = false;
          componentInfo.path = filePath;

          babelPath.stop();

          return;
        }

        // is "export default fn();"
        if (
          !componentUsedAsNamed &&
          babelTypes.isCallExpression(babelPath.node.declaration)
        ) {
          setInfoOfComponentWrappedWithFn(
            componentInfo,
            { ast, usedName: componentUseName, filePath, isIsland },
            babelPath,
          );

          babelPath.stop();

          return;
        }

        if (!componentUsedAsNamed || !isObjectDefaultExport(babelPath)) {
          return;
        }

        // is "export default {};"
        babelPath.node.declaration.properties.forEach((property) => {
          if (
            !babelTypes.isObjectProperty(property) ||
            !babelTypes.isIdentifier(property.value) ||
            !babelTypes.isIdentifier(property.key) ||
            property.computed ||
            property.key.name !== componentUseName
          ) {
            return;
          }

          const previousShouldUseDisplayName =
            componentInfo.shouldUseDisplayName;

          Object.assign(
            componentInfo,
            assignComponentInfo(
              { ast, path: filePath },
              {
                name: property.value.name,
                importName: componentUseName,
                namedExport: true,
                isIsland,
                path: componentInfo.path || filePath,
              },
            ),
          );

          if (previousShouldUseDisplayName) {
            componentInfo.shouldUseDisplayName = true;
          }

          babelPath.stop();
        });
      },
      ExportNamespaceSpecifier(babelPath) {
        if (componentUsedAsNamed) {
          return;
        }

        // is "export * as Component from 'path';"
        const component = getCompleteComponentInfoByFilePath(
          getScriptFullPathWithExtension(
            babelPath.parentPath.node.source.value,
            nodePath.dirname(filePath),
          ),
          {
            name: componentUseName,
            namedExport: false,
            isIsland,
          },
        );

        if (!component.name) {
          return;
        }

        Object.assign(componentInfo, component);
        babelPath.stop();
      },
      ExportAllDeclaration(babelPath) {
        // is "export * from 'path';"
        const component = getCompleteComponentInfoByFilePath(
          getScriptFullPathWithExtension(
            babelPath.node.source.value,
            nodePath.dirname(filePath),
          ),
          {
            name: componentUseName,
            namedExport: componentUsedAsNamed,
            isIsland,
          },
        );

        if (!component.name && component.path) {
          return;
        }

        Object.assign(componentInfo, component);
        babelPath.stop();
      },
    });

    return componentInfo;
  } catch (error) {
    throw new NordicIslandsError(
      getLookingForComponentPathError(componentUseName, error.message),
      {
        type: APP_FILE_MEMBERS.COMPONENT,
        path: filePath,
      },
      error,
    );
  }
};

module.exports = {
  assignComponentInfo,
  getCompleteComponentInfoByFilePath,
};
