// @nordic/fs-cache/src/source-map/index.ts
import { IS_VITE } from '../utils/constants';
import { getMappedPosition as getWebpackMappedPosition } from './mapped-position/webpack';
import { getMappedPosition as getViteMappedPosition } from './mapped-position';

export type CodeLocation = {
  url: string;
  file: string;
  line?: string;
  column?: string;
};

/* v8 ignore next -- Temp condition while webpack support is maintained */
const getMappedPosition = IS_VITE
  ? getViteMappedPosition
  : getWebpackMappedPosition;

class SourceMapCache {
  #nodeModulesCache = new Map<string, string>();

  async getCodePositionPath({ url, file, line, column }: CodeLocation) {
    const mappedPosition = await getMappedPosition(this.#nodeModulesCache, {
      url,
      file,
      line: line ? +line : 1,
      column: column ? +column : 0,
    });

    if (!mappedPosition) {
      return null;
    }

    return `${mappedPosition.source}:${mappedPosition.line}:${mappedPosition.column}`;
  }
}

export const sourceMapCache = new SourceMapCache();
