// @andes/textfield/build/esm/masks/cl/index.development.js
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';

/** @type {import('../../..').TextfieldMask} */
var money = {
  mask: 'N',
  tokens: {
    N: {
      pattern: /\d/,
      multiple: true
    },
    D: {
      pattern: /\d/,
      optional: true
    }
  },
  preProcess: val => val.replace(/\./g, ''),
  postProcess: val => {
    if (!val) return '';
    const sub = 3 - (val.includes(',') ? val.length - val.indexOf(',') : 0);
    return Intl.NumberFormat('es-AR', {
      // Se deja el format de AR ya que el de chile no funciona
      style: 'currency',
      currency: 'ARS'
    }).format(Number(val.replace(',', '.'))).replace('$', '').replace(/\s/g, '').slice(0, sub ? -sub : undefined);
  }
};

/** @type {import('../../..').TextfieldMask} */
var rut = {
  mask: 'X',
  tokens: {
    X: {
      pattern: /[\dkK]/,
      multiple: true
    }
  },
  eager: true,
  preProcess: val => val.replace(/[^\dkK]/gi, '').slice(0, 9),
  postProcess: rawInput => {
    if (!rawInput) return '';
    const sanitized = rawInput.replace(/[^0-9kK]/g, '').toUpperCase().replace(/^0+/, '');
    if (!sanitized) return '';
    if (sanitized.length < 8) {
      const _ref = sanitized.match(/^(\d{1,2})(\d{0,3})?(\d{0,3})?/) || [],
        _ref2 = _slicedToArray(_ref, 4),
        _ref2$ = _ref2[1],
        part1 = _ref2$ === void 0 ? '' : _ref2$,
        _ref2$2 = _ref2[2],
        part2 = _ref2$2 === void 0 ? '' : _ref2$2,
        _ref2$3 = _ref2[3],
        part3 = _ref2$3 === void 0 ? '' : _ref2$3;
      return [part1, part2, part3].filter(Boolean).join('.');
    }
    const numberPart = sanitized.slice(0, -1);
    const verifier = sanitized.slice(-1);
    return `${numberPart.replace(/^(\d{1,2})(\d{3})(\d{3})$/, '$1.$2.$3')}-${verifier}`;
  }
};
export { rut as RUT, money };
