// @nordic/env/dist/esm/index.production.mjs
import env$1 from 'frontend-env';
import fs from 'node:fs';
import crypto from 'node:crypto';

/* eslint-disable security/detect-non-literal-fs-filename */
const VERSION_REGEX = /"version":\s*"(.+)"/;
function readFileValueByRegex(filePath, regex) {
    if (!fs.existsSync(filePath)) {
        return '';
    }
    return regex.exec(fs.readFileSync(filePath, 'utf8'))?.[1] ?? '';
}
function readAppPackageField(field) {
    return readFileValueByRegex('package.json', 
    // eslint-disable-next-line security/detect-non-literal-regexp
    new RegExp(`"${field}":\\s*"(.+?)"`));
}
function readFuryAppName() {
    return readFileValueByRegex('.fury', /application_name:\s*(.*)/).trim();
}
function readNordicVersion() {
    return readFileValueByRegex('node_modules/nordic/package.json', VERSION_REGEX);
}
function readNordicModulesVersion() {
    return readFileValueByRegex('node_modules/@nordic/layout/package.json', VERSION_REGEX);
}
function readNordicWebkitVersion() {
    return readFileValueByRegex(
    // TODO: Change with `@nordic/native-action` in the next major version
    'node_modules/@nordic/webnative-core/package.json', VERSION_REGEX);
}
function readReactVersion() {
    return readFileValueByRegex('node_modules/react/package.json', VERSION_REGEX);
}

class NordicEnvError extends Error {
    name = 'NordicEnvError';
}

function getAppValue({ nordicVersion, key, value, }) {
    const finalValue = value || readAppPackageField(key);
    if (!finalValue && nordicVersion) {
        throw new NordicEnvError(`Application ${key} is required,` +
            ` make sure it is set in the "${key}" environment variable` +
            ` or "${key}" field in the \`package.json\` file.`);
    }
    return finalValue;
}
function getAppName(nordicVersion) {
    const { APPLICATION, npm_package_name: pkgName } = env$1;
    return getAppValue({
        nordicVersion,
        value: APPLICATION || readFuryAppName() || pkgName,
        key: 'name',
    });
}
function getAppNameHash(appName) {
    return appName && crypto.hash('md5', appName, 'base64url');
}
function getAppVersion(nordicVersion) {
    const { VERSION, npm_package_version: pkgVersion } = env$1;
    return getAppValue({
        nordicVersion,
        value: VERSION || pkgVersion,
        key: 'version',
    });
}

const nordicVersion = readNordicVersion();
const appName = getAppName(nordicVersion);
// @ts-expect-error force union with `NodeJS.ProcessEnv`
const env = {
    ...env$1,
    NORDIC_VERSION: nordicVersion,
    NORDIC_MODULES_VERSION: readNordicModulesVersion(),
    NORDIC_WEBKIT_VERSION: readNordicWebkitVersion(),
    REACT_VERSION: readReactVersion(),
    APP_NAME_HASH: getAppNameHash(appName),
    APPLICATION: appName,
    VERSION: getAppVersion(nordicVersion),
};

export { env };
