// node-melitk-otel/lib/providers/meter-provider-singleton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MeterProviderSingleton = void 0;
exports.getMetricAllowKeys = getMetricAllowKeys;
const sdk_metrics_1 = require("@opentelemetry/sdk-metrics");
const api_1 = require("@opentelemetry/api");
const grpc_js_1 = require("@grpc/grpc-js");
const otel_1 = require("../constants/otel");
const delta_metric_exporter_1 = require("./delta-metric-exporter");
function getMetricAllowKeys(envKey, defaults) {
    const envValue = process.env[envKey];
    if (envValue) {
        const keys = envValue
            .split(',')
            .map((k) => k.trim())
            .filter(Boolean);
        if (keys.length > 0) {
            return keys;
        }
    }
    return defaults;
}
function getAllowListProcessor(envKey, defaults) {
    return (0, sdk_metrics_1.createAllowListAttributesProcessor)(getMetricAllowKeys(envKey, defaults));
}
class MeterProviderSingleton {
    static instance;
    static getInstance() {
        if (!MeterProviderSingleton.instance) {
            const meterProvider = this.newMeterProvider();
            api_1.metrics.setGlobalMeterProvider(meterProvider);
            MeterProviderSingleton.instance = meterProvider;
        }
        return MeterProviderSingleton.instance;
    }
    static newMeterProvider() {
        const metricReader = new sdk_metrics_1.PeriodicExportingMetricReader({
            exporter: new delta_metric_exporter_1.DeltaMetricExporter({
                url: `${otel_1.OTEL_AGENT_HOST}:${otel_1.OTEL_AGENT_PORT}`,
                credentials: grpc_js_1.credentials.createInsecure(),
                metadata: new grpc_js_1.Metadata({ waitForReady: true }),
            }),
            exportIntervalMillis: otel_1.EXPORT_INTERVAL_MILLIS,
            exportTimeoutMillis: otel_1.EXPORT_TIMEOUT_MILLIS,
        });
        return new sdk_metrics_1.MeterProvider({
            views: this.createMetricViews(),
            readers: [metricReader],
        });
    }
    static createMetricViews() {
        const clientDurationProcessor = getAllowListProcessor(otel_1.HTTP_CLIENT_DURATION_ALLOW_KEYS_ENV, otel_1.DEFAULT_HTTP_CLIENT_DURATION_ATTRIBUTES);
        const serverDurationProcessor = getAllowListProcessor(otel_1.HTTP_SERVER_DURATION_ALLOW_KEYS_ENV, otel_1.DEFAULT_HTTP_SERVER_DURATION_ATTRIBUTES);
        const serverRequestDurationProcessor = getAllowListProcessor(otel_1.HTTP_SERVER_REQUEST_DURATION_ALLOW_KEYS_ENV, otel_1.DEFAULT_HTTP_SERVER_REQUEST_DURATION_ATTRIBUTES);
        const clientRequestDurationProcessor = getAllowListProcessor(otel_1.HTTP_CLIENT_REQUEST_DURATION_ALLOW_KEYS_ENV, otel_1.DEFAULT_HTTP_CLIENT_REQUEST_DURATION_ATTRIBUTES);
        return [
            // ============================================================
            // Explicit bucket histogram views (original metrics)
            // ============================================================
            // http.server.duration (old semconv, milliseconds) - explicit buckets
            {
                instrumentType: sdk_metrics_1.InstrumentType.HISTOGRAM,
                instrumentName: otel_1.HTTP_SERVER_DURATION_METRIC_NAME,
                aggregation: {
                    type: sdk_metrics_1.AggregationType.EXPLICIT_BUCKET_HISTOGRAM,
                    options: { boundaries: otel_1.HISTOGRAM_BUCKETS_MSEC },
                },
                attributesProcessors: [serverDurationProcessor],
            },
            // http.client.duration (old semconv, milliseconds) - explicit buckets
            {
                instrumentType: sdk_metrics_1.InstrumentType.HISTOGRAM,
                instrumentName: 'http.client.duration',
                aggregation: {
                    type: sdk_metrics_1.AggregationType.EXPLICIT_BUCKET_HISTOGRAM,
                    options: { boundaries: otel_1.HISTOGRAM_BUCKETS_MSEC },
                },
                attributesProcessors: [clientDurationProcessor],
            },
            // http.server.request.duration (new semconv, seconds) - explicit buckets
            {
                instrumentType: sdk_metrics_1.InstrumentType.HISTOGRAM,
                instrumentName: 'http.server.request.duration',
                aggregation: {
                    type: sdk_metrics_1.AggregationType.EXPLICIT_BUCKET_HISTOGRAM,
                    options: { boundaries: otel_1.HISTOGRAM_BUCKETS_SEC },
                },
                attributesProcessors: [serverRequestDurationProcessor],
            },
            // http.client.request.duration (new semconv, seconds) - explicit buckets
            {
                instrumentType: sdk_metrics_1.InstrumentType.HISTOGRAM,
                instrumentName: 'http.client.request.duration',
                aggregation: {
                    type: sdk_metrics_1.AggregationType.EXPLICIT_BUCKET_HISTOGRAM,
                    options: { boundaries: otel_1.HISTOGRAM_BUCKETS_SEC },
                },
                attributesProcessors: [clientRequestDurationProcessor],
            },
            // ============================================================
            // Exponential histogram views (renamed metrics for Datadog migration)
            // ============================================================
            // http.server.duration -> http.server.duration_exponential
            {
                instrumentType: sdk_metrics_1.InstrumentType.HISTOGRAM,
                instrumentName: otel_1.HTTP_SERVER_DURATION_METRIC_NAME,
                name: 'http.server.duration_exponential',
                aggregation: {
                    type: sdk_metrics_1.AggregationType.EXPONENTIAL_HISTOGRAM,
                    options: {
                        maxSize: otel_1.EXPONENTIAL_HISTOGRAM_MAX_SIZE,
                        recordMinMax: otel_1.EXPONENTIAL_HISTOGRAM_RECORD_MIN_MAX,
                    },
                },
                attributesProcessors: [serverDurationProcessor],
            },
            // http.server.request.duration -> http.server.request.duration_exponential
            {
                instrumentType: sdk_metrics_1.InstrumentType.HISTOGRAM,
                instrumentName: 'http.server.request.duration',
                name: 'http.server.request.duration_exponential',
                aggregation: {
                    type: sdk_metrics_1.AggregationType.EXPONENTIAL_HISTOGRAM,
                    options: {
                        maxSize: otel_1.EXPONENTIAL_HISTOGRAM_MAX_SIZE,
                        recordMinMax: otel_1.EXPONENTIAL_HISTOGRAM_RECORD_MIN_MAX,
                    },
                },
                attributesProcessors: [serverRequestDurationProcessor],
            },
        ];
    }
}
exports.MeterProviderSingleton = MeterProviderSingleton;
