// node-melitk-otel/lib/providers/tracer-provider-singleton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TracerProviderSingleton = void 0;
const grpc_js_1 = require("@grpc/grpc-js");
const exporter_trace_otlp_grpc_1 = require("@opentelemetry/exporter-trace-otlp-grpc");
const api_1 = require("@opentelemetry/api");
const core_1 = require("@opentelemetry/core");
const B3MultiPropagator_1 = require("@opentelemetry/propagator-b3/build/src/B3MultiPropagator");
const context_async_hooks_1 = require("@opentelemetry/context-async-hooks");
const otel_1 = require("../constants/otel");
const configs_1 = require("../configs");
const orchestrator_1 = require("../orchestrator");
class TracerProviderSingleton {
    static instance;
    static getInstance() {
        if (!TracerProviderSingleton.instance) {
            TracerProviderSingleton.instance = this.newTracerProvider();
        }
        return TracerProviderSingleton.instance;
    }
    static newTracerProvider() {
        const collectorOptions = {
            url: `${otel_1.OTEL_AGENT_HOST}:${otel_1.OTEL_AGENT_PORT}`,
            credentials: grpc_js_1.credentials.createInsecure(),
            metadata: new grpc_js_1.Metadata({ waitForReady: true }), // prevent UNAVAILABLE
            timeoutMillis: configs_1.OTEL_AGENT_DEFAULT_TIMEOUT,
        };
        const tracerExporter = new exporter_trace_otlp_grpc_1.OTLPTraceExporter(collectorOptions);
        const tracerProvider = (0, orchestrator_1.orchestrateTracerProvider)(tracerExporter);
        api_1.trace.setGlobalTracerProvider(tracerProvider);
        api_1.propagation.setGlobalPropagator(new core_1.CompositePropagator({
            propagators: [
                new core_1.W3CTraceContextPropagator(),
                new core_1.W3CBaggagePropagator(),
                new B3MultiPropagator_1.B3MultiPropagator(),
            ],
        }));
        const contextManager = new context_async_hooks_1.AsyncHooksContextManager().enable();
        api_1.context.setGlobalContextManager(contextManager);
        return tracerProvider;
    }
}
exports.TracerProviderSingleton = TracerProviderSingleton;
