// @andes/common/build/cjs/metrics/node-index.development.js
'use strict';

var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var _objectWithoutProperties__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutProperties);
const _excluded = ["type"];
/* eslint-disable import/no-extraneous-dependencies */
const _require = require('frontend-statsd/internal'),
  statsd = _require.statsd;
const env = require('frontend-env');
const _require2 = require('@andes/common/build/utils/encodeBase32'),
  encodeStringToBase32 = _require2.encodeStringToBase32;
const transformToKebabCase = require('@andes/common/build/utils/transformToKebabCase');
const statName = 'andes.frontend_andes_ui.customization';
const initialValue = 1;
const sampleRate = 0.25;

/** @typedef {import('./metrics').CustomizationData} CustomizationData */

/** @param {CustomizationData} data */
const trackCustomization = _ref => {
  let type = _ref.type,
    other = _objectWithoutProperties__default["default"](_ref, _excluded);
  statsd.increment(statName, initialValue, sampleRate, _objectSpread__default["default"]({
    application: env.APPLICATION,
    platform: env.PLATFORM || '',
    customizationType: type
  }, other), () => {});
};

/** @type {Record<string, import('./metrics.d.ts').CustomizationType>} */
const CUSTOMIZATION_TYPE = {
  CUSTOM: 'custom',
  HEADLESS: 'headless',
  STYLE_OVERRIDES: 'styleOverrides',
  TOKENS: 'tokens',
  SLOTS: 'slots'
};
const MAX_TAG_LENGTH = 200;

/**
 * Shortens the properties of an object to fit within a specified maximum length
 *
 * @param {Record<string, string>} obj
 * @param {number} maxLength
 */
const shortenObjectToMaxLength = function () {
  let obj = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let maxLength = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  const shortenObject = /** @type {Record<string, string>} */{};
  const objectString = JSON.stringify(obj);
  const objectBase32 = encodeStringToBase32(objectString);
  if (objectBase32.length <= maxLength) {
    return obj;
  }
  let accumulatedCharacters = 0;
  // eslint-disable-next-line no-restricted-syntax
  for (const _ref2 of Object.entries(obj)) {
    var _ref3 = _slicedToArray__default["default"](_ref2, 2);
    const key = _ref3[0];
    const value = _ref3[1];
    const stringProperty = `"${key}":${JSON.stringify(value)},`;
    if (accumulatedCharacters + stringProperty.length <= maxLength) {
      shortenObject[key] = value;
      accumulatedCharacters += stringProperty.length;
    } else {
      break;
    }
  }
  return shortenObject;
};

/**
 * Encode the values of an object in Base32
 *
 * @param {Record<string, string>} obj
 */
const parseObjectValuesToBase32 = function () {
  let obj = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return Object.entries(obj).reduce((acc, _ref4) => {
    let _ref5 = _slicedToArray__default["default"](_ref4, 2),
      key = _ref5[0],
      value = _ref5[1];
    acc[key] = encodeStringToBase32(value);
    return acc;
  }, /** @type {Record<string, string>} */{});
};

/**
 * Creates an object with methods to process and record various types of customization metrics.
 *
 * @param {(data?: CustomizationData) => void} tracker - Tracking function used to record metrics.
 * @returns {Object} An object with methods to send different types of customization metrics:
 *   - `sendCustom`: Processes and records custom metrics.
 *   - `sendHeadless`: Processes and records metrics for headless components.
 *   - `sendStyleOverrides`: Processes and records style override metrics.
 *   - `sendTokens`: Processes and records 'tokens' metrics.
 */
const senderMetrics = tracker => {
  /**
   * Process data and record 'custom' metrics.
   *
   * @type {import('./metrics').SendCustom}
   */
  const sendCustom = _ref6 => {
    let component = _ref6.component,
      _ref6$details = _ref6.details,
      details = _ref6$details === void 0 ? {} : _ref6$details;
    const tagKey = 'custom-properties';
    const properties = shortenObjectToMaxLength(details, MAX_TAG_LENGTH - tagKey.length);
    const propertiesString = JSON.stringify(properties);
    const propertiesBase32 = encodeStringToBase32(propertiesString);
    const hasMoreProperties = Object.keys(properties).length < Object.keys(details).length;
    const data = {
      [tagKey]: propertiesBase32,
      'more-properties': hasMoreProperties
    };
    tracker === null || tracker === void 0 || tracker(_objectSpread__default["default"]({
      type: CUSTOMIZATION_TYPE.CUSTOM,
      // @ts-expect-error
      component: typeof component === 'string' ? transformToKebabCase(component) : undefined
    }, data));
  };

  /**
   * Processes and records metrics for headless components.
   *
   * @type {import('./metrics').SendHeadless}
   */
  const sendHeadless = componentName => componentProps => {
    const shouldSendHeadlessComponentMetric = 'extend' in componentProps && componentProps.extend;
    if (shouldSendHeadlessComponentMetric) {
      tracker === null || tracker === void 0 || tracker({
        type: CUSTOMIZATION_TYPE.HEADLESS,
        component: componentName
      });
    }
  };

  /**
   * Processes and records metrics for slots components.
   *
   * @type {import('./metrics').SendSlots}
   */
  const sendSlots = (componentName, slotConditionFn) => componentProps => {
    const shouldSlotComponentMetric = typeof slotConditionFn === 'function' ? slotConditionFn(componentProps) : false;
    if (shouldSlotComponentMetric) {
      tracker === null || tracker === void 0 || tracker({
        type: CUSTOMIZATION_TYPE.SLOTS,
        component: componentName
      });
    }
  };

  /**
   * Process data and record 'styleOverrides' metrics.
   *
   * @type {import('./metrics').SendStyleOverrides}
   */
  const sendStyleOverrides = _ref7 => {
    let component = _ref7.component;
    tracker === null || tracker === void 0 || tracker({
      type: CUSTOMIZATION_TYPE.STYLE_OVERRIDES,
      // @ts-expect-error
      component: transformToKebabCase(component)
    });
  };

  /**
   * Process data and record 'tokens' metrics.
   *
   * @type {import('./metrics').SendTokens}
   */
  const sendTokens = _ref8 => {
    let _ref8$details = _ref8.details,
      details = _ref8$details === void 0 ? {} : _ref8$details;
    const tokensStringToMaxLength = Object.entries(details).reduce((acc, _ref9) => {
      let _ref0 = _slicedToArray__default["default"](_ref9, 2),
        key = _ref0[0],
        value = _ref0[1];
      acc[key] = typeof value === 'string' ? value.slice(0, Math.max(MAX_TAG_LENGTH - key.length, 0)) : value;
      return acc;
    }, /** @type {Record<string, string>} */{});
    const tokens = parseObjectValuesToBase32(tokensStringToMaxLength);
    tracker === null || tracker === void 0 || tracker(_objectSpread__default["default"]({
      type: CUSTOMIZATION_TYPE.TOKENS
    }, tokens));
  };
  return {
    sendCustom,
    sendHeadless,
    sendSlots,
    sendStyleOverrides,
    sendTokens
  };
};
module.exports = {
  trackCustomization,
  senderMetrics
};
