// @andes/list/build/esm/generators/itemNodeGenerator.development.js
import { Children } from 'react';
import warning from '@andes/common/build/utils/warning';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';

/** @typedef {import('../..').RightContent} RightContent */

/**
 * @param {import('react').ReactNode | undefined} slot
 * @param {Pick<import('../..').ListItemContentProps, 'title' | 'description' | 'rightContent'>} layoutProps
 * @param {{ legacyBehavior?: boolean }} options
 */
const hasSlot = (slot, _ref, _ref2) => {
  let title = _ref.title,
    description = _ref.description,
    rightContent = _ref.rightContent;
  let _ref2$legacyBehavior = _ref2.legacyBehavior,
    legacyBehavior = _ref2$legacyBehavior === void 0 ? false : _ref2$legacyBehavior;
  const hasRightContent = typeof rightContent === 'object' ? /** @type {RightContent} */(rightContent === null || rightContent === void 0 ? void 0 : rightContent.content) || (/** @type {RightContent} */rightContent === null || rightContent === void 0 ? void 0 : rightContent.description) : rightContent;
  const hasDefaultContent = title || description || hasRightContent;
  const isSlotted = legacyBehavior ? slot && !hasDefaultContent : slot;
  return !!isSlotted;
};
const DEFAULT_POSITION_TOP = 'top';

/**
 * @typedef {Object} RightContentConfig
 * @prop {'top' | 'centered'} position
 * @prop {import('react').ReactNode} [content]
 * @prop {string} [description]
 */

/**
 * @typedef {Object} RightContent
 * @prop {string} [content]
 * @prop {string} [description]
 */

/**
 * @typedef {Object} Props
 * @prop {import('react').ReactNode | RightContent} [rightContent]
 * @prop {'centered' | 'top'} [rightContentPosition]
 */

/** @type {(value: import('react').ReactNode | RightContent) => value is RightContent} */
const isRightContentType = value => /** @type {RightContent} */(value === null || value === void 0 ? void 0 : value.content) !== undefined;

/** @type {(props:Props) => RightContentConfig} */
const computeRightContent = _ref => {
  let rightContentProp = _ref.rightContent,
    rightContentPositionProp = _ref.rightContentPosition;
  if (!rightContentProp) return {
    position: DEFAULT_POSITION_TOP
  };
  const computedRightContentPosition = {
    position: rightContentPositionProp || DEFAULT_POSITION_TOP
  };
  if (isRightContentType(rightContentProp)) {
    var _rightContentProp$des;
    return _objectSpread(_objectSpread({}, computedRightContentPosition), {}, {
      content: rightContentProp === null || rightContentProp === void 0 ? void 0 : rightContentProp.content,
      description: (_rightContentProp$des = rightContentProp === null || rightContentProp === void 0 ? void 0 : rightContentProp.description) !== null && _rightContentProp$des !== void 0 ? _rightContentProp$des : undefined
    });
  }
  return _objectSpread(_objectSpread({}, computedRightContentPosition), {}, {
    content: rightContentProp || null,
    position: rightContentPositionProp || DEFAULT_POSITION_TOP
  });
};
var computeRightContent$1 = computeRightContent;

/** @typedef {{ type: string; element: import('react').ReactNode }} ChildNodeItem */

/**
 * @typedef {Object} NativeProps
 * @prop {string} [`aria-label`]
 */

/** @typedef {import('../../index').ListItemProps} ListItemProps */

/** @param {ListItemProps} props */
const hasChildNodes = _ref => {
  let title = _ref.title,
    children = _ref.children;
  return title && Children.count(children) > 0;
};

/** @param {ListItemProps & NativeProps} props */
const itemNodeGenerator = function* getCollectionNode(props) {
  var _computeRightContent;
  const children = props.children,
    ariaLabel = props['aria-label'],
    title = props.title,
    description = props.description,
    rightContent = props.rightContent;
  const rendered = title || children;
  const textValue = (typeof rendered === 'string' ? rendered : '') || ariaLabel || '';
  const isSlotted = hasSlot(children, {
    title,
    description,
    rightContent: (_computeRightContent = computeRightContent$1({
      rightContent
    })) === null || _computeRightContent === void 0 ? void 0 : _computeRightContent.content
  }, {
    legacyBehavior: true
  });
  warning(isSlotted || textValue, '<ListItem> needs some plain text contents for accessibility reasons. Please add a `title` or `aria-label` prop.');
  yield {
    type: 'item',
    key: props.value,
    props,
    rendered,
    textValue,
    'aria-label': ariaLabel || title,
    hasChildNodes,
    *childNodes() {
      if (title) {
        /** @type {ChildNodeItem[]} */
        const items = [];
        Children.forEach(children, child => {
          items.push({
            type: 'item',
            element: child
          });
        });
        yield* items;
      }
    }
  };
};
export { itemNodeGenerator as default };
