// @nordic/env/src/fs.ts
/* eslint-disable security/detect-non-literal-fs-filename */
import fs from 'node:fs';

const VERSION_REGEX = /"version":\s*"(.+)"/;

function readFileValueByRegex(filePath: string, regex: RegExp) {
  if (!fs.existsSync(filePath)) {
    return '';
  }

  return regex.exec(fs.readFileSync(filePath, 'utf8'))?.[1] ?? '';
}

export function readAppPackageField(field: string) {
  return readFileValueByRegex(
    'package.json',
    // eslint-disable-next-line security/detect-non-literal-regexp
    new RegExp(`"${field}":\\s*"(.+?)"`),
  );
}
export function readFuryAppName() {
  return readFileValueByRegex('.fury', /application_name:\s*(.*)/).trim();
}

export function readNordicVersion() {
  return readFileValueByRegex(
    'node_modules/nordic/package.json',
    VERSION_REGEX,
  );
}

export function readNordicModulesVersion() {
  return readFileValueByRegex(
    'node_modules/@nordic/layout/package.json',
    VERSION_REGEX,
  );
}

export function readNordicWebkitVersion() {
  return readFileValueByRegex(
    // TODO: Change with `@nordic/native-action` in the next major version
    'node_modules/@nordic/webnative-core/package.json',
    VERSION_REGEX,
  );
}

export function readReactVersion() {
  return readFileValueByRegex('node_modules/react/package.json', VERSION_REGEX);
}
