// @nordic/env/src/index.ts
import type { EnvVariables } from 'frontend-env';

import baseEnv from 'frontend-env';

import {
  readNordicVersion,
  readNordicModulesVersion,
  readNordicWebkitVersion,
  readReactVersion,
} from './fs';
import { getAppName, getAppNameHash, getAppVersion } from './getters';

const nordicVersion = readNordicVersion();
const appName = getAppName(nordicVersion);

export type NordicEnvVariables = EnvVariables &
  NodeJS.ProcessEnv & {
    NORDIC_VERSION: string;
    NORDIC_MODULES_VERSION: string;
    NORDIC_WEBKIT_VERSION: string;
    REACT_VERSION: string;
    APP_NAME_HASH: string;
    APPLICATION: string;
    VERSION: string;
  };

// @ts-expect-error force union with `NodeJS.ProcessEnv`
export const env: NordicEnvVariables = {
  ...baseEnv,
  NORDIC_VERSION: nordicVersion,
  NORDIC_MODULES_VERSION: readNordicModulesVersion(),
  NORDIC_WEBKIT_VERSION: readNordicWebkitVersion(),
  REACT_VERSION: readReactVersion(),
  APP_NAME_HASH: getAppNameHash(appName),
  APPLICATION: appName,
  VERSION: getAppVersion(nordicVersion),
};
