// nordic-pages-discovery/lib/page/http-error-page.js
const { getDuplicatedPageFiles } = require('../helpers/file/inconsistency');
const { throwDuplicatedPage } = require('../helpers/duplicate');
const { Page } = require('./index');

/**
 * @typedef { keyof typeof import('../constants/device').DEVICE_TYPES } DeviceType
 * @typedef { import('./paths').Paths } Paths
 *
 * @typedef { { [device in DeviceType]?: Paths } } PagePathsByDevice
 * @typedef { { statusError?: string, paths?: PagePathsByDevice } } HttpErrorPageProps
 * @typedef { { [statusError:string]: { paths: PagePathsByDevice } } } DiscoveredHttpErrorPages
 *
 * @typedef { object } ThrowDuplicatedPageParams - Entity with the duplicated data
 * @property { string } currentFile - Full path of current file being analized
 * @property { string } previousFile - Full path of the file previously processed
 * @property { string } baseDir - Base directory
 */

/**
 * Representation of an error page accessible through error status code
 * @extends Page
 */
class HttpErrorPage extends Page {
  /**
   * @param { HttpErrorPageProps } [props={}]
   */
  constructor({ statusError = '', paths = {} } = {}) {
    super(paths);

    /**
     * @type { string }
     */
    this.statusError = statusError;
  }

  /**
   * @returns { boolean } Wether or not the error page is valid
   */
  validate() {
    return !!this.statusError?.trim() && super.validate();
  }

  /**
   * @param { string } pagesDir
   * @param { string } pagesBuildDir
   */
  completePaths(pagesDir, pagesBuildDir) {
    const pageLogPortion = `http error status code: ${this.statusError}`;

    super.completePaths(pageLogPortion, pagesDir, pagesBuildDir);
  }

  /**
   * Throws duplicated error page by file type extension or index naming convention.
   * @param { ThrowDuplicatedPageParams } duplicatedPageData
   *
   * @example
   * this.statusError = 404;
   * const currentFile = '404/index.mobile.js';
   * const previousFile = '404.mobile.js';
   * const dir = 'app/nordic-pages';
   *
   * throwDuplicated({ currentFile, previousFile, dir })
   * => Duplicated error page files by index naming convention
   */
  throwDuplicated({ baseDir, currentFile, previousFile }) {
    throwDuplicatedPage({
      baseDir,
      previousFile,
      currentFile,
      pageLogPortion: `${this.statusError} http error status code`,
    });
  }

  /**
   * Checks if the error page is duplicated.
   * If it is, returns the files of the duplicated page;
   * otherwise it returns an empty array
   * @param { DiscoveredHttpErrorPages } currentDiscoveredHttpErrorPages - Discovered error pages at the time
   * @returns { string[] } Files of the duplicated page
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  checkIfDuplicated(currentDiscoveredHttpErrorPages, dataId) {
    const previousPage = currentDiscoveredHttpErrorPages[this.statusError];

    if (!this.validate() || !previousPage) {
      return [];
    }

    return getDuplicatedPageFiles(
      new HttpErrorPage({
        statusError: this.statusError,
        paths: previousPage.paths,
      }),
      this,
    );
  }
}

module.exports = { HttpErrorPage };
