// nordic-pages-discovery/lib/page/index.js
const { SPECIFIC_DEVICE_TYPES, DEVICE_TYPES } = require('../constants/device');
const { log } = require('../helpers/logger');
const { getFileDeviceType } = require('../helpers/file/device');
const { Paths } = require('./paths');
const {
  getNoPageViewExistsForBaseStyleSheetLog,
  getNoBaseViewForBaseStyleSheetLog,
  getNoBasePathsFoundForPageLog,
} = require('../helpers/file/inconsistency');
const {
  getPriorityDeviceData,
  removeDeviceFromClientName,
} = require('../helpers/device');
const {
  DuplicatedFileByFileTypeError,
} = require('../error/duplicated-file-by-file-type-error');
const { PAGE_MEMBERS } = require('../constants/page');

/**
 * @typedef { keyof DEVICE_TYPES } DeviceType
 * @typedef { keyof PAGE_MEMBERS } PageMemberType
 * @typedef { typeof import('../constants/page').PAGE_MEMBERS.view } PageViewMember
 * @typedef { typeof import('../constants/page').PAGE_MEMBERS.styles } PageStylesMember
 * @typedef { PageViewMember | PageStylesMember } PagePathsMemberType
 * @typedef { import('../discovery/nordic-pages-discovery').PageInfoByView[string] } PageInfoByView
 *
 * @typedef { object } PageFileParams
 * @property { string } filePath - Relative path to file
 * @property { string } dir - Directory used to get the full path of the received file
 * @property { PagePathsMemberType } pageMember
 * @property {PageInfoByView} [pageInfo]
 *
 * @typedef { object } FilesByPageMemberData
 * @property { string[] } files - List of files of the page member type
 * @property { PagePathsMemberType } pageMember - Page member to which the files belongs
 * @property { string } dir - Directory path used to get files full path
 * @property { PageInfoByView } [pageInfoByViewFile]
 *
 * @typedef { object } ThrowDuplicatedPageParams - Entity with the duplicated data
 * @property { string } currentFile - Full path of current file being analized
 * @property { string } previousFile - Full path of the file previously processed
 * @property { string } baseDir - Base directory
 *
 * @typedef { object } AddPageFilesParams
 * @property { { [pageMember in PageMemberType]?: string[] } } files - Page files classified by page member
 * @property { string } pagesDir - Directory path used to get files full path
 * @property { string | undefined } pagesBuildDir - Directory path used to get build page files full path
 * @property { PageInfoByView } pageInfoByViewFile
 */

/**
 * Page template
 */
class Page {
  /**
   * @param { { [device in DeviceType]?: Paths } } [paths={}]
   */
  constructor(paths = {}) {
    if (this.constructor === Page) {
      throw new Error(
        "Page is an abstract class and abstract classes can't be instantiated",
      );
    }

    /**
     * @type { { [device in DeviceType]?: Paths } } - Page paths by device type
     */
    this.paths = paths;
  }

  /**
   * Throws duplicated page by file type extension or index naming convention.
   * @param { ThrowDuplicatedPageParams } duplicatedPageData
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars, class-methods-use-this
  throwDuplicated({ baseDir, currentFile, previousFile }) {
    throw new Error('throwDuplicatedPage is an abstract method');
  }

  /**
   * Add received page file to the corresponding paths object
   * based on the file device type and page member.
   * Throws an error if the file cannot be added because of a duplicate case
   * @param { PageFileParams } pageFileParams
   */
  addFileToCorrespondingPaths({ filePath, dir, pageMember, pageInfo = {} }) {
    try {
      const fileDeviceType = getFileDeviceType(filePath);
      const { islands, hydrate, canUseBootstrapPage } = pageInfo;
      const {
        islands: islandsInPath,
        hydrate: hydrateInPath,
        canUseBootstrapPage: canUseBootstrapPageInPath,
      } = this.paths[fileDeviceType] || {};

      const devicePaths = new Paths({
        ...this.paths[fileDeviceType],
        islands: islands?.length ? islands : islandsInPath,
        hydrate: typeof hydrate === 'undefined' ? hydrateInPath : hydrate,
        canUseBootstrapPage:
          typeof canUseBootstrapPage === 'undefined'
            ? canUseBootstrapPageInPath
            : canUseBootstrapPage,
      });

      devicePaths.addPagePath({ filePath, dir, pageMember });

      this.paths[fileDeviceType] = devicePaths;
    } catch (error) {
      if (!(error instanceof DuplicatedFileByFileTypeError)) {
        throw error;
      }

      this.throwDuplicated({
        baseDir: error.baseDir,
        currentFile: error.currentFile,
        previousFile: error.previousFile,
      });
    }
  }

  /**
   * @private
   * Completes the base paths object.
   * If an inconsistency with the paths object is detected, a warning is displayed.
   * Possible inconsistencies:
   * - StyleSheet without a view file partner
   * - No base view present
   * - No base files present
   * @param { string } pageLogPortion - Log portion with the page identifier (ex. route or statusError)
   */
  completePathsBaseObject(pageLogPortion) {
    let warnLog;

    const { device, paths: priorityPaths } = getPriorityDeviceData(this.paths);

    if (this.paths?.base?.view || (!this.paths?.base && !device)) {
      return;
    }

    if (!this.paths?.base && device) {
      const basePaths = new Paths({
        ...priorityPaths,
        client: removeDeviceFromClientName(priorityPaths.client),
      });

      this.paths.base = basePaths;

      warnLog = getNoBasePathsFoundForPageLog({
        device,
        priorityDeviceView: basePaths.view,
        pageLogPortion,
      });
    } else if (device === null) {
      warnLog = getNoPageViewExistsForBaseStyleSheetLog(
        this.paths.base.styles,
        pageLogPortion,
      );

      delete this.paths.base;
    } else {
      this.paths.base.view = priorityPaths.view;
      this.paths.base.hydrate = priorityPaths.hydrate;
      this.paths.base.canUseBootstrapPage = priorityPaths.canUseBootstrapPage;

      warnLog = getNoBaseViewForBaseStyleSheetLog({
        device,
        priorityDeviceView: priorityPaths.view,
        pageLogPortion,
        baseStyleSheet: this.paths.base.styles,
      });
    }

    log.warn(warnLog);
  }

  /**
   * @private
   */
  completeDevicePathsObject() {
    Object.keys(SPECIFIC_DEVICE_TYPES).forEach(
      (/** @type { DeviceType} */ device) => {
        /**
         * @type { Paths }
         */
        const devicePaths = this.paths[device];

        if (!devicePaths || devicePaths.view) {
          return;
        }

        devicePaths.complete(device, this.paths);
      },
    );
  }

  /**
   * @param { string } pagesDir
   * @param { string } pagesBuildDir
   */
  completePathsObject(pagesDir, pagesBuildDir) {
    Object.values(DEVICE_TYPES).forEach((deviceType) => {
      const devicePaths = this.paths[deviceType];

      if (!devicePaths) {
        return;
      }

      if (pagesBuildDir) {
        devicePaths.fillTranspiledView(pagesDir, pagesBuildDir);
      }
    });
  }

  /**
   * @param { string } pageLogPortion - Log portion with the page identifier (ex. route or statusError)
   * @param { string } pagesDir
   * @param { string } pagesBuildDir
   */
  completePaths(pageLogPortion, pagesDir, pagesBuildDir) {
    this.completeDevicePathsObject();
    this.completePathsBaseObject(pageLogPortion);
    this.completePathsObject(pagesDir, pagesBuildDir);
  }

  /**
   * @returns { boolean } Wether or not the page is valid
   */
  validate() {
    return this.paths && Object.keys(this.paths).length !== 0;
  }

  /**
   * @returns { string[] } List of all page files use in paths objects
   */
  getAllFilesOfPage() {
    return Object.keys(DEVICE_TYPES)
      .map((device) => {
        const { view, styles } = this.paths[device] || {};

        return [view || [], styles || []];
      })
      .flat(2);
  }

  /**
   * Adds each discovered file of the page to its corresponding location
   * @param { AddPageFilesParams } addPageFilesParams
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  addPageFiles({ files, pagesDir, pagesBuildDir, pageInfoByViewFile }) {
    this.addFilesByPageMember({
      files: files[PAGE_MEMBERS.view] || [],
      pageMember: PAGE_MEMBERS.view,
      dir: pagesDir,
      pageInfoByViewFile,
    });

    this.addFilesByPageMember({
      files: files[PAGE_MEMBERS.styles] || [],
      pageMember: PAGE_MEMBERS.styles,
      dir: pagesDir,
    });
  }

  /**
   * @private
   * Adds each file of a certain page member type to the corresponding paths object of the page
   * @param { FilesByPageMemberData } filesByPageMemberData
   */
  addFilesByPageMember({ files, pageMember, dir, pageInfoByViewFile = {} }) {
    files.forEach((filePath) => {
      this.addFileToCorrespondingPaths({
        pageMember,
        filePath,
        dir,
        pageInfo: pageInfoByViewFile[filePath],
      });
    });
  }
}

module.exports = { Page };
