// nordic-pages-discovery/lib/page/navigable-page.js
const { PAGE_MEMBERS } = require('../constants/page');
const {
  getDuplicatedFilesByExtensionLog,
  throwDuplicatedPage,
} = require('../helpers/duplicate');
const { getFileFullPath } = require('../helpers/file/path-detail');
const { getDuplicatedPageFiles } = require('../helpers/file/inconsistency');
const { getTranspiledPath } = require('../helpers/build');
const { Page } = require('./index');

/**
 * @typedef { keyof typeof import('../constants/device').DEVICE_TYPES } DeviceType
 * @typedef { import('./paths').Paths } Paths
 * @typedef { keyof PAGE_MEMBERS } PageMemberType
 * @typedef { { [device in DeviceType]?: Paths } } PagePathsByDevice
 * @typedef { { params: Array<string>, regexRoute: string | null } } CatchAllMetadata
 *
 * @typedef { object } NavigablePageProps
 * @property { string | RegExp } [route]
 * @property { string } [hooks]
 * @property { PagePathsByDevice } [paths]
 * @property { CatchAllMetadata } [catchAllMetadata]
 *
 * @typedef { Object } ThrowDuplicatedPageParams
 * @property { string } currentFile - Full path of current file being analized
 * @property { string } previousFile - Full path of the file previously processed
 * @property { string } baseDir - Base directory
 *
 * @typedef { object } AddPageFilesParams
 * @property { { [pageMember in PageMemberType]?: string[] } } files - Page files classified by page member
 * @property { import('../discovery/nordic-pages-discovery').PageInfoByView } pageInfoByViewFile
 * @property { string } pagesDir - Directory path used to get files full path
 * @property { string | undefined } pagesBuildDir - Directory path used to get build page files full path
 */

/**
 * Representation of a navigable page
 * @extends Page
 */
class NavigablePage extends Page {
  /**
   * @param { NavigablePageProps } [props={}]
   */
  constructor({
    route = '',
    hooks = '',
    paths = {},
    catchAllMetadata = null,
  } = {}) {
    super(paths);
    /**
     * @type { boolean }
     */
    this.isRegexRoute = route instanceof RegExp;
    /**
     * @type { string }
     */
    this.route = this.isRegexRoute ? route.source : route;
    /**
     * @type { string } Path of the page hooks file
     */
    this.hooks = hooks;
    /**
     * @type { CatchAllMetadata | null }
     */
    this.catchAllMetadata = catchAllMetadata;
  }

  /**
   * Sets the page hooks file if it was not set previously.
   * If it was, a duplicated hooks file error is thrown.
   * @param { string } hooksFile
   * @param { string } pagesDir - Directory path used to get the hooks full path
   * @param { string } pagesBuildDir - Directory path used to get the hooks build full path
   */
  setHooksFile(hooksFile, pagesDir, pagesBuildDir) {
    if (!this.hooks) {
      const fileFullPath = getFileFullPath(hooksFile, pagesDir);

      this.hooks =
        getTranspiledPath(fileFullPath, pagesDir, pagesBuildDir) ||
        fileFullPath;

      return;
    }

    const errorMessage = getDuplicatedFilesByExtensionLog({
      files: [this.hooks, hooksFile],
      filesPageMember: PAGE_MEMBERS.hooks,
      pageLogPortion: `${this.route} route`,
    });

    throw new Error(errorMessage);
  }

  /**
   * @param { string } pagesDir
   * @param { string } pagesBuildDir
   */
  completePaths(pagesDir, pagesBuildDir) {
    super.completePaths(`route: ${this.route}`, pagesDir, pagesBuildDir);
  }

  /**
   * @returns { boolean } Wether or not the navigable page is valid
   */
  validate() {
    return !!this.route?.trim() && super.validate();
  }

  /**
   * Checks if the navigable page is duplicated.
   * If it is, returns the files of the duplicated page;
   * otherwise it returns an empty array
   * @param { { [dataId: string]: NavigablePage } } currentNavigablePagesData - Discovered navigable pages data at the time
   * @param { string } dataId - Navigable page data id
   * @returns { string[] } Files of the duplicated page
   */
  checkIfDuplicated(currentDiscoveredNavigablePages, dataId) {
    const previousPage = currentDiscoveredNavigablePages[dataId];

    if (!this.validate() || !previousPage) {
      return [];
    }

    return getDuplicatedPageFiles(previousPage, this);
  }

  /**
   * Throws duplicated route error by file type extension or index naming convention.
   * @param { ThrowDuplicatedPageParams } throwDuplicatedPageParams
   *
   * @example
   * this.route = '/home';
   * const currentFile = 'home/index.mobile.js';
   * const duplicatedFile = 'home.mobile.js';
   * const dir = 'app/nordic-pages';
   *
   * throwDuplicated({ currentFile, duplicatedFile, dir, })
   * => Duplicated navigable page files by index naming convention
   */
  throwDuplicated({ baseDir, currentFile, previousFile }) {
    throwDuplicatedPage({
      baseDir,
      previousFile,
      currentFile,
      pageLogPortion: `${this.route} route`,
    });
  }

  /**
   * Adds each discovered file of the page to its corresponding location
   * @param { AddPageFilesParams } addPageFilesParams
   */
  addPageFiles({ files, pageInfoByViewFile, pagesDir, pagesBuildDir }) {
    super.addPageFiles({
      files,
      pageInfoByViewFile,
      pagesDir,
      pagesBuildDir,
    });

    (files[PAGE_MEMBERS.hooks] || []).forEach((pageHooks) => {
      this.setHooksFile(pageHooks, pagesDir, pagesBuildDir);
    });
  }

  /**
   * @returns { string[] } List of all page files use in paths objects
   */
  getAllFilesOfPage() {
    const pageFiles = super.getAllFilesOfPage();

    if (!this.hooks) {
      return pageFiles;
    }

    pageFiles.push(this.hooks);

    return pageFiles;
  }
}

module.exports = { NavigablePage };
