// nordic-pages-discovery/lib/page/paths.js
const { DEVICE_TYPES } = require('../constants/device');
const {
  SCRIPT_OR_STYLES_EXTENSIONS_REGEX,
  ENDS_WITH_STYLES_REGEX,
  ENDS_WITH_STYLES_DEVICE_REGEX,
  STARTS_WITH_STYLES_REGEX,
  STARTS_WITH_STYLES_DEVICE_REGEX,
} = require('../constants/regexs');
const { getTranspiledPath } = require('../helpers/build');
const {
  DuplicatedFileByFileTypeError,
} = require('../error/duplicated-file-by-file-type-error');
const { getFileFullPath } = require('../helpers/file/path-detail');
const {
  getNoBaseViewForDeviceStyleSheetLog,
} = require('../helpers/file/inconsistency');

/**
 * @typedef { typeof import('../constants/page').PAGE_MEMBERS.view } PageViewMember
 * @typedef { typeof import('../constants/page').PAGE_MEMBERS.styles } PageStylesMember
 * @typedef { PageViewMember | PageStylesMember } PagePathsMemberType
 * @typedef { keyof DEVICE_TYPES } DeviceType
 * @typedef { import('../helpers/page-info').HydrateValidValues } HydrateValidValues
 *
 * @typedef { object } PathsProps
 * @property { string } [client]
 * @property { string } [view]
 * @property { string } [styles]
 * @property { string | null } [transpiledView]
 * @property { string[] | undefined } [islands]
 * @property { HydrateValidValues } [hydrate]
 * @property { boolean } [canUseBootstrapPage]
 *
 * @typedef { { [device in DeviceType]?: Paths } } PagePaths
 *
 * @typedef { object } FileToAddData
 * @property { string } filePath - Relative path of the file to be added
 * @property { PagePathsMemberType } pageMember - Page member to which the file belongs
 * @property { string } dir - Directory use to get the file full path
 */

/**
 * Page paths object for a certain device
 */
class Paths {
  /**
   * @param { PathsProps } [props={}]
   */
  constructor({
    client = '',
    view = '',
    styles = '',
    hydrate = true,
    canUseBootstrapPage = false,
    transpiledView = null,
    islands = [],
  } = {}) {
    /**
     * @type {string}
     */
    this.view = view;
    /**
     * @type {string}
     */
    this.styles = styles;
    /**
     * @type {string}
     */
    this.client = client;
    /**
     * @type {string | null}
     */
    this.transpiledView = transpiledView;
    /**
     * @type {string[]}
     */
    this.islands = islands;
    /**
     * @type { HydrateType }
     */
    this.hydrate = hydrate;
    /**
     * @type {boolean}
     */
    this.canUseBootstrapPage = canUseBootstrapPage;
  }

  /**
   * Add file to paths only if no other file of the same page member was assign previously,
   * if it was, a duplicated file error is thrown.
   * @param {FileToAddData} fileToAddData
   */
  addPagePath({ filePath, dir, pageMember }) {
    const fileFullPath = getFileFullPath(filePath, dir);
    const isDuplicatedPagePart = !!this[pageMember];

    if (!isDuplicatedPagePart) {
      this[pageMember] = fileFullPath;

      this.setClientIfApplies(filePath);

      return;
    }

    throw new DuplicatedFileByFileTypeError({
      currentFile: filePath,
      previousFile: this[pageMember],
      baseDir: dir,
    });
  }

  /**
   * @private
   * Set the client name only if client has no value
   * @param {string} filePath - Relative path of file use to define the client name
   */
  setClientIfApplies(filePath) {
    if (this.client) {
      return;
    }

    this.client = filePath
      .replace(SCRIPT_OR_STYLES_EXTENSIONS_REGEX, '')
      .replace(STARTS_WITH_STYLES_REGEX, 'index')
      .replace(ENDS_WITH_STYLES_REGEX, '/index')
      .replace(STARTS_WITH_STYLES_DEVICE_REGEX, 'index.')
      .replace(ENDS_WITH_STYLES_DEVICE_REGEX, '/index.')
      .replace(/\//g, '-')
      .replace(/\[|\]/g, '')
      .replace(/\.\.\./g, '___');
  }

  /**
   * Completes the paths object with other page paths information.
   * If it cannot complete it, an error is thrown
   * @param { DeviceType } device - Device of the current paths
   * @param { PagePaths } pagePaths - Paths of the page by device
   */
  complete(device, pagePaths) {
    if (device === DEVICE_TYPES.webview && pagePaths?.mobile?.view) {
      this.view = pagePaths.mobile.view;
      this.islands = pagePaths.mobile.islands;
      this.hydrate = pagePaths.mobile.hydrate;
      this.canUseBootstrapPage = pagePaths.mobile.canUseBootstrapPage;

      return;
    }

    if (pagePaths?.base?.view) {
      this.view = pagePaths.base.view;
      this.islands = pagePaths.base.islands;
      this.hydrate = pagePaths.base.hydrate;
      this.canUseBootstrapPage = pagePaths.base.canUseBootstrapPage;

      return;
    }

    throw new Error(getNoBaseViewForDeviceStyleSheetLog(this.styles));
  }

  /**
   * @param { string } pagesDir
   * @param { string } pagesBuildDir
   */
  fillTranspiledView(pagesDir, pagesBuildDir) {
    const equivalentView = getTranspiledPath(
      this.view,
      pagesDir,
      pagesBuildDir,
    );

    if (!equivalentView) {
      return;
    }

    this.transpiledView = equivalentView;
  }
}

module.exports = { Paths };
