// @nordic/rendering-context/src/server/layout/fonts.ts
import type { LayoutViewHeadSlots } from './types';

import { getFonts as getGlobalThemeFonts } from '@global-theme/context/server';

const FONTS_CSS_COMMENT =
  '/*! * Web Fonts from Fontspring.com * * All OpenType features and all extended glyphs have been removed. *' +
  ' Fully installable fonts can be purchased at https://www.fontspring.com * *' +
  ' The fonts included in this stylesheet are subject to the End User License you purchased * from Fontspring.' +
  ' The fonts are protected under domestic and international trademark and * copyright law.' +
  ' You are prohibited from modifying, reverse engineering, duplicating, or * distributing this font software.' +
  ' * * (c) 2010-2023 Fontspring * * * * * The fonts included are copyrighted by the vendor listed below.' +
  ' * * Vendor: Mark Simonson Studio * License URL: https://www.fontspring.com/licenses/mark-simonson-studio/webfont * * */';

const FONTS_DATA = [
  { name: 'light', weight: 300 },
  { name: 'regular', weight: 400 },
  { name: 'semibold', weight: 600 },
];

export async function getFontResources(cdn: string) {
  const links: LayoutViewHeadSlots['links'] = [];
  const globalThemeFonts = await getGlobalThemeFonts(cdn);

  if (globalThemeFonts) {
    globalThemeFonts.preload?.forEach((href) => {
      links.push({
        href,
        rel: 'preload',
        as: 'font',
        type: 'font/woff2',
        crossOrigin: 'anonymous',
      });
    });

    return { links, style: globalThemeFonts.css };
  }

  const css = FONTS_DATA.reduce((acc, { name, weight }) => {
    const url = `${cdn}/ui/webfonts/v3.0.0/proxima-nova/proximanova-${name}`;

    links.push({
      href: `${url}.woff2`,
      rel: 'preload',
      as: 'font',
      type: 'font/woff2',
      crossOrigin: 'anonymous',
    });

    return (
      `${acc}@font-face{font-family:'Proxima Nova';font-weight:${weight};font-display:swap;font-style:normal;` +
      `src:url(${url}.woff2) format("woff2"),url(${url}.woff) format("woff")}`
    );
  }, '');

  return {
    links,
    style: { id: '__FONTS__', children: `${FONTS_CSS_COMMENT} ${css}` },
  };
}
