// @nordic/rendering-context/src/server/layout/types.ts
import type {
  ComponentType,
  LinkHTMLAttributes,
  MetaHTMLAttributes,
  PropsWithChildren,
} from 'react';

export type LayoutViewSlotData<T = Record<string, unknown>> =
  | T
  | false
  | null
  | undefined;

export type LayoutViewSlotProps<T> =
  | Array<LayoutViewSlotProps<T>>
  | LayoutViewSlotData<T>
  | (() => Array<LayoutViewSlotProps<T>> | LayoutViewSlotData<T>);

export type LayoutViewHeadSlots<TScript = unknown, TStyle = unknown> = {
  title: ComponentType;
  meta: Array<LayoutViewSlotProps<MetaHTMLAttributes<HTMLMetaElement>>>;
  links: Array<LayoutViewSlotProps<LinkHTMLAttributes<HTMLLinkElement>>>;
  styles: Array<LayoutViewSlotProps<TStyle>>;
  scripts: Array<LayoutViewSlotProps<TScript>>;
  noscript?: string;
};

export type LayoutViewBodySlots<TScript = unknown> = {
  start: Array<ComponentType>;
  main: ComponentType<PropsWithChildren>;
  end: Array<ComponentType>;
  scripts: Array<LayoutViewSlotProps<TScript>>;
};

export type LayoutViewStore<TScript = unknown, TStyle = unknown> = {
  lang: string;
  headSlots: LayoutViewHeadSlots<TScript, TStyle>;
  bodySlots: LayoutViewBodySlots<TScript>;
};
