// nordic-pages-discovery/lib/islands/get-islands-from-file/imported-components/dynamic-import-declaration/index.js
const traverse = require('@babel/traverse').default;
const babelTypes = require('@babel/types');

const { IMPORT_TYPES } = require('../../../../constants/islands');
const {
  getComponentTypeByPath,
} = require('../../../../helpers/islands/component');
const {
  getScriptFullPathWithExtension,
} = require('../../../../helpers/file/path-detail');
const { searchImportName } = require('./import-name');

/**
 * @param {babelTypes.NodePath} path
 * @param {babelTypes.Identifier} nodeId
 * @param {babelTypes.CallExpression} init
 * @returns {boolean} Is dynamic import declaration usage
 */
const isPotentialDynamicImportDeclarationUsage = (path, nodeId, init) => {
  if (
    !(
      babelTypes.isProgram(path.parentPath.parent) ||
      babelTypes.isProgram(path.parentPath.parentPath.parent)
    ) ||
    !babelTypes.isIdentifier(nodeId) ||
    !babelTypes.isCallExpression(init) ||
    !(
      babelTypes.isIdentifier(init.callee) ||
      babelTypes.isMemberExpression(init.callee)
    )
  ) {
    return false;
  }

  const [firstArg, secondArg, thirdArg] = init.arguments;

  if (
    !firstArg ||
    thirdArg ||
    (!babelTypes.isArrowFunctionExpression(firstArg) &&
      !babelTypes.isFunctionExpression(firstArg)) ||
    (secondArg && !babelTypes.isObjectExpression(secondArg))
  ) {
    return false;
  }

  return true;
};

/**
 * @param {babelTypes.NodePath } nodePath
 * @param {{ file: string, dir: string }} origin
 * @returns {import('../../../../helpers/islands/component').ComponentManifest | {}} Dynamic import component info
 */
const getPotentialDynamicImportDeclarationUsage = (nodePath, origin) => {
  let dynamicImportPath = null;
  /** @type {babelTypes.ArrowFunctionExpression | babelTypes.FunctionExpression | null | undefined} */
  let resolveComponentsNode = null;

  if (
    !babelTypes.isVariableDeclarator(nodePath.node) &&
    !babelTypes.isAssignmentExpression(nodePath.node)
  ) {
    return {};
  }

  const init = nodePath.node.init || nodePath.node.right;
  const nodeId = nodePath.node.id || nodePath.node.left;

  traverse(init, {
    noScope: true,
    Import(path) {
      // is "import('path')"
      if (
        !babelTypes.isStringLiteral(path.parent.arguments[0]) ||
        getComponentTypeByPath(path.parent.arguments[0].value) !==
          IMPORT_TYPES.RELATIVE
      ) {
        return;
      }

      dynamicImportPath = getScriptFullPathWithExtension(
        path.parent.arguments[0].value,
        origin.dir,
      );

      if (
        path.parentPath &&
        babelTypes.isMemberExpression(path.parentPath.parent) &&
        babelTypes.isIdentifier(path.parentPath.parent.property) &&
        path.parentPath.parent.property.name === 'then' &&
        babelTypes.isCallExpression(path.parentPath.parentPath.parent)
      ) {
        [resolveComponentsNode] = path.parentPath.parentPath.parent.arguments;
      }

      path.stop();
    },
  });

  if (!dynamicImportPath) {
    return {};
  }

  let moduleName;
  let originalFunctionName;

  if (babelTypes.isIdentifier(init.callee)) {
    const binding = nodePath.scope.getBinding(init.callee.name);

    if (binding?.path && babelTypes.isImportSpecifier(binding.path)) {
      moduleName = binding.path.parent.source.value;
      originalFunctionName =
        binding.path.node.imported.name || binding.path.node.imported;
    }
  } else if (babelTypes.isMemberExpression(init.callee)) {
    const binding = nodePath.scope.getBinding(init.callee.object.name);

    if (binding?.path && babelTypes.isImportDefaultSpecifier(binding.path)) {
      moduleName = binding.path.parent.source.value;
      originalFunctionName = init.callee.property.name;
    }
  }

  /** @type {import('../../../../helpers/islands/component').Component} */
  const component = {
    importName: IMPORT_TYPES.DEFAULT,
    path: dynamicImportPath,
    namedExport: false,
    isIsland: false,
    importDynamicCaller: {
      module: moduleName,
      type: originalFunctionName,
      file: origin.file,
    },
  };

  const [, secondArg] = init.arguments;

  if (!resolveComponentsNode && secondArg) {
    resolveComponentsNode = secondArg.properties.find(
      (property) =>
        babelTypes.isObjectProperty(property) &&
        babelTypes.isIdentifier(property.key) &&
        (babelTypes.isArrowFunctionExpression(property.value) ||
          babelTypes.isFunctionExpression(property.value)) &&
        property.key.name === 'resolveComponent',
    )?.value;
  }

  if (resolveComponentsNode) {
    const importName = searchImportName(resolveComponentsNode);

    if (importName) {
      component.importName = importName;
      component.namedExport = true;
    }
  }

  return {
    [nodeId.name]: component,
  };
};

module.exports = {
  isPotentialDynamicImportDeclarationUsage,
  getPotentialDynamicImportDeclarationUsage,
};
