// @global-theme/context/dist/cjs/server.production.js
'use strict';

var storage = require('./storage-7GJlLLs9.js');
var serialize = require('serialize-javascript');
require('node:async_hooks');

const SERIALIZE_OPTIONS = { isJSON: true };
/**
 * Safely serializes a value to JSON for embedding in HTML scripts.
 * Uses serialize-javascript to prevent XSS and script injection.
 *
 * @param data - The value to serialize
 * @returns Safe JSON string for script embedding
 */
function serializeJSON(data) {
    return serialize(data, SERIALIZE_OPTIONS);
}

/**
 * Generates the global theme context initialization script for server-side rendering.
 * This script should be embedded in the HTML to hydrate the client-side context.
 *
 * @returns Script object with id and serialized theme context, or undefined if context is not initialized
 */
async function getGlobalThemeContextScript() {
    const context = storage.getContext();
    if (!context) {
        return undefined;
    }
    const theme = await context.theme;
    const fonts = await context.fonts;
    const supportsSuperscript = context.supportsSuperscript ??
        !fonts.some(({ name }) => name.toLowerCase().replace(/\s+/g, '') === 'proximanova');
    const contextScript = `window._gt={ctx:${serializeJSON({ theme, supportsSuperscript })}};`;
    return {
        id: '__GLOBAL_THEME_CTX__',
        children: contextScript,
    };
}

const PROXIMA_NOVA_FONTS_CSS_COMMENT = '/*! * Web Fonts from Fontspring.com * * All OpenType features and all extended glyphs have been removed. *' +
    ' Fully installable fonts can be purchased at https://www.fontspring.com * *' +
    ' The fonts included in this stylesheet are subject to the End User License you purchased * from Fontspring.' +
    ' The fonts are protected under domestic and international trademark and * copyright law.' +
    ' You are prohibited from modifying, reverse engineering, duplicating, or * distributing this font software.' +
    ' * * (c) 2010-2023 Fontspring * * * * * The fonts included are copyrighted by the vendor listed below.' +
    ' * * Vendor: Mark Simonson Studio * License URL: https://www.fontspring.com/licenses/mark-simonson-studio/webfont * * */';

async function getFonts(cdn) {
    const context = storage.getContext();
    if (!context) {
        return undefined;
    }
    const fonts = await context.fonts;
    if (fonts.length === 0) {
        return {};
    }
    const preloadFonts = new Set();
    const cssArray = [];
    fonts.forEach(({ weights }) => {
        weights.forEach(({ getUrl, getFontFaceStyle }) => {
            preloadFonts.add(getUrl(cdn));
            cssArray.push(getFontFaceStyle(cdn));
        });
    });
    const css = cssArray.join('');
    // TODO: Improve this handling of fonts comments
    const hasProximaNova = fonts.some(({ name }) => name.toLowerCase().replace(/\s+/g, '') === 'proximanova');
    context.supportsSuperscript = !hasProximaNova;
    return {
        css: {
            id: '__FONTS__',
            children: hasProximaNova
                ? `${PROXIMA_NOVA_FONTS_CSS_COMMENT} ${css}`
                : css,
        },
        preload: preloadFonts,
    };
}
async function getStyles() {
    const context = storage.getContext();
    if (!context) {
        return undefined;
    }
    const styles = await context.styles;
    const themeMetadata = await context.theme;
    context.themeCSS = styles.children;
    context.themeMetadata = themeMetadata;
    context.themeMode = themeMetadata?.mode;
    return styles;
}
function getThemeCSS() {
    const context = storage.getContext();
    if (!context) {
        return undefined;
    }
    return context.themeCSS;
}
function getThemeMode() {
    const context = storage.getContext();
    if (!context) {
        return undefined;
    }
    return context.themeMode;
}

exports.getContext = storage.getContext;
exports.runWithGlobalThemeContext = storage.runWithGlobalThemeContext;
exports.getFonts = getFonts;
exports.getGlobalThemeContextScript = getGlobalThemeContextScript;
exports.getStyles = getStyles;
exports.getThemeCSS = getThemeCSS;
exports.getThemeMode = getThemeMode;
