// @andes/checkbox/build/cjs/index.production.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _extends = require('@babel/runtime/helpers/extends');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var React = require('react');
var classNames = require('classnames');
var context = require('@andes/context');
var visuallyHidden = require('@andes/visually-hidden');
var themes = require('@andes/themes');
var FeedbackPositive16 = require('@andes/icons/FeedbackPositive16');
var computeDataAndesState = require('@andes/common/build/utils/computeDataAndesState');
var warning = require('@andes/common/build/utils/warning');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var FeedbackPositive16__default = /*#__PURE__*/_interopDefaultLegacy(FeedbackPositive16);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
const IndeterminateIcon = () => /*#__PURE__*/React__default["default"].createElement("svg", {
  "aria-hidden": "true",
  width: "16",
  height: "16",
  viewBox: "0 0 16 16",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/React__default["default"].createElement("rect", {
  width: "10",
  height: "1.6",
  x: "3",
  y: "7.2",
  fill: "currentColor"
}));
var IndeterminateIcon$1 = IndeterminateIcon;
const namespace = 'andes-checkbox';
const WARNING_LABEL = `<Checkbox> needs some plain text contents for accessibility reasons.
Please add a \`label\` or \`srLabel\` prop.`;
const LABEL_POSITION_LEFT = 'left';
const Checkbox = /*#__PURE__*/function () {
  const Checkbox = props => {
    const checked = props.checked,
      className = props.className,
      defaultChecked = props.defaultChecked,
      _props$disabled = props.disabled,
      disabled = _props$disabled === void 0 ? false : _props$disabled,
      _props$error = props.error,
      error = _props$error === void 0 ? false : _props$error,
      _props$highlight = props.highlight,
      highlight = _props$highlight === void 0 ? false : _props$highlight,
      id = props.id,
      _props$indeterminate = props.indeterminate,
      indeterminate = _props$indeterminate === void 0 ? false : _props$indeterminate,
      _props$inputProps = props.inputProps,
      inputProps = _props$inputProps === void 0 ? {} : _props$inputProps,
      label = props.label,
      _props$labelPosition = props.labelPosition,
      labelPosition = _props$labelPosition === void 0 ? 'right' : _props$labelPosition,
      onChange = props.onChange,
      _props$srHidden = props.srHidden,
      srHidden = _props$srHidden === void 0 ? false : _props$srHidden,
      srLabel = props.srLabel,
      ref = props.ref;
    const _useThemeContext = themes.useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const _useRef = React.useRef(typeof checked !== 'undefined'),
      isControlled = _useRef.current;
    const _useState = React.useState(defaultChecked || checked),
      _useState2 = _slicedToArray__default["default"](_useState, 2),
      isChecked = _useState2[0],
      setIsChecked = _useState2[1];
    warning__default["default"](Boolean(label || srLabel || srHidden), WARNING_LABEL);
    React.useEffect(() => {
      if (isControlled && checked !== undefined) {
        setIsChecked(checked);
      }
    }, [checked, isControlled]);
    const componentId = context.useId(id);
    const srLabelId = `${componentId}-srLabel`;
    const inputRef = React.useRef(null);
    const handleChange = event => {
      if (!isControlled) setIsChecked(event.target.checked);
      if (typeof onChange === 'function') onChange(event);
    };
    const checkboxClassnames = classNames__default["default"](namespace, className, {
      [`${namespace}--error`]: error,
      [`${namespace}--highlight`]: highlight,
      [`${namespace}--label-left`]: labelPosition === LABEL_POSITION_LEFT,
      [`${namespace}--indeterminate`]: indeterminate && isChecked,
      [`${namespace}--disabled`]: disabled
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('checkbox'));
    const labelClassnames = classNames__default["default"](`${namespace}__label`, {
      [`${namespace}__label-text`]: label
    });
    const Wrapper = label ? 'label' : 'div';
    React.useEffect(() => {
      if (inputRef !== null && inputRef !== void 0 && inputRef.current) {
        inputRef.current.indeterminate = indeterminate && isChecked || false;
      }
    }, [inputRef, indeterminate, isChecked]);
    return /*#__PURE__*/React__default["default"].createElement(Wrapper, _extends__default["default"]({
      className: checkboxClassnames,
      ref: ref,
      "data-andes-checkbox": true,
      "data-andes-checkbox-variant": highlight ? 'highlight' : undefined
    }, computeDataAndesState.computeDataAndesState({
      checked: isChecked,
      disabled,
      indeterminate: indeterminate && isChecked,
      error: error && !isChecked
    }), themeHash), /*#__PURE__*/React__default["default"].createElement("span", {
      className: `${namespace}__checkbox`,
      "data-andes-checkbox-container": true
    }, /*#__PURE__*/React__default["default"].createElement("input", _extends__default["default"]({}, inputProps, {
      ref: inputRef,
      checked: checked,
      className: `${namespace}__input`,
      defaultChecked: defaultChecked,
      disabled: disabled,
      id: componentId,
      onChange: handleChange,
      type: "checkbox"
    }, srLabel ? {
      'aria-labelledby': srLabelId
    } : {}, {
      "data-andes-checkbox-input": true
    })), /*#__PURE__*/React__default["default"].createElement("span", {
      className: `${namespace}__icon`
    }, indeterminate && isChecked ? /*#__PURE__*/React__default["default"].createElement(IndeterminateIcon$1, null) : /*#__PURE__*/React__default["default"].createElement(FeedbackPositive16__default["default"], {
      color: "currentColor"
    }))), label && /*#__PURE__*/React__default["default"].createElement("span", {
      "aria-hidden": !!srLabel || undefined,
      className: labelClassnames,
      "data-andes-checkbox-label": true
    }, label), srLabel && /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, {
      id: srLabelId
    }, srLabel));
  };
  Checkbox.displayName = 'Checkbox';
  return Checkbox;
}();
var Checkbox$1 = Checkbox;
exports.Checkbox = Checkbox$1;
