// @nordic/context/dist/cjs/index.browser.development.js
'use strict';

var coreContext = require('@nordic/core-context');
var renderingContext = require('@nordic/rendering-context');
var react = require('react');
var freeze = require('@nordic/utils/freeze');
var validation = require('@nordic/utils/validation');
var error = require('@nordic/utils/error');

/**
 * Returns the prefix for the assets for the current page.
 *
 * **⚠️ WARNING**: This function only works within React components, rendering stage, or client-side.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getassetsprefix | Documentation}
 *
 * @example
 *
 * import { getAssetsPrefix } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const assetsPrefix = getAssetsPrefix();
 *
 *  return (
 *    <h1>
 *      Assets Prefix: {assetsPrefix}
 *    </h1>
 *  );
 * };
 */
function getAssetsPrefix() {
  return renderingContext.getRenderingAssets().prefix;
}
/**
 * Returns the generated CSRF token for the current page.
 *
 * **⚠️ WARNING**: This function only works within React components, rendering stage, or client-side.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getcsrftoken | Documentation}
 *
 * @example
 *
 * import { getCsrfToken } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const csrfToken = getCsrfToken();
 *
 *  return (
 *    <h1>
 *      CSRF Token: {csrfToken}
 *    </h1>
 *  );
 * };
 */
const getCsrfToken = renderingContext.getRenderingCsrfToken;
/**
 * Returns the generated nonce token for the current page.
 *
 * **⚠️ WARNING**: This function only works within React components, rendering stage, or client-side.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getnonce | Documentation}
 *
 * @example
 *
 * import { getNonce } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const nonce = getNonce();
 *
 *  return (
 *    <h1>
 *      Nonce: {nonce}
 *    </h1>
 *  );
 * };
 */
const getNonce = renderingContext.getRenderingNonce;
/**
 * Provides the device type anywhere, including React Components.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getdevicetype | Documentation}
 *
 * @example
 *
 * import { getDeviceType } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const deviceType = getDeviceType();
 *
 *  return (
 *    <h1>
 *      Device Type: {deviceType}
 *    </h1>
 *  );
 * };
 */
const getDeviceType = coreContext.getDeviceType;
/**
 * Provides the platform type anywhere, including React Components.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getplatformtype | Documentation}
 *
 * @example
 *
 * import { getPlatformType } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const platformType = getPlatformType();
 *
 *  return (
 *    <h1>
 *      Platform Type: {platformType}
 *    </h1>
 *  );
 * };
 */
const getPlatformType = coreContext.getPlatformType;
/**
 * Provides the webview information anywhere, including React Components.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#getwebviewinfo | Documentation}
 *
 * @example
 *
 * import { getWebviewInfo } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const webviewInfo = getWebviewInfo();
 *
 *  return (
 *    <h1>
 *      Webview Info: {JSON.stringify(webviewInfo)}
 *    </h1>
 *  );
 * };
 */
const getWebviewInfo = coreContext.getWebviewInfo;

class NordicAssetNotFoundError extends error.OwnerStackError {
  constructor(assetName) {
    super(`Asset "${assetName}" not found.` + ' Please make sure that you have built all your assets before running your app,' + ' or check if you are trying to load a non-existing bundle.');
    this.name = 'NordicAssetNotFoundError';
  }
}

class NordicContextError extends error.OwnerStackError {
  name = 'NordicContextError';
}

/**
 * Returns the props of the current page, defined in the `getServerSideProps()` function.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#usepageprops | Documentation}
 *
 * @example
 *
 * import { usePageProps } from 'nordic/context';
 *
 * const FooComponent = () => {
 *  const { foo } = usePageProps();
 *
 *  return (
 *    <h1>
 *      Foo value: {foo}
 *    </h1>
 *  );
 * };
 */
function usePageProps() {
  const flags = renderingContext.getRenderingFlags();
  if (!validation.isTestEnv() && !validation.isStorybook() && !flags.staticMarkup && !flags.stream && !renderingContext.getRenderingMainEntry()) {
    throw new NordicContextError('`usePageProps()` requires the `layout.mainEntry` option. Check the render options.');
  }
  return renderingContext.getRenderingAppProps()?.pageProps || {};
}
/**
 * Provides access to the configured assets with prefix.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#useassetsprefix | Documentation}
 *
 * @example
 *
 * import { useAssetsPrefix } from 'nordic/context';
 *
 * const SomeComponent = () => {
 *  const assetsPrefix = useAssetsPrefix();
 *
 *  return (
 *    <h1>
 *      Asset with prefix: {assetsPrefix.join('some-asset.png')}
 *    </h1>
 *  );
 * };
 */
function useAssetsPrefix() {
  return react.useMemo(() => {
    const assets = renderingContext.getRenderingAssets();
    return freeze.freezeObject({
      value: assets.prefix,
      join: assetName => {
        if (validation.isExternalUrl(assetName)) {
          return assetName;
        }
        const fileName = assets.manifest.get(assetName);
        const fileNameWithPrefix = assets.prefix + (fileName || assetName);
        // isTestEnv()` is only for development code, to allow unit tests execution without app build
        if (!fileName && !validation.isTestEnv() && !validation.isStorybook() && renderingContext.getRenderingFramework() === 'nordic') {
          throw new NordicAssetNotFoundError(assetName);
        }
        return fileNameWithPrefix;
      }
    });
  }, []);
}

/**
 * Returns whether the current context is a webview.
 *
 * @see {@link https://nordic.adminml.com/docs/api/Nordic%20Modules/nordic-context#iswebview | Documentation}
 *
 * @example
 *
 * import { isWebview } from 'nordic/context';
 *
 * const SomeComponent = () => (
 *   <h1>
 *     Is Webview: {isWebview() ? 'Yes' : 'No'}
 *   </h1>
 * );
 */
const isWebview = () => !!getWebviewInfo();

exports.getAssetsPrefix = getAssetsPrefix;
exports.getCsrfToken = getCsrfToken;
exports.getDeviceType = getDeviceType;
exports.getNonce = getNonce;
exports.getPlatformType = getPlatformType;
exports.getWebviewInfo = getWebviewInfo;
exports.isWebview = isWebview;
exports.useAssetsPrefix = useAssetsPrefix;
exports.usePageProps = usePageProps;
