// nordic-pages-discovery/lib/islands/get-islands-from-file/imported-components/index.js
const traverse = require('@babel/traverse').default;

const { APP_FILE_MEMBERS } = require('../../../constants/file');
const { NordicIslandsError } = require('../../../error/nordic-islands-error');
const { getFileDirname } = require('../../../helpers/file/path-detail');
const {
  getLookingForIslandsAndComponentsError,
} = require('../../../error/get-discovery-error');
const { getImportDeclarationUsage } = require('./import-declaration');
const {
  isPotentialDynamicImportDeclarationUsage,
  getPotentialDynamicImportDeclarationUsage,
} = require('./dynamic-import-declaration');
const { isImportType } = require('../../../helpers/babel/typescript');

/**
 * @typedef {{ importName: string, path: string, namedExport: boolean, islands: import('../../../cache/imports-cache').ComponentIslands | undefined }} ComponentInstance
 * @typedef {Record<string, ComponentInstance>} ComponentManifest
 */

/**
 * @param {string} filePath
 * @param {import('../../../cache/ast-cache').BabelAST} ast
 * @returns {{ importedComponents: ComponentManifest | {}, islandWrapper: string? }}
 */
const getImportedComponentsFromFile = (filePath, ast) => {
  try {
    const originDir = getFileDirname(filePath);
    const importedComponents = {};

    let islandWrapper = null;

    traverse(ast, {
      noScope: true,
      ImportDeclaration(path) {
        const { appComponentsUsage, islandWrapperUsage } =
          getImportDeclarationUsage(
            {
              source: path.node.source,
              specifiers: path.node.specifiers,
              isType: isImportType(path.node),
            },
            originDir,
          );

        Object.assign(importedComponents, appComponentsUsage);

        islandWrapper = islandWrapperUsage || islandWrapper;
      },
      VariableDeclarator(path) {
        if (
          !isPotentialDynamicImportDeclarationUsage(
            path,
            path.node.id,
            path.node.init,
          )
        ) {
          return;
        }

        // example 1: const MyComponent = loadable(() => import('./my-component'));
        // example 2: const MyComponent = loadable(() => { return import('./my-component'); });
        // example 3: const MyComponent = loadable(function() { return import('./my-component'); });
        // example 4: const MyComponent = loadable(function fn() { return import('./my-component'); });
        // example 5: const MyComponent = loadable(() => import('./my-component').then(...));
        // example 6: const MyComponent = loadable(() => { return import('./my-component').then(...); });
        // example 7: const MyComponent = loadable(function() { return import('./my-component').then(...); });
        // example 8: const MyComponent = loadable(function fn() { return import('./my-component').then(...); });
        const appComponentUsage = getPotentialDynamicImportDeclarationUsage(
          path,
          { file: filePath, dir: originDir },
        );

        Object.assign(importedComponents, appComponentUsage);
      },
    });

    return { importedComponents, islandWrapper };
  } catch (error) {
    throw new NordicIslandsError(
      getLookingForIslandsAndComponentsError(error.message),
      {
        type: `${APP_FILE_MEMBERS.PAGE} or ${APP_FILE_MEMBERS.COMPONENT}`,
        path: filePath,
      },
      error,
    );
  }
};

module.exports = {
  getImportedComponentsFromFile,
};
