// @andes/floating-menu/build/esm/module.js
import _extends from '@babel/runtime/helpers/extends';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { useMemo, Children, useState, useEffect, useCallback, cloneElement, useRef } from 'react';
import classNames from 'classnames';
import { useAndesContext, useId } from '@andes/context';
import { useThemeContext } from '@andes/themes';
import { Popper } from '@andes/popper';
import { withTracker } from '@andes/with-tracker';
import { SPACE, ENTER, ARROW_UP, ARROW_DOWN, ESCAPE, TAB } from '@andes/common/build/constants/keyboardCodes';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import includes from '@andes/common/build/utils/includes';
import usePreventScroll from '@andes/common/build/hooks/usePreventScroll';
import { FocusRing, FocusScope } from '@react-aria/focus';
import { VisuallyHidden, VisuallyHiddenFocusable } from '@andes/visually-hidden';
import warning from '@andes/common/build/utils/warning';
import Close16 from '@andes/icons/Close16';
import Search16 from '@andes/icons/Search16';
import Close20 from '@andes/icons/Close20';
import Search20 from '@andes/icons/Search20';
import { Checkbox } from '@andes/checkbox';
import { ListBox } from '@andes/listbox';
import { List, ListGroup, ListActionItem, ListItem } from '@andes/list';
import { senderMetrics, trackCustomization } from '@andes/common/build/metrics';
import groupNodeGenerator from '@andes/list/build/generators/groupNodeGenerator';
import itemNodeGenerator from '@andes/list/build/generators/itemNodeGenerator';
const ITEM_DISPLAY_NAME = 'FloatingMenuItem';
const GROUP_DISPLAY_NAME = 'FloatingMenuGroup';
const SUPPORTED_GROUP_TYPES = ['FloatingMenuGroup', 'DropdownGroup'];
const TYPE_MENU = 'menu';
const TYPE_SELECT = 'select';
const SEARCH_BOX_SIZE = 52;
const LIST_ITEM_SIZE = 48;
const listHoverClass = 'andes-floating-menu-item-hover';
const useFloatingMenu = (triggerRef, menuRef, searchBoxRef, componentProps) => {
  const allowUnselect = componentProps.allowUnselect,
    bottomContent = componentProps.bottomContent,
    children = componentProps.children,
    defaultOpen = componentProps.defaultOpen,
    defaultValue = componentProps.defaultValue,
    listId = componentProps.listId,
    menuMaxHeight = componentProps.menuMaxHeight,
    multiple = componentProps.multiple,
    offset = componentProps.offset,
    onChange = componentProps.onChange,
    onClose = componentProps.onClose,
    onOpen = componentProps.onOpen,
    open = componentProps.open,
    popperOptions = componentProps.popperOptions,
    searchConfig = componentProps.searchConfig,
    style = componentProps.style,
    topContent = componentProps.topContent,
    type = componentProps.type,
    value = componentProps.value,
    width = componentProps.width;
  const isGroup = useMemo(() => {
    var _firstOption$type;
    if (!children) return false;
    const firstOption = Children.toArray(children)[0];
    return includes(SUPPORTED_GROUP_TYPES, firstOption === null || firstOption === void 0 || (_firstOption$type = firstOption.type) === null || _firstOption$type === void 0 ? void 0 : _firstOption$type.displayName);
  }, [children]);
  const _useState = useState(value || defaultValue),
    _useState2 = _slicedToArray(_useState, 2),
    currentValues = _useState2[0],
    setCurrentValues = _useState2[1];
  const _useState3 = useState(false),
    _useState4 = _slicedToArray(_useState3, 2),
    isMenuVisible = _useState4[0],
    setIsMenuVisible = _useState4[1];
  const _useState5 = useState(undefined),
    _useState6 = _slicedToArray(_useState5, 2),
    searchItemFocused = _useState6[0],
    setSearchItemFocused = _useState6[1];
  const _useState7 = useState(typeof searchConfig === 'object' && (searchConfig === null || searchConfig === void 0 ? void 0 : searchConfig.defaultValue)),
    _useState8 = _slicedToArray(_useState7, 2),
    searchText = _useState8[0],
    setSearchText = _useState8[1];
  usePreventScroll({
    isDisabled: type === TYPE_MENU || type === TYPE_SELECT && !isMenuVisible
  });
  const isUncontrolled = useMemo(() => value === undefined, []);
  const hasTopSlot = typeof topContent !== 'undefined';
  const hasBottomSlot = typeof bottomContent !== 'undefined';
  const isSlotted = hasTopSlot || hasBottomSlot || !!searchConfig;
  const role = type === 'select' ? 'combobox' : undefined;
  useEffect(() => {
    if (!isUncontrolled) {
      setCurrentValues(value);
    }
  }, [value, isUncontrolled]);
  const isOpenUncontrolled = useMemo(() => open === undefined, []);
  const isEffectiveOpen = isOpenUncontrolled ? isMenuVisible : open !== null && open !== void 0 ? open : false;
  const ariaHasPopupValue = isSlotted || type === 'menu' && !multiple ? 'dialog' : 'listbox';
  const getItemToFocus = menu => {
    const optionsList = Array.prototype.slice.call(menu.querySelectorAll('.andes-list__item'));
    if (optionsList.length) {
      let focusableOption;
      if (type === TYPE_MENU) {
        focusableOption = optionsList[0].querySelector('.andes-list__item-action, .andes-list__item-anchor, .andes-list__item-actionable');
        if (focusableOption) return focusableOption;
      }
      focusableOption = optionsList.find(option => option.tabIndex === 0);
      if (!focusableOption) {
        focusableOption = optionsList.find(option => option.className.indexOf('andes-list__item--selected') !== -1);
      }
      if (!focusableOption) {
        var _optionsList = _slicedToArray(optionsList, 1);
        focusableOption = _optionsList[0];
      }
      return focusableOption;
    }
    return null;
  };
  const handleSelectionChange = (event, values) => {
    const shouldAvoidUnselect = !allowUnselect && !multiple && (values === undefined || !(values !== null && values !== void 0 && values.length)) && currentValues !== undefined;
    if (shouldAvoidUnselect) return;
    if (isUncontrolled) setCurrentValues(values);
    let newValue = values;
    if (!multiple && Array.isArray(values)) {
      var _values = _slicedToArray(values, 1);
      newValue = _values[0];
    }
    if (searchConfig) setSearchItemFocused(state => !Array.isArray(newValue) ? newValue : state);
    if (typeof onChange === 'function') onChange(event, newValue);
  };
  const setPopoverFocus = () => {
    if (searchConfig) {
      var _searchBoxRef$current;
      const label = searchBoxRef === null || searchBoxRef === void 0 || (_searchBoxRef$current = searchBoxRef.current) === null || _searchBoxRef$current === void 0 ? void 0 : _searchBoxRef$current.querySelector('label');
      setTimeout(() => label === null || label === void 0 ? void 0 : label.focus(), 0);
      return;
    }
    const menu = menuRef.current;
    if (menu) {
      const itemToFocus = getItemToFocus(menu);
      if (itemToFocus) {
        itemToFocus.focus();
        menu.scrollTop = itemToFocus.offsetTop - LIST_ITEM_SIZE;
      }
    }
  };
  const onSearchChangeText = text => setSearchText(text);
  const fitsSearchValue = useCallback(item => {
    if (!searchText) return true;
    const optionText = item.props.title;
    if (typeof searchConfig === 'object' && typeof searchConfig.filterCondition === 'function') {
      return searchConfig.filterCondition(searchText, optionText, item);
    }
    return optionText && includes(optionText.toLowerCase(), searchText.toLowerCase());
  }, [searchText, typeof searchConfig === 'object' && (searchConfig === null || searchConfig === void 0 ? void 0 : searchConfig.filterCondition)]);
  const options = useMemo(() => {
    if (!children) return null;
    if (isGroup) {
      let index = 0;
      const groups = [];
      Children.forEach(children, (group, groupIndex) => {
        const newChildren = [];
        Children.forEach(group === null || group === void 0 ? void 0 : group.props.children, item => {
          if (fitsSearchValue(item)) {
            newChildren.push(/*#__PURE__*/cloneElement(item, {
              index,
              groupIndex
            }));
          }
          index += 1;
        });
        if (newChildren.length > 0) {
          groups.push(/*#__PURE__*/cloneElement(group, {
            children: newChildren
          }));
        }
      });
      return groups;
    }
    const newChildren = [];
    Children.forEach(children, (child, index) => {
      if (fitsSearchValue(child)) {
        newChildren.push(/*#__PURE__*/cloneElement(child, {
          index
        }));
      }
    });
    return newChildren;
  }, [children, isGroup, fitsSearchValue]);
  const hoverElement = (list, itemSelected) => {
    const addRemoveFocus = (items, i, index) => {
      if (items[i].classList.contains(listHoverClass)) {
        items[i].classList.remove(listHoverClass);
      }
      if (itemSelected === (index === undefined ? i : index)) {
        items[i].classList.add(listHoverClass);
      }
    };
    if (isGroup) {
      let index = 0;
      for (let j = 0; j < list.length; j += 1) {
        const groupItems = list[j].children[1].children;
        for (let i = 0; i < groupItems.length; i += 1) {
          addRemoveFocus(groupItems, i, index);
          index += 1;
        }
      }
      return;
    }
    for (let i = 0; i < list.length; i += 1) {
      addRemoveFocus(list, i);
    }
  };
  const onSearchFocus = event => {
    setTimeout(() => {
      var _searchBoxRef$current2;
      return (_searchBoxRef$current2 = searchBoxRef.current) === null || _searchBoxRef$current2 === void 0 ? void 0 : _searchBoxRef$current2.focus();
    }, 0);
    const list = isGroup ? options === null || options === void 0 ? void 0 : options.reduce((previousValue, currentValue) => [...previousValue, ...currentValue.props.children], []) : options || [];
    if (!list) return;
    const listLength = list.length;
    setSearchItemFocused(state => {
      var _menuRef$current, _list$focusIndex;
      let index = list.findIndex(_ref => {
        let props = _ref.props;
        return props.value === state;
      });
      switch (event === null || event === void 0 ? void 0 : event.code) {
        case ARROW_DOWN:
          index += 1;
          break;
        case ARROW_UP:
          if (index === -1 || index === 0) {
            index = listLength - 1;
            break;
          }
          index -= 1;
          break;
        default:
          index = 0;
          break;
      }
      const focusIndex = index % listLength;
      hoverElement((menuRef === null || menuRef === void 0 || (_menuRef$current = menuRef.current) === null || _menuRef$current === void 0 ? void 0 : _menuRef$current.children) || [], focusIndex);
      return (_list$focusIndex = list[focusIndex]) === null || _list$focusIndex === void 0 || (_list$focusIndex = _list$focusIndex.props) === null || _list$focusIndex === void 0 ? void 0 : _list$focusIndex.value;
    });
  };
  const getFocusedItem = () => searchItemFocused;
  useEffect(() => {
    if (isMenuVisible && searchConfig && searchItemFocused !== undefined) {
      const menu = menuRef.current;
      if (menu) {
        var _document;
        const itemToScrollTop = (_document = document) === null || _document === void 0 ? void 0 : _document.getElementById(`${listId}-option-${searchItemFocused}`);
        if (itemToScrollTop && menuMaxHeight) {
          const listHeight = menuMaxHeight - SEARCH_BOX_SIZE;
          const itemsToTop = Math.floor(listHeight / LIST_ITEM_SIZE) - 1;
          menu.scrollTop = itemToScrollTop.offsetTop - itemsToTop * LIST_ITEM_SIZE;
        }
      }
    }
  }, [searchConfig, menuRef, searchItemFocused, listId, isMenuVisible]);
  const onSearchSelectionChange = event => {
    if (!searchItemFocused) return;
    if ((multiple || allowUnselect) && Array.isArray(currentValues)) {
      if ((currentValues || []).indexOf(searchItemFocused) > -1) {
        handleSelectionChange(event, currentValues.filter(item => item !== searchItemFocused));
        return;
      }
    }
    if (multiple) {
      handleSelectionChange(event, [...(currentValues || []), searchItemFocused]);
      return;
    }
    handleSelectionChange(event, [searchItemFocused]);
  };
  const resetSearchValues = () => {
    var _searchBoxRef$current3;
    const label = searchBoxRef === null || searchBoxRef === void 0 || (_searchBoxRef$current3 = searchBoxRef.current) === null || _searchBoxRef$current3 === void 0 ? void 0 : _searchBoxRef$current3.querySelector('label');
    if (label !== null && label !== void 0 && label.control) {
      const labelControl = label.control;
      labelControl.value = '';
    }
    setSearchText('');
    if (typeof searchConfig === 'object' && typeof searchConfig.onChange === 'function') searchConfig.onChange('');
  };
  const openMenu = event => {
    setIsMenuVisible(true);
    if (!hasTopSlot) {
      setTimeout(setPopoverFocus, 0);
    }
    if (typeof onOpen === 'function') onOpen(event);
  };
  const closeMenu = (_, preventFocus) => {
    const trigger = triggerRef.current;
    setIsMenuVisible(false);
    if (searchConfig) resetSearchValues();
    if (trigger && !preventFocus) setTimeout(() => trigger.focus(), 0);
    if (typeof onClose === 'function') onClose();
  };
  const handleClickOutside = _ => {
    closeMenu(_, true);
  };
  const toggleMenu = event => {
    if (isOpenUncontrolled) {
      if (isMenuVisible) {
        closeMenu();
      } else {
        openMenu(event);
      }
      return;
    }
    if (open) {
      if (typeof onClose === 'function') onClose();
    } else if (typeof onOpen === 'function') {
      onOpen(event);
    }
  };
  useEffect(() => {
    if (defaultOpen) setTimeout(openMenu, 0);
  }, []);
  useEffect(() => {
    if (!isOpenUncontrolled) {
      if (open) {
        setTimeout(openMenu, 0);
      } else if (isMenuVisible) {
        closeMenu();
      }
    }
  }, [open]);
  const getMenuMaxHeightRelatedStyles = () => {
    if (!menuMaxHeight) return {};
    return {
      maxHeight: menuMaxHeight,
      overflowY: 'auto'
    };
  };
  const handleOpenPopup = event => {
    event.preventDefault();
    if (isOpenUncontrolled) {
      openMenu(event);
      return;
    }
    if (typeof onOpen === 'function') onOpen(event);
  };
  const handleClosePopup = () => {
    if (isOpenUncontrolled) {
      closeMenu();
      return;
    }
    if (typeof onClose === 'function') onClose();
  };
  const handleTriggerKeyDown = event => {
    switch (event.code) {
      case ESCAPE:
        handleClosePopup();
        break;
      case ARROW_DOWN:
      case ARROW_UP:
        if (type === 'select') handleOpenPopup(event);
        break;
      case ENTER:
      case SPACE:
        handleOpenPopup(event);
        break;
    }
  };
  const getValuesArray = () => {
    if (Array.isArray(currentValues)) return currentValues;
    if (typeof currentValues === 'string' || typeof currentValues === 'number') return [currentValues];
    return [];
  };
  const getAriaControls = () => {
    if (type !== 'select') return undefined;
    if (!isEffectiveOpen) return undefined;
    if (role && isSlotted) {
      return `${componentProps === null || componentProps === void 0 ? void 0 : componentProps.id}-dialog`;
    }
    return `${componentProps === null || componentProps === void 0 ? void 0 : componentProps.id}-list`;
  };
  const menuProps = {
    visible: isMenuVisible,
    onSelectionChange: handleSelectionChange,
    options,
    selectedKeys: getValuesArray(),
    style: _objectSpread(_objectSpread({}, getMenuMaxHeightRelatedStyles()), {}, {
      width
    }, style),
    popperOptions,
    plugins: [{
      name: 'hideOnEsc',
      defaultValue: true,
      fn(_ref2) {
        let popper = _ref2.popper;
        function localOnKeyDown(event) {
          if (event.key === ESCAPE) {
            handleClosePopup();
          }
        }
        return {
          onShow() {
            popper.addEventListener('keydown', localOnKeyDown);
          },
          onHide() {
            popper.removeEventListener('keydown', localOnKeyDown);
          }
        };
      }
    }],
    offset,
    onClickOutside: isOpenUncontrolled ? handleClickOutside : onClose,
    closeMenu: isOpenUncontrolled ? closeMenu : onClose,
    onSearchFocus: searchConfig ? onSearchFocus : undefined,
    onSearchSelectionChange: searchConfig ? onSearchSelectionChange : undefined,
    onSearchChangeText: searchConfig ? onSearchChangeText : undefined,
    getFocusedItem: searchConfig ? getFocusedItem : undefined,
    onMount: searchConfig ? componentProps.onMount : undefined
  };
  const triggerProps = {
    role,
    'aria-expanded': isMenuVisible,
    'aria-controls': getAriaControls(),
    'aria-haspopup': ariaHasPopupValue,
    onClick: toggleMenu,
    onKeyDown: handleTriggerKeyDown
  };
  return {
    menuProps,
    triggerProps
  };
};
var useFloatingMenu$1 = useFloatingMenu;
const namespace = 'andes-floating-menu';
const namespaceMenuMultiple = `${namespace}--multiple`;
const formNamespace = 'andes-form-control';
const namespaceMenuSearch = `${namespace}--search`;
const WARNING_DIALOG_SR_LABEL = `<FloatingMenu> needs some plain text content for the dialog element for accessibility reasons.
Please add a 'dialogSrLabel' prop.`;
const WARNING_SEARCHBOX_DISABLED = `If you are using the 'disabled' property to temporarily block the SearchBox, it is recommended to use loading instead.
The disabled state will be removed in future versions.`;
var enTranslations = {
  "andes-floating-menu\u0004No hay resultados": [null, "No results"],
  "andes-floating-menu\u0004Abrir buscador": [null, "Open search tool"],
  "andes-floating-menu\u0004Buscar": [null, "Search"],
  "andes-floating-menu\u0004Borrar": [null, "Delete search"],
  "andes-floating-menu\u0004Cerrar menú": [null, "Close menu"]
};
var esArTranslations = {
  "andes-floating-menu\u0004No hay resultados": [null, "No hay resultados"],
  "andes-floating-menu\u0004Abrir buscador": [null, "Abrir buscador"],
  "andes-floating-menu\u0004Buscar": [null, "Buscar"],
  "andes-floating-menu\u0004Borrar": [null, "Borrar"],
  "andes-floating-menu\u0004Cerrar menú": [null, "Cerrar menú"]
};
var ptBrTranslations = {
  "andes-floating-menu\u0004No hay resultados": [null, "Sem resultados"],
  "andes-floating-menu\u0004Abrir buscador": [null, "Abrir mecanismo de busca"],
  "andes-floating-menu\u0004Buscar": [null, "Procurar"],
  "andes-floating-menu\u0004Borrar": [null, "Excluir"],
  "andes-floating-menu\u0004Cerrar menú": [null, "Fechar menu"]
};
var zhCNTranslations = {
  "andes-floating-menu\u0004No hay resultados": [null, "无结果"],
  "andes-floating-menu\u0004Abrir buscador": [null, "打开搜索工具"],
  "andes-floating-menu\u0004Buscar": [null, "搜索"],
  "andes-floating-menu\u0004Borrar": [null, "删除搜索"],
  "andes-floating-menu\u0004Cerrar menú": [null, "关闭菜单"]
};
const SearchBox = /*#__PURE__*/function () {
  const SearchBox = _ref => {
    let className = _ref.className,
      defaultValue = _ref.defaultValue,
      _ref$inputProps = _ref.inputProps,
      inputProps = _ref$inputProps === void 0 ? {} : _ref$inputProps,
      onChange = _ref.onChange,
      onClear = _ref.onClear,
      onKeyDown = _ref.onKeyDown,
      placeholder = _ref.placeholder,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? 'small' : _ref$size,
      srLabel = _ref.srLabel,
      value = _ref.value,
      ref = _ref.ref;
    const _useAndesContext = useAndesContext('@andes/floating-menu', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations,
        'zh-CN': zhCNTranslations
      }),
      device = _useAndesContext.device,
      i18n = _useAndesContext.i18n;
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const inputRef = useRef(null);
    const isControlled = useMemo(() => typeof value !== 'undefined', []);
    const _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      focused = _useState2[0],
      setFocused = _useState2[1];
    const _useState3 = useState(false),
      _useState4 = _slicedToArray(_useState3, 2),
      showClear = _useState4[0],
      setShowClear = _useState4[1];
    const firstRender = useRef(true);
    const isWebMobileIOS = (device === null || device === void 0 ? void 0 : device.osName) === 'ios';
    warning(!(inputProps !== null && inputProps !== void 0 && inputProps.disabled), WARNING_SEARCHBOX_DISABLED);
    useEffect(() => {
      if (!isControlled && defaultValue) {
        setShowClear(true);
      }
      firstRender.current = false;
    }, []);
    useEffect(() => {
      if (isControlled) {
        const showClearButton = !!value && (value === null || value === void 0 ? void 0 : value.length) > 0;
        setShowClear(showClearButton);
      }
    }, [isControlled, value, setShowClear]);
    useEffect(() => {
      if (focused) {
        var _inputRef$current;
        (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 || _inputRef$current.focus();
      }
    }, [focused]);
    const SearchIcon = () => {
      if (size === 'small') {
        return /*#__PURE__*/React.createElement(Search16, {
          color: "currentColor"
        });
      }
      return /*#__PURE__*/React.createElement(Search20, {
        color: "currentColor"
      });
    };
    const CloseIcon = () => {
      if (size === 'small') {
        return /*#__PURE__*/React.createElement(Close16, {
          color: "currentColor"
        });
      }
      return /*#__PURE__*/React.createElement(Close20, {
        color: "currentColor"
      });
    };
    const handleChange = event => {
      var _inputRef$current2, _inputRef$current3;
      const showClearButton = !!(inputRef !== null && inputRef !== void 0 && (_inputRef$current2 = inputRef.current) !== null && _inputRef$current2 !== void 0 && _inputRef$current2.value) && (inputRef === null || inputRef === void 0 || (_inputRef$current3 = inputRef.current) === null || _inputRef$current3 === void 0 ? void 0 : _inputRef$current3.value.length) > 0;
      setShowClear(showClearButton);
      if (typeof onChange === 'function') onChange(event);
    };
    const handleClear = event => {
      var _inputRef$current4;
      event.preventDefault();
      if (!isControlled && inputRef.current) inputRef.current.value = '';
      event.target.value = '';
      setShowClear(false);
      if (typeof onClear === 'function') onClear(event);
      if (typeof onChange === 'function') onChange(event);
      (_inputRef$current4 = inputRef.current) === null || _inputRef$current4 === void 0 || _inputRef$current4.focus();
    };
    const handleBlur = event => {
      const onBlur = inputProps.onBlur;
      setFocused(false);
      if (typeof onBlur === 'function') onBlur(event);
    };
    const handleFocus = event => {
      const onFocus = inputProps.onFocus;
      setFocused(true);
      if (typeof onFocus === 'function') onFocus(event);
    };
    const dismissClassNames = classNames(`${formNamespace}__close-icon`, {
      [`${formNamespace}__close-icon-visible`]: showClear
    });
    const wrapperClassNames = classNames(namespaceMenuSearch, `${namespaceMenuSearch}--size-${size}`, className, {
      [`${namespaceMenuSearch}--web-mobile-ios`]: isWebMobileIOS,
      [`${namespaceMenuSearch}--focused`]: focused
    });
    return /*#__PURE__*/React.createElement("div", _extends({
      ref: ref,
      className: getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('searchbox')
    }, themeHash), /*#__PURE__*/React.createElement(FocusRing, {
      within: true,
      isTextInput: true,
      focusRingClass: "focus-visible"
    }, /*#__PURE__*/React.createElement("div", {
      className: wrapperClassNames,
      "data-andes-searchbox": true,
      "data-andes-searchbox-variant": "simple",
      "data-andes-searchbox-size": size
    }, /*#__PURE__*/React.createElement("label", null, /*#__PURE__*/React.createElement(VisuallyHidden, null, srLabel || i18n.pgettext(namespace, 'Buscar')), /*#__PURE__*/React.createElement("input", _extends({}, inputProps, {
      className: `${formNamespace}__field`,
      value: value,
      defaultValue: defaultValue,
      placeholder: placeholder || `${i18n.pgettext(namespace, 'Buscar')}`,
      onChange: handleChange,
      onBlur: handleBlur,
      onFocus: handleFocus,
      ref: inputRef,
      onKeyDown: onKeyDown,
      "data-andes-searchbox-input": true
    }))), /*#__PURE__*/React.createElement("button", {
      className: dismissClassNames,
      onClick: handleClear,
      onMouseDown: handleClear,
      "aria-label": `${i18n.pgettext(namespace, 'Borrar')}`,
      type: "button",
      "data-andes-searchbox-close": true
    }, /*#__PURE__*/React.createElement(CloseIcon, null)), /*#__PURE__*/React.createElement("div", {
      className: `${formNamespace}__search-icon`,
      "data-andes-searchbox-search": true
    }, /*#__PURE__*/React.createElement(SearchIcon, null)))));
  };
  SearchBox.displayName = 'SearchBox';
  return SearchBox;
}();
var SearchBox$1 = SearchBox;
const FloatingMenuListDialogContent = _ref => {
  let visible = _ref.visible,
    children = _ref.children,
    id = _ref.id,
    dialogSrLabel = _ref.dialogSrLabel,
    closeMenu = _ref.closeMenu;
  warning(dialogSrLabel, WARNING_DIALOG_SR_LABEL);
  const _useAndesContext = useAndesContext('@andes/floating-menu', {
      en: enTranslations,
      'es-AR': esArTranslations,
      'pt-BR': ptBrTranslations,
      'zh-CN': zhCNTranslations
    }),
    device = _useAndesContext.device,
    i18n = _useAndesContext.i18n;
  const shouldDisableFocusTrap = (device === null || device === void 0 ? void 0 : device.osName) === 'android' && (device === null || device === void 0 ? void 0 : device.mobile);
  const hiddenCloseButton = /*#__PURE__*/React.createElement(VisuallyHiddenFocusable, {
    onClick: closeMenu,
    type: "button"
  }, i18n.pgettext(namespace, 'Cerrar menú'));
  return /*#__PURE__*/React.createElement("div", {
    role: "dialog",
    "aria-modal": "true",
    "aria-label": dialogSrLabel,
    id: `${id}-dialog`
  }, /*#__PURE__*/React.createElement(FocusScope, {
    contain: !shouldDisableFocusTrap && visible,
    restoreFocus: true,
    autoFocus: true
  }, shouldDisableFocusTrap && hiddenCloseButton, children));
};
const FloatingMenuListSlotted = _ref2 => {
  let children = _ref2.children,
    visible = _ref2.visible,
    topContent = _ref2.topContent,
    bottomContent = _ref2.bottomContent,
    dialog = _ref2.dialog,
    dialogSrLabel = _ref2.dialogSrLabel,
    id = _ref2.id,
    closeMenu = _ref2.closeMenu;
  const hasTopSlot = typeof topContent !== 'undefined';
  const hasBottomSlot = typeof bottomContent !== 'undefined';
  const isSlotted = hasTopSlot || hasBottomSlot;
  const content = /*#__PURE__*/React.createElement(React.Fragment, null, topContent, children, bottomContent);
  if (dialog === false) {
    return content;
  }
  if (dialog || isSlotted) {
    if (hasTopSlot) {
      return /*#__PURE__*/React.createElement(React.Fragment, null, visible && /*#__PURE__*/React.createElement(FloatingMenuListDialogContent, {
        visible: visible,
        id: id,
        dialogSrLabel: dialogSrLabel,
        closeMenu: closeMenu
      }, content));
    }
    return /*#__PURE__*/React.createElement(FloatingMenuListDialogContent, {
      visible: visible,
      id: id,
      dialogSrLabel: dialogSrLabel,
      closeMenu: closeMenu
    }, content);
  }
  return content;
};
var FloatingMenuListSlotted$1 = FloatingMenuListSlotted;
const FloatingMenuSearch = /*#__PURE__*/function () {
  const FloatingMenuSearch = _ref => {
    let defaultValue = _ref.defaultValue,
      getFocusedItem = _ref.getFocusedItem,
      handleKeyDown = _ref.handleKeyDown,
      list = _ref.list,
      listId = _ref.listId,
      onChange = _ref.onChange,
      placeholder = _ref.placeholder,
      visible = _ref.visible,
      dialogSrLabel = _ref.dialogSrLabel,
      id = _ref.id,
      ref = _ref.ref;
    const inputProps = {
      role: 'combobox',
      'aria-label': placeholder,
      'aria-autocomplete': 'list',
      'aria-controls': listId,
      'aria-activedescendant': getFocusedItem() ? `${listId}-option-${getFocusedItem()}` : undefined,
      'aria-expanded': true
    };
    return /*#__PURE__*/React.createElement(FloatingMenuListSlotted$1, {
      id: id,
      dialogSrLabel: dialogSrLabel,
      visible: visible,
      topContent: /*#__PURE__*/React.createElement("div", {
        className: `${namespace}__search-box`
      }, /*#__PURE__*/React.createElement(SearchBox$1, {
        ref: ref,
        defaultValue: defaultValue,
        onChange: onChange,
        placeholder: placeholder,
        onKeyDown: handleKeyDown,
        inputProps: inputProps
      }))
    }, list);
  };
  FloatingMenuSearch.displayName = 'FloatingMenuSearch';
  return FloatingMenuSearch;
}();
var FloatingMenuSearch$1 = FloatingMenuSearch;
const _excluded$3 = ["children", "className", "defaultValue", "onSelect", "value", "ref"];
const FloatingMenuMultiple = /*#__PURE__*/function () {
  const FloatingMenuMultiple = _ref => {
    let children = _ref.children,
      className = _ref.className,
      defaultValue = _ref.defaultValue,
      onSelect = _ref.onSelect,
      value = _ref.value,
      ref = _ref.ref,
      listProps = _objectWithoutProperties(_ref, _excluded$3);
    const _useState = useState(value || defaultValue),
      _useState2 = _slicedToArray(_useState, 2),
      currentValues = _useState2[0],
      setCurrentValues = _useState2[1];
    const isUncontrolled = useMemo(() => value === undefined, []);
    useEffect(() => {
      if (!isUncontrolled) {
        setCurrentValues(value);
      }
    }, [value]);
    const customOnSelect = (event, selectedValues, itemDetails) => {
      if (isUncontrolled) setCurrentValues(selectedValues);
      if (typeof onSelect === 'function') onSelect(event, selectedValues, itemDetails);
    };
    const getValuesArray = () => Array.isArray(currentValues) ? currentValues : [];
    const stateProps = {};
    if (!isUncontrolled) {
      stateProps.selectedKeys = getValuesArray();
    } else if (defaultValue) {
      stateProps.defaultSelectedKeys = getValuesArray();
    }
    const floatingMenuMultipleClassNames = classNames(namespaceMenuMultiple, className);
    return /*#__PURE__*/React.createElement(ListBox, _extends({
      className: floatingMenuMultipleClassNames,
      selectionComponent: /*#__PURE__*/React.createElement(Checkbox, {
        srHidden: true
      }),
      selectionMode: "multiple",
      ref: ref,
      onSelect: customOnSelect
    }, listProps, stateProps), children);
  };
  FloatingMenuMultiple.displayName = 'FloatingMenuMultiple';
  return FloatingMenuMultiple;
}();
var FloatingMenuMultiple$1 = FloatingMenuMultiple;
const FloatingMenuList = /*#__PURE__*/function () {
  const FloatingMenuList = _ref => {
    let closeMenu = _ref.closeMenu,
      bottomContent = _ref.bottomContent,
      i18n = _ref.i18n,
      id = _ref.id,
      _ref$multiple = _ref.multiple,
      multiple = _ref$multiple === void 0 ? false : _ref$multiple,
      onKeyDown = _ref.onKeyDown,
      onSelect = _ref.onSelect,
      options = _ref.options,
      _ref$search = _ref.search,
      search = _ref$search === void 0 ? false : _ref$search,
      selectedKeys = _ref.selectedKeys,
      srLabel = _ref.srLabel,
      topContent = _ref.topContent,
      _ref$type = _ref.type,
      type = _ref$type === void 0 ? 'menu' : _ref$type,
      visible = _ref.visible,
      dialogSrLabel = _ref.dialogSrLabel,
      ref = _ref.ref;
    if (!options || !options.length) {
      return /*#__PURE__*/React.createElement(FloatingMenuListSlotted$1, {
        visible: visible,
        topContent: topContent,
        bottomContent: bottomContent,
        dialog: !search,
        id: id,
        dialogSrLabel: dialogSrLabel
      }, /*#__PURE__*/React.createElement("div", {
        className: `${namespace}__no-results`,
        "data-andes-floating-menu-no-result": true
      }, i18n.pgettext(namespace, 'No hay resultados')));
    }
    const commonProps = {
      id,
      ref,
      className: namespace,
      onKeyDown,
      srLabel
    };
    if (multiple) {
      return /*#__PURE__*/React.createElement(FloatingMenuListSlotted$1, {
        visible: visible,
        topContent: topContent,
        bottomContent: bottomContent,
        dialog: !search,
        id: id,
        dialogSrLabel: dialogSrLabel
      }, /*#__PURE__*/React.createElement(FloatingMenuMultiple$1, _extends({}, commonProps, {
        onSelect: onSelect,
        value: selectedKeys,
        shouldUseVirtualFocus: !!search
      }), options));
    }
    if (type === 'select' || search) {
      const isSlotted = typeof topContent !== 'undefined' || typeof bottomContent !== 'undefined';
      return /*#__PURE__*/React.createElement(FloatingMenuListSlotted$1, {
        visible: visible,
        topContent: topContent,
        bottomContent: bottomContent,
        dialog: isSlotted && !search,
        id: id,
        dialogSrLabel: dialogSrLabel
      }, /*#__PURE__*/React.createElement(ListBox, _extends({}, commonProps, {
        onSelect: onSelect,
        selectionMode: multiple ? 'multiple' : 'single',
        selectedKeys: selectedKeys,
        shouldUseVirtualFocus: !!search
      }), options));
    }
    const handleMenuClick = (childOnSelect, event, value) => {
      if (typeof childOnSelect === 'function') childOnSelect(event, value);
      if (typeof onSelect === 'function') onSelect(event, value);
    };
    const getItem = item => /*#__PURE__*/cloneElement(item, {
      onClick: (evt, value) => handleMenuClick(item.props.onClick, evt, value),
      tabIndex: undefined
    });
    return /*#__PURE__*/React.createElement(FloatingMenuListSlotted$1, {
      visible: visible,
      topContent: topContent,
      bottomContent: bottomContent,
      dialog: true,
      id: id,
      dialogSrLabel: dialogSrLabel,
      closeMenu: closeMenu
    }, /*#__PURE__*/React.createElement(List, commonProps, React.Children.map(options, child => {
      var _child$type;
      const isGroup = (child === null || child === void 0 || (_child$type = child.type) === null || _child$type === void 0 ? void 0 : _child$type.displayName) === GROUP_DISPLAY_NAME;
      if (isGroup) {
        const items = React.Children.map(child.props.children, getItem);
        return /*#__PURE__*/cloneElement(child, {
          children: items
        });
      }
      return getItem(child);
    })));
  };
  FloatingMenuList.displayName = 'FloatingMenuList';
  return FloatingMenuList;
}();
var FloatingMenuList$1 = FloatingMenuList;
const computeSearchProps = _ref => {
  let searchProps = _ref.search;
  if (!searchProps) return undefined;
  if (typeof searchProps === 'boolean') {
    return {
      search: true
    };
  }
  return _objectSpread(_objectSpread({}, searchProps), {}, {
    search: true
  });
};
var computeSearchProps$1 = computeSearchProps;
var _metrics$sendSlots;
const METRIC_COMPONENT = 'Floating-Menu';
const isSlotted = _ref => {
  let topContent = _ref.topContent,
    bottomContent = _ref.bottomContent,
    search = _ref.search;
  return typeof topContent !== 'undefined' && !search || typeof bottomContent !== 'undefined';
};
const metrics = senderMetrics(trackCustomization);
const FloatingMenuMetrics = {
  SSR: [metrics === null || metrics === void 0 || (_metrics$sendSlots = metrics.sendSlots) === null || _metrics$sendSlots === void 0 ? void 0 : _metrics$sendSlots.call(metrics, METRIC_COMPONENT, isSlotted)]
};
const _excluded$2 = ["ref"],
  _excluded2 = ["closeMenu", "getFocusedItem", "onSearchChangeText", "onSearchFocus", "onSearchSelectionChange", "onSelectionChange", "options", "selectedKeys"];
const FloatingMenu = /*#__PURE__*/function () {
  const FloatingMenu = _ref => {
    let ref = _ref.ref,
      props = _objectWithoutProperties(_ref, _excluded$2);
    const _props$allowUnselect = props.allowUnselect,
      allowUnselect = _props$allowUnselect === void 0 ? false : _props$allowUnselect,
      appendTo = props.appendTo,
      bottomContent = props.bottomContent,
      className = props.className,
      _props$defaultOpen = props.defaultOpen,
      defaultOpen = _props$defaultOpen === void 0 ? false : _props$defaultOpen,
      id = props.id,
      menuMaxHeight = props.menuMaxHeight,
      _props$multiple = props.multiple,
      multiple = _props$multiple === void 0 ? false : _props$multiple,
      _props$placement = props.placement,
      placement = _props$placement === void 0 ? 'bottom' : _props$placement,
      _props$search = props.search,
      search = _props$search === void 0 ? false : _props$search,
      srLabel = props.srLabel,
      topContent = props.topContent,
      trigger = props.trigger,
      _props$type = props.type,
      type = _props$type === void 0 ? 'menu' : _props$type,
      _props$width = props.width,
      width = _props$width === void 0 ? 280 : _props$width,
      dialogSrLabel = props.dialogSrLabel;
    const componentId = useId(id);
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    getThemedComponentClasses === null || getThemedComponentClasses === void 0 || getThemedComponentClasses('list');
    const searchConfig = computeSearchProps$1({
      search
    });
    const _useAndesContext = useAndesContext('@andes/floating-menu', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations,
        'zh-CN': zhCNTranslations
      }),
      i18n = _useAndesContext.i18n;
    const listId = `${componentId}-list`;
    const triggerRef = useRef(null);
    const menuRef = useRef(null);
    const searchBoxRef = useRef(null);
    const hasSearched = useRef(false);
    const _useFloatingMenu = useFloatingMenu$1(triggerRef, menuRef, searchBoxRef, _objectSpread(_objectSpread({}, props), {}, {
        allowUnselect,
        defaultOpen,
        multiple,
        placement,
        type,
        width,
        listId,
        id: componentId,
        searchConfig
      })),
      menuProps = _useFloatingMenu.menuProps,
      triggerProps = _useFloatingMenu.triggerProps;
    const closeMenu = menuProps.closeMenu,
      getFocusedItem = menuProps.getFocusedItem,
      onSearchChangeText = menuProps.onSearchChangeText,
      onSearchFocus = menuProps.onSearchFocus,
      onSearchSelectionChange = menuProps.onSearchSelectionChange,
      onSelectionChange = menuProps.onSelectionChange,
      options = menuProps.options,
      selectedKeys = menuProps.selectedKeys,
      otherMenuProps = _objectWithoutProperties(menuProps, _excluded2);
    useEffect(() => {
      if (hasSearched.current) {
        hasSearched.current = false;
        if (typeof onSearchFocus === 'function') onSearchFocus();
      }
    }, [hasSearched.current]);
    useEffect(() => {
      const menu = menuRef.current;
      if (menu && searchConfig) {
        requestAnimationFrame(() => {
          menu.style.maxHeight = menuMaxHeight ? `${menuMaxHeight - SEARCH_BOX_SIZE}px` : '';
          menu.style.overflowY = 'auto';
        });
      }
    }, [menuRef.current, menuMaxHeight, searchConfig]);
    const handleSelectionChange = (event, values) => {
      if (values) {
        onSelectionChange(event, values);
      } else if (typeof onSearchSelectionChange === 'function') {
        onSearchSelectionChange(event);
      }
      if (!multiple && typeof closeMenu === 'function') closeMenu();
    };
    const handleKeyDown = event => {
      switch (event.code) {
        case TAB:
          if (type !== 'menu') event.preventDefault();
          break;
        case ARROW_UP:
        case ARROW_DOWN:
          if (searchConfig) onSearchFocus(event);
          break;
        case ENTER:
          if (searchConfig) handleSelectionChange(event);
          break;
      }
    };
    const handleSearchChange = event => {
      const value = event.target.value;
      hasSearched.current = true;
      if (typeof onSearchChangeText === 'function') onSearchChangeText(value);
      if (typeof (searchConfig === null || searchConfig === void 0 ? void 0 : searchConfig.onChange) === 'function') searchConfig.onChange(value);
    };
    const getContent = () => {
      const list = /*#__PURE__*/React.createElement(FloatingMenuList$1, {
        closeMenu: closeMenu,
        id: listId,
        multiple: multiple,
        onKeyDown: handleKeyDown,
        onSelect: handleSelectionChange,
        options: options,
        ref: menuRef,
        selectedKeys: selectedKeys,
        type: type,
        search: searchConfig === null || searchConfig === void 0 ? void 0 : searchConfig.search,
        srLabel: srLabel,
        visible: menuProps.visible,
        i18n: i18n,
        topContent: !searchConfig ? topContent : undefined,
        bottomContent: bottomContent,
        dialogSrLabel: dialogSrLabel
      });
      if (!searchConfig) return list;
      return /*#__PURE__*/React.createElement(FloatingMenuSearch$1, _extends({
        getFocusedItem: getFocusedItem,
        handleKeyDown: handleKeyDown,
        list: list,
        listId: listId,
        ref: searchBoxRef,
        visible: menuProps.visible,
        id: componentId
      }, searchConfig, {
        onChange: handleSearchChange,
        dialogSrLabel: dialogSrLabel
      }));
    };
    const containerClassNames = classNames(namespace, className, {
      [`${namespace}--multiple`]: multiple
    });
    const contentClassName = classNames(`${namespace}__popper-content`, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('floatingMenu'));
    return /*#__PURE__*/React.createElement("div", {
      className: containerClassNames,
      ref: ref
    }, /*#__PURE__*/React.createElement(Popper, _extends({
      appendTo: appendTo,
      content: getContent(),
      interactive: true,
      placement: placement,
      id: `${componentId}-popper`,
      focusScope: true,
      classNames: {
        content: contentClassName
      },
      attrs: {
        root: themeHash,
        content: _objectSpread({
          'data-andes-floating-menu': true
        }, themeHash)
      }
    }, otherMenuProps), /*#__PURE__*/cloneElement(trigger, _objectSpread({
      ref: triggerRef
    }, triggerProps))));
  };
  FloatingMenu.displayName = 'FloatingMenu';
  return FloatingMenu;
}();
var FloatingMenu$1 = withTracker(FloatingMenu, FloatingMenuMetrics);
const _excluded$1 = ["children"];
const FloatingMenuGroup = /*#__PURE__*/function () {
  const FloatingMenuGroup = _ref => {
    let children = _ref.children,
      props = _objectWithoutProperties(_ref, _excluded$1);
    return /*#__PURE__*/React.createElement(ListGroup, props, children);
  };
  FloatingMenuGroup.getCollectionNode = groupNodeGenerator;
  FloatingMenuGroup.displayName = GROUP_DISPLAY_NAME;
  return FloatingMenuGroup;
}();
var FloatingMenuGroup$1 = FloatingMenuGroup;
const _excluded = ["children", "description", "disabled", "groupIndex", "href", "index", "onClick", "rightContent", "selected", "title", "value"];
const FloatingMenuItem = /*#__PURE__*/function () {
  const FloatingMenuItem = _ref => {
    let children = _ref.children,
      description = _ref.description,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      groupIndex = _ref.groupIndex,
      href = _ref.href,
      index = _ref.index,
      onClick = _ref.onClick,
      rightContent = _ref.rightContent,
      _ref$selected = _ref.selected,
      selected = _ref$selected === void 0 ? false : _ref$selected,
      title = _ref.title,
      value = _ref.value,
      props = _objectWithoutProperties(_ref, _excluded);
    const itemProps = _objectSpread({
      description,
      disabled,
      groupIndex,
      index,
      rightContent,
      selected,
      title,
      value
    }, props);
    const onClickHandler = e => {
      if (!disabled) {
        if (typeof onClick === 'function') {
          onClick(e, value);
        }
      }
    };
    if (href || onClick) {
      return /*#__PURE__*/React.createElement(ListActionItem, _extends({}, itemProps, {
        href: href,
        onClick: onClickHandler,
        legacyBehaviorSlot: true,
        chevron: false
      }), children);
    }
    return /*#__PURE__*/React.createElement(ListItem, _extends({}, itemProps, {
      onClick: onClickHandler
    }), children);
  };
  FloatingMenuItem.getCollectionNode = itemNodeGenerator;
  FloatingMenuItem.displayName = ITEM_DISPLAY_NAME;
  return FloatingMenuItem;
}();
var FloatingMenuItem$1 = FloatingMenuItem;
export { FloatingMenu$1 as FloatingMenu, FloatingMenuGroup$1 as FloatingMenuGroup, FloatingMenuItem$1 as FloatingMenuItem };
