// @andes/tag/build/esm/module.js
import _extends from '@babel/runtime/helpers/extends';
import React, { useState, useMemo } from 'react';
import classNames from 'classnames';
import { useId, useAndesContext } from '@andes/context';
import Close16 from '@andes/icons/Close16';
import { VisuallyHidden } from '@andes/visually-hidden';
import { computeDataAndesState } from '@andes/common/build/utils/computeDataAndesState';
import includes from '@andes/common/build/utils/includes';
import warning from '@andes/common/build/utils/warning';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { Thumbnail } from '@andes/thumbnail';
import { useThemeContext } from '@andes/themes';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { ESCAPE } from '@andes/common/build/constants/keyboardCodes';
import { Popper } from '@andes/popper';
import FeedbackPositive16 from '@andes/icons/FeedbackPositive16';
const DEPRECATED_COLOR_GREEN = 'green';
const DEPRECATED_COLOR_RED = 'red';
const DEPRECATED_COLOR_ORANGE = 'orange';
const COLOR_POSITIVE = 'positive';
const COLOR_CAUTION = 'caution';
const COLOR_NEGATIVE = 'negative';
const ALLOWED_COLORS = [COLOR_POSITIVE, COLOR_CAUTION, COLOR_NEGATIVE];
const DEPRECATED_COLORS = [DEPRECATED_COLOR_GREEN, DEPRECATED_COLOR_RED, DEPRECATED_COLOR_ORANGE];
[...ALLOWED_COLORS, ...DEPRECATED_COLORS];
const WARNING_COLOR_GREEN = `
'${DEPRECATED_COLOR_GREEN}' will be removed from 'color' options in a future major version.
Please, use '${COLOR_POSITIVE}' option instead
`;
const WARNING_COLOR_ORANGE = `
'${DEPRECATED_COLOR_ORANGE}' will be removed from 'color' options in a future major version.
Please, use '${COLOR_CAUTION}' option instead
`;
const WARNING_COLOR_RED = `
'${DEPRECATED_COLOR_RED}' will be removed from 'color' options in a future major version.
Please, use '${COLOR_NEGATIVE}' option instead
`;
const DEPRECATED_COLOR_WARNINGS = {
  [`${DEPRECATED_COLOR_GREEN}`]: WARNING_COLOR_GREEN,
  [`${DEPRECATED_COLOR_ORANGE}`]: WARNING_COLOR_ORANGE,
  [`${DEPRECATED_COLOR_RED}`]: WARNING_COLOR_RED
};
const namespace = 'andes-tag';
const namespaceCollapsed = `${namespace}-collapsed`;
const ThumbnailAsset = _ref => {
  let icon = _ref.icon,
    imageUrl = _ref.imageUrl,
    text = _ref.text,
    color = _ref.color,
    imageAlt = _ref.imageAlt;
  const _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    imageError = _useState2[0],
    setImageError = _useState2[1];
  const onImageError = () => setImageError(true);
  const renderContent = () => {
    if (icon) return icon;
    if (text) return /*#__PURE__*/React.createElement("span", null, text);
    if (imageUrl) {
      return imageError ? /*#__PURE__*/React.createElement(React.Fragment, null) : /*#__PURE__*/React.createElement("img", {
        src: imageUrl,
        alt: imageAlt,
        onError: onImageError
      });
    }
    return /*#__PURE__*/React.createElement(React.Fragment, null);
  };
  return /*#__PURE__*/React.createElement(Thumbnail, {
    className: `${namespace}--internal-thumbnail`,
    size: "24",
    color: color
  }, renderContent());
};
var ThumbnailAsset$1 = ThumbnailAsset;
const TagIcon = _ref => {
  let icon = _ref.icon;
  const classnames = classNames([`${namespace}__custom-icon`]);
  return /*#__PURE__*/React.createElement("div", {
    className: classnames
  }, icon);
};
var TagIcon$1 = TagIcon;
const computeColor = color => {
  switch (color) {
    case DEPRECATED_COLOR_RED:
    case COLOR_NEGATIVE:
      return COLOR_NEGATIVE;
    case DEPRECATED_COLOR_ORANGE:
    case COLOR_CAUTION:
      return COLOR_CAUTION;
    case DEPRECATED_COLOR_GREEN:
    case COLOR_POSITIVE:
      return COLOR_POSITIVE;
    default:
      return undefined;
  }
};
var computeColor$1 = computeColor;
var enTranslations = {
  "andes-tag\u0004quitar {0}": [null, "remove {0}"],
  "andes-tag\u0004error": [null, "error"],
  "andes-tag\u0004deshabilitado": [null, "disabled"],
  "andes-tag\u0004mostrar {0} mas": [null, "show {0} more"]
};
var zhCnTranslations = {
  "andes-tag\u0004quitar {0}": [null, "移除 {0}"],
  "andes-tag\u0004error": [null, "错误"],
  "andes-tag\u0004deshabilitado": [null, "已禁用"],
  "andes-tag\u0004mostrar {0} mas": [null, "再显示 {0} 条"]
};
var esArTranslations = {
  "andes-tag\u0004quitar {0}": [null, "quitar {0}"],
  "andes-tag\u0004error": [null, "error"],
  "andes-tag\u0004deshabilitado": [null, "deshabilitado"],
  "andes-tag\u0004mostrar {0} mas": [null, "mostrar {0} más"]
};
var ptBrTranslations = {
  "andes-tag\u0004quitar {0}": [null, "tirar {0}"],
  "andes-tag\u0004error": [null, "erro"],
  "andes-tag\u0004deshabilitado": [null, "desativado"],
  "andes-tag\u0004mostrar {0} mas": [null, "mostrar mais {0}"]
};
const getContainerCustomColorStyle = color => {
  if (!color) return;
  const backgroundColor = color.backgroundColor,
    borderColor = color.borderColor,
    textColor = color.color,
    closeIconColorHover = color.closeIconColorHover,
    closeIconColorActive = color.closeIconColorActive;
  return {
    style: {
      '--andes-tag-color-text-default': textColor,
      '--andes-tag-color-fill-default': backgroundColor,
      '--andes-tag-color-border-default': borderColor,
      '--andes-tag-color-close-default': textColor,
      '--andes-tag-color-close-hover': closeIconColorHover,
      '--andes-tag-color-close-active': closeIconColorActive,
      '--andes-tag-color-icon-default': textColor
    }
  };
};
const getColor = color => ({
  customColor: typeof color === 'object' ? color : undefined,
  computedColor: color && typeof color === 'string' ? computeColor$1(color) : undefined
});
const Tag = /*#__PURE__*/function () {
  const Tag = _ref => {
    let andesDataAttributes = _ref.andesDataAttributes,
      ariaLive = _ref['aria-live'],
      asset = _ref.asset,
      icon = _ref.icon,
      children = _ref.children,
      className = _ref.className,
      _ref$clickable = _ref.clickable,
      clickable = _ref$clickable === void 0 ? false : _ref$clickable,
      closeButtonSrLabel = _ref.closeButtonSrLabel,
      color = _ref.color,
      component = _ref.component,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      id = _ref.id,
      label = _ref.label,
      onClick = _ref.onClick,
      onClose = _ref.onClose,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? 'large' : _ref$size,
      srLabel = _ref.srLabel,
      ref = _ref.ref;
    const componentId = useId(id);
    const _useAndesContext = useAndesContext('@andes/tag', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations,
        'zh-CN': zhCnTranslations
      }),
      i18n = _useAndesContext.i18n;
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const _getColor = getColor(color),
      customColor = _getColor.customColor,
      computedColor = _getColor.computedColor;
    if (typeof color === 'string') {
      warning(!includes(DEPRECATED_COLORS, color), DEPRECATED_COLOR_WARNINGS[color]);
    }
    const handleClose = event => {
      event.stopPropagation();
      if (!disabled && typeof onClose === 'function') onClose(event);
    };
    const getClose = () => /*#__PURE__*/React.createElement("button", {
      type: "button",
      className: `${namespace}__close`,
      onClick: handleClose,
      disabled: disabled,
      "aria-label": closeButtonSrLabel || i18n.pgettext(namespace, 'quitar {0}', label),
      "data-andes-tag-close": true
    }, /*#__PURE__*/React.createElement("span", {
      "aria-hidden": true,
      className: `${namespace}__icon-container`
    }, /*#__PURE__*/React.createElement(Close16, {
      color: "currentColor"
    })));
    const tagClassName = classNames(namespace, className, {
      [`${namespace}--${size}`]: size,
      [`${namespace}__color--${computedColor}`]: computedColor,
      [`${namespace}--disabled`]: disabled,
      [`${namespace}--clickable`]: (typeof onClick === 'function' || clickable) && !disabled,
      [`${namespace}--custom-color`]: customColor
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('tag'));
    const messageDisabled = () => disabled && /*#__PURE__*/React.createElement(VisuallyHidden, null, i18n.pgettext(namespace, 'deshabilitado'), ":");
    const screenReaderLabel = () => srLabel && /*#__PURE__*/React.createElement(VisuallyHidden, null, srLabel);
    const content = children ? /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__child`
    }, messageDisabled(), screenReaderLabel(), children) : /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__label`
    }, messageDisabled(), screenReaderLabel(), label);
    let tagProps;
    let Component;
    if (typeof onClick === 'function' && !disabled) {
      tagProps = {
        className: tagClassName,
        onClick,
        type: 'button'
      };
      Component = component || 'button';
    } else {
      tagProps = {
        className: tagClassName
      };
      Component = component || 'div';
    }
    return /*#__PURE__*/React.createElement(Component, _extends({}, tagProps, themeHash, andesDataAttributes, {
      ref: ref,
      id: componentId,
      "aria-live": ariaLive,
      "data-andes-tag": true,
      "data-andes-tag-type": "simple",
      "data-andes-tag-size": size,
      "data-andes-tag-color": computedColor
    }, computeDataAndesState({
      disabled
    }), getContainerCustomColorStyle(customColor)), /*#__PURE__*/React.createElement(React.Fragment, null, asset && size === 'large' && /*#__PURE__*/React.createElement(ThumbnailAsset$1, _extends({}, asset, {
      color: asset === null || asset === void 0 ? void 0 : asset.color,
      imageAlt: label
    })), icon && size === 'large' && /*#__PURE__*/React.createElement(TagIcon$1, {
      icon: icon
    }), content, typeof onClose === 'function' && getClose()));
  };
  Tag.displayName = 'Tag';
  return Tag;
}();
var Tag$1 = Tag;
class TagCollapsed extends React.Component {
  constructor(props) {
    super(props);
    _defineProperty(this, "getKeyByIndex", index => {
      const id = this.props.id;
      return `${id}-item-${index}`;
    });
    this.state = {
      moreTagsVisible: false
    };
    this.buttonRef = /*#__PURE__*/React.createRef();
    this.showMoreTags = this.showMoreTags.bind(this);
    this.hideMoreTags = this.hideMoreTags.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
  }
  getVisibleTags(tags) {
    const _this$props$limit = this.props.limit,
      limit = _this$props$limit === void 0 ? 2 : _this$props$limit;
    return tags.slice(0, limit);
  }
  getEllipseTag(tags) {
    var _tags$;
    const _this$props = this.props,
      _this$props$limit2 = _this$props.limit,
      limit = _this$props$limit2 === void 0 ? 2 : _this$props$limit2,
      _this$props$disabled = _this$props.disabled,
      disabled = _this$props$disabled === void 0 ? false : _this$props$disabled,
      i18n = _this$props.i18n;
    const moreTagsVisible = this.state.moreTagsVisible;
    const size = ((_tags$ = tags[0]) === null || _tags$ === void 0 || (_tags$ = _tags$.props) === null || _tags$ === void 0 ? void 0 : _tags$.size) || 'large';
    const label = `+${tags.length - limit}`;
    const buttonTagClassName = classNames(namespace, `${namespace}--${size}`, {
      [`${namespace}--selectable`]: !disabled,
      [`${namespace}--clickable`]: !disabled,
      [`${namespace}--disabled`]: disabled
    });
    const popperOptions = {
      modifiers: [{
        name: 'offset',
        options: {
          offset: [0, 6]
        }
      }]
    };
    const listItemTagClassName = classNames(`${namespaceCollapsed}__wrap`, `${namespaceCollapsed}__list-item`);
    return /*#__PURE__*/React.createElement("li", {
      className: listItemTagClassName,
      onKeyDown: this.onKeyDown
    }, /*#__PURE__*/React.createElement(Popper, {
      interactive: true,
      placement: "bottom-start",
      trigger: "click",
      visible: moreTagsVisible,
      popperOptions: popperOptions,
      onClickOutside: this.hideMoreTags,
      zIndex: 10000,
      content: /*#__PURE__*/React.createElement("div", {
        className: `${namespaceCollapsed}__more`,
        role: "dialog",
        onBlur: this.onBlur
      }, /*#__PURE__*/React.createElement("ul", {
        className: `${namespaceCollapsed}__list`
      }, tags.slice(limit).map((tag, index) => /*#__PURE__*/React.createElement("li", {
        key: this.getKeyByIndex(index),
        className: `${namespaceCollapsed}__list-item`
      }, tag)))),
      focusScope: true
    }, /*#__PURE__*/React.createElement("button", {
      onClick: moreTagsVisible ? this.hideMoreTags : this.showMoreTags,
      "aria-haspopup": "dialog",
      "aria-expanded": moreTagsVisible,
      "aria-label": i18n.pgettext(namespace, 'mostrar {0} mas', tags.length - limit),
      className: buttonTagClassName,
      ref: this.buttonRef,
      disabled: disabled,
      type: "button",
      "data-andes-tag-collapsed-ellipse": true
    }, /*#__PURE__*/React.createElement("span", {
      className: `${namespace}__label`
    }, label))));
  }
  showMoreTags() {
    const _this$props$disabled2 = this.props.disabled,
      disabled = _this$props$disabled2 === void 0 ? false : _this$props$disabled2;
    if (disabled) return;
    this.setState({
      moreTagsVisible: true
    });
  }
  hideMoreTags() {
    this.setState({
      moreTagsVisible: false
    });
  }
  focusInCurrentTarget(event) {
    const relatedTarget = event.relatedTarget,
      currentTarget = event.currentTarget;
    if (relatedTarget === null) return false;
    let target = relatedTarget.parentNode;
    while (target !== null) {
      if (target === currentTarget) return true;
      target = target.parentNode;
    }
    return false;
  }
  onBlur(event) {
    if (!this.focusInCurrentTarget(event)) {
      this.hideMoreTags();
    }
  }
  onKeyDown(event) {
    var _this$buttonRef;
    const code = event.code;
    const button = (_this$buttonRef = this.buttonRef) === null || _this$buttonRef === void 0 ? void 0 : _this$buttonRef.current;
    if (code === ESCAPE && button) {
      this.hideMoreTags();
      button.focus();
    }
  }
  render() {
    const _this$props2 = this.props,
      children = _this$props2.children,
      _this$props2$limit = _this$props2.limit,
      limit = _this$props2$limit === void 0 ? 2 : _this$props2$limit,
      srLabel = _this$props2.srLabel,
      innerRef = _this$props2.innerRef;
    const collapsedTagClass = classNames(`${namespaceCollapsed}`, `${namespaceCollapsed}__list`, `${namespaceCollapsed}__list-main`);
    const tagChildren = React.Children.toArray(children);
    const exceedLimit = tagChildren.length > limit;
    return /*#__PURE__*/React.createElement("ul", {
      className: collapsedTagClass,
      "aria-label": srLabel,
      ref: innerRef
    }, this.getVisibleTags(tagChildren).map((tag, index) => /*#__PURE__*/React.createElement("li", {
      key: this.getKeyByIndex(index),
      className: `${namespaceCollapsed}__list-item`
    }, tag)), exceedLimit && this.getEllipseTag(tagChildren));
  }
}
TagCollapsed.displayName = 'TagCollapsed';
var TagCollapsed$1 = TagCollapsed;
const _excluded = ["ref"];
const TagCollapsedContainer = /*#__PURE__*/function () {
  const TagCollapsedContainer = _ref => {
    let ref = _ref.ref,
      props = _objectWithoutProperties(_ref, _excluded);
    const componentId = useId(props.id);
    const _useAndesContext = useAndesContext('@andes/tag', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations,
        'zh-CN': zhCnTranslations
      }),
      i18n = _useAndesContext.i18n;
    return /*#__PURE__*/React.createElement(TagCollapsed$1, _extends({}, props, {
      id: componentId,
      i18n: i18n,
      innerRef: ref
    }));
  };
  TagCollapsedContainer.displayName = 'TagCollapsedContainer';
  return TagCollapsedContainer;
}();
var TagCollapsedContainer$1 = TagCollapsedContainer;
const TagChoice = /*#__PURE__*/function () {
  const TagChoice = _ref => {
    let asset = _ref.asset,
      className = _ref.className,
      defaultSelected = _ref.defaultSelected,
      id = _ref.id,
      label = _ref.label,
      onClick = _ref.onClick,
      selected = _ref.selected,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? 'large' : _ref$size,
      srLabel = _ref.srLabel,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      ref = _ref.ref;
    const componentId = useId(id);
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const isUncontrolled = useMemo(() => selected === undefined, []);
    const _useState = useState(defaultSelected),
      _useState2 = _slicedToArray(_useState, 2),
      uncontrolledSelected = _useState2[0],
      setUncontrolledSelected = _useState2[1];
    const isSelected = () => {
      if (isUncontrolled) return uncontrolledSelected;
      return selected;
    };
    const handleClick = event => {
      if (isUncontrolled) {
        setUncontrolledSelected(state => !state);
      }
      if (typeof onClick === 'function') onClick(event, !isSelected());
    };
    const tagClassName = classNames(namespace, `${namespace}--choice`, className, {
      [`${namespace}--clickable`]: !disabled,
      [`${namespace}--${size}`]: size,
      [`${namespace}--selected`]: isSelected(),
      [`${namespace}--disabled`]: disabled
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('tag'));
    return /*#__PURE__*/React.createElement("button", _extends({
      className: tagClassName,
      ref: ref,
      id: componentId,
      onClick: handleClick,
      "aria-pressed": isSelected(),
      type: "button",
      disabled: disabled,
      "data-andes-tag": true,
      "data-andes-tag-type": "choice"
    }, themeHash, computeDataAndesState({
      disabled,
      selected: isSelected()
    })), asset && size === 'large' && /*#__PURE__*/React.createElement(ThumbnailAsset$1, _extends({}, asset, {
      imageAlt: label
    })), /*#__PURE__*/React.createElement("div", {
      className: `${namespace}__label`
    }, srLabel && /*#__PURE__*/React.createElement(VisuallyHidden, null, srLabel), label), isSelected() && !disabled && /*#__PURE__*/React.createElement("span", {
      "aria-hidden": true,
      className: `${namespace}__icon-container ${namespace}__icon-container-check`,
      "data-andes-tag-check-icon": true
    }, /*#__PURE__*/React.createElement(FeedbackPositive16, {
      color: "currentColor"
    })));
  };
  TagChoice.displayName = 'TagChoice';
  return TagChoice;
}();
var TagChoice$1 = TagChoice;
export { Tag$1 as Tag, TagChoice$1 as TagChoice, TagCollapsedContainer$1 as TagCollapsed };
