// nordic-pages-discovery/lib/islands/get-islands-from-file/islands-of-file/index.js
const { APP_FILE_MEMBERS } = require('../../../constants/file');
const { NordicIslandsError } = require('../../../error/nordic-islands-error');
const { importsCache } = require('../../../cache/imports-cache');
const { astCache } = require('../../../cache/ast-cache');
const {
  getCompleteComponentInfoByFilePath,
} = require('../../../helpers/babel/component-info');
const { getCompleteIslandChild } = require('./island-child');
const { getImportedComponentsFromFile } = require('../imported-components');
const { getSortedJSXElementsOfComponent } = require('../jsx-elements');
const {
  getLookingForIslandsInComponentError,
  getIncorrectUseOfESMError,
} = require('../../../error/get-discovery-error');

/**
 * @param {string} filePath
 * @param {{ importName: string, namedExport: boolean, property: string | null | undefined }} componentInfo
 */
const getIslandsFromComponent = (
  filePath,
  { importName, namedExport, property },
) => {
  try {
    const componentIslandsFromCache = property
      ? null
      : importsCache.getImport({
          importName,
          path: filePath,
          namedExport,
        });

    if (componentIslandsFromCache) {
      return componentIslandsFromCache;
    }

    const ast = astCache.getAstByFilePath(filePath);

    const { islandWrapper, importedComponents } = getImportedComponentsFromFile(
      filePath,
      ast,
    );

    if (!Object.keys(importedComponents).length && !islandWrapper) {
      return {};
    }

    const sortedComponents = getSortedJSXElementsOfComponent(
      ast,
      filePath,
      { importedComponents, islandWrapper },
      { importName, namedExport, property },
    );

    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return getIslands(sortedComponents, filePath);
  } catch (error) {
    throw new NordicIslandsError(
      getLookingForIslandsInComponentError(error.message),
      {
        type: `${APP_FILE_MEMBERS.PAGE} or ${APP_FILE_MEMBERS.COMPONENT}`,
        path: filePath,
      },
      error,
    );
  }
};

/**
 * @param {import('../jsx-elements/elements-of-ast').JSXElementsManifest} components
 * @param {string} filePath
 * @returns {import('../../../helpers/file-system/islands-manifest').IslandsManifest | {}}
 */
const getIslands = (components, filePath) => {
  const islands = {};

  Object.keys(components).forEach((componentNameInFile) => {
    const component = components[componentNameInFile];

    if (component.islands) {
      Object.assign(islands, component.islands);

      return;
    }

    if (!component.path) {
      return;
    }

    const componentInfo = getCompleteComponentInfoByFilePath(component.path, {
      name: component.importName,
      namedExport: component.namedExport,
      isIsland: component.isIsland,
    });

    if (!componentInfo.path) {
      return;
    }

    if (!componentInfo.exportedName) {
      throw new NordicIslandsError(
        getIncorrectUseOfESMError({
          importName: component.importName,
          originPath: filePath,
          finalPath: componentInfo.path,
        }),
        {
          type: `${APP_FILE_MEMBERS.PAGE} or ${APP_FILE_MEMBERS.COMPONENT}`,
          path: filePath,
        },
      );
    }

    if (component.isIsland) {
      const islandChild = getCompleteIslandChild(componentInfo);

      Object.assign(islands, islandChild);

      importsCache.addImport({
        ...component,
        componentPath: componentInfo.path,
        initialPath: filePath,
        islands: islandChild,
      });

      return;
    }

    const componentIslands = getIslandsFromComponent(componentInfo.path, {
      importName: componentInfo.exportedName,
      namedExport: componentInfo.namedExport,
      property: component.property,
    });

    importsCache.addImport({
      ...component,
      componentPath: componentInfo.path,
      initialPath: component.path,
      islands: componentIslands,
      localNamedExport: componentInfo.namedExport,
    });

    Object.assign(islands, componentIslands);
  });

  return islands;
};

module.exports = {
  getIslands,
};
