// nordic-pages-discovery/lib/islands/get-islands-from-file/jsx-elements/react-component-nodes/export-declaration.js
const babelTypes = require('@babel/types');

const { IMPORT_TYPES } = require('../../../../constants/islands');
const { assignNodeValuesToDestination } = require('./nodes-assigner');
const {
  isIdentifierDefaultExport,
  isFunctionDefaultExport,
  isObjectDefaultExport,
  isVariableNamedExport,
  isFunctionExpressionNamedExport,
  isObjectExport,
} = require('../../../../helpers/babel/export-statement');

/**
 * @param {import('.').NodesManifest} nodes
 * @param {babel.NodePath<babelTypes.ExportDefaultDeclaration>} nodePath
 */
const addNodeExportedAsDefaultToDestination = (nodes, nodePath) => {
  const exportedNodesData = [];

  // is "export default Component;"
  if (isIdentifierDefaultExport(nodePath)) {
    exportedNodesData.push({
      importKeys: {
        importKey: IMPORT_TYPES.DEFAULT,
        localKey: nodePath.node.declaration.name,
      },
      info: { namedExport: false },
      nodesWithProperty: [nodePath.node.declaration.name],
    });
  }

  // is "export default (...) => {...};" or "export default function (...) {...}"
  if (isFunctionDefaultExport(nodePath)) {
    exportedNodesData.push({
      importKeys: {
        importKey: IMPORT_TYPES.DEFAULT,
        localKey: IMPORT_TYPES.DEFAULT,
      },
      info: { node: nodePath.node, namedExport: false },
    });
  }

  // is "export default {};"
  if (isObjectDefaultExport(nodePath)) {
    nodePath.node.declaration.properties?.forEach((property) => {
      if (
        !babelTypes.isObjectProperty(property) ||
        !babelTypes.isIdentifier(property.value) ||
        !babelTypes.isIdentifier(property.key)
      ) {
        return;
      }

      exportedNodesData.push({
        importKeys: {
          importKey: property.key.name,
          localKey: property.value.name,
        },
        info: { namedExport: true },
        nodesWithProperty: [property.value.name, property.key.name],
      });
    });
  }

  assignNodeValuesToDestination(nodes, exportedNodesData);
};

/**
 * @param {import('.').NodesManifest} nodes
 * @param {babel.NodePath<babelTypes.ExportNamedDeclaration>} nodePath
 */
const addNodeExportedAsNamedToDestination = (nodes, nodePath) => {
  const exportedNodesData = [];

  // is "export const Component = ..."
  if (isVariableNamedExport(nodePath)) {
    nodePath.node.declaration.declarations.forEach((declaration) => {
      if (!babelTypes.isIdentifier(declaration.id)) {
        return;
      }

      // example 1: export const Component = () => {...}
      // example 2: export const Component = React.memo(OtherComponent)
      // example 3: export const Component = React.memo(() => ...)
      // example 4: export const Component = React.memo(function () {...})
      if (
        (babelTypes.isCallExpression(declaration.init) &&
          !!declaration.init.arguments.filter(
            (arg) =>
              babelTypes.isArrowFunctionExpression(arg) ||
              babelTypes.isFunctionExpression(arg) ||
              babelTypes.isIdentifier(arg),
          ).length) ||
        babelTypes.isArrowFunctionExpression(declaration.init)
      ) {
        exportedNodesData.push({
          importKeys: {
            importKey: declaration.id.name,
            localKey: declaration.id.name,
          },
          info: { node: declaration, namedExport: true },
        });
      }
    });
  }

  // is "export function Component() {}"
  if (isFunctionExpressionNamedExport(nodePath)) {
    exportedNodesData.push({
      importKeys: {
        importKey: nodePath.node.declaration.id.name,
        localKey: nodePath.node.declaration.id.name,
      },
      info: { node: nodePath.node.declaration, namedExport: true },
    });
  }

  // is "export { Component };" or "export { OtherComponent as Component };"
  if (isObjectExport(nodePath)) {
    nodePath.node.specifiers?.forEach((specifier) => {
      if (
        babelTypes.isExportSpecifier(specifier) &&
        babelTypes.isIdentifier(specifier.local) &&
        babelTypes.isIdentifier(specifier.exported)
      ) {
        exportedNodesData.push({
          importKeys: {
            importKey: specifier.exported.name,
            localKey: specifier.local.name,
          },
          info: { namedExport: true },
          nodesWithProperty: [specifier.local.name, specifier.exported.name],
        });
      }
    });
  }

  assignNodeValuesToDestination(nodes, exportedNodesData);
};

module.exports = {
  addNodeExportedAsDefaultToDestination,
  addNodeExportedAsNamedToDestination,
};
