// nordic-pages-discovery/lib/islands/get-islands-from-file/jsx-elements/react-component-nodes/index.js
const traverse = require('@babel/traverse').default;
const babelTypes = require('@babel/types');

const { IMPORT_TYPES } = require('../../../../constants/islands');
const { assignNodeValuesToDestination } = require('./nodes-assigner');
const {
  addNodeExportedAsDefaultToDestination,
  addNodeExportedAsNamedToDestination,
} = require('./export-declaration');
const {
  isPotencialReactClassDeclaration,
  isPotencialReactClassNamedExport,
  isPotencialReactClassDefaultExport,
} = require('../../../../helpers/babel/export-statement');

/**
 * @typedef {{ importName: string, node: import('../../../../cache/ast-cache').BabelAST, namedExport: boolean, key?: string }} ComponentNode
 * @typedef {Record<string, ComponentNode | undefined>} NodesManifest
 */

const addNodeDeclarationDataFromObject = (
  nodes,
  declaration,
  nodesDeclarationData,
  declarationId,
  propertyKeyName,
) => {
  declaration.properties.forEach((property) => {
    if (
      !babelTypes.isObjectProperty(property) ||
      babelTypes.isLiteral(property.value)
    ) {
      return;
    }

    let propertyKey = propertyKeyName ? `${propertyKeyName}-` : '';

    if (!babelTypes.isIdentifier(property.key) && property.computed) {
      propertyKey = `${propertyKey}computed`;
    } else {
      propertyKey = `${propertyKey}${property.key.name}`;
    }

    if (babelTypes.isObjectExpression(property.value)) {
      // example: const myComponents = { Components: { ComponentC: MyComponent } };
      addNodeDeclarationDataFromObject(
        nodes,
        property.value,
        nodesDeclarationData,
        declarationId,
        propertyKey,
      );
    } else {
      nodesDeclarationData.push({
        importKeys: {
          importKey: nodes[declarationId]?.key
            ? `${nodes[declarationId].key}-${propertyKey}`
            : null,
          localKey: `${declarationId}-${propertyKey}`,
        },
        info: { node: property },
      });
    }
  });
};

/**
 * @param {import('../../../../cache/ast-cache').BabelAST} ast
 * @returns {NodesManifest}
 */
const getPotentialReactComponentNodes = (ast) => {
  const nodes = {};

  traverse(ast, {
    noScope: true,
    AssignmentExpression(path) {
      if (
        !babelTypes.isExpressionStatement(path.parent) ||
        !babelTypes.isProgram(path.parentPath.parent)
      ) {
        return;
      }

      const { left } = path.node;

      // is "varName = ..."
      if (babelTypes.isIdentifier(left)) {
        assignNodeValuesToDestination(nodes, [
          {
            importKeys: {
              localKey: left.name,
            },
            info: { node: path.node },
          },
        ]);

        return;
      }

      if (
        !babelTypes.isMemberExpression(left) ||
        !babelTypes.isIdentifier(left.object)
      ) {
        return;
      }

      // is "varName.prop = ..." or "varName[prop] = ..."
      const { object: obj, property } = left;

      assignNodeValuesToDestination(nodes, [
        {
          importKeys: {
            localKey: obj.name,
          },
          info: { node: path.node },
        },
      ]);

      if (!babelTypes.isIdentifier(property)) {
        return;
      }

      assignNodeValuesToDestination(nodes, [
        {
          importKeys: {
            localKey: `${obj.name}-${property.name}`,
          },
          info: { node: path.node },
        },
      ]);
    },
    VariableDeclaration(path) {
      if (
        !babelTypes.isProgram(path.parent) &&
        !babelTypes.isExportDeclaration(path.parent)
      ) {
        return;
      }

      path.node.declarations.forEach((declaration) => {
        if (
          !babelTypes.isIdentifier(declaration.id) ||
          (!babelTypes.isFunctionExpression(declaration.init) &&
            !babelTypes.isArrowFunctionExpression(declaration.init) &&
            !babelTypes.isObjectExpression(declaration.init) &&
            !babelTypes.isCallExpression(declaration.init)) ||
          (babelTypes.isCallExpression(declaration.init) &&
            !declaration.init.arguments.filter(
              (arg) =>
                babelTypes.isArrowFunctionExpression(arg) ||
                babelTypes.isFunctionExpression(arg) ||
                babelTypes.isIdentifier(arg) ||
                (babelTypes.isMemberExpression(arg) &&
                  babelTypes.isIdentifier(arg.object)),
            ).length)
        ) {
          return;
        }

        // example 1: const MyComponent = function () {};
        // example 2: const MyComponent = () => {};
        // example 3: const MyComponent = fn(() => {});
        // example 4: const MyComponent = fn(MyComponent);
        // example 5: const MyComponent = fn(function () {});
        // example 6: const MyComponent = fn(components[i]);
        // example 7: const MyComponent = fn(components.MyComponent);
        const nodesDeclarationData = [
          {
            importKeys: {
              importKey: nodes[declaration.id.name]?.key,
              localKey: declaration.id.name,
            },
            info: { node: declaration.init },
          },
        ];

        // example: const myComponents = { ComponentA,  MyComponent: ComponentB, Others: { ComponentC: OtherComponent } };
        if (babelTypes.isObjectExpression(declaration.init)) {
          addNodeDeclarationDataFromObject(
            nodes,
            declaration.init,
            nodesDeclarationData,
            declaration.id.name,
          );
        }

        assignNodeValuesToDestination(nodes, nodesDeclarationData);
      });
    },
    FunctionDeclaration(path) {
      if (
        !babelTypes.isProgram(path.parent) ||
        !babelTypes.isIdentifier(path.node.id)
      ) {
        return;
      }

      // example: function myFunction() {}
      assignNodeValuesToDestination(nodes, [
        {
          importKeys: {
            importKey: nodes[path.node.id.name]?.key,
            localKey: path.node.id.name,
          },
          info: { node: path.node },
        },
      ]);
    },
    ExportNamedDeclaration(path) {
      if (!babelTypes.isProgram(path.parent)) {
        return;
      }

      addNodeExportedAsNamedToDestination(nodes, path);
    },
    ExportDefaultDeclaration(path) {
      if (!babelTypes.isProgram(path.parent)) {
        return;
      }

      // example: export default MyComponent;
      addNodeExportedAsDefaultToDestination(nodes, path);
    },
    ClassDeclaration(path) {
      if (
        (!babelTypes.isProgram(path.parent) &&
          !babelTypes.isExportDefaultDeclaration(path.parent) &&
          !babelTypes.isExportNamedDeclaration(path.parent)) ||
        (!babelTypes.isProgram(path.parentPath.parent) &&
          (babelTypes.isExportDefaultDeclaration(path.parent) ||
            babelTypes.isExportNamedDeclaration(path.parent))) ||
        !isPotencialReactClassDeclaration(path)
      ) {
        return;
      }

      let importKey = nodes[path.node.id.name]?.key;
      let namedExport = null;

      // example: export class Component extends ReactClass { render(){} }
      if (isPotencialReactClassNamedExport(path)) {
        importKey = path.node.id.name;
        namedExport = true;
      }

      // example: export default class Component extends ReactClass { render(){} }
      if (isPotencialReactClassDefaultExport(path)) {
        importKey = IMPORT_TYPES.DEFAULT;
        namedExport = false;
      }

      assignNodeValuesToDestination(nodes, [
        {
          importKeys: {
            importKey,
            localKey: path.node.id.name,
          },
          info: { node: path.node, namedExport },
        },
      ]);
    },
  });

  return nodes;
};

module.exports = {
  getPotentialReactComponentNodes,
};
