// babel-plugin-nordic-module-resolver/src/normalize-options/index.js
const path = require('path');

const { createSelector } = require('reselect');

const { defaultExtensions } = require('./constants');
const {
  normalizeCwd,
  normalizeRoot,
  normalizeAlias,
  normalizeTransformedFunctions,
  normalizeLoglevel,
} = require('./normalizer');

module.exports = createSelector(
  // The currentFile should have an extension; otherwise it's considered a special value
  (currentFile) =>
    currentFile.includes('.') ? path.dirname(currentFile) : currentFile,
  (_, opts) => opts,
  (currentFile, opts) => {
    const cwd = normalizeCwd(opts.cwd, currentFile);
    const extensions = opts.extensions || defaultExtensions;

    return {
      cwd,
      extensions,
      root: normalizeRoot(opts.root, cwd),
      alias: normalizeAlias(opts.alias),
      loglevel: normalizeLoglevel(opts.loglevel),
      transformFunctions: normalizeTransformedFunctions(
        opts.transformFunctions,
      ),
      stripExtensions: opts.stripExtensions || extensions,
      customResolvePath: opts.resolvePath,
    };
  },
);
