// babel-plugin-nordic-module-resolver/src/normalize-options/normalizer.js
const fs = require('fs');
const path = require('path');

const findBabelConfig = require('find-babel-config');
const glob = require('glob');

const { defaultTransformedFunctions } = require('./constants');
const { getAliasSubstitute, getAliasTarget } = require('./alias-data');
const { findPackageJson } = require('../utils');

function isRegExp(string) {
  return string.startsWith('^') || string.endsWith('$');
}

const specialCwd = {
  babelrc: (startPath) => findBabelConfig.sync(startPath).file,
  packagejson: (startPath) => findPackageJson({ cwd: startPath }),
};

function normalizeCwd(optsCwd, currentFile) {
  let cwd;

  if (optsCwd in specialCwd) {
    const startPath = currentFile === 'unknown' ? './' : currentFile;

    const computedCwd = specialCwd[optsCwd](startPath);

    cwd = computedCwd ? path.dirname(computedCwd) : null;
  } else {
    cwd = optsCwd;
  }

  return cwd || process.cwd();
}

function normalizeRoot(optsRoot, cwd) {
  if (!optsRoot) {
    return [];
  }

  const rootArray = Array.isArray(optsRoot) ? optsRoot : [optsRoot];

  return rootArray
    .map((dirPath) => path.resolve(cwd, dirPath))
    .reduce((resolvedDirs, absDirPath) => {
      if (!glob.hasMagic(absDirPath)) {
        return [...resolvedDirs, absDirPath];
      }

      const roots = glob
        .sync(absDirPath)
        .filter((resolvedPath) => fs.lstatSync(resolvedPath).isDirectory());

      return [...resolvedDirs, ...roots];
    }, []);
}

function normalizeAlias(optsAlias) {
  if (!optsAlias) {
    return [];
  }

  const aliasArray = Array.isArray(optsAlias) ? optsAlias : [optsAlias];

  return aliasArray.reduce((aliasPairs, alias) => {
    Object.entries(alias).forEach(([key, value]) => {
      const isKeyRegExp = isRegExp(key);

      aliasPairs.push([
        getAliasTarget(key, isKeyRegExp),
        getAliasSubstitute(value, isKeyRegExp),
      ]);
    });

    return aliasPairs;
  }, []);
}

function normalizeTransformedFunctions(optsTransformFunctions) {
  if (!optsTransformFunctions) {
    return defaultTransformedFunctions;
  }

  return defaultTransformedFunctions.concat(optsTransformFunctions);
}

function normalizeLoglevel(optsLoglevel) {
  return optsLoglevel || 'warn';
}

module.exports = {
  normalizeCwd,
  normalizeRoot,
  normalizeAlias,
  normalizeTransformedFunctions,
  normalizeLoglevel,
};
