// nordic-csp-engine/lib/cron/manager/manager.js
const { CronJob } = require('cron');
const { circularDeepEqual } = require('fast-equals');
const { MeliStatsD } = require('@meli/statsd');

const { CRON_CONFIG_SCHEMA } = require('../../constants/cron');
const { logger } = require('../../logger');
const cronStorage = require('../storage');
const { TELEMETRY_CRON_JOB_ERROR } = require('../../constants/telemetry');

const statsd = new MeliStatsD();

/**
 * Manages cron jobs and handles ticks
 * @typedef {{key: string, value: any}} CronHandlerData
 * @typedef {CronHandlerData | null} CronHandlerResult
 * @typedef {() => Promise<CronHandlerResult> | CronHandlerResult} CronHandler
 * @typedef {{name: string, cronTime: string, handler: CronHandler}} CronConfig
 * @typedef {(data: CronHandlerData) => void} onTickCallback
 */
class CronManager {
  /**
   * @param {{onTick: onTickCallback}} options
   */
  constructor({ onTick }) {
    /**
     * @private
     * @type {onTickCallback}
     */
    this.onTickCallback = onTick;
    /**
     * @private
     * @type {Map<string, CronJob>}
     */
    this.cronJobs = new Map();
    /** @private */
    this.baseConfig = {
      cronTime: '0 0 * * *',
    };
  }

  /**
   * @public
   * @param {CronConfig} config
   */
  addJob(config) {
    const { cronTime, name, handler } = { ...this.baseConfig, ...config };

    this.validateConfig(config);

    const cronJob = new CronJob(cronTime, async () =>
      this.onTick(name, handler),
    );

    this.cronJobs.set(name, cronJob);
  }

  /** @public */
  start() {
    this.cronJobs.forEach((cronJob) => {
      if (!cronJob.running) {
        cronJob.start();
        void cronJob.fireOnTick();
      }
    });
  }

  /** @public */
  stop() {
    this.cronJobs.forEach((cronJob) => cronJob.stop());
  }

  /**
   * @public
   * @param {string} name
   */
  stopByName(name) {
    const cronJob = this.cronJobs.get(name);

    if (!cronJob) {
      throw new Error(`CronStopError: job with name ${name} not found`);
    }

    cronJob.stop();
  }

  /**
   * @private
   * @param {string} name
   * @param {() => Promise<CronHandlerResult>} handler
   */
  async onTick(name, handler) {
    try {
      const data = await handler();

      if (!data) {
        logger.debug(
          `CronHandlerEmptyData: Cron ${name} handler returned empty result data`,
        );

        return;
      }

      const { key, value } = data;

      const storedValue = await cronStorage.get(key);

      if (circularDeepEqual(value, storedValue)) {
        logger.debug(
          `CronHandlerEqualDataValue: Cron ${name} handler returned equal value as the stored value`,
        );

        return;
      }

      await cronStorage.set(data.key, data.value);

      this.onTickCallback(data);
    } catch (err) {
      logger.debug(
        `CronHandlerError: Cron ${name} handler failed, ${err.message}`,
      );

      statsd.increment(TELEMETRY_CRON_JOB_ERROR, {
        error: 'CRON_JOB_ERROR',
      });
    }
  }

  /**
   * @private
   * @param {CronConfig} config
   */
  validateConfig(config) {
    const { error } = CRON_CONFIG_SCHEMA.validate(config);

    if (error) {
      error.name = `CronConfig${error.name}`;

      throw error;
    }

    if (this.cronJobs.get(config.name)) {
      throw new Error(
        `CronNameError: job with name ${config.name} already exists`,
      );
    }
  }
}

module.exports = CronManager;
