// @andes/textfield/build/esm/masks/index.production.js
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
const validateDate = value => {
  const _value$split = value.split('/'),
    _value$split2 = _slicedToArray(_value$split, 3),
    day = _value$split2[0],
    month = _value$split2[1],
    year = _value$split2[2];
  const dayNumber = parseInt(day, 10);
  const monthNumber = parseInt(month, 10);
  const yearNumber = parseInt(year, 10);
  let newValue = value;
  if ((day === null || day === void 0 ? void 0 : day[0]) > '3' && !(day !== null && day !== void 0 && day[1])) {
    newValue = `0${day[0]}`;
  }
  if (dayNumber > 31 || dayNumber === 0) {
    return day[0];
  }
  if ((month === null || month === void 0 ? void 0 : month[0]) > '1' && !(month !== null && month !== void 0 && month[1])) {
    newValue = `${day}/0${month[0]}`;
  }
  if (monthNumber > 12 || monthNumber === 0) {
    return `${day}/${(month === null || month === void 0 ? void 0 : month[0]) === '1' || (month === null || month === void 0 ? void 0 : month[0]) === '0' ? month === null || month === void 0 ? void 0 : month[0] : ''}`;
  }
  const daysInMonth = [31, Number(year) % 4 === 0 ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  if (monthNumber && Number(String(monthNumber)[0]) !== 0 && dayNumber > daysInMonth[monthNumber - 1]) {
    return `${day}/${month[0]}`;
  }
  if (yearNumber === 0) return `${day}/${month}`;
  if (yearNumber) {
    const firstDigit = year === null || year === void 0 ? void 0 : year[0];
    const secondDigit = year === null || year === void 0 ? void 0 : year[1];
    const thirdDigit = year === null || year === void 0 ? void 0 : year[2];
    const fourthDigit = year === null || year === void 0 ? void 0 : year[3];
    if (firstDigit && firstDigit !== '1' && firstDigit !== '2') {
      return `${day}/${month}/`;
    }
    if (secondDigit && !(firstDigit === '1' && secondDigit === '9' || firstDigit === '2' && (secondDigit === '1' || secondDigit === '0'))) {
      return `${day}/${month}/${firstDigit}`;
    }
    if (thirdDigit && !(secondDigit === '9' || secondDigit === '0' || secondDigit === '1' && thirdDigit === '0')) {
      return `${day}/${month}/${firstDigit}${secondDigit}`;
    }
    if (fourthDigit && !(secondDigit === '9' || secondDigit === '0' || secondDigit === '1' && fourthDigit === '0')) {
      return `${day}/${month}/${firstDigit}${secondDigit}${thirdDigit}`;
    }
  }
  return newValue;
};
const validateDateTime = value => {
  const _value$split3 = value.split(' '),
    _value$split4 = _slicedToArray(_value$split3, 2),
    dateStr = _value$split4[0],
    timeStr = _value$split4[1];
  const _ref = timeStr ? timeStr.split(':') : [],
    _ref2 = _slicedToArray(_ref, 3),
    hour = _ref2[0],
    minute = _ref2[1],
    seconds = _ref2[2];
  const dateValue = validateDate(dateStr);
  const newValue = timeStr ? `${dateValue} ${timeStr}` : dateValue;
  if (parseInt(hour === null || hour === void 0 ? void 0 : hour[0], 10) > 2) return dateValue;
  const hourNumber = parseInt(hour, 10);
  if (hourNumber > 23) return `${dateValue} ${hour[0]}`;
  if (parseInt(minute === null || minute === void 0 ? void 0 : minute[0], 10) > 5) return `${dateValue} ${hour}`;
  if (parseInt(seconds === null || seconds === void 0 ? void 0 : seconds[0], 10) > 5) return `${dateValue} ${hour}:${minute}`;
  return newValue;
};
var date$2 = {
  mask: 'DD/DD/DDDD',
  tokens: {
    D: {
      pattern: /\d/
    }
  },
  eager: true,
  preProcess: value => value.replace(/[^\d/:]/g, ''),
  postProcess: value => validateDate(value)
};
var money$4 = {
  mask: 'N,DD',
  tokens: {
    N: {
      pattern: /\d/,
      multiple: true
    },
    D: {
      pattern: /\d/,
      optional: true
    }
  },
  preProcess: val => val.replace(/\./g, ''),
  postProcess: val => {
    if (!val || val === ',') return '';
    const sub = 3 - (val.includes(',') ? val.length - val.indexOf(',') : 0);
    return Intl.NumberFormat('es-AR', {
      style: 'currency',
      currency: 'ARS'
    }).format(Number(val.replace(',', '.'))).replace('$', '').replace(/\s/g, '').slice(0, sub ? -sub : undefined);
  }
};
var phone$2 = {
  mask: '(ZZ) DDD-DDDD',
  eager: true,
  tokens: {
    Z: {
      pattern: /[1-9]/
    },
    D: {
      pattern: /\d/
    }
  }
};
var zipcode$2 = {
  mask: 'ADDDD',
  tokens: {
    A: {
      pattern: /[a-z]/
    },
    D: {
      pattern: /\d/
    }
  }
};
var dni = {
  mask: 'IDDDDDDD',
  tokens: {
    I: {
      pattern: /[M|F|Xm|f|x0-9]/
    },
    D: {
      pattern: /\d/
    }
  }
};
var cuil = {
  mask: 'ZD-ZDDDDDDD-D',
  eager: true,
  tokens: {
    Z: {
      pattern: /[1-9]/
    },
    D: {
      pattern: /\d/
    }
  }
};
var datetime$1 = {
  mask: 'DD/DD/DDDD DD:DD:DD',
  tokens: {
    D: {
      pattern: /\d/
    }
  },
  eager: true,
  preProcess: value => value.replace(/[^\d/:]/g, ''),
  postProcess: value => validateDateTime(value)
};
var index$8 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  date: date$2,
  money: money$4,
  PHONE: phone$2,
  ZIPCODE: zipcode$2,
  DNI: dni,
  CUIL: cuil,
  datetime: datetime$1
});
var date$1 = {
  mask: 'DD/DD/DDDD',
  tokens: {
    D: {
      pattern: /\d/
    }
  },
  eager: true,
  preProcess: value => value.replace(/[^\d/:]/g, ''),
  postProcess: value => validateDate(value)
};
var money$3 = {
  mask: 'N,DD',
  tokens: {
    N: {
      pattern: /\d/,
      multiple: true
    },
    D: {
      pattern: /\d/,
      optional: true
    }
  },
  preProcess: val => val.replace(/\./g, ''),
  postProcess: val => {
    if (!val || val === ',') return '';
    const sub = 3 - (val.includes(',') ? val.length - val.indexOf(',') : 0);
    return Intl.NumberFormat('pt-BR', {
      style: 'currency',
      currency: 'BRL'
    }).format(Number(val.replace(',', '.'))).replace('R$', '').replace(/\s/g, '').slice(0, sub ? -sub : undefined);
  }
};
var phone$1 = {
  mask: '(ZZ) DDD-DDDD',
  eager: true,
  tokens: {
    Z: {
      pattern: /[1-9]/
    },
    D: {
      pattern: /\d/
    }
  }
};
var mobile = {
  mask: '(ZZ) DDDDD-DDDD',
  eager: true,
  tokens: {
    Z: {
      pattern: /[1-9]/
    },
    D: {
      pattern: /\d/
    }
  }
};
var zipcode$1 = {
  mask: 'DDDDD-DDD',
  eager: true,
  tokens: {
    A: {
      pattern: /[a-z]/
    },
    D: {
      pattern: /\d/
    }
  }
};
var cnpj = {
  mask: 'DD.DDD.DDD/DDDD-DD',
  eager: true,
  tokens: {
    D: {
      pattern: /\d/
    }
  }
};
var cpf = {
  mask: 'DDD.DDD.DDD-DD',
  eager: true,
  tokens: {
    D: {
      pattern: /\d/
    }
  }
};
var rg = {
  mask: 'DD.DDD.DDD-D',
  eager: true,
  tokens: {
    D: {
      pattern: /\d/
    }
  }
};
var datetime = {
  mask: 'DD/DD/DDDD DD:DD:DD',
  tokens: {
    D: {
      pattern: /\d/
    }
  },
  eager: true,
  preProcess: value => value.replace(/[^\d/:]/g, ''),
  postProcess: value => validateDateTime(value)
};
var index$7 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  date: date$1,
  money: money$3,
  PHONE: phone$1,
  MOBILE: mobile,
  ZIPCODE: zipcode$1,
  CNPJ: cnpj,
  CPF: cpf,
  RG: rg,
  datetime: datetime
});
var money$2 = {
  mask: 'N',
  tokens: {
    N: {
      pattern: /\d/,
      multiple: true
    },
    D: {
      pattern: /\d/,
      optional: true
    }
  },
  preProcess: val => val.replace(/\./g, ''),
  postProcess: val => {
    if (!val) return '';
    const sub = 3 - (val.includes(',') ? val.length - val.indexOf(',') : 0);
    return Intl.NumberFormat('es-AR', {
      style: 'currency',
      currency: 'ARS'
    }).format(Number(val.replace(',', '.'))).replace('$', '').replace(/\s/g, '').slice(0, sub ? -sub : undefined);
  }
};
var rut$1 = {
  mask: 'X',
  tokens: {
    X: {
      pattern: /[\dkK]/,
      multiple: true
    }
  },
  eager: true,
  preProcess: val => val.replace(/[^\dkK]/gi, '').slice(0, 9),
  postProcess: rawInput => {
    if (!rawInput) return '';
    const sanitized = rawInput.replace(/[^0-9kK]/g, '').toUpperCase().replace(/^0+/, '');
    if (!sanitized) return '';
    if (sanitized.length < 8) {
      const _ref = sanitized.match(/^(\d{1,2})(\d{0,3})?(\d{0,3})?/) || [],
        _ref2 = _slicedToArray(_ref, 4),
        _ref2$ = _ref2[1],
        part1 = _ref2$ === void 0 ? '' : _ref2$,
        _ref2$2 = _ref2[2],
        part2 = _ref2$2 === void 0 ? '' : _ref2$2,
        _ref2$3 = _ref2[3],
        part3 = _ref2$3 === void 0 ? '' : _ref2$3;
      return [part1, part2, part3].filter(Boolean).join('.');
    }
    const numberPart = sanitized.slice(0, -1);
    const verifier = sanitized.slice(-1);
    return `${numberPart.replace(/^(\d{1,2})(\d{3})(\d{3})$/, '$1.$2.$3')}-${verifier}`;
  }
};
var index$6 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  money: money$2,
  RUT: rut$1
});
var money$1 = {
  mask: 'N',
  tokens: {
    N: {
      pattern: /\d/,
      multiple: true
    },
    D: {
      pattern: /\d/,
      optional: true
    }
  },
  preProcess: val => val.replace(/\./g, ''),
  postProcess: val => {
    if (!val) return '';
    const sub = 3 - (val.includes(',') ? val.length - val.indexOf(',') : 0);
    return Intl.NumberFormat('es-AR', {
      style: 'currency',
      currency: 'ARS'
    }).format(Number(val.replace(',', '.'))).replace('$', '').replace(/\s/g, '').slice(0, sub ? -sub : undefined);
  }
};
var nit = {
  mask: 'DDD.DDD.DDD-D',
  eager: true,
  tokens: {
    D: {
      pattern: /\d/
    }
  }
};
var index$5 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  money: money$1,
  NIT: nit
});
var ruc$1 = {
  mask: 'DDDDDDDDDDDDD',
  eager: true,
  tokens: {
    D: {
      pattern: /\d/
    }
  }
};
var index$4 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  RUC: ruc$1
});
var date = {
  mask: 'DD/DD/DDDD',
  tokens: {
    D: {
      pattern: /\d/
    }
  },
  eager: true,
  preProcess: value => value.replace(/[^\d/:]/g, ''),
  postProcess: value => validateDate(value)
};
var money = {
  mask: 'N.DD',
  tokens: {
    N: {
      pattern: /\d/,
      multiple: true
    },
    D: {
      pattern: /\d/,
      optional: true
    }
  },
  preProcess: val => val.replace(/,/g, ''),
  postProcess: val => {
    if (!val || val === '.') return '';
    const sub = 3 - (val.includes('.') ? val.length - val.indexOf('.') : 0);
    return Intl.NumberFormat('es-MX', {
      style: 'currency',
      currency: 'MXN'
    }).format(Number(val.replace(',', '.'))).replace('$', '').replace(/\s/g, '').slice(0, sub ? -sub : undefined);
  }
};
var phone = {
  mask: '(ZZ) DDD-DDDD',
  eager: true,
  tokens: {
    Z: {
      pattern: /[1-9]/
    },
    D: {
      pattern: /\d/
    }
  }
};
var zipcode = {
  mask: 'DDDDD',
  tokens: {
    D: {
      pattern: /\d/
    }
  }
};
var rfc = {
  mask: 'LLLLDDDDDDLLL',
  eager: true,
  tokens: {
    L: {
      pattern: /[A-ZÑ&]/
    },
    D: {
      pattern: /\d/
    }
  }
};
var index$3 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  date: date,
  money: money,
  PHONE: phone,
  ZIPCODE: zipcode,
  RFC: rfc
});
var ruc = {
  mask: 'ZDDDDDDDDDDD',
  eager: true,
  tokens: {
    Z: {
      pattern: /[1-9]/
    },
    D: {
      pattern: /\d/
    }
  }
};
var index$2 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  RUC: ruc
});
var rut = {
  mask: 'DD.DDD.DDD.DDD',
  eager: true,
  tokens: {
    D: {
      pattern: /\d/
    }
  }
};
var index$1 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  RUT: rut
});
var rif = {
  mask: 'C-DDDDDDDD-D',
  eager: true,
  tokens: {
    C: {
      pattern: /[VEJGEP]/
    },
    D: {
      pattern: /\d/
    }
  },
  preProcess: val => {
    const upper = val.toUpperCase().replace(/[^A-Z0-9]/g, '');
    const firstChar = upper.charAt(0);
    if (!/[VEJGEP]/.test(firstChar)) {
      return '';
    }
    const digits = upper.slice(1).replace(/\D/g, '');
    return (firstChar + digits).slice(0, 10);
  }
};
var index = /*#__PURE__*/Object.freeze({
  __proto__: null,
  RIF: rif
});
export { index$8 as AR, index$7 as BR, index$6 as CL, index$5 as CO, index$4 as EC, index$3 as MX, index$2 as PE, index$1 as UY, index as VE };
