// @andes/common/build/esm/hooks/useModal.development.js
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { useState, useEffect } from 'react';

// Provide functionality of Array.includes to all browsers

/**
 * @param {string[] | string} array
 * @param {string} value
 */
const includes = (array, value) => array.indexOf(value) !== -1;
var includes$1 = includes;

/* eslint-disable quotes */
/* eslint-disable no-console */

/**
 * @param {boolean} condition
 * @param {string} message
 * @param {string} [type]
 */
const print = function (condition, message) {
  let type = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'error';
  if (condition) {
    return;
  }
  if (typeof console !== 'undefined') {
    let logFunction = /** @type {function | undefined} */console[(/** @type {keyof typeof console} */type)];
    if (typeof logFunction !== 'function') logFunction = console.error;
    logFunction(message);
  }
  try {
    throw new Error(message);
    // eslint-disable-next-line no-empty
  } catch (x) {}
};
const warning = process.env.NODE_ENV !== "'production'" && process.env.NODE_ENV !== "production" ? print : () => {};
var warning$1 = warning;
const WARNING_INERT_ATTRIBUTE = 'The \'inert\' attribute is not supported by this browser To ensure accessibility and prevent the background from being interactive, add the polyfill "wicg-inert": https://github.com/WICG/inert';
const TAG_NAME_INVALID = ['STYLE', 'SCRIPT'];
const isInertSupported = () => typeof window !== 'undefined' && 'inert' in HTMLElement.prototype;

/** @typedef {import('./hooks').UseModalProps} UseModalProps */
/** @typedef {import('./hooks').AddInertProps} AddInertProps */
/** @typedef {import('./hooks').RemoveInertProps} RemoveInertProps */
/** @typedef {import('./hooks').AddAriaHiddenProps} AddAriaHiddenProps */
/** @typedef {import('./hooks').RemoveAriaHiddenProps} RemoveAriaHiddenProps */

/**
 * @param {string} portalSelector
 * @param {HTMLElement | ShadowRoot | undefined} portalContainer
 */
const getElementsBySelector = (portalSelector, portalContainer) => {
  var _document;
  if (typeof window === 'undefined' || typeof document === 'undefined') {
    return {
      bodyElements: [],
      portalElements: []
    };
  }
  const bodyElements = /** @type {HTMLElement []} */(_document = document) !== null && _document !== void 0 && (_document = _document.body) !== null && _document !== void 0 && _document.children ? [].slice.call(document.body.children) : [];
  let portalElements = /** @type {Node []} */[];
  if (portalContainer instanceof ShadowRoot) {
    const portalInShadow = portalContainer.querySelector(portalSelector);
    if (portalInShadow) {
      portalElements = [portalInShadow];
    }
  } else {
    var _document2;
    portalElements = /** @type {Node []} */(_document2 = document) !== null && _document2 !== void 0 && _document2.querySelectorAll(portalSelector) ? [].slice.call(document.querySelectorAll(portalSelector)) : [];
  }
  return {
    bodyElements,
    portalElements
  };
};

/**
 * Recursively traverses all descendants of node (including shadow DOM),
 * skipping portalContainer, and applies a callback.
 * @param {HTMLElement} node
 * @param {HTMLElement | ShadowRoot | undefined} portalContainer
 * @param {(el: HTMLElement) => void} callback
 * @param {ShadowRoot | undefined} originalShadowRoot
 */
const traverseDOM = (node, portalContainer, callback, originalShadowRoot) => {
  if (!node) return;

  // If portalContainer is a ShadowRoot and this node is its host
  // Skip applying callback to host, but traverse inside the ShadowRoot
  if (originalShadowRoot && node.shadowRoot === originalShadowRoot) {
    const shadowChildren = Array.from(originalShadowRoot.children || []);
    for (let i = 0; i < shadowChildren.length; i += 1) {
      const child = shadowChildren[i];
      if (child instanceof HTMLElement) {
        traverseDOM(child, portalContainer, callback, originalShadowRoot);
      }
    }
    return;
  }
  callback(node);
  if (node === portalContainer) return;
  const children = [...Array.from(node.children || []), ...(node.shadowRoot && node.shadowRoot !== originalShadowRoot ? Array.from(node.shadowRoot.children || []) : [])];
  for (let i = 0; i < children.length; i += 1) {
    const child = children[i];
    if (child instanceof HTMLElement) {
      traverseDOM(child, portalContainer, callback, originalShadowRoot);
    }
  }
};

/**
 * Removes the given attribute from portalContainer and its ancestors up to rootNode.
 * @param {HTMLElement} portalEl
 * @param {HTMLElement} rootNode
 * @param {string} attribute
 */
const removeInertFromAncestors = (portalEl, rootNode, attribute) => {
  let current = portalEl;
  while (current && current !== rootNode.parentElement) {
    if (current.hasAttribute && current.hasAttribute(attribute)) {
      current.removeAttribute(attribute);
    }
    if (current === rootNode) break;
    // Walk up through shadow DOM hosts as well as parent elements
    let parent = null;
    if (current.parentElement) {
      parent = current.parentElement;
    } else if (current.getRootNode && current.getRootNode() instanceof ShadowRoot) {
      const root = current.getRootNode();
      if (root instanceof ShadowRoot && 'host' in root) {
        parent = root.host;
      }
    }
    if (!parent) break;
    if (parent instanceof HTMLElement) {
      current = parent;
    } else {
      break;
    }
  }
};

/**
 * Gets the actual portal element when portalContainer is a ShadowRoot
 * @param {HTMLElement | ShadowRoot | undefined} portalContainer
 * @param {string} portalSelector
 * @returns {HTMLElement | ShadowRoot | undefined}
 */
const getActualPortalElement = (portalContainer, portalSelector) => {
  if (!portalContainer) return undefined;
  if (portalContainer instanceof ShadowRoot) {
    const portals = portalContainer.querySelectorAll(portalSelector);
    const portalElement = portals.length > 0 ? portals[portals.length - 1] : null;
    return /** @type {HTMLElement | null} */portalElement || portalContainer;
  }
  return portalContainer;
};

/**
 * Recursively sets the given attribute on all descendants of node except those containing the portalContainer.
 * When portalContainer is found, removes the attribute from it and its ancestors up to node.
 * @param {HTMLElement} node
 * @param {HTMLElement | ShadowRoot | undefined} portalContainer
 * @param {string} attribute
 * @param {string} portalSelector
 */
const setAttributeRecursively = (node, portalContainer, attribute, portalSelector) => {
  const actualPortalElement = getActualPortalElement(portalContainer, portalSelector);
  const originalShadowRoot = portalContainer instanceof ShadowRoot ? portalContainer : undefined;
  traverseDOM(node, actualPortalElement, el => {
    if (el === actualPortalElement) {
      removeInertFromAncestors(el, node, attribute);
      return;
    }
    if (!includes$1(TAG_NAME_INVALID, el.tagName)) {
      el.setAttribute(attribute, 'true');
    }
  }, originalShadowRoot);
};

/**
 * Recursively removes inert/aria-hidden from all descendants of node except those containing the portalContainer.
 * @param {HTMLElement} node
 * @param {HTMLElement | ShadowRoot | undefined} portalContainer
 * @param {string} attribute
 */
const removeAttributeRecursively = (node, portalContainer, attribute) => {
  const originalShadowRoot = portalContainer instanceof ShadowRoot ? portalContainer : undefined;
  traverseDOM(node, portalContainer, el => {
    if (!includes$1(TAG_NAME_INVALID, el.tagName) && el.hasAttribute(attribute)) {
      el.removeAttribute(attribute);
    }
  }, originalShadowRoot);
};

/** @param {AddAriaHiddenProps} Props */
const addAriaHidden = _ref => {
  let _ref$outsideElements = _ref.outsideElements,
    outsideElements = _ref$outsideElements === void 0 ? [] : _ref$outsideElements,
    portalContainer = _ref.portalContainer,
    _ref$portalElements = _ref.portalElements,
    portalElements = _ref$portalElements === void 0 ? [] : _ref$portalElements;
  outsideElements.forEach(node => {
    if (!node) return;
    if (includes$1(TAG_NAME_INVALID, node.tagName)) return;
    if (portalElements.length && portalElements.some(n => n.isEqualNode(node))) return;
    if (portalContainer && node.contains(portalContainer)) return;
    node.setAttribute('aria-hidden', 'true');
  });
};

/** @param {RemoveAriaHiddenProps} Props */
const removeAriaHidden = _ref2 => {
  let _ref2$outsideElements = _ref2.outsideElements,
    outsideElements = _ref2$outsideElements === void 0 ? [] : _ref2$outsideElements;
  outsideElements.forEach(node => {
    if (!node) return;
    if (includes$1(TAG_NAME_INVALID, node.tagName)) return;
    if (node.hasAttribute('aria-hidden')) {
      node.removeAttribute('aria-hidden');
    }
  });
};

/** @param {AddInertProps} Props */
const addInert = _ref3 => {
  let _ref3$outsideElements = _ref3.outsideElements,
    outsideElements = _ref3$outsideElements === void 0 ? [] : _ref3$outsideElements,
    portalContainer = _ref3.portalContainer,
    _ref3$portalElements = _ref3.portalElements,
    portalElements = _ref3$portalElements === void 0 ? [] : _ref3$portalElements,
    portalSelector = _ref3.portalSelector;
  if (!isInertSupported()) {
    warning$1(false, WARNING_INERT_ATTRIBUTE);
  }
  const inertAttr = isInertSupported() ? 'inert' : 'aria-hidden';
  const actualPortalElement = getActualPortalElement(portalContainer, portalSelector);
  const shadowHost = portalContainer instanceof ShadowRoot ? portalContainer.host : null;
  outsideElements.forEach(node => {
    if (!node) return;
    if (includes$1(TAG_NAME_INVALID, node.tagName)) return;
    if (portalElements.length && portalElements.some(n => n.isEqualNode(node))) return;
    if (shadowHost && node === shadowHost) return;
    if (portalContainer && (node.tagName === 'MAIN' || node.classList.contains('odin-wrapper'))) {
      setAttributeRecursively(node, portalContainer, inertAttr, portalSelector);
      return;
    }
    if (actualPortalElement && node.contains(actualPortalElement)) return;
    node.setAttribute(inertAttr, 'true');
  });
};

/** @param {RemoveInertProps} Props */
const removeInert = _ref4 => {
  let _ref4$outsideElements = _ref4.outsideElements,
    outsideElements = _ref4$outsideElements === void 0 ? [] : _ref4$outsideElements,
    portalContainer = _ref4.portalContainer;
  const inertAttr = isInertSupported() ? 'inert' : 'aria-hidden';
  const shadowHost = portalContainer instanceof ShadowRoot ? portalContainer.host : null;
  outsideElements.forEach(node => {
    if (!node) return;
    if (includes$1(TAG_NAME_INVALID, node.tagName)) return;
    if (shadowHost && node === shadowHost && node.shadowRoot) {
      removeAttributeRecursively(node, portalContainer, inertAttr);
      return;
    }
    if (node.hasAttribute(inertAttr)) {
      node.removeAttribute(inertAttr);
    }
    if (portalContainer && (node.tagName === 'MAIN' || node.classList.contains('odin-wrapper'))) {
      removeAttributeRecursively(node, portalContainer, inertAttr);
    }
  });
};

/** @param {UseModalProps} Props */
const useModal = _ref5 => {
  let isOpen = _ref5.isOpen,
    portalSelector = _ref5.portalSelector,
    portalContainer = _ref5.portalContainer;
  const _useState = useState(/** @type {HTMLElement []} */[]),
    _useState2 = _slicedToArray(_useState, 2),
    outsideElements = _useState2[0],
    setOutsideElements = _useState2[1];
  const _useState3 = useState(/** @type {Node []} */[]),
    _useState4 = _slicedToArray(_useState3, 2),
    portalElements = _useState4[0],
    setPortalElements = _useState4[1];
  useEffect(() => {
    const _getElementsBySelecto = getElementsBySelector(portalSelector, portalContainer),
      bodyElements = _getElementsBySelecto.bodyElements,
      _portalElements = _getElementsBySelecto.portalElements;
    setOutsideElements(bodyElements);
    setPortalElements(_portalElements);
  }, [portalSelector, portalContainer]);
  useEffect(() => {
    if (isOpen) {
      addInert({
        outsideElements,
        portalContainer,
        portalElements,
        portalSelector
      });
    }
    return () => {
      removeInert({
        outsideElements,
        portalContainer
      });
    };
  }, [isOpen, portalContainer, outsideElements, portalElements, portalSelector]);
};
export { addAriaHidden, addInert, useModal as default, getElementsBySelector, removeAriaHidden, removeInert };
