// @andes/common/build/esm/hooks/useTransitionMap.development.js
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { useState, useRef, useCallback } from 'react';
const PRE_ENTER = 0;
const ENTERING = 1;
const ENTERED = 2;
const PRE_EXIT = 3;
const EXITING = 4;
const EXITED = 5;
const UNMOUNTED = 6;
const STATUS = /** @type {const} */['preEnter', 'entering', 'entered', 'preExit', 'exiting', 'exited', 'unmounted'];

/** @typedef {PRE_ENTER | ENTERING | ENTERED | PRE_EXIT | EXITING | EXITED |  UNMOUNTED} Status */

/**
 * @typedef {Object} TimeoutObject
 * @property {number} [enter]
 * @property {number} [exit]
 */

/** @typedef {TimeoutObject | number } Timeout */

/**
 * @typedef {Object} State
 * @prop {number} _s
 * @prop {typeof STATUS[number]} status
 * @prop {boolean} isEnter
 * @prop {boolean} isMounted
 * @prop {boolean} isResolved
 */

/**
 * @param {import('../constants/transitionsStatuses').Status} status
 */
const getState = status => ({
  _s: status,
  status: STATUS[status],
  isEnter: status < PRE_EXIT,
  isMounted: status !== UNMOUNTED,
  isResolved: status === ENTERED || status > EXITING
});

/** @param {boolean} [unmounted] */
const startOrEnd = unmounted => unmounted ? UNMOUNTED : EXITED;

/**
 * @param {number} status
 * @param {boolean} [unmountOnExit]
 */
const getEndStatus = (status, unmountOnExit) => {
  switch (status) {
    case ENTERING:
    case PRE_ENTER:
      return ENTERED;
    case EXITING:
    case PRE_EXIT:
      return startOrEnd(unmountOnExit);
  }
};

/** @param {Timeout} [timeout] */
const getTimeout = timeout => typeof timeout === 'object' ? [timeout.enter, timeout.exit] : [timeout, timeout];

/**
 * @param {Function} transitState
 * @param {import('../constants/transitionsStatuses').Status} status
 */
const nextTick = (transitState, status) => setTimeout(() => {
  // Reading document.body.offsetTop can force browser to repaint before transition to the next state
  // eslint-disable-next-line no-unused-expressions
  Number.isNaN(document.body.offsetTop) || transitState(status + 1);
}, 0);
const initialStateMap = new Map();
const initialConfigMap = new Map();

/**
 * @typedef {Object} ConfigObject
 * @property {boolean} [initialEntered]
 */

/**
 * @param {any} key
 * @param {import('../constants/transitionsStatuses').Status} status
 * @param {Function} setStateMap
 * @param {import('react').RefObject<Map<any, any>>} latestStateMap
 * @param {NodeJS.Timeout} [timeoutId]
 * @param {({key, current}: {key: any; current: import('../utils/transitionHelpers').State}) => void} [onChange]
 */
const updateState = (key, status, setStateMap, latestStateMap, timeoutId, onChange) => {
  clearTimeout(timeoutId);
  const state = getState(status);
  const stateMap = new Map(latestStateMap.current);
  stateMap.set(key, state);
  setStateMap(stateMap);
  latestStateMap.current = stateMap;
  if (typeof onChange === 'function') {
    onChange({
      key,
      current: state
    });
  }
};

/** @type {typeof import('@andes/common/build/hooks/useTransitionMap').useTransitionMap} */
const useTransitionMap = function () {
  let _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
    allowMultiple = _ref.allowMultiple,
    _ref$enter = _ref.enter,
    enter = _ref$enter === void 0 ? true : _ref$enter,
    _ref$exit = _ref.exit,
    exit = _ref$exit === void 0 ? true : _ref$exit,
    preEnter = _ref.preEnter,
    preExit = _ref.preExit,
    timeout = _ref.timeout,
    initialEntered = _ref.initialEntered,
    mountOnEnter = _ref.mountOnEnter,
    unmountOnExit = _ref.unmountOnExit,
    onChange = _ref.onStateChange;
  const _useState = useState(initialStateMap),
    _useState2 = _slicedToArray(_useState, 2),
    stateMap = _useState2[0],
    setStateMap = _useState2[1];
  const latestStateMap = useRef(stateMap);
  const configMap = useRef(initialConfigMap);
  const _getTimeout = getTimeout(timeout),
    _getTimeout2 = _slicedToArray(_getTimeout, 2),
    enterTimeout = _getTimeout2[0],
    exitTimeout = _getTimeout2[1];
  const setItem = useCallback(
  /**
   * @param {any} key
   * @param {ConfigObject} [config]
   */
  (key, config) => {
    const _ref2 = config || {},
      _ref2$initialEntered = _ref2.initialEntered,
      _initialEntered = _ref2$initialEntered === void 0 ? initialEntered : _ref2$initialEntered;
    const status = _initialEntered ? ENTERED : startOrEnd(mountOnEnter);
    updateState(key, status, setStateMap, latestStateMap);
    configMap.current.set(key, {});
  }, [initialEntered, mountOnEnter]);
  const deleteItem = useCallback(/** @param {any} key */
  key => {
    const newStateMap = new Map(latestStateMap.current);
    if (newStateMap.delete(key)) {
      setStateMap(newStateMap);
      latestStateMap.current = newStateMap;
      configMap.current.delete(key);
      return true;
    }
    return false;
  }, []);
  const endTransition = useCallback(/** @param {any} key */
  key => {
    const stateObj = latestStateMap.current.get(key);
    if (!stateObj) {
      if (process.env.NODE_ENV !== 'production') {
        console.error(`[UseTransition Hook] invalid key: ${key}`);
      }
      return;
    }
    const _configMap$current$ge = configMap.current.get(key),
      timeoutId = _configMap$current$ge.timeoutId;
    // eslint-disable-next-line no-underscore-dangle
    const status = getEndStatus(stateObj._s, unmountOnExit);
    if (status) {
      updateState(key, status, setStateMap, latestStateMap, timeoutId, onChange);
    }
  }, [onChange, unmountOnExit]);
  const toggle = useCallback(
  /**
   * @param {any} key
   * @param {boolean} [toEnter]
   */
  (key, toEnter) => {
    const stateObj = latestStateMap.current.get(key);
    if (!stateObj) {
      if (process.env.NODE_ENV !== 'production') {
        console.error(`[UseTransition Hook] invalid key: ${key}`);
      }
      return;
    }
    const config = configMap.current.get(key);
    /** @param {import('../constants/transitionsStatuses').Status} status */
    const transitState = status => {
      updateState(key, status, setStateMap, latestStateMap, config.timeoutId, onChange);
      switch (status) {
        case ENTERING:
          if (enterTimeout && enterTimeout >= 0) {
            config.timeoutId = setTimeout(() => endTransition(key), enterTimeout);
          }
          break;
        case EXITING:
          if (exitTimeout && exitTimeout >= 0) {
            config.timeoutId = setTimeout(() => endTransition(key), exitTimeout);
          }
          break;
        case PRE_ENTER:
        case PRE_EXIT:
          config.timeoutId = nextTick(transitState, status);
          break;
      }
    };
    const enterStage = stateObj.isEnter;
    // eslint-disable-next-line no-param-reassign
    if (typeof toEnter !== 'boolean') toEnter = !enterStage;
    if (toEnter) {
      if (!enterStage) {
        // eslint-disable-next-line no-nested-ternary
        transitState(enter ? preEnter ? PRE_ENTER : ENTERING : ENTERED);
        if (!allowMultiple) {
          latestStateMap.current.forEach((_, _key) => _key !== key && toggle(_key, false));
        }
      }
    } else if (enterStage) {
      // eslint-disable-next-line no-nested-ternary
      transitState(exit ? preExit ? PRE_EXIT : EXITING : startOrEnd(unmountOnExit));
    }
  }, [onChange, endTransition, allowMultiple, enter, exit, preEnter, preExit, enterTimeout, exitTimeout, unmountOnExit]);
  const toggleAll = useCallback(/** @param {boolean} [toEnter] */
  toEnter => {
    if (!allowMultiple && toEnter !== false) return;
    Array.from(latestStateMap.current.keys()).forEach(key => toggle(key, toEnter));
  }, [allowMultiple, toggle]);
  return {
    stateMap,
    toggle,
    toggleAll,
    endTransition,
    setItem,
    deleteItem
  };
};
export { useTransitionMap };
