// @andes/common/build/esm/hooks/useTransitionMap.production.js
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { useState, useRef, useCallback } from 'react';
const PRE_ENTER = 0;
const ENTERING = 1;
const ENTERED = 2;
const PRE_EXIT = 3;
const EXITING = 4;
const EXITED = 5;
const UNMOUNTED = 6;
const STATUS = ['preEnter', 'entering', 'entered', 'preExit', 'exiting', 'exited', 'unmounted'];
const getState = status => ({
  _s: status,
  status: STATUS[status],
  isEnter: status < PRE_EXIT,
  isMounted: status !== UNMOUNTED,
  isResolved: status === ENTERED || status > EXITING
});
const startOrEnd = unmounted => unmounted ? UNMOUNTED : EXITED;
const getEndStatus = (status, unmountOnExit) => {
  switch (status) {
    case ENTERING:
    case PRE_ENTER:
      return ENTERED;
    case EXITING:
    case PRE_EXIT:
      return startOrEnd(unmountOnExit);
  }
};
const getTimeout = timeout => typeof timeout === 'object' ? [timeout.enter, timeout.exit] : [timeout, timeout];
const nextTick = (transitState, status) => setTimeout(() => {
  Number.isNaN(document.body.offsetTop) || transitState(status + 1);
}, 0);
const initialStateMap = new Map();
const initialConfigMap = new Map();
const updateState = (key, status, setStateMap, latestStateMap, timeoutId, onChange) => {
  clearTimeout(timeoutId);
  const state = getState(status);
  const stateMap = new Map(latestStateMap.current);
  stateMap.set(key, state);
  setStateMap(stateMap);
  latestStateMap.current = stateMap;
  if (typeof onChange === 'function') {
    onChange({
      key,
      current: state
    });
  }
};
const useTransitionMap = function () {
  let _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
    allowMultiple = _ref.allowMultiple,
    _ref$enter = _ref.enter,
    enter = _ref$enter === void 0 ? true : _ref$enter,
    _ref$exit = _ref.exit,
    exit = _ref$exit === void 0 ? true : _ref$exit,
    preEnter = _ref.preEnter,
    preExit = _ref.preExit,
    timeout = _ref.timeout,
    initialEntered = _ref.initialEntered,
    mountOnEnter = _ref.mountOnEnter,
    unmountOnExit = _ref.unmountOnExit,
    onChange = _ref.onStateChange;
  const _useState = useState(initialStateMap),
    _useState2 = _slicedToArray(_useState, 2),
    stateMap = _useState2[0],
    setStateMap = _useState2[1];
  const latestStateMap = useRef(stateMap);
  const configMap = useRef(initialConfigMap);
  const _getTimeout = getTimeout(timeout),
    _getTimeout2 = _slicedToArray(_getTimeout, 2),
    enterTimeout = _getTimeout2[0],
    exitTimeout = _getTimeout2[1];
  const setItem = useCallback((key, config) => {
    const _ref2 = config || {},
      _ref2$initialEntered = _ref2.initialEntered,
      _initialEntered = _ref2$initialEntered === void 0 ? initialEntered : _ref2$initialEntered;
    const status = _initialEntered ? ENTERED : startOrEnd(mountOnEnter);
    updateState(key, status, setStateMap, latestStateMap);
    configMap.current.set(key, {});
  }, [initialEntered, mountOnEnter]);
  const deleteItem = useCallback(key => {
    const newStateMap = new Map(latestStateMap.current);
    if (newStateMap.delete(key)) {
      setStateMap(newStateMap);
      latestStateMap.current = newStateMap;
      configMap.current.delete(key);
      return true;
    }
    return false;
  }, []);
  const endTransition = useCallback(key => {
    const stateObj = latestStateMap.current.get(key);
    if (!stateObj) {
      if (process.env.NODE_ENV !== 'production') {
        console.error(`[UseTransition Hook] invalid key: ${key}`);
      }
      return;
    }
    const _configMap$current$ge = configMap.current.get(key),
      timeoutId = _configMap$current$ge.timeoutId;
    const status = getEndStatus(stateObj._s, unmountOnExit);
    if (status) {
      updateState(key, status, setStateMap, latestStateMap, timeoutId, onChange);
    }
  }, [onChange, unmountOnExit]);
  const toggle = useCallback((key, toEnter) => {
    const stateObj = latestStateMap.current.get(key);
    if (!stateObj) {
      if (process.env.NODE_ENV !== 'production') {
        console.error(`[UseTransition Hook] invalid key: ${key}`);
      }
      return;
    }
    const config = configMap.current.get(key);
    const transitState = status => {
      updateState(key, status, setStateMap, latestStateMap, config.timeoutId, onChange);
      switch (status) {
        case ENTERING:
          if (enterTimeout && enterTimeout >= 0) {
            config.timeoutId = setTimeout(() => endTransition(key), enterTimeout);
          }
          break;
        case EXITING:
          if (exitTimeout && exitTimeout >= 0) {
            config.timeoutId = setTimeout(() => endTransition(key), exitTimeout);
          }
          break;
        case PRE_ENTER:
        case PRE_EXIT:
          config.timeoutId = nextTick(transitState, status);
          break;
      }
    };
    const enterStage = stateObj.isEnter;
    if (typeof toEnter !== 'boolean') toEnter = !enterStage;
    if (toEnter) {
      if (!enterStage) {
        transitState(enter ? preEnter ? PRE_ENTER : ENTERING : ENTERED);
        if (!allowMultiple) {
          latestStateMap.current.forEach((_, _key) => _key !== key && toggle(_key, false));
        }
      }
    } else if (enterStage) {
      transitState(exit ? preExit ? PRE_EXIT : EXITING : startOrEnd(unmountOnExit));
    }
  }, [onChange, endTransition, allowMultiple, enter, exit, preEnter, preExit, enterTimeout, exitTimeout, unmountOnExit]);
  const toggleAll = useCallback(toEnter => {
    if (!allowMultiple && toEnter !== false) return;
    Array.from(latestStateMap.current.keys()).forEach(key => toggle(key, toEnter));
  }, [allowMultiple, toggle]);
  return {
    stateMap,
    toggle,
    toggleAll,
    endTransition,
    setItem,
    deleteItem
  };
};
export { useTransitionMap };
