// @meli/input-validation-storage/dist/callerTracker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getModuleScope = getModuleScope;
const path_1 = require("path");
const statsd_1 = require("@meli/statsd");
const constants_1 = require("./constants");
/**
 * Checks if a module path belongs to any of the SDK packages.
 * Supports both legacy (node-sdk-input-validation) and new (@meli/input-validation) package names.
 * @param {string} modulePath - The module path to check
 * @returns {boolean} True if the path contains any of the SDK package names
 */
function isSDKPackagePath(modulePath) {
    return constants_1.SDK_PACKAGE_NAMES.some((pkgName) => modulePath.includes(pkgName));
}
/**
 * Returns true for V8 native built-in frames that carry no real source location
 * (e.g. Array.from, .some, .filter report as "<anonymous>"; older runtimes as "native").
 */
function isNativeFrame(modulePath) {
    return modulePath.startsWith('<') || modulePath === 'native';
}
const statsd = new statsd_1.MeliStatsD();
/**
 * Truncates an absolute module path so that it starts at the project root
 * directory name. If the project root cannot be determined or is not found
 * in the path, the original path is returned. Resolves `process.cwd()` lazily
 * on each call so tests that mock cwd (and the rare app that uses
 * `process.chdir()`) keep working.
 * @param {string} modulePath - The absolute module path to truncate
 * @returns {string} The truncated module path
 */
function truncateToProjectRoot(modulePath) {
    let cwd;
    try {
        cwd = process.cwd();
    }
    catch {
        return modulePath;
    }
    const projectRootName = (0, path_1.basename)(cwd);
    if (!projectRootName) {
        return modulePath;
    }
    const separatorIndex = modulePath.indexOf(`/${projectRootName}/`);
    if (separatorIndex !== -1) {
        return modulePath.substring(separatorIndex + 1);
    }
    if (modulePath.startsWith(`${projectRootName}/`)) {
        return modulePath;
    }
    // Monorepo fallback: module is outside cwd (e.g., root node_modules
    // when cwd is a nested workspace). Find the common ancestor directory.
    const cwdParts = cwd.split('/');
    const modParts = modulePath.split('/');
    let commonDepth = 0;
    while (commonDepth < cwdParts.length &&
        commonDepth < modParts.length &&
        cwdParts[commonDepth] === modParts[commonDepth]) {
        commonDepth += 1;
    }
    if (commonDepth > 1) {
        return modParts.slice(commonDepth - 1).join('/');
    }
    return modulePath;
}
/**
 * Walks the stack from `startIndex` skipping native built-in frames (always)
 * and, when `skipSdkFrames` is true, SDK package frames as well. Returns the
 * first frame that represents a real application caller, or UNKNOWN_PATH when
 * the stack is exhausted without finding a non-native frame.
 *
 * Native frames (e.g. "Array.from (<anonymous>)", "Array.prototype.forEach (native)")
 * are always skipped regardless of `skipSdkFrames` because they never carry a
 * real source location.
 */
function resolveCallerFrame(stackLines, startIndex, skipSdkFrames) {
    let index = startIndex;
    let match = constants_1.STACK_FRAME_REGEX.exec(stackLines[index]);
    let modulePath = match ? match[1] : constants_1.UNKNOWN_PATH;
    let isSdkMiddlewareCaller = false;
    while (modulePath !== constants_1.UNKNOWN_PATH) {
        if (isNativeFrame(modulePath)) {
            // native built-in frame — skip, fall through to advance
        }
        else if (skipSdkFrames && isSDKPackagePath(modulePath)) {
            if (modulePath.includes(constants_1.SDK_MIDDLEWARE_PATH)) {
                isSdkMiddlewareCaller = true;
                break;
            }
            // SDK package frame — skip, fall through to advance
        }
        else {
            break; // real application caller found
        }
        if (index + 1 >= stackLines.length) {
            // Stack exhausted while skipping frames — no real caller found
            modulePath = constants_1.UNKNOWN_PATH;
            break;
        }
        index += 1;
        match = constants_1.STACK_FRAME_REGEX.exec(stackLines[index]);
        modulePath = match ? match[1] : constants_1.UNKNOWN_PATH;
    }
    return { modulePath, isSdkMiddlewareCaller };
}
/**
 * Gets the module scope information from the current call stack
 * @param {number} [targetLine] - The target line in the stack to analyze
 * @returns {ModuleScopeInfo} Module scope information including type, scope and path
 */
function getModuleScope(targetLine) {
    const defaultReturn = {
        moduleType: constants_1.MODULE_TYPE.APP_CODE,
        scope: constants_1.GLOBAL_SCOPE,
        modulePath: constants_1.UNKNOWN_PATH,
        line: undefined,
    };
    try {
        const target = {};
        Error.captureStackTrace(target, getModuleScope);
        const stackLines = target.stack?.split('\n') || [];
        const startIndex = targetLine || constants_1.DEFAULT_TARGET_LINE;
        if (startIndex > stackLines.length) {
            return defaultReturn;
        }
        const skipSdkFrames = targetLine === undefined;
        const { modulePath: rawModulePath, isSdkMiddlewareCaller } = resolveCallerFrame(stackLines, startIndex, skipSdkFrames);
        const lineMatch = /:(\d+):\d+$/.exec(rawModulePath);
        const line = lineMatch ? Number.parseInt(lineMatch[1], 10) : undefined;
        let modulePath = rawModulePath;
        if (modulePath === constants_1.UNKNOWN_PATH) {
            statsd.increment(constants_1.METRIC_MODULE_SCOPE_ERROR, {
                error: 'UNKNOWN_CALLER_FRAME',
            });
            return defaultReturn;
        }
        modulePath = modulePath.replace(/:\d+:\d+$/, '');
        modulePath = truncateToProjectRoot(modulePath);
        const moduleType = modulePath.includes('node_modules')
            ? constants_1.MODULE_TYPE.EXTERNAL_PACKAGE
            : constants_1.MODULE_TYPE.APP_CODE;
        if (isSdkMiddlewareCaller) {
            return {
                moduleType: constants_1.MODULE_TYPE.APP_CODE,
                stackLines,
                scope: constants_1.GLOBAL_SCOPE,
                modulePath,
                line,
            };
        }
        if (moduleType === constants_1.MODULE_TYPE.APP_CODE) {
            return {
                moduleType,
                stackLines,
                scope: constants_1.GLOBAL_SCOPE,
                modulePath,
                line,
            };
        }
        const moduleMatches = Array.from(modulePath.matchAll(constants_1.NODE_MODULES_REGEX));
        if (moduleMatches.length === 0) {
            statsd.increment(constants_1.METRIC_MODULE_SCOPE_ERROR, {
                error: 'UNKNOWN_MODULE',
            });
            return {
                moduleType: constants_1.MODULE_TYPE.EXTERNAL_PACKAGE,
                scope: constants_1.UNKNOWN_MODULE,
                modulePath,
                line,
            };
        }
        const [, scope] = moduleMatches[moduleMatches.length - 1];
        return {
            moduleType,
            stackLines,
            scope,
            modulePath,
            line,
        };
    }
    catch (error) {
        statsd.increment(constants_1.METRIC_MODULE_SCOPE_ERROR, {
            error: 'UNKNOWN_ERROR',
        });
        return defaultReturn;
    }
}
