// @meli/input-validation-storage/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SDK_MIDDLEWARE_PATH = exports.SDK_PACKAGE_NAMES = exports.STACK_FRAME_REGEX = exports.NODE_MODULES_REGEX = exports.DEFAULT_TARGET_LINE = exports.METRIC_MODULE_SCOPE_ERROR = exports.METRIC_ASYNC_LOCAL_STORAGE_ERROR = exports.UNKNOWN_MODULE = exports.UNKNOWN_PATH = exports.GLOBAL_SCOPE = exports.MODULE_TYPE = void 0;
exports.MODULE_TYPE = {
    APP_CODE: 'APP_CODE',
    EXTERNAL_PACKAGE: 'EXTERNAL_PACKAGE',
};
exports.GLOBAL_SCOPE = 'GLOBAL';
exports.UNKNOWN_PATH = 'UNKNOWN';
exports.UNKNOWN_MODULE = 'UNKNOWN_MODULE';
exports.METRIC_ASYNC_LOCAL_STORAGE_ERROR = 'node.websecurity.iv.storage.async_local_storage_error';
exports.METRIC_MODULE_SCOPE_ERROR = 'node.websecurity.iv.storage.module_scope_error';
// stack trace line to start the analysis of the module scope
exports.DEFAULT_TARGET_LINE = 3;
// regex to match node_modules paths
exports.NODE_MODULES_REGEX = /node_modules\/((@[^/]+\/[^/]+)|([^/]+))/g;
// regex to match stack frame
// eslint-disable-next-line security/detect-unsafe-regex
exports.STACK_FRAME_REGEX = /at (?:(?:.+? \()?([^)]+)(?:\))?)/;
// names of the SDK packages (used to identify the SDK frames in the stack trace)
// supports both legacy and new package names for backwards compatibility
exports.SDK_PACKAGE_NAMES = [
    'node-sdk-input-validation', // legacy package name
    'input-validation', // new package name (@meli/input-validation)
];
exports.SDK_MIDDLEWARE_PATH = 'middleware/requestValidation';
