// @meli/input-validation-storage/dist/validationState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationState = void 0;
const async_hooks_1 = require("async_hooks");
const statsd_1 = require("@meli/statsd");
const constants_1 = require("./constants");
const callerTracker_1 = require("./callerTracker");
const statsd = new statsd_1.MeliStatsD();
const asyncLocalStorage = new async_hooks_1.AsyncLocalStorage();
/**
 * Gets the current validation store from AsyncLocalStorage
 * @returns {Map<unknown, ValidationStateStore> | null} The current validation store or null if not initialized
 */
function current() {
    const store = asyncLocalStorage.getStore();
    if (!store) {
        statsd.increment(constants_1.METRIC_ASYNC_LOCAL_STORAGE_ERROR, {
            error: 'localstorage_not_initialized',
        });
        return null;
    }
    return store;
}
/**
 * Runs a callback within a new validation context using AsyncLocalStorage
 * @param {Function} callback - The callback to run within the validation context
 * @returns {T} The result of the callback
 * @template T - The return type of the callback
 */
function run(callback) {
    return asyncLocalStorage.run(new Map(), callback);
}
/**
 * Checks if an input is currently being validated
 * @param {unknown} input - The input to check
 * @returns {boolean} Whether the input is being validated
 */
function isValidating(input) {
    const store = current();
    const validationState = store?.get(input);
    return validationState ? validationState.state : false;
}
/**
 * Checks if a property has been validated for a given input and scope
 * @param {unknown} input - The input to check
 * @param {string} propertyName - The name of the property to check
 * @param {string} [scope=GLOBAL_SCOPE] - The scope to check in
 * @returns {boolean} Whether the property has been validated
 */
function isPropertyValidated(input, propertyName, scope = constants_1.GLOBAL_SCOPE) {
    if (!input || !propertyName) {
        return false;
    }
    const store = current();
    if (!store) {
        return false;
    }
    const validationState = store.get(input);
    if (!validationState) {
        return false;
    }
    if (scope === constants_1.GLOBAL_SCOPE) {
        return (validationState.fullyValidated ||
            validationState.validatedProps.has(propertyName));
    }
    const scopedState = validationState.scopes.get(scope);
    return (scopedState?.fullyValidated ||
        scopedState?.validatedProps.has(propertyName) ||
        false);
}
/**
 * Validates if a property has been validated in the current module scope
 * @param {unknown} input - The input object to validate
 * @param {string} propertyName - The name of the property to validate
 * @param {ValidationOptions} [options={}] - Validation options including global scope preference
 * @returns {boolean} Whether the property is validated in the current scope
 */
function validateInScope(input, propertyName, options = {}) {
    if (!input || !propertyName) {
        return false;
    }
    // Fast path: check global scope first (avoids expensive getModuleScope stack trace)
    if (isPropertyValidated(input, propertyName, constants_1.GLOBAL_SCOPE)) {
        return true;
    }
    const { scope } = options.global ? { scope: constants_1.GLOBAL_SCOPE } : (0, callerTracker_1.getModuleScope)();
    if (scope === constants_1.GLOBAL_SCOPE) {
        // Already verified via fast path above — property is not in GLOBAL_SCOPE
        return false;
    }
    return isPropertyValidated(input, propertyName, scope);
}
/**
 * Gets the current scope for an input
 * @param {unknown} input - The input to get the scope for
 * @param {ValidationOptions} [options={}] - Options
 * @returns {string} The current scope
 */
function getCurrentScope(input, options = {}) {
    const { scope } = options.global ? { scope: constants_1.GLOBAL_SCOPE } : (0, callerTracker_1.getModuleScope)();
    return scope;
}
/**
 * Marks an input as fully validated, meaning all property accesses are considered valid.
 * Used for non-object types (arrays, maps, sets) where individual property tracking is impractical.
 * @param {unknown} input - The input to mark as fully validated
 * @param {string} [scope=GLOBAL_SCOPE] - The scope to mark in
 */
function markFullyValidated(input, scope = constants_1.GLOBAL_SCOPE) {
    const store = current();
    if (!store || !input) {
        return;
    }
    const validationState = store.get(input) ?? {
        state: false,
        fullyValidated: false,
        validatedProps: new Set(),
        scopes: new Map(),
    };
    if (scope === constants_1.GLOBAL_SCOPE) {
        validationState.fullyValidated = true;
    }
    else {
        const scopedState = validationState.scopes.get(scope) ?? {
            validatedProps: new Set(),
        };
        scopedState.fullyValidated = true;
        if (!validationState.scopes.has(scope)) {
            validationState.scopes.set(scope, scopedState);
        }
    }
    store.set(input, validationState);
}
/**
 * ValidationState - Main export object providing validation state management
 *
 * This object provides methods for managing validation states using AsyncLocalStorage.
 * All methods are thread-safe and work within the current validation context.
 */
exports.ValidationState = {
    getModuleScope: callerTracker_1.getModuleScope,
    current,
    run,
    isValidating,
    isPropertyValidated,
    validateInScope,
    getCurrentScope,
    markFullyValidated,
};
