// @andes/context/build/cjs/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var React = require('react');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var warning = require('@andes/common/build/utils/warning');
var context = require('@nordic/context');
var internal = require('@nordic/context/internal');
var i18n = require('@nordic/i18n');
var startsWith = require('@andes/common/build/utils/startsWith');
var PropTypes = require('prop-types');
var isMobile = require('@andes/common/build/utils/isMobile');
var getNativeApp = require('@andes/common/build/utils/getNativeApp');
var getOsName = require('@andes/common/build/utils/getOsName');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var startsWith__default = /*#__PURE__*/_interopDefaultLegacy(startsWith);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var isMobile__default = /*#__PURE__*/_interopDefaultLegacy(isMobile);
var getNativeApp__default = /*#__PURE__*/_interopDefaultLegacy(getNativeApp);
var getOsName__default = /*#__PURE__*/_interopDefaultLegacy(getOsName);

/** @param {string} [value] */
const generateId = value => {
  if (process.env.NODE_ENV !== 'test') return value;
  return typeof value === 'undefined' ? 'test-id' : value;
};
var generateId$1 = generateId;

/** @type {import('../..').UseId} */
const useIdUtil = id => {
  const reactId = React.useId();
  return generateId$1(id) || reactId;
};
var useIdUtil$1 = useIdUtil;
const useNordicContext = () => {
  try {
    const csrfToken = context.getCsrfToken();
    const deviceInfo = internal.getDeviceInfo();
    const i18n$1 = i18n.getI18n();
    const locale = i18n$1 === null || i18n$1 === void 0 ? void 0 : i18n$1.locale;
    const device = {
      mobile: (deviceInfo === null || deviceInfo === void 0 ? void 0 : deviceInfo.type) === 'mobile',
      nativeApp: {
        os: deviceInfo === null || deviceInfo === void 0 ? void 0 : deviceInfo.context
      },
      osName: deviceInfo === null || deviceInfo === void 0 ? void 0 : deviceInfo.context,
      type: deviceInfo === null || deviceInfo === void 0 ? void 0 : deviceInfo.type,
      webView: deviceInfo === null || deviceInfo === void 0 ? void 0 : deviceInfo.webView
    };
    return {
      csrfToken,
      device,
      i18n: i18n$1,
      locale
    };
  } catch (_error) {
    return undefined;
  }
};
var useNordicProvider = useNordicContext;
const EN_US = 'en';
const ES_AR = 'es-AR';
const PT_BR = 'pt-BR';
const ZH_CN = 'zh-CN';
const AndesContext = /*#__PURE__*/React__default["default"].createContext({});
var AndesContext$1 = AndesContext;

/** @typedef {import('../..').ContextValue} ContextValue */
/** @typedef {import('../..').CreateI18nPgettext} CreateI18nPgettext */
/** @typedef {import('../..').CreateI18nNPgettext} CreateI18nNPgettext */
/** @typedef {import('../..').PrimitiveTypes} PrimitiveTypes */
/** @typedef {import('../..').I18nValues} I18nValues */

const REPLACEMENT_REGEX = /\{(\d+)\}/g;

/**
 * @param {string} text
 * @param {Array<PrimitiveTypes | I18nValues>} replacements */
const replaceArguments = (text, replacements) => text.replace(REPLACEMENT_REGEX, (_, index) => `${replacements[+index]}`);

/**
 * @param {import('@meli/i18n').I18n} i18n
 * @return {CreateI18nPgettext}
 */
const createI18nPgettext = i18n => function (context, text) {
  // @ts-expect-error: disabled the rule because 'translations' is a custom property.
  const translationFound = i18n.andes.translations[`${context}\u0004${text}`];
  const usedTranslation = translationFound ? translationFound[1] : text;
  for (var _len = arguments.length, replacements = new Array(_len > 2 ? _len - 2 : 0), _key = 2; _key < _len; _key++) {
    replacements[_key - 2] = arguments[_key];
  }
  return replaceArguments(usedTranslation, replacements.flat());
};

/**
 * @param {import('@meli/i18n').I18n} i18n
 * @return {CreateI18nNPgettext}
 */
const createI18nNPgettext = i18n => function (context, singular, plural, quantity) {
  // @ts-expect-error: disabled the rule because 'translations' is a custom property.
  const translationFound = i18n.andes.translations[`${context}\u0004${singular}`];
  let text;
  if (translationFound) text = quantity === 1 ? translationFound[1] : translationFound[2];
  const usedTranslation = text !== null && text !== void 0 ? text : quantity === 1 ? singular : plural;
  for (var _len2 = arguments.length, replacements = new Array(_len2 > 4 ? _len2 - 4 : 0), _key2 = 4; _key2 < _len2; _key2++) {
    replacements[_key2 - 4] = arguments[_key2];
  }
  return replaceArguments(usedTranslation, replacements.flat());
};

/** @param {import('@meli/i18n').Locale} [locale] */
const getBestLocale = locale => {
  if (!locale) return ES_AR;
  if (startsWith__default["default"](locale, 'en')) return EN_US;
  if (startsWith__default["default"](locale, 'zh')) return ZH_CN;
  return locale === PT_BR ? PT_BR : ES_AR;
};

/** @type {import('../..').InjectTranslations} */
const injectTranslations = (i18n, contextLocale, moduleName, availableTranslations) => {
  const bestLocale = getBestLocale(contextLocale);
  const componentTranslations = availableTranslations[bestLocale];

  // @ts-expect-error: disabled the rule because 'andes' is a custom property.
  if (!i18n.andes) {
    // @ts-expect-error: disabled the rule because 'andes' is a custom property.
    i18n.andes = {};
    i18n.pgettext = function (context, text) {
      for (var _len3 = arguments.length, replacements = new Array(_len3 > 2 ? _len3 - 2 : 0), _key3 = 2; _key3 < _len3; _key3++) {
        replacements[_key3 - 2] = arguments[_key3];
      }
      return createI18nPgettext(i18n)(context, text, ...replacements);
    };
    i18n.npgettext = function (context, singular, plural, quantity) {
      for (var _len4 = arguments.length, replacements = new Array(_len4 > 4 ? _len4 - 4 : 0), _key4 = 4; _key4 < _len4; _key4++) {
        replacements[_key4 - 4] = arguments[_key4];
      }
      return createI18nNPgettext(i18n)(context, singular, plural, quantity, ...replacements);
    };
  }

  // @ts-expect-error: disabled the rule because 'andes' is a custom property.
  if (!i18n.andes[String(moduleName)]) {
    // @ts-expect-error: translations is not an exposed type, it is added in a custom way.
    i18n.andes.translations = _objectSpread__default["default"](_objectSpread__default["default"]({}, i18n.andes.translations), componentTranslations);
    // @ts-expect-error: disabled the rule because 'andes' is a custom property.
    i18n.andes[String(moduleName)] = true;
  }
};

/** @type {import('../..').GetPatchedI18n} */
const getPatchedI18n = (contextValue, moduleName, availableTranslations) => {
  const i18n = (/** @type {ContextValue} */contextValue).i18n,
    _contextValue$locale = (/** @type {ContextValue} */contextValue).locale,
    locale = _contextValue$locale === void 0 ? 'es-AR' : _contextValue$locale;
  injectTranslations(i18n, locale, moduleName, availableTranslations);
  return i18n;
};
const DEFAULT_LOCALE = ES_AR;

/** @typedef {import('../..').ContextValue} ContextValue */
/** @typedef {import('@meli/i18n').I18n} I18n */

/** @type {import('../..').UseCallAndesContext} */
const useCallAndesContext = (moduleName, andesTranslations) => {
  const contextValue = React.useContext(AndesContext$1);
  if (Object.keys(contextValue).length === 0) return;
  const patchedI18n = andesTranslations ? getPatchedI18n(/** @type {ContextValue} */contextValue, moduleName, andesTranslations) : {
    translations: {}
  };
  return _objectSpread__default["default"](_objectSpread__default["default"]({}, contextValue), {}, {
    i18n: (/** @type {I18n} */patchedI18n)
  });
};

/** @type {import('../..').UseAndesContext} */
const useAndesContext = function () {
  let moduleName = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '@andes';
  let andesTranslations = arguments.length > 1 ? arguments[1] : undefined;
  const nordicContext = useNordicProvider();

  // In Nordic environments, the NordicProvider context is imported.
  if (nordicContext) {
    return nordicContext;
  }

  // AndesProvider is imported in contexts outside Nordic.
  const andesContext = useCallAndesContext(moduleName, andesTranslations);
  if (andesContext) {
    return andesContext;
  }
  warning__default["default"](false, `${moduleName} needs a valid configuration of <AndesProvider> on your application.
      Check the documentation about @andes/context and <AndesProvider> on:
      https://github.com/melisource/fury_frontend-andes-ui/blob/master/packages/context/README.md`);

  // Create a default i18n object with essential functions to prevent runtime errors
  const defaultI18n = {
    translations: {},
    pgettext: (/** @type {string} */context, /** @type {string} */text) => text,
    npgettext: (/** @type {string} */context, /** @type {string} */singular, /** @type {string} */plural, /** @type {number} */quantity) => quantity === 1 ? singular : plural
  };
  return {
    i18n: (/** @type {I18n} */ /** @type {unknown} */defaultI18n),
    locale: DEFAULT_LOCALE,
    device: {
      mobile: undefined,
      type: undefined,
      webView: undefined
    },
    csrfToken: ''
  };
};
var useAndesContext$1 = useAndesContext;

/** @param {import('../..').Device} [device] */
const getDevice = device => {
  if (typeof device !== 'undefined') return device;
  const basicDeviceData = {
    mobile: isMobile__default["default"](),
    webView: typeof MobileWebKit !== 'undefined',
    nativeApp: getNativeApp__default["default"](),
    osName: getOsName__default["default"]()
  };
  return basicDeviceData;
};
var getDevice$1 = getDevice;

/** @type {typeof import('..').AndesProvider} */
const AndesProvider = _ref => {
  let children = _ref.children,
    locale = _ref.locale,
    device = _ref.device,
    csrfToken = _ref.csrfToken;
  const localI18n = {
    translations: {},
    locale
  };
  const value = {
    locale,
    i18n: localI18n,
    device: getDevice$1(device),
    csrfToken
  };
  return /*#__PURE__*/React__default["default"].createElement(AndesContext$1.Provider, {
    value: value
  }, React.Children.only(children));
};
AndesProvider.propTypes = {
  children: PropTypes__default["default"].element.isRequired,
  // https://furydocs.io/m10e-service/latest/guide/#/wiki/terms?id=locales
  locale: PropTypes__default["default"].oneOf(/** @type {const} */['es-AR', 'pt-BR', 'es-PE', 'es-BO', 'es-CL', 'es-CO', 'es-CR', 'es-DO', 'es-EC', 'es-GT', 'es-HN', 'es-NI', 'es-PA', 'es-PR', 'es-PY', 'es-SV', 'es-UY', 'es-VE', 'es-MX', 'en', 'en-US', 'zh-CN']).isRequired,
  // https://github.com/melisource/fury_frontend-device-detection#reqdevice
  device: PropTypes__default["default"].shape({
    type: PropTypes__default["default"].string,
    webView: PropTypes__default["default"].bool,
    mobile: PropTypes__default["default"].bool,
    osName: PropTypes__default["default"].string,
    nativeApp: PropTypes__default["default"].shape({
      name: PropTypes__default["default"].string,
      os: PropTypes__default["default"].string,
      version: PropTypes__default["default"].string
    })
  }),
  csrfToken: PropTypes__default["default"].string
};
AndesProvider.displayName = 'AndesProvider';
var AndesProvider$1 = AndesProvider;
exports.AndesContext = AndesContext$1;
exports.AndesProvider = AndesProvider$1;
exports.useAndesContext = useAndesContext$1;
exports.useId = useIdUtil$1;
