// @andes/list/build/generators/itemNodeGenerator.js
'use strict';

var react = require('react');
var warning = require('@andes/common/build/utils/warning');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
const hasSlot = (slot, _ref, _ref2) => {
  let title = _ref.title,
    description = _ref.description,
    rightContent = _ref.rightContent;
  let _ref2$legacyBehavior = _ref2.legacyBehavior,
    legacyBehavior = _ref2$legacyBehavior === void 0 ? false : _ref2$legacyBehavior;
  const hasRightContent = typeof rightContent === 'object' ? (rightContent === null || rightContent === void 0 ? void 0 : rightContent.content) || (rightContent === null || rightContent === void 0 ? void 0 : rightContent.description) : rightContent;
  const hasDefaultContent = title || description || hasRightContent;
  const isSlotted = legacyBehavior ? slot && !hasDefaultContent : slot;
  return !!isSlotted;
};
const DEFAULT_POSITION_TOP = 'top';
const isRightContentType = value => (value === null || value === void 0 ? void 0 : value.content) !== undefined;
const computeRightContent = _ref => {
  let rightContentProp = _ref.rightContent,
    rightContentPositionProp = _ref.rightContentPosition;
  if (!rightContentProp) return {
    position: DEFAULT_POSITION_TOP
  };
  const computedRightContentPosition = {
    position: rightContentPositionProp || DEFAULT_POSITION_TOP
  };
  if (isRightContentType(rightContentProp)) {
    var _rightContentProp$des;
    return _objectSpread__default["default"](_objectSpread__default["default"]({}, computedRightContentPosition), {}, {
      content: rightContentProp === null || rightContentProp === void 0 ? void 0 : rightContentProp.content,
      description: (_rightContentProp$des = rightContentProp === null || rightContentProp === void 0 ? void 0 : rightContentProp.description) !== null && _rightContentProp$des !== void 0 ? _rightContentProp$des : undefined
    });
  }
  return _objectSpread__default["default"](_objectSpread__default["default"]({}, computedRightContentPosition), {}, {
    content: rightContentProp || null,
    position: rightContentPositionProp || DEFAULT_POSITION_TOP
  });
};
var computeRightContent$1 = computeRightContent;
const hasChildNodes = _ref => {
  let title = _ref.title,
    children = _ref.children;
  return title && react.Children.count(children) > 0;
};
const itemNodeGenerator = function* getCollectionNode(props) {
  var _computeRightContent;
  const children = props.children,
    ariaLabel = props['aria-label'],
    title = props.title,
    description = props.description,
    rightContent = props.rightContent;
  const rendered = title || children;
  const textValue = (typeof rendered === 'string' ? rendered : '') || ariaLabel || '';
  const isSlotted = hasSlot(children, {
    title,
    description,
    rightContent: (_computeRightContent = computeRightContent$1({
      rightContent
    })) === null || _computeRightContent === void 0 ? void 0 : _computeRightContent.content
  }, {
    legacyBehavior: true
  });
  warning__default["default"](isSlotted || textValue, '<ListItem> needs some plain text contents for accessibility reasons. Please add a `title` or `aria-label` prop.');
  yield {
    type: 'item',
    key: props.value,
    props,
    rendered,
    textValue,
    'aria-label': ariaLabel || title,
    hasChildNodes,
    *childNodes() {
      if (title) {
        const items = [];
        react.Children.forEach(children, child => {
          items.push({
            type: 'item',
            element: child
          });
        });
        yield* items;
      }
    }
  };
};
module.exports = itemNodeGenerator;
