// @meli/statsd/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MeliStatsD = void 0;
const hot_shots_1 = require("hot-shots");
const frontend_env_1 = __importDefault(require("frontend-env"));
const error_handler_1 = require("./error-handler");
const mainClient = new hot_shots_1.StatsD({
    host: 'datadog',
    maxBufferSize: frontend_env_1.default.FURY && frontend_env_1.default.PRODUCTION ? 8192 : 0,
    mock: !frontend_env_1.default.FURY || frontend_env_1.default.IS_CI,
    errorHandler: error_handler_1.errorHandler,
});
const privateData = new WeakMap();
// eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging, @typescript-eslint/no-extraneous-class
class MeliStatsD {
    constructor(options = {}) {
        const { application, scope, ...globalTags } = (options.globalTags ||
            {});
        const client = mainClient.childClient({
            prefix: options.prefix,
            suffix: options.suffix,
            globalTags,
        });
        privateData.set(this, { client });
    }
}
exports.MeliStatsD = MeliStatsD;
Object.getOwnPropertyNames(hot_shots_1.StatsD.prototype).forEach((key) => {
    if (key === 'constructor') {
        return;
    }
    MeliStatsD.prototype[key] = function statsdMethod(...args) {
        const { client } = privateData.get(this);
        return client[key](...args);
    };
});
