// @nordic/core-context/dist/esm/getters-CuJl4rad.js
import { AsyncLocalStorage } from 'node:async_hooks';
import { isTestEnv } from '@nordic/utils/validation';
import { OwnerStackError } from '@nordic/utils/error';

class NordicCoreContextError extends OwnerStackError {
    name = 'NordicCoreContextError';
}
function throwContextNotInitializedError() {
    throw new NordicCoreContextError('Nordic Core Context is not initialized.');
}

function getTestContextScope() {
    return global._n.ctx;
}

const storage = new AsyncLocalStorage();
function getContext() {
    const context = storage.getStore();
    if (isTestEnv()) {
        return getTestContextScope().c;
    }
    if (!context) {
        throwContextNotInitializedError();
    }
    return context;
}

function getDeviceType() {
    return getContext().deviceType;
}
function getPlatformType() {
    return getContext().platformType;
}
function getWebviewInfo() {
    return getContext().webviewInfo;
}
function getDeviceInfo() {
    return getContext().deviceInfo;
}
function getSafeAreaInsets() {
    return getContext().safeAreaInsets;
}
function getI18n() {
    return getContext().i18n;
}
function getNetworkState() {
    return getContext().networkState;
}
function getPlatform() {
    return getContext().platform;
}

export { getDeviceInfo as a, getDeviceType as b, getI18n as c, getNetworkState as d, getPlatform as e, getPlatformType as f, getContext as g, getSafeAreaInsets as h, getTestContextScope as i, getWebviewInfo as j, storage as s };
