// @nordic/core-context/dist/esm/server.development.mjs
import { serializeJSON } from '@nordic/utils/serialize';
import { g as getContext, s as storage, a as getDeviceInfo$1, j as getWebviewInfo } from './getters-CuJl4rad.js';
import { getReq, getRes } from '@meli/server-context';
import { parseCookieValue } from '@nordic/utils/cookie';
import { config } from 'frontend-config';
import { UAParser } from 'ua-parser-js';
import 'node:async_hooks';
import '@nordic/utils/validation';
import '@nordic/utils/error';

function getSerializableContext(staticMarkup) {
    const context = getContext();
    const result = {
        deviceType: context.deviceType,
        platformType: context.platformType,
        platform: context.platform,
        webviewInfo: context.webviewInfo,
        deviceInfo: context.deviceInfo,
        networkState: context.networkState,
        safeAreaInsets: context.safeAreaInsets,
    };
    if (staticMarkup) {
        return result;
    }
    result.i18n = context.i18n;
    return result;
}

function getCoreContextInitScript(staticMarkup) {
    return {
        id: '__NORDIC_CORE_CTX__',
        children: `_n.ctx.c=${serializeJSON(getSerializableContext(staticMarkup))};`,
    };
}

const KNOWN_TYPES = new Set([
    'desktop',
    'mobile',
    'tablet',
]);
function getDeviceType() {
    const { type } = getReq().device;
    if (KNOWN_TYPES.has(type)) {
        return type;
    }
    return 'desktop';
}

const KNOWN_OS$1 = new Set(['android', 'ios']);
function getPlatformType() {
    const { nativeApp } = getReq().device;
    if (!nativeApp) {
        return 'web';
    }
    if (nativeApp.name.toLowerCase() === 'smartpos') {
        return 'smartpos';
    }
    const os = nativeApp.os.toLowerCase();
    if (KNOWN_OS$1.has(os)) {
        return os;
    }
    return 'web';
}

const DC_COOKIE = 'x-nni-dc';
const CONTEXT_COOKIE = 'nni-ctx';
const DC_COOKIE_DICTIONARY = {
    back: {
        key: 'cam',
        type: 'binary',
    },
    perfCategory: {
        key: 'clss',
    },
    front: {
        key: 'fcam',
        type: 'binary',
    },
    flash: {
        key: 'flash',
        type: 'binary',
    },
    model: {
        key: 'mdl',
    },
    nfc: {
        key: 'nfc',
        type: 'binary',
    },
    rooted: {
        key: 'root',
        type: 'binary',
    },
    effectiveType: {
        key: 'etype',
    },
    connectionType: {
        key: 'ctype',
    },
    isConnectionSlow: {
        key: 'cslw',
        type: 'binary',
    },
    isConnectionExpensive: {
        key: 'cexp',
        type: 'binary',
    },
    osName: {
        key: 'os',
    },
    osVersion: {
        key: 'osv',
    },
    source: {
        key: 'src',
    },
};
const DEVICE_CAPABILITIES_GROUPS_DICTIONARY = {
    back: {
        main: 'deviceInfo',
        subs: ['capabilities', 'camera'],
    },
    perfCategory: {
        main: 'deviceInfo',
    },
    front: {
        main: 'deviceInfo',
        subs: ['capabilities', 'camera'],
    },
    flash: {
        main: 'deviceInfo',
        subs: ['capabilities', 'camera'],
    },
    model: {
        main: 'deviceInfo',
        formatter: (value) => value.replaceAll('+', ' '),
    },
    nfc: {
        main: 'deviceInfo',
        subs: ['capabilities'],
    },
    rooted: {
        main: 'deviceInfo',
        subs: ['os'],
    },
    effectiveType: {
        main: 'networkState',
    },
    connectionType: {
        main: 'networkState',
    },
    isConnectionSlow: {
        main: 'networkState',
    },
    isConnectionExpensive: {
        main: 'networkState',
    },
};
const LAST_COMMA_REGEX = /,$/;

function concatCookieValue(key, value) {
    return value ? `${key}=${value},` : '';
}
function setCookie(cookieName, cookieValue) {
    if (!cookieValue) {
        return;
    }
    const { cookies, platform } = getReq();
    const currentCookieValue = cookies[cookieName];
    if (currentCookieValue !== cookieValue) {
        const cookieOptions = {
            httpOnly: true,
            secure: true,
            sameSite: false,
        };
        if (platform) {
            cookieOptions.domain = config.get('url.mainDomain.link', platform.id, platform.siteId);
        }
        getRes().cookie(cookieName, cookieValue, cookieOptions);
    }
}

const VERSION_REGEX = /^v\d+$/;
/**
 * Parse webkit engine version.
 *
 * Valid versions are:
 * - v<number> (e.g., 'v2')
 * - 'legacy'.
 *
 * When engine received is not valid, 'v1' is returned.
 */
function parseEngineVersion(value) {
    return value && VERSION_REGEX.test(value) ? value : 'v1';
}
const SCOPE_DICTIONARY = {
    '0': 'view',
    '1': 'component',
    '2': 'page',
    '3': 'bottomsheet',
};
function parseScopeValue(value) {
    return SCOPE_DICTIONARY[value];
}

const WEBKIT_CONTEXT_DICTIONARY = {
    engine: { key: 'e', preparser: parseEngineVersion },
    version: { key: 'v' },
    debugMode: { key: 'd', type: 'binary' },
    fitToContent: { key: 'ftc', type: 'binary' },
    insetTop: { key: 'it', type: 'number' },
    insetRight: { key: 'ir', type: 'number' },
    insetBottom: { key: 'ib', type: 'number' },
    insetLeft: { key: 'il', type: 'number' },
    source: { key: 'src' },
    scope: { key: 's', preparser: parseScopeValue },
};
const CONTEXT_INFO_REGEX = /(e|v|ftc)=[^,]+,?/g;
function getWebkitInfoFromCookie(cookieValue) {
    if (!cookieValue) {
        return null;
    }
    return parseCookieValue(cookieValue, WEBKIT_CONTEXT_DICTIONARY);
}
function getWebviewProfile(platformType) {
    if (platformType === 'web') {
        return undefined;
    }
    const req = getReq();
    const { nativeApp, webkit } = req.device;
    const { engineVersion, nativeVersion } = webkit ?? {};
    const engineValue = engineVersion === '2.0.0' ? 'v2' : 'v1';
    const fitToContentValue = req.header('x-wk-ftc');
    const appInfo = { name: nativeApp.name, version: nativeApp.version };
    const webkitInfo = getWebkitInfoFromCookie(req.cookies[CONTEXT_COOKIE]);
    if (!webkitInfo) {
        return {
            nativeApp: appInfo,
            webkit: {
                engine: engineValue,
                version: nativeVersion,
                fitToContent: fitToContentValue === '1',
            },
        };
    }
    const safeAreaInsets = {
        top: webkitInfo.insetTop,
        right: webkitInfo.insetRight,
        bottom: webkitInfo.insetBottom,
        left: webkitInfo.insetLeft,
    };
    if (req.ajax) {
        return {
            safeAreaInsets,
            nativeApp: appInfo,
            webkit: {
                debugMode: webkitInfo.debugMode,
                version: webkitInfo.version,
                engine: webkitInfo.engine,
            },
        };
    }
    if (webkitInfo.source !== 'wk') {
        return {
            safeAreaInsets,
            nativeApp: appInfo,
            webkit: {
                debugMode: webkitInfo.debugMode,
                version: nativeVersion,
                fitToContent: fitToContentValue === '1',
                engine: engineValue,
            },
        };
    }
    return {
        safeAreaInsets,
        nativeApp: appInfo,
        webkit: {
            debugMode: webkitInfo.debugMode,
            version: webkitInfo.version,
            fitToContent: webkitInfo.fitToContent,
            engine: webkitInfo.engine,
            scope: webkitInfo.scope,
        },
    };
}
function setWebkitContextCookie(webkitInfo) {
    const currentCookieValue = getReq().cookies[CONTEXT_COOKIE];
    const engineCookiePart = concatCookieValue('e', webkitInfo.engine);
    const versionCookiePart = concatCookieValue('v', webkitInfo.version);
    const fitToContentCookiePart = concatCookieValue('ftc', webkitInfo.fitToContent ? '1' : '0');
    const updatedCookieValue = currentCookieValue
        ? currentCookieValue.replace(CONTEXT_INFO_REGEX, (_, key) => {
            switch (key) {
                case 'e':
                    return engineCookiePart;
                case 'v':
                    return versionCookiePart;
                default:
                    return fitToContentCookiePart;
            }
        })
        : engineCookiePart + versionCookiePart + fitToContentCookiePart;
    setCookie(CONTEXT_COOKIE, updatedCookieValue.replace(LAST_COMMA_REGEX, ''));
}

const KNOWN_OS = new Set([
    'Android',
    'Chrome OS',
    'Chromium OS',
    'iOS',
    'Linux',
    'macOS',
    'Windows',
]);
const DEVICE_DETECTION_CLIENT_HINTS = 'device-memory, dpr, viewport-width, rtt, downlink, ect, save-data';
/** HTTP headers **/
const ACCEPT_CH_HEADER = 'Accept-CH';
const DEVICE_MODEL_HEADER = 'Sec-CH-UA-Model';
const OS_PLATFORM_NAME_HEADER = 'Sec-CH-UA-Platform';
const OS_PLATFORM_VERSION_HEADER = 'Sec-CH-UA-Platform-Version';
const USER_AGENT = 'user-agent';

/**
 * Add hints to client hints request headers
 * define by device-detection [here](https://github.com/melisource/fury_frontend-device-detection/blob/master/src/client-hints.js#L44).
 */
function setClientHints() {
    getRes().setHeader(ACCEPT_CH_HEADER, `${DEVICE_DETECTION_CLIENT_HINTS}, ${OS_PLATFORM_NAME_HEADER}, ${OS_PLATFORM_VERSION_HEADER}, ${DEVICE_MODEL_HEADER}`);
}

function getDeviceModel() {
    const { headers } = getReq();
    const model = headers[DEVICE_MODEL_HEADER];
    if (typeof model === 'string' && model) {
        return model;
    }
    return UAParser(headers[USER_AGENT]).device.model;
}

function getDeviceOS() {
    const { device, headers } = getReq();
    const { osName, osVersion } = device;
    if (osName) {
        return { name: osName, version: osVersion };
    }
    const name = headers[OS_PLATFORM_NAME_HEADER];
    const version = headers[OS_PLATFORM_VERSION_HEADER];
    if (typeof name !== 'string' || !KNOWN_OS.has(name)) {
        return undefined;
    }
    return { name: name.toLowerCase(), version };
}

function getDeviceInfo() {
    return {
        os: getDeviceOS(),
        model: getDeviceModel(),
    };
}

const SLOW_EFFECTIVE_TYPES = new Set(['slow-2g', '2g', '3g']);
/**
 * Returns an initial value of the network state.
 * The final network state value is obtained from [@nordic/network](https://github.com/melisource/fury_nordic-webkit/tree/master/packages/nordic-network).
 */
function getNetworkState() {
    const { ect: effectiveType, downlink: downlinkStr, saveData, } = getReq().device;
    const downlink = typeof downlinkStr === 'string' ? +downlinkStr : downlinkStr;
    const isConnectionExpensive = saveData === 'on';
    const isConnectionSlow = typeof downlink === 'number' && downlink > 0 && downlink < 2;
    if (!effectiveType) {
        return { isConnectionExpensive, isConnectionSlow };
    }
    return {
        effectiveType,
        isConnectionExpensive,
        isConnectionSlow: isConnectionSlow || SLOW_EFFECTIVE_TYPES.has(effectiveType),
    };
}

function getDeviceInfoFromCookie(cookieValue) {
    if (!cookieValue) {
        return null;
    }
    return parseCookieValue(cookieValue, DC_COOKIE_DICTIONARY);
}
function getDeviceOSFromWebview() {
    const { cookies, device } = getReq();
    const { source, osName, osVersion } = getDeviceInfoFromCookie(cookies[DC_COOKIE]) || {};
    if (source === 'wk') {
        return osName ? { name: osName, version: osVersion } : undefined;
    }
    const name = device.nativeApp?.os || device.webviewOSInfo?.name || osName;
    const version = device.webviewOSInfo?.version || device.osVersion || osVersion;
    return name ? { name, version } : undefined;
}

function getProfileWithWebkitCookieInfo() {
    const cookieValue = getReq().cookies[DC_COOKIE];
    const os = getDeviceOSFromWebview();
    const profile = os ? { deviceInfo: { os } } : {};
    if (!cookieValue) {
        return profile;
    }
    const parsedValue = parseCookieValue(cookieValue, DC_COOKIE_DICTIONARY);
    return Object.entries(parsedValue).reduce((acc, [key, value]) => {
        const group = DEVICE_CAPABILITIES_GROUPS_DICTIONARY[key];
        if (!group) {
            return acc;
        }
        const { main: mainGroup, subs: subgroups, formatter } = group;
        if (!acc[mainGroup]) {
            acc[mainGroup] = {};
        }
        let profileBlock = acc[mainGroup];
        subgroups?.forEach((subgroup) => {
            profileBlock[subgroup] = profileBlock[subgroup] || {};
            profileBlock = profileBlock[subgroup];
        });
        profileBlock[key] =
            typeof value === 'string' && typeof formatter === 'function'
                ? formatter(value)
                : value;
        return acc;
    }, profile);
}

const OS_INFO_REGEX = /osv?=[^,]+,?/g;
/**
 * Recollects device profile information from the request, cookies and HTTP headers.
 */
function getDeviceProfile(platformType) {
    if (platformType !== 'web') {
        return getProfileWithWebkitCookieInfo();
    }
    setClientHints();
    return {
        deviceInfo: getDeviceInfo(),
        networkState: getNetworkState(),
    };
}
function setDeviceInfoCookie(deviceOS) {
    const currentCookieValue = getReq().cookies[DC_COOKIE];
    const osCookiePart = concatCookieValue('os', deviceOS.name);
    const osvCookiePart = concatCookieValue('osv', deviceOS.version);
    const updatedCookieValue = currentCookieValue
        ? currentCookieValue.replace(OS_INFO_REGEX, (match) => match.startsWith('os=') ? osCookiePart : osvCookiePart)
        : osCookiePart + osvCookiePart;
    setCookie(DC_COOKIE, updatedCookieValue.replace(LAST_COMMA_REGEX, ''));
}

const coreContextMiddleware = (req, res, next) => {
    const platformType = getPlatformType();
    const deviceProfile = getDeviceProfile(platformType);
    const webviewProfile = getWebviewProfile(platformType);
    const context = {
        deviceType: getDeviceType(),
        deviceInfo: deviceProfile.deviceInfo,
        networkState: deviceProfile.networkState,
        platformType,
    };
    if (webviewProfile) {
        const { safeAreaInsets } = webviewProfile;
        if (safeAreaInsets &&
            Object.values(safeAreaInsets).some((value) => typeof value === 'number')) {
            context.safeAreaInsets = safeAreaInsets;
        }
        context.webviewInfo = {
            nativeApp: webviewProfile.nativeApp,
            webkit: webviewProfile.webkit,
        };
    }
    return storage.run(context, next);
};
const webviewCookiesMiddleware = (req, res, next) => {
    if (!req.device.webView) {
        return next();
    }
    const deviceOS = getDeviceInfo$1()?.os;
    const webkitInfo = getWebviewInfo()?.webkit;
    if (deviceOS) {
        setDeviceInfoCookie(deviceOS);
    }
    if (webkitInfo) {
        setWebkitContextCookie(webkitInfo);
    }
    return next();
};

function getLayoutMiddlewareOptions() {
    return getContext().layoutMiddlewareOptions;
}

function setLayoutMiddlewareOptions(layoutMiddlewareOptions) {
    getContext().layoutMiddlewareOptions = layoutMiddlewareOptions;
}

export { coreContextMiddleware, getCoreContextInitScript, getLayoutMiddlewareOptions, setLayoutMiddlewareOptions, webviewCookiesMiddleware };
