// frontend-config/src/constants/regexps/ua/identified-in-apps.js
const facebookTags = [
  'FBAN',
  'FBIOS',
  'FBAV',
  'FBBV',
  'FBDV',
  'FBMD',
  'FBSN',
  'FBSV',
  'FBSS',
  'FBCR',
  'FBID',
  'FBLC',
  'FBOP',
  'FB_IAB',
];

exports.atlassian = '((A|a)tlassian)|((J|j)ira)|JIRA';

exports.discord = '(D|d)iscord(B|b)ot';

exports.facebookMessenger = `(${facebookTags.join('|')}).*(Messenger|MESSENGER)`;

exports.facebook = `(${facebookTags.join('|')})(?!.*(Messenger|MESSENGER))`;

exports.instagram = 'Instagram';

exports.linkedIn = 'LinkedIn(App|Bot)?';

exports.outlookMail = 'Outlook\\s*Mail';

exports.pinterest = `Pinterest(${[
  // TODO: revisar regla desactivada
  // eslint-disable-next-line no-useless-escape
  '\/(Android|iOS)',
  '\\s+for\\s+(Android|iOS)',
].join('|')})`;

// TODO: revisar regla desactivada
// eslint-disable-next-line no-useless-escape
exports.slack = 'Slack(\/|\\s)';

exports.snapchat = 'Snapchat';

exports.telegram = 'TelegramBot';

exports.tikTok = [
  'TikTok',
  'BytedanceWebview',
  'musical_ly',
].join('|');

exports.twitter = [
  'Twitter(?!(B|b)ot)',
  '(?!like\\s+)Twitter(B|b)ot',
].join('|');

exports.weChat = [
  'micromessenger',
  'MicroMessenger',
  'WeChat',
].join('|');

exports.whatsApp = 'WhatsApp';
