// @nordic/rendering-context/src/common/types.ts
export type NordicRenderingContextAssets = {
  cdn: string;
  prefix: string;
  manifest: Map<string, string>;
  mainAssetsNames: {
    scripts: Set<string>;
    styles: Set<string>;
  };
};

export type NordicRenderingContextAppProps = {
  pageProps?: Record<string, unknown>;
  sharedState?: Record<string, unknown>;
};

export type NordicRenderingContextFlags = {
  isBot?: boolean;
  loadable?: boolean;
  islands?: boolean;
  staticMarkup?: boolean;
  pipeableStream?: boolean;
  stream?: boolean;
};

export type NordicRenderingContextSupport = {
  customElements: {
    builtIn: boolean;
    define: boolean;
  };
  imgLoading: boolean;
  prefetch: boolean;
  sameSite: boolean;
  share: boolean;
  outdated: boolean;
};

export type NordicRenderingContext = {
  assets: NordicRenderingContextAssets;
  flags: NordicRenderingContextFlags;
  support: NordicRenderingContextSupport;
  title: string;
  framework: 'nextjs' | 'nordic';
  appProps?: NordicRenderingContextAppProps;
  mainEntry?: string;
  nonce?: string;
  csrfToken?: string;
  trustedTypesRegex?: string | null;
  // TODO: NEXT MAJOR remove this
  wrapperProps?: {
    id?: string;
    className: string;
  };
};

declare global {
  interface NordicWindowContext {
    r: NordicRenderingContext;
  }
}
