// @andes/textfield/build/cjs/index.production.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _extends = require('@babel/runtime/helpers/extends');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var React = require('react');
var classNames = require('classnames');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var focus = require('@react-aria/focus');
var context = require('@andes/context');
var keyboardCodes = require('@andes/common/build/constants/keyboardCodes');
var autofocus = require('@andes/common/build/utils/autofocus');
var utils = require('@andes/common/build/webkit/utils');
var isMobile = require('@andes/common/build/utils/isMobile');
var includes = require('@andes/common/build/utils/includes');
var useSrLabel = require('@andes/common/build/hooks/useSrLabel');
var visuallyHidden = require('@andes/visually-hidden');
var FeedbackPositive12 = require('@andes/icons/FeedbackPositive12');
var FeedbackPositive20 = require('@andes/icons/FeedbackPositive20');
var form = require('@andes/form');
var themes = require('@andes/themes');
var computeDataAndesState = require('@andes/common/build/utils/computeDataAndesState');
require('@andes/common/build/utils/propTypesValidators');
var VisibilityON20 = require('@andes/icons/VisibilityON20');
var VisibilityOFF20 = require('@andes/icons/VisibilityOFF20');
var transformToKebabCase = require('@andes/common/build/utils/transformToKebabCase');
var dropdown = require('@andes/dropdown');
var progressIndicatorCircular = require('@andes/progress-indicator-circular');
var autosize = require('autosize');
var Close12 = require('@andes/icons/Close12');
var Close20 = require('@andes/icons/Close20');
var warning = require('@andes/common/build/utils/warning');
require('prop-types');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var _objectWithoutProperties__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutProperties);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var includes__default = /*#__PURE__*/_interopDefaultLegacy(includes);
var useSrLabel__default = /*#__PURE__*/_interopDefaultLegacy(useSrLabel);
var FeedbackPositive12__default = /*#__PURE__*/_interopDefaultLegacy(FeedbackPositive12);
var FeedbackPositive20__default = /*#__PURE__*/_interopDefaultLegacy(FeedbackPositive20);
var VisibilityON20__default = /*#__PURE__*/_interopDefaultLegacy(VisibilityON20);
var VisibilityOFF20__default = /*#__PURE__*/_interopDefaultLegacy(VisibilityOFF20);
var transformToKebabCase__default = /*#__PURE__*/_interopDefaultLegacy(transformToKebabCase);
var autosize__default = /*#__PURE__*/_interopDefaultLegacy(autosize);
var Close12__default = /*#__PURE__*/_interopDefaultLegacy(Close12);
var Close20__default = /*#__PURE__*/_interopDefaultLegacy(Close20);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var enTranslations = {
  "andes-textfield\u0004{0} caracteres restantes": [null, "{0} remaining characters"],
  "andes-textfield\u0004Maximo {0} caracteres": [null, "Maximum of {0} characters"],
  "andes-textfield\u0004Campo de texto valido": [null, "Valid text field"],
  "andes-textfield\u0004Campo de texto procesando": [null, "Text field processing"],
  "andes-textfield\u0004Borrar": [null, "Remove"],
  "andes-textfield\u0004Dígito {0}": [null, "Digit {0}"],
  "andes-textfield\u0004Mostrar contraseña": [null, "Show Password"]
};
var zhCnTranslations = {
  "andes-textfield\u0004{0} caracteres restantes": [null, "剩余 {0} 个字符"],
  "andes-textfield\u0004Maximo {0} caracteres": [null, "最多可输入 {0} 个字符"],
  "andes-textfield\u0004Campo de texto valido": [null, "有效文本字段"],
  "andes-textfield\u0004Campo de texto procesando": [null, "文本字段处理中"],
  "andes-textfield\u0004Borrar": [null, "移除"],
  "andes-textfield\u0004Dígito {0}": [null, "数字 {0}"],
  "andes-textfield\u0004Mostrar contraseña": [null, "显示密码"]
};
var esArTranslations = {
  "andes-textfield\u0004{0} caracteres restantes": [null, "{0} caracteres restantes"],
  "andes-textfield\u0004Maximo {0} caracteres": [null, "Máximo {0} caracteres"],
  "andes-textfield\u0004Campo de texto valido": [null, "Campo de texto valido"],
  "andes-textfield\u0004Campo de texto procesando": [null, "Campo de texto procesando"],
  "andes-textfield\u0004Borrar": [null, "Borrar"],
  "andes-textfield\u0004Dígito {0}": [null, "Dígito {0}"],
  "andes-textfield\u0004Mostrar contraseña": [null, "Mostrar Contraseña"]
};
var ptBrTranslations = {
  "andes-textfield\u0004{0} caracteres restantes": [null, "{0} caracteres restantes"],
  "andes-textfield\u0004Maximo {0} caracteres": [null, "Máximo de {0} caracteres"],
  "andes-textfield\u0004Campo de texto valido": [null, "Campo de texto válido"],
  "andes-textfield\u0004Campo de texto procesando": [null, "Processamento de campo de texto"],
  "andes-textfield\u0004Borrar": [null, "Excluir"],
  "andes-textfield\u0004Dígito {0}": [null, "Dígito {0}"],
  "andes-textfield\u0004Mostrar contraseña": [null, "Mostrar Senha"]
};
const formControlNamespace = 'andes-form-control';
const splitButtonNamespace = `${formControlNamespace}__split-button`;
const namespace$3 = 'andes-textfield';
const ValidatedIcon = _ref => {
  let notificationsLabel = _ref.notificationsLabel,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'large' : _ref$size;
  const srLabelVisibility = useSrLabel__default["default"]();
  return /*#__PURE__*/React__default["default"].createElement("span", {
    role: "status",
    "aria-live": "polite"
  }, /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, null, srLabelVisibility && notificationsLabel), /*#__PURE__*/React__default["default"].createElement("span", {
    className: `${formControlNamespace}__validated-icon`
  }, size === 'large' ? /*#__PURE__*/React__default["default"].createElement(FeedbackPositive20__default["default"], {
    color: "currentColor"
  }) : /*#__PURE__*/React__default["default"].createElement(FeedbackPositive12__default["default"], {
    color: "currentColor"
  })));
};
var ValidatedIcon$1 = ValidatedIcon;
const WARNING_DEPRECATED_MODIFIER_WARNING = `\`warning\` modifier will be deprecated in a future major version.
Please use \`caution\` instead.`;
const numericNamespace = 'andes-form-control--numeric';
const namespace$2 = 'andes-form-control';
const translationsNamespace = 'andes-textfield';
const onlyNumberRegex = /^[0-9]*$/;
const DEFAULT_FOCUS_INDEX = 0;
const NOT_FOCUSABLE_INDEX = -1;
const numbersRegex = /[^0-9]/gi;
const specialCharacters = ['Backspace', 'Tab', 'Enter', 'Control', 'Shift', 'Escape', 'Meta', 'Alt', 'CapsLock', 'ArrowLeft', 'ArrowRight'];
const getFocusIndexOrFalsy = focus => focus === false ? NOT_FOCUSABLE_INDEX : Number(focus === true ? DEFAULT_FOCUS_INDEX : focus);
const getDefaultValue = digits => {
  if (!digits || typeof digits !== 'number') return [];
  return new Array(digits).fill('');
};
const isEmpty = value => {
  const array = typeof value === 'string' ? value.split('') : value;
  return array.every(item => !item);
};
const parsedValues = (value, digits) => {
  if (typeof value === 'string') {
    let parsedValue = value.replace(numbersRegex, '').split('');
    if (digits && parsedValue.length < digits) {
      const diff = digits - parsedValue.length;
      const arrayEmpty = new Array(diff).fill('');
      parsedValue = parsedValue.concat(arrayEmpty);
    }
    return parsedValue;
  }
  if (Array.isArray(value)) return value;
};
const Numeric = /*#__PURE__*/function () {
  const Numeric = _ref => {
    let _ref$centered = _ref.centered,
      centered = _ref$centered === void 0 ? false : _ref$centered,
      className = _ref.className,
      defaultValue = _ref.defaultValue,
      _ref$digits = _ref.digits,
      digits = _ref$digits === void 0 ? 1 : _ref$digits,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      _ref$focus = _ref.focus,
      focus$1 = _ref$focus === void 0 ? DEFAULT_FOCUS_INDEX : _ref$focus,
      helper = _ref.helper,
      id = _ref.id,
      label = _ref.label,
      modifier = _ref.modifier,
      onChange = _ref.onChange,
      onPaste = _ref.onPaste,
      srLabel = _ref.srLabel,
      _ref$strict = _ref.strict,
      strict = _ref$strict === void 0 ? false : _ref$strict,
      value = _ref.value,
      ref = _ref.ref,
      andesDataAttributes = _ref.andesDataAttributes;
    const _useState = React.useState(parsedValues(defaultValue, digits) || getDefaultValue(digits)),
      _useState2 = _slicedToArray__default["default"](_useState, 2),
      inputValue = _useState2[0],
      setInputValue = _useState2[1];
    const _useState3 = React.useState(getFocusIndexOrFalsy(focus$1)),
      _useState4 = _slicedToArray__default["default"](_useState3, 2),
      focusedItem = _useState4[0],
      setFocusedItem = _useState4[1];
    const _useRef = React.useRef(typeof value !== 'undefined'),
      isControlled = _useRef.current;
    const currentValue = isControlled ? parsedValues(value, digits) || [] : inputValue;
    const inputsRef = React.useRef([]);
    const firstRender = React.useRef(true);
    const _useFocusRing = focus.useFocusRing(),
      isFocusVisible = _useFocusRing.isFocusVisible,
      focusProps = _useFocusRing.focusProps;
    const _useAndesContext = context.useAndesContext('@andes/textfield', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations,
        'zh-CN': zhCnTranslations
      }),
      i18n = _useAndesContext.i18n,
      device = _useAndesContext.device;
    const isMobile$1 = device === null || device === void 0 ? void 0 : device.mobile;
    const componentId = context.useId(id);
    const moveFocusBackwards = (event, index) => {
      const newIndex = index - 1;
      const newValue = [...currentValue];
      if (newIndex >= 0) newValue[newIndex] = '';
      setFocusedItem(state => newIndex > -1 ? newIndex : state);
      if (!isControlled) {
        setInputValue(newValue);
      }
      if (typeof onChange === 'function') {
        onChange(event, newValue);
      }
    };
    const moveFocusForward = () => {
      setFocusedItem(state => {
        const newIndex = state + 1;
        return newIndex < digits ? newIndex : state;
      });
    };
    const assignPastedValues = (event, values) => {
      const newValues = currentValue;
      const newFocus = values.length > digits ? digits - 1 : values.length - 1;
      for (let i = 0; i < currentValue.length; i += 1) {
        var _values$i;
        newValues[i] = (_values$i = values[i]) !== null && _values$i !== void 0 ? _values$i : '';
      }
      setInputValue(newValues);
      setFocusedItem(newFocus);
      if (typeof onPaste === 'function') onPaste(event, newValues);
    };
    const handlePaste = event => {
      var _event$clipboardData;
      const textPlainItems = Array.from((event === null || event === void 0 || (_event$clipboardData = event.clipboardData) === null || _event$clipboardData === void 0 ? void 0 : _event$clipboardData.items) || []).filter(item => item.type === 'text/plain' || item.type === 'text');
      if (textPlainItems.length > 0) {
        var _textPlainItems$;
        (_textPlainItems$ = textPlainItems[0]) === null || _textPlainItems$ === void 0 || _textPlainItems$.getAsString(text => {
          const pastedValues = text.replace(numbersRegex, '').split('');
          assignPastedValues(event, pastedValues);
        });
      }
    };
    const onInputHandler = event => {
      const eventValue = event.target.value;
      if ((eventValue === null || eventValue === void 0 ? void 0 : eventValue.length) === digits) {
        setFocusedItem(digits - 1);
        return;
      }
      if (!onlyNumberRegex.test(eventValue) || eventValue.length > 1 && eventValue.length !== digits) {
        return;
      }
      if ((eventValue === null || eventValue === void 0 ? void 0 : eventValue.trim()) !== '') {
        moveFocusForward();
      }
    };
    const onChangeHandler = (event, index) => {
      const eventValue = event.target.value;
      if (!onlyNumberRegex.test(eventValue) || eventValue.length > 1 && eventValue.length !== digits) {
        return;
      }
      const newValue = [...currentValue];
      if (eventValue.length === digits) {
        const values = eventValue.split('');
        for (let i = 0; i < currentValue.length; i += 1) {
          var _values$i2;
          newValue[i] = (_values$i2 = values[i]) !== null && _values$i2 !== void 0 ? _values$i2 : '';
        }
      } else {
        newValue[index] = eventValue;
      }
      if (!isControlled) {
        setInputValue(newValue);
      }
      if (typeof onChange === 'function') {
        onChange(event, newValue);
      }
      return newValue;
    };
    const onKeyDownHandler = (event, index) => {
      const eventValue = event.target.value;
      if (event.code === keyboardCodes.BACKSPACE || event.key === 'Backspace') {
        if (eventValue === '') {
          moveFocusBackwards(event, index);
          return;
        }
      }
      if (event.code === keyboardCodes.TAB || event.key === 'Tab') {
        if (index === digits - 1 || index === 0) {
          setFocusedItem(NOT_FOCUSABLE_INDEX);
          return;
        }
      }
      const isPasting = (event.metaKey || event.ctrlKey) && event.code === keyboardCodes.KEY_V;
      if (isPasting) return;
      if (includes__default["default"](specialCharacters, event.key)) return;
      const isInvalid = event.key.match(numbersRegex);
      if (isInvalid) {
        event.preventDefault();
      }
    };
    const focusValidItem = index => {
      const item = inputsRef.current[index];
      if (item) {
        if (!firstRender.current) {
          item.focus();
        } else {
          autofocus.autofocus(item, firstRender.current);
          firstRender.current = false;
        }
      }
    };
    const onClickHandler = event => {
      if (isEmpty(currentValue) && (utils.isWebView() || !isMobile.hasTouchSupport() && !isMobile$1)) {
        event.preventDefault();
        if (focusedItem === DEFAULT_FOCUS_INDEX) {
          focusValidItem(DEFAULT_FOCUS_INDEX);
        } else {
          setFocusedItem(DEFAULT_FOCUS_INDEX);
        }
      }
    };
    const onFocusHandler = (event, index) => {
      setFocusedItem(index);
      if (typeof focusProps.onFocus === 'function') focusProps.onFocus(event);
    };
    const handleBlur = event => {
      setFocusedItem(NOT_FOCUSABLE_INDEX);
      if (typeof focusProps.onBlur === 'function') focusProps.onBlur(event);
    };
    React.useEffect(() => {
      if (focusedItem !== NOT_FOCUSABLE_INDEX) {
        focusValidItem(focusedItem);
      }
    }, [focusedItem]);
    const containerClassNames = classNames__default["default"](numericNamespace, className, {
      [`${namespace$2}--${modifier}`]: modifier,
      [`${namespace$2}--centered`]: centered
    });
    return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
      className: containerClassNames,
      role: "group",
      "aria-label": srLabel,
      "aria-labelledby": label ? `${componentId}-label` : undefined,
      "aria-describedby": helper ? `${componentId}-message` : undefined,
      onPaste: handlePaste,
      ref: ref
    }, andesDataAttributes), label && /*#__PURE__*/React__default["default"].createElement("span", {
      id: `${componentId}-label`,
      className: `${namespace$2}__label`,
      "data-andes-codeinput-label": true
    }, label), /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${numericNamespace}-wrapper`
    }, [...Array(digits)].map((_, index) => {
      const textFieldPadding = classNames__default["default"](`${namespace$2}__textfield-space`, {
        [`${namespace$2}__textfield-space--big`]: digits === 6 && index === 2
      });
      const inputWrapperClasses = classNames__default["default"](namespace$2, `${namespace$2}--textfield`, {
        [`${namespace$2}--centered`]: centered,
        [`${namespace$2}--focused`]: focusedItem === index,
        [`${namespace$2}--visible-focused`]: isFocusVisible && focusedItem === index,
        [`${namespace$2}--${modifier}`]: modifier,
        [`${namespace$2}--disabled`]: disabled
      });
      return /*#__PURE__*/React__default["default"].createElement("div", {
        className: textFieldPadding,
        key: `${componentId}-${index}`
      }, /*#__PURE__*/React__default["default"].createElement("div", {
        className: inputWrapperClasses
      }, /*#__PURE__*/React__default["default"].createElement("div", {
        className: `${namespace$2}__control`
      }, /*#__PURE__*/React__default["default"].createElement("input", _extends__default["default"]({
        className: `${namespace$2}__field`,
        "aria-label": i18n.pgettext(translationsNamespace, 'Dígito {0}', index + 1),
        type: strict ? 'tel' : 'number',
        pattern: "[0-9]*",
        value: currentValue[index],
        id: `${componentId}-digit-${index}`,
        onChange: evt => onChangeHandler(evt, index),
        onInput: onInputHandler,
        onKeyDown: evt => onKeyDownHandler(evt, index),
        onFocus: evt => onFocusHandler(evt, index),
        onBlur: handleBlur,
        onMouseDown: onClickHandler,
        disabled: disabled,
        ref: el => {
          inputsRef.current[index] = el;
        },
        "aria-invalid": modifier === 'error'
      }, index === 0 && {
        autoComplete: 'one-time-code',
        autoFocus: focus$1 !== false
      }, {
        "data-andes-codeinput-input": true
      })), modifier === 'completed' && /*#__PURE__*/React__default["default"].createElement(ValidatedIcon$1, {
        notificationsLabel: i18n.pgettext(translationsNamespace, 'Campo de texto valido')
      }))));
    })), helper && /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${namespace$2}__bottom`
    }, /*#__PURE__*/React__default["default"].createElement(form.Helper, {
      id: `${id}-message`,
      modifier: modifier === 'error' ? modifier : undefined,
      andesDataAttributes: {
        'data-andes-codeinput-helper': true
      }
    }, helper)));
  };
  Numeric.displayName = 'Numeric';
  return Numeric;
}();
var Numeric$1 = Numeric;
const _excluded$5 = ["className", "id", "ref"];
const namespace$1 = 'andes-code-input';
const CodeInput = /*#__PURE__*/function () {
  const CodeInput = _ref => {
    let className = _ref.className,
      id = _ref.id,
      ref = _ref.ref,
      otherProps = _objectWithoutProperties__default["default"](_ref, _excluded$5);
    const _useThemeContext = themes.useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const codeInputClassNames = classNames__default["default"](namespace$1, className, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('textfield'));
    const componentId = context.useId(id);
    const disabled = otherProps.disabled,
      modifier = otherProps.modifier;
    return /*#__PURE__*/React__default["default"].createElement(Numeric$1, _extends__default["default"]({
      className: codeInputClassNames,
      id: componentId,
      ref: ref,
      andesDataAttributes: _objectSpread__default["default"](_objectSpread__default["default"]({
        'data-andes-codeinput': true,
        'data-andes-codeinput-modifier': modifier
      }, computeDataAndesState.computeDataAndesState({
        disabled,
        error: modifier === 'error'
      })), themeHash)
    }, otherProps));
  };
  CodeInput.displayName = 'CodeInput';
  return CodeInput;
}();
var CodeInput$1 = CodeInput;
var handlers = {
  onFocus: function onFocus(element) {
    element === null || element === void 0 || element.classList.add('andes-form-control--focused');
  },
  onBlur: function onBlur(element, value) {
    element === null || element === void 0 || element.classList.remove('andes-form-control--focused');
    if (value && value.length) {
      element === null || element === void 0 || element.classList.add('andes-form-control--floated');
    } else {
      element === null || element === void 0 || element.classList.remove('andes-form-control--floated');
    }
  },
  onChange: function onChange(value, inputHiddenRef) {
    if (inputHiddenRef) {
      inputHiddenRef.value = value;
    }
  }
};
const _excluded$4 = ["autoFocus", "className", "defaultValue", "helper", "i18n", "id", "innerRef", "inputProps", "label", "maxLength", "modifier", "onClickShowPassword", "showHide", "srLabel", "value", "andesDataAttributes"];
const namespace = 'andes-form-control';
const iconAttributes = {
  color: 'currentColor'
};
class Password extends React__default["default"].Component {
  constructor(props) {
    super(props);
    this.state = {
      focused: props.autoFocus || false,
      showPassword: false
    };
    this.isUncontrolledValue = props.value === undefined;
    this.setField = this.setField.bind(this);
    this.setContainer = this.setContainer.bind(this);
    this.onBlurHandler = this.onBlurHandler.bind(this);
    this.onFocusHandler = this.onFocusHandler.bind(this);
    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.onChangeIconHandler = this.onChangeIconHandler.bind(this);
    this.renderShowHidePasswordButton = this.renderShowHidePasswordButton.bind(this);
    this.setHiddenPass = this.setHiddenPass.bind(this);
  }
  onBlurHandler(event) {
    const container = this.container,
      field = this.field;
    const value = field.value;
    const onBlur = this.props.onBlur;
    handlers.onBlur(container, value);
    this.setState({
      focused: false
    });
    if (onBlur) onBlur(event);
  }
  onFocusHandler(event) {
    const container = this.container;
    const onFocus = this.props.onFocus;
    handlers.onFocus(container);
    this.setState({
      focused: true
    });
    if (onFocus) onFocus(event);
  }
  onChangeHandler(event) {
    const value = event.target.value;
    const _this$props = this.props,
      _this$props$showHide = _this$props.showHide,
      showHide = _this$props$showHide === void 0 ? false : _this$props$showHide,
      onChange = _this$props.onChange;
    const hiddenPass = this.hiddenPass;
    if (showHide && this.isUncontrolledValue) handlers.onChange(value, hiddenPass);
    if (onChange) onChange(event);
  }
  onChangeIconHandler() {
    const onClickShowPassword = this.props.onClickShowPassword;
    this.setState(prevState => _objectSpread__default["default"](_objectSpread__default["default"]({}, prevState), {}, {
      showPassword: !prevState.showPassword
    }));
    if (onClickShowPassword) onClickShowPassword();
  }
  setField(fieldRef) {
    this.field = fieldRef;
  }
  setContainer(containerRef) {
    const innerRef = this.props.innerRef;
    this.container = containerRef;
    if (innerRef) innerRef.current = containerRef;
  }
  setHiddenPass(passRef) {
    this.hiddenPass = passRef;
  }
  renderShowHidePasswordButton(i18n) {
    const id = this.props.id;
    const showPassword = this.state.showPassword;
    return /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${namespace}__password-embedded`
    }, /*#__PURE__*/React__default["default"].createElement("input", {
      checked: showPassword,
      className: `${namespace}__password-checkbox`,
      id: `${id}-checkbox`,
      type: "checkbox",
      onChange: this.onChangeIconHandler,
      "aria-label": i18n.pgettext('andes-textfield', 'Mostrar contraseña')
    }), /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${namespace}__password-icon`
    }, showPassword ? /*#__PURE__*/React__default["default"].createElement(VisibilityON20__default["default"], iconAttributes) : /*#__PURE__*/React__default["default"].createElement(VisibilityOFF20__default["default"], iconAttributes)));
  }
  render() {
    const _this$props2 = this.props,
      _this$props2$autoFocu = _this$props2.autoFocus,
      autoFocus = _this$props2$autoFocu === void 0 ? false : _this$props2$autoFocu,
      className = _this$props2.className,
      defaultValue = _this$props2.defaultValue,
      helper = _this$props2.helper,
      i18n = _this$props2.i18n,
      id = _this$props2.id;
    _this$props2.innerRef;
    const _this$props2$inputPro = _this$props2.inputProps,
      inputProps = _this$props2$inputPro === void 0 ? {} : _this$props2$inputPro,
      label = _this$props2.label,
      _this$props2$maxLengt = _this$props2.maxLength,
      maxLength = _this$props2$maxLengt === void 0 ? 120 : _this$props2$maxLengt,
      modifier = _this$props2.modifier;
    _this$props2.onClickShowPassword;
    const _this$props2$showHide = _this$props2.showHide,
      showHide = _this$props2$showHide === void 0 ? false : _this$props2$showHide,
      srLabel = _this$props2.srLabel,
      value = _this$props2.value,
      andesDataAttributes = _this$props2.andesDataAttributes,
      other = _objectWithoutProperties__default["default"](_this$props2, _excluded$4);
    const _this$state = this.state,
      focused = _this$state.focused,
      showPassword = _this$state.showPassword;
    const fieldProps = _objectSpread__default["default"](_objectSpread__default["default"](_objectSpread__default["default"]({}, other), inputProps), {}, {
      'aria-describedby': helper ? `${id}-message` : undefined,
      'aria-invalid': modifier === 'error',
      'aria-label': srLabel,
      autoFocus,
      className: `${namespace}__field`,
      id,
      maxLength: maxLength,
      onBlur: this.onBlurHandler,
      onChange: this.onChangeHandler,
      onFocus: this.onFocusHandler,
      ref: this.setField,
      type: showPassword ? 'text' : 'password',
      value,
      defaultValue,
      autoComplete: inputProps.autoComplete || 'current-password'
    });
    const fieldHiddenProps = {
      style: {
        display: 'none'
      },
      type: 'password',
      id: 'password-hidden',
      value,
      defaultValue,
      ref: this.setHiddenPass,
      autoComplete: inputProps.autoComplete || 'current-password'
    };
    const classes = classNames__default["default"](namespace, className, `${namespace}--textfield`, {
      [`${namespace}--password`]: showHide,
      [`${namespace}--focused`]: autoFocus,
      [`${namespace}--${modifier}`]: modifier
    });
    return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
      className: classes,
      ref: this.setContainer,
      "data-andes-password": true
    }, andesDataAttributes), label && /*#__PURE__*/React__default["default"].createElement("label", {
      htmlFor: id,
      className: `${namespace}__label`
    }, label), /*#__PURE__*/React__default["default"].createElement(focus.FocusRing, {
      within: true,
      isTextInput: true,
      focusRingClass: focused ? 'visible-focused' : ''
    }, /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${namespace}__control`
    }, /*#__PURE__*/React__default["default"].createElement("input", _extends__default["default"]({
      key: "field"
    }, fieldProps, {
      "data-andes-password-input": true
    })), /*#__PURE__*/React__default["default"].createElement("input", _extends__default["default"]({
      key: "field-password-hidden"
    }, fieldHiddenProps)), showHide && this.renderShowHidePasswordButton(i18n))), (helper || showHide) && /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${namespace}__bottom`
    }, /*#__PURE__*/React__default["default"].createElement(form.Helper, {
      id: `${id}-message`,
      modifier: modifier === 'error' ? modifier : undefined,
      andesDataAttributes: {
        'data-andes-password-helper': true
      }
    }, helper)));
  }
}
var Password$1 = Password;
const _excluded$3 = ["ref", "id", "className"];
const PasswordContainer = /*#__PURE__*/function () {
  const PasswordContainer = _ref => {
    let ref = _ref.ref,
      id = _ref.id,
      className = _ref.className,
      otherProps = _objectWithoutProperties__default["default"](_ref, _excluded$3);
    const componentId = context.useId(id);
    const _useThemeContext = themes.useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const _useAndesContext = context.useAndesContext('@andes/textfield', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations,
        'zh-CN': zhCnTranslations
      }),
      i18n = _useAndesContext.i18n;
    return /*#__PURE__*/React__default["default"].createElement(Password$1, _extends__default["default"]({
      innerRef: ref
    }, otherProps, {
      i18n: i18n,
      id: componentId,
      className: classNames__default["default"](className, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('textfield')),
      andesDataAttributes: themeHash
    }));
  };
  PasswordContainer.displayName = 'PasswordContainer';
  return PasswordContainer;
}();
var PasswordContainer$1 = PasswordContainer;
const ClearButton = _ref => {
  let namespace = _ref.namespace,
    onClick = _ref.onClick,
    _ref$ariaLabel = _ref['aria-label'],
    ariaLabel = _ref$ariaLabel === void 0 ? '' : _ref$ariaLabel,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'large' : _ref$size;
  return /*#__PURE__*/React__default["default"].createElement("button", {
    "aria-label": ariaLabel,
    className: `${namespace}__clear-icon`,
    onClick: onClick,
    onMouseDown: e => e.preventDefault(),
    type: "button"
  }, size === 'large' ? /*#__PURE__*/React__default["default"].createElement(Close20__default["default"], {
    color: "currentColor"
  }) : /*#__PURE__*/React__default["default"].createElement(Close12__default["default"], {
    color: "currentColor"
  }));
};
var ClearButton$1 = ClearButton;
const DEPRECATED_MODIFIERS = ['warning'];
const computeModifier = modifier => {
  warning__default["default"](!modifier || typeof modifier === 'string' && !DEPRECATED_MODIFIERS.includes(modifier), WARNING_DEPRECATED_MODIFIER_WARNING);
  switch (modifier) {
    case 'warning':
    case 'caution':
      return 'caution';
    case 'error':
      return 'error';
    default:
      return undefined;
  }
};
const _excluded$2 = ["centered", "children", "className", "clear", "countdown", "disabled", "defaultValue", "embeddedChildren", "focus", "helper", "id", "innerRef", "i18n", "inputProps", "label", "maxLength", "modifier", "multiline", "numeric", "onChange", "placeholder", "prefix", "rows", "select", "size", "split", "srLabel", "suffix", "value", "autoFocus", "andesDataAttributes"];
const emptyEvent = {
  target: {
    value: ''
  }
};
class TextField extends React__default["default"].Component {
  constructor(props) {
    super(props);
    const value = String(props.defaultValue || props.value || '');
    this.state = {
      height: 'auto',
      fieldRef: null,
      countdown: value.length,
      focused: props.autoFocus || false,
      prevProps: props
    };
    this.isUncontrolledValue = props.value === undefined;
    this.setField = this.setField.bind(this);
    this.setRef = this.setRef.bind(this);
    this.onBlurHandler = this.onBlurHandler.bind(this);
    this.onFocusHandler = this.onFocusHandler.bind(this);
    this.onInputHandler = this.onInputHandler.bind(this);
    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.onClearHandler = this.onClearHandler.bind(this);
  }
  componentDidMount() {
    const _this$props = this.props,
      _this$props$multiline = _this$props.multiline,
      multiline = _this$props$multiline === void 0 ? false : _this$props$multiline,
      _this$props$focus = _this$props.focus,
      focus = _this$props$focus === void 0 ? false : _this$props$focus,
      _this$props$select = _this$props.select,
      select = _this$props$select === void 0 ? false : _this$props$select,
      _this$props$autoFocus = _this$props.autoFocus,
      autoFocus = _this$props$autoFocus === void 0 ? false : _this$props$autoFocus;
    const field = this.field;
    if (multiline && field) autosize__default["default"](field);
    if (focus || autoFocus) {
      autofocus.autofocus(field);
    }
    if (select) field === null || field === void 0 || field.select();
  }
  static getDerivedStateFromProps(nextProps, prevState) {
    const fieldRef = prevState.fieldRef,
      prevProps = prevState.prevProps;
    const _nextProps$multiline = nextProps.multiline,
      multiline = _nextProps$multiline === void 0 ? false : _nextProps$multiline,
      nextValue = nextProps.value,
      _nextProps$focus = nextProps.focus,
      focus = _nextProps$focus === void 0 ? false : _nextProps$focus,
      _nextProps$select = nextProps.select,
      select = _nextProps$select === void 0 ? false : _nextProps$select,
      _nextProps$countdown = nextProps.countdown,
      countdown = _nextProps$countdown === void 0 ? false : _nextProps$countdown;
    const isUncontrolledValue = prevState.isUncontrolledValue;
    const propsToCheck = ['multiline', 'focus', 'select', 'value', 'countdown'];
    const changedProps = propsToCheck.some(prop => nextProps[prop] !== prevProps[prop]);
    if (changedProps) {
      if (multiline && fieldRef) autosize__default["default"].update(fieldRef);
      if (focus) fieldRef === null || fieldRef === void 0 || fieldRef.focus();
      if (select) fieldRef === null || fieldRef === void 0 || fieldRef.select();
      if (countdown && !isUncontrolledValue && nextValue !== (fieldRef === null || fieldRef === void 0 ? void 0 : fieldRef.value)) {
        return {
          countdown: String(nextValue).length,
          prevProps: nextProps
        };
      }
    }
    return changedProps ? {
      prevProps: nextProps
    } : null;
  }
  onBlurHandler(event) {
    const onBlur = this.props.onBlur;
    this.setState({
      focused: false
    });
    if (onBlur) onBlur(event);
  }
  onFocusHandler(event) {
    const onFocus = this.props.onFocus;
    this.setState({
      focused: true
    });
    if (onFocus) onFocus(event);
  }
  onInputHandler(event) {
    const onInput = this.props.onInput;
    const value = event.target.value;
    this.setState({
      countdown: value.length
    });
    if (typeof onInput === 'function') onInput(event);
  }
  onChangeHandler(event) {
    const onChange = this.props.onChange;
    if (typeof onChange === 'function') onChange(event);
  }
  onClearHandler() {
    const onChange = this.props.onChange;
    const field = this.field,
      isUncontrolledValue = this.isUncontrolledValue,
      onInputHandler = this.onInputHandler,
      onChangeHandler = this.onChangeHandler;
    if (isUncontrolledValue) field.value = '';
    onInputHandler(emptyEvent);
    if (onChange) onChangeHandler(emptyEvent);
  }
  setField(field) {
    this.field = field;
    this.setState({
      fieldRef: field
    });
  }
  setRef(ref) {
    const innerRef = this.props.innerRef;
    if (typeof innerRef === 'function') {
      innerRef(ref);
      return;
    }
    if (innerRef) innerRef.current = ref;
  }
  renderChildren() {
    const _this$props2 = this.props,
      children = _this$props2.children,
      _this$props2$embedded = _this$props2.embeddedChildren,
      embeddedChildren = _this$props2$embedded === void 0 ? true : _this$props2$embedded;
    if (embeddedChildren) {
      return /*#__PURE__*/React__default["default"].createElement("div", {
        className: `${formControlNamespace}__embedded`,
        "data-andes-textfield-right-content": true
      }, children);
    }
    return children;
  }
  renderHelper() {
    const _this$props3 = this.props,
      helper = _this$props3.helper,
      id = _this$props3.id,
      modifier = _this$props3.modifier;
    return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, helper && /*#__PURE__*/React__default["default"].createElement(form.Helper, {
      id: `${id}-message`,
      modifier: computeModifier(modifier),
      className: `${formControlNamespace}--internal-helper`,
      andesDataAttributes: {
        'data-andes-textfield-helper': true
      }
    }, helper));
  }
  render() {
    const _this$props4 = this.props,
      _this$props4$centered = _this$props4.centered,
      centered = _this$props4$centered === void 0 ? false : _this$props4$centered,
      children = _this$props4.children,
      className = _this$props4.className,
      _this$props4$clear = _this$props4.clear,
      clear = _this$props4$clear === void 0 ? false : _this$props4$clear,
      _this$props4$countdow = _this$props4.countdown,
      countdown = _this$props4$countdow === void 0 ? false : _this$props4$countdow,
      _this$props4$disabled = _this$props4.disabled,
      disabled = _this$props4$disabled === void 0 ? false : _this$props4$disabled,
      defaultValue = _this$props4.defaultValue;
    _this$props4.embeddedChildren;
    const focus$1 = _this$props4.focus,
      helper = _this$props4.helper,
      id = _this$props4.id;
    _this$props4.innerRef;
    const i18n = _this$props4.i18n,
      _this$props4$inputPro = _this$props4.inputProps,
      inputProps = _this$props4$inputPro === void 0 ? {} : _this$props4$inputPro,
      label = _this$props4.label,
      _this$props4$maxLengt = _this$props4.maxLength,
      maxLength = _this$props4$maxLengt === void 0 ? 120 : _this$props4$maxLengt,
      modifier = _this$props4.modifier,
      _this$props4$multilin = _this$props4.multiline,
      multiline = _this$props4$multilin === void 0 ? false : _this$props4$multilin,
      _this$props4$numeric = _this$props4.numeric,
      numeric = _this$props4$numeric === void 0 ? false : _this$props4$numeric;
    _this$props4.onChange;
    const placeholder = _this$props4.placeholder,
      _this$props4$prefix = _this$props4.prefix,
      prefix = _this$props4$prefix === void 0 ? '' : _this$props4$prefix,
      _this$props4$rows = _this$props4.rows,
      rows = _this$props4$rows === void 0 ? 1 : _this$props4$rows;
    _this$props4.select;
    const _this$props4$size = _this$props4.size,
      size = _this$props4$size === void 0 ? 'large' : _this$props4$size,
      _this$props4$split = _this$props4.split,
      split = _this$props4$split === void 0 ? false : _this$props4$split,
      srLabel = _this$props4.srLabel,
      _this$props4$suffix = _this$props4.suffix,
      suffix = _this$props4$suffix === void 0 ? '' : _this$props4$suffix,
      value = _this$props4.value,
      autoFocus = _this$props4.autoFocus,
      andesDataAttributes = _this$props4.andesDataAttributes,
      otherProps = _objectWithoutProperties__default["default"](_this$props4, _excluded$2);
    const _this$state = this.state,
      focused = _this$state.focused,
      height = _this$state.height,
      count = _this$state.countdown;
    const showBottomInfo = !numeric && !split && modifier !== 'readonly' && !disabled && (countdown || helper);
    const classes = classNames__default["default"](formControlNamespace, `${formControlNamespace}--textfield`, className, {
      [`${formControlNamespace}--multiline`]: multiline,
      [`${formControlNamespace}--countdown`]: countdown,
      [`${formControlNamespace}--centered`]: centered,
      [`${formControlNamespace}--focused`]: focused,
      [`${formControlNamespace}--${modifier}`]: modifier,
      [`${formControlNamespace}--disabled`]: disabled,
      [`${formControlNamespace}--medium`]: size === 'medium',
      [`${formControlNamespace}--textfield-with-prefix`]: !!prefix,
      [`${formControlNamespace}--textfield-with-suffix`]: !!suffix
    });
    const getAriaDescribedby = () => {
      const ariaDescribedby = inputProps['aria-describedby'];
      if (split) return ariaDescribedby;
      if (helper && countdown) return `${id}-message ${id}-count`;
      if (countdown) return `${id}-count`;
      if (helper) return `${id}-message`;
    };
    const suffixClasses = classNames__default["default"](`${formControlNamespace}__suffix`);
    const prefixClasses = classNames__default["default"](`${formControlNamespace}__prefix`);
    const fieldClasses = classNames__default["default"](`${formControlNamespace}__field`, {
      [`${formControlNamespace}__field--multiline`]: multiline,
      [`${formControlNamespace}--medium`]: size === 'medium'
    });
    const bottomInfoClasses = classNames__default["default"](`${formControlNamespace}__bottom`, {
      [`${formControlNamespace}--medium`]: size === 'medium'
    });
    const fieldProps = _objectSpread__default["default"](_objectSpread__default["default"](_objectSpread__default["default"]({}, otherProps), inputProps), {}, {
      id,
      defaultValue,
      value,
      className: fieldClasses,
      disabled: disabled || modifier === 'readonly',
      maxLength: maxLength,
      style: multiline ? {
        height
      } : {},
      placeholder,
      onBlur: this.onBlurHandler,
      onFocus: this.onFocusHandler,
      onChange: this.onChangeHandler,
      ref: this.setField,
      'aria-invalid': modifier === 'error' || modifier === 'warning' || modifier === 'caution' || (inputProps === null || inputProps === void 0 ? void 0 : inputProps['aria-invalid']) || otherProps['aria-invalid'],
      'aria-label': srLabel,
      'aria-describedby': getAriaDescribedby(),
      rows: rows,
      autoFocus: focus$1 || autoFocus
    });
    const Input = multiline ? 'textarea' : 'input';
    return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
      className: classes,
      ref: this.setRef
    }, andesDataAttributes, {
      "data-andes-textfield": true,
      "data-andes-textfield-modifier": modifier
    }, computeDataAndesState.computeDataAndesState({
      disabled,
      error: modifier === 'error',
      warning: modifier === 'warning'
    })), label && /*#__PURE__*/React__default["default"].createElement("label", {
      htmlFor: id,
      "data-andes-textfield-label": true
    }, /*#__PURE__*/React__default["default"].createElement("span", {
      className: `${formControlNamespace}__label`
    }, label)), /*#__PURE__*/React__default["default"].createElement(focus.FocusRing, {
      within: true,
      isTextInput: true,
      focusRingClass: focused ? 'visible-focused' : ''
    }, /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${formControlNamespace}__control`
    }, prefix && /*#__PURE__*/React__default["default"].createElement("span", {
      className: prefixClasses,
      "data-andes-textfield-prefix": true
    }, prefix), /*#__PURE__*/React__default["default"].createElement(Input, _extends__default["default"]({
      key: "field"
    }, fieldProps, {
      onInput: this.onInputHandler,
      "data-andes-textfield-input": true,
      "data-andes-textfield-input-type": Input
    })), suffix && /*#__PURE__*/React__default["default"].createElement("span", {
      className: suffixClasses,
      "data-andes-textfield-suffix": true
    }, suffix), modifier === 'completed' && /*#__PURE__*/React__default["default"].createElement(ValidatedIcon$1, {
      notificationsLabel: i18n.pgettext(namespace$3, 'Campo de texto valido'),
      size: size
    }), modifier === 'indeterminate' && /*#__PURE__*/React__default["default"].createElement(progressIndicatorCircular.ProgressIndicatorCircular, {
      className: `${formControlNamespace}--textfield-progress`,
      size: "xsmall",
      srAnnouncement: i18n.pgettext(namespace$3, 'Campo de texto procesando'),
      srLabel: i18n.pgettext(namespace$3, 'Campo de texto procesando')
    }), clear && /*#__PURE__*/React__default["default"].createElement(ClearButton$1, {
      "aria-label": i18n.pgettext(namespace$3, 'Borrar'),
      namespace: formControlNamespace,
      onClick: this.onClearHandler,
      size: size
    }), children && this.renderChildren())), showBottomInfo && /*#__PURE__*/React__default["default"].createElement("div", {
      className: bottomInfoClasses,
      "data-andes-textfield-bottom-info": true
    }, this.renderHelper(), countdown && /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement("span", {
      id: `${id}-countdown-progress`,
      className: `${formControlNamespace}__countdown`,
      role: "progressbar",
      "aria-valuenow": count,
      "aria-valuemin": 0,
      "aria-valuemax": Number(maxLength),
      "aria-label": i18n.pgettext(namespace$3, 'Maximo {0} caracteres', maxLength),
      "data-andes-textfield-countdown": true
    }, `${count} / ${maxLength}`), /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, {
      id: `${id}-count`,
      "aria-live": "polite",
      "aria-atomic": "true"
    }, i18n.pgettext(namespace$3, '{0} caracteres restantes', Number(maxLength) - count)))));
  }
}
TextField.displayName = 'TextField';
var TextField$1 = TextField;
const _excluded$1 = ["ref"],
  _excluded2 = ["autoFocus", "children", "className", "defaultSelection", "defaultValue", "disabled", "helper", "inputProps", "id", "label", "list", "modifier", "placeholder", "onBlur", "onChange", "onDropdownOpen", "onFocus", "splitIcon", "splitLabel", "splitPosition", "srLabel", "value"];
const SplitTextfield = /*#__PURE__*/function () {
  const SplitTextfield = _ref => {
    let ref = _ref.ref,
      props = _objectWithoutProperties__default["default"](_ref, _excluded$1);
    const _props$autoFocus = props.autoFocus,
      autoFocus = _props$autoFocus === void 0 ? false : _props$autoFocus,
      children = props.children,
      className = props.className,
      _props$defaultSelecti = props.defaultSelection,
      defaultSelection = _props$defaultSelecti === void 0 ? 0 : _props$defaultSelecti,
      defaultValue = props.defaultValue,
      _props$disabled = props.disabled,
      disabled = _props$disabled === void 0 ? false : _props$disabled,
      helper = props.helper,
      inputProps = props.inputProps,
      id = props.id,
      label = props.label,
      list = props.list,
      modifier = props.modifier,
      placeholder = props.placeholder,
      onBlur = props.onBlur,
      onChange = props.onChange,
      onDropdownOpen = props.onDropdownOpen,
      onFocus = props.onFocus,
      splitIcon = props.splitIcon,
      splitLabel = props.splitLabel,
      _props$splitPosition = props.splitPosition,
      splitPosition = _props$splitPosition === void 0 ? 'left' : _props$splitPosition,
      srLabel = props.srLabel,
      value = props.value,
      other = _objectWithoutProperties__default["default"](props, _excluded2);
    const getDefaultValue = () => {
      var _list$content$default;
      return list === null || list === void 0 || (_list$content$default = list.content[defaultSelection]) === null || _list$content$default === void 0 ? void 0 : _list$content$default.title;
    };
    const ownContainerRef = React.useRef(null);
    const containerRef = ref || ownContainerRef;
    const fieldRef = React.useRef(undefined);
    const componentId = context.useId(id);
    const _useFocusRing = focus.useFocusRing({
        isTextInput: true
      }),
      isFocusVisible = _useFocusRing.isFocusVisible,
      focusProps = _useFocusRing.focusProps;
    const isReadonly = modifier === 'readonly';
    const _useThemeContext = themes.useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const _useAndesContext = context.useAndesContext('@andes/textfield', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations,
        'zh-CN': zhCnTranslations
      }),
      i18n = _useAndesContext.i18n;
    const containerClassNames = classNames__default["default"](formControlNamespace, className, `${formControlNamespace}--textfield`, `${formControlNamespace}--split`, {
      [`${formControlNamespace}--split__embedded`]: children,
      [`${formControlNamespace}--${modifier}`]: modifier,
      [`${formControlNamespace}--focused`]: autoFocus,
      [`${formControlNamespace}--disabled`]: disabled
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('textfield'));
    const splitControlClassNames = classNames__default["default"](`${formControlNamespace}__control`, {
      [`${formControlNamespace}__control--visible-focused`]: isFocusVisible
    });
    const splitButtonClassNames = classNames__default["default"](splitButtonNamespace, {
      [`${splitButtonNamespace}--right`]: splitPosition === 'right',
      [`${splitButtonNamespace}--readonly`]: isReadonly
    });
    const fieldClassNames = classNames__default["default"](`${formControlNamespace}__field ${formControlNamespace}__split-field`, {
      [`${formControlNamespace}__split-field--readonly`]: isReadonly
    });
    const iconClassNames = classNames__default["default"](`${splitButtonNamespace}-content--icon`, {
      [`${splitButtonNamespace}-content--icon-disabled`]: disabled
    });
    const splitItemClassNames = `${formControlNamespace}--split-item`;
    const customOnChange = (_, itemValue) => {
      if (!list) return;
      const content = list.content,
        onClick = list.onClick;
      if (typeof onClick === 'function' && itemValue) {
        onClick(content.find(_ref2 => {
          let title = _ref2.title;
          return title === itemValue;
        }));
      }
    };
    const customOnBlur = event => {
      var _fieldRef$current;
      handlers.onBlur(containerRef.current, (_fieldRef$current = fieldRef.current) === null || _fieldRef$current === void 0 ? void 0 : _fieldRef$current.field.value);
      if (typeof focusProps.onBlur === 'function') focusProps.onBlur(event);
      if (typeof onBlur === 'function') onBlur(event);
    };
    const customOnFocus = event => {
      handlers.onFocus(containerRef.current);
      if (typeof focusProps.onFocus === 'function') focusProps.onFocus(event);
      if (typeof onFocus === 'function') onFocus(event);
    };
    const renderCustomTrigger = triggerContent => {
      let adaptedTriggerContent = triggerContent;
      if (splitLabel) adaptedTriggerContent = splitLabel;
      if (splitIcon) {
        adaptedTriggerContent = /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement("span", {
          className: iconClassNames
        }, splitIcon), adaptedTriggerContent);
      }
      return adaptedTriggerContent;
    };
    const fieldProps = _objectSpread__default["default"](_objectSpread__default["default"]({}, other), {}, {
      inputProps: _objectSpread__default["default"]({
        'aria-describedby': helper ? `${componentId}-message` : undefined
      }, inputProps),
      'aria-invalid': modifier === 'error',
      'aria-labelledby': `${componentId}-dropdown-trigger`,
      srLabel: srLabel || label,
      className: fieldClassNames,
      children,
      defaultValue,
      disabled,
      id: componentId,
      modifier: modifier !== 'error' ? modifier : undefined,
      onBlur: customOnBlur,
      onChange,
      onFocus: customOnFocus,
      placeholder,
      ref: fieldRef,
      value,
      i18n
    });
    return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
      className: containerClassNames,
      ref: containerRef
    }, themeHash, {
      "data-andes-split-textfield": true
    }), label && /*#__PURE__*/React__default["default"].createElement("label", {
      htmlFor: componentId
    }, /*#__PURE__*/React__default["default"].createElement("span", {
      className: `${formControlNamespace}__label`
    }, label)), /*#__PURE__*/React__default["default"].createElement("div", {
      className: splitControlClassNames
    }, splitPosition === 'right' && /*#__PURE__*/React__default["default"].createElement(TextField$1, _extends__default["default"]({}, fieldProps, {
      split: true
    })), /*#__PURE__*/React__default["default"].createElement(dropdown.DropdownStandalone, {
      className: splitButtonClassNames,
      label: srLabel || label,
      menuAlignment: "bottom",
      defaultValue: getDefaultValue(),
      onOpen: onDropdownOpen,
      disabled: disabled || isReadonly,
      id: `${componentId}-dropdown`,
      renderCustomTrigger: renderCustomTrigger,
      onChange: customOnChange
    }, list === null || list === void 0 ? void 0 : list.content.map(item => /*#__PURE__*/
    React__default["default"].createElement(dropdown.DropdownItem, {
      value: item.title,
      className: splitItemClassNames,
      key: transformToKebabCase__default["default"](item.title),
      title: item.title,
      leftContent: item.icon && /*#__PURE__*/React__default["default"].createElement(item.icon, null)
    }))), splitPosition === 'left' && /*#__PURE__*/React__default["default"].createElement(TextField$1, _extends__default["default"]({}, fieldProps, {
      split: true
    }))), helper && !disabled && !isReadonly && /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${formControlNamespace}__bottom`
    }, /*#__PURE__*/React__default["default"].createElement(form.Helper, {
      id: `${componentId}-message`,
      modifier: computeModifier(modifier),
      andesDataAttributes: {
        'data-andes-split-textfield-helper': true
      }
    }, helper)));
  };
  SplitTextfield.displayName = 'SplitTextfield';
  return SplitTextfield;
}();
var SplitTextfield$1 = SplitTextfield;
const _excluded = ["ref", "id", "className"];
const TextfieldContainer = /*#__PURE__*/function () {
  const TextfieldContainer = _ref => {
    let ref = _ref.ref,
      id = _ref.id,
      className = _ref.className,
      otherProps = _objectWithoutProperties__default["default"](_ref, _excluded);
    const componentId = context.useId(id);
    const _useAndesContext = context.useAndesContext('@andes/textfield', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations,
        'zh-CN': zhCnTranslations
      }),
      i18n = _useAndesContext.i18n;
    const _useThemeContext = themes.useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    return /*#__PURE__*/React__default["default"].createElement(TextField$1, _extends__default["default"]({
      i18n: i18n,
      innerRef: ref
    }, otherProps, {
      id: componentId,
      className: classNames__default["default"](className, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('textfield')),
      andesDataAttributes: themeHash
    }));
  };
  TextfieldContainer.displayName = 'TextfieldContainer';
  return TextfieldContainer;
}();
var TextfieldContainer$1 = TextfieldContainer;
exports.CodeInput = CodeInput$1;
exports.Password = PasswordContainer$1;
exports.SplitTextfield = SplitTextfield$1;
exports.TextField = TextfieldContainer$1;
