// @nordic/rendering-context/dist/esm/getters-Cg82FgpK.js
import { AsyncLocalStorage } from 'node:async_hooks';

const storage = new AsyncLocalStorage();
function getContext() {
    const context = storage.getStore();
    {
        return context;
    }
}

function getRenderingAssets() {
    return getContext().assets;
}
function getRenderingAppProps() {
    return getContext().appProps;
}
function getRenderingFlags() {
    return getContext().flags;
}
function getRenderingSupport() {
    return getContext().support;
}
function getRenderingTitle() {
    return getContext().title;
}
function getRenderingMainEntry() {
    return getContext().mainEntry;
}
function getRenderingNonce() {
    return getContext().nonce;
}
function getRenderingCsrfToken() {
    return getContext().csrfToken;
}
function getRenderingWrapperProps() {
    return getContext().wrapperProps;
}
function getRenderingFramework() {
    return getContext().framework;
}

export { getRenderingAppProps as a, getRenderingAssets as b, getRenderingCsrfToken as c, getRenderingFlags as d, getRenderingFramework as e, getRenderingMainEntry as f, getContext as g, getRenderingNonce as h, getRenderingSupport as i, getRenderingTitle as j, getRenderingWrapperProps as k, storage as s };
