// @nordic/rendering-context/dist/esm/getters-Dj-Oy54B.js
import { AsyncLocalStorage } from 'node:async_hooks';
import { getTestContextScope } from '@nordic/core-context';
import { isTestEnv } from '@nordic/utils/validation';
import { OwnerStackError } from '@nordic/utils/error';

class NordicRenderingContextError extends OwnerStackError {
    name = 'NordicRenderingContextError';
}
function throwContextNotInitializedError() {
    throw new NordicRenderingContextError('Nordic Rendering Context is not initialized.');
}

const storage = new AsyncLocalStorage();
function getContext() {
    const context = storage.getStore();
    if (isTestEnv()) {
        return getTestContextScope().r;
    }
    if (!context) {
        throwContextNotInitializedError();
    }
    return context;
}

function getRenderingAssets() {
    return getContext().assets;
}
function getRenderingAppProps() {
    return getContext().appProps;
}
function getRenderingFlags() {
    return getContext().flags;
}
function getRenderingSupport() {
    return getContext().support;
}
function getRenderingTitle() {
    return getContext().title;
}
function getRenderingMainEntry() {
    return getContext().mainEntry;
}
function getRenderingNonce() {
    return getContext().nonce;
}
function getRenderingCsrfToken() {
    return getContext().csrfToken;
}
function getRenderingWrapperProps() {
    return getContext().wrapperProps;
}
function getRenderingFramework() {
    return getContext().framework;
}

export { NordicRenderingContextError as N, getRenderingAppProps as a, getRenderingAssets as b, getRenderingCsrfToken as c, getRenderingFlags as d, getRenderingFramework as e, getRenderingMainEntry as f, getContext as g, getRenderingNonce as h, getRenderingSupport as i, getRenderingTitle as j, getRenderingWrapperProps as k, storage as s };
