// @nordic/rendering-context/dist/esm/index.browser.development.mjs
import 'core-js/modules/es.regexp.flags.js';
import { OwnerStackError } from '@nordic/utils/error';

class NordicRenderingContextError extends OwnerStackError {
  name = 'NordicRenderingContextError';
}
function throwContextNotInitializedError() {
  throw new NordicRenderingContextError('Nordic Rendering Context is not initialized.');
}

function getContext() {
  const context = window._n.ctx.r;
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- Can be undefined in case the render has not been executed into the rendering context
  if (!context) {
    throwContextNotInitializedError();
  }
  return context;
}

function getRenderingAssets() {
  return getContext().assets;
}
function getRenderingAppProps() {
  return getContext().appProps;
}
function getRenderingFlags() {
  return getContext().flags;
}
function getRenderingSupport() {
  return getContext().support;
}
function getRenderingTitle() {
  return getContext().title;
}
function getRenderingMainEntry() {
  return getContext().mainEntry;
}
function getRenderingNonce() {
  return getContext().nonce;
}
function getRenderingCsrfToken() {
  return getContext().csrfToken;
}
function getRenderingWrapperProps() {
  return getContext().wrapperProps;
}
function getRenderingFramework() {
  return getContext().framework;
}

function setRenderingAppProps(appProps) {
  getContext().appProps = appProps;
}

export { getRenderingAppProps, getRenderingAssets, getRenderingCsrfToken, getRenderingFlags, getRenderingFramework, getRenderingMainEntry, getRenderingNonce, getRenderingSupport, getRenderingTitle, getRenderingWrapperProps, setRenderingAppProps };
