// @nordic/rendering-context/dist/esm/index.browser.production.mjs
import 'core-js/modules/es.regexp.flags.js';

function getContext() {
  const context = window._n.ctx.r;
  return context;
}

function getRenderingAssets() {
  return getContext().assets;
}
function getRenderingAppProps() {
  return getContext().appProps;
}
function getRenderingFlags() {
  return getContext().flags;
}
function getRenderingSupport() {
  return getContext().support;
}
function getRenderingTitle() {
  return getContext().title;
}
function getRenderingMainEntry() {
  return getContext().mainEntry;
}
function getRenderingNonce() {
  return getContext().nonce;
}
function getRenderingCsrfToken() {
  return getContext().csrfToken;
}
function getRenderingWrapperProps() {
  return getContext().wrapperProps;
}
function getRenderingFramework() {
  return getContext().framework;
}

function setRenderingAppProps(appProps) {
  getContext().appProps = appProps;
}

export { getRenderingAppProps, getRenderingAssets, getRenderingCsrfToken, getRenderingFlags, getRenderingFramework, getRenderingMainEntry, getRenderingNonce, getRenderingSupport, getRenderingTitle, getRenderingWrapperProps, setRenderingAppProps };
