// @andes/common/build/esm/utils/webviewTopbarHandler.production.js
const isValidWebkitEnvironment = () => typeof MobileWebKit !== 'undefined' && typeof MobileWebKit.executeNative === 'function' && typeof MobileWebKit.events !== 'undefined' && typeof MobileWebKit.events.subscribeMethod === 'function' && typeof MobileWebKit.events.unsubscribe === 'function';
const isWebkit2 = () => {
  var _MobileWebKit, _MobileWebKit2;
  return isValidWebkitEnvironment() && ((_MobileWebKit = MobileWebKit) === null || _MobileWebKit === void 0 ? void 0 : _MobileWebKit.engine) === ((_MobileWebKit2 = MobileWebKit) === null || _MobileWebKit2 === void 0 || (_MobileWebKit2 = _MobileWebKit2.Engines) === null || _MobileWebKit2 === void 0 ? void 0 : _MobileWebKit2.V200);
};
const disableWebviewTopbar = _ref => {
  var _MobileWebKit3, _MobileWebKit4;
  let onBack = _ref.onBack,
    _ref$opacity = _ref.opacity,
    opacity = _ref$opacity === void 0 ? 0.8 : _ref$opacity,
    _ref$animated = _ref.animated,
    animated = _ref$animated === void 0 ? true : _ref$animated,
    _ref$duration = _ref.duration,
    duration = _ref$duration === void 0 ? 200 : _ref$duration;
  if (!isWebkit2()) return;
  (_MobileWebKit3 = MobileWebKit) === null || _MobileWebKit3 === void 0 || _MobileWebKit3.topbar.disable({
    opacity,
    animated,
    duration
  }).catch(() => {});
  const backActionSubscriptionId = (_MobileWebKit4 = MobileWebKit) === null || _MobileWebKit4 === void 0 ? void 0 : _MobileWebKit4.navigation.subscribeToBackAction(onBack);
  return {
    backActionSubscriptionId
  };
};
const enableWebviewTopbar = function () {
  var _MobileWebKit5;
  let _ref2 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
    _ref2$animated = _ref2.animated,
    animated = _ref2$animated === void 0 ? true : _ref2$animated,
    _ref2$duration = _ref2.duration,
    duration = _ref2$duration === void 0 ? 200 : _ref2$duration,
    subscriptionIds = _ref2.subscriptionIds;
  if (!isWebkit2()) return;
  (_MobileWebKit5 = MobileWebKit) === null || _MobileWebKit5 === void 0 || _MobileWebKit5.topbar.enable({
    animated,
    duration
  }).catch(() => {});
  if (subscriptionIds !== null && subscriptionIds !== void 0 && subscriptionIds.backActionSubscriptionId) {
    var _MobileWebKit6;
    (_MobileWebKit6 = MobileWebKit) === null || _MobileWebKit6 === void 0 || _MobileWebKit6.navigation.unsubscribeToBackAction(subscriptionIds.backActionSubscriptionId).catch(() => {});
  }
};
const hideWebviewTopbar = _ref3 => {
  var _MobileWebKit7, _MobileWebKit9;
  let onBack = _ref3.onBack,
    _ref3$animated = _ref3.animated,
    animated = _ref3$animated === void 0 ? true : _ref3$animated,
    _ref3$duration = _ref3.duration,
    animation_duration = _ref3$duration === void 0 ? 350 : _ref3$duration;
  if (!isWebkit2()) return;
  if (typeof ((_MobileWebKit7 = MobileWebKit) === null || _MobileWebKit7 === void 0 || (_MobileWebKit7 = _MobileWebKit7.topbar) === null || _MobileWebKit7 === void 0 ? void 0 : _MobileWebKit7.fullScreenMode) === 'function') {
    MobileWebKit.topbar.fullScreenMode(true, {
      animated,
      animation_duration
    }).catch(() => {});
  } else {
    var _MobileWebKit8;
    (_MobileWebKit8 = MobileWebKit) === null || _MobileWebKit8 === void 0 || _MobileWebKit8.topbar.disable({
      opacity: 0.8,
      animated,
      duration: 200
    }).catch(() => {});
  }
  const backActionSubscriptionId = (_MobileWebKit9 = MobileWebKit) === null || _MobileWebKit9 === void 0 ? void 0 : _MobileWebKit9.navigation.subscribeToBackAction(onBack);
  return {
    backActionSubscriptionId
  };
};
const showWebviewTopbar = function () {
  var _MobileWebKit0;
  let _ref4 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
    _ref4$animated = _ref4.animated,
    animated = _ref4$animated === void 0 ? true : _ref4$animated,
    _ref4$duration = _ref4.duration,
    animation_duration = _ref4$duration === void 0 ? 350 : _ref4$duration,
    subscriptionIds = _ref4.subscriptionIds;
  if (!isWebkit2()) return;
  if (typeof ((_MobileWebKit0 = MobileWebKit) === null || _MobileWebKit0 === void 0 || (_MobileWebKit0 = _MobileWebKit0.topbar) === null || _MobileWebKit0 === void 0 ? void 0 : _MobileWebKit0.fullScreenMode) === 'function') {
    MobileWebKit.topbar.fullScreenMode(false, {
      animated,
      animation_duration
    }).catch(() => {});
  } else {
    var _MobileWebKit1;
    (_MobileWebKit1 = MobileWebKit) === null || _MobileWebKit1 === void 0 || _MobileWebKit1.topbar.enable({
      animated,
      duration: 200
    }).catch(() => {});
  }
  if (subscriptionIds !== null && subscriptionIds !== void 0 && subscriptionIds.backActionSubscriptionId) {
    var _MobileWebKit10;
    (_MobileWebKit10 = MobileWebKit) === null || _MobileWebKit10 === void 0 || _MobileWebKit10.navigation.unsubscribeToBackAction(subscriptionIds.backActionSubscriptionId).catch(() => {});
  }
};
export { disableWebviewTopbar, enableWebviewTopbar, hideWebviewTopbar, showWebviewTopbar };
