// node-melitk-otel/lib/events/config/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadConfigs = exports.metricDefaultAttributes = void 0;
const frontend_env_1 = __importDefault(require("frontend-env"));
const package_json_1 = __importDefault(require("../../../package.json"));
/**
 * Gets the value of the environment variable or the default value if it is not defined.
 * @param key The key of the environment variable.
 * @param defaultValue The default value to return if the environment variable is not defined.
 * @returns The value of the environment variable or the default value if it is not defined.
 */
const getEnvOrDefault = (key, defaultValue) => {
    const value = frontend_env_1.default[key];
    if (value === undefined) {
        return defaultValue;
    }
    return value;
};
exports.metricDefaultAttributes = {
    technology: 'nodejs',
    sdk_version: package_json_1.default.version,
};
/**
 * Gets the value of the environment variable or 'unknown' if it is not defined.
 * @param key The key of the environment variable.
 * @returns The value of the environment variable or 'unknown' if it is not defined.
 */
const getEnvOrUnknown = (key) => getEnvOrDefault(key, 'unknown');
const OTEL_EVENTS_COLLECTOR_DEFAULT_ENDPOINT = getEnvOrDefault('O11Y_EVENTS_OTEL_EXPORTER_OTLP_ENDPOINT', 'https://otel-events-collector.furycloud.io:4317');
const ACTIVATOR_URL_BY_SCOPE_CRITICALITY = {
    test: 'http://flow-discovery-test.furycloud.io',
    low: 'http://flow-discovery-low.furycloud.io',
    medium: 'http://flow-discovery-medium.furycloud.io',
    high: 'http://flow-discovery-high.furycloud.io',
    'bu-critical': 'http://flow-discovery-bu-critical.furycloud.io',
    'company-critical': 'http://flow-discovery-company-critical.furycloud.io',
    unknown: 'http://flow-discovery.furycloud.io',
};
const EVENTS_ACTIVATOR_DEFAULT_SERVICES_ENDPOINT = getEnvOrDefault('O11Y_EVENTS_ACTIVATOR_SERVICES_ENDPOINT', '/v1/activator');
/**
 * Loads the configs for the o11y events.
 * @returns {O11YDefaults} The default configs for the o11y events.
 */
const loadConfigs = () => {
    const configs = {
        o11yEvents: {
            enabled: getEnvOrDefault('O11Y_EVENTS_ENABLED', true),
            debugMode: getEnvOrDefault('O11Y_EVENTS_DEBUG_MODE', false),
            otel: {
                exporter: {
                    otlp: {
                        endpoint: OTEL_EVENTS_COLLECTOR_DEFAULT_ENDPOINT,
                        exportTimeout: getEnvOrDefault('O11Y_EVENTS_OTEL_EXPORTER_OTLP_EXPORT_TIMEOUT', 1000),
                        retry: {
                            enabled: getEnvOrDefault('O11Y_EVENTS_OTEL_EXPORTER_OTLP_RETRY_ENABLED', true),
                            initialInterval: getEnvOrDefault('O11Y_EVENTS_OTEL_EXPORTER_OTLP_RETRY_INITIAL_INTERVAL', 400),
                            maxInterval: getEnvOrDefault('O11Y_EVENTS_OTEL_EXPORTER_OTLP_RETRY_MAX_INTERVAL', 900),
                            maxElapsedTime: getEnvOrDefault('O11Y_EVENTS_OTEL_EXPORTER_OTLP_RETRY_MAX_ELAPSED_TIME', 1000),
                        },
                    },
                },
                processor: {
                    maxQueueSize: getEnvOrDefault('O11Y_EVENTS_OTEL_PROCESSOR_MAX_QUEUE_SIZE', 1024),
                    scheduleDelay: getEnvOrDefault('O11Y_EVENTS_OTEL_PROCESSOR_SCHEDULED_DELAY', 5000),
                    exportTimeout: getEnvOrDefault('O11Y_EVENTS_OTEL_PROCESSOR_EXPORT_TIMEOUT', 5000),
                    maxExportBatchSize: getEnvOrDefault('O11Y_EVENTS_OTEL_PROCESSOR_MAX_EXPORT_BATCH_SIZE', 512),
                    blockOnQueueFull: getEnvOrDefault('O11Y_EVENTS_OTEL_PROCESSOR_BLOCK_ON_QUEUE_FULL', false),
                },
                attributes: {
                    serviceName: `${getEnvOrUnknown('SERVICE')}.${getEnvOrUnknown('APPLICATION')}`,
                    serviceVersion: getEnvOrUnknown('VERSION'),
                    furyApplication: getEnvOrUnknown('APPLICATION'),
                    furyService: getEnvOrUnknown('SERVICE'),
                    furyCriticality: getEnvOrUnknown('SCOPE_CRITICALITY'),
                },
            },
            activator: {
                period: getEnvOrDefault('O11Y_EVENTS_ACTIVATOR_PERIOD', 5 * 60 * 1000), // 5 minutes
                timeout: getEnvOrDefault('O11Y_EVENTS_ACTIVATOR_TIMEOUT', 5000),
                maxRetries: getEnvOrDefault('O11Y_EVENTS_ACTIVATOR_MAX_RETRIES', 3),
                url: ACTIVATOR_URL_BY_SCOPE_CRITICALITY[getEnvOrUnknown('SCOPE_CRITICALITY')],
                servicesEndpoint: EVENTS_ACTIVATOR_DEFAULT_SERVICES_ENDPOINT,
            },
        },
    };
    return configs;
};
exports.loadConfigs = loadConfigs;
