// @andes/button/build/esm/module.js
import _extends from '@babel/runtime/helpers/extends';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { useRef, useState, useEffect, cloneElement } from 'react';
import classNames from 'classnames';
import { ProgressIndicatorCircular } from '@andes/progress-indicator-circular';
import { VisuallyHidden } from '@andes/visually-hidden';
import warning from '@andes/common/build/utils/warning';
import { useId, useAndesContext } from '@andes/context';
import { useThemeContext } from '@andes/themes';
import { computeDataAndesState } from '@andes/common/build/utils/computeDataAndesState';
import { FloatingMenu, FloatingMenuItem } from '@andes/floating-menu';
import transformToKebabCase from '@andes/common/build/utils/transformToKebabCase';
const namespace$2 = 'andes-button';
const namespaceDropdownButton = 'andes-button-dropdown';
const WARNING_A11Y_LABELS = `<DropdownButton> needs some plain text contents for accessibility reasons.
Please add an 'srLabel' or 'children' prop.`;
const WARNING_A11Y_PROGRESS_LABEL = `<ProgressButton> needs to announce its loading progress for accessibility reasons.
Please use a 'progressLabel' or add 'progressSrAnnouncement' prop.`;
const WARNING_HIERARCHY_VALUE = `'hierarchy' with 'transparent' value will be deprecated in a future major version. 
Please, use 'mute' value on <Button> instead`;
const computeHierarchy = hierarchy => {
  if (hierarchy === 'transparent') {
    warning(false, WARNING_HIERARCHY_VALUE);
    return 'mute';
  }
  return hierarchy;
};
const _excluded$3 = ["children", "className", "disabled", "fullWidth", "hierarchy", "href", "id", "loading", "onClick", "size", "srAnnouncement", "srLabel", "type", "ref", "backgroundType", "data-andes-state"];
const progressIndicatorSize = {
  small: 'xsmall',
  medium: 'xsmall',
  large: 'small'
};
const ButtonText = _ref => {
  let children = _ref.children;
  return /*#__PURE__*/React.createElement("span", {
    className: `${namespace$2}__text`,
    "data-andes-button-text": true
  }, children);
};
const Button = /*#__PURE__*/function () {
  const Button = props => {
    const children = props.children,
      className = props.className,
      _props$disabled = props.disabled,
      disabled = _props$disabled === void 0 ? false : _props$disabled,
      _props$fullWidth = props.fullWidth,
      fullWidth = _props$fullWidth === void 0 ? false : _props$fullWidth,
      _props$hierarchy = props.hierarchy,
      hierarchy = _props$hierarchy === void 0 ? 'loud' : _props$hierarchy,
      href = props.href,
      id = props.id,
      loading = props.loading,
      onClick = props.onClick,
      _props$size = props.size,
      size = _props$size === void 0 ? 'large' : _props$size,
      srAnnouncement = props.srAnnouncement,
      srLabel = props.srLabel,
      _props$type = props.type,
      type = _props$type === void 0 ? 'button' : _props$type,
      ref = props.ref,
      _props$backgroundType = props.backgroundType,
      backgroundType = _props$backgroundType === void 0 ? 'default' : _props$backgroundType,
      dataAndesState = props['data-andes-state'],
      otherProps = _objectWithoutProperties(props, _excluded$3);
    const componentId = useId(id);
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const srLabelTimeout = useRef(undefined);
    const firstRender = useRef(true);
    const _useState = useState(''),
      _useState2 = _slicedToArray(_useState, 2),
      srAnnouncementLabel = _useState2[0],
      setSrAnnouncementLabel = _useState2[1];
    const _useState3 = useState(false),
      _useState4 = _slicedToArray(_useState3, 2),
      loadingComplete = _useState4[0],
      setLoadingComplete = _useState4[1];
    const computedHierarchy = computeHierarchy(hierarchy);
    useEffect(() => {
      if (!loading) {
        setSrAnnouncementLabel('');
        if (!firstRender.current) setLoadingComplete(true);
      }
      if (firstRender) firstRender.current = false;
    }, [loading]);
    useEffect(() => {
      let finishTimeout;
      if (loadingComplete) {
        finishTimeout = setTimeout(() => {
          setLoadingComplete(false);
        }, 200);
      }
      return () => clearTimeout(finishTimeout);
    }, [loadingComplete]);
    useEffect(() => () => {
      clearTimeout(srLabelTimeout.current);
    }, []);
    warning(!loading || loading && srAnnouncement, `<ButtonWithSpinner> needs some plain text contents for accessibility reasons.
    Please add a 'srAnnouncement' prop.`);
    const classes = classNames(namespace$2, className, {
      [`${namespace$2}--${size}`]: size,
      [`${namespace$2}--${computedHierarchy}`]: computedHierarchy,
      [`${namespace$2}--disabled`]: disabled,
      [`${namespace$2}--full-width`]: fullWidth,
      [`${namespace$2}--loading`]: loading,
      [`${namespace$2}--loading-complete`]: loadingComplete,
      [`${namespace$2}--color-on-background`]: backgroundType === 'complexBackground'
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('button'));
    const sharedProps = _objectSpread(_objectSpread(_objectSpread({
      className: classes,
      id: componentId,
      'aria-label': srLabel
    }, otherProps), {}, {
      'data-andes-button': true,
      'data-andes-button-hierarchy': computedHierarchy,
      'data-andes-button-size': size
    }, computeDataAndesState(_objectSpread(_objectSpread({}, dataAndesState ? {
      otherDataAndesState: dataAndesState
    } : {}), {}, {
      disabled,
      loading
    }))), themeHash);
    const content = /*#__PURE__*/React.createElement("span", {
      className: `${namespace$2}__content`,
      "data-andes-button-content": true
    }, children);
    if (href) {
      const anchorProps = _objectSpread(_objectSpread({}, disabled || loading ? {
        role: 'link',
        'aria-disabled': true
      } : {
        href
      }), {}, {
        ref: ref,
        onClick: onClick
      }, sharedProps);
      return /*#__PURE__*/React.createElement("a", anchorProps, content);
    }
    const buttonProps = _objectSpread({
      disabled: disabled || loading,
      ref: ref,
      type: type
    }, sharedProps);
    const customOnClick = event => {
      const label = srAnnouncement || '';
      srLabelTimeout.current = setTimeout(() => setSrAnnouncementLabel(label), 30);
      if (typeof onClick === 'function') onClick(event);
    };
    const progressSize = progressIndicatorSize[size];
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("button", _extends({}, buttonProps, {
      onClick: customOnClick
    }), content, loading && /*#__PURE__*/React.createElement(ProgressIndicatorCircular, {
      ariaLabelledBy: componentId,
      size: progressSize
    })), loading && /*#__PURE__*/React.createElement(VisuallyHidden, {
      "aria-live": "assertive"
    }, srAnnouncementLabel));
  };
  Button.displayName = 'Button';
  return Button;
}();
const _excluded$2 = ["className", "dialogSrLabel", "id", "listContent", "namespace", "onClickOutside", "placement", "popperOptions", "srLabel", "trigger", "visible"];
const ButtonListContent = _ref => {
  let className = _ref.className,
    dialogSrLabel = _ref.dialogSrLabel,
    id = _ref.id,
    listContent = _ref.listContent,
    namespace = _ref.namespace;
  _ref.onClickOutside;
  let placement = _ref.placement,
    popperOptions = _ref.popperOptions,
    srLabel = _ref.srLabel,
    trigger = _ref.trigger,
    visible = _ref.visible,
    props = _objectWithoutProperties(_ref, _excluded$2);
  return /*#__PURE__*/React.createElement(FloatingMenu, _extends({
    className: className,
    dialogSrLabel: dialogSrLabel,
    placement: placement,
    open: visible,
    popperOptions: popperOptions,
    trigger: trigger,
    id: id,
    srLabel: srLabel
  }, props), listContent.map((_ref2, i) => {
    let title = _ref2.title,
      description = _ref2.description,
      href = _ref2.href,
      onClick = _ref2.onClick,
      leftContent = _ref2.leftContent;
    return /*#__PURE__*/React.createElement(FloatingMenuItem, {
      id: `${namespace}__menu-item-${i}`,
      href: href,
      className: `${namespace}__menu-item`,
      key: title,
      value: title,
      title: title,
      description: description,
      onClick: onClick,
      leftContent: leftContent
    });
  }));
};
var ButtonListContent$1 = ButtonListContent;
var enTranslations = {
  "andes-button-dropdown\u0004menú desplegable": [null, "Dropdown menu"],
  "andes-button-split\u0004menú desplegable": [null, "Dropdown menu"]
};
var zhCnTranslations = {
  "andes-button-dropdown\u0004menú desplegable": [null, "下拉菜单"],
  "andes-button-split\u0004menú desplegable": [null, "下拉菜单"]
};
var esArTranslations = {
  "andes-button-dropdown\u0004menú desplegable": [null, "Menú desplegable"],
  "andes-button-split\u0004menú desplegable": [null, "Menú desplegable"]
};
var ptBrTranslations = {
  "andes-button-dropdown\u0004menú desplegable": [null, "Menu suspenso"],
  "andes-button-split\u0004menú desplegable": [null, "Menu suspenso"]
};
const ChevronDown$1 = () => /*#__PURE__*/React.createElement("svg", {
  width: "20px",
  height: "20px",
  viewBox: "0 0 20 20",
  version: "1.1",
  xmlns: "http://www.w3.org/2000/svg",
  className: `${namespaceDropdownButton}__chevron`
}, /*#__PURE__*/React.createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd",
  fillOpacity: "0.8"
}, /*#__PURE__*/React.createElement("polygon", {
  fill: "currentColor",
  fillRule: "nonzero",
  transform: "translate(10.000000, 10.000000) rotate(-90.000000) translate(-10.000000, -10.000000) ",
  points: "8.4482158 9.99750228 13.243845 5.20187305 12.3953169 4.35334491 6.75115953 9.99750228 12.4003123 15.6466551 13.2488405 14.798127"
})));
const ChevronUP$1 = () => /*#__PURE__*/React.createElement("svg", {
  width: "20px",
  height: "20px",
  viewBox: "0 0 20 20",
  version: "1.1",
  xmlns: "http://www.w3.org/2000/svg",
  className: `${namespaceDropdownButton}__chevron`
}, /*#__PURE__*/React.createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd",
  fillOpacity: "0.8"
}, /*#__PURE__*/React.createElement("polygon", {
  fill: "currentColor",
  fillRule: "nonzero",
  transform: "translate(10.001873, 9.995000) rotate(90.000000) translate(-10.001873, -9.995000) ",
  points: "8.45008885 9.99250228 13.2457181 5.19687305 12.3971899 4.34834491 6.75303257 9.99250228 12.4021854 15.6416551 13.2507135 14.793127"
})));
const DropdownButton = /*#__PURE__*/function () {
  const DropdownButton = _ref => {
    let appendTo = _ref.appendTo,
      backgroundType = _ref.backgroundType,
      className = _ref.className,
      children = _ref.children,
      dialogSrLabel = _ref.dialogSrLabel,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      _ref$fullWidth = _ref.fullWidth,
      fullWidth = _ref$fullWidth === void 0 ? false : _ref$fullWidth,
      _ref$hierarchy = _ref.hierarchy,
      hierarchy = _ref$hierarchy === void 0 ? 'loud' : _ref$hierarchy,
      id = _ref.id,
      listContent = _ref.listContent,
      menuWidth = _ref.menuWidth,
      _ref$side = _ref.side,
      side = _ref$side === void 0 ? 'bottom' : _ref$side,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? 'large' : _ref$size,
      srLabel = _ref.srLabel,
      ref = _ref.ref;
    const componentId = useId(id);
    const _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      showOptions = _useState2[0],
      setShowOptions = _useState2[1];
    warning(children || srLabel, WARNING_A11Y_LABELS);
    const _useAndesContext = useAndesContext('@andes/button', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations,
        'zh-CN': zhCnTranslations
      }),
      i18n = _useAndesContext.i18n;
    const handleOpen = () => {
      setShowOptions(true);
    };
    const handleClose = () => {
      setShowOptions(false);
    };
    const classes = classNames(namespaceDropdownButton, className);
    const listClasses = classNames(`${namespaceDropdownButton}__menu`, className, {
      [`${namespaceDropdownButton}__menu--${side}`]: side
    });
    const trigger = /*#__PURE__*/React.createElement(Button, {
      backgroundType: backgroundType,
      className: classes,
      hierarchy: hierarchy,
      size: size,
      "aria-expanded": showOptions,
      srLabel: srLabel,
      disabled: disabled,
      fullWidth: fullWidth,
      id: componentId,
      "data-andes-dropdown-button": true,
      "data-andes-state": showOptions ? 'expanded' : 'collapsed'
    }, children && /*#__PURE__*/React.createElement(ButtonText, null, children), !showOptions ? /*#__PURE__*/React.createElement(ChevronDown$1, null) : /*#__PURE__*/React.createElement(ChevronUP$1, null));
    return /*#__PURE__*/React.createElement("div", {
      ref: ref
    }, /*#__PURE__*/React.createElement(ButtonListContent$1, {
      appendTo: appendTo,
      dialogSrLabel: dialogSrLabel,
      placement: side,
      listContent: listContent,
      onOpen: handleOpen,
      onClose: handleClose,
      popperOptions: {
        modifiers: [{
          name: 'sameWidth',
          enabled: !menuWidth,
          fn: _ref2 => {
            let state = _ref2.state;
            state.styles.popper.width = `${state.rects.reference.width}px`;
          },
          phase: 'beforeWrite',
          requires: ['computeStyles'],
          effect: _ref3 => {
            let state = _ref3.state;
            state.elements.popper.style.width = `${state.elements.reference.clientWidth}px`;
          }
        }]
      },
      trigger: trigger,
      namespace: namespaceDropdownButton,
      id: `${componentId}-menu`,
      className: listClasses,
      srLabel: srLabel || i18n.pgettext(namespaceDropdownButton, 'menú desplegable'),
      width: menuWidth || '100%'
    }));
  };
  DropdownButton.displayName = 'DropdownButton';
  return DropdownButton;
}();
var DropdownButton$1 = DropdownButton;
const namespace$1 = 'andes-button-split';
const ChevronDown = () => /*#__PURE__*/React.createElement("svg", {
  width: "20px",
  height: "20px",
  viewBox: "0 0 20 20",
  version: "1.1",
  xmlns: "http://www.w3.org/2000/svg",
  className: `${namespace$1}__chevron`
}, /*#__PURE__*/React.createElement("title", null, "Abrir men\xFA de opciones / Mostrar opciones"), /*#__PURE__*/React.createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd",
  fillOpacity: "0.8"
}, /*#__PURE__*/React.createElement("polygon", {
  fill: "currentColor",
  fillRule: "nonzero",
  transform: "translate(10.000000, 10.000000) rotate(-90.000000) translate(-10.000000, -10.000000) ",
  points: "8.4482158 9.99750228 13.243845 5.20187305 12.3953169 4.35334491 6.75115953 9.99750228 12.4003123 15.6466551 13.2488405 14.798127"
})));
const ChevronUP = () => /*#__PURE__*/React.createElement("svg", {
  width: "20px",
  height: "20px",
  viewBox: "0 0 20 20",
  version: "1.1",
  xmlns: "http://www.w3.org/2000/svg",
  className: `${namespace$1}__chevron`
}, /*#__PURE__*/React.createElement("title", null, "Cerrar men\xFA de opciones / Ocultar opciones"), /*#__PURE__*/React.createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd",
  fillOpacity: "0.8"
}, /*#__PURE__*/React.createElement("polygon", {
  fill: "currentColor",
  fillRule: "nonzero",
  transform: "translate(10.001873, 9.995000) rotate(90.000000) translate(-10.001873, -9.995000) ",
  points: "8.45008885 9.99250228 13.2457181 5.19687305 12.3971899 4.34834491 6.75303257 9.99250228 12.4021854 15.6416551 13.2507135 14.793127"
})));
const placement = {
  top: 'top-start',
  bottom: 'bottom-start',
  bottomRight: 'bottom-start',
  topRight: 'top-start'
};
const SplitButton = /*#__PURE__*/function () {
  const SplitButton = _ref => {
    let _ref$backgroundType = _ref.backgroundType,
      backgroundType = _ref$backgroundType === void 0 ? 'default' : _ref$backgroundType,
      children = _ref.children,
      className = _ref.className,
      dialogSrLabel = _ref.dialogSrLabel,
      disabled = _ref.disabled,
      _ref$hierarchy = _ref.hierarchy,
      hierarchy = _ref$hierarchy === void 0 ? 'loud' : _ref$hierarchy,
      id = _ref.id,
      listContent = _ref.listContent,
      onClick = _ref.onClick,
      _ref$side = _ref.side,
      side = _ref$side === void 0 ? 'bottomRight' : _ref$side,
      srLabel = _ref.srLabel,
      ref = _ref.ref;
    const componentId = useId(id);
    const _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      showOptions = _useState2[0],
      setShowOptions = _useState2[1];
    const _useAndesContext = useAndesContext('@andes/button', {
        en: enTranslations,
        'es-AR': esArTranslations,
        'pt-BR': ptBrTranslations,
        'zh-CN': zhCnTranslations
      }),
      i18n = _useAndesContext.i18n;
    const show = () => {
      setShowOptions(true);
    };
    const hide = () => {
      setShowOptions(false);
    };
    const classes = classNames(namespace$1, className);
    const popperOptions = side === 'top' || side === 'bottom' ? {
      modifiers: [{
        name: 'sameWidth',
        enabled: true,
        fn: _ref2 => {
          let state = _ref2.state;
          state.styles.popper.left = `-${state.elements.reference.parentElement.offsetLeft}px`;
          state.styles.popper.width = `${state.elements.reference.parentElement.offsetLeft + state.rects.reference.width}px`;
        },
        phase: 'beforeWrite',
        requires: ['computeStyles'],
        effect: _ref3 => {
          let state = _ref3.state;
          state.elements.popper.style.width = `${state.elements.reference.parentElement.offsetLeft + state.elements.reference.clientWidth}px`;
        }
      }]
    } : undefined;
    const listClasses = classNames(`${namespace$1}__menu`, className, {
      [`${namespace$1}__menu--${transformToKebabCase(side)}`]: side
    });
    const trigger = () => /*#__PURE__*/React.createElement(Button, {
      backgroundType: backgroundType,
      className: `${namespace$1}-chevron`,
      disabled: disabled,
      hierarchy: hierarchy,
      "aria-expanded": showOptions,
      srLabel: srLabel || i18n.pgettext(namespace$1, 'menú desplegable'),
      id: `${componentId}-chevron`,
      "data-andes-split-button-trigger": true
    }, !showOptions ? /*#__PURE__*/React.createElement(ChevronDown, null) : /*#__PURE__*/React.createElement(ChevronUP, null));
    const buttonListContentPlacement = placement[`${side}`];
    return /*#__PURE__*/React.createElement("div", {
      className: classes,
      ref: ref,
      "data-andes-split-button": true,
      "data-andes-state": showOptions ? 'expanded' : 'collapsed'
    }, /*#__PURE__*/React.createElement(Button, {
      backgroundType: backgroundType,
      className: `${namespace$1}-text`,
      disabled: disabled,
      onClick: onClick,
      hierarchy: hierarchy,
      id: `${componentId}-text`
    }, /*#__PURE__*/React.createElement(ButtonText, null, children)), /*#__PURE__*/React.createElement(ButtonListContent$1, {
      placement: buttonListContentPlacement,
      dialogSrLabel: dialogSrLabel,
      listContent: listContent,
      visible: showOptions,
      onClickOutside: hide,
      onOpen: show,
      onClose: hide,
      popperOptions: popperOptions,
      trigger: trigger(),
      namespace: namespace$1,
      id: `${componentId}-menu`,
      srLabel: srLabel || i18n.pgettext(namespace$1, 'menú desplegable'),
      className: listClasses,
      width: side === 'top' || side === 'bottom' ? '100%' : 'max-content'
    }));
  };
  SplitButton.displayName = 'SplitButton';
  return SplitButton;
}();
var SplitButton$1 = SplitButton;
const _excluded$1 = ["children", "className", "disabled", "hierarchy", "id", "loading", "progressLabel", "progressSrAnnouncement", "timeout", "ref"];
const STATES = {
  LOADING: 'loading',
  ENTERED: 'entered',
  FINISHED: 'finished'
};
const DEFAULT_HIERARCHY$1 = 'loud';
const HIERARCHY_ALLOWED = {
  [DEFAULT_HIERARCHY$1]: true,
  quiet: true
};
const buildClassesObject = function (stages) {
  let defaultClasses = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return stages.reduce((classes, stage) => _objectSpread(_objectSpread({}, classes), {}, {
    [stage]: true
  }), defaultClasses);
};
const ProgressButton = /*#__PURE__*/function () {
  const ProgressButton = _ref => {
    let children = _ref.children,
      className = _ref.className,
      disabled = _ref.disabled,
      _ref$hierarchy = _ref.hierarchy,
      hierarchy = _ref$hierarchy === void 0 ? 'loud' : _ref$hierarchy,
      id = _ref.id,
      _ref$loading = _ref.loading,
      loading = _ref$loading === void 0 ? false : _ref$loading,
      progressLabel = _ref.progressLabel,
      progressSrAnnouncement = _ref.progressSrAnnouncement,
      _ref$timeout = _ref.timeout,
      timeout = _ref$timeout === void 0 ? 1000 : _ref$timeout,
      ref = _ref.ref,
      otherProps = _objectWithoutProperties(_ref, _excluded$1);
    const componentId = useId(id);
    const mounted = useRef(false);
    const _useState = useState([STATES.ENTERED]),
      _useState2 = _slicedToArray(_useState, 2),
      stages = _useState2[0],
      setStages = _useState2[1];
    const hierarchySelected = HIERARCHY_ALLOWED[hierarchy] ? hierarchy : DEFAULT_HIERARCHY$1;
    warning(Boolean(progressLabel || progressSrAnnouncement), WARNING_A11Y_PROGRESS_LABEL);
    useEffect(() => {
      if (mounted.current) {
        if (loading) {
          setStages([STATES.LOADING]);
        } else {
          setStages([STATES.LOADING, STATES.FINISHED]);
          const finishTimeout = setTimeout(() => {
            setStages([STATES.FINISHED]);
          }, timeout);
          return () => clearTimeout(finishTimeout);
        }
      }
    }, [loading, timeout]);
    useEffect(() => {
      mounted.current = true;
      if (loading) {
        setStages([STATES.LOADING]);
      }
    }, []);
    const classObject = buildClassesObject(stages, {
      'andes-button--progress__label': Boolean(progressLabel)
    });
    const classes = classNames('andes-button--progress', className, classObject);
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Button, _extends({
      id: componentId,
      hierarchy: hierarchySelected,
      className: classes,
      ref: ref,
      disabled: disabled || loading
    }, otherProps, loading ? {
      'data-andes-state': 'progress'
    } : {}), progressLabel && loading ? /*#__PURE__*/React.createElement("span", {
      className: "andes-button--progress__content"
    }, progressLabel) : children), /*#__PURE__*/React.createElement(VisuallyHidden, {
      "aria-live": "assertive"
    }, (progressSrAnnouncement || progressLabel) && loading ? progressSrAnnouncement || progressLabel : ''));
  };
  ProgressButton.displayName = 'ProgressButton';
  return ProgressButton;
}();
var ProgressButton$1 = ProgressButton;
const _excluded = ["behavior", "className", "hierarchy", "icon", "size", "srLabel", "text", "ref"];
const DEFAULT_HIERARCHY = 'loud';
const namespace = 'andes-button';
const FloatingActionButton = /*#__PURE__*/function () {
  const FloatingActionButton = props => {
    const _props$behavior = props.behavior,
      behavior = _props$behavior === void 0 ? 'expanded' : _props$behavior,
      className = props.className,
      _props$hierarchy = props.hierarchy,
      hierarchy = _props$hierarchy === void 0 ? DEFAULT_HIERARCHY : _props$hierarchy,
      icon = props.icon,
      _props$size = props.size,
      size = _props$size === void 0 ? 'large' : _props$size,
      srLabel = props.srLabel,
      text = props.text,
      ref = props.ref,
      otherProps = _objectWithoutProperties(props, _excluded);
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    warning(text || srLabel, `<FloatingActionButton> needs some plain text contents for accessibility reasons.
      Please add an 'srLabel' or 'text' prop.`);
    const containerClasses = classNames(`${namespace}-fab`, `${namespace}-fab--${behavior}`, `${namespace}-fab--${hierarchy}`, className, {
      [`${namespace}-fab--${size}`]: size
    }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('fab'));
    const content = /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("span", {
      className: `${namespace}--icon-container`
    }, /*#__PURE__*/cloneElement(icon, {
      'aria-hidden': true
    })), text && /*#__PURE__*/React.createElement(ButtonText, null, text));
    return /*#__PURE__*/React.createElement("div", _extends({
      className: containerClasses,
      "data-andes-fab-button-container": true,
      "data-andes-fab-button-hierarchy": hierarchy,
      "data-andes-fab-button-size": size
    }, themeHash, computeDataAndesState({
      expanded: behavior === 'expanded',
      collapsed: behavior === 'collapsed'
    })), /*#__PURE__*/React.createElement(Button, _extends({}, otherProps, {
      "data-andes-fab-button": true,
      hierarchy: hierarchy,
      size: size,
      srLabel: srLabel || text,
      ref: ref
    }), content));
  };
  FloatingActionButton.displayName = 'FloatingActionButton';
  return FloatingActionButton;
}();
var FloatingActionButton$1 = FloatingActionButton;
export { Button, ButtonText, DropdownButton$1 as DropdownButton, FloatingActionButton$1 as FloatingActionButton, ProgressButton$1 as ProgressButton, SplitButton$1 as SplitButton };
