// node-melitk-otel/lib/events/activator/scheduledActivator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScheduledActivator = void 0;
const api_1 = require("@opentelemetry/api");
const telemetry_1 = __importDefault(require("./telemetry"));
const client_1 = require("./client/client");
const configs_1 = require("../../configs");
/**
 * Represents the scheduled activator class.
 * @class
 * @implements {Activator}
 */
class ScheduledActivator {
    config;
    /**
     * Represents the status of the activator.
     * @type {boolean}
     * @private
     */
    active;
    /**
     * Represents the telemetry for the activator.
     * @type {ActivatorTelemetry}
     * @private
     * @readonly
     */
    telemetry;
    /**
     * Represents the rest client for the activator.
     * @type {ActivatorClient}
     * @private
     * @readonly
     */
    client;
    /**
     * Creates an instance of ScheduledActivator.
     * @param {ActivatorConfig} config The {@link ActivatorConfig} for the activator.
     * @param {string} service The name of the service.
     * @constructor
     * @returns {ScheduledActivator} The instance of ScheduledActivator.
     */
    constructor(config) {
        this.config = config;
        this.active = false;
        this.client = new client_1.ActivatorClient(this.config.configs);
        this.telemetry = new telemetry_1.default();
        this.telemetry.observeActivatorStatusMetric(() => this.active);
    }
    /**
     * Gets the status of the activator.
     * @returns {boolean} The status of the activator.
     * @public
     * @method
     * @override
     */
    isActive() {
        return this.active;
    }
    /**
     * Runs the activator.
     * Calls the `events-flow-discovery` service on {@link checkService} and sets the boolean result on {@link isActive} property.
     * @returns {void}
     * @public
     * @method
     */
    async run() {
        api_1.diag.debug(`${configs_1.O11Y_EVENTS_LOG_PREFIX}Running activator for ${this.config.service}`);
        this.telemetry.incrementCheckServiceMetric();
        await this.client
            .checkService(this.config.service)
            .then((response) => {
            this.telemetry.incrementCheckServiceSucceedMetric();
            this.active = response.status === 'enabled';
        })
            .catch((error) => {
            if (error.response !== undefined && error.response.status === 404) {
                api_1.diag.debug(`${configs_1.O11Y_EVENTS_LOG_PREFIX}Service ${this.config.service} not found in activation service`);
                this.active = false;
                this.telemetry.incrementCheckServiceNotFoundMetric();
                return;
            }
            this.telemetry.incrementCheckServiceFailedMetric();
            api_1.diag.debug(`${configs_1.O11Y_EVENTS_LOG_PREFIX}Failed to check activation status for ${this.config.service}: ${error.status}`);
        });
    }
}
exports.ScheduledActivator = ScheduledActivator;
