// node-melitk-otel/lib/events/activator/telemetry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const metric_1 = require("../common/metric");
const config_1 = require("../config/config");
/**
 * Represents the ActivatorTelemetry class which extends O11yMetrics.
 * This class provides metrics related to the activator's behavior.
 * @class
 * @extends {O11yMetrics}
 */
class ActivatorTelemetry extends metric_1.O11yMetrics {
    /**
     * Represents the metric `activator.check_service`.
     * This metric represents the number of times the activator checks if the service should be active.
     * @type {Counter}
     * @private
     * @readonly
     */
    _checkServiceMetric;
    /**
     * Represents the metric `activator.check_service.failed`.
     * This metric represents the number of times the activator checks if the service should be active and fails.
     * @type {Counter}
     * @private
     * @readonly
     */
    _checkServiceFailedMetric;
    /**
     * Represents the metric `activator.check_service.not_found`.
     * This metric represents the number of times the activator checks if the service should be active and receives a 404 status response.
     * @type {Counter}
     * @private
     * @readonly
     */
    _checkServiceNotFoundMetric;
    /**
     * Represents the metric `activator.check_service.succeed`.
     * This metric represents the number of times the activator checks if the service should be active and succeeds.
     * @type {Counter}
     * @private
     * @readonly
     */
    _checkServiceSucceedMetric;
    /**
     * Represents the metric `activator.is_active`.
     * This metric represents if the service is active, metric with value 1, or not, metric with value 0.
     * @type {ObservableGauge}
     * @private
     * @readonly
     */
    _activatorStatusMetric;
    /**
     * Creates an instance of ActivatorTelemetry.
     */
    constructor() {
        super();
        this._checkServiceMetric = this.newCounter('activator.check_service', {
            description: 'This metric represents the number of times the activator checks if the service should be active.',
            unit: '1',
        });
        this._checkServiceFailedMetric = this.newCounter('activator.check_service.failed', {
            description: 'This metric represents the number of times the activator checks if the service should be active and fails.',
            unit: '1',
        });
        this._checkServiceNotFoundMetric = this.newCounter('activator.check_service.not_found', {
            description: 'This metric represents the number of times the activator checks if the service should be active and receives a 404 status response',
            unit: '1',
        });
        this._checkServiceSucceedMetric = this.newCounter('activator.check_service.succeed', {
            description: 'This metric represents the number of times the activator checks if the service should be active and succeeds.',
            unit: '1',
        });
        this._activatorStatusMetric = this.newObservableGauge('activator.is_active', {
            description: 'This metric represents if the service is active, metric with value 1, or not, metric with value 0.',
            unit: '1',
        });
    }
    /**
     * Increments the metric `activator.check_service` by 1.
     * @returns {void}
     */
    incrementCheckServiceMetric() {
        this._checkServiceMetric.add(1, config_1.metricDefaultAttributes);
    }
    /**
     * Increments the metric `activator.check_service.failed` by 1.
     * @returns {void}
     */
    incrementCheckServiceFailedMetric = () => {
        this._checkServiceFailedMetric.add(1, config_1.metricDefaultAttributes);
    };
    /**
     * Increments the metric `activator.check_service.not_found` by 1.
     * @returns {void}
     */
    incrementCheckServiceNotFoundMetric = () => {
        this._checkServiceNotFoundMetric.add(1, config_1.metricDefaultAttributes);
    };
    /**
     * Increments the metric `activator.check_service.succeed` by 1.
     * @returns {void}
     */
    incrementCheckServiceSucceedMetric = () => {
        this._checkServiceSucceedMetric.add(1, config_1.metricDefaultAttributes);
    };
    /**
     * Creates an observer for the return of the {@link isActive}.
     * Changes the metric `activator.is_active` with the value 1 if the service is active, otherwise 0.
     * @param {() => boolean} isActive A function that returns true if the service is active, otherwise false.
     * @returns {void}
     */
    observeActivatorStatusMetric = (isActive) => {
        this._activatorStatusMetric.addCallback((observer) => {
            observer.observe(isActive() ? 1 : 0, config_1.metricDefaultAttributes);
        });
    };
}
exports.default = ActivatorTelemetry;
