// @nordic/core-context/src/server/middleware.ts
import type { RequestHandler } from 'express';
import type { DeviceInfo } from './device/types';
import type { NordicCoreServerContext } from './storage';

import { getDeviceInfo, getWebviewInfo } from '../common/entries/getters';
import { getDeviceType } from './utils/device';
import { getPlatformType } from './utils/platform';
import { getWebviewProfile, setWebkitContextCookie } from './utils/webview';
import { getDeviceProfile, setDeviceInfoCookie } from './device';
import { storage } from './storage';

export const coreContextMiddleware: RequestHandler = (req, res, next) => {
  const platformType = getPlatformType();
  const deviceProfile = getDeviceProfile(platformType);
  const webviewProfile = getWebviewProfile(platformType);
  const context: NordicCoreServerContext = {
    deviceType: getDeviceType(),
    deviceInfo: deviceProfile.deviceInfo,
    networkState: deviceProfile.networkState,
    platformType,
  };

  if (webviewProfile) {
    const { safeAreaInsets } = webviewProfile;

    if (
      safeAreaInsets &&
      Object.values(safeAreaInsets).some((value) => typeof value === 'number')
    ) {
      context.safeAreaInsets = safeAreaInsets;
    }

    context.webviewInfo = {
      nativeApp: webviewProfile.nativeApp,
      webkit: webviewProfile.webkit,
    };
  }

  return storage.run(context, next);
};

export const webviewCookiesMiddleware: RequestHandler = (req, res, next) => {
  if (!req.device.webView) {
    return next();
  }

  const deviceOS = getDeviceInfo<DeviceInfo>()?.os;
  const webkitInfo = getWebviewInfo()?.webkit;

  if (deviceOS) {
    setDeviceInfoCookie(deviceOS);
  }

  if (webkitInfo) {
    setWebkitContextCookie(webkitInfo);
  }

  return next();
};
