// @meli/input-validation/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.nullable = exports.optional = exports.nil = exports.undefined = exports.stringBoolean = exports.coerce = exports.nativeEnumeration = exports.enumeration = exports.literal = exports.date = exports.bigint = exports.boolean = exports.number = exports.nullish = exports.discriminatedUnion = exports.intersection = exports.union = exports.record = exports.object = exports.string = exports.array = exports.createValidationMiddleware = exports.requestValidation = exports.ivStringBoolean = exports.ivNullable = exports.ivOptional = exports.ivNull = exports.ivUndefined = exports.ivCoerce = exports.ivNativeEnum = exports.ivEnum = exports.ivLiteral = exports.ivDate = exports.ivBigint = exports.ivBoolean = exports.ivNumber = exports.ivNullish = exports.ivDiscriminatedUnion = exports.ivIntersection = exports.ivUnion = exports.IvRecord = exports.ivRecord = exports.IvObject = exports.ivObject = exports.IvString = exports.ivString = exports.IvArray = exports.ivArray = exports.schemaBuilder = void 0;
const zod_1 = require("zod");
const ivArray_1 = __importStar(require("./iv/entity/ivArray"));
exports.ivArray = ivArray_1.default;
exports.array = ivArray_1.default;
Object.defineProperty(exports, "IvArray", { enumerable: true, get: function () { return ivArray_1.IvArray; } });
const ivString_1 = __importStar(require("./iv/entity/ivString"));
exports.ivString = ivString_1.default;
exports.string = ivString_1.default;
Object.defineProperty(exports, "IvString", { enumerable: true, get: function () { return ivString_1.IvString; } });
const ivObject_1 = __importStar(require("./iv/entity/ivObject"));
exports.ivObject = ivObject_1.default;
exports.object = ivObject_1.default;
Object.defineProperty(exports, "IvObject", { enumerable: true, get: function () { return ivObject_1.IvObject; } });
const ivRecord_1 = __importStar(require("./iv/entity/ivRecord"));
exports.ivRecord = ivRecord_1.default;
exports.record = ivRecord_1.default;
Object.defineProperty(exports, "IvRecord", { enumerable: true, get: function () { return ivRecord_1.IvRecord; } });
const ivCoerce_1 = __importDefault(require("./iv/entity/ivCoerce"));
exports.ivCoerce = ivCoerce_1.default;
exports.coerce = ivCoerce_1.default;
const ivStringBoolean_1 = __importDefault(require("./iv/entity/ivStringBoolean"));
exports.ivStringBoolean = ivStringBoolean_1.default;
exports.stringBoolean = ivStringBoolean_1.default;
const ivUnion_1 = __importDefault(require("./iv/operator/ivUnion"));
exports.ivUnion = ivUnion_1.default;
exports.union = ivUnion_1.default;
const ivIntersection_1 = __importDefault(require("./iv/operator/ivIntersection"));
exports.ivIntersection = ivIntersection_1.default;
exports.intersection = ivIntersection_1.default;
const ivDiscriminatedUnion_1 = __importDefault(require("./iv/operator/ivDiscriminatedUnion"));
exports.ivDiscriminatedUnion = ivDiscriminatedUnion_1.default;
exports.discriminatedUnion = ivDiscriminatedUnion_1.default;
const ivNullish_1 = __importDefault(require("./iv/modifier/ivNullish"));
exports.ivNullish = ivNullish_1.default;
exports.nullish = ivNullish_1.default;
const ivOptional_1 = __importDefault(require("./iv/modifier/ivOptional"));
exports.ivOptional = ivOptional_1.default;
exports.optional = ivOptional_1.default;
const ivNullable_1 = __importDefault(require("./iv/modifier/ivNullable"));
exports.ivNullable = ivNullable_1.default;
exports.nullable = ivNullable_1.default;
const requestValidation_1 = __importStar(require("./iv/middleware/requestValidation"));
exports.requestValidation = requestValidation_1.default;
Object.defineProperty(exports, "createValidationMiddleware", { enumerable: true, get: function () { return requestValidation_1.createValidationMiddleware; } });
const secureTypeMetric_1 = require("./utils/secureTypeMetric");
const ivNumber = (0, secureTypeMetric_1.createSecureType)(zod_1.z.number);
exports.ivNumber = ivNumber;
exports.number = ivNumber;
const ivBoolean = (0, secureTypeMetric_1.createSecureType)(zod_1.z.boolean);
exports.ivBoolean = ivBoolean;
exports.boolean = ivBoolean;
const ivBigint = (0, secureTypeMetric_1.createSecureType)(zod_1.z.bigint);
exports.ivBigint = ivBigint;
exports.bigint = ivBigint;
const ivDate = (0, secureTypeMetric_1.createSecureType)(zod_1.z.date);
exports.ivDate = ivDate;
exports.date = ivDate;
const ivLiteral = (0, secureTypeMetric_1.createSecureType)(zod_1.z.literal);
exports.ivLiteral = ivLiteral;
exports.literal = ivLiteral;
const ivEnum = (0, secureTypeMetric_1.createSecureType)(zod_1.z.enum);
exports.ivEnum = ivEnum;
exports.enumeration = ivEnum;
const ivNativeEnum = (0, secureTypeMetric_1.createSecureType)(zod_1.z.nativeEnum);
exports.ivNativeEnum = ivNativeEnum;
exports.nativeEnumeration = ivNativeEnum;
const ivUndefined = (0, secureTypeMetric_1.createSecureType)(zod_1.z.undefined);
exports.ivUndefined = ivUndefined;
exports.undefined = ivUndefined;
const ivNull = (0, secureTypeMetric_1.createSecureType)(zod_1.z.null);
exports.ivNull = ivNull;
exports.nil = ivNull;
exports.schemaBuilder = {
    ivArray: ivArray_1.default,
    ivString: ivString_1.default,
    ivObject: ivObject_1.default,
    ivRecord: ivRecord_1.default,
    ivUnion: ivUnion_1.default,
    ivIntersection: ivIntersection_1.default,
    ivDiscriminatedUnion: ivDiscriminatedUnion_1.default,
    ivNullish: ivNullish_1.default,
    ivNumber,
    ivBoolean,
    ivBigint,
    ivDate,
    ivLiteral,
    ivEnum,
    ivNativeEnum,
    ivCoerce: ivCoerce_1.default,
    ivUndefined,
    ivNull,
    ivOptional: ivOptional_1.default,
    ivNullable: ivNullable_1.default,
    ivStringBoolean: ivStringBoolean_1.default,
};
