// @andes/button/build/cjs/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _extends = require('@babel/runtime/helpers/extends');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var React = require('react');
var PropTypes = require('prop-types');
var classNames = require('classnames');
var progressIndicatorCircular = require('@andes/progress-indicator-circular');
var visuallyHidden = require('@andes/visually-hidden');
var warning = require('@andes/common/build/utils/warning');
var context = require('@andes/context');
var themes = require('@andes/themes');
var computeDataAndesState = require('@andes/common/build/utils/computeDataAndesState');
var floatingMenu = require('@andes/floating-menu');
var transformToKebabCase = require('@andes/common/build/utils/transformToKebabCase');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _objectWithoutProperties__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutProperties);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var transformToKebabCase__default = /*#__PURE__*/_interopDefaultLegacy(transformToKebabCase);
const namespace$2 = 'andes-button';
const namespaceDropdownButton = 'andes-button-dropdown';
const WARNING_A11Y_LABELS = `<DropdownButton> needs some plain text contents for accessibility reasons.
Please add an 'srLabel' or 'children' prop.`;
const WARNING_A11Y_PROGRESS_LABEL = `<ProgressButton> needs to announce its loading progress for accessibility reasons.
Please use a 'progressLabel' or add 'progressSrAnnouncement' prop.`;
const WARNING_HIERARCHY_VALUE = `'hierarchy' with 'transparent' value will be deprecated in a future major version. 
Please, use 'mute' value on <Button> instead`;

/**
 * @param {import('../..').ButtonProps['hierarchy']} [hierarchy]
 */
const computeHierarchy = hierarchy => {
  if (hierarchy === 'transparent') {
    warning__default["default"](false, WARNING_HIERARCHY_VALUE);
    return 'mute';
  }
  return hierarchy;
};
const _excluded$3 = ["children", "className", "disabled", "fullWidth", "hierarchy", "href", "id", "loading", "onClick", "size", "srAnnouncement", "srLabel", "type", "ref", "backgroundType", "data-andes-state"];
const progressIndicatorSize = {
  small: 'xsmall',
  medium: 'xsmall',
  large: 'small'
};

/**
 * Button text component
 * Utility component only to be used when Button content
 * includes more than just text (could be an icon or spinner)
 *
 * @type {typeof import('../index').ButtonText}
 */
const ButtonText = _ref => {
  let children = _ref.children;
  return /*#__PURE__*/React__default["default"].createElement("span", {
    className: `${namespace$2}__text`,
    "data-andes-button-text": true
  }, children);
};
ButtonText.propTypes = {
  children: PropTypes__default["default"].node.isRequired
};

/** @type {typeof import('..').Button} */
const Button = props => {
  const children = props.children,
    className = props.className,
    _props$disabled = props.disabled,
    disabled = _props$disabled === void 0 ? false : _props$disabled,
    _props$fullWidth = props.fullWidth,
    fullWidth = _props$fullWidth === void 0 ? false : _props$fullWidth,
    _props$hierarchy = props.hierarchy,
    hierarchy = _props$hierarchy === void 0 ? 'loud' : _props$hierarchy,
    href = props.href,
    id = props.id,
    loading = props.loading,
    onClick = props.onClick,
    _props$size = props.size,
    size = _props$size === void 0 ? 'large' : _props$size,
    srAnnouncement = props.srAnnouncement,
    srLabel = props.srLabel,
    _props$type = props.type,
    type = _props$type === void 0 ? 'button' : _props$type,
    ref = props.ref,
    _props$backgroundType = props.backgroundType,
    backgroundType = _props$backgroundType === void 0 ? 'default' : _props$backgroundType,
    dataAndesState = props['data-andes-state'],
    otherProps = _objectWithoutProperties__default["default"](props, _excluded$3);
  const componentId = context.useId(id);
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;

  /** @type {React.RefObject<NodeJS.Timeout | undefined>} */
  const srLabelTimeout = React.useRef(undefined);
  const firstRender = React.useRef(true);
  const _useState = React.useState(''),
    _useState2 = _slicedToArray__default["default"](_useState, 2),
    srAnnouncementLabel = _useState2[0],
    setSrAnnouncementLabel = _useState2[1];
  const _useState3 = React.useState(false),
    _useState4 = _slicedToArray__default["default"](_useState3, 2),
    loadingComplete = _useState4[0],
    setLoadingComplete = _useState4[1];
  const computedHierarchy = computeHierarchy(hierarchy);

  // Resets label when the component finishes loading
  React.useEffect(() => {
    if (!loading) {
      setSrAnnouncementLabel('');
      if (!firstRender.current) setLoadingComplete(true);
    }
    if (firstRender) firstRender.current = false;
  }, [loading]);
  React.useEffect(() => {
    /** @type {NodeJS.Timeout} */
    let finishTimeout;
    if (loadingComplete) {
      finishTimeout = setTimeout(() => {
        setLoadingComplete(false);
        /**
         * The exit animation takes 0.2 seconds to avoid animation failures it must be the same value.
         * TODO: When supported by more browsers I will change this implementation to animationend to be more precise.
         * https://caniuse.com/?search=animationend
         */
      }, 200);
    }
    return () => clearTimeout(finishTimeout);
  }, [loadingComplete]);
  React.useEffect(() => () => {
    clearTimeout(srLabelTimeout.current);
  }, []);
  warning__default["default"](!loading || loading && srAnnouncement, `<ButtonWithSpinner> needs some plain text contents for accessibility reasons.
    Please add a 'srAnnouncement' prop.`);
  const classes = classNames__default["default"](namespace$2, className, {
    [`${namespace$2}--${size}`]: size,
    [`${namespace$2}--${computedHierarchy}`]: computedHierarchy,
    [`${namespace$2}--disabled`]: disabled,
    [`${namespace$2}--full-width`]: fullWidth,
    [`${namespace$2}--loading`]: loading,
    [`${namespace$2}--loading-complete`]: loadingComplete,
    [`${namespace$2}--color-on-background`]: backgroundType === 'complexBackground'
  }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('button'));
  const sharedProps = _objectSpread__default["default"](_objectSpread__default["default"](_objectSpread__default["default"]({
    className: classes,
    id: componentId,
    'aria-label': srLabel
  }, otherProps), {}, {
    'data-andes-button': true,
    'data-andes-button-hierarchy': computedHierarchy,
    'data-andes-button-size': size
  }, computeDataAndesState.computeDataAndesState(_objectSpread__default["default"](_objectSpread__default["default"]({}, dataAndesState ? {
    otherDataAndesState: dataAndesState
  } : {}), {}, {
    disabled,
    loading
  }))), themeHash);
  const content = /*#__PURE__*/React__default["default"].createElement("span", {
    className: `${namespace$2}__content`,
    "data-andes-button-content": true
  }, children);
  if (href) {
    /** @type {React.AnchorHTMLAttributes<HTMLAnchorElement> & React.RefAttributes<HTMLAnchorElement | null>} */
    const anchorProps = _objectSpread__default["default"](_objectSpread__default["default"]({}, disabled || loading ? {
      role: 'link',
      'aria-disabled': true
    } : {
      href
    }), {}, {
      // eslint-disable-next-line object-shorthand
      ref: (/** @type {React.RefObject<HTMLAnchorElement | null>} */ /** @type {unknown} */ref),
      // eslint-disable-next-line object-shorthand
      onClick: (/** @type {React.MouseEventHandler<HTMLAnchorElement>} */ /** @type {unknown} */onClick)
    }, sharedProps);
    return /*#__PURE__*/React__default["default"].createElement("a", anchorProps, content);
  }

  /** @type {React.ButtonHTMLAttributes<HTMLButtonElement> & React.RefAttributes<HTMLButtonElement | null>} */
  const buttonProps = _objectSpread__default["default"]({
    disabled: disabled || loading,
    // eslint-disable-next-line object-shorthand
    ref: (/** @type {React.RefObject<HTMLButtonElement | null>} */ /** @type {unknown} */ref),
    // eslint-disable-next-line object-shorthand
    type: (/** @type {React.ButtonHTMLAttributes<HTMLButtonElement>['type']} */type)
  }, sharedProps);

  /** @type {React.MouseEventHandler<HTMLButtonElement>} */
  const customOnClick = event => {
    const label = srAnnouncement || '';
    srLabelTimeout.current = setTimeout(() => setSrAnnouncementLabel(label), 30);
    if (typeof onClick === 'function') onClick(event);
  };
  const progressSize = /** @type {'small' | 'medium' | 'large'} */progressIndicatorSize[size];
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement("button", _extends__default["default"]({}, buttonProps, {
    onClick: customOnClick
  }), content, loading && /*#__PURE__*/React__default["default"].createElement(progressIndicatorCircular.ProgressIndicatorCircular
  // @ts-expect-error It is a prop for internal use and is not typed.
  , {
    ariaLabelledBy: componentId,
    size: progressSize
  })), loading && /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, {
    "aria-live": "assertive"
  }, srAnnouncementLabel));
};
Button.displayName = 'Button';
Button.propTypes = {
  backgroundType: PropTypes__default["default"].oneOf(['default', 'complexBackground']),
  children: PropTypes__default["default"].node.isRequired,
  className: PropTypes__default["default"].string,
  disabled: PropTypes__default["default"].bool,
  fullWidth: PropTypes__default["default"].bool,
  hierarchy: PropTypes__default["default"].oneOf(['loud', 'quiet', 'mute']),
  href: PropTypes__default["default"].string,
  id: PropTypes__default["default"].string,
  loading: PropTypes__default["default"].bool,
  onClick: PropTypes__default["default"].func,
  size: PropTypes__default["default"].oneOf(['small', 'medium', 'large']),
  srAnnouncement: PropTypes__default["default"].string,
  srLabel: PropTypes__default["default"].string,
  type: PropTypes__default["default"].string
};
const _excluded$2 = ["className", "dialogSrLabel", "id", "listContent", "namespace", "onClickOutside", "placement", "popperOptions", "srLabel", "trigger", "visible"];

/** @typedef {(e: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>, value: string) => void} OnClickHandler */

/**
 * @typedef {Object} ListContent
 * @prop {string} title
 * @prop {string} [description]
 * @prop {string} [href]
 * @prop {OnClickHandler} [onClick]
 * @prop {import('@andes/floating-menu').FloatingMenuItemProps['leftContent']} [leftContent]
 */

/**
 * @typedef {Object} ButtonListContentProps
 * @prop {() => HTMLElement} [appendTo]
 * @prop {string} [dialogSrLabel]
 * @prop {string} id
 * @prop {string} namespace
 * @prop {ListContent[]} listContent
 * @prop {boolean} [visible]
 * @prop {() => void} [onClickOutside]
 * @prop {import('@andes/floating-menu').FloatingMenuProps['popperOptions']} [popperOptions]
 */

/**
 * @type {React.FunctionComponent<ButtonListContentProps
 * & Pick<import('../../floating-menu').FloatingMenuProps,
 * 'id' | 'placement' | 'trigger' | 'className' | 'srLabel' | 'width' | 'onOpen' | 'onClose'>>}
 */
const ButtonListContent = _ref => {
  let className = _ref.className,
    dialogSrLabel = _ref.dialogSrLabel,
    id = _ref.id,
    listContent = _ref.listContent,
    namespace = _ref.namespace;
  _ref.onClickOutside;
  let placement = _ref.placement,
    popperOptions = _ref.popperOptions,
    srLabel = _ref.srLabel,
    trigger = _ref.trigger,
    visible = _ref.visible,
    props = _objectWithoutProperties__default["default"](_ref, _excluded$2);
  return /*#__PURE__*/React__default["default"].createElement(floatingMenu.FloatingMenu, _extends__default["default"]({
    className: className,
    dialogSrLabel: dialogSrLabel,
    placement: placement,
    open: visible,
    popperOptions: popperOptions,
    trigger: trigger,
    id: id,
    srLabel: srLabel
  }, props), listContent.map((_ref2, i) => {
    let title = _ref2.title,
      description = _ref2.description,
      href = _ref2.href,
      onClick = _ref2.onClick,
      leftContent = _ref2.leftContent;
    return /*#__PURE__*/React__default["default"].createElement(floatingMenu.FloatingMenuItem, {
      id: `${namespace}__menu-item-${i}`,
      href: href,
      className: `${namespace}__menu-item`,
      key: title,
      value: title,
      title: title,
      description: description,
      onClick: onClick,
      leftContent: leftContent
    });
  }));
};
ButtonListContent.propTypes = {
  appendTo: PropTypes__default["default"].func,
  className: PropTypes__default["default"].string,
  dialogSrLabel: PropTypes__default["default"].string,
  id: PropTypes__default["default"].string.isRequired,
  listContent: PropTypes__default["default"].arrayOf(PropTypes__default["default"].shape({
    description: PropTypes__default["default"].string,
    href: PropTypes__default["default"].string,
    onClick: PropTypes__default["default"].func,
    title: PropTypes__default["default"].string.isRequired,
    leftContent: PropTypes__default["default"].oneOfType([PropTypes__default["default"].node, PropTypes__default["default"].shape({
      title: PropTypes__default["default"].string,
      src: PropTypes__default["default"].string,
      alt: PropTypes__default["default"].string,
      src2x: PropTypes__default["default"].string
    })])
  }).isRequired).isRequired,
  namespace: PropTypes__default["default"].string.isRequired,
  onClickOutside: PropTypes__default["default"].func,
  placement: PropTypes__default["default"].oneOf(['bottom', 'bottom-start', 'bottom-end', 'top', 'top-start', 'top-end', 'left', 'left-start', 'left-end', 'right', 'right-start', 'right-end']),
  popperOptions: PropTypes__default["default"].shape({}),
  srLabel: PropTypes__default["default"].string,
  trigger: PropTypes__default["default"].element.isRequired,
  visible: PropTypes__default["default"].bool,
  width: PropTypes__default["default"].oneOfType([PropTypes__default["default"].number, PropTypes__default["default"].string])
};
var ButtonListContent$1 = ButtonListContent;
var enTranslations = {
  "andes-button-dropdown\u0004menú desplegable": [null, "Dropdown menu"],
  "andes-button-split\u0004menú desplegable": [null, "Dropdown menu"]
};
var zhCnTranslations = {
  "andes-button-dropdown\u0004menú desplegable": [null, "下拉菜单"],
  "andes-button-split\u0004menú desplegable": [null, "下拉菜单"]
};
var esArTranslations = {
  "andes-button-dropdown\u0004menú desplegable": [null, "Menú desplegable"],
  "andes-button-split\u0004menú desplegable": [null, "Menú desplegable"]
};
var ptBrTranslations = {
  "andes-button-dropdown\u0004menú desplegable": [null, "Menu suspenso"],
  "andes-button-split\u0004menú desplegable": [null, "Menu suspenso"]
};
const ChevronDown$1 = () => /*#__PURE__*/React__default["default"].createElement("svg", {
  width: "20px",
  height: "20px",
  viewBox: "0 0 20 20",
  version: "1.1",
  xmlns: "http://www.w3.org/2000/svg",
  className: `${namespaceDropdownButton}__chevron`
}, /*#__PURE__*/React__default["default"].createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd",
  fillOpacity: "0.8"
}, /*#__PURE__*/React__default["default"].createElement("polygon", {
  fill: "currentColor",
  fillRule: "nonzero",
  transform: "translate(10.000000, 10.000000) rotate(-90.000000) translate(-10.000000, -10.000000) "
  // eslint-disable-next-line max-len
  ,

  points: "8.4482158 9.99750228 13.243845 5.20187305 12.3953169 4.35334491 6.75115953 9.99750228 12.4003123 15.6466551 13.2488405 14.798127"
})));
const ChevronUP$1 = () => /*#__PURE__*/React__default["default"].createElement("svg", {
  width: "20px",
  height: "20px",
  viewBox: "0 0 20 20",
  version: "1.1",
  xmlns: "http://www.w3.org/2000/svg",
  className: `${namespaceDropdownButton}__chevron`
}, /*#__PURE__*/React__default["default"].createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd",
  fillOpacity: "0.8"
}, /*#__PURE__*/React__default["default"].createElement("polygon", {
  fill: "currentColor",
  fillRule: "nonzero",
  transform: "translate(10.001873, 9.995000) rotate(90.000000) translate(-10.001873, -9.995000) "
  // eslint-disable-next-line max-len
  ,

  points: "8.45008885 9.99250228 13.2457181 5.19687305 12.3971899 4.34834491 6.75303257 9.99250228 12.4021854 15.6416551 13.2507135 14.793127"
})));

/**
 * @typedef {Object} DropdownButtonInternalProps
 * @prop {() => HTMLElement} [appendTo]
*/

/** @type {React.FunctionComponent<import('..').DropdownButtonProps & DropdownButtonInternalProps>} */
const DropdownButton = _ref => {
  let appendTo = _ref.appendTo,
    backgroundType = _ref.backgroundType,
    className = _ref.className,
    children = _ref.children,
    dialogSrLabel = _ref.dialogSrLabel,
    _ref$disabled = _ref.disabled,
    disabled = _ref$disabled === void 0 ? false : _ref$disabled,
    _ref$fullWidth = _ref.fullWidth,
    fullWidth = _ref$fullWidth === void 0 ? false : _ref$fullWidth,
    _ref$hierarchy = _ref.hierarchy,
    hierarchy = _ref$hierarchy === void 0 ? 'loud' : _ref$hierarchy,
    id = _ref.id,
    listContent = _ref.listContent,
    menuWidth = _ref.menuWidth,
    _ref$side = _ref.side,
    side = _ref$side === void 0 ? 'bottom' : _ref$side,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'large' : _ref$size,
    srLabel = _ref.srLabel,
    ref = _ref.ref;
  const componentId = context.useId(id);
  const _useState = React.useState(false),
    _useState2 = _slicedToArray__default["default"](_useState, 2),
    showOptions = _useState2[0],
    setShowOptions = _useState2[1];
  warning__default["default"](children || srLabel, WARNING_A11Y_LABELS);
  const _useAndesContext = context.useAndesContext('@andes/button', {
      en: enTranslations,
      'es-AR': esArTranslations,
      'pt-BR': ptBrTranslations,
      'zh-CN': zhCnTranslations
    }),
    i18n = _useAndesContext.i18n;
  const handleOpen = () => {
    setShowOptions(true);
  };
  const handleClose = () => {
    setShowOptions(false);
  };
  const classes = classNames__default["default"](namespaceDropdownButton, className);
  const listClasses = classNames__default["default"](`${namespaceDropdownButton}__menu`, className, {
    [`${namespaceDropdownButton}__menu--${side}`]: side
  });
  const trigger = /*#__PURE__*/React__default["default"].createElement(Button, {
    backgroundType: backgroundType,
    className: classes,
    hierarchy: hierarchy,
    size: size,
    "aria-expanded": showOptions,
    srLabel: srLabel,
    disabled: disabled,
    fullWidth: fullWidth,
    id: componentId,
    "data-andes-dropdown-button": true,
    "data-andes-state": showOptions ? 'expanded' : 'collapsed'
  }, children && /*#__PURE__*/React__default["default"].createElement(ButtonText, null, children), !showOptions ? /*#__PURE__*/React__default["default"].createElement(ChevronDown$1, null) : /*#__PURE__*/React__default["default"].createElement(ChevronUP$1, null));
  return /*#__PURE__*/React__default["default"].createElement("div", {
    ref: ref
  }, /*#__PURE__*/React__default["default"].createElement(ButtonListContent$1, {
    appendTo: appendTo,
    dialogSrLabel: dialogSrLabel,
    placement: side,
    listContent: listContent,
    onOpen: handleOpen,
    onClose: handleClose,
    popperOptions: {
      modifiers: [{
        name: 'sameWidth',
        enabled: !menuWidth,
        fn: _ref2 => {
          let state = _ref2.state;
          state.styles.popper.width = `${state.rects.reference.width}px`;
        },
        phase: 'beforeWrite',
        requires: ['computeStyles'],
        effect: _ref3 => {
          let state = _ref3.state;
          // @ts-expect-error: clientWidth exists in our implementation
          state.elements.popper.style.width = `${state.elements.reference.clientWidth}px`;
        }
      }]
    },
    trigger: trigger,
    namespace: namespaceDropdownButton,
    id: `${componentId}-menu`,
    className: listClasses,
    srLabel: srLabel || i18n.pgettext(namespaceDropdownButton, 'menú desplegable'),
    width: menuWidth || '100%'
  }));
};
DropdownButton.displayName = 'DropdownButton';
DropdownButton.propTypes = {
  appendTo: PropTypes__default["default"].func,
  backgroundType: PropTypes__default["default"].oneOf(['default', 'complexBackground']),
  className: PropTypes__default["default"].string,
  children: PropTypes__default["default"].string,
  dialogSrLabel: PropTypes__default["default"].string,
  disabled: PropTypes__default["default"].bool,
  fullWidth: PropTypes__default["default"].bool,
  hierarchy: PropTypes__default["default"].oneOf(['loud', 'quiet']),
  id: PropTypes__default["default"].string,
  listContent: PropTypes__default["default"].arrayOf(PropTypes__default["default"].shape({
    href: PropTypes__default["default"].string,
    onClick: PropTypes__default["default"].func,
    description: PropTypes__default["default"].string,
    title: PropTypes__default["default"].string.isRequired
  }).isRequired).isRequired,
  menuWidth: PropTypes__default["default"].oneOfType([PropTypes__default["default"].number, PropTypes__default["default"].string]),
  side: PropTypes__default["default"].oneOf(['top-start', 'top', 'top-end', 'bottom-start', 'bottom', 'bottom-end', 'left-start', 'left', 'left-end', 'right-start', 'right', 'right-end']),
  size: PropTypes__default["default"].oneOf(['medium', 'large']),
  srLabel: PropTypes__default["default"].string
};
var DropdownButton$1 = DropdownButton;

/** @typedef {import('@popperjs/core').ModifierPhases} ModifierPhases */

const namespace$1 = 'andes-button-split';
const ChevronDown = () => /*#__PURE__*/React__default["default"].createElement("svg", {
  width: "20px",
  height: "20px",
  viewBox: "0 0 20 20",
  version: "1.1",
  xmlns: "http://www.w3.org/2000/svg",
  className: `${namespace$1}__chevron`
}, /*#__PURE__*/React__default["default"].createElement("title", null, "Abrir men\xFA de opciones / Mostrar opciones"), /*#__PURE__*/React__default["default"].createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd",
  fillOpacity: "0.8"
}, /*#__PURE__*/React__default["default"].createElement("polygon", {
  fill: "currentColor",
  fillRule: "nonzero",
  transform: "translate(10.000000, 10.000000) rotate(-90.000000) translate(-10.000000, -10.000000) "
  // eslint-disable-next-line max-len
  ,

  points: "8.4482158 9.99750228 13.243845 5.20187305 12.3953169 4.35334491 6.75115953 9.99750228 12.4003123 15.6466551 13.2488405 14.798127"
})));
const ChevronUP = () => /*#__PURE__*/React__default["default"].createElement("svg", {
  width: "20px",
  height: "20px",
  viewBox: "0 0 20 20",
  version: "1.1",
  xmlns: "http://www.w3.org/2000/svg",
  className: `${namespace$1}__chevron`
}, /*#__PURE__*/React__default["default"].createElement("title", null, "Cerrar men\xFA de opciones / Ocultar opciones"), /*#__PURE__*/React__default["default"].createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd",
  fillOpacity: "0.8"
}, /*#__PURE__*/React__default["default"].createElement("polygon", {
  fill: "currentColor",
  fillRule: "nonzero",
  transform: "translate(10.001873, 9.995000) rotate(90.000000) translate(-10.001873, -9.995000) "
  // eslint-disable-next-line max-len
  ,

  points: "8.45008885 9.99250228 13.2457181 5.19687305 12.3971899 4.34834491 6.75303257 9.99250228 12.4021854 15.6416551 13.2507135 14.793127"
})));
const placement = {
  top: 'top-start',
  bottom: 'bottom-start',
  bottomRight: 'bottom-start',
  topRight: 'top-start'
};

/** @type {typeof import('..').SplitButton} */
const SplitButton = _ref => {
  let _ref$backgroundType = _ref.backgroundType,
    backgroundType = _ref$backgroundType === void 0 ? 'default' : _ref$backgroundType,
    children = _ref.children,
    className = _ref.className,
    dialogSrLabel = _ref.dialogSrLabel,
    disabled = _ref.disabled,
    _ref$hierarchy = _ref.hierarchy,
    hierarchy = _ref$hierarchy === void 0 ? 'loud' : _ref$hierarchy,
    id = _ref.id,
    listContent = _ref.listContent,
    onClick = _ref.onClick,
    _ref$side = _ref.side,
    side = _ref$side === void 0 ? 'bottomRight' : _ref$side,
    srLabel = _ref.srLabel,
    ref = _ref.ref;
  const componentId = context.useId(id);
  const _useState = React.useState(false),
    _useState2 = _slicedToArray__default["default"](_useState, 2),
    showOptions = _useState2[0],
    setShowOptions = _useState2[1];
  const _useAndesContext = context.useAndesContext('@andes/button', {
      en: enTranslations,
      'es-AR': esArTranslations,
      'pt-BR': ptBrTranslations,
      'zh-CN': zhCnTranslations
    }),
    i18n = _useAndesContext.i18n;
  const show = () => {
    setShowOptions(true);
  };
  const hide = () => {
    setShowOptions(false);
  };
  const classes = classNames__default["default"](namespace$1, className);
  const popperOptions = side === 'top' || side === 'bottom' ? {
    modifiers: [{
      name: 'sameWidth',
      enabled: true,
      // @ts-expect-error state has any type
      fn: _ref2 => {
        let state = _ref2.state;
        state.styles.popper.left = `-${state.elements.reference.parentElement.offsetLeft}px`;
        state.styles.popper.width = `${state.elements.reference.parentElement.offsetLeft + state.rects.reference.width}px`;
      },
      phase: (/** @type {ModifierPhases} */'beforeWrite'),
      requires: ['computeStyles'],
      // @ts-expect-error state has any type
      effect: _ref3 => {
        let state = _ref3.state;
        state.elements.popper.style.width = `${state.elements.reference.parentElement.offsetLeft + state.elements.reference.clientWidth}px`;
      }
    }]
  } : undefined;
  const listClasses = classNames__default["default"](`${namespace$1}__menu`, className, {
    [`${namespace$1}__menu--${transformToKebabCase__default["default"](side)}`]: side
  });
  const trigger = () => /*#__PURE__*/React__default["default"].createElement(Button, {
    backgroundType: backgroundType,
    className: `${namespace$1}-chevron`,
    disabled: disabled,
    hierarchy: hierarchy,
    "aria-expanded": showOptions,
    srLabel: srLabel || i18n.pgettext(namespace$1, 'menú desplegable'),
    id: `${componentId}-chevron`,
    "data-andes-split-button-trigger": true
  }, !showOptions ? /*#__PURE__*/React__default["default"].createElement(ChevronDown, null) : /*#__PURE__*/React__default["default"].createElement(ChevronUP, null));
  const buttonListContentPlacement = /** @type {'bottom-start' | 'top-start'} */placement[`${side}`];
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: classes,
    ref: ref,
    "data-andes-split-button": true,
    "data-andes-state": showOptions ? 'expanded' : 'collapsed'
  }, /*#__PURE__*/React__default["default"].createElement(Button, {
    backgroundType: backgroundType,
    className: `${namespace$1}-text`,
    disabled: disabled,
    onClick: onClick,
    hierarchy: hierarchy,
    id: `${componentId}-text`
  }, /*#__PURE__*/React__default["default"].createElement(ButtonText, null, children)), /*#__PURE__*/React__default["default"].createElement(ButtonListContent$1, {
    placement: buttonListContentPlacement,
    dialogSrLabel: dialogSrLabel,
    listContent: listContent,
    visible: showOptions,
    onClickOutside: hide,
    onOpen: show,
    onClose: hide,
    popperOptions: popperOptions,
    trigger: trigger(),
    namespace: namespace$1,
    id: `${componentId}-menu`,
    srLabel: srLabel || i18n.pgettext(namespace$1, 'menú desplegable'),
    className: listClasses,
    width: side === 'top' || side === 'bottom' ? '100%' : 'max-content'
  }));
};
SplitButton.propTypes = {
  backgroundType: PropTypes__default["default"].oneOf(['default', 'complexBackground']),
  children: PropTypes__default["default"].string.isRequired,
  className: PropTypes__default["default"].string,
  dialogSrLabel: PropTypes__default["default"].string,
  disabled: PropTypes__default["default"].bool,
  hierarchy: PropTypes__default["default"].oneOf(['loud', 'quiet']),
  id: PropTypes__default["default"].string,
  listContent: PropTypes__default["default"].arrayOf(PropTypes__default["default"].shape({
    href: PropTypes__default["default"].string,
    onClick: PropTypes__default["default"].func,
    description: PropTypes__default["default"].string,
    title: PropTypes__default["default"].string.isRequired,
    leftContent: PropTypes__default["default"].oneOfType([PropTypes__default["default"].node, PropTypes__default["default"].shape({
      title: PropTypes__default["default"].string,
      src: PropTypes__default["default"].string,
      alt: PropTypes__default["default"].string,
      src2x: PropTypes__default["default"].string
    })])
  }).isRequired).isRequired,
  onClick: PropTypes__default["default"].func,
  side: PropTypes__default["default"].oneOf(['top', 'bottom', 'bottomRight', 'topRight']),
  srLabel: PropTypes__default["default"].string
};
SplitButton.displayName = 'SplitButton';
var SplitButton$1 = SplitButton;
const _excluded$1 = ["children", "className", "disabled", "hierarchy", "id", "loading", "progressLabel", "progressSrAnnouncement", "timeout", "ref"];
const STATES = {
  LOADING: 'loading',
  ENTERED: 'entered',
  FINISHED: 'finished'
};
const DEFAULT_HIERARCHY$1 = 'loud';
const HIERARCHY_ALLOWED = {
  [DEFAULT_HIERARCHY$1]: true,
  quiet: true
};

/**
 * @param {string[]} stages
 * @param {Record<string, unknown>} defaultClasses
 */
const buildClassesObject = function (stages) {
  let defaultClasses = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return stages.reduce((classes, stage) => _objectSpread__default["default"](_objectSpread__default["default"]({}, classes), {}, {
    [stage]: true
  }), defaultClasses);
};

/** @type {typeof import('..').ProgressButton} */
const ProgressButton = _ref => {
  let children = _ref.children,
    className = _ref.className,
    disabled = _ref.disabled,
    _ref$hierarchy = _ref.hierarchy,
    hierarchy = _ref$hierarchy === void 0 ? 'loud' : _ref$hierarchy,
    id = _ref.id,
    _ref$loading = _ref.loading,
    loading = _ref$loading === void 0 ? false : _ref$loading,
    progressLabel = _ref.progressLabel,
    progressSrAnnouncement = _ref.progressSrAnnouncement,
    _ref$timeout = _ref.timeout,
    timeout = _ref$timeout === void 0 ? 1000 : _ref$timeout,
    ref = _ref.ref,
    otherProps = _objectWithoutProperties__default["default"](_ref, _excluded$1);
  const componentId = context.useId(id);
  const mounted = React.useRef(false);
  const _useState = React.useState([STATES.ENTERED]),
    _useState2 = _slicedToArray__default["default"](_useState, 2),
    stages = _useState2[0],
    setStages = _useState2[1];
  const hierarchySelected = HIERARCHY_ALLOWED[hierarchy] ? hierarchy : DEFAULT_HIERARCHY$1;
  warning__default["default"](Boolean(progressLabel || progressSrAnnouncement), WARNING_A11Y_PROGRESS_LABEL);
  React.useEffect(() => {
    if (mounted.current) {
      if (loading) {
        setStages([STATES.LOADING]);
      } else {
        setStages([STATES.LOADING, STATES.FINISHED]);
        const finishTimeout = setTimeout(() => {
          setStages([STATES.FINISHED]);
        }, timeout);
        return () => clearTimeout(finishTimeout);
      }
    }
  }, [loading, timeout]);
  React.useEffect(() => {
    mounted.current = true;
    if (loading) {
      setStages([STATES.LOADING]);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  const classObject = buildClassesObject(stages, {
    'andes-button--progress__label': Boolean(progressLabel)
  });
  const classes = classNames__default["default"]('andes-button--progress', className, classObject);
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(Button, _extends__default["default"]({
    id: componentId,
    hierarchy: hierarchySelected,
    className: classes,
    ref: ref,
    disabled: disabled || loading
  }, otherProps, loading ? {
    'data-andes-state': 'progress'
  } : {}), progressLabel && loading ? /*#__PURE__*/React__default["default"].createElement("span", {
    className: "andes-button--progress__content"
  }, progressLabel) : children), /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, {
    "aria-live": "assertive"
  }, (progressSrAnnouncement || progressLabel) && loading ? progressSrAnnouncement || progressLabel : ''));
};
ProgressButton.propTypes = {
  children: PropTypes__default["default"].node,
  className: PropTypes__default["default"].string,
  disabled: PropTypes__default["default"].bool,
  hierarchy: PropTypes__default["default"].oneOf(['quiet', 'loud']),
  id: PropTypes__default["default"].string,
  loading: PropTypes__default["default"].bool,
  progressLabel: PropTypes__default["default"].node,
  progressSrAnnouncement: PropTypes__default["default"].string,
  timeout: PropTypes__default["default"].number
};
ProgressButton.displayName = 'ProgressButton';
var ProgressButton$1 = ProgressButton;
const _excluded = ["behavior", "className", "hierarchy", "icon", "size", "srLabel", "text", "ref"];
const DEFAULT_HIERARCHY = 'loud';
const namespace = 'andes-button';

/** @type {typeof import('..').FloatingActionButton} */
const FloatingActionButton = props => {
  const _props$behavior = props.behavior,
    behavior = _props$behavior === void 0 ? 'expanded' : _props$behavior,
    className = props.className,
    _props$hierarchy = props.hierarchy,
    hierarchy = _props$hierarchy === void 0 ? DEFAULT_HIERARCHY : _props$hierarchy,
    icon = props.icon,
    _props$size = props.size,
    size = _props$size === void 0 ? 'large' : _props$size,
    srLabel = props.srLabel,
    text = props.text,
    ref = props.ref,
    otherProps = _objectWithoutProperties__default["default"](props, _excluded);
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;
  warning__default["default"](text || srLabel, `<FloatingActionButton> needs some plain text contents for accessibility reasons.
      Please add an 'srLabel' or 'text' prop.`);
  const containerClasses = classNames__default["default"](`${namespace}-fab`, `${namespace}-fab--${behavior}`, `${namespace}-fab--${hierarchy}`, className, {
    [`${namespace}-fab--${size}`]: size
  }, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('fab'));
  const content = /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement("span", {
    className: `${namespace}--icon-container`
  }, /*#__PURE__*/React.cloneElement(/** @type {React.ReactElement} */icon, {
    'aria-hidden': true
  })), text && /*#__PURE__*/React__default["default"].createElement(ButtonText, null, text));
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    className: containerClasses,
    "data-andes-fab-button-container": true,
    "data-andes-fab-button-hierarchy": hierarchy,
    "data-andes-fab-button-size": size
  }, themeHash, computeDataAndesState.computeDataAndesState({
    expanded: behavior === 'expanded',
    collapsed: behavior === 'collapsed'
  })), /*#__PURE__*/React__default["default"].createElement(Button, _extends__default["default"]({}, otherProps, {
    "data-andes-fab-button": true,
    hierarchy: hierarchy,
    size: size,
    srLabel: srLabel || text,
    ref: ref
  }), content));
};
FloatingActionButton.displayName = 'FloatingActionButton';
FloatingActionButton.propTypes = {
  behavior: PropTypes__default["default"].oneOf(['expanded', 'collapsed']),
  className: PropTypes__default["default"].string,
  icon: PropTypes__default["default"].element.isRequired,
  id: PropTypes__default["default"].string,
  onClick: PropTypes__default["default"].func,
  size: PropTypes__default["default"].oneOf(['small', 'large']),
  hierarchy: PropTypes__default["default"].oneOf(['loud', 'quiet']),
  srLabel: PropTypes__default["default"].string,
  text: PropTypes__default["default"].string
};
var FloatingActionButton$1 = FloatingActionButton;
exports.Button = Button;
exports.ButtonText = ButtonText;
exports.DropdownButton = DropdownButton$1;
exports.FloatingActionButton = FloatingActionButton$1;
exports.ProgressButton = ProgressButton$1;
exports.SplitButton = SplitButton$1;
