// nordic-csp-engine/lib/csp-config.js
const { env } = require('@nordic/env');
const {
  RestClient: createRestClient,
  buildFlowStarterContext,
} = require('frontend-restclient');

const { detectFramework } = require('./framework');
const {
  RESTCLIENT_TIMEOUT,
  RESTCLIENT_WEB_SEC_SCOPE,
  RESTCLIENT_WEB_SEC_BASE_URL,
} = require('./constants/restclient');
const {
  CRON_CSP_CONFIG_NAME,
  CRON_CSP_STORAGE_KEY,
  CRON_CSP_TIME,
} = require('./constants/cron');

const framework = detectFramework();

const restclient = createRestClient({
  baseURL: RESTCLIENT_WEB_SEC_BASE_URL,
  timeout: RESTCLIENT_TIMEOUT,
  retry: {
    maxRetries: 2,
  },
  logErrors: false,
});

async function handler() {
  const queryParams = framework ? { framework } : {};

  const { data } = await restclient.get(
    `/applications/${env.APPLICATION}/scopes/${RESTCLIENT_WEB_SEC_SCOPE}/headers`,
    {
      context: buildFlowStarterContext(),
      params: queryParams,
    },
  );

  const cspConfig = data?.csp;

  if (!cspConfig) {
    return null;
  }

  return {
    key: CRON_CSP_STORAGE_KEY,
    value: cspConfig,
  };
}

module.exports = {
  name: CRON_CSP_CONFIG_NAME,
  cronTime: CRON_CSP_TIME,
  handler,
};
