// nordic-csp-engine/lib/headers-builder.js
const crypto = require('node:crypto');

const { DEFAULT_PLATFORM } = require('./constants/platform');

function generateCSPNonce() {
  return crypto.randomBytes(16).toString('base64');
}

/**
 * @typedef {import('frontend-platform_detection').Platform} Platform
 * @typedef {{traceRequestId: string, platform?: Platform,  userId?: number}} ContextData
 * @param {string} nonce
 * @param {ContextData} contextData
 */
function getPropertiesReplacement(nonce, { traceRequestId, userId }) {
  return [
    {
      regex: /\$nonce_id/g,
      value: nonce,
    },
    {
      regex: /\$request_id/g,
      value: traceRequestId,
    },
    {
      regex: /\$user_id/g,
      value: userId,
    },
  ];
}

/**
 * @typedef {import('frontend-platform_detection').PlatformId[]} PlatformIdList
 * @typedef {import('frontend-platform_detection').SiteId[]} SiteIdList
 * @typedef {{name: string, value: string, platforms: PlatformIdList, sites: SiteIdList}} CSPRawHeader
 * @typedef {{headers: CSPRawHeader[]}} CSPConfig
 */
class CSPHeadersBuilder {
  /**
   * @param {CSPConfig} cspConfig
   * @param {ContextData} contextData
   */
  constructor(cspConfig, contextData) {
    /**
     * @private
     * @type {CSPConfig}
     */
    this.cspConfig = cspConfig;
    /**
     * @private
     * @type {ContextData}
     */
    this.contextData = contextData;
    /**
     * @private
     * @type {string}
     */
    this.nonce = generateCSPNonce();
    /**
     * @private
     * @type {ReturnType<getPropertiesReplacement>}
     */
    this.propertiesReplacement = getPropertiesReplacement(
      this.nonce,
      contextData,
    );
  }

  /**
   * @public
   */
  getNonce() {
    return this.nonce;
  }

  /**
   * Build a CSP headers with directives for the given config
   * @public
   * @throws {Error} If the platform is not defined
   */
  build() {
    let { platform } = this.contextData;

    if (!platform) {
      platform = DEFAULT_PLATFORM;
    }

    const rawHeaders = this.filterCSPRawHeaders(platform);

    if (!rawHeaders.length) {
      throw new Error(
        `Empty headers for platform ${platform.id} and site ${platform.siteId}`,
      );
    }

    return this.buildCSPHeaders(rawHeaders);
  }

  /**
   * @private
   * @param {Platform} platform
   */
  filterCSPRawHeaders({ id: platformId, siteId }) {
    return this.cspConfig.headers.filter(
      ({ platforms, sites }) =>
        platforms.includes(platformId) && sites.includes(siteId),
    );
  }

  /**
   * @private
   * @param {CSPRawHeader[]} rawHeaders
   */
  buildCSPHeaders(rawHeaders) {
    return rawHeaders.map(({ name, value }) => ({
      name,
      value: this.buildHeaderValue(value),
    }));
  }

  /**
   * @private
   * @param {string} headerRawValue
   */
  buildHeaderValue(headerRawValue) {
    return this.propertiesReplacement.reduce(
      (acc, { regex, value }) => acc.replace(regex, value || ''),
      headerRawValue,
    );
  }
}

module.exports = CSPHeadersBuilder;
