// nordic-csp-engine/lib/middleware.js
const { MeliStatsD } = require('@meli/statsd');

const CSPHeadersBuilder = require('./headers-builder');
const {
  CRON_CSP_STORAGE_KEY,
  CRON_SECURITY_STORAGE_KEY,
} = require('./constants/cron');
const {
  TELEMETRY_CSP_MIDDLEWARE_ERROR,
  TELEMETRY_CSP_CONFIG_MISSING,
} = require('./constants/telemetry');
const { logger } = require('./logger');
const cronStorage = require('./cron/storage');

const statsd = new MeliStatsD();

/**
 * Checks if a path matches any backend path pattern
 * @param {string} url - The current request URL (can include query params)
 * @param {Array<{ type: 'exact' | 'prefix' | 'regex'; value: string | RegExp; }> | undefined} backendPaths - Array of backend path patterns
 * @returns {boolean} True if the path matches any backend path pattern
 */
function matchesBackendPath(url, backendPaths) {
  // Extract path without query params
  const path = url.split('?')[0];

  const matchers = {
    exact: (value) => path === value,
    prefix: (value) => path.startsWith(value),
    regex: (value) => {
      try {
        // eslint-disable-next-line security/detect-non-literal-regexp
        const regex = new RegExp(value);

        return regex.test(path);
      } catch (err) {
        return false;
      }
    },
  };

  return backendPaths
    .filter(({ value }) => typeof value === 'string')
    .some(({ type, value }) => matchers[type](value));
}

/**
 * Sets CSP (content security policy) headers on the response if applies.
 */
function cspMiddleware() {
  /** @type {import('express').RequestHandler} */
  return async function CSPMiddleware(req, res, next) {
    try {
      const { traceRequestId, platform, auth } = req;

      // Fetch both configurations in parallel
      const [securityConfig, cspConfig] = await Promise.all([
        cronStorage.get(CRON_SECURITY_STORAGE_KEY),
        cronStorage.get(CRON_CSP_STORAGE_KEY),
      ]);

      if (securityConfig) {
        const { strictPolicy } = securityConfig.csp;

        if (
          strictPolicy &&
          matchesBackendPath(req.originalUrl, strictPolicy.backendPaths)
        ) {
          res.append('Content-Security-Policy', strictPolicy.value);

          return;
        }
      } else if (
        matchesBackendPath(req.originalUrl, [
          // Accept /api or any /api/ segment (e.g. */api/*), but reject /apixxx
          { type: 'regex', value: '\\/api(?:\\/|$|\\?)' },
        ])
      ) {
        return;
      }

      if (!cspConfig) {
        statsd.increment(TELEMETRY_CSP_CONFIG_MISSING, {
          error: 'CSP_CONFIG_MISSING',
        });

        return;
      }

      const cspHeadersBuilder = new CSPHeadersBuilder(cspConfig, {
        traceRequestId,
        platform,
        userId: auth?.user?.id,
      });
      const cspHeaders = cspHeadersBuilder.build();

      cspHeaders.forEach(({ name, value }) => {
        res.append(name, value);
      });

      req.csp = {
        nonce: cspHeadersBuilder.getNonce(),
      };
    } catch (err) {
      logger.debug(err.message);

      statsd.increment(TELEMETRY_CSP_MIDDLEWARE_ERROR, {
        error: 'ERROR_SETTING_CSP_HEADERS',
      });
    } finally {
      next();
    }
  };
}

module.exports = cspMiddleware;
