// nordic-csp-engine/lib/security-config.js
const { env } = require('@nordic/env');
const { MeliStatsD } = require('@meli/statsd');
const {
  RestClient: createRestClient,
  buildFlowStarterContext,
} = require('frontend-restclient');

const {
  RESTCLIENT_WEB_SEC_SCOPE,
  RESTCLIENT_TIMEOUT,
  RESTCLIENT_WEB_SEC_BASE_URL,
} = require('./constants/restclient');
const {
  CRON_SECURITY_CONFIG_NAME,
  CRON_SECURITY_STORAGE_KEY,
  CRON_SECURITY_TIME,
} = require('./constants/cron');
const { TELEMETRY_SECURITY_CONFIG_ERROR } = require('./constants/telemetry');

/**
 * @typedef {Object} SecurityConfigAPI
 * @property {Object} input_validation
 * @property {boolean} input_validation.enabled
 * @property {string} [input_validation.deadline]
 * @property {'report' | 'warning' | 'block'} [input_validation.mode]
 * @property {string[]} [input_validation.allowed_modules]
 * @property {Object} csrf
 * @property {boolean} csrf.enabled
 * @property {Array<{ type: 'regex' | 'string'; value: string; }>} [csrf.ignore_paths]
 * @property {Object} trusted_types
 * @property {string[] | null} trusted_types.regexs
 * @property {Object} csp
 * @property {Object} [csp.strict_policy]
 * @property {string} csp.strict_policy.value
 * @property {Array<{ type: 'exact' | 'prefix' | 'regex'; value: string; }>} csp.strict_policy.backend_paths
 */

const restclient = createRestClient({
  baseURL: RESTCLIENT_WEB_SEC_BASE_URL,
  timeout: RESTCLIENT_TIMEOUT,
  retry: {
    maxRetries: 2,
  },
  logErrors: false,
});

const statsd = new MeliStatsD();

/**
 * Processes ignore paths for CSRF protection
 * @param {Array<{ type: 'regex' | 'string'; value: string; }>} ignoredPaths - Array of paths to be processed
 * @returns {Array<RegExp | string>} Array of processed paths (RegExp or string)
 * @note Strings are placed before RegExp for performance (exact/prefix matches are faster than regex)
 */
function processIgnoredPaths(ignoredPaths) {
  if (!ignoredPaths) {
    return [];
  }

  const strings = [];
  const regexes = [];

  for (const ignoredPath of ignoredPaths) {
    if (ignoredPath.type === 'regex') {
      try {
        // eslint-disable-next-line security/detect-non-literal-regexp
        regexes.push(new RegExp(ignoredPath.value));
      } catch (error) {
        statsd.increment(TELEMETRY_SECURITY_CONFIG_ERROR, {
          error: 'CSRF_INVALID_REGEX',
        });
      }
    } else {
      strings.push(ignoredPath.value);
    }
  }

  // Strings first (faster), then regexes (slower)
  return [...strings, ...regexes];
}

/**
 * Processes trusted types regex patterns into a single RegExp
 * @param {Array<string>} regexs - Array of regex patterns to be combined
 * @returns {string|null} A single string pattern that combines all regexes or null if invalid
 */
function processTrustedTypesRegexs(regexs) {
  if (!regexs || regexs.length === 0) {
    return null;
  }

  try {
    const regex = regexs.join('|');

    // Validate the regex pattern by creating a RegExp instance
    // but return the string pattern as the consuming system expects a string
    // eslint-disable-next-line no-new, security/detect-non-literal-regexp
    new RegExp(regex);

    return regex;
  } catch (error) {
    statsd.increment(TELEMETRY_SECURITY_CONFIG_ERROR, {
      error: 'TRUSTED_TYPES_INVALID_REGEX',
    });

    return null;
  }
}

/**
 * Processes backend paths for strict policy CSP
 * @param {Array<{ type: 'exact' | 'prefix' | 'regex'; value: string; }  | undefined>} backendPaths - Array of backend paths to be processed
 * @returns {Array<{ type: 'exact' | 'prefix' | 'regex'; value: string | RegExp; }>} Array of processed backend paths
 * @note Paths are ordered by comparison speed: exact, prefix, regex for optimal performance
 */
function processBackendPaths(backendPaths) {
  const exactPaths = [];
  const prefixPaths = [];
  const regexPaths = [];

  for (const backendPath of backendPaths) {
    if (!backendPath?.type || !backendPath?.value) {
      return [];
    }

    const { type, value } = backendPath;

    switch (type) {
      case 'exact':
        exactPaths.push({ type, value });

        break;

      case 'prefix':
        prefixPaths.push({ type, value });

        break;

      case 'regex':
        try {
          RegExp(value);

          // Store raw value; it will be converted lazily where it's used
          regexPaths.push({ type, value });
        } catch (error) {
          statsd.increment(TELEMETRY_SECURITY_CONFIG_ERROR, {
            error: 'BACKEND_PATHS_INVALID_REGEX',
          });
        }

        break;

      default:
        // Ignore unknown types
        break;
    }
  }

  // Order by comparison speed
  return [...exactPaths, ...prefixPaths, ...regexPaths];
}

/**
 * Processes strict policy configuration for CSP
 * @param {Object} strictPolicy - Raw strict policy configuration
 * @param {string} strictPolicy.value - CSP policy value
 * @param {Array<{ type: 'exact' | 'prefix' | 'regex'; value: string; }>} strictPolicy.backend_paths - Backend paths where policy applies
 * @returns {Object|undefined} Processed strict policy or undefined if invalid
 */
function processStrictPolicy(strictPolicy) {
  if (!strictPolicy) {
    return undefined;
  }

  if (!strictPolicy.value) {
    statsd.increment(TELEMETRY_SECURITY_CONFIG_ERROR, {
      error: 'CSP_STRICT_POLICY_MISSING_VALUE',
    });

    return undefined;
  }

  if (!strictPolicy.backend_paths) {
    statsd.increment(TELEMETRY_SECURITY_CONFIG_ERROR, {
      error: 'CSP_STRICT_POLICY_MISSING_BACKEND_PATHS',
    });

    return undefined;
  }

  const processedBackendPaths = processBackendPaths(strictPolicy.backend_paths);

  if (processedBackendPaths.length === 0) {
    statsd.increment(TELEMETRY_SECURITY_CONFIG_ERROR, {
      error: 'CSP_STRICT_POLICY_INVALID_BACKEND_PATHS',
    });

    return undefined;
  }

  return {
    value: strictPolicy.value,
    backendPaths: processedBackendPaths,
  };
}

/**
 * Processes the raw security configuration
 * @param {SecurityConfigAPI} config - Raw configuration from the server
 * @returns {import('./types').WebsecConfig} Processed configuration
 */
function processSecurityConfig(config) {
  const result = {
    inputValidation: {
      enabled: config.input_validation.enabled,
      deadline: config.input_validation.deadline,
      mode: config.input_validation.mode,
      allowedModules: config.input_validation.allowed_modules,
    },
    csrf: {
      enabled: config.csrf.enabled,
      ignoredPaths: config.csrf.ignored_paths
        ? processIgnoredPaths(config.csrf.ignored_paths)
        : [],
    },
    trustedTypes: {
      regex: config.trusted_types.regexs
        ? processTrustedTypesRegexs(config.trusted_types.regexs)
        : null,
    },
    csp: {
      strictPolicy: processStrictPolicy(config.csp.strict_policy),
    },
  };

  return result;
}

async function handler() {
  const { data } = await restclient.get(
    `/applications/${env.APPLICATION}/scopes/${RESTCLIENT_WEB_SEC_SCOPE}/configurations`,
    {
      context: buildFlowStarterContext(),
    },
  );

  if (!data) {
    statsd.increment(TELEMETRY_SECURITY_CONFIG_ERROR, {
      error: 'EMPTY_CONFIG_RECEIVED',
    });

    return null;
  }

  return {
    key: CRON_SECURITY_STORAGE_KEY,
    value: processSecurityConfig(data),
  };
}

module.exports = {
  name: CRON_SECURITY_CONFIG_NAME,
  cronTime: CRON_SECURITY_TIME,
  handler,
};
