// @andes/themes/build/esm/index.development.js
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import React, { useContext, createContext, useMemo } from 'react';
import PropTypes from 'prop-types';
import { useId } from '@andes/context';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import { registerMetric } from '@andes/themes/build/metrics';
const ThemeContext = /*#__PURE__*/createContext(/** @type {import('../..').ThemeContextValue} */{});

/** @type {typeof import('../..').useThemeContext} */
const useThemeContext = () => useContext(ThemeContext);
const andesThemePrefix = 'andes';
const componentVariantPrefix = 'component-variant-';

/**
 * @param {unknown} item
 * @returns { item is Record<keyof any, unknown> }
 */
const isObject = item => {
  if (typeof item !== 'object' || item === null) {
    return false;
  }
  const prototype = Object.getPrototypeOf(item);
  return (prototype === null || prototype === Object.prototype || Object.getPrototypeOf(prototype) === null) && !(Symbol.toStringTag in item) && !(Symbol.iterator in item);
};

/**
 * @param {unknown} source
 */
const cloneDeep = source => {
  if (!isObject(source)) {
    return source;
  }

  /** @type {Record<keyof any, unknown>} */
  const output = {};
  Object.keys(source).forEach(key => {
    output[key] = cloneDeep(source[key]);
  });
  return output;
};

/**
 * @typedef {Object} MergeOptions
 * @prop {boolean} [clone]
 * @prop {boolean} [allowOnlySourceKeys]
 */

/**
 * @param {Record<string, any>} target
 * @param {Record<string, any>} source
 * @param {MergeOptions} [options]
 */
const mergeDeep = function (target, source) {
  let options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {
    clone: true,
    allowOnlySourceKeys: false
  };
  const _options$clone = options.clone,
    clone = _options$clone === void 0 ? true : _options$clone,
    _options$allowOnlySou = options.allowOnlySourceKeys,
    allowOnlySourceKeys = _options$allowOnlySou === void 0 ? false : _options$allowOnlySou;
  const output = clone ? _objectSpread({}, target) : target;
  if (isObject(target) && isObject(source)) {
    if (allowOnlySourceKeys) {
      Object.keys(output).forEach(key => {
        if (!(key in source)) {
          delete output[key];
        }
      });
    }
    Object.keys(source).forEach(key => {
      if (key === '__proto__') {
        return;
      }
      if (isObject(source[key]) && key in target && isObject(target[key])) {
        // @ts-ignore
        output[key] = mergeDeep(target[key], source[key], options);
      } else if (clone) {
        output[key] = isObject(source[key]) ? cloneDeep(source[key]) : source[key];
      } else {
        output[key] = source[key];
      }
    });
  }
  return output;
};

/** @param {string} str */
const camelCaseToKebabCase = str => typeof str === 'string' ? str.replace(/[A-Z]/g, x => `-${x.toLowerCase()}`) : '';

/**
 * @param {string} prop
 * @param {string} value
 */
const toCssValue = (prop, value) => `${camelCaseToKebabCase(prop)}: ${value};`;

/**
 * @param {Record<string, any>} target
 * @param {Record<string, any>} source
 */
const diffDeep = function () {
  let target = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let source = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  const keysInTargetAndSource = /** @type {Record<string, any>} */{};
  const keysOnlyInTarget = /** @type {Record<string, any>} */{};

  /**
   * @param {Record<string, any>} targetObj
   * @param {Record<string, any>} sourceObj
   * @param {Record<string, any>} keysInBothAcc
   * @param {Record<string, any>} keysInTargetAcc
   */
  const findCommonKeys = (targetObj, sourceObj, keysInBothAcc, keysInTargetAcc) => {
    Object.entries(targetObj).forEach(_ref => {
      let _ref2 = _slicedToArray(_ref, 2),
        key = _ref2[0],
        valueTargetObjEntry = _ref2[1];
      const valueSourceObjEntry = sourceObj[key];
      if (valueSourceObjEntry !== undefined) {
        if (isObject(valueTargetObjEntry) && isObject(valueSourceObjEntry)) {
          const targetSubObj = {};
          const sourceSubObj = {};
          findCommonKeys(valueTargetObjEntry, valueSourceObjEntry, targetSubObj, sourceSubObj);
          if (Object.keys(targetSubObj).length > 0) {
            keysInBothAcc[key] = targetSubObj;
          }
          if (Object.keys(sourceSubObj).length > 0) {
            keysInTargetAcc[key] = sourceSubObj;
          }
        } else if (typeof valueTargetObjEntry === typeof valueSourceObjEntry) {
          keysInBothAcc[key] = valueTargetObjEntry;
        } else {
          keysInTargetAcc[key] = valueTargetObjEntry;
        }
      } else {
        keysInTargetAcc[key] = valueTargetObjEntry;
      }
    });
  };
  findCommonKeys(target, source, keysInTargetAndSource, keysOnlyInTarget);
  return {
    keysInTargetAndSource,
    keysOnlyInTarget
  };
};

/** @param {string} name */
const toVarReference = name => `var(${name})`;

/** @param {string} value */
const toVarDefinition = function (value) {
  let prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
  return `--${prefix ? `${prefix}-` : ''}${value}`;
};

/** @param {string} value */
const isVarReference = value => typeof value === 'string' && value.startsWith('{');

/** @param {string} value */
const formatVarReferenceName = value => typeof value === 'string' ? value.slice(1, -1).split('.').join('-') : '';

/** @param {string} value */
const extractPrefix = value => typeof value === 'string' ? value.split('-')[0] || '' : '';

/**
 * @param {string} value
 * @param {Record<string, string>} varReferences
 */
const resolveVarReference = (value, varReferences) => {
  const valueAsToken = camelCaseToKebabCase(value);
  const referenceValue = varReferences[valueAsToken];
  if (referenceValue) {
    return toVarReference(toVarDefinition(valueAsToken));
  }
  return value;
};

/**
 * @param {Record<string, any>} obj
 * @param {string} prefix
 *
 * @returns {Record<string, unknown>}
 */
const toVarNames = function (obj) {
  let prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
  return Object.entries(obj).reduce((vars, _ref) => {
    let _ref2 = _slicedToArray(_ref, 2),
      key = _ref2[0],
      value = _ref2[1];
    const keyName = key === 'base' ? `${prefix}` : `${prefix}-${key}`;
    if (typeof value === 'object') {
      vars[key] = toVarNames(value, `${prefix ? `${prefix}-` : ''}${key}`);
    } else {
      vars[key] = toVarReference(toVarDefinition(camelCaseToKebabCase(keyName)));
    }
    return vars;
  }, /** @type {Record<string, unknown>} */{});
};

/**
 * @param {Record<string, any>} obj
 * @param {string} prefix
 *
 * @returns {Record<string, string>}
 */
const toVarsMap = function () {
  let obj = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
  return Object.entries(obj).reduce((vars, _ref3) => {
    let _ref4 = _slicedToArray(_ref3, 2),
      key = _ref4[0],
      value = _ref4[1];
    if (typeof value === 'object') {
      const nestedVars = toVarsMap(value, `${prefix ? `${prefix}-` : ''}${key}`);
      return _objectSpread(_objectSpread({}, vars), nestedVars);
    }
    const varValue = isVarReference(value) ? `${extractPrefix(prefix)}-${formatVarReferenceName(value).replace('-base', '')}` : value;
    const keyName = key === 'base' ? `${prefix}` : `${prefix}-${key}`;
    return _objectSpread(_objectSpread({}, vars), {}, {
      [camelCaseToKebabCase(keyName)]: varValue
    });
  }, {});
};

/**
 * @param {Record<string, string>} obj
 * @param {Record<string, string>} varReferences
 *
 * @returns {string}
 */
const toCssVars = function () {
  let obj = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let varReferences = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return Object.entries(obj).map(_ref5 => {
    let _ref6 = _slicedToArray(_ref5, 2),
      key = _ref6[0],
      value = _ref6[1];
    const varName = toVarDefinition(key);
    const varValue = (typeof value === 'string' ? value.startsWith(extractPrefix(key)) : '') ? resolveVarReference(value, varReferences) : value;
    return `${varName}: ${varValue};`;
  }).join(' ');
};

/** @param {string} value */
const isAndesCssVar = value => typeof value === 'string' && value.startsWith(`--${andesThemePrefix}`);
const _excluded = ["styleOverrides", "variants"];
const separateComponentStylesFromComponentEntries = function () {
  let componentEntriesObj = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  const styleOverrides = componentEntriesObj.styleOverrides,
    _componentEntriesObj$ = componentEntriesObj.variants,
    componentVariants = _componentEntriesObj$ === void 0 ? [] : _componentEntriesObj$,
    componentStyles = _objectWithoutProperties(componentEntriesObj, _excluded);
  const componentOverrides = typeof styleOverrides === 'object' ? {
    styleOverrides
  } : {};
  return {
    componentStyles,
    componentOverrides,
    componentVariants
  };
};

/** @param {Record<string, any>} customStyles */
const separateCssStylesFromCustomVars = customStyles => {
  const cssStyles = Object.fromEntries(Object.entries(customStyles).filter(_ref => {
    let _ref2 = _slicedToArray(_ref, 2),
      key = _ref2[0],
      value = _ref2[1];
    return typeof value !== 'object' && !isAndesCssVar(key);
  }));
  const otherStyles = Object.fromEntries(Object.entries(customStyles).filter(_ref3 => {
    let _ref4 = _slicedToArray(_ref3, 2),
      value = _ref4[1];
    return typeof value === 'object';
  }));
  return {
    cssStyles,
    otherStyles
  };
};

/** @typedef {import('../../..').AndesProviderTheme} Theme */

/**
 * @param {Theme["colorSchemes"]} colorSchemes
 * @param {(colorScheme: Record<string, any>) => Record<string, unknown>} varMapper
 *
 * @returns {{ semanticLightVars: Record<string, unknown>, semanticDarkVars: Record<string, unknown> }}
 */
const getColorSchemeVars = (colorSchemes, varMapper) => {
  let semanticLightVars = /** @type {Record<string, unknown>} */{};
  let semanticDarkVars = /** @type {Record<string, unknown>} */{};
  if (typeof colorSchemes === 'object') {
    if (typeof colorSchemes.light === 'object') {
      semanticLightVars = varMapper(colorSchemes.light);
    }

    // @ts-expect-error dark is not a property of colorSchemes until now
    if (typeof colorSchemes.dark === 'object') {
      // @ts-expect-error dark is not a property of colorSchemes until now
      semanticDarkVars = varMapper(colorSchemes.dark);
    }
  }
  return {
    semanticLightVars,
    semanticDarkVars
  };
};

/**
 * @param {Theme} theme
 * @param {string} prefix
 * @param {(obj: Record<string, unknown>, prefix: string) => Record<string, unknown>} varMapper
 */
const mapCoreAndSemanticVars = (theme, prefix, varMapper) => {
  const _theme$core = theme.core,
    core = _theme$core === void 0 ? {} : _theme$core,
    _theme$colorSchemes = theme.colorSchemes,
    colorSchemes = _theme$colorSchemes === void 0 ? {} : _theme$colorSchemes;

  // core tier variables
  const coreVars = varMapper(core, prefix);

  // semantic tier variables
  const _getColorSchemeVars = getColorSchemeVars(colorSchemes, colorScheme => varMapper(colorScheme, prefix)),
    semanticLightVars = _getColorSchemeVars.semanticLightVars;
  return {
    coreVars,
    semanticLightVars
  };
};

/**
 * @param {Theme} theme
 * @param {string} prefix

 */
const createThemeVarsMaps = (theme, prefix) => {
  const _mapCoreAndSemanticVa = mapCoreAndSemanticVars(theme, prefix, toVarsMap),
    coreVars = _mapCoreAndSemanticVa.coreVars,
    semanticLightVars = _mapCoreAndSemanticVa.semanticLightVars;

  // component tier variables mapper

  const componentVarsMapper = (/** @type {Record<string, any>} */componentStyles) => toVarsMap(componentStyles, prefix);
  return (
    /** @type {{
    coreVars: Record<string, string>,
    semanticLightVars: Record<string, string>,
    componentVarsMapper: (componentStyles: Record<string, any>) => Record<string, string>
    }} */
    {
      coreVars,
      semanticLightVars,
      componentVarsMapper
    }
  );
};

/**
 * @param {Theme} theme
 * @param {string} prefix
 */
const createThemeVarsNames = (theme, prefix) => {
  const _mapCoreAndSemanticVa2 = mapCoreAndSemanticVars(theme, prefix, toVarNames),
    coreVars = _mapCoreAndSemanticVa2.coreVars,
    semanticLightVars = _mapCoreAndSemanticVa2.semanticLightVars;
  return {
    core: coreVars,
    semantic: semanticLightVars
  };
};
var core$1 = {
  color: {
    blue: {
      "100": "rgba(65, 137, 230, 0.1)",
      "150": "rgba(65, 137, 230, 0.15)",
      "200": "rgba(65, 137, 230, 0.2)",
      "300": "rgba(65, 137, 230, 0.3)",
      "400": "rgba(65, 137, 230, 0.4)",
      "500": "#3483fa",
      "600": "#2968c8",
      "700": "#1f4e96",
      "800": "#183c73"
    },
    cyan: {
      "100": "rgba(71, 154, 209, 0.1)",
      "150": "rgba(71, 154, 209, 0.15)",
      "200": "rgba(71, 154, 209, 0.2)",
      "300": "rgba(71, 154, 209, 0.3)",
      "400": "rgba(71, 154, 209, 0.4)",
      "500": "#009ee3",
      "600": "#007eb5",
      "700": "#005e88",
      "800": "#004766"
    },
    white: {
      "300": "rgba(255, 255, 255, 0.3)",
      base: "#ffffff"
    },
    gray: {
      "40": "rgba(0, 0, 0, 0.04)",
      "70": "rgba(0, 0, 0, 0.07)",
      "100": "rgba(0, 0, 0, 0.1)",
      "150": "rgba(0, 0, 0, 0.15)",
      "250": "rgba(0, 0, 0, 0.25)",
      "450": "rgba(0, 0, 0, 0.45)",
      "550": "rgba(0, 0, 0, 0.55)",
      "800": "rgba(0, 0, 0, 0.8)",
      "900": "rgba(0, 0, 0, 0.9)",
      solid: {
        "40": "#f5f5f5",
        "70": "#ededed",
        "100": "#e5e5e5",
        "150": "#d9d9d9",
        "250": "#bfbfbf",
        "450": "#8c8c8c",
        "550": "#737373",
        "800": "#333333",
        "900": "#1a1a1a"
      }
    },
    transparent: "rgba(255, 255, 255, 0)",
    green: {
      "100": "rgba(0, 166, 80, 0.1)",
      "150": "rgba(0, 166, 80, 0.15)",
      "200": "rgba(0, 166, 80, 0.2)",
      "300": "rgba(0, 166, 80, 0.3)",
      "400": "rgba(0, 166, 80, 0.4)",
      "500": "#00a650",
      "600": "#008744",
      "700": "#006633",
      "800": "#004d27"
    },
    orange: {
      "100": "rgba(255, 119, 51, 0.1)",
      "150": "rgba(255, 119, 51, 0.15)",
      "200": "rgba(255, 119, 51, 0.2)",
      "300": "rgba(255, 119, 51, 0.3)",
      "400": "rgba(255, 119, 51, 0.4)",
      "500": "#ff7733",
      "600": "#e6540b",
      "700": "#cc3e0a",
      "800": "#a62a08"
    },
    red: {
      "100": "rgba(242, 61, 79, 0.1)",
      "150": "rgba(242, 61, 79, 0.15)",
      "200": "rgba(242, 61, 79, 0.2)",
      "300": "rgba(242, 61, 79, 0.3)",
      "400": "rgba(242, 61, 79, 0.4)",
      "500": "#f23d4f",
      "600": "#d12440",
      "700": "#a61d33",
      "800": "#801627"
    },
    yellow: {
      "500": "#ffe600"
    }
  }
};
var colorSchemes$1 = {
  light: {
    color: {
      background: {
        primary: "{color.gray.solid.70}",
        secondary: "{color.gray.solid.40}",
        tertiary: "{color.white.base}",
        overlay: "{color.gray.800}"
      },
      surface: {
        primary: "{color.white.base}",
        secondary: "{color.gray.solid.40}",
        tertiary: "{color.gray.solid.70}",
        hover: "{color.gray.40}",
        active: "{color.gray.70}"
      },
      fill: {
        primary: "{color.white.base}",
        secondary: "{color.gray.40}",
        tertiary: "{color.gray.100}",
        inverse: "{color.gray.solid.900}",
        disabled: "{color.gray.100}",
        accent: "{color.blue.500}"
      },
      text: {
        primary: "{color.gray.900}",
        secondary: "{color.gray.550}",
        accent: "{color.blue.500}",
        inverse: "{color.white.base}",
        disabled: "{color.gray.250}",
        link: {
          "default": {
            base: "{color.blue.500}",
            inverse: "{color.white.base}"
          },
          hover: {
            base: "{color.blue.600}",
            inverse: "{color.white.base}"
          },
          active: {
            base: "{color.blue.700}",
            inverse: "{color.white.base}"
          }
        }
      },
      border: {
        primary: "{color.gray.100}",
        inverse: "{color.white.base}",
        disabled: "{color.gray.250}"
      },
      icon: {
        primary: "{color.gray.900}",
        secondary: "{color.gray.550}",
        accent: "{color.blue.500}",
        inverse: "{color.white.base}",
        disabled: "{color.gray.250}"
      },
      feedback: {
        fill: {
          informative: {
            loud: {
              base: "{color.blue.500}",
              onFill: "{color.blue.600}"
            },
            quiet: "{color.blue.100}"
          },
          positive: {
            loud: {
              base: "{color.green.500}",
              onFill: "{color.green.600}"
            },
            quiet: "{color.green.100}"
          },
          caution: {
            loud: {
              base: "{color.orange.500}",
              onFill: "{color.orange.600}"
            },
            quiet: "{color.orange.100}"
          },
          negative: {
            loud: {
              base: "{color.red.500}",
              onFill: "{color.red.600}"
            },
            quiet: "{color.red.100}"
          }
        },
        text: {
          informative: {
            loud: "{color.blue.500}"
          },
          positive: {
            loud: "{color.green.500}"
          },
          caution: {
            loud: "{color.orange.500}"
          },
          negative: {
            loud: "{color.red.500}"
          }
        },
        border: {
          informative: {
            loud: {
              base: "{color.blue.500}",
              onFill: "{color.blue.600}"
            }
          },
          positive: {
            loud: {
              base: "{color.green.500}",
              onFill: "{color.green.600}"
            }
          },
          caution: {
            loud: {
              base: "{color.orange.500}",
              onFill: "{color.orange.600}"
            }
          },
          negative: {
            loud: {
              base: "{color.red.500}",
              onFill: "{color.red.600}"
            }
          }
        },
        icon: {
          informative: {
            loud: "{color.blue.500}"
          },
          positive: {
            loud: "{color.green.500}"
          },
          caution: {
            loud: "{color.orange.500}"
          },
          negative: {
            loud: "{color.red.500}"
          }
        }
      },
      interactive: {
        fill: {
          loud: {
            "default": "{color.blue.500}",
            hover: "{color.blue.600}",
            active: "{color.blue.700}"
          },
          quiet: {
            "default": "{color.blue.150}",
            hover: "{color.blue.200}",
            active: "{color.blue.300}"
          },
          mute: {
            "default": "{color.transparent}",
            hover: "{color.blue.100}",
            active: "{color.blue.200}"
          }
        },
        border: {
          "default": "{color.gray.250}",
          hover: "{color.gray.550}",
          active: "{color.blue.500}"
        },
        icon: {
          "default": {
            base: "{color.gray.550}",
            accent: "{color.blue.500}",
            inverse: "{color.white.base}"
          },
          hover: {
            base: "{color.gray.800}",
            accent: "{color.blue.600}"
          },
          active: {
            base: "{color.gray.900}",
            accent: "{color.blue.700}"
          }
        }
      },
      selected: {
        fill: {
          "default": "{color.blue.500}",
          hover: "{color.blue.600}",
          active: "{color.blue.700}"
        },
        text: {
          "default": "{color.blue.500}",
          hover: "{color.blue.600}",
          active: "{color.blue.700}"
        },
        border: {
          "default": "{color.blue.500}",
          hover: "{color.blue.600}",
          active: "{color.blue.700}"
        },
        icon: {
          "default": "{color.blue.500}",
          hover: "{color.blue.600}",
          active: "{color.blue.700}"
        }
      },
      brand: {
        text: {
          onFill: "{color.gray.900}"
        },
        icon: {
          onFill: "{color.gray.900}"
        },
        fill: "{color.yellow.500}"
      }
    }
  }
};
var components$1 = {
  accordion: {
    color: {
      surface: {
        "default": "{color.transparent}",
        hover: "{color.surface.hover}",
        active: "{color.surface.active}",
        disabled: "{color.transparent}"
      },
      title: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      },
      description: {
        "default": "{color.text.secondary}",
        disabled: "{color.text.disabled}"
      },
      border: "{color.border.primary}"
    }
  },
  amountField: {
    color: {
      currency: {
        "default": "{color.text.primary}",
        active: "{color.text.primary}",
        error: "{color.text.primary}",
        errorShake: "{color.feedback.text.negative.loud}"
      },
      amount: {
        "default": "{color.text.primary}",
        active: "{color.text.primary}",
        error: "{color.text.primary}",
        errorShake: "{color.feedback.text.negative.loud}"
      },
      suffix: {
        "default": "{color.text.secondary}",
        active: "{color.text.secondary}",
        error: "{color.text.secondary}",
        errorShake: "{color.feedback.text.negative.loud}"
      },
      percentage: {
        "default": "{color.text.primary}",
        active: "{color.text.primary}",
        error: "{color.text.primary}",
        errorShake: "{color.feedback.text.negative.loud}"
      },
      placeholder: {
        "default": "{color.text.secondary}"
      }
    },
    converter: {
      color: {
        equivalence: {
          "default": "{color.text.secondary}"
        },
        suffix: {
          "default": "{color.text.secondary}"
        }
      },
      swapButton: {
        color: {
          fill: {
            "default": "{color.interactive.fill.quiet.default}",
            hover: "{color.interactive.fill.quiet.hover}",
            active: "{color.interactive.fill.quiet.active}"
          },
          icon: {
            "default": "{color.icon.accent}"
          }
        }
      }
    }
  },
  badge: {
    loud: {
      color: {
        fill: {
          informative: {
            base: "{color.feedback.fill.informative.loud.base}",
            onComplexBg: "{color.feedback.fill.informative.loud.onFill}"
          },
          positive: {
            base: "{color.feedback.fill.positive.loud.base}",
            onComplexBg: "{color.feedback.fill.positive.loud.onFill}"
          },
          caution: {
            base: "{color.feedback.fill.caution.loud.base}",
            onComplexBg: "{color.feedback.fill.caution.loud.onFill}"
          },
          negative: {
            base: "{color.feedback.fill.negative.loud.base}",
            onComplexBg: "{color.feedback.fill.negative.loud.onFill}"
          },
          neutral: {
            base: "{color.gray.solid.550}",
            onComplexBg: "{color.gray.solid.900}"
          },
          inverse: "{color.fill.primary}"
        },
        text: {
          informative: {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          positive: {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          caution: {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          negative: {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          neutral: {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          inverse: "{color.text.primary}"
        },
        icon: {
          informative: {
            base: "{color.icon.inverse}",
            onComplexBg: "{color.icon.inverse}"
          },
          positive: {
            base: "{color.icon.inverse}",
            onComplexBg: "{color.icon.inverse}"
          },
          caution: {
            base: "{color.icon.inverse}",
            onComplexBg: "{color.icon.inverse}"
          },
          negative: {
            base: "{color.icon.inverse}",
            onComplexBg: "{color.icon.inverse}"
          },
          neutral: {
            base: "{color.icon.inverse}",
            onComplexBg: "{color.icon.inverse}"
          },
          inverse: "{color.icon.primary}"
        }
      }
    },
    quiet: {
      color: {
        fill: {
          informative: "{color.feedback.fill.informative.quiet}",
          positive: "{color.feedback.fill.positive.quiet}",
          caution: "{color.feedback.fill.caution.quiet}",
          negative: "{color.feedback.fill.negative.quiet}",
          neutral: "{color.gray.100}"
        },
        text: {
          informative: "{color.feedback.text.informative.loud}",
          positive: "{color.feedback.text.positive.loud}",
          caution: "{color.feedback.text.caution.loud}",
          negative: "{color.feedback.text.negative.loud}",
          neutral: "{color.text.secondary}"
        },
        icon: {
          informative: "{color.feedback.icon.informative.loud}",
          positive: "{color.feedback.icon.positive.loud}",
          caution: "{color.feedback.icon.caution.loud}",
          negative: "{color.feedback.icon.negative.loud}",
          neutral: "{color.gray.solid.550}"
        }
      }
    },
    dot: {
      quiet: {
        color: {
          fill: {
            informative: "{color.blue.400}",
            positive: "{color.green.400}",
            caution: "{color.orange.400}",
            negative: "{color.red.400}",
            neutral: "{color.gray.250}"
          }
        }
      }
    }
  },
  bottomSheet: {
    header: {
      color: {
        fill: "{color.fill.primary}",
        drag: {
          fill: "{color.gray.250}"
        },
        title: "{color.text.primary}",
        back: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}"
        },
        close: {
          "default": {
            base: "{color.interactive.icon.default.base}",
            inverse: "{color.interactive.icon.default.inverse}"
          },
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}"
        }
      }
    },
    color: {
      fill: "{color.fill.primary}"
    },
    dimmer: {
      color: {
        background: "{color.background.overlay}"
      }
    }
  },
  boxSelector: {
    color: {
      selected: {
        border: {
          "default": "{color.selected.border.default}",
          hover: "{color.selected.border.hover}",
          active: "{color.selected.border.active}",
          disabled: "{color.gray.100}"
        }
      },
      unselected: {
        border: {
          "default": "{color.interactive.border.default}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.gray.100}"
        }
      }
    }
  },
  breadcrumb: {
    currentLevel: {
      color: {
        text: "{color.text.primary}"
      }
    },
    fatherLevel: {
      color: {
        text: {
          "default": "{color.text.link.default.base}",
          hover: "{color.text.link.hover.base}",
          active: "{color.text.link.active.base}"
        }
      }
    },
    color: {
      chevron: "{color.icon.secondary}"
    }
  },
  button: {
    loud: {
      color: {
        fill: {
          "default": {
            base: "{color.interactive.fill.loud.default}",
            onComplexBg: "{color.gray.450}"
          },
          hover: {
            base: "{color.interactive.fill.loud.hover}",
            onComplexBg: "{color.gray.550}"
          },
          active: {
            base: "{color.interactive.fill.loud.active}",
            onComplexBg: "{color.gray.800}"
          },
          progress: {
            base: "{color.blue.600}",
            onComplexBg: "{color.gray.550}"
          },
          disabled: {
            base: "{color.fill.disabled}",
            onComplexBg: "{color.gray.150}"
          }
        },
        text: {
          "default": {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          disabled: {
            base: "{color.text.disabled}",
            onComplexBg: "{color.text.disabled}"
          }
        },
        spinner: {
          base: "{color.fill.primary}",
          onComplexBg: "{color.fill.primary}"
        }
      }
    },
    quiet: {
      color: {
        fill: {
          "default": {
            base: "{color.interactive.fill.quiet.default}",
            onComplexBg: "{color.gray.150}"
          },
          hover: {
            base: "{color.interactive.fill.quiet.hover}",
            onComplexBg: "{color.gray.250}"
          },
          active: {
            base: "{color.interactive.fill.quiet.active}",
            onComplexBg: "{color.gray.450}"
          },
          progress: {
            base: "{color.blue.300}",
            onComplexBg: "{color.gray.250}"
          },
          disabled: {
            base: "{color.fill.disabled}",
            onComplexBg: "{color.gray.150}"
          }
        },
        text: {
          "default": {
            base: "{color.text.accent}",
            onComplexBg: "{color.text.inverse}"
          },
          disabled: {
            base: "{color.text.disabled}",
            onComplexBg: "{color.text.disabled}"
          }
        },
        spinner: {
          base: "{color.feedback.fill.informative.loud.base}",
          onComplexBg: "{color.fill.primary}"
        }
      }
    },
    mute: {
      color: {
        fill: {
          "default": {
            base: "{color.interactive.fill.mute.default}",
            onComplexBg: "{color.transparent}"
          },
          hover: {
            base: "{color.interactive.fill.mute.hover}",
            onComplexBg: "{color.gray.70}"
          },
          active: {
            base: "{color.interactive.fill.mute.active}",
            onComplexBg: "{color.gray.150}"
          },
          disabled: {
            base: "{color.transparent}",
            onComplexBg: "{color.transparent}"
          }
        },
        text: {
          "default": {
            base: "{color.text.accent}",
            onComplexBg: "{color.text.inverse}"
          },
          disabled: {
            base: "{color.text.disabled}",
            onComplexBg: "{color.text.disabled}"
          }
        },
        spinner: {
          base: "{color.feedback.fill.informative.loud.base}",
          onComplexBg: "{color.fill.primary}"
        }
      }
    },
    icon: {
      loud: {
        color: {
          icon: {
            "default": {
              base: "{color.icon.inverse}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      },
      quiet: {
        color: {
          icon: {
            "default": {
              base: "{color.icon.accent}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      },
      mute: {
        color: {
          icon: {
            "default": {
              base: "{color.icon.accent}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      }
    },
    split: {
      loud: {
        color: {
          fill: {
            "default": {
              base: "{color.interactive.fill.loud.default}",
              onComplexBg: "{color.gray.450}"
            },
            hover: {
              base: "{color.interactive.fill.loud.hover}",
              onComplexBg: "{color.gray.550}"
            },
            active: {
              base: "{color.interactive.fill.loud.active}",
              onComplexBg: "{color.gray.800}"
            },
            disabled: {
              base: "{color.fill.disabled}",
              onComplexBg: "{color.gray.150}"
            }
          },
          chevron: {
            "default": {
              base: "{color.icon.inverse}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      },
      quiet: {
        color: {
          fill: {
            "default": {
              base: "{color.interactive.fill.quiet.default}",
              onComplexBg: "{color.gray.150}"
            },
            hover: {
              base: "{color.interactive.fill.quiet.hover}",
              onComplexBg: "{color.gray.250}"
            },
            active: {
              base: "{color.interactive.fill.quiet.active}",
              onComplexBg: "{color.gray.450}"
            },
            disabled: {
              base: "{color.fill.disabled}",
              onComplexBg: "{color.gray.150}"
            }
          },
          chevron: {
            "default": {
              base: "{color.icon.accent}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      },
      color: {
        border: {
          "default": "{color.border.primary}"
        }
      }
    },
    dropdown: {
      loud: {
        color: {
          chevron: {
            "default": {
              base: "{color.icon.inverse}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      },
      quiet: {
        color: {
          chevron: {
            "default": {
              base: "{color.icon.accent}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      }
    }
  },
  card: {
    container: {
      color: {
        surface: {
          primary: "{color.surface.primary}",
          secondary: "{color.surface.secondary}",
          tertiary: "{color.surface.tertiary}"
        }
      }
    },
    color: {
      surface: {
        primary: {
          "default": "{color.transparent}",
          hover: "{color.surface.hover}",
          active: "{color.surface.active}"
        },
        secondary: {
          "default": "{color.transparent}",
          hover: "{color.surface.hover}",
          active: "{color.surface.active}"
        },
        tertiary: {
          "default": "{color.transparent}",
          hover: "{color.surface.hover}",
          active: "{color.surface.active}"
        }
      },
      border: "{color.border.primary}"
    },
    pipe: {
      color: {
        fill: {
          informative: "{color.feedback.fill.informative.loud.base}",
          positive: "{color.feedback.fill.positive.loud.base}",
          caution: "{color.feedback.fill.caution.loud.base}",
          negative: "{color.feedback.fill.negative.loud.base}"
        }
      }
    },
    extended: {
      header: {
        color: {
          title: "{color.text.primary}",
          border: "{color.border.primary}"
        }
      },
      footer: {
        color: {
          chevron: {
            "default": "{color.interactive.icon.default.accent}",
            hover: "{color.interactive.icon.hover.accent}",
            active: "{color.interactive.icon.active.accent}"
          },
          border: "{color.border.primary}"
        }
      }
    }
  },
  carousel: {
    controls: {
      color: {
        fill: {
          "default": "{color.fill.primary}"
        },
        icon: {
          "default": "{color.interactive.icon.default.accent}",
          hover: "{color.interactive.icon.hover.accent}"
        }
      }
    }
  },
  checkbox: {
    color: {
      selected: {
        fill: {
          "default": "{color.selected.fill.default}",
          hover: "{color.selected.fill.hover}",
          active: "{color.selected.fill.active}",
          disabled: "{color.fill.disabled}"
        }
      },
      indeterminate: {
        fill: {
          "default": "{color.selected.fill.default}",
          hover: "{color.selected.fill.hover}",
          active: "{color.selected.fill.active}",
          disabled: "{color.fill.disabled}"
        }
      },
      unselected: {
        fill: {
          "default": "{color.fill.primary}",
          disabled: "{color.gray.40}"
        },
        border: {
          "default": "{color.interactive.border.default}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.gray.100}"
        }
      },
      icon: {
        "default": "{color.icon.inverse}",
        disabled: "{color.icon.disabled}"
      },
      text: {
        "default": "{color.text.primary}",
        error: "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    highlight: {
      color: {
        fill: {
          "default": "{color.blue.200}",
          disabled: "{color.gray.40}",
          error: "{color.red.300}"
        },
        selected: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}",
            disabled: "{color.fill.disabled}"
          }
        },
        indeterminate: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}",
            disabled: "{color.fill.disabled}"
          }
        },
        unselected: {
          fill: {
            "default": "{color.fill.primary}",
            disabled: "{color.gray.40}"
          },
          border: {
            "default": "{color.blue.500}",
            hover: "{color.blue.600}",
            active: "{color.blue.700}",
            error: "{color.feedback.border.negative.loud.base}",
            disabled: "{color.gray.100}"
          }
        },
        icon: {
          "default": "{color.icon.inverse}",
          disabled: "{color.icon.disabled}"
        },
        text: {
          "default": "{color.text.primary}",
          error: "{color.text.primary}",
          disabled: "{color.text.disabled}"
        }
      }
    }
  },
  datePicker: {
    day: {
      color: {
        selected: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          },
          text: {
            "default": "{color.text.inverse}",
            hover: "{color.text.inverse}",
            active: "{color.text.inverse}"
          }
        },
        unselected: {
          fill: {
            "default": "{color.interactive.fill.mute.default}",
            hover: "{color.interactive.fill.mute.hover}",
            active: "{color.interactive.fill.mute.active}",
            disabled: "{color.transparent}"
          },
          text: {
            "default": "{color.text.primary}",
            hover: "{color.text.primary}",
            active: "{color.text.primary}",
            disabled: "{color.text.disabled}"
          }
        }
      },
      event: {
        color: {
          selected: {
            fill: {
              "default": "{color.selected.fill.default}",
              hover: "{color.selected.fill.hover}",
              active: "{color.selected.fill.active}"
            },
            text: {
              "default": "{color.text.inverse}",
              hover: "{color.text.inverse}",
              active: "{color.text.inverse}"
            }
          },
          unselected: {
            fill: {
              "default": "{color.interactive.fill.mute.default}",
              hover: "{color.interactive.fill.mute.hover}",
              active: "{color.interactive.fill.mute.active}",
              disabled: "{color.transparent}"
            },
            text: {
              "default": "{color.text.primary}",
              hover: "{color.text.primary}",
              active: "{color.text.primary}",
              disabled: "{color.text.disabled}"
            }
          }
        },
        dot: {
          color: {
            selected: {
              fill: {
                "default": "{color.fill.primary}",
                hover: "{color.fill.primary}",
                active: "{color.fill.primary}"
              }
            },
            unselected: {
              fill: {
                "default": "{color.selected.fill.default}",
                hover: "{color.selected.fill.hover}",
                active: "{color.selected.fill.active}",
                disabled: "{color.gray.250}"
              }
            }
          }
        }
      }
    },
    rangeDay: {
      color: {
        fill: {
          "default": "{color.blue.150}"
        }
      },
      from: {
        color: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          },
          text: {
            "default": "{color.text.inverse}",
            hover: "{color.text.inverse}",
            active: "{color.text.inverse}"
          }
        }
      },
      middle: {
        color: {
          fill: {
            "default": "{color.interactive.fill.mute.default}",
            hover: "{color.interactive.fill.mute.hover}",
            active: "{color.interactive.fill.mute.active}"
          },
          text: {
            "default": "{color.text.primary}",
            hover: "{color.text.primary}",
            active: "{color.text.primary}"
          }
        }
      },
      to: {
        color: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          },
          text: {
            "default": "{color.text.inverse}",
            hover: "{color.text.inverse}",
            active: "{color.text.inverse}"
          }
        }
      }
    },
    month: {
      color: {
        selected: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          },
          text: {
            "default": "{color.text.inverse}",
            hover: "{color.text.inverse}",
            active: "{color.text.inverse}"
          }
        },
        unselected: {
          fill: {
            "default": "{color.interactive.fill.mute.default}",
            hover: "{color.interactive.fill.mute.hover}",
            active: "{color.interactive.fill.mute.active}",
            disabled: "{color.transparent}"
          },
          text: {
            "default": "{color.text.primary}",
            hover: "{color.text.primary}",
            active: "{color.text.primary}",
            disabled: "{color.text.disabled}"
          }
        }
      }
    },
    standalone: {
      color: {
        fill: "{color.transparent}",
        border: "{color.border.primary}"
      },
      dateSelector: {
        color: {
          text: {
            "default": "{color.text.primary}",
            hover: "{color.text.primary}",
            active: "{color.text.primary}",
            disabled: "{color.text.primary}"
          },
          chevron: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}",
            disabled: "{color.icon.disabled}"
          }
        }
      },
      weekdays: {
        color: {
          text: {
            "default": "{color.text.secondary}"
          }
        }
      },
      controls: {
        color: {
          chevron: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}",
            disabled: "{color.icon.disabled}"
          }
        }
      }
    },
    field: {
      color: {
        fill: {
          "default": "{color.fill.primary}",
          disabled: "{color.gray.40}"
        },
        label: {
          "default": "{color.text.primary}",
          error: "{color.feedback.text.negative.loud}",
          disabled: "{color.text.disabled}",
          readOnly: "{color.text.secondary}"
        },
        placeholder: {
          "default": "{color.text.secondary}"
        },
        calendar: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}",
          disabled: "{color.icon.disabled}"
        },
        border: {
          "default": "{color.interactive.border.default}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.border.disabled}"
        },
        value: {
          "default": "{color.text.primary}",
          disabled: "{color.text.disabled}"
        }
      }
    },
    color: {
      fill: "{color.fill.primary}"
    },
    extended: {
      color: {
        fill: "{color.fill.primary}",
        border: "{color.border.primary}"
      }
    }
  },
  dropdown: {
    color: {
      fill: {
        "default": "{color.fill.primary}",
        disabled: "{color.gray.40}"
      },
      label: {
        "default": "{color.text.primary}",
        error: "{color.feedback.text.negative.loud}",
        disabled: "{color.text.disabled}",
        readOnly: "{color.text.secondary}"
      },
      placeholder: {
        "default": "{color.text.secondary}"
      },
      chevron: {
        "default": "{color.interactive.icon.default.accent}",
        hover: "{color.interactive.icon.hover.accent}",
        active: "{color.interactive.icon.active.accent}",
        disabled: "{color.icon.disabled}"
      },
      icon: {
        "default": "{color.icon.primary}",
        disabled: "{color.icon.disabled}"
      },
      border: {
        "default": "{color.interactive.border.default}",
        hover: "{color.interactive.border.hover}",
        active: "{color.interactive.border.active}",
        error: "{color.feedback.border.negative.loud.base}",
        disabled: "{color.border.disabled}"
      },
      value: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    standalone: {
      color: {
        chevron: {
          "default": "{color.interactive.icon.default.base}",
          active: "{color.interactive.icon.active.base}",
          disabled: "{color.icon.disabled}"
        },
        icon: {
          "default": "{color.icon.primary}",
          disabled: "{color.icon.disabled}"
        },
        value: {
          "default": "{color.text.primary}",
          disabled: "{color.text.disabled}"
        }
      }
    }
  },
  empty: {
    states: {
      color: {
        title: "{color.text.primary}",
        description: "{color.text.primary}"
      }
    }
  },
  expandableSection: {
    color: {
      text: {
        "default": {
          base: "{color.text.primary}",
          accent: "{color.text.accent}"
        },
        disabled: "{color.text.disabled}"
      },
      chevron: {
        "default": {
          base: "{color.interactive.icon.default.base}",
          accent: "{color.interactive.icon.default.accent}"
        },
        hover: {
          base: "{color.interactive.icon.hover.base}",
          accent: "{color.interactive.icon.hover.accent}"
        },
        active: {
          base: "{color.interactive.icon.active.base}",
          accent: "{color.interactive.icon.active.accent}"
        },
        disabled: "{color.icon.disabled}"
      }
    },
    highlighted: {
      controls: {
        color: {
          fill: {
            "default": "{color.fill.primary}"
          },
          icon: {
            "default": "{color.interactive.icon.default.accent}",
            hover: "{color.interactive.icon.hover.accent}"
          }
        }
      }
    }
  },
  feedbackScreens: {
    header: {
      color: {
        title: "{color.text.primary}",
        description: "{color.text.primary}",
        overline: "{color.text.primary}",
        highlighted: {
          positive: "{color.feedback.text.positive.loud}",
          caution: "{color.feedback.text.caution.loud}",
          negative: "{color.feedback.text.negative.loud}"
        }
      },
      asset: {
        color: {
          fill: {
            positive: "{color.fill.primary}",
            caution: "{color.fill.primary}",
            negative: "{color.fill.primary}"
          },
          icon: {
            positive: "{color.feedback.icon.positive.loud}",
            caution: "{color.feedback.icon.caution.loud}",
            negative: "{color.feedback.icon.negative.loud}"
          },
          border: {
            positive: "{color.feedback.border.positive.loud.base}",
            caution: "{color.feedback.border.caution.loud.base}",
            negative: "{color.feedback.border.negative.loud.base}"
          }
        }
      }
    },
    statusBar: {
      color: {
        fill: "{color.brand.fill}",
        icon: "{color.icon.primary}"
      }
    },
    color: {
      background: {
        base: "{color.background.tertiary}",
        secondary: "{color.background.secondary}"
      },
      close: {
        "default": "{color.interactive.icon.default.base}",
        active: "{color.interactive.icon.active.base}"
      }
    },
    extended: {
      color: {
        background: "{color.background.secondary}"
      }
    },
    congrats: {
      header: {
        color: {
          title: "{color.text.primary}",
          description: "{color.text.primary}",
          overline: "{color.text.primary}",
          highlighted: "{color.feedback.text.positive.loud}"
        },
        asset: {
          color: {
            fill: "{color.fill.primary}",
            icon: "{color.feedback.icon.positive.loud}",
            border: "{color.feedback.border.positive.loud.base}"
          }
        }
      },
      statusBar: {
        color: {
          fill: "{color.feedback.fill.positive.loud.base}",
          icon: "{color.icon.inverse}"
        }
      },
      color: {
        background: "{color.background.secondary}",
        close: {
          "default": "{color.icon.inverse}"
        }
      }
    }
  },
  fab: {
    container: {
      loud: {
        color: {
          fill: "{color.fill.accent}"
        }
      },
      quiet: {
        color: {
          fill: "{color.fill.primary}"
        }
      }
    },
    loud: {
      color: {
        fill: {
          "default": "{color.interactive.fill.loud.default}",
          hover: "{color.interactive.fill.loud.hover}",
          active: "{color.interactive.fill.loud.active}"
        },
        text: {
          "default": "{color.text.inverse}"
        },
        icon: {
          "default": "{color.icon.inverse}"
        }
      }
    },
    quiet: {
      color: {
        fill: {
          "default": "{color.fill.primary}",
          hover: "{color.interactive.fill.quiet.hover}",
          active: "{color.interactive.fill.quiet.active}"
        },
        text: {
          "default": "{color.text.accent}"
        },
        icon: {
          "default": "{color.icon.accent}"
        }
      }
    }
  },
  fileUploader: {
    dropzone: {
      color: {
        fill: {
          "default": "{color.transparent}",
          hover: "{color.interactive.fill.quiet.hover}"
        },
        action: {
          "default": "{color.text.accent}",
          disabled: "{color.text.disabled}"
        },
        dropInstruction: {
          "default": "{color.text.primary}",
          disabled: "{color.text.disabled}"
        },
        description: {
          "default": "{color.text.secondary}",
          disabled: "{color.text.disabled}"
        },
        icon: {
          "default": "{color.icon.accent}",
          disabled: "{color.icon.disabled}"
        },
        border: {
          "default": "{color.feedback.border.informative.loud.base}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.interactive.border.default}"
        }
      }
    },
    thumbnail: {
      close: {
        color: {
          fill: {
            "default": "{color.fill.primary}"
          },
          icon: {
            "default": "{color.icon.primary}"
          },
          border: "{color.gray.solid.70}"
        }
      }
    },
    color: {
      label: {
        "default": "{color.text.primary}",
        error: "{color.feedback.text.negative.loud}",
        disabled: "{color.text.disabled}"
      }
    },
    tag: {
      color: {
        fileSize: "{color.text.secondary}"
      }
    }
  },
  floatingMenu: {
    color: {
      fill: "{color.fill.primary}"
    },
    textResult: {
      color: {
        text: "{color.text.secondary}"
      }
    }
  },
  helper: {
    color: {
      text: {
        "default": "{color.text.secondary}",
        caution: "{color.text.secondary}",
        error: "{color.feedback.text.negative.loud}"
      }
    }
  },
  inputStepper: {
    color: {
      fill: {
        "default": "{color.fill.primary}",
        disabled: "{color.gray.40}",
        readOnly: "{color.transparent}"
      },
      complementary: {
        "default": "{color.text.secondary}",
        readOnly: "{color.text.primary}",
        disabled: "{color.text.disabled}"
      },
      label: {
        "default": "{color.text.primary}",
        error: "{color.feedback.text.negative.loud}",
        disabled: "{color.text.disabled}",
        readOnly: "{color.text.secondary}"
      },
      helper: {
        "default": "{color.text.secondary}",
        error: "{color.feedback.text.negative.loud}"
      },
      placeholder: "{color.text.secondary}",
      spinner: "{color.feedback.fill.informative.loud.base}",
      border: {
        "default": "{color.interactive.border.default}",
        hover: "{color.interactive.border.hover}",
        disabled: "{color.interactive.border.default}",
        active: "{color.feedback.border.informative.loud.base}",
        error: "{color.feedback.border.negative.loud.base}"
      },
      value: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    controls: {
      color: {
        fill: {
          "default": "{color.interactive.fill.mute.default}",
          hover: "{color.interactive.fill.mute.hover}",
          active: "{color.interactive.fill.mute.active}"
        },
        icon: {
          "default": "{color.icon.accent}",
          disabled: "{color.icon.disabled}"
        }
      }
    }
  },
  list: {
    color: {
      surface: {
        "default": "{color.transparent}",
        hover: "{color.surface.hover}",
        active: "{color.surface.active}",
        disabled: "{color.transparent}"
      },
      title: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      },
      description: {
        "default": "{color.text.secondary}",
        disabled: "{color.text.disabled}"
      },
      icon: {
        right: "{color.icon.primary}",
        left: "{color.icon.primary}",
        disabled: "{color.icon.disabled}"
      },
      border: "{color.border.primary}"
    },
    chevron: {
      color: {
        icon: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}",
          disabled: "{color.icon.disabled}"
        }
      }
    },
    pickList: {
      color: {
        selected: {
          title: {
            "default": "{color.selected.text.default}",
            hover: "{color.selected.text.hover}",
            active: "{color.selected.text.active}"
          },
          icon: {
            "default": "{color.selected.icon.default}",
            hover: "{color.selected.icon.hover}",
            active: "{color.selected.icon.active}"
          },
          pipe: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          }
        }
      }
    },
    group: {
      header: {
        color: {
          fill: "{color.transparent}",
          text: "{color.text.secondary}",
          border: "{color.border.primary}"
        }
      }
    }
  },
  message: {
    loud: {
      color: {
        fill: {
          informative: "{color.feedback.fill.informative.loud.base}",
          positive: "{color.feedback.fill.positive.loud.base}",
          caution: "{color.feedback.fill.caution.loud.base}",
          negative: "{color.feedback.fill.negative.loud.base}"
        },
        title: {
          informative: "{color.text.inverse}",
          positive: "{color.text.inverse}",
          caution: "{color.text.inverse}",
          negative: "{color.text.inverse}"
        },
        description: {
          informative: "{color.text.inverse}",
          positive: "{color.text.inverse}",
          caution: "{color.text.inverse}",
          negative: "{color.text.inverse}"
        },
        close: {
          informative: {
            "default": "{color.interactive.icon.default.inverse}"
          },
          positive: {
            "default": "{color.interactive.icon.default.inverse}"
          },
          caution: {
            "default": "{color.interactive.icon.default.inverse}"
          },
          negative: {
            "default": "{color.interactive.icon.default.inverse}"
          }
        }
      }
    },
    quiet: {
      color: {
        fill: {
          informative: "{color.fill.secondary}",
          positive: "{color.fill.secondary}",
          caution: "{color.fill.secondary}",
          negative: "{color.fill.secondary}"
        },
        title: {
          informative: "{color.text.primary}",
          positive: "{color.text.primary}",
          caution: "{color.text.primary}",
          negative: "{color.text.primary}"
        },
        description: {
          informative: "{color.text.primary}",
          positive: "{color.text.primary}",
          caution: "{color.text.primary}",
          negative: "{color.text.primary}"
        },
        close: {
          informative: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}"
          },
          positive: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}"
          },
          caution: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}"
          },
          negative: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}"
          }
        }
      },
      pipe: {
        color: {
          fill: {
            informative: "{color.feedback.fill.informative.loud.base}",
            positive: "{color.feedback.fill.positive.loud.base}",
            caution: "{color.feedback.fill.caution.loud.base}",
            negative: "{color.feedback.fill.negative.loud.base}"
          }
        }
      }
    }
  },
  moneyAmount: {
    color: {
      amount: "{color.text.primary}",
      currency: "{color.text.primary}",
      iso: "{color.text.primary}",
      suffix: "{color.text.secondary}",
      previous: "{color.text.secondary}",
      discount: "{color.feedback.text.positive.loud}"
    }
  },
  pagination: {
    color: {
      selected: {
        fill: {
          "default": "{color.fill.primary}",
          hover: "{color.blue.150}",
          active: "{color.blue.200}"
        },
        border: {
          "default": "{color.selected.border.default}",
          hover: "{color.selected.border.hover}",
          active: "{color.selected.border.active}"
        },
        text: {
          "default": "{color.text.primary}"
        }
      },
      unselected: {
        fill: {
          "default": "{color.transparent}",
          hover: "{color.blue.150}",
          active: "{color.blue.200}"
        },
        text: {
          "default": "{color.text.secondary}"
        }
      }
    },
    controls: {
      color: {
        fill: {
          "default": "{color.transparent}",
          hover: "{color.blue.150}",
          active: "{color.blue.200}",
          disabled: "{color.transparent}"
        },
        text: {
          "default": "{color.text.secondary}",
          disabled: "{color.text.disabled}"
        },
        chevron: {
          "default": "{color.icon.secondary}",
          disabled: "{color.icon.disabled}"
        }
      }
    }
  },
  pageViewer: {
    dot: {
      color: {
        selected: {
          fill: {
            base: "{color.selected.fill.default}",
            inverse: "{color.fill.primary}"
          },
          border: {
            inverse: "{color.gray.250}"
          }
        },
        unselected: {
          fill: {
            base: "{color.fill.tertiary}",
            inverse: "{color.gray.250}"
          },
          border: {
            inverse: "{color.border.inverse}"
          }
        }
      }
    }
  },
  progressIndicator: {
    circular: {
      color: {
        fill: {
          base: "{color.fill.tertiary}",
          inverse: "{color.white.300}"
        },
        label: {
          base: "{color.text.primary}",
          inverse: "{color.text.inverse}"
        },
        counter: {
          base: "{color.text.primary}",
          inverse: "{color.text.inverse}"
        },
        icon: {
          negative: "{color.feedback.icon.negative.loud}",
          positive: "{color.feedback.icon.positive.loud}"
        }
      },
      progress: {
        color: {
          fill: {
            "default": "{color.feedback.fill.informative.loud.base}",
            inverse: "{color.fill.primary}",
            negative: "{color.feedback.fill.negative.loud.base}",
            positive: "{color.feedback.fill.positive.loud.base}"
          }
        }
      }
    },
    linear: {
      progress: {
        color: {
          fill: {
            "default": "{color.feedback.fill.informative.loud.base}",
            inverse: "{color.fill.primary}",
            negative: "{color.feedback.fill.negative.loud.base}",
            positive: "{color.feedback.fill.positive.loud.base}"
          }
        }
      },
      color: {
        fill: {
          base: "{color.fill.tertiary}",
          inverse: "{color.white.300}"
        },
        label: {
          base: "{color.text.primary}",
          inverse: "{color.text.inverse}"
        },
        counter: {
          large: {
            base: "{color.text.primary}",
            inverse: "{color.text.inverse}"
          },
          small: {
            base: "{color.text.secondary}",
            inverse: "{color.text.inverse}"
          }
        }
      },
      highlight: {
        color: {
          fill: "{color.feedback.fill.informative.loud.base}"
        }
      },
      comparison: {
        color: {
          fill: {
            "default": "{color.gray.550}"
          },
          label: "{color.text.secondary}"
        }
      }
    }
  },
  radioButton: {
    color: {
      selected: {
        fill: {
          "default": "{color.fill.primary}"
        },
        border: {
          "default": "{color.selected.border.default}",
          hover: "{color.selected.border.hover}",
          active: "{color.selected.border.active}",
          disabled: "{color.gray.100}"
        }
      },
      unselected: {
        fill: {
          "default": "{color.fill.primary}",
          disabled: "{color.gray.40}"
        },
        border: {
          "default": "{color.interactive.border.default}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.gray.100}"
        }
      },
      text: {
        "default": "{color.text.primary}",
        error: "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    tick: {
      color: {
        selected: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}",
            disabled: "{color.fill.disabled}"
          }
        }
      }
    }
  },
  searchbox: {
    color: {
      fill: {
        "default": "{color.fill.primary}"
      },
      inputText: {
        "default": "{color.text.primary}",
        loading: "{color.text.secondary}"
      },
      placeholder: {
        "default": "{color.text.secondary}"
      },
      search: {
        "default": "{color.interactive.icon.default.base}",
        hover: "{color.interactive.icon.hover.base}",
        active: "{color.interactive.icon.active.base}"
      },
      clear: {
        "default": "{color.interactive.icon.default.base}",
        hover: "{color.interactive.icon.hover.base}",
        active: "{color.interactive.icon.active.base}"
      },
      border: {
        "default": "{color.interactive.border.default}",
        hover: "{color.interactive.border.hover}",
        active: "{color.interactive.border.active}"
      }
    },
    split: {
      color: {
        divider: {
          "default": "{color.border.primary}"
        }
      }
    }
  },
  segmentedControl: {
    color: {
      fill: {
        base: "{color.fill.secondary}",
        onComplexBg: "{color.gray.550}"
      },
      border: {
        base: "{color.gray.40}",
        onComplexBg: "{color.gray.250}"
      }
    },
    segment: {
      color: {
        selected: {
          fill: {
            "default": {
              base: "{color.fill.primary}",
              onComplexBg: "{color.fill.primary}"
            },
            disabled: {
              base: "{color.gray.40}",
              onComplexBg: "{color.white.300}"
            }
          },
          text: {
            "default": {
              base: "{color.text.primary}",
              onComplexBg: "{color.text.primary}"
            },
            disabled: {
              base: "{color.text.disabled}",
              onComplexBg: "{color.gray.450}"
            }
          },
          icon: {
            "default": {
              base: "{color.icon.primary}",
              onComplexBg: "{color.icon.primary}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.gray.450}"
            }
          }
        },
        unselected: {
          fill: {
            "default": {
              base: "{color.transparent}",
              onComplexBg: "{color.transparent}"
            },
            hover: {
              base: "{color.surface.hover}",
              onComplexBg: "{color.gray.100}"
            },
            active: {
              base: "{color.surface.active}",
              onComplexBg: "{color.gray.250}"
            },
            disabled: {
              base: "{color.transparent}",
              onComplexBg: "{color.transparent}"
            }
          },
          text: {
            "default": {
              base: "{color.text.secondary}",
              onComplexBg: "{color.text.inverse}"
            },
            disabled: {
              base: "{color.text.disabled}",
              onComplexBg: "{color.white.300}"
            },
            hover: {
              onComplexBg: "{color.text.inverse}"
            },
            active: {
              onComplexBg: "{color.text.inverse}"
            }
          },
          icon: {
            "default": {
              base: "{color.icon.secondary}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.white.300}"
            },
            hover: {
              onComplexBg: "{color.icon.inverse}"
            },
            active: {
              onComplexBg: "{color.icon.inverse}"
            }
          }
        }
      }
    }
  },
  slider: {
    progress: {
      color: {
        fill: {
          "default": "{color.feedback.fill.informative.loud.base}",
          disabled: "{color.gray.solid.250}"
        }
      }
    },
    handle: {
      front: {
        color: {
          fill: {
            "default": "{color.fill.primary}",
            disabled: "{color.fill.primary}"
          }
        }
      },
      back: {
        color: {
          fill: {
            "default": "{color.feedback.fill.informative.loud.base}",
            active: "{color.blue.300}",
            disabled: "{color.gray.solid.250}"
          }
        }
      }
    },
    color: {
      fill: {
        "default": "{color.fill.tertiary}",
        disabled: "{color.fill.disabled}"
      },
      label: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      },
      icon: {
        "default": "{color.icon.primary}",
        disabled: "{color.icon.disabled}"
      },
      text: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    step: {
      color: {
        fill: {
          "default": "{color.gray.250}"
        }
      }
    },
    reference: {
      color: {
        fill: {
          active: "{color.fill.inverse}"
        },
        text: {
          "default": "{color.text.primary}",
          active: "{color.text.inverse}"
        }
      }
    }
  },
  snackbar: {
    color: {
      fill: {
        base: "{color.fill.inverse}",
        positive: "{color.feedback.fill.positive.loud.base}",
        negative: "{color.feedback.fill.negative.loud.base}"
      },
      text: {
        base: "{color.text.inverse}",
        positive: "{color.text.inverse}",
        negative: "{color.text.inverse}"
      }
    },
    action: {
      color: {
        text: {
          "default": "{color.text.link.default.inverse}",
          positive: {
            "default": "{color.text.link.default.inverse}"
          },
          negative: {
            "default": "{color.text.link.default.inverse}"
          }
        }
      }
    }
  },
  "switch": {
    color: {
      selected: {
        fill: {
          "default": "{color.selected.fill.default}",
          disabled: "{color.blue.300}"
        }
      },
      unselected: {
        fill: {
          "default": "{color.fill.tertiary}",
          disabled: "{color.fill.disabled}"
        }
      },
      text: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    handle: {
      color: {
        fill: {
          "default": "{color.fill.primary}"
        }
      }
    }
  },
  tabs: {
    color: {
      fill: "{color.transparent}"
    },
    tab: {
      color: {
        selected: {
          fill: {
            "default": "{color.interactive.fill.mute.default}",
            hover: "{color.interactive.fill.mute.hover}",
            active: "{color.interactive.fill.mute.active}"
          },
          pipe: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          },
          text: {
            "default": "{color.selected.text.default}",
            hover: "{color.selected.text.hover}",
            active: "{color.selected.text.active}"
          }
        },
        unselected: {
          fill: {
            "default": "{color.interactive.fill.mute.default}",
            hover: "{color.interactive.fill.mute.hover}",
            active: "{color.interactive.fill.mute.active}"
          },
          text: {
            "default": "{color.text.primary}",
            hover: "{color.selected.text.hover}",
            active: "{color.selected.text.active}"
          }
        },
        border: "{color.border.primary}"
      }
    },
    controls: {
      color: {
        chevron: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}",
          disabled: "{color.icon.disabled}"
        }
      }
    }
  },
  tag: {
    color: {
      fill: {
        "default": "{color.transparent}",
        hover: "{color.fill.secondary}",
        active: "{color.fill.tertiary}",
        disabled: "{color.transparent}",
        positive: {
          "default": "{color.feedback.fill.positive.quiet}",
          hover: "{color.green.150}",
          active: "{color.green.200}"
        },
        caution: {
          "default": "{color.feedback.fill.caution.quiet}",
          hover: "{color.orange.150}",
          active: "{color.orange.200}"
        },
        negative: {
          "default": "{color.feedback.fill.negative.quiet}",
          hover: "{color.red.150}",
          active: "{color.red.200}"
        }
      },
      text: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}",
        positive: {
          "default": "{color.feedback.text.positive.loud}",
          hover: "{color.green.600}",
          active: "{color.green.700}"
        },
        caution: {
          "default": "{color.feedback.text.caution.loud}",
          hover: "{color.orange.600}",
          active: "{color.orange.700}"
        },
        negative: {
          "default": "{color.feedback.text.negative.loud}",
          hover: "{color.red.600}",
          active: "{color.red.700}"
        }
      },
      icon: {
        "default": "{color.icon.primary}",
        disabled: "{color.icon.disabled}",
        positive: {
          "default": "{color.feedback.icon.positive.loud}",
          hover: "{color.green.600}",
          active: "{color.green.700}"
        },
        caution: {
          "default": "{color.feedback.icon.caution.loud}",
          hover: "{color.orange.600}",
          active: "{color.orange.700}"
        },
        negative: {
          "default": "{color.feedback.icon.negative.loud}",
          hover: "{color.red.600}",
          active: "{color.red.700}"
        }
      },
      close: {
        "default": "{color.interactive.icon.default.base}",
        hover: "{color.interactive.icon.hover.base}",
        active: "{color.interactive.icon.active.base}",
        disabled: "{color.icon.disabled}",
        positive: {
          "default": "{color.feedback.icon.positive.loud}",
          hover: "{color.green.600}",
          active: "{color.green.700}"
        },
        caution: {
          "default": "{color.feedback.icon.caution.loud}",
          hover: "{color.orange.600}",
          active: "{color.orange.700}"
        },
        negative: {
          "default": "{color.feedback.icon.negative.loud}",
          hover: "{color.red.600}",
          active: "{color.red.700}"
        }
      },
      border: {
        "default": "{color.gray.250}",
        disabled: "{color.gray.100}",
        positive: {
          "default": "{color.feedback.border.positive.loud.base}",
          hover: "{color.green.600}",
          active: "{color.green.700}"
        },
        caution: {
          "default": "{color.feedback.border.caution.loud.base}",
          hover: "{color.orange.600}",
          active: "{color.orange.700}"
        },
        negative: {
          "default": "{color.feedback.border.negative.loud.base}",
          hover: "{color.red.600}",
          active: "{color.red.700}"
        }
      }
    },
    choice: {
      color: {
        selected: {
          fill: {
            "default": "{color.blue.100}",
            disabled: "{color.transparent}"
          },
          text: {
            "default": "{color.selected.text.default}",
            hover: "{color.selected.text.hover}",
            active: "{color.selected.text.active}",
            disabled: "{color.text.disabled}"
          },
          check: {
            "default": "{color.selected.icon.default}",
            hover: "{color.selected.icon.hover}",
            active: "{color.selected.icon.active}"
          },
          icon: {
            "default": "{color.selected.icon.default}",
            hover: "{color.selected.icon.hover}",
            active: "{color.selected.icon.active}",
            disabled: "{color.icon.disabled}"
          },
          border: {
            "default": "{color.selected.border.default}",
            hover: "{color.selected.border.hover}",
            active: "{color.selected.border.active}",
            disabled: "{color.gray.100}"
          }
        },
        unselected: {
          fill: {
            "default": "{color.transparent}"
          },
          text: {
            "default": "{color.text.primary}",
            disabled: "{color.text.disabled}"
          },
          icon: {
            "default": "{color.icon.primary}",
            disabled: "{color.icon.disabled}"
          },
          border: {
            "default": "{color.gray.250}",
            hover: "{color.gray.550}",
            active: "{color.blue.500}",
            disabled: "{color.gray.100}"
          }
        }
      }
    }
  },
  technicalError: {
    section: {
      color: {
        title: "{color.text.primary}",
        description: "{color.text.primary}",
        instruction: "{color.text.secondary}"
      },
      asset: {
        color: {
          fill: "{color.fill.primary}",
          icon: "{color.icon.secondary}",
          border: "{color.gray.550}"
        }
      },
      code: {
        color: {
          fill: {
            "default": "{color.transparent}",
            hover: "{color.fill.secondary}",
            active: "{color.fill.tertiary}"
          },
          border: {
            "default": "{color.gray.250}"
          },
          text: {
            "default": "{color.text.secondary}"
          },
          icon: {
            "default": "{color.interactive.icon.default.accent}",
            hover: "{color.interactive.icon.hover.accent}",
            active: "{color.interactive.icon.active.accent}"
          }
        }
      }
    },
    screen: {
      statusBar: {
        color: {
          fill: "{color.fill.brand}",
          icon: "{color.icon.primary}"
        }
      },
      color: {
        background: "{color.background.secondary}"
      }
    },
    snackbar: {
      color: {
        instruction: "{color.text.inverse}"
      },
      code: {
        color: {
          fill: {
            "default": "{color.transparent}",
            active: "{color.fill.primary}"
          },
          border: {
            "default": "{color.border.inverse}",
            active: "{color.border.inverse}"
          },
          text: {
            "default": "{color.text.link.default.inverse}",
            active: "{color.text.primary}"
          },
          icon: {
            "default": "{color.interactive.icon.default.inverse}",
            active: "{color.feedback.icon.positive.loud}"
          }
        }
      }
    }
  },
  textfield: {
    color: {
      fill: {
        "default": "{color.fill.primary}",
        disabled: "{color.fill.secondary}"
      },
      label: {
        "default": "{color.text.primary}",
        error: "{color.feedback.text.negative.loud}",
        disabled: "{color.text.disabled}",
        readOnly: "{color.text.secondary}"
      },
      placeholder: {
        "default": "{color.text.secondary}"
      },
      inputText: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      },
      prefix: {
        "default": "{color.text.secondary}",
        disabled: "{color.text.disabled}"
      },
      suffix: {
        "default": "{color.text.secondary}",
        disabled: "{color.text.disabled}"
      },
      text: {
        checkbox: {
          "default": "{color.text.primary}",
          disabled: "{color.text.disabled}"
        }
      },
      icon: {
        "default": "{color.icon.primary}",
        disabled: "{color.icon.disabled}"
      },
      tooltip: {
        "default": "{color.interactive.icon.default.accent}",
        hover: "{color.interactive.icon.hover.accent}",
        active: "{color.interactive.icon.active.accent}"
      },
      validated: "{color.feedback.icon.positive.loud}",
      clear: {
        "default": "{color.interactive.icon.default.base}",
        hover: "{color.interactive.icon.hover.base}",
        active: "{color.interactive.icon.active.base}"
      },
      border: {
        "default": "{color.interactive.border.default}",
        hover: "{color.interactive.border.hover}",
        active: "{color.interactive.border.active}",
        error: "{color.feedback.border.negative.loud.base}",
        disabled: "{color.border.disabled}"
      }
    },
    split: {
      color: {
        divider: {
          "default": "{color.border.primary}",
          disabled: "{color.border.disabled}"
        }
      }
    },
    inline: {
      color: {
        fill: {
          "default": "{color.transparent}"
        },
        inputText: {
          "default": "{color.text.primary}",
          caution: "{color.feedback.text.caution.loud}",
          error: "{color.feedback.text.negative.loud}"
        },
        edit: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}"
        },
        border: {
          "default": "{color.transparent}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          caution: "{color.feedback.border.caution.loud.base}",
          error: "{color.feedback.border.negative.loud.base}"
        }
      }
    },
    password: {
      color: {
        icon: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}"
        }
      }
    }
  },
  timePicker: {
    intervals: {
      color: {
        fill: {
          disabled: "{color.gray.40}"
        },
        label: {
          "default": "{color.text.primary}",
          error: "{color.feedback.text.negative.loud}",
          disabled: "{color.text.disabled}",
          readOnly: "{color.text.secondary}"
        },
        placeholder: "{color.text.secondary}",
        border: {
          "default": "{color.interactive.border.default}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.border.disabled}",
          readOnly: "{color.transparent}"
        }
      },
      selector: {
        color: {
          fill: {
            disabled: "{color.gray.40}"
          },
          chevron: {
            "default": "{color.interactive.icon.default.accent}",
            hover: "{color.interactive.icon.hover.accent}",
            active: "{color.interactive.icon.active.accent}",
            disabled: "{color.icon.disabled}"
          },
          border: {
            "default": "{color.interactive.border.default}",
            hover: "{color.interactive.border.hover}",
            active: "{color.interactive.border.active}",
            error: "{color.feedback.border.negative.loud.base}",
            disabled: "{color.border.disabled}",
            readOnly: "{color.transparent}"
          },
          value: {
            "default": "{color.text.primary}",
            disabled: "{color.text.disabled}"
          }
        }
      }
    }
  },
  thumbnail: {
    image: {
      color: {
        fill: {
          "default": "{color.fill.primary}"
        }
      }
    },
    loud: {
      color: {
        fill: {
          "default": "{color.blue.500}",
          disabled: "{color.fill.disabled}"
        },
        text: {
          "default": "{color.text.inverse}",
          disabled: "{color.text.disabled}"
        },
        icon: {
          "default": "{color.icon.inverse}",
          disabled: "{color.icon.disabled}"
        }
      }
    },
    quiet: {
      color: {
        fill: {
          "default": "{color.blue.100}",
          disabled: "{color.fill.disabled}"
        },
        text: {
          "default": "{color.text.accent}",
          disabled: "{color.text.disabled}"
        },
        icon: {
          "default": "{color.icon.accent}",
          disabled: "{color.icon.disabled}"
        }
      }
    },
    mute: {
      color: {
        fill: {
          "default": "{color.fill.primary}",
          disabled: "{color.fill.primary}"
        },
        text: {
          "default": "{color.text.primary}",
          disabled: "{color.text.disabled}"
        },
        icon: {
          "default": "{color.icon.primary}",
          disabled: "{color.icon.disabled}"
        }
      }
    },
    color: {
      border: {
        "default": "{color.border.primary}"
      }
    },
    badge: {
      color: {
        text: {
          informative: "{color.feedback.text.informative.loud}",
          positive: "{color.feedback.text.positive.loud}",
          caution: "{color.feedback.text.caution.loud}",
          negative: "{color.feedback.text.negative.loud}"
        },
        icon: {
          informative: "{color.feedback.icon.informative.loud}",
          positive: "{color.feedback.icon.positive.loud}",
          caution: "{color.feedback.icon.caution.loud}",
          negative: "{color.feedback.icon.negative.loud}"
        },
        border: {
          informative: {
            base: "{color.feedback.border.informative.loud.base}",
            onComplexBg: "{color.feedback.border.informative.loud.onFill}"
          },
          positive: {
            base: "{color.feedback.border.positive.loud.base}",
            onComplexBg: "{color.feedback.border.positive.loud.onFill}"
          },
          caution: {
            base: "{color.feedback.border.caution.loud.base}",
            onComplexBg: "{color.feedback.border.caution.loud.onFill}"
          },
          negative: {
            base: "{color.feedback.border.negative.loud.base}",
            onComplexBg: "{color.feedback.border.negative.loud.onFill}"
          }
        }
      }
    },
    grid: {
      color: {
        fill: "{color.fill.secondary}"
      },
      overflow: {
        color: {
          fill: "{color.gray.solid.70}",
          text: "{color.text.secondary}",
          border: "{color.border.primary}"
        }
      }
    },
    stacked: {
      overflow: {
        color: {
          fill: "{color.gray.solid.70}",
          text: "{color.text.secondary}",
          border: "{color.border.primary}"
        }
      }
    }
  },
  tooltip: {
    color: {
      fill: {
        base: "{color.fill.primary}",
        inverse: "{color.fill.inverse}",
        accent: "{color.feedback.fill.informative.loud.base}"
      },
      title: {
        base: "{color.text.primary}",
        inverse: "{color.text.inverse}",
        accent: "{color.text.inverse}"
      },
      text: {
        base: "{color.text.primary}",
        inverse: "{color.text.inverse}",
        accent: "{color.text.inverse}"
      },
      close: {
        "default": {
          base: "{color.interactive.icon.default.base}",
          inverse: "{color.interactive.icon.default.inverse}",
          accent: "{color.interactive.icon.default.inverse}"
        },
        hover: "{color.interactive.icon.hover.base}",
        active: "{color.interactive.icon.active.base}"
      }
    }
  }
};
var mlTheme = {
  core: core$1,
  colorSchemes: colorSchemes$1,
  components: components$1
};

/** @typedef {import('../../..').ThemeHash} ThemeHash */
/** @typedef {import('../../..').AndesProviderTheme} Theme */

/**
 * @param {Record<string, any>} styleOverrides
 *
 * @returns {string}
 */
const stylesObjectToCssStyles = styleOverrides => Object.entries(styleOverrides).map(_ref => {
  let _ref2 = _slicedToArray(_ref, 2),
    key = _ref2[0],
    value = _ref2[1];
  const sanitizedKey = camelCaseToKebabCase(key);
  if (typeof value === 'object') {
    return `${sanitizedKey} {
          ${stylesObjectToCssStyles(value)}
        }`;
  }
  return toCssValue(sanitizedKey, value);
}).join(' ');

/** @param {Record<string, any>} object */
const getObjectStyles = object => {
  if (typeof object !== 'object') {
    return '';
  }
  return stylesObjectToCssStyles(object);
};

/**
 * @param {string} component
 * @param {Record<string, any>} varsObject
 * @param {(componentStyles: Record<string, any>) => string} varMapper
 */
const getComponentVariables = (component, varsObject, varMapper) => varMapper({
  [component]: varsObject
});

/**
 * @param {Theme["components"]} components
 * @param {import('../../..').GetComponentsStyles} componentClasses
 * @param {ThemeHash} themeHash
 * @param {string} prefix
 * @param {(componentStyles: Record<string, any>) => string} varMapper
 */
const buildComponentsStyles = (components, componentClasses, themeHash, prefix, varMapper) => {
  if (typeof components !== 'object') {
    return '';
  }
  return Object.entries(components).filter(_ref3 => {
    let _ref4 = _slicedToArray(_ref3, 1),
      component = _ref4[0];
    return componentClasses.has(component);
  }).map(_ref5 => {
    var _componentClasses$get, _componentClasses$get2;
    let _ref6 = _slicedToArray(_ref5, 2),
      component = _ref6[0],
      componentEntriesObj = _ref6[1];
    const _separateComponentSty = separateComponentStylesFromComponentEntries(componentEntriesObj),
      componentStyles = _separateComponentSty.componentStyles,
      componentOverrides = _separateComponentSty.componentOverrides;
    const componentVariants = ((_componentClasses$get = componentClasses.get(component)) === null || _componentClasses$get === void 0 ? void 0 : _componentClasses$get.variants) || [];
    const _diffDeep = diffDeep(componentStyles, mlTheme.components[(/** @type {keyof typeof mlTheme.components} */component)] || {}),
      whitelistedComponentStyles = _diffDeep.keysInTargetAndSource,
      customStyles = _diffDeep.keysOnlyInTarget;
    const _separateCssStylesFro = separateCssStylesFromCustomVars(customStyles),
      cssStyles = _separateCssStylesFro.cssStyles,
      otherStyles = _separateCssStylesFro.otherStyles;
    const computedStyle = `
              ${getComponentVariables(component, whitelistedComponentStyles, varMapper)}
              ${getComponentVariables(component, otherStyles, varMapper)}
              ${getObjectStyles(cssStyles)}
              ${getObjectStyles(componentOverrides === null || componentOverrides === void 0 ? void 0 : componentOverrides.styleOverrides)}
              ${componentVariants.map(variant => `&.${componentVariantPrefix}${variant.id} {${getObjectStyles(variant.styles)}}`).join(' ')}
      `.trim();
    if (computedStyle.length === 0) {
      return '';
    }
    const className = (((_componentClasses$get2 = componentClasses.get(component)) === null || _componentClasses$get2 === void 0 ? void 0 : _componentClasses$get2.className) || '').split(' ').join('.');
    const _Object$entries$ = _slicedToArray(Object.entries(themeHash)[0], 2),
      hashKey = _Object$entries$[0],
      hashValue = _Object$entries$[1];
    const hashSelector = `${hashKey}=${hashValue}`;
    const cssSelectors = `
      .${className}[class^=${prefix}],
      .${className}[class^=${prefix}][${hashSelector}]
      `;
    return `${cssSelectors} {${computedStyle}}`;
  }).join(' ');
};

/**
 * @param {Theme} theme
 * @param {string} prefix
 * @param {ThemeHash} themeHash
 */
const createCssVarsThemeStyles = (theme, prefix, themeHash) => {
  const _theme$components = theme.components,
    components = _theme$components === void 0 ? {} : _theme$components;
  const _createThemeVarsMaps = createThemeVarsMaps(theme, prefix),
    coreVars = _createThemeVarsMaps.coreVars,
    semanticLightVars = _createThemeVarsMaps.semanticLightVars,
    componentVarsMapper = _createThemeVarsMaps.componentVarsMapper;
  const themeVariables = createThemeVarsNames(theme, prefix);
  const coreAndSemanticVars = _objectSpread(_objectSpread({}, coreVars), semanticLightVars);

  // const getDarkMode = () =>
  //   (Object.keys(semanticDarkVars).length > 0
  //     ? `@media(prefers-color-scheme: dark){${toCSSVars(semanticDarkVars, coreVars)}}`
  //     : '');

  const getCoreStyles = () => `
  ${toCssVars(coreVars)}
  ${toCssVars(semanticLightVars, coreVars)}`;

  /** @param {import('../../..').GetComponentsStyles} styledComponents */
  const getComponentsStyles = styledComponents => buildComponentsStyles(components, styledComponents, themeHash, prefix, componentStyles => toCssVars(componentVarsMapper(componentStyles), coreAndSemanticVars));
  return {
    getCoreStyles,
    getComponentsStyles,
    themeVariables
  };
};

/** @typedef {import('../../..').AndesProviderTheme} Theme */

/**
 * @param {Theme["components"]} components
 * @param {string} id
 * @param {string} prefix
 *
 * @returns {import('../../..').GetComponentsStyles}
 */
const registerComponentsClasses = (components, id, prefix) => {
  if (typeof components !== 'object') {
    return new Map();
  }
  return new Map(Object.entries(components).map(_ref => {
    let _ref2 = _slicedToArray(_ref, 2),
      component = _ref2[0],
      componentStyles = _ref2[1];
    return [component, {
      className: `${prefix}-${id}-${component}`,
      variants: (componentStyles === null || componentStyles === void 0 ? void 0 : componentStyles.variants) || []
    }];
  }));
};

/**
  * @param {import('../../..').CustomVariant} variant
  * @param {object} componentProps
  */
const variantMatch = (variant, componentProps) => typeof variant.propsMatcher === 'function' ? variant.propsMatcher(componentProps) : Object.keys(variant.propsMatcher).every(key =>
// eslint-disable-next-line max-len
componentProps[(/** @type {keyof typeof componentProps} */key)] === /** @type { Record<string, unknown> } */variant.propsMatcher[key]);

/**
 * @param {Theme} theme
 * @param {string} id
 */
const createTheme = (theme, id) => {
  const _theme$components = theme.components,
    components = _theme$components === void 0 ? {} : _theme$components;
  const hashKey = 'data-t-hash';
  const themeHash = {
    [hashKey]: id
  };
  const componentsClasses = registerComponentsClasses(components, id, andesThemePrefix);
  const _createCssVarsThemeSt = createCssVarsThemeStyles(theme, andesThemePrefix, themeHash),
    getCoreStyles = _createCssVarsThemeSt.getCoreStyles,
    getComponentsStyles = _createCssVarsThemeSt.getComponentsStyles,
    themeVariables = _createCssVarsThemeSt.themeVariables;

  /**
   * @param {string} component
   * @param {object} componentProps
   */
  const getComponentVariants = (component, componentProps) => {
    let variants = /** @type import('../../..').CustomVariant[] */[];
    if (Object.keys(componentProps).length > 0) {
      var _components$component;
      const componentVariants = ((_components$component = components[component]) === null || _components$component === void 0 ? void 0 : _components$component.variants) || [];
      variants = componentVariants.filter(variant => variantMatch(variant, componentProps));
    }
    return variants;
  };

  /**
   * @param {import('../../..').GetComponentsStyles} styledComponents
   * @param {boolean} [global]
   */
  const getStyles = function () {
    let styledComponents = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : componentsClasses;
    let global = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
    const style = (getCoreStyles() + getComponentsStyles(styledComponents)).replace(/ {2,}|\n/g, '');
    return style.length > 0 ? `${global ? 'body' : `div#${id}, [data-andes-portal][${hashKey}=${id}], [data-andes-popper-outside-root][${hashKey}=${id}]`}{${style}}` : '';
  };
  return {
    getStyles,
    componentsClasses,
    themeHash,
    getComponentVariants,
    themeVariables
  };
};
var core = {
  color: {
    blue: {
      "100": "rgba(65, 137, 230, 0.1)",
      "150": "rgba(65, 137, 230, 0.15)",
      "200": "rgba(65, 137, 230, 0.2)",
      "300": "rgba(65, 137, 230, 0.3)",
      "400": "rgba(65, 137, 230, 0.4)",
      "500": "#3483fa",
      "600": "#2968c8",
      "700": "#1f4e96",
      "800": "#183c73"
    },
    cyan: {
      "100": "rgba(71, 154, 209, 0.1)",
      "150": "rgba(71, 154, 209, 0.15)",
      "200": "rgba(71, 154, 209, 0.2)",
      "300": "rgba(71, 154, 209, 0.3)",
      "400": "rgba(71, 154, 209, 0.4)",
      "500": "#009ee3",
      "600": "#007eb5",
      "700": "#005e88",
      "800": "#004766"
    },
    white: {
      "300": "rgba(255, 255, 255, 0.3)",
      base: "#ffffff"
    },
    gray: {
      "40": "rgba(0, 0, 0, 0.04)",
      "70": "rgba(0, 0, 0, 0.07)",
      "100": "rgba(0, 0, 0, 0.1)",
      "150": "rgba(0, 0, 0, 0.15)",
      "250": "rgba(0, 0, 0, 0.25)",
      "450": "rgba(0, 0, 0, 0.45)",
      "550": "rgba(0, 0, 0, 0.55)",
      "800": "rgba(0, 0, 0, 0.8)",
      "900": "rgba(0, 0, 0, 0.9)",
      solid: {
        "40": "#f5f5f5",
        "70": "#ededed",
        "100": "#e5e5e5",
        "150": "#d9d9d9",
        "250": "#bfbfbf",
        "450": "#8c8c8c",
        "550": "#737373",
        "800": "#333333",
        "900": "#1a1a1a"
      }
    },
    transparent: "rgba(255, 255, 255, 0)",
    green: {
      "100": "rgba(0, 166, 80, 0.1)",
      "150": "rgba(0, 166, 80, 0.15)",
      "200": "rgba(0, 166, 80, 0.2)",
      "300": "rgba(0, 166, 80, 0.3)",
      "400": "rgba(0, 166, 80, 0.4)",
      "500": "#00a650",
      "600": "#008744",
      "700": "#006633",
      "800": "#004d27"
    },
    orange: {
      "100": "rgba(255, 119, 51, 0.1)",
      "150": "rgba(255, 119, 51, 0.15)",
      "200": "rgba(255, 119, 51, 0.2)",
      "300": "rgba(255, 119, 51, 0.3)",
      "400": "rgba(255, 119, 51, 0.4)",
      "500": "#ff7733",
      "600": "#e6540b",
      "700": "#cc3e0a",
      "800": "#a62a08"
    },
    red: {
      "100": "rgba(242, 61, 79, 0.1)",
      "150": "rgba(242, 61, 79, 0.15)",
      "200": "rgba(242, 61, 79, 0.2)",
      "300": "rgba(242, 61, 79, 0.3)",
      "400": "rgba(242, 61, 79, 0.4)",
      "500": "#f23d4f",
      "600": "#d12440",
      "700": "#a61d33",
      "800": "#801627"
    },
    yellow: {
      "500": "#ffe600"
    }
  }
};
var colorSchemes = {
  light: {
    color: {
      background: {
        primary: "{color.gray.solid.70}",
        secondary: "{color.gray.solid.40}",
        tertiary: "{color.white.base}",
        overlay: "{color.gray.800}"
      },
      surface: {
        primary: "{color.white.base}",
        secondary: "{color.gray.solid.40}",
        tertiary: "{color.gray.solid.70}",
        hover: "{color.gray.40}",
        active: "{color.gray.70}"
      },
      fill: {
        primary: "{color.white.base}",
        secondary: "{color.gray.40}",
        tertiary: "{color.gray.100}",
        inverse: "{color.gray.solid.900}",
        disabled: "{color.gray.100}",
        accent: "{color.cyan.500}"
      },
      text: {
        primary: "{color.gray.900}",
        secondary: "{color.gray.550}",
        accent: "{color.cyan.500}",
        inverse: "{color.white.base}",
        disabled: "{color.gray.250}",
        link: {
          "default": {
            base: "{color.cyan.500}",
            inverse: "{color.white.base}"
          },
          hover: {
            base: "{color.cyan.600}",
            inverse: "{color.white.base}"
          },
          active: {
            base: "{color.cyan.700}",
            inverse: "{color.white.base}"
          }
        }
      },
      border: {
        primary: "{color.gray.100}",
        inverse: "{color.white.base}",
        disabled: "{color.gray.250}"
      },
      icon: {
        primary: "{color.gray.900}",
        secondary: "{color.gray.550}",
        accent: "{color.cyan.500}",
        inverse: "{color.white.base}",
        disabled: "{color.gray.250}"
      },
      feedback: {
        fill: {
          informative: {
            loud: {
              base: "{color.cyan.500}",
              onFill: "{color.cyan.600}"
            },
            quiet: "{color.cyan.100}"
          },
          positive: {
            loud: {
              base: "{color.green.500}",
              onFill: "{color.green.600}"
            },
            quiet: "{color.green.100}"
          },
          caution: {
            loud: {
              base: "{color.orange.500}",
              onFill: "{color.orange.600}"
            },
            quiet: "{color.orange.100}"
          },
          negative: {
            loud: {
              base: "{color.red.500}",
              onFill: "{color.red.600}"
            },
            quiet: "{color.red.100}"
          }
        },
        text: {
          informative: {
            loud: "{color.cyan.500}"
          },
          positive: {
            loud: "{color.green.500}"
          },
          caution: {
            loud: "{color.orange.500}"
          },
          negative: {
            loud: "{color.red.500}"
          }
        },
        border: {
          informative: {
            loud: {
              base: "{color.cyan.500}",
              onFill: "{color.cyan.600}"
            }
          },
          positive: {
            loud: {
              base: "{color.green.500}",
              onFill: "{color.green.600}"
            }
          },
          caution: {
            loud: {
              base: "{color.orange.500}",
              onFill: "{color.orange.600}"
            }
          },
          negative: {
            loud: {
              base: "{color.red.500}",
              onFill: "{color.red.600}"
            }
          }
        },
        icon: {
          informative: {
            loud: "{color.cyan.500}"
          },
          positive: {
            loud: "{color.green.500}"
          },
          caution: {
            loud: "{color.orange.500}"
          },
          negative: {
            loud: "{color.red.500}"
          }
        }
      },
      interactive: {
        fill: {
          loud: {
            "default": "{color.cyan.500}",
            hover: "{color.cyan.600}",
            active: "{color.cyan.700}"
          },
          quiet: {
            "default": "{color.cyan.150}",
            hover: "{color.cyan.200}",
            active: "{color.cyan.300}"
          },
          mute: {
            "default": "{color.transparent}",
            hover: "{color.cyan.100}",
            active: "{color.cyan.200}"
          }
        },
        border: {
          "default": "{color.gray.250}",
          hover: "{color.gray.550}",
          active: "{color.cyan.500}"
        },
        icon: {
          "default": {
            base: "{color.gray.550}",
            accent: "{color.cyan.500}",
            inverse: "{color.white.base}"
          },
          hover: {
            base: "{color.gray.800}",
            accent: "{color.cyan.600}"
          },
          active: {
            base: "{color.gray.900}",
            accent: "{color.cyan.700}"
          }
        }
      },
      selected: {
        fill: {
          "default": "{color.cyan.500}",
          hover: "{color.cyan.600}",
          active: "{color.cyan.700}"
        },
        text: {
          "default": "{color.cyan.500}",
          hover: "{color.cyan.600}",
          active: "{color.cyan.700}"
        },
        border: {
          "default": "{color.cyan.500}",
          hover: "{color.cyan.600}",
          active: "{color.cyan.700}"
        },
        icon: {
          "default": "{color.cyan.500}",
          hover: "{color.cyan.600}",
          active: "{color.cyan.700}"
        }
      },
      brand: {
        text: {
          onFill: "{color.white.base}"
        },
        icon: {
          onFill: "{color.white.base}"
        },
        fill: "{color.cyan.500}"
      }
    }
  }
};
var components = {
  accordion: {
    color: {
      surface: {
        "default": "{color.transparent}",
        hover: "{color.surface.hover}",
        active: "{color.surface.active}",
        disabled: "{color.transparent}"
      },
      title: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      },
      description: {
        "default": "{color.text.secondary}",
        disabled: "{color.text.disabled}"
      },
      border: "{color.border.primary}"
    }
  },
  amountField: {
    color: {
      currency: {
        "default": "{color.text.primary}",
        active: "{color.text.primary}",
        error: "{color.text.primary}",
        errorShake: "{color.feedback.text.negative.loud}"
      },
      amount: {
        "default": "{color.text.primary}",
        active: "{color.text.primary}",
        error: "{color.text.primary}",
        errorShake: "{color.feedback.text.negative.loud}"
      },
      suffix: {
        "default": "{color.text.secondary}",
        active: "{color.text.secondary}",
        error: "{color.text.secondary}",
        errorShake: "{color.feedback.text.negative.loud}"
      },
      percentage: {
        "default": "{color.text.primary}",
        active: "{color.text.primary}",
        error: "{color.text.primary}",
        errorShake: "{color.feedback.text.negative.loud}"
      },
      placeholder: {
        "default": "{color.text.secondary}"
      }
    },
    converter: {
      color: {
        equivalence: {
          "default": "{color.text.secondary}"
        },
        suffix: {
          "default": "{color.text.secondary}"
        }
      },
      swapButton: {
        color: {
          fill: {
            "default": "{color.interactive.fill.quiet.default}",
            hover: "{color.interactive.fill.quiet.hover}",
            active: "{color.interactive.fill.quiet.active}"
          },
          icon: {
            "default": "{color.icon.accent}"
          }
        }
      }
    }
  },
  badge: {
    loud: {
      color: {
        fill: {
          informative: {
            base: "{color.feedback.fill.informative.loud.base}",
            onComplexBg: "{color.feedback.fill.informative.loud.onFill}"
          },
          positive: {
            base: "{color.feedback.fill.positive.loud.base}",
            onComplexBg: "{color.feedback.fill.positive.loud.onFill}"
          },
          caution: {
            base: "{color.feedback.fill.caution.loud.base}",
            onComplexBg: "{color.feedback.fill.caution.loud.onFill}"
          },
          negative: {
            base: "{color.feedback.fill.negative.loud.base}",
            onComplexBg: "{color.feedback.fill.negative.loud.onFill}"
          },
          neutral: {
            base: "{color.gray.solid.550}",
            onComplexBg: "{color.gray.solid.900}"
          },
          inverse: "{color.fill.primary}"
        },
        text: {
          informative: {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          positive: {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          caution: {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          negative: {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          neutral: {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          inverse: "{color.text.primary}"
        },
        icon: {
          informative: {
            base: "{color.icon.inverse}",
            onComplexBg: "{color.icon.inverse}"
          },
          positive: {
            base: "{color.icon.inverse}",
            onComplexBg: "{color.icon.inverse}"
          },
          caution: {
            base: "{color.icon.inverse}",
            onComplexBg: "{color.icon.inverse}"
          },
          negative: {
            base: "{color.icon.inverse}",
            onComplexBg: "{color.icon.inverse}"
          },
          neutral: {
            base: "{color.icon.inverse}",
            onComplexBg: "{color.icon.inverse}"
          },
          inverse: "{color.icon.primary}"
        }
      }
    },
    quiet: {
      color: {
        fill: {
          informative: "{color.feedback.fill.informative.quiet}",
          positive: "{color.feedback.fill.positive.quiet}",
          caution: "{color.feedback.fill.caution.quiet}",
          negative: "{color.feedback.fill.negative.quiet}",
          neutral: "{color.gray.100}"
        },
        text: {
          informative: "{color.feedback.text.informative.loud}",
          positive: "{color.feedback.text.positive.loud}",
          caution: "{color.feedback.text.caution.loud}",
          negative: "{color.feedback.text.negative.loud}",
          neutral: "{color.text.secondary}"
        },
        icon: {
          informative: "{color.feedback.icon.informative.loud}",
          positive: "{color.feedback.icon.positive.loud}",
          caution: "{color.feedback.icon.caution.loud}",
          negative: "{color.feedback.icon.negative.loud}",
          neutral: "{color.gray.solid.550}"
        }
      }
    },
    dot: {
      quiet: {
        color: {
          fill: {
            informative: "{color.cyan.400}",
            positive: "{color.green.400}",
            caution: "{color.orange.400}",
            negative: "{color.red.400}",
            neutral: "{color.gray.250}"
          }
        }
      }
    }
  },
  bottomSheet: {
    header: {
      color: {
        fill: "{color.fill.primary}",
        drag: {
          fill: "{color.gray.250}"
        },
        title: "{color.text.primary}",
        back: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}"
        },
        close: {
          "default": {
            base: "{color.interactive.icon.default.base}",
            inverse: "{color.interactive.icon.default.inverse}"
          },
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}"
        }
      }
    },
    color: {
      fill: "{color.fill.primary}"
    },
    dimmer: {
      color: {
        background: "{color.background.overlay}"
      }
    }
  },
  boxSelector: {
    color: {
      selected: {
        border: {
          "default": "{color.selected.border.default}",
          hover: "{color.selected.border.hover}",
          active: "{color.selected.border.active}",
          disabled: "{color.gray.100}"
        }
      },
      unselected: {
        border: {
          "default": "{color.interactive.border.default}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.gray.100}"
        }
      }
    }
  },
  breadcrumb: {
    currentLevel: {
      color: {
        text: "{color.text.primary}"
      }
    },
    fatherLevel: {
      color: {
        text: {
          "default": "{color.text.link.default.base}",
          hover: "{color.text.link.hover.base}",
          active: "{color.text.link.active.base}"
        }
      }
    },
    color: {
      chevron: "{color.icon.secondary}"
    }
  },
  button: {
    loud: {
      color: {
        fill: {
          "default": {
            base: "{color.interactive.fill.loud.default}",
            onComplexBg: "{color.gray.450}"
          },
          hover: {
            base: "{color.interactive.fill.loud.hover}",
            onComplexBg: "{color.gray.550}"
          },
          active: {
            base: "{color.interactive.fill.loud.active}",
            onComplexBg: "{color.gray.800}"
          },
          progress: {
            base: "{color.cyan.600}",
            onComplexBg: "{color.gray.550}"
          },
          disabled: {
            base: "{color.fill.disabled}",
            onComplexBg: "{color.gray.150}"
          }
        },
        text: {
          "default": {
            base: "{color.text.inverse}",
            onComplexBg: "{color.text.inverse}"
          },
          disabled: {
            base: "{color.text.disabled}",
            onComplexBg: "{color.text.disabled}"
          }
        },
        spinner: {
          base: "{color.fill.primary}",
          onComplexBg: "{color.fill.primary}"
        }
      }
    },
    quiet: {
      color: {
        fill: {
          "default": {
            base: "{color.interactive.fill.quiet.default}",
            onComplexBg: "{color.gray.150}"
          },
          hover: {
            base: "{color.interactive.fill.quiet.hover}",
            onComplexBg: "{color.gray.250}"
          },
          active: {
            base: "{color.interactive.fill.quiet.active}",
            onComplexBg: "{color.gray.450}"
          },
          progress: {
            base: "{color.cyan.300}",
            onComplexBg: "{color.gray.250}"
          },
          disabled: {
            base: "{color.fill.disabled}",
            onComplexBg: "{color.gray.150}"
          }
        },
        text: {
          "default": {
            base: "{color.text.accent}",
            onComplexBg: "{color.text.inverse}"
          },
          disabled: {
            base: "{color.text.disabled}",
            onComplexBg: "{color.text.disabled}"
          }
        },
        spinner: {
          base: "{color.feedback.fill.informative.loud.base}",
          onComplexBg: "{color.fill.primary}"
        }
      }
    },
    mute: {
      color: {
        fill: {
          "default": {
            base: "{color.interactive.fill.mute.default}",
            onComplexBg: "{color.transparent}"
          },
          hover: {
            base: "{color.interactive.fill.mute.hover}",
            onComplexBg: "{color.gray.70}"
          },
          active: {
            base: "{color.interactive.fill.mute.active}",
            onComplexBg: "{color.gray.150}"
          },
          disabled: {
            base: "{color.transparent}",
            onComplexBg: "{color.transparent}"
          }
        },
        text: {
          "default": {
            base: "{color.text.accent}",
            onComplexBg: "{color.text.inverse}"
          },
          disabled: {
            base: "{color.text.disabled}",
            onComplexBg: "{color.text.disabled}"
          }
        },
        spinner: {
          base: "{color.feedback.fill.informative.loud.base}",
          onComplexBg: "{color.fill.primary}"
        }
      }
    },
    icon: {
      loud: {
        color: {
          icon: {
            "default": {
              base: "{color.icon.inverse}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      },
      quiet: {
        color: {
          icon: {
            "default": {
              base: "{color.icon.accent}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      },
      mute: {
        color: {
          icon: {
            "default": {
              base: "{color.icon.accent}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      }
    },
    split: {
      loud: {
        color: {
          fill: {
            "default": {
              base: "{color.interactive.fill.loud.default}",
              onComplexBg: "{color.gray.450}"
            },
            hover: {
              base: "{color.interactive.fill.loud.hover}",
              onComplexBg: "{color.gray.550}"
            },
            active: {
              base: "{color.interactive.fill.loud.active}",
              onComplexBg: "{color.gray.800}"
            },
            disabled: {
              base: "{color.fill.disabled}",
              onComplexBg: "{color.gray.150}"
            }
          },
          chevron: {
            "default": {
              base: "{color.icon.inverse}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      },
      quiet: {
        color: {
          fill: {
            "default": {
              base: "{color.interactive.fill.quiet.default}",
              onComplexBg: "{color.gray.150}"
            },
            hover: {
              base: "{color.interactive.fill.quiet.hover}",
              onComplexBg: "{color.gray.250}"
            },
            active: {
              base: "{color.interactive.fill.quiet.active}",
              onComplexBg: "{color.gray.450}"
            },
            disabled: {
              base: "{color.fill.disabled}",
              onComplexBg: "{color.gray.150}"
            }
          },
          chevron: {
            "default": {
              base: "{color.icon.accent}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      },
      color: {
        border: {
          "default": "{color.border.primary}"
        }
      }
    },
    dropdown: {
      loud: {
        color: {
          chevron: {
            "default": {
              base: "{color.icon.inverse}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      },
      quiet: {
        color: {
          chevron: {
            "default": {
              base: "{color.icon.accent}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.icon.disabled}"
            }
          }
        }
      }
    }
  },
  card: {
    container: {
      color: {
        surface: {
          primary: "{color.surface.primary}",
          secondary: "{color.surface.secondary}",
          tertiary: "{color.surface.tertiary}"
        }
      }
    },
    color: {
      surface: {
        primary: {
          "default": "{color.transparent}",
          hover: "{color.surface.hover}",
          active: "{color.surface.active}"
        },
        secondary: {
          "default": "{color.transparent}",
          hover: "{color.surface.hover}",
          active: "{color.surface.active}"
        },
        tertiary: {
          "default": "{color.transparent}",
          hover: "{color.surface.hover}",
          active: "{color.surface.active}"
        }
      },
      border: "{color.border.primary}"
    },
    pipe: {
      color: {
        fill: {
          informative: "{color.feedback.fill.informative.loud.base}",
          positive: "{color.feedback.fill.positive.loud.base}",
          caution: "{color.feedback.fill.caution.loud.base}",
          negative: "{color.feedback.fill.negative.loud.base}"
        }
      }
    },
    extended: {
      header: {
        color: {
          title: "{color.text.primary}",
          border: "{color.border.primary}"
        }
      },
      footer: {
        color: {
          chevron: {
            "default": "{color.interactive.icon.default.accent}",
            hover: "{color.interactive.icon.hover.accent}",
            active: "{color.interactive.icon.active.accent}"
          },
          border: "{color.border.primary}"
        }
      }
    }
  },
  carousel: {
    controls: {
      color: {
        fill: {
          "default": "{color.fill.primary}"
        },
        icon: {
          "default": "{color.interactive.icon.default.accent}",
          hover: "{color.interactive.icon.hover.accent}"
        }
      }
    }
  },
  checkbox: {
    color: {
      selected: {
        fill: {
          "default": "{color.selected.fill.default}",
          hover: "{color.selected.fill.hover}",
          active: "{color.selected.fill.active}",
          disabled: "{color.fill.disabled}"
        }
      },
      indeterminate: {
        fill: {
          "default": "{color.selected.fill.default}",
          hover: "{color.selected.fill.hover}",
          active: "{color.selected.fill.active}",
          disabled: "{color.fill.disabled}"
        }
      },
      unselected: {
        fill: {
          "default": "{color.fill.primary}",
          disabled: "{color.gray.40}"
        },
        border: {
          "default": "{color.interactive.border.default}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.gray.100}"
        }
      },
      icon: {
        "default": "{color.icon.inverse}",
        disabled: "{color.icon.disabled}"
      },
      text: {
        "default": "{color.text.primary}",
        error: "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    highlight: {
      color: {
        fill: {
          "default": "{color.cyan.200}",
          disabled: "{color.gray.40}",
          error: "{color.red.300}"
        },
        selected: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}",
            disabled: "{color.fill.disabled}"
          }
        },
        indeterminate: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}",
            disabled: "{color.fill.disabled}"
          }
        },
        unselected: {
          fill: {
            "default": "{color.fill.primary}",
            disabled: "{color.gray.40}"
          },
          border: {
            "default": "{color.cyan.500}",
            hover: "{color.cyan.600}",
            active: "{color.cyan.700}",
            error: "{color.feedback.border.negative.loud.base}",
            disabled: "{color.gray.100}"
          }
        },
        icon: {
          "default": "{color.icon.inverse}",
          disabled: "{color.icon.disabled}"
        },
        text: {
          "default": "{color.text.primary}",
          error: "{color.text.primary}",
          disabled: "{color.text.disabled}"
        }
      }
    }
  },
  datePicker: {
    day: {
      color: {
        selected: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          },
          text: {
            "default": "{color.text.inverse}",
            hover: "{color.text.inverse}",
            active: "{color.text.inverse}"
          }
        },
        unselected: {
          fill: {
            "default": "{color.interactive.fill.mute.default}",
            hover: "{color.interactive.fill.mute.hover}",
            active: "{color.interactive.fill.mute.active}",
            disabled: "{color.transparent}"
          },
          text: {
            "default": "{color.text.primary}",
            hover: "{color.text.primary}",
            active: "{color.text.primary}",
            disabled: "{color.text.disabled}"
          }
        }
      },
      event: {
        color: {
          selected: {
            fill: {
              "default": "{color.selected.fill.default}",
              hover: "{color.selected.fill.hover}",
              active: "{color.selected.fill.active}"
            },
            text: {
              "default": "{color.text.inverse}",
              hover: "{color.text.inverse}",
              active: "{color.text.inverse}"
            }
          },
          unselected: {
            fill: {
              "default": "{color.interactive.fill.mute.default}",
              hover: "{color.interactive.fill.mute.hover}",
              active: "{color.interactive.fill.mute.active}",
              disabled: "{color.transparent}"
            },
            text: {
              "default": "{color.text.primary}",
              hover: "{color.text.primary}",
              active: "{color.text.primary}",
              disabled: "{color.text.disabled}"
            }
          }
        },
        dot: {
          color: {
            selected: {
              fill: {
                "default": "{color.fill.primary}",
                hover: "{color.fill.primary}",
                active: "{color.fill.primary}"
              }
            },
            unselected: {
              fill: {
                "default": "{color.selected.fill.default}",
                hover: "{color.selected.fill.hover}",
                active: "{color.selected.fill.active}",
                disabled: "{color.gray.250}"
              }
            }
          }
        }
      }
    },
    rangeDay: {
      color: {
        fill: {
          "default": "{color.cyan.150}"
        }
      },
      from: {
        color: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          },
          text: {
            "default": "{color.text.inverse}",
            hover: "{color.text.inverse}",
            active: "{color.text.inverse}"
          }
        }
      },
      middle: {
        color: {
          fill: {
            "default": "{color.interactive.fill.mute.default}",
            hover: "{color.interactive.fill.mute.hover}",
            active: "{color.interactive.fill.mute.active}"
          },
          text: {
            "default": "{color.text.primary}",
            hover: "{color.text.primary}",
            active: "{color.text.primary}"
          }
        }
      },
      to: {
        color: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          },
          text: {
            "default": "{color.text.inverse}",
            hover: "{color.text.inverse}",
            active: "{color.text.inverse}"
          }
        }
      }
    },
    month: {
      color: {
        selected: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          },
          text: {
            "default": "{color.text.inverse}",
            hover: "{color.text.inverse}",
            active: "{color.text.inverse}"
          }
        },
        unselected: {
          fill: {
            "default": "{color.interactive.fill.mute.default}",
            hover: "{color.interactive.fill.mute.hover}",
            active: "{color.interactive.fill.mute.active}",
            disabled: "{color.transparent}"
          },
          text: {
            "default": "{color.text.primary}",
            hover: "{color.text.primary}",
            active: "{color.text.primary}",
            disabled: "{color.text.disabled}"
          }
        }
      }
    },
    standalone: {
      color: {
        fill: "{color.transparent}",
        border: "{color.border.primary}"
      },
      dateSelector: {
        color: {
          text: {
            "default": "{color.text.primary}",
            hover: "{color.text.primary}",
            active: "{color.text.primary}",
            disabled: "{color.text.primary}"
          },
          chevron: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}",
            disabled: "{color.icon.disabled}"
          }
        }
      },
      weekdays: {
        color: {
          text: {
            "default": "{color.text.secondary}"
          }
        }
      },
      controls: {
        color: {
          chevron: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}",
            disabled: "{color.icon.disabled}"
          }
        }
      }
    },
    field: {
      color: {
        fill: {
          "default": "{color.fill.primary}",
          disabled: "{color.gray.40}"
        },
        label: {
          "default": "{color.text.primary}",
          error: "{color.feedback.text.negative.loud}",
          disabled: "{color.text.disabled}",
          readOnly: "{color.text.secondary}"
        },
        placeholder: {
          "default": "{color.text.secondary}"
        },
        calendar: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}",
          disabled: "{color.icon.disabled}"
        },
        border: {
          "default": "{color.interactive.border.default}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.border.disabled}"
        },
        value: {
          "default": "{color.text.primary}",
          disabled: "{color.text.disabled}"
        }
      }
    },
    color: {
      fill: "{color.fill.primary}"
    },
    extended: {
      color: {
        fill: "{color.fill.primary}",
        border: "{color.border.primary}"
      }
    }
  },
  dropdown: {
    color: {
      fill: {
        "default": "{color.fill.primary}",
        disabled: "{color.gray.40}"
      },
      label: {
        "default": "{color.text.primary}",
        error: "{color.feedback.text.negative.loud}",
        disabled: "{color.text.disabled}",
        readOnly: "{color.text.secondary}"
      },
      placeholder: {
        "default": "{color.text.secondary}"
      },
      chevron: {
        "default": "{color.interactive.icon.default.accent}",
        hover: "{color.interactive.icon.hover.accent}",
        active: "{color.interactive.icon.active.accent}",
        disabled: "{color.icon.disabled}"
      },
      icon: {
        "default": "{color.icon.primary}",
        disabled: "{color.icon.disabled}"
      },
      border: {
        "default": "{color.interactive.border.default}",
        hover: "{color.interactive.border.hover}",
        active: "{color.interactive.border.active}",
        error: "{color.feedback.border.negative.loud.base}",
        disabled: "{color.border.disabled}"
      },
      value: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    standalone: {
      color: {
        chevron: {
          "default": "{color.interactive.icon.default.base}",
          active: "{color.interactive.icon.active.base}",
          disabled: "{color.icon.disabled}"
        },
        icon: {
          "default": "{color.icon.primary}",
          disabled: "{color.icon.disabled}"
        },
        value: {
          "default": "{color.text.primary}",
          disabled: "{color.text.disabled}"
        }
      }
    }
  },
  empty: {
    states: {
      color: {
        title: "{color.text.primary}",
        description: "{color.text.primary}"
      }
    }
  },
  expandableSection: {
    color: {
      text: {
        "default": {
          base: "{color.text.primary}",
          accent: "{color.text.accent}"
        },
        disabled: "{color.text.disabled}"
      },
      chevron: {
        "default": {
          base: "{color.interactive.icon.default.base}",
          accent: "{color.interactive.icon.default.accent}"
        },
        hover: {
          base: "{color.interactive.icon.hover.base}",
          accent: "{color.interactive.icon.hover.accent}"
        },
        active: {
          base: "{color.interactive.icon.active.base}",
          accent: "{color.interactive.icon.active.accent}"
        },
        disabled: "{color.icon.disabled}"
      }
    },
    highlighted: {
      controls: {
        color: {
          fill: {
            "default": "{color.fill.primary}"
          },
          icon: {
            "default": "{color.interactive.icon.default.accent}",
            hover: "{color.interactive.icon.hover.accent}"
          }
        }
      }
    }
  },
  feedbackScreens: {
    header: {
      color: {
        title: "{color.text.primary}",
        description: "{color.text.primary}",
        overline: "{color.text.primary}",
        highlighted: {
          positive: "{color.feedback.text.positive.loud}",
          caution: "{color.feedback.text.caution.loud}",
          negative: "{color.feedback.text.negative.loud}"
        }
      },
      asset: {
        color: {
          fill: {
            positive: "{color.fill.primary}",
            caution: "{color.fill.primary}",
            negative: "{color.fill.primary}"
          },
          icon: {
            positive: "{color.feedback.icon.positive.loud}",
            caution: "{color.feedback.icon.caution.loud}",
            negative: "{color.feedback.icon.negative.loud}"
          },
          border: {
            positive: "{color.feedback.border.positive.loud.base}",
            caution: "{color.feedback.border.caution.loud.base}",
            negative: "{color.feedback.border.negative.loud.base}"
          }
        }
      }
    },
    statusBar: {
      color: {
        fill: "{color.brand.fill}",
        icon: "{color.icon.primary}"
      }
    },
    color: {
      background: {
        base: "{color.background.tertiary}",
        secondary: "{color.background.secondary}"
      },
      close: {
        "default": "{color.interactive.icon.default.base}",
        active: "{color.interactive.icon.active.base}"
      }
    },
    extended: {
      color: {
        background: "{color.background.secondary}"
      }
    },
    congrats: {
      header: {
        color: {
          title: "{color.text.primary}",
          description: "{color.text.primary}",
          overline: "{color.text.primary}",
          highlighted: "{color.feedback.text.positive.loud}"
        },
        asset: {
          color: {
            fill: "{color.fill.primary}",
            icon: "{color.feedback.icon.positive.loud}",
            border: "{color.feedback.border.positive.loud.base}"
          }
        }
      },
      statusBar: {
        color: {
          fill: "{color.feedback.fill.positive.loud.base}",
          icon: "{color.icon.inverse}"
        }
      },
      color: {
        background: "{color.background.secondary}",
        close: {
          "default": "{color.icon.inverse}"
        }
      }
    }
  },
  fab: {
    container: {
      loud: {
        color: {
          fill: "{color.fill.accent}"
        }
      },
      quiet: {
        color: {
          fill: "{color.fill.primary}"
        }
      }
    },
    loud: {
      color: {
        fill: {
          "default": "{color.interactive.fill.loud.default}",
          hover: "{color.interactive.fill.loud.hover}",
          active: "{color.interactive.fill.loud.active}"
        },
        text: {
          "default": "{color.text.inverse}"
        },
        icon: {
          "default": "{color.icon.inverse}"
        }
      }
    },
    quiet: {
      color: {
        fill: {
          "default": "{color.fill.primary}",
          hover: "{color.interactive.fill.quiet.hover}",
          active: "{color.interactive.fill.quiet.active}"
        },
        text: {
          "default": "{color.text.accent}"
        },
        icon: {
          "default": "{color.icon.accent}"
        }
      }
    }
  },
  fileUploader: {
    dropzone: {
      color: {
        fill: {
          "default": "{color.transparent}",
          hover: "{color.interactive.fill.quiet.hover}"
        },
        action: {
          "default": "{color.text.accent}",
          disabled: "{color.text.disabled}"
        },
        dropInstruction: {
          "default": "{color.text.primary}",
          disabled: "{color.text.disabled}"
        },
        description: {
          "default": "{color.text.secondary}",
          disabled: "{color.text.disabled}"
        },
        icon: {
          "default": "{color.icon.accent}",
          disabled: "{color.icon.disabled}"
        },
        border: {
          "default": "{color.feedback.border.informative.loud.base}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.interactive.border.default}"
        }
      }
    },
    thumbnail: {
      close: {
        color: {
          fill: {
            "default": "{color.fill.primary}"
          },
          icon: {
            "default": "{color.icon.primary}"
          },
          border: "{color.gray.solid.70}"
        }
      }
    },
    color: {
      label: {
        "default": "{color.text.primary}",
        error: "{color.feedback.text.negative.loud}",
        disabled: "{color.text.disabled}"
      }
    },
    tag: {
      color: {
        fileSize: "{color.text.secondary}"
      }
    }
  },
  floatingMenu: {
    color: {
      fill: "{color.fill.primary}"
    },
    textResult: {
      color: {
        text: "{color.text.secondary}"
      }
    }
  },
  helper: {
    color: {
      text: {
        "default": "{color.text.secondary}",
        caution: "{color.text.secondary}",
        error: "{color.feedback.text.negative.loud}"
      }
    }
  },
  inputStepper: {
    color: {
      fill: {
        "default": "{color.fill.primary}",
        disabled: "{color.gray.40}",
        readOnly: "{color.transparent}"
      },
      complementary: {
        "default": "{color.text.secondary}",
        readOnly: "{color.text.primary}",
        disabled: "{color.text.disabled}"
      },
      label: {
        "default": "{color.text.primary}",
        error: "{color.feedback.text.negative.loud}",
        disabled: "{color.text.disabled}",
        readOnly: "{color.text.secondary}"
      },
      helper: {
        "default": "{color.text.secondary}",
        error: "{color.feedback.text.negative.loud}"
      },
      placeholder: "{color.text.secondary}",
      spinner: "{color.feedback.fill.informative.loud.base}",
      border: {
        "default": "{color.interactive.border.default}",
        hover: "{color.interactive.border.hover}",
        disabled: "{color.interactive.border.default}",
        active: "{color.feedback.border.informative.loud.base}",
        error: "{color.feedback.border.negative.loud.base}"
      },
      value: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    controls: {
      color: {
        fill: {
          "default": "{color.interactive.fill.mute.default}",
          hover: "{color.interactive.fill.mute.hover}",
          active: "{color.interactive.fill.mute.active}"
        },
        icon: {
          "default": "{color.icon.accent}",
          disabled: "{color.icon.disabled}"
        }
      }
    }
  },
  list: {
    color: {
      surface: {
        "default": "{color.transparent}",
        hover: "{color.surface.hover}",
        active: "{color.surface.active}",
        disabled: "{color.transparent}"
      },
      title: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      },
      description: {
        "default": "{color.text.secondary}",
        disabled: "{color.text.disabled}"
      },
      icon: {
        right: "{color.icon.primary}",
        left: "{color.icon.primary}",
        disabled: "{color.icon.disabled}"
      },
      border: "{color.border.primary}"
    },
    chevron: {
      color: {
        icon: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}",
          disabled: "{color.icon.disabled}"
        }
      }
    },
    pickList: {
      color: {
        selected: {
          title: {
            "default": "{color.selected.text.default}",
            hover: "{color.selected.text.hover}",
            active: "{color.selected.text.active}"
          },
          icon: {
            "default": "{color.selected.icon.default}",
            hover: "{color.selected.icon.hover}",
            active: "{color.selected.icon.active}"
          },
          pipe: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          }
        }
      }
    },
    group: {
      header: {
        color: {
          fill: "{color.transparent}",
          text: "{color.text.secondary}",
          border: "{color.border.primary}"
        }
      }
    }
  },
  message: {
    loud: {
      color: {
        fill: {
          informative: "{color.feedback.fill.informative.loud.base}",
          positive: "{color.feedback.fill.positive.loud.base}",
          caution: "{color.feedback.fill.caution.loud.base}",
          negative: "{color.feedback.fill.negative.loud.base}"
        },
        title: {
          informative: "{color.text.inverse}",
          positive: "{color.text.inverse}",
          caution: "{color.text.inverse}",
          negative: "{color.text.inverse}"
        },
        description: {
          informative: "{color.text.inverse}",
          positive: "{color.text.inverse}",
          caution: "{color.text.inverse}",
          negative: "{color.text.inverse}"
        },
        close: {
          informative: {
            "default": "{color.interactive.icon.default.inverse}"
          },
          positive: {
            "default": "{color.interactive.icon.default.inverse}"
          },
          caution: {
            "default": "{color.interactive.icon.default.inverse}"
          },
          negative: {
            "default": "{color.interactive.icon.default.inverse}"
          }
        }
      }
    },
    quiet: {
      color: {
        fill: {
          informative: "{color.fill.secondary}",
          positive: "{color.fill.secondary}",
          caution: "{color.fill.secondary}",
          negative: "{color.fill.secondary}"
        },
        title: {
          informative: "{color.text.primary}",
          positive: "{color.text.primary}",
          caution: "{color.text.primary}",
          negative: "{color.text.primary}"
        },
        description: {
          informative: "{color.text.primary}",
          positive: "{color.text.primary}",
          caution: "{color.text.primary}",
          negative: "{color.text.primary}"
        },
        close: {
          informative: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}"
          },
          positive: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}"
          },
          caution: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}"
          },
          negative: {
            "default": "{color.interactive.icon.default.base}",
            hover: "{color.interactive.icon.hover.base}",
            active: "{color.interactive.icon.active.base}"
          }
        }
      },
      pipe: {
        color: {
          fill: {
            informative: "{color.feedback.fill.informative.loud.base}",
            positive: "{color.feedback.fill.positive.loud.base}",
            caution: "{color.feedback.fill.caution.loud.base}",
            negative: "{color.feedback.fill.negative.loud.base}"
          }
        }
      }
    }
  },
  moneyAmount: {
    color: {
      amount: "{color.text.primary}",
      currency: "{color.text.primary}",
      iso: "{color.text.primary}",
      suffix: "{color.text.secondary}",
      previous: "{color.text.secondary}",
      discount: "{color.feedback.text.positive.loud}"
    }
  },
  pagination: {
    color: {
      selected: {
        fill: {
          "default": "{color.fill.primary}",
          hover: "{color.cyan.150}",
          active: "{color.cyan.200}"
        },
        border: {
          "default": "{color.selected.border.default}",
          hover: "{color.selected.border.hover}",
          active: "{color.selected.border.active}"
        },
        text: {
          "default": "{color.text.primary}"
        }
      },
      unselected: {
        fill: {
          "default": "{color.transparent}",
          hover: "{color.cyan.150}",
          active: "{color.cyan.200}"
        },
        text: {
          "default": "{color.text.secondary}"
        }
      }
    },
    controls: {
      color: {
        fill: {
          "default": "{color.transparent}",
          hover: "{color.cyan.150}",
          active: "{color.cyan.200}",
          disabled: "{color.transparent}"
        },
        text: {
          "default": "{color.text.secondary}",
          disabled: "{color.text.disabled}"
        },
        chevron: {
          "default": "{color.icon.secondary}",
          disabled: "{color.icon.disabled}"
        }
      }
    }
  },
  pageViewer: {
    dot: {
      color: {
        selected: {
          fill: {
            base: "{color.selected.fill.default}",
            inverse: "{color.fill.primary}"
          },
          border: {
            inverse: "{color.gray.250}"
          }
        },
        unselected: {
          fill: {
            base: "{color.fill.tertiary}",
            inverse: "{color.gray.250}"
          },
          border: {
            inverse: "{color.border.inverse}"
          }
        }
      }
    }
  },
  progressIndicator: {
    circular: {
      color: {
        fill: {
          base: "{color.fill.tertiary}",
          inverse: "{color.white.300}"
        },
        label: {
          base: "{color.text.primary}",
          inverse: "{color.text.inverse}"
        },
        counter: {
          base: "{color.text.primary}",
          inverse: "{color.text.inverse}"
        },
        icon: {
          negative: "{color.feedback.icon.negative.loud}",
          positive: "{color.feedback.icon.positive.loud}"
        }
      },
      progress: {
        color: {
          fill: {
            "default": "{color.feedback.fill.informative.loud.base}",
            inverse: "{color.fill.primary}",
            negative: "{color.feedback.fill.negative.loud.base}",
            positive: "{color.feedback.fill.positive.loud.base}"
          }
        }
      }
    },
    linear: {
      progress: {
        color: {
          fill: {
            "default": "{color.feedback.fill.informative.loud.base}",
            inverse: "{color.fill.primary}",
            negative: "{color.feedback.fill.negative.loud.base}",
            positive: "{color.feedback.fill.positive.loud.base}"
          }
        }
      },
      color: {
        fill: {
          base: "{color.fill.tertiary}",
          inverse: "{color.white.300}"
        },
        label: {
          base: "{color.text.primary}",
          inverse: "{color.text.inverse}"
        },
        counter: {
          large: {
            base: "{color.text.primary}",
            inverse: "{color.text.inverse}"
          },
          small: {
            base: "{color.text.secondary}",
            inverse: "{color.text.inverse}"
          }
        }
      },
      highlight: {
        color: {
          fill: "{color.feedback.fill.informative.loud.base}"
        }
      },
      comparison: {
        color: {
          fill: {
            "default": "{color.gray.550}"
          },
          label: "{color.text.secondary}"
        }
      }
    }
  },
  radioButton: {
    color: {
      selected: {
        fill: {
          "default": "{color.fill.primary}"
        },
        border: {
          "default": "{color.selected.border.default}",
          hover: "{color.selected.border.hover}",
          active: "{color.selected.border.active}",
          disabled: "{color.gray.100}"
        }
      },
      unselected: {
        fill: {
          "default": "{color.fill.primary}",
          disabled: "{color.gray.40}"
        },
        border: {
          "default": "{color.interactive.border.default}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.gray.100}"
        }
      },
      text: {
        "default": "{color.text.primary}",
        error: "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    tick: {
      color: {
        selected: {
          fill: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}",
            disabled: "{color.fill.disabled}"
          }
        }
      }
    }
  },
  searchbox: {
    color: {
      fill: {
        "default": "{color.fill.primary}"
      },
      inputText: {
        "default": "{color.text.primary}",
        loading: "{color.text.secondary}"
      },
      placeholder: {
        "default": "{color.text.secondary}"
      },
      search: {
        "default": "{color.interactive.icon.default.base}",
        hover: "{color.interactive.icon.hover.base}",
        active: "{color.interactive.icon.active.base}"
      },
      clear: {
        "default": "{color.interactive.icon.default.base}",
        hover: "{color.interactive.icon.hover.base}",
        active: "{color.interactive.icon.active.base}"
      },
      border: {
        "default": "{color.interactive.border.default}",
        hover: "{color.interactive.border.hover}",
        active: "{color.interactive.border.active}"
      }
    },
    split: {
      color: {
        divider: {
          "default": "{color.border.primary}"
        }
      }
    }
  },
  segmentedControl: {
    color: {
      fill: {
        base: "{color.fill.secondary}",
        onComplexBg: "{color.gray.550}"
      },
      border: {
        base: "{color.gray.40}",
        onComplexBg: "{color.gray.250}"
      }
    },
    segment: {
      color: {
        selected: {
          fill: {
            "default": {
              base: "{color.fill.primary}",
              onComplexBg: "{color.fill.primary}"
            },
            disabled: {
              base: "{color.gray.40}",
              onComplexBg: "{color.white.300}"
            }
          },
          text: {
            "default": {
              base: "{color.text.primary}",
              onComplexBg: "{color.text.primary}"
            },
            disabled: {
              base: "{color.text.disabled}",
              onComplexBg: "{color.gray.450}"
            }
          },
          icon: {
            "default": {
              base: "{color.icon.primary}",
              onComplexBg: "{color.icon.primary}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.gray.450}"
            }
          }
        },
        unselected: {
          fill: {
            "default": {
              base: "{color.transparent}",
              onComplexBg: "{color.transparent}"
            },
            hover: {
              base: "{color.surface.hover}",
              onComplexBg: "{color.gray.100}"
            },
            active: {
              base: "{color.surface.active}",
              onComplexBg: "{color.gray.250}"
            },
            disabled: {
              base: "{color.transparent}",
              onComplexBg: "{color.transparent}"
            }
          },
          text: {
            "default": {
              base: "{color.text.secondary}",
              onComplexBg: "{color.text.inverse}"
            },
            disabled: {
              base: "{color.text.disabled}",
              onComplexBg: "{color.white.300}"
            },
            hover: {
              onComplexBg: "{color.text.inverse}"
            },
            active: {
              onComplexBg: "{color.text.inverse}"
            }
          },
          icon: {
            "default": {
              base: "{color.icon.secondary}",
              onComplexBg: "{color.icon.inverse}"
            },
            disabled: {
              base: "{color.icon.disabled}",
              onComplexBg: "{color.white.300}"
            },
            hover: {
              onComplexBg: "{color.icon.inverse}"
            },
            active: {
              onComplexBg: "{color.icon.inverse}"
            }
          }
        }
      }
    }
  },
  slider: {
    progress: {
      color: {
        fill: {
          "default": "{color.feedback.fill.informative.loud.base}",
          disabled: "{color.gray.solid.250}"
        }
      }
    },
    handle: {
      front: {
        color: {
          fill: {
            "default": "{color.fill.primary}",
            disabled: "{color.fill.primary}"
          }
        }
      },
      back: {
        color: {
          fill: {
            "default": "{color.feedback.fill.informative.loud.base}",
            active: "{color.cyan.300}",
            disabled: "{color.gray.solid.250}"
          }
        }
      }
    },
    color: {
      fill: {
        "default": "{color.fill.tertiary}",
        disabled: "{color.fill.disabled}"
      },
      label: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      },
      icon: {
        "default": "{color.icon.primary}",
        disabled: "{color.icon.disabled}"
      },
      text: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    step: {
      color: {
        fill: {
          "default": "{color.gray.250}"
        }
      }
    },
    reference: {
      color: {
        fill: {
          active: "{color.fill.inverse}"
        },
        text: {
          "default": "{color.text.primary}",
          active: "{color.text.inverse}"
        }
      }
    }
  },
  snackbar: {
    color: {
      fill: {
        base: "{color.fill.inverse}",
        positive: "{color.feedback.fill.positive.loud.base}",
        negative: "{color.feedback.fill.negative.loud.base}"
      },
      text: {
        base: "{color.text.inverse}",
        positive: "{color.text.inverse}",
        negative: "{color.text.inverse}"
      }
    },
    action: {
      color: {
        text: {
          "default": "{color.text.link.default.inverse}",
          positive: {
            "default": "{color.text.link.default.inverse}"
          },
          negative: {
            "default": "{color.text.link.default.inverse}"
          }
        }
      }
    }
  },
  "switch": {
    color: {
      selected: {
        fill: {
          "default": "{color.selected.fill.default}",
          disabled: "{color.cyan.300}"
        }
      },
      unselected: {
        fill: {
          "default": "{color.fill.tertiary}",
          disabled: "{color.fill.disabled}"
        }
      },
      text: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      }
    },
    handle: {
      color: {
        fill: {
          "default": "{color.fill.primary}"
        }
      }
    }
  },
  tabs: {
    color: {
      fill: "{color.transparent}"
    },
    tab: {
      color: {
        selected: {
          fill: {
            "default": "{color.interactive.fill.mute.default}",
            hover: "{color.interactive.fill.mute.hover}",
            active: "{color.interactive.fill.mute.active}"
          },
          pipe: {
            "default": "{color.selected.fill.default}",
            hover: "{color.selected.fill.hover}",
            active: "{color.selected.fill.active}"
          },
          text: {
            "default": "{color.selected.text.default}",
            hover: "{color.selected.text.hover}",
            active: "{color.selected.text.active}"
          }
        },
        unselected: {
          fill: {
            "default": "{color.interactive.fill.mute.default}",
            hover: "{color.interactive.fill.mute.hover}",
            active: "{color.interactive.fill.mute.active}"
          },
          text: {
            "default": "{color.text.primary}",
            hover: "{color.selected.text.hover}",
            active: "{color.selected.text.active}"
          }
        },
        border: "{color.border.primary}"
      }
    },
    controls: {
      color: {
        chevron: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}",
          disabled: "{color.icon.disabled}"
        }
      }
    }
  },
  tag: {
    color: {
      fill: {
        "default": "{color.transparent}",
        hover: "{color.fill.secondary}",
        active: "{color.fill.tertiary}",
        disabled: "{color.transparent}",
        positive: {
          "default": "{color.feedback.fill.positive.quiet}",
          hover: "{color.green.150}",
          active: "{color.green.200}"
        },
        caution: {
          "default": "{color.feedback.fill.caution.quiet}",
          hover: "{color.orange.150}",
          active: "{color.orange.200}"
        },
        negative: {
          "default": "{color.feedback.fill.negative.quiet}",
          hover: "{color.red.150}",
          active: "{color.red.200}"
        }
      },
      text: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}",
        positive: {
          "default": "{color.feedback.text.positive.loud}",
          hover: "{color.green.600}",
          active: "{color.green.700}"
        },
        caution: {
          "default": "{color.feedback.text.caution.loud}",
          hover: "{color.orange.600}",
          active: "{color.orange.700}"
        },
        negative: {
          "default": "{color.feedback.text.negative.loud}",
          hover: "{color.red.600}",
          active: "{color.red.700}"
        }
      },
      icon: {
        "default": "{color.icon.primary}",
        disabled: "{color.icon.disabled}",
        positive: {
          "default": "{color.feedback.icon.positive.loud}",
          hover: "{color.green.600}",
          active: "{color.green.700}"
        },
        caution: {
          "default": "{color.feedback.icon.caution.loud}",
          hover: "{color.orange.600}",
          active: "{color.orange.700}"
        },
        negative: {
          "default": "{color.feedback.icon.negative.loud}",
          hover: "{color.red.600}",
          active: "{color.red.700}"
        }
      },
      close: {
        "default": "{color.interactive.icon.default.base}",
        hover: "{color.interactive.icon.hover.base}",
        active: "{color.interactive.icon.active.base}",
        disabled: "{color.icon.disabled}",
        positive: {
          "default": "{color.feedback.icon.positive.loud}",
          hover: "{color.green.600}",
          active: "{color.green.700}"
        },
        caution: {
          "default": "{color.feedback.icon.caution.loud}",
          hover: "{color.orange.600}",
          active: "{color.orange.700}"
        },
        negative: {
          "default": "{color.feedback.icon.negative.loud}",
          hover: "{color.red.600}",
          active: "{color.red.700}"
        }
      },
      border: {
        "default": "{color.gray.250}",
        disabled: "{color.gray.100}",
        positive: {
          "default": "{color.feedback.border.positive.loud.base}",
          hover: "{color.green.600}",
          active: "{color.green.700}"
        },
        caution: {
          "default": "{color.feedback.border.caution.loud.base}",
          hover: "{color.orange.600}",
          active: "{color.orange.700}"
        },
        negative: {
          "default": "{color.feedback.border.negative.loud.base}",
          hover: "{color.red.600}",
          active: "{color.red.700}"
        }
      }
    },
    choice: {
      color: {
        selected: {
          fill: {
            "default": "{color.cyan.100}",
            disabled: "{color.transparent}"
          },
          text: {
            "default": "{color.selected.text.default}",
            hover: "{color.selected.text.hover}",
            active: "{color.selected.text.active}",
            disabled: "{color.text.disabled}"
          },
          check: {
            "default": "{color.selected.icon.default}",
            hover: "{color.selected.icon.hover}",
            active: "{color.selected.icon.active}"
          },
          icon: {
            "default": "{color.selected.icon.default}",
            hover: "{color.selected.icon.hover}",
            active: "{color.selected.icon.active}",
            disabled: "{color.icon.disabled}"
          },
          border: {
            "default": "{color.selected.border.default}",
            hover: "{color.selected.border.hover}",
            active: "{color.selected.border.active}",
            disabled: "{color.gray.100}"
          }
        },
        unselected: {
          fill: {
            "default": "{color.transparent}"
          },
          text: {
            "default": "{color.text.primary}",
            disabled: "{color.text.disabled}"
          },
          icon: {
            "default": "{color.icon.primary}",
            disabled: "{color.icon.disabled}"
          },
          border: {
            "default": "{color.gray.250}",
            hover: "{color.gray.550}",
            active: "{color.cyan.500}",
            disabled: "{color.gray.100}"
          }
        }
      }
    }
  },
  technicalError: {
    section: {
      color: {
        title: "{color.text.primary}",
        description: "{color.text.primary}",
        instruction: "{color.text.secondary}"
      },
      asset: {
        color: {
          fill: "{color.fill.primary}",
          icon: "{color.icon.secondary}",
          border: "{color.gray.550}"
        }
      },
      code: {
        color: {
          fill: {
            "default": "{color.transparent}",
            hover: "{color.fill.secondary}",
            active: "{color.fill.tertiary}"
          },
          border: {
            "default": "{color.gray.250}"
          },
          text: {
            "default": "{color.text.secondary}"
          },
          icon: {
            "default": "{color.interactive.icon.default.accent}",
            hover: "{color.interactive.icon.hover.accent}",
            active: "{color.interactive.icon.active.accent}"
          }
        }
      }
    },
    screen: {
      statusBar: {
        color: {
          fill: "{color.fill.brand}",
          icon: "{color.icon.primary}"
        }
      },
      color: {
        background: "{color.background.secondary}"
      }
    },
    snackbar: {
      color: {
        instruction: "{color.text.inverse}"
      },
      code: {
        color: {
          fill: {
            "default": "{color.transparent}",
            active: "{color.fill.primary}"
          },
          border: {
            "default": "{color.border.inverse}",
            active: "{color.border.inverse}"
          },
          text: {
            "default": "{color.text.link.default.inverse}",
            active: "{color.text.primary}"
          },
          icon: {
            "default": "{color.interactive.icon.default.inverse}",
            active: "{color.feedback.icon.positive.loud}"
          }
        }
      }
    }
  },
  textfield: {
    color: {
      fill: {
        "default": "{color.fill.primary}",
        disabled: "{color.fill.secondary}"
      },
      label: {
        "default": "{color.text.primary}",
        error: "{color.feedback.text.negative.loud}",
        disabled: "{color.text.disabled}",
        readOnly: "{color.text.secondary}"
      },
      placeholder: {
        "default": "{color.text.secondary}"
      },
      inputText: {
        "default": "{color.text.primary}",
        disabled: "{color.text.disabled}"
      },
      prefix: {
        "default": "{color.text.secondary}",
        disabled: "{color.text.disabled}"
      },
      suffix: {
        "default": "{color.text.secondary}",
        disabled: "{color.text.disabled}"
      },
      text: {
        checkbox: {
          "default": "{color.text.primary}",
          disabled: "{color.text.disabled}"
        }
      },
      icon: {
        "default": "{color.icon.primary}",
        disabled: "{color.icon.disabled}"
      },
      tooltip: {
        "default": "{color.interactive.icon.default.accent}",
        hover: "{color.interactive.icon.hover.accent}",
        active: "{color.interactive.icon.active.accent}"
      },
      validated: "{color.feedback.icon.positive.loud}",
      clear: {
        "default": "{color.interactive.icon.default.base}",
        hover: "{color.interactive.icon.hover.base}",
        active: "{color.interactive.icon.active.base}"
      },
      border: {
        "default": "{color.interactive.border.default}",
        hover: "{color.interactive.border.hover}",
        active: "{color.interactive.border.active}",
        error: "{color.feedback.border.negative.loud.base}",
        disabled: "{color.border.disabled}"
      }
    },
    split: {
      color: {
        divider: {
          "default": "{color.border.primary}",
          disabled: "{color.border.disabled}"
        }
      }
    },
    inline: {
      color: {
        fill: {
          "default": "{color.transparent}"
        },
        inputText: {
          "default": "{color.text.primary}",
          caution: "{color.feedback.text.caution.loud}",
          error: "{color.feedback.text.negative.loud}"
        },
        edit: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}"
        },
        border: {
          "default": "{color.transparent}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          caution: "{color.feedback.border.caution.loud.base}",
          error: "{color.feedback.border.negative.loud.base}"
        }
      }
    },
    password: {
      color: {
        icon: {
          "default": "{color.interactive.icon.default.base}",
          hover: "{color.interactive.icon.hover.base}",
          active: "{color.interactive.icon.active.base}"
        }
      }
    }
  },
  timePicker: {
    intervals: {
      color: {
        fill: {
          disabled: "{color.gray.40}"
        },
        label: {
          "default": "{color.text.primary}",
          error: "{color.feedback.text.negative.loud}",
          disabled: "{color.text.disabled}",
          readOnly: "{color.text.secondary}"
        },
        placeholder: "{color.text.secondary}",
        border: {
          "default": "{color.interactive.border.default}",
          hover: "{color.interactive.border.hover}",
          active: "{color.interactive.border.active}",
          error: "{color.feedback.border.negative.loud.base}",
          disabled: "{color.border.disabled}",
          readOnly: "{color.transparent}"
        }
      },
      selector: {
        color: {
          fill: {
            disabled: "{color.gray.40}"
          },
          chevron: {
            "default": "{color.interactive.icon.default.accent}",
            hover: "{color.interactive.icon.hover.accent}",
            active: "{color.interactive.icon.active.accent}",
            disabled: "{color.icon.disabled}"
          },
          border: {
            "default": "{color.interactive.border.default}",
            hover: "{color.interactive.border.hover}",
            active: "{color.interactive.border.active}",
            error: "{color.feedback.border.negative.loud.base}",
            disabled: "{color.border.disabled}",
            readOnly: "{color.transparent}"
          },
          value: {
            "default": "{color.text.primary}",
            disabled: "{color.text.disabled}"
          }
        }
      }
    }
  },
  thumbnail: {
    image: {
      color: {
        fill: {
          "default": "{color.fill.primary}"
        }
      }
    },
    loud: {
      color: {
        fill: {
          "default": "{color.cyan.500}",
          disabled: "{color.fill.disabled}"
        },
        text: {
          "default": "{color.text.inverse}",
          disabled: "{color.text.disabled}"
        },
        icon: {
          "default": "{color.icon.inverse}",
          disabled: "{color.icon.disabled}"
        }
      }
    },
    quiet: {
      color: {
        fill: {
          "default": "{color.cyan.100}",
          disabled: "{color.fill.disabled}"
        },
        text: {
          "default": "{color.text.accent}",
          disabled: "{color.text.disabled}"
        },
        icon: {
          "default": "{color.icon.accent}",
          disabled: "{color.icon.disabled}"
        }
      }
    },
    mute: {
      color: {
        fill: {
          "default": "{color.fill.primary}",
          disabled: "{color.fill.primary}"
        },
        text: {
          "default": "{color.text.primary}",
          disabled: "{color.text.disabled}"
        },
        icon: {
          "default": "{color.icon.primary}",
          disabled: "{color.icon.disabled}"
        }
      }
    },
    color: {
      border: {
        "default": "{color.border.primary}"
      }
    },
    badge: {
      color: {
        text: {
          informative: "{color.feedback.text.informative.loud}",
          positive: "{color.feedback.text.positive.loud}",
          caution: "{color.feedback.text.caution.loud}",
          negative: "{color.feedback.text.negative.loud}"
        },
        icon: {
          informative: "{color.feedback.icon.informative.loud}",
          positive: "{color.feedback.icon.positive.loud}",
          caution: "{color.feedback.icon.caution.loud}",
          negative: "{color.feedback.icon.negative.loud}"
        },
        border: {
          informative: {
            base: "{color.feedback.border.informative.loud.base}",
            onComplexBg: "{color.feedback.border.informative.loud.onFill}"
          },
          positive: {
            base: "{color.feedback.border.positive.loud.base}",
            onComplexBg: "{color.feedback.border.positive.loud.onFill}"
          },
          caution: {
            base: "{color.feedback.border.caution.loud.base}",
            onComplexBg: "{color.feedback.border.caution.loud.onFill}"
          },
          negative: {
            base: "{color.feedback.border.negative.loud.base}",
            onComplexBg: "{color.feedback.border.negative.loud.onFill}"
          }
        }
      }
    },
    grid: {
      color: {
        fill: "{color.fill.secondary}"
      },
      overflow: {
        color: {
          fill: "{color.gray.solid.70}",
          text: "{color.text.secondary}",
          border: "{color.border.primary}"
        }
      }
    },
    stacked: {
      overflow: {
        color: {
          fill: "{color.gray.solid.70}",
          text: "{color.text.secondary}",
          border: "{color.border.primary}"
        }
      }
    }
  },
  tooltip: {
    color: {
      fill: {
        base: "{color.fill.primary}",
        inverse: "{color.fill.inverse}",
        accent: "{color.feedback.fill.informative.loud.base}"
      },
      title: {
        base: "{color.text.primary}",
        inverse: "{color.text.inverse}",
        accent: "{color.text.inverse}"
      },
      text: {
        base: "{color.text.primary}",
        inverse: "{color.text.inverse}",
        accent: "{color.text.inverse}"
      },
      close: {
        "default": {
          base: "{color.interactive.icon.default.base}",
          inverse: "{color.interactive.icon.default.inverse}",
          accent: "{color.interactive.icon.default.inverse}"
        },
        hover: "{color.interactive.icon.hover.base}",
        active: "{color.interactive.icon.active.base}"
      }
    }
  }
};
var mpTheme = {
  core: core,
  colorSchemes: colorSchemes,
  components: components
};

/** @typedef {import('../../').AndesProviderTheme} Theme */
/** @typedef {import('../../').AndesBaseTheme} AndesBaseTheme */

/**
 * @param {Theme} childTheme
 * @param {Theme} parentTheme
 */
const mergeChildAndParentThemes = (childTheme, parentTheme) => {
  const newTheme = /** @type {Theme} */{};
  Object.entries(childTheme).forEach(_ref => {
    let _ref2 = _slicedToArray(_ref, 2),
      key = _ref2[0],
      value = _ref2[1];
    if (parentTheme[key] !== undefined) {
      newTheme[key] = mergeDeep(/** @type {Record<string, any>}* */parentTheme[key], /** @type {Record<string, any>}* */value, {
        allowOnlySourceKeys: true
      });
    }
  });
  return newTheme;
};

/**
 * @param {Theme} theme
 * @param {AndesBaseTheme} defaultThemeObject
 * @param {boolean} [override]
 *
 * @returns {Theme}
 */
const mergeThemes = (theme, defaultThemeObject, override) => {
  const _useContext = useContext(ThemeContext),
    parentTheme = _useContext.theme;
  if (override || !parentTheme) {
    return /** @type {Theme} */mergeDeep(defaultThemeObject, theme);
  }
  return mergeChildAndParentThemes(theme, parentTheme);
};

/** @type {typeof import('../..').useCreateTheme} */
const useCreateTheme = (theme, defaultTheme, id, override) => {
  /** @type {AndesBaseTheme} */
  const defaultThemeObject = defaultTheme === 'MP' ? mpTheme : mlTheme;
  const mergedTheme = mergeThemes(theme, defaultThemeObject, override);
  return _objectSpread({
    theme: mergedTheme
  }, createTheme(mergedTheme, id));
};

/** @returns {[
 * ReadonlyMap<string, import("../..").RegisterComponentsClassesObject>,
 * (component: string, componentClass: string, variant: Array<import("../..").CustomVariant>) => void]} */
const useComponentRegister = () => {
  const themedComponents = /** @type {Map<string, import("../..").RegisterComponentsClassesObject>} */
  new Map();

  /**
   * @param {string} component
   * @param {string} componentClass
   */
  const registerComponentClassName = (component, componentClass) => {
    if (!themedComponents.has(component)) {
      themedComponents.set(component, {
        className: componentClass,
        variants: []
      });
    }
  };

  /**
   * @param {string} component
   * @param {Array<import("../..").CustomVariant>} variants
   */
  const registerComponentVariant = (component, variants) => {
    if (themedComponents.has(component)) {
      const oldValue = themedComponents.get(component);
      if (oldValue) {
        const existingVariantIds = new Set(oldValue.variants.map(variant => variant.id));
        const filteredVariants = variants.filter(variant => !existingVariantIds.has(variant.id));
        const updatedVariants = [...oldValue.variants, ...filteredVariants];
        const newValue = {
          className: oldValue.className,
          variants: updatedVariants
        };
        themedComponents.set(component, newValue);
      }
    }
  };

  /**
   * @param {string} component
   * @param {string} componentClass
   * @param {Array<import("../..").CustomVariant>} variants
   */
  const registerComponent = (component, componentClass, variants) => {
    registerComponentClassName(component, componentClass);
    registerComponentVariant(component, variants);
  };
  return [themedComponents, registerComponent];
};

/* eslint-disable react/prop-types */

/** @type {React.FunctionComponent} */
const ThemeStyles = () => {
  const _useThemeContext = useThemeContext(),
    getThemeStyles = _useThemeContext.getThemeStyles;
  if (typeof getThemeStyles !== 'function') {
    return null;
  }
  return (/*#__PURE__*/
    // eslint-disable-next-line nordic/no-style-tags
    React.createElement("style", null, getThemeStyles())
  );
};

/** @type {typeof import('../..').AndesThemeProvider} */
const ThemeProvider = _ref => {
  let _ref$baseTheme = _ref.baseTheme,
    baseTheme = _ref$baseTheme === void 0 ? 'ML' : _ref$baseTheme,
    children = _ref.children,
    id = _ref.id,
    _ref$override = _ref.override,
    override = _ref$override === void 0 ? false : _ref$override,
    stylesOptions = _ref.stylesOptions,
    _ref$theme = _ref.theme,
    theme = _ref$theme === void 0 ? {} : _ref$theme;
  const _useContext = useContext(ThemeContext),
    parentGetThemedComponentClasses = _useContext.getThemedComponentClasses;
  const customId = useId(id).replace(/:/g, '');
  const _useCreateTheme = useCreateTheme(theme, baseTheme, customId, override),
    providerTheme = _useCreateTheme.theme,
    getCreateThemeStyles = _useCreateTheme.getStyles,
    componentsClasses = _useCreateTheme.componentsClasses,
    getComponentVariants = _useCreateTheme.getComponentVariants,
    themeHash = _useCreateTheme.themeHash,
    themeVariables = _useCreateTheme.themeVariables;
  const _useComponentRegister = useComponentRegister(),
    _useComponentRegister2 = _slicedToArray(_useComponentRegister, 2),
    themedComponents = _useComponentRegister2[0],
    registerComponent = _useComponentRegister2[1];
  const _ref2 = stylesOptions || {},
    externalInject = _ref2.externalInject;

  /**
   * @param {string} component
   * @param {object} props
   */
  const getThemedComponentClasses = function (component) {
    var _componentsClasses$ge;
    let props = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    const parentThemeClass = typeof parentGetThemedComponentClasses === 'function' ? parentGetThemedComponentClasses(component, props) : '';
    const componentClass = ((_componentsClasses$ge = componentsClasses.get(component)) === null || _componentsClasses$ge === void 0 ? void 0 : _componentsClasses$ge.className) || '';
    let componentClassNames = componentClass;
    if (parentThemeClass) {
      componentClassNames = componentClass ? `${parentThemeClass} ${componentClass}` : parentThemeClass;
    }
    if (!componentClassNames) {
      return '';
    }
    const componentVariants = getComponentVariants(component, props);
    registerComponent(component, componentClassNames, componentVariants);
    const themedComponent = themedComponents.get(component);
    if (!themedComponent) {
      return '';
    }
    const componentBaseClassName = themedComponent === null || themedComponent === void 0 ? void 0 : themedComponent.className;
    const componentVariantsClassName = componentVariants.reduce((classNames, variant) => {
      classNames.push(`${componentVariantPrefix}${variant.id}`);
      return classNames;
    }, /** @type {string[] } */[]).join(' ');
    const className = `${componentBaseClassName} ${componentVariantsClassName}`.trim();
    return className;
  };
  const getThemeStyles = () => {
    const _ref3 = stylesOptions || {},
      _ref3$buildOn = _ref3.buildOn,
      buildOn = _ref3$buildOn === void 0 ? 'SSR' : _ref3$buildOn,
      global = _ref3.global;
    const compliesRenderStylesCondition = buildOn === 'SSR' ? true // Always render styles on server side to avoid missmatch with client side
    : typeof window !== 'undefined';
    return compliesRenderStylesCondition ? getCreateThemeStyles(themedComponents, global) : '';
  };
  const contextValue = useMemo(() => ({
    theme: providerTheme,
    getThemedComponentClasses,
    getThemeStyles,
    themeHash,
    themeVariables
  }),
  // eslint-disable-next-line react-hooks/exhaustive-deps
  [theme]);
  registerMetric(theme);
  return /*#__PURE__*/React.createElement(ThemeContext.Provider, {
    value: contextValue
  }, /*#__PURE__*/React.createElement("div", {
    id: customId,
    style: {
      display: 'contents'
    }
  }, children, !externalInject ? /*#__PURE__*/React.createElement(ThemeStyles, null) : null));
};
ThemeProvider.displayName = 'AndesThemeProvider';
ThemeProvider.propTypes = {
  children: PropTypes.node.isRequired,
  baseTheme: PropTypes.oneOf(['ML', 'MP']),
  id: PropTypes.string,
  stylesOptions: PropTypes.shape({}),
  theme: PropTypes.shape({}),
  override: PropTypes.bool
};
export { ThemeProvider as AndesThemeProvider, useCreateTheme, useThemeContext };
