// @andes/list/build/esm/module.js
import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { useContext, createContext, useMemo, Children, cloneElement, isValidElement, useRef, useState, useEffect } from 'react';
import classNames from 'classnames';
import { useId } from '@andes/context';
import { useThemeContext } from '@andes/themes';
import { mergeProps } from '@react-aria/utils';
import warning from '@andes/common/build/utils/warning';
import includes from '@andes/common/build/utils/includes';
import { computeDataAndesState } from '@andes/common/build/utils/computeDataAndesState';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { Image } from '@andes/image';
import ChevronRight20 from '@andes/icons/ChevronRight20';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import { withTracker } from '@andes/with-tracker';
import { senderMetrics, trackCustomization } from '@andes/common/build/metrics';
const namespace = 'andes-list';
const listGroupNamespace = 'andes-list__group';
const listItemNamespace = 'andes-list__item';
const LIST_ACTION_ITEM_DISPLAY_NAME = 'ListActionItem';
const LIST_BASE_ITEM_DISPLAY_NAME = 'ListBaseItem';
const LIST_DISPLAY_NAME = 'List';
const LIST_GROUP_DISPLAY_NAME = 'ListGroup';
const LIST_ITEM_CONTENT_DISPLAY_NAME = 'ListItemContent';
const LIST_ITEM_DISPLAY_NAME = 'ListItem';
const ListContext = /*#__PURE__*/createContext({});
function useListContext() {
  return useContext(ListContext);
}
function ListProvider(_ref) {
  let children = _ref.children,
    listType = _ref.listType;
  const value = useMemo(() => ({
    listType
  }), [listType]);
  return /*#__PURE__*/React.createElement(ListContext.Provider, {
    value: value
  }, children);
}
const getAttributesByListType = function () {
  let listType = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'standalone';
  return function () {
    let _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      size = _ref.size;
    let listName = 'list';
    switch (listType) {
      case 'standalone':
        listName = 'list';
        break;
      case 'listbox':
        listName = 'listbox';
        break;
      case 'checkbox':
        listName = 'checkbox-list';
        break;
      case 'radio':
        listName = 'radio-list';
        break;
      default:
        listName = 'list';
        break;
    }
    return {
      list: {
        [`data-andes-${listName}`]: true
      },
      group: {
        'data-andes-list-group': true
      },
      header: {
        'data-andes-list-group-header': true
      },
      sublist: {
        'data-andes-list-group-sublist': true
      },
      item: {
        [`data-andes-${listName}-item`]: true,
        [`data-andes-${listName}-item-size`]: size
      },
      chevron: {
        'data-andes-list-chevron': true
      },
      selection: {
        [`data-andes-${listName}-${listType === 'listbox' ? 'checkbox' : listType}`]: true
      },
      title: {
        [`data-andes-${listName}-title`]: true
      },
      description: {
        [`data-andes-${listName}-description`]: true
      },
      leftContent: {
        [`data-andes-${listName}-left-content`]: true
      },
      rightContent: {
        [`data-andes-${listName}-right-content`]: true
      }
    };
  };
};
const _excluded$4 = ["ref"],
  _excluded2$1 = ["children", "className", "component", "id", "listType", "selectable", "srLabel", "type", "withDividers"];
const List = /*#__PURE__*/function () {
  const List = _ref => {
    let ref = _ref.ref,
      props = _objectWithoutProperties(_ref, _excluded$4);
    const children = props.children,
      className = props.className,
      _props$component = props.component,
      Component = _props$component === void 0 ? 'ul' : _props$component,
      id = props.id,
      _props$listType = props.listType,
      listType = _props$listType === void 0 ? 'standalone' : _props$listType,
      _props$selectable = props.selectable,
      selectable = _props$selectable === void 0 ? true : _props$selectable,
      srLabel = props.srLabel,
      _props$type = props.type,
      type = _props$type === void 0 ? 'default' : _props$type,
      _props$withDividers = props.withDividers,
      withDividers = _props$withDividers === void 0 ? false : _props$withDividers,
      other = _objectWithoutProperties(props, _excluded2$1);
    const _useThemeContext = useThemeContext(),
      getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
      themeHash = _useThemeContext.themeHash;
    const componentId = useId(id);
    const _getAttributesByListT = getAttributesByListType(listType)(),
      list = _getAttributesByListT.list;
    const listClassNames = classNames(namespace, className, `${namespace}--${type}`, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('list'), {
      [`${namespace}--selectable`]: selectable,
      [`${namespace}-with-dividers`]: withDividers
    });
    return /*#__PURE__*/React.createElement(ListProvider, {
      listType: listType
    }, /*#__PURE__*/React.createElement(Component, _extends({
      className: listClassNames,
      id: componentId,
      ref: ref
    }, other, themeHash, list, {
      "aria-label": srLabel
    }), React.Children.map(children, (child, index) => {
      let childNewId;
      if (props.role !== 'listbox') {
        var _child$props;
        const childId = child === null || child === void 0 || (_child$props = child.props) === null || _child$props === void 0 ? void 0 : _child$props.id;
        if (childId) {
          childNewId = childId;
        } else if (componentId) {
          childNewId = `${componentId}-${index}`;
        }
      }
      return child && /*#__PURE__*/React.cloneElement(child, {
        id: childNewId
      });
    })));
  };
  List.displayName = LIST_DISPLAY_NAME;
  return List;
}();
var List$1 = List;
const hasChildNodes = _ref => {
  let label = _ref.label,
    children = _ref.children;
  return label && Children.count(children) > 0;
};
const groupNodeGenerator = function* getCollectionNode(props) {
  const children = props.children,
    label = props.label,
    ariaLabel = props['aria-label'];
  const rendered = label || children;
  const textValue = (typeof rendered === 'string' ? rendered : '') || ariaLabel || '';
  warning(textValue, '<ListGroup> needs some plain text contents for accessibility reasons. Please add a `label` or `aria-label` prop.');
  yield {
    type: 'section',
    props,
    rendered,
    textValue,
    'aria-label': ariaLabel || label,
    hasChildNodes,
    *childNodes() {
      if (label) {
        const items = [];
        Children.forEach(children, child => {
          items.push({
            type: 'item',
            element: child
          });
        });
        yield* items;
      }
    }
  };
};
var groupNodeGenerator$1 = groupNodeGenerator;
const SMALL = 'small';
const MEDIUM = 'medium';
const PADDING_SIZE_0 = 0;
const PADDING_SIZE_16 = 16;
const PADDING_SIZE_24 = 24;
const PADDING_SIZE_32 = 32;
const PADDING_SIZE_40 = 40;
const DEPRECATED_PADDING_SIZE_0 = '0';
const DEPRECATED_PADDING_SIZE_16 = '16';
const DEPRECATED_PADDING_SIZE_24 = '24';
const DEPRECATED_PADDING_SIZE_32 = '32';
const DEPRECATED_PADDING_SIZE_40 = '40';
const DEFAULT_POSITION_TOP = 'top';
const primarySizes = {
  small: 18,
  medium: 20,
  large: 22
};
const WARNING_CHEVRON = `'chevron' prop will be deprecated in an upcoming major version.
Please use the '<ListActionItem>' component instead.`;
const WARNING_DISABLED = `'disabled' prop will be deprecated in an upcoming major version.
Please use the '<ListActionItem>' component instead.`;
const WARNING_HREF = `'href' prop will be deprecated in an upcoming major version.
Please use the '<ListActionItem>' component instead.`;
const WARNING_TARGET = `'target' prop will be deprecated in an upcoming major version.
Please use the '<ListActionItem>' component instead.`;
const WARNING_ON_CLICK = `'onClick' prop will be deprecated in an upcoming major version.
Please use the '<ListActionItem>' component instead.`;
const WARNING_BUTTON_SWITCH_ITEM_SMALL = `A <ListItem> with size="small" does not allow Buttons or Switches as
rightContent. Please, use a bigger <ListItem> instead`;
const WARNING_BUTTON_SIZE = '<ListItem> only supports Buttons with size="medium"';
const WARNING_BADGE_SIZE = '<ListItem> only supports Badges with size="small"';
const WARNING_LIST_BASE_ITEM = '\'component\' prop will be deprecated in a future major version. Please remove this prop to ensure future compatibility.';
const WARNING_LIST_ITEM = `<ListItem> needs to be wrapped in a '<List>' for accessibility reasons. 
Add the '<List>' component.`;
const WARNING_LIST_ITEM_CHILDREN = 'Children can\'t be used in conjunction with props title, description, leftContent and rightContent, in future versions children will have precedence over these properties';
const WARNING_LIST_ITEM_LEFT_CONTENT = 'The current object format \'{ title: string, src: string, alt: string, src2x: string }\' will be deprecated from the \'leftContent\' options. Please use a \'node\' instead.';
const WARNING_LIST_GROUP = `<ListGroup> needs to be wrapped in a '<List>' for accessibility reasons. 
Add the '<List>' component.`;
const WARNING_LIST_ACTION_ITEM_ACTION = '<ListActionItem> needs an action. Please add a `href` or `onClick` prop.';
const WARNING_PADDING_SIZE_0 = `
  '${DEPRECATED_PADDING_SIZE_0}' option will be removed from 'paddingSize' options in a future major version.
  Please, use ${PADDING_SIZE_0} as number instead.
`;
const WARNING_PADDING_SIZE_16 = `
  '${DEPRECATED_PADDING_SIZE_16}' option will be removed from 'paddingSize' options in a future major version.
  Please, use ${PADDING_SIZE_16} as number instead.
`;
const WARNING_PADDING_SIZE_24 = `
  '${DEPRECATED_PADDING_SIZE_24}' option will be removed from 'paddingSize' options in a future major version.
  Please, use ${PADDING_SIZE_24} as number instead.
`;
const WARNING_PADDING_SIZE_32 = `
  '${DEPRECATED_PADDING_SIZE_32}' option will be removed from 'paddingSize' options in a future major version.
  Please, use ${PADDING_SIZE_32} as number instead.
`;
const WARNING_PADDING_SIZE_40 = `
  '${DEPRECATED_PADDING_SIZE_40}' option will be removed from 'paddingSize' options in a future major version.
  Please, use ${PADDING_SIZE_40} as number instead.
`;
const DEPRECATED_PADDING_SIZES_WARNINGS = {
  [`${DEPRECATED_PADDING_SIZE_0}`]: WARNING_PADDING_SIZE_0,
  [`${DEPRECATED_PADDING_SIZE_16}`]: WARNING_PADDING_SIZE_16,
  [`${DEPRECATED_PADDING_SIZE_24}`]: WARNING_PADDING_SIZE_24,
  [`${DEPRECATED_PADDING_SIZE_32}`]: WARNING_PADDING_SIZE_32,
  [`${DEPRECATED_PADDING_SIZE_40}`]: WARNING_PADDING_SIZE_40
};
const _excluded$3 = ["children", "className", "component", "groupProps", "headingProps", "id", "itemProps", "label", "paddingSize", "size", "withDividers"];
const ListGroup = /*#__PURE__*/function () {
  const ListGroup = props => {
    const children = props.children,
      className = props.className,
      _props$component = props.component,
      Component = _props$component === void 0 ? 'ul' : _props$component,
      groupProps = props.groupProps,
      headingProps = props.headingProps,
      id = props.id,
      itemProps = props.itemProps,
      label = props.label,
      paddingSize = props.paddingSize,
      _props$size = props.size,
      size = _props$size === void 0 ? MEDIUM : _props$size,
      _props$withDividers = props.withDividers,
      withDividers = _props$withDividers === void 0 ? false : _props$withDividers,
      other = _objectWithoutProperties(props, _excluded$3);
    const listContext = useListContext();
    if (typeof paddingSize === 'string') {
      warning(!includes(Object.keys(DEPRECATED_PADDING_SIZES_WARNINGS), paddingSize), DEPRECATED_PADDING_SIZES_WARNINGS[paddingSize]);
    }
    const listGroupClassNames = classNames(listGroupNamespace, className, `${listGroupNamespace}--size-${size}`, {
      [`${listGroupNamespace}--padding-${paddingSize}`]: paddingSize,
      [`${listGroupNamespace}--with-dividers`]: withDividers
    });
    const _useMemo = useMemo(() => getAttributesByListType()(), []),
      group = _useMemo.group,
      header = _useMemo.header,
      sublist = _useMemo.sublist;
    warning(!!(listContext !== null && listContext !== void 0 && listContext.listType), WARNING_LIST_GROUP);
    return /*#__PURE__*/React.createElement("li", _extends({
      className: listGroupClassNames
    }, mergeProps(itemProps, other), group), /*#__PURE__*/React.createElement("span", _extends({
      className: `${listGroupNamespace}--header`,
      id: id
    }, headingProps, header), label), /*#__PURE__*/React.createElement(Component, _extends({
      className: `${listGroupNamespace}--sublist`,
      "aria-labelledby": id
    }, groupProps, sublist), Children.map(children, (child, index) => {
      var _child$props;
      let childNewId;
      const childId = (_child$props = child.props) === null || _child$props === void 0 ? void 0 : _child$props.id;
      if (childId) {
        childNewId = childId;
      } else if (id) {
        childNewId = `${id}-${index}`;
      }
      return /*#__PURE__*/cloneElement(child, {
        id: childNewId
      });
    })));
  };
  ListGroup.getCollectionNode = groupNodeGenerator$1;
  ListGroup.displayName = LIST_GROUP_DISPLAY_NAME;
  return ListGroup;
}();
var ListGroup$1 = ListGroup;
const getListBaseItemClassNames = props => {
  const className = props.className,
    _props$disabled = props.disabled,
    disabled = _props$disabled === void 0 ? false : _props$disabled,
    _props$hasDescription = props.hasDescription,
    hasDescription = _props$hasDescription === void 0 ? false : _props$hasDescription,
    _props$hasLeftContent = props.hasLeftContentImage,
    hasLeftContentImage = _props$hasLeftContent === void 0 ? false : _props$hasLeftContent,
    _props$interactive = props.interactive,
    interactive = _props$interactive === void 0 ? false : _props$interactive,
    paddingSize = props.paddingSize,
    _props$selected = props.selected,
    selected = _props$selected === void 0 ? false : _props$selected,
    _props$size = props.size,
    size = _props$size === void 0 ? MEDIUM : _props$size,
    slotted = props.slotted;
  const computedPaddingSize = Number(paddingSize);
  const listItemClassNames = classNames(listItemNamespace, className, `${listItemNamespace}--size-${size}`, {
    [`${listItemNamespace}--selected-padding-0`]: selected && computedPaddingSize === 0,
    [`${listItemNamespace}--padding-${computedPaddingSize}`]: computedPaddingSize || computedPaddingSize === 0,
    [`${listItemNamespace}--disabled`]: disabled,
    [`${listItemNamespace}--selected`]: selected,
    [`${listItemNamespace}-with-image`]: hasLeftContentImage,
    [`${listItemNamespace}-with-secondary`]: hasDescription && size !== 'small',
    [`${listItemNamespace}--selectable`]: interactive,
    [`${listItemNamespace}--slotted`]: slotted
  });
  return listItemClassNames;
};
var getListItemClassNames = getListBaseItemClassNames;
const _excluded$2 = ["children", "className", "component", "disabled", "forwardedRef", "hasDescription", "hasLeftContentImage", "interactive", "paddingSize", "selected", "size", "slotted", "srLabel", "disableSelection", "checked", "ref"];
const ListBaseItem = /*#__PURE__*/function () {
  const ListBaseItem = props => {
    const children = props.children,
      className = props.className,
      _props$component = props.component,
      Component = _props$component === void 0 ? 'li' : _props$component,
      disabled = props.disabled;
    props.forwardedRef;
    const hasDescription = props.hasDescription,
      hasLeftContentImage = props.hasLeftContentImage,
      interactive = props.interactive,
      paddingSize = props.paddingSize,
      selected = props.selected,
      _props$size = props.size,
      size = _props$size === void 0 ? MEDIUM : _props$size,
      _props$slotted = props.slotted,
      slotted = _props$slotted === void 0 ? false : _props$slotted;
    props.srLabel;
    const _props$disableSelecti = props.disableSelection,
      disableSelection = _props$disableSelecti === void 0 ? false : _props$disableSelecti,
      _props$checked = props.checked,
      checked = _props$checked === void 0 ? false : _props$checked,
      ref = props.ref,
      other = _objectWithoutProperties(props, _excluded$2);
    const _useListContext = useListContext(),
      _useListContext$listT = _useListContext.listType,
      listType = _useListContext$listT === void 0 ? 'standalone' : _useListContext$listT;
    warning(Component === 'li', WARNING_LIST_BASE_ITEM);
    if (typeof paddingSize === 'string') {
      warning(!includes(Object.keys(DEPRECATED_PADDING_SIZES_WARNINGS), paddingSize), DEPRECATED_PADDING_SIZES_WARNINGS[paddingSize]);
    }
    const listItemClassNames = getListItemClassNames({
      className,
      disabled,
      hasDescription,
      hasLeftContentImage,
      interactive,
      paddingSize,
      selected,
      size,
      slotted
    });
    const _useMemo = useMemo(() => getAttributesByListType(listType)({
        size
      }), [size]),
      item = _useMemo.item;
    return /*#__PURE__*/React.createElement(Component, _extends({
      className: listItemClassNames,
      ref: ref
    }, other, item, computeDataAndesState({
      checked: checked || false,
      disabled: disabled || disableSelection || false,
      selected
    })), children);
  };
  ListBaseItem.displayName = LIST_BASE_ITEM_DISPLAY_NAME;
  return ListBaseItem;
}();
var ListBaseItem$1 = ListBaseItem;
const getVerifiedRightContent = (rightContent, itemSize) => {
  if (/*#__PURE__*/isValidElement(rightContent)) {
    var _rightContent$type, _rightContent$type2, _rightContent$props2, _rightContent$type3;
    if (((_rightContent$type = rightContent.type) === null || _rightContent$type === void 0 ? void 0 : _rightContent$type.displayName) === 'Button') {
      var _rightContent$props;
      if (itemSize === 'small') {
        warning(false, WARNING_BUTTON_SWITCH_ITEM_SMALL);
        return undefined;
      }
      if (((_rightContent$props = rightContent.props) === null || _rightContent$props === void 0 ? void 0 : _rightContent$props.size) !== 'medium') {
        warning(false, WARNING_BUTTON_SIZE);
        return /*#__PURE__*/cloneElement(rightContent, {
          size: 'medium'
        });
      }
    }
    if (includes(['Pill', 'PillIcon'], (_rightContent$type2 = rightContent.type) === null || _rightContent$type2 === void 0 ? void 0 : _rightContent$type2.displayName) && (rightContent === null || rightContent === void 0 || (_rightContent$props2 = rightContent.props) === null || _rightContent$props2 === void 0 ? void 0 : _rightContent$props2.size) !== 'small') {
      warning(false, WARNING_BADGE_SIZE);
      return /*#__PURE__*/cloneElement(rightContent, {
        size: 'small'
      });
    }
    if (((_rightContent$type3 = rightContent.type) === null || _rightContent$type3 === void 0 ? void 0 : _rightContent$type3.displayName) === 'Switch' && itemSize === 'small') {
      warning(false, WARNING_BUTTON_SWITCH_ITEM_SMALL);
      return undefined;
    }
  }
  return rightContent;
};
var getVerifiedRightContent$1 = getVerifiedRightContent;
const hasSlot = (slot, _ref, _ref2) => {
  let title = _ref.title,
    description = _ref.description,
    rightContent = _ref.rightContent;
  let _ref2$legacyBehavior = _ref2.legacyBehavior,
    legacyBehavior = _ref2$legacyBehavior === void 0 ? false : _ref2$legacyBehavior;
  const hasRightContent = typeof rightContent === 'object' ? (rightContent === null || rightContent === void 0 ? void 0 : rightContent.content) || (rightContent === null || rightContent === void 0 ? void 0 : rightContent.description) : rightContent;
  const hasDefaultContent = title || description || hasRightContent;
  const isSlotted = legacyBehavior ? slot && !hasDefaultContent : slot;
  return !!isSlotted;
};
const isRightContentType = value => (value === null || value === void 0 ? void 0 : value.content) !== undefined;
const computeRightContent = _ref => {
  let rightContentProp = _ref.rightContent,
    rightContentPositionProp = _ref.rightContentPosition;
  if (!rightContentProp) return {
    position: DEFAULT_POSITION_TOP
  };
  const computedRightContentPosition = {
    position: rightContentPositionProp || DEFAULT_POSITION_TOP
  };
  if (isRightContentType(rightContentProp)) {
    var _rightContentProp$des;
    return _objectSpread(_objectSpread({}, computedRightContentPosition), {}, {
      content: rightContentProp === null || rightContentProp === void 0 ? void 0 : rightContentProp.content,
      description: (_rightContentProp$des = rightContentProp === null || rightContentProp === void 0 ? void 0 : rightContentProp.description) !== null && _rightContentProp$des !== void 0 ? _rightContentProp$des : undefined
    });
  }
  return _objectSpread(_objectSpread({}, computedRightContentPosition), {}, {
    content: rightContentProp || null,
    position: rightContentPositionProp || DEFAULT_POSITION_TOP
  });
};
var computeRightContent$1 = computeRightContent;
const ListItemContent = /*#__PURE__*/function () {
  const ListItemContent = props => {
    var _leftContent$type;
    const _props$addPrimaryLabe = props.addPrimaryLabel,
      addLabel = _props$addPrimaryLabe === void 0 ? false : _props$addPrimaryLabe,
      _props$baseList = props.baseList,
      baseList = _props$baseList === void 0 ? false : _props$baseList,
      _props$chevron = props.chevron,
      chevron = _props$chevron === void 0 ? false : _props$chevron,
      children = props.children,
      _props$component = props.component,
      Component = _props$component === void 0 ? React.Fragment : _props$component,
      _props$componentProps = props.componentProps,
      componentProps = _props$componentProps === void 0 ? {} : _props$componentProps,
      _props$description = props.description,
      description = _props$description === void 0 ? '' : _props$description,
      id = props.id,
      leftContent = props.leftContent,
      _props$legacyBehavior = props.legacyBehavior,
      legacyBehavior = _props$legacyBehavior === void 0 ? false : _props$legacyBehavior,
      maxLines = props.maxLines,
      _props$rightContent = props.rightContent,
      rightContent = _props$rightContent === void 0 ? '' : _props$rightContent,
      _props$rightContentPo = props.rightContentPosition,
      rightContentPosition = _props$rightContentPo === void 0 ? 'top' : _props$rightContentPo,
      _props$size = props.size,
      size = _props$size === void 0 ? MEDIUM : _props$size,
      _props$title = props.title,
      title = _props$title === void 0 ? '' : _props$title,
      value = props.value;
    const _useListContext = useListContext(),
      _useListContext$listT = _useListContext.listType,
      listType = _useListContext$listT === void 0 ? 'standalone' : _useListContext$listT;
    const attributes = getAttributesByListType(listType)();
    const hasDescription = description && size !== SMALL;
    const isThumbnail = /*#__PURE__*/isValidElement(leftContent) && (leftContent === null || leftContent === void 0 || (_leftContent$type = leftContent.type) === null || _leftContent$type === void 0 ? void 0 : _leftContent$type.displayName) === 'Thumbnail';
    const leftContentHasSrc = (leftContent === null || leftContent === void 0 ? void 0 : leftContent.src) && typeof leftContent.src === 'string';
    const computedRightContent = computeRightContent$1({
      rightContent,
      rightContentPosition
    });
    const stringRightContent = typeof (computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.content) === 'string';
    const centerRightContent = leftContentHasSrc || isThumbnail;
    const isSlotted = hasSlot(children, {
      title,
      description,
      rightContent: (computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.content) || (computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.description)
    }, {
      legacyBehavior
    });
    const primaryTextRef = useRef(null);
    const _useState = useState(),
      _useState2 = _slicedToArray(_useState, 2),
      primaryHeight = _useState2[0],
      setPrimaryHeight = _useState2[1];
    useEffect(() => {
      if (!centerRightContent) {
        var _primaryTextRef$curre;
        setPrimaryHeight((_primaryTextRef$curre = primaryTextRef.current) === null || _primaryTextRef$curre === void 0 ? void 0 : _primaryTextRef$curre.clientHeight);
      }
    }, []);
    const primaryClassNames = classNames(`${listItemNamespace}-primary`, {
      [`${listItemNamespace}-primary--max-lines-${maxLines}`]: maxLines
    });
    const secondColumnClassNames = classNames(`${listItemNamespace}-second-column`, {
      [`${listItemNamespace}-second-column-with-centered-text`]: centerRightContent && stringRightContent
    });
    const tertiaryContainerClassNames = classNames(`${listItemNamespace}-tertiary-container`, `${listItemNamespace}-tertiary-container--${centerRightContent ? 'centered' : computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.position}`);
    const tertiaryClassNames = classNames(`${listItemNamespace}-tertiary`, `${listItemNamespace}-tertiary--${centerRightContent ? 'centered' : computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.position}`);
    const assetClassNames = classNames(`${listItemNamespace}-asset`, {
      [`${listItemNamespace}-asset--icon`]: !centerRightContent,
      [`${listItemNamespace}-asset--icon-one-line`]: !hasDescription && primaryHeight && primaryHeight <= primarySizes[size]
    });
    const chevronClassNames = classNames(`${listItemNamespace}-chevron`, `${listItemNamespace}-chevron--${centerRightContent ? 'centered' : computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.position}`);
    const getTitle = () => addLabel ? /*#__PURE__*/
    React.createElement("label", _extends({
      className: primaryClassNames,
      htmlFor: `${id}-selection-component-${value}`,
      ref: primaryTextRef,
      onClick: event => event.stopPropagation()
    }, attributes === null || attributes === void 0 ? void 0 : attributes.title), title) : /*#__PURE__*/React.createElement("span", _extends({
      className: primaryClassNames,
      ref: primaryTextRef
    }, attributes === null || attributes === void 0 ? void 0 : attributes.title), title);
    const getLeftContent = () => {
      if (leftContentHasSrc) {
        return /*#__PURE__*/React.createElement(Image, _extends({}, leftContent, {
          className: `${listItemNamespace}-image`,
          "aria-hidden": true
        }));
      }
      return leftContent;
    };
    if (isSlotted) {
      return /*#__PURE__*/React.createElement(Component, componentProps, /*#__PURE__*/React.createElement("div", {
        className: `${listItemNamespace}-content`,
        id: addLabel ? `${id}-content-${value}` : `${id}-content`
      }, children), chevron && /*#__PURE__*/React.createElement("div", _extends({
        className: `${listItemNamespace}-chevron`
      }, attributes === null || attributes === void 0 ? void 0 : attributes.chevron), /*#__PURE__*/React.createElement(ChevronRight20, {
        color: "currentColor"
      })));
    }
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Component, componentProps, (title || description) && /*#__PURE__*/React.createElement("div", {
      className: `${listItemNamespace}-first-column`,
      id: `${id}-content`
    }, leftContent && /*#__PURE__*/React.createElement("div", _extends({
      className: assetClassNames,
      "aria-hidden": !(isThumbnail && leftContent.props.srLabel) || undefined
    }, attributes === null || attributes === void 0 ? void 0 : attributes.leftContent), getLeftContent()), /*#__PURE__*/React.createElement("div", {
      className: `${listItemNamespace}-text`
    }, title && getTitle(), hasDescription && /*#__PURE__*/React.createElement("span", _extends({
      className: `${listItemNamespace}-secondary`
    }, attributes === null || attributes === void 0 ? void 0 : attributes.description), description))), !baseList && ((computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.content) || chevron) && /*#__PURE__*/React.createElement("div", {
      className: secondColumnClassNames
    }, (computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.content) && /*#__PURE__*/React.createElement("span", _extends({
      className: tertiaryClassNames
    }, attributes === null || attributes === void 0 ? void 0 : attributes.rightContent, {
      id: `${id}-right-content`
    }), getVerifiedRightContent$1(computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.content, size)), chevron && /*#__PURE__*/React.createElement("div", {
      className: chevronClassNames
    }, /*#__PURE__*/React.createElement(ChevronRight20, {
      color: "currentColor"
    })))), children, baseList && ((computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.content) || chevron) && /*#__PURE__*/React.createElement("div", {
      className: secondColumnClassNames
    }, (computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.content) && /*#__PURE__*/React.createElement("div", _extends({
      className: tertiaryContainerClassNames
    }, attributes === null || attributes === void 0 ? void 0 : attributes.rightContent, {
      id: `${id}-right-content`
    }), /*#__PURE__*/React.createElement("span", {
      className: tertiaryClassNames
    }, getVerifiedRightContent$1(computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.content, size)), stringRightContent && (computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.description) && size !== 'small' && /*#__PURE__*/React.createElement("span", {
      className: `${listItemNamespace}-tertiary-description`
    }, computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.description)), chevron && /*#__PURE__*/React.createElement("div", _extends({
      className: chevronClassNames
    }, attributes === null || attributes === void 0 ? void 0 : attributes.chevron), /*#__PURE__*/React.createElement(ChevronRight20, {
      color: "currentColor"
    }))));
  };
  ListItemContent.displayName = LIST_ITEM_CONTENT_DISPLAY_NAME;
  return ListItemContent;
}();
var ListItemContent$1 = ListItemContent;
const hasImage = leftContent => {
  var _leftContent$type;
  const isThumbnail = /*#__PURE__*/isValidElement(leftContent) && (leftContent === null || leftContent === void 0 || (_leftContent$type = leftContent.type) === null || _leftContent$type === void 0 ? void 0 : _leftContent$type.displayName) === 'Thumbnail';
  const isImage = typeof (leftContent === null || leftContent === void 0 ? void 0 : leftContent.src) === 'string';
  return !!(isThumbnail || isImage);
};
const isLegacyLeftContent = leftContent => {
  const legacyLeftContent = (leftContent === null || leftContent === void 0 ? void 0 : leftContent.title) || (leftContent === null || leftContent === void 0 ? void 0 : leftContent.src) || (leftContent === null || leftContent === void 0 ? void 0 : leftContent.alt) || (leftContent === null || leftContent === void 0 ? void 0 : leftContent.src2x);
  return !!legacyLeftContent;
};
var _metrics$sendSlots;
const METRIC_COMPONENT = 'List';
const isSlotted = _ref => {
  let title = _ref.title,
    description = _ref.description,
    rightContent = _ref.rightContent,
    children = _ref.children,
    legacyBehavior = _ref.legacyBehavior;
  const defaultContent = title || description || rightContent;
  const shouldSendSlotMetric = legacyBehavior ? children && !defaultContent : children;
  return !!shouldSendSlotMetric;
};
const metrics = senderMetrics(trackCustomization);
const ListItemMetrics = {
  SSR: [metrics === null || metrics === void 0 || (_metrics$sendSlots = metrics.sendSlots) === null || _metrics$sendSlots === void 0 ? void 0 : _metrics$sendSlots.call(metrics, METRIC_COMPONENT, isSlotted)]
};
const _excluded$1 = ["chevron", "children", "className", "description", "disabled", "href", "id", "leftContent", "legacyBehavior", "legacyBehaviorSlot", "maxLines", "onClick", "paddingSize", "rightContent", "rightContentPosition", "size", "target", "title", "value", "groupIndex", "index", "ref"];
const ListActionItem = /*#__PURE__*/function () {
  const ListActionItem = props => {
    const _props$chevron = props.chevron,
      chevron = _props$chevron === void 0 ? true : _props$chevron,
      children = props.children,
      className = props.className,
      description = props.description,
      _props$disabled = props.disabled,
      disabled = _props$disabled === void 0 ? false : _props$disabled,
      href = props.href,
      id = props.id,
      leftContent = props.leftContent,
      legacyBehavior = props.legacyBehavior,
      legacyBehaviorSlot = props.legacyBehaviorSlot,
      maxLines = props.maxLines,
      onClick = props.onClick,
      paddingSize = props.paddingSize,
      rightContent = props.rightContent,
      rightContentPosition = props.rightContentPosition,
      _props$size = props.size,
      size = _props$size === void 0 ? MEDIUM : _props$size,
      target = props.target,
      title = props.title;
    props.value;
    props.groupIndex;
    props.index;
    const ref = props.ref,
      other = _objectWithoutProperties(props, _excluded$1);
    const componentId = useId(id);
    const hasDescription = !!description;
    const hasLeftContentImage = hasImage(leftContent);
    const isSlotted = hasSlot(children, {
      title,
      description,
      rightContent
    }, {
      legacyBehavior: legacyBehavior || legacyBehaviorSlot
    });
    const legacyLeftContent = isLegacyLeftContent(leftContent);
    warning(!legacyLeftContent, WARNING_LIST_ITEM_LEFT_CONTENT);
    warning(!!(href || onClick), WARNING_LIST_ACTION_ITEM_ACTION);
    const listBaseItemDeprecatedProps = legacyBehavior ? {
      selected: props === null || props === void 0 ? void 0 : props.selected
    } : {};
    const getPrimaryAction = () => !legacyBehavior && (href ? /*#__PURE__*/
    React.createElement("a", _extends({
      "aria-labelledby": `${componentId}-content`,
      className: `${listItemNamespace}-actionable`
    }, disabled ? {
      role: 'link',
      'aria-disabled': true
    } : {
      href,
      target,
      onClick
    })) : /*#__PURE__*/React.createElement("button", {
      "aria-labelledby": `${componentId}-content`,
      className: `${listItemNamespace}-actionable`,
      disabled: disabled,
      onClick: onClick,
      type: "button"
    }));
    const getLegacyWrapper = () => {
      if (!legacyBehavior || !(href || onClick)) return {};
      if (href) {
        return {
          component: 'a',
          componentProps: _objectSpread({
            className: `${listItemNamespace}-anchor`,
            onClick,
            target
          }, disabled ? {
            role: 'link',
            'aria-disabled': true
          } : {
            href
          })
        };
      }
      return {
        component: 'button',
        componentProps: {
          className: `${listItemNamespace}-action`,
          disabled,
          onClick,
          type: 'button'
        }
      };
    };
    return /*#__PURE__*/React.createElement(ListBaseItem$1, _extends({
      className: className,
      id: componentId,
      ref: ref,
      disabled: disabled,
      hasDescription: hasDescription,
      hasLeftContentImage: hasLeftContentImage,
      interactive: !legacyBehavior,
      paddingSize: paddingSize,
      size: size,
      slotted: isSlotted
    }, listBaseItemDeprecatedProps, other), getPrimaryAction(), /*#__PURE__*/React.createElement(ListItemContent$1, _extends({
      title: title,
      description: description,
      leftContent: leftContent,
      rightContent: rightContent,
      rightContentPosition: rightContentPosition,
      maxLines: maxLines,
      size: size,
      baseList: true,
      chevron: chevron,
      legacyBehavior: legacyBehavior,
      id: componentId
    }, getLegacyWrapper()), children));
  };
  ListActionItem.displayName = LIST_ACTION_ITEM_DISPLAY_NAME;
  return ListActionItem;
}();
var ListActionItem$1 = withTracker(ListActionItem, ListItemMetrics);
const _excluded = ["ref"],
  _excluded2 = ["chevron", "children", "className", "description", "disabled", "groupIndex", "href", "id", "leftContent", "maxLines", "onClick", "onSelect", "paddingSize", "rightContent", "rightContentPosition", "selected", "size", "target", "title", "value"];
const ListItem = /*#__PURE__*/function () {
  const ListItem = _ref => {
    let ref = _ref.ref,
      props = _objectWithoutProperties(_ref, _excluded);
    const chevron = props.chevron,
      children = props.children,
      className = props.className,
      description = props.description,
      disabled = props.disabled;
    props.groupIndex;
    const href = props.href,
      id = props.id,
      leftContent = props.leftContent,
      maxLines = props.maxLines,
      onClick = props.onClick;
    props.onSelect;
    const paddingSize = props.paddingSize,
      rightContent = props.rightContent,
      _props$rightContentPo = props.rightContentPosition,
      rightContentPosition = _props$rightContentPo === void 0 ? 'top' : _props$rightContentPo,
      _props$selected = props.selected,
      selected = _props$selected === void 0 ? false : _props$selected,
      _props$size = props.size,
      size = _props$size === void 0 ? MEDIUM : _props$size,
      target = props.target,
      title = props.title,
      value = props.value,
      other = _objectWithoutProperties(props, _excluded2);
    const listContext = useListContext();
    const computedRightContent = computeRightContent$1({
      rightContent,
      rightContentPosition
    });
    const hasDescription = !!description;
    const hasLeftContentImage = hasImage(leftContent);
    const isSlotted = hasSlot(children, {
      title,
      description,
      rightContent: (computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.content) || (computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.description)
    }, {
      legacyBehavior: true
    });
    const listBaseItemDeprecatedProps = {
      selected,
      disabled
    };
    const listActionItemDeprecatedProps = {
      selected
    };
    const listItemContentDeprecatedProps = {
      chevron: !!chevron,
      value: value === null || value === void 0 ? void 0 : value.toString()
    };
    warning(!!(listContext !== null && listContext !== void 0 && listContext.listType), WARNING_LIST_ITEM);
    if ((listContext === null || listContext === void 0 ? void 0 : listContext.listType) === 'standalone') {
      warning(typeof disabled === 'undefined', WARNING_DISABLED);
      warning(typeof onClick === 'undefined', WARNING_ON_CLICK);
    }
    warning(!(children && (title || description || rightContent)), WARNING_LIST_ITEM_CHILDREN);
    warning(typeof chevron === 'undefined', WARNING_CHEVRON);
    const legacyLeftContent = isLegacyLeftContent(leftContent);
    warning(!legacyLeftContent, WARNING_LIST_ITEM_LEFT_CONTENT);
    warning(typeof href === 'undefined', WARNING_HREF);
    warning(typeof target === 'undefined', WARNING_TARGET);
    if (href || onClick) {
      return /*#__PURE__*/React.createElement(ListActionItem$1, _extends({
        className: className,
        ref: ref,
        id: id,
        legacyBehavior: true,
        maxLines: maxLines,
        paddingSize: paddingSize,
        size: size,
        chevron: !!chevron,
        href: href,
        target: target,
        disabled: disabled,
        onClick: onClick,
        title: title,
        description: description,
        leftContent: leftContent,
        rightContent: computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.content,
        rightContentPosition: computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.position
      }, listActionItemDeprecatedProps, other), children);
    }
    return /*#__PURE__*/React.createElement(ListBaseItem$1, _extends({
      className: className,
      id: id,
      ref: ref,
      hasDescription: hasDescription,
      hasLeftContentImage: hasLeftContentImage,
      paddingSize: paddingSize,
      size: size,
      slotted: isSlotted
    }, listBaseItemDeprecatedProps, other), /*#__PURE__*/React.createElement(ListItemContent$1, _extends({
      title: title,
      description: description,
      leftContent: leftContent,
      rightContent: computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.content,
      rightContentPosition: computedRightContent === null || computedRightContent === void 0 ? void 0 : computedRightContent.position,
      maxLines: maxLines,
      size: size,
      baseList: true,
      legacyBehavior: true,
      id: id
    }, listItemContentDeprecatedProps), children));
  };
  ListItem.displayName = LIST_ITEM_DISPLAY_NAME;
  return ListItem;
}();
var ListItem$1 = withTracker(ListItem, ListItemMetrics);
export { List$1 as List, ListActionItem$1 as ListActionItem, ListBaseItem$1 as ListBaseItem, ListGroup$1 as ListGroup, ListItem$1 as ListItem, ListItemContent$1 as ListItemContent };
