// node-melitk-otel/lib/node-melitk-otel.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const env = __importStar(require("frontend-env"));
const api_1 = require("@opentelemetry/api");
const instrumentation_http_1 = require("@opentelemetry/instrumentation-http");
const instrumentation_express_1 = require("@opentelemetry/instrumentation-express");
const instrumentation_runtime_node_1 = require("@opentelemetry/instrumentation-runtime-node");
const instrumentation_1 = require("@opentelemetry/instrumentation");
const meter_provider_singleton_1 = require("./providers/meter-provider-singleton");
const tracer_provider_singleton_1 = require("./providers/tracer-provider-singleton");
const otel_1 = require("./constants/otel");
/**
 * Returns a new console logger instance for the OpenTelemetry diagnostics.
 * @returns {DiagConsoleLogger} The console logger for the OpenTelemetry diagnostics.
 */
const getConsoleLogger = () => new api_1.DiagConsoleLogger();
/**
 * Returns the OpenTelemetry log level based on the environment configuration.
 * If the O11Y_EVENTS_DEBUG_MODE is set to 'true', the log level will be set to INFO.
 * Otherwise, the log level will be set to the value of the `OTEL_LOG_LEVEL` environment variable,
 * or {@link DiagLogLevel.ERROR} if the environment variable is not set or invalid.
 * @returns {DiagLogLevel} The log level for the OpenTelemetry diagnostics.
 */
const getLogLevel = () => {
    if (env.O11Y_EVENTS_DEBUG_MODE === 'true') {
        return api_1.DiagLogLevel.INFO;
    }
    const logLevel = api_1.DiagLogLevel[env.OTEL_LOG_LEVEL?.toUpperCase()] || api_1.DiagLogLevel.ERROR;
    return logLevel;
};
/**
 * Initializes the OpenTelemetry implementations for tracing and metrics.
 * @typedef {Object} OTel
 * @property {NodeTracerProvider} tracerProvider - The initialized tracer provider.
 * @property {MeterProvider} meterProvider - The initialized meter provider.
 *
 * @returns {OTel} The initialized tracer and meter providers.
 */
const initializer = () => {
    // Set default exporter none
    env.OTEL_TRACES_EXPORTER = env.OTEL_TRACES_EXPORTER ?? 'none';
    // For troubleshooting, change the log level to debug
    api_1.diag.setLogger(getConsoleLogger(), getLogLevel());
    const meterProvider = meter_provider_singleton_1.MeterProviderSingleton.getInstance();
    const tracerProvider = tracer_provider_singleton_1.TracerProviderSingleton.getInstance();
    const hostsToIgnore = [
        // Collector route used by tracing
        'collector.newrelic.com',
    ];
    const headersToPropagate = [
        'x-request-id',
        'x-site-id',
        // Propagate B3 and W3C header values to the span attributes
        // so we can troubleshoot the trace context propagation.
        // B3 Single-Header format: {traceId}-{spanId}-{sampled}-{parentSpanId}
        'b3',
        // B3 Multi-Header Format
        'x-b3-sampled',
        'x-b3-traceid',
        'x-b3-spanid',
        'x-b3-parentspanid',
        'x-b3-flags',
        // W3C TraceContext format: "00-{trace-id}-{parent-id}-{trace-flags}"
        'traceparent',
        'tracestate',
        // W3C Baggage format: "key1=value1;key2=value2;..."
        'baggage',
    ];
    // Register the OpenTelemetry instrumentations for HTTP and Express
    (0, instrumentation_1.registerInstrumentations)({
        instrumentations: [
            new instrumentation_http_1.HttpInstrumentation({
                headersToSpanAttributes: {
                    server: {
                        requestHeaders: headersToPropagate,
                    },
                    client: {
                        requestHeaders: headersToPropagate,
                    },
                },
                ignoreOutgoingRequestHook: (reqOpts) => hostsToIgnore.some((hostToIgnore) => (reqOpts.host ?? '').includes(hostToIgnore)),
            }),
            new instrumentation_express_1.ExpressInstrumentation(),
            new instrumentation_runtime_node_1.RuntimeNodeInstrumentation({
                monitoringPrecision: otel_1.MONITORING_PRECISION,
            }),
        ],
        tracerProvider,
        meterProvider,
    });
    return { tracerProvider, meterProvider };
};
exports.default = initializer;
