// @andes/common/build/cjs/constants/countries.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
const COUNTRY_AR = 'AR'; // Argentina
const COUNTRY_BO = 'BO'; // Bolivia
const COUNTRY_BR = 'BR'; // Brasil
const COUNTRY_CL = 'CL'; // Chile
const COUNTRY_CO = 'CO'; // Colombia
const COUNTRY_CR = 'CR'; // Costa Rica
const COUNTRY_EC = 'EC'; // Ecuador
const COUNTRY_SV = 'SV'; // El Salvador
const COUNTRY_GT = 'GT'; // Guatemala
const COUNTRY_HN = 'HN'; // Honduras
const COUNTRY_MX = 'MX'; // Mexico
const COUNTRY_NI = 'NI'; // Nicaragua
const COUNTRY_PA = 'PA'; // Panama
const COUNTRY_PY = 'PY'; // Paraguay
const COUNTRY_PE = 'PE'; // Perú
const COUNTRY_DO = 'DO'; // República Dominicana
const COUNTRY_UY = 'UY'; // Uruguay
const COUNTRY_VE = 'VE'; // Venezuela

const AVAILABLE_COUNTRIES = [COUNTRY_AR, COUNTRY_BO, COUNTRY_BR, COUNTRY_CL, COUNTRY_CO, COUNTRY_CR, COUNTRY_EC, COUNTRY_SV, COUNTRY_GT, COUNTRY_HN, COUNTRY_MX, COUNTRY_NI, COUNTRY_PA, COUNTRY_PY, COUNTRY_PE, COUNTRY_DO, COUNTRY_UY, COUNTRY_VE];
exports.AVAILABLE_COUNTRIES = AVAILABLE_COUNTRIES;
exports.COUNTRY_AR = COUNTRY_AR;
exports.COUNTRY_BO = COUNTRY_BO;
exports.COUNTRY_BR = COUNTRY_BR;
exports.COUNTRY_CL = COUNTRY_CL;
exports.COUNTRY_CO = COUNTRY_CO;
exports.COUNTRY_CR = COUNTRY_CR;
exports.COUNTRY_DO = COUNTRY_DO;
exports.COUNTRY_EC = COUNTRY_EC;
exports.COUNTRY_GT = COUNTRY_GT;
exports.COUNTRY_HN = COUNTRY_HN;
exports.COUNTRY_MX = COUNTRY_MX;
exports.COUNTRY_NI = COUNTRY_NI;
exports.COUNTRY_PA = COUNTRY_PA;
exports.COUNTRY_PE = COUNTRY_PE;
exports.COUNTRY_PY = COUNTRY_PY;
exports.COUNTRY_SV = COUNTRY_SV;
exports.COUNTRY_UY = COUNTRY_UY;
exports.COUNTRY_VE = COUNTRY_VE;
