// @andes/list/build/cjs/utils/getAttributesByListType.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

/**
 * @param {'standalone' | 'listbox' | 'checkbox' | 'radio'} [listType]
 */
const getAttributesByListType = function () {
  let listType = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'standalone';
  return /** @param {Pick<import('../..').ListItemProps, 'size'>} param */function () {
    let _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      size = _ref.size;
    let listName = 'list';
    switch (listType) {
      case 'standalone':
        listName = 'list';
        break;
      case 'listbox':
        listName = 'listbox';
        break;
      case 'checkbox':
        listName = 'checkbox-list';
        break;
      case 'radio':
        listName = 'radio-list';
        break;
      default:
        listName = 'list';
        break;
    }
    return {
      list: {
        [`data-andes-${listName}`]: true
      },
      group: {
        'data-andes-list-group': true
      },
      header: {
        'data-andes-list-group-header': true
      },
      sublist: {
        'data-andes-list-group-sublist': true
      },
      item: {
        [`data-andes-${listName}-item`]: true,
        [`data-andes-${listName}-item-size`]: size
      },
      chevron: {
        'data-andes-list-chevron': true
      },
      selection: {
        [`data-andes-${listName}-${listType === 'listbox' ? 'checkbox' : listType}`]: true
      },
      title: {
        [`data-andes-${listName}-title`]: true
      },
      description: {
        [`data-andes-${listName}-description`]: true
      },
      leftContent: {
        [`data-andes-${listName}-left-content`]: true
      },
      rightContent: {
        [`data-andes-${listName}-right-content`]: true
      }
    };
  };
};
exports.getAttributesByListType = getAttributesByListType;
