// @nordic/fs-cache/src/assets/manifest/webpack.ts
// TODO: Remove file when webpack support is dropped
import type { WritableDeep } from 'type-fest';
import type { StatsCompilation } from 'webpack';
import type { MainAssets, MainAssetsData } from './types';

import { freezeObject } from '@nordic/utils/freeze';
import { serializeJSON } from '@nordic/utils/serialize';

import { readJsonFileSync } from '../../utils/fs';
import { MANIFEST_ENTRYPOINTS_EXTENSIONS_REGEX } from '../regex';

type Stats = Required<
  Pick<
    StatsCompilation,
    'assetsByChunkName' | 'entrypoints' | 'namedChunkGroups'
  >
>;

const MANIFEST_EXCLUDE_EXTENSIONS_REGEX = /\.(?:map|json)$/;

const DEFAULT_STATS: Stats = {
  entrypoints: {},
  assetsByChunkName: {},
  namedChunkGroups: {},
};

function createBaseManifestEntries() {
  const manifest = readJsonFileSync<Record<string, string>>(
    'build/manifest.json',
  );

  if (!manifest) {
    return [];
  }

  return Object.entries(manifest).filter(
    ([, value]) =>
      !MANIFEST_EXCLUDE_EXTENSIONS_REGEX.test(value) &&
      !value.includes('node_modules'),
  );
}

function createMainAssetsManifestEntries(
  namedChunkGroups: Stats['namedChunkGroups'],
  manifestEntries: Array<[string, string]>,
): Array<[string, MainAssets]> {
  const invertedManifest = new Map(
    manifestEntries.map(([key, value]) => [value, key]),
  );

  return Object.entries(namedChunkGroups).map(([key, value]) => {
    const scripts: WritableDeep<MainAssetsData> = {
      names: [],
      sources: [],
      serializedNames: '',
    };
    const styles: WritableDeep<MainAssetsData> = {
      names: [],
      sources: [],
      serializedNames: '',
    };

    value.assets?.forEach(({ name: fileName }) => {
      const { sources, names } = fileName.endsWith('.js') ? scripts : styles;
      const assetName = invertedManifest.get(fileName);

      if (__DEV__ && fileName.endsWith('.hot-update.js')) {
        return;
      }

      sources.push(fileName);

      if (assetName) {
        names.push(assetName);
      }
    });

    return [
      key,
      {
        scripts: {
          names: freezeObject(scripts.names),
          sources: freezeObject(scripts.sources),
          serializedNames: serializeJSON(scripts.names),
        },
        styles: {
          names: freezeObject(styles.names),
          sources: freezeObject(styles.sources),
          serializedNames: serializeJSON(styles.names),
        },
      },
    ];
  });
}

export function createManifestEntries() {
  const stats = readJsonFileSync<Stats>('build/stats.json') ?? DEFAULT_STATS;
  const baseManifestEntries = createBaseManifestEntries();
  const mainAssetsManifestEntries = createMainAssetsManifestEntries(
    stats.namedChunkGroups,
    baseManifestEntries,
  );
  const entrypointsValues = Object.values(stats.entrypoints);

  return {
    baseManifestEntries: baseManifestEntries.filter(
      ([, value]) =>
        !MANIFEST_ENTRYPOINTS_EXTENSIONS_REGEX.test(value) ||
        entrypointsValues.some(({ assets }) =>
          assets?.some(({ name }) => name === value),
        ),
    ),
    mainAssetsManifestEntries,
  };
}
