// nordic-pages-discovery/lib/helpers/forward-ref.js
// TODO: Delete file when "forwardRef" support is removed
const babelTypes = require('@babel/types');
const traverse = require('@babel/traverse').default;

const { IMPORT_TYPES } = require('../constants/islands');

const FORWARD_REF_NAME = 'forwardRef';

/**
 * @typedef {{ name: string?, isLocal: boolean }} ForwardRefInfo
 */

/**
 * @param {import('../cache/ast-cache').BabelAST} ast
 * @param {import('@babel/types').CallExpression} callExpression
 * @returns {ForwardRefInfo} fnInfo
 */
const getFnInfoIfForwardRef = (ast, { callee, arguments: args }) => {
  const fnInfo = {
    name: null,
    isLocal: false,
  };
  const forwardRefUsage = {
    localName: null,
    namedExport: false,
  };

  traverse(ast, {
    noScope: true,
    ImportDeclaration(path) {
      const { source, specifiers } = path.node;

      if (source.value !== 'react') {
        if (forwardRefUsage.localName) {
          path.stop();
        }

        return;
      }

      specifiers.forEach((specifier) => {
        // example 1: import * as React from 'react';
        // example 2: import React from 'react';
        // example 3: import { forwardRef } from 'react';
        // example 4: import { forwardRef as forwardRefFn } from 'react';
        const namedExport =
          babelTypes.isImportSpecifier(specifier) &&
          babelTypes.isIdentifier(specifier.imported) &&
          specifier.imported.name !== IMPORT_TYPES.DEFAULT;

        if (namedExport && specifier.imported.name === FORWARD_REF_NAME) {
          forwardRefUsage.namedExport = true;
          forwardRefUsage.localName = specifier.local.name;

          path.stop();

          return;
        }

        if (!namedExport) {
          forwardRefUsage.namedExport = false;
          forwardRefUsage.localName = specifier.local.name;
        }
      });
    },
  });

  if (
    !forwardRefUsage.localName ||
    (!(
      forwardRefUsage.namedExport &&
      babelTypes.isIdentifier(callee) &&
      callee.name === forwardRefUsage.localName
    ) &&
      !(
        !forwardRefUsage.namedExport &&
        babelTypes.isMemberExpression(callee) &&
        babelTypes.isIdentifier(callee.object) &&
        callee.object.name === forwardRefUsage.localName &&
        babelTypes.isIdentifier(callee.property) &&
        callee.property.name === FORWARD_REF_NAME
      ))
  ) {
    return fnInfo;
  }

  const [fn] = args;

  if (!babelTypes.isIdentifier(fn)) {
    fnInfo.name =
      babelTypes.isFunctionExpression(fn) && babelTypes.isIdentifier(fn.id)
        ? fn.id.name
        : '';
    fnInfo.isLocal = true;

    return fnInfo;
  }

  fnInfo.name = fn.name;
  fnInfo.isLocal = false;

  traverse(ast, {
    noScope: true,
    VariableDeclarator(path) {
      const { id, init } = path.node;

      if (
        !babelTypes.isIdentifier(id) ||
        id.name !== fnInfo.name ||
        !babelTypes.isVariableDeclaration(path.parent) ||
        !babelTypes.isProgram(path.parentPath.parent)
      ) {
        return;
      }

      if (!babelTypes.isArrowFunctionExpression(init)) {
        path.stop();

        return;
      }

      fnInfo.isLocal = true;

      path.stop();
    },
  });

  return fnInfo;
};

module.exports = {
  getFnInfoIfForwardRef,
};
