// @andes/close-listener/build/esm/index.development.js
import React from 'react';
import PropTypes from 'prop-types';

/* global document */

/** @typedef {import('../..').CloseListenerProps} CloseListenerProps */

/** @extends React.Component<CloseListenerProps> */
class CloseListener extends React.Component {
  /** @param {CloseListenerProps} props */
  constructor(props) {
    super(props);
    this.handleRef = this.handleRef.bind(this);
    this.handleDocumentClick = this.handleDocumentClick.bind(this);
    this.pressEscToClose = this.pressEscToClose.bind(this);
  }
  componentDidMount() {
    document.addEventListener('click', this.handleDocumentClick);
    document.addEventListener('keydown', this.pressEscToClose, false);
  }
  componentWillUnmount() {
    document.removeEventListener('click', this.handleDocumentClick);
    document.removeEventListener('keydown', this.pressEscToClose, false);
  }

  /** @param {KeyboardEvent} event */
  pressEscToClose(event) {
    const _this$props = this.props,
      _this$props$handleEsc = _this$props.handleEsc,
      handleEsc = _this$props$handleEsc === void 0 ? true : _this$props$handleEsc,
      onCloseOutside = _this$props.onCloseOutside;
    if (handleEsc) {
      if (event.code === 'Escape') {
        // esc key
        onCloseOutside(event);
      }
    }
  }

  /** @param {MouseEvent} event */
  handleDocumentClick(event) {
    const _this$props2 = this.props,
      _this$props2$handleCl = _this$props2.handleClick,
      handleClick = _this$props2$handleCl === void 0 ? true : _this$props2$handleCl,
      onCloseOutside = _this$props2.onCloseOutside;
    if (handleClick) {
      if (this.element && !this.element.contains(/** @type {Node} */event.target)) {
        onCloseOutside(event);
      }
    }
  }

  /** @param {HTMLElement} el */
  handleRef(el) {
    const children = this.props.children;
    this.element = el;
    // @ts-expect-error ref prop doesn't exist on ReactElement
    if (children.ref && typeof children.ref === 'function') {
      // @ts-expect-error ref prop doesn't exist on ReactElement
      children.ref(el);
    }
  }
  render() {
    const children = this.props.children;
    return /*#__PURE__*/React.cloneElement(/** @type {React.ReactElement} */children, {
      ref: this.handleRef
    });
  }
}
CloseListener.propTypes = {
  children: PropTypes.element.isRequired,
  onCloseOutside: PropTypes.func.isRequired,
  handleClick: PropTypes.bool,
  handleEsc: PropTypes.bool
};
var CloseListener$1 = CloseListener;
export { CloseListener$1 as CloseListener };
