// nordic-babel-alias-resolver/src/aliases.js
const path = require('path');

const { getTsconfig } = require('get-tsconfig');

const resolveAliases = require('./resolveAliases');

/**
 * @param {string} str
 */
function escapeRegex(str) {
  return str.replace(/[-\\^$*+?.()|[\]{}]/g, '\\$&');
}

function getCompilerOptions(searchPath = '.') {
  const tsConfig = getTsconfig(searchPath)?.config;
  const config = tsConfig || getTsconfig(searchPath, 'jsconfig.json')?.config;

  const { paths, baseUrl } = config?.compilerOptions || {};

  if (!paths) {
    throw new Error(
      'ModulePathAliases: Could not resolve paths from tsconfig.json or jsconfig.json.',
    );
  }

  if (baseUrl && path.resolve(baseUrl) !== path.resolve()) {
    throw new Error(
      'ModulePathAliases: baseUrl must be set to the project root directory.',
    );
  }

  return { paths, baseUrl };
}

/**
 * ts-jest adapted pathsToModuleNameMapper function.
 * https://github.com/kulshekhar/ts-jest/blob/v29.1.2/src/config/paths-to-module-name-mapper.ts
 * @param {Record<string,string[]>} mapping
 * @param {string} prefix
 */
function pathsToModuleNameMapper(mapping, prefix) {
  /** @type {Record<string,string[]|string>} */
  const jestMap = {};

  const enrichedPrefix =
    prefix !== '' && !prefix.endsWith('/') ? `${prefix}/` : prefix;

  Object.entries(mapping).forEach(([fromPath, toPaths]) => {
    if (!toPaths.length) {
      throw new Error(
        `ModulePathAliases: No target path specified for alias: ${fromPath}`,
      );
    }

    // this string split will never return length 0
    const segments = fromPath.split(/\*/g);

    if (segments.length === 1) {
      const paths = toPaths.map((target) => `${enrichedPrefix}${target}`);

      const cjsPattern = `^${escapeRegex(fromPath)}$`;

      jestMap[cjsPattern] = paths.length === 1 ? paths[0] : paths;

      return;
    }

    if (segments.length === 2) {
      const paths = toPaths.map((target) => {
        const enrichedTarget =
          target.startsWith('./') && prefix !== ''
            ? target.substring(target.indexOf('/') + 1)
            : target;

        return `${enrichedPrefix}${enrichedTarget.replace(/\*/g, '$1')}`;
      });

      const cjsPattern = `^${escapeRegex(segments[0])}(.*)${escapeRegex(segments[1])}$`;

      jestMap[cjsPattern] = paths.length === 1 ? paths[0] : paths;

      return;
    }

    throw new Error(
      `ModulePathAliases: Multiple wildcard * are not supported in alias: ${fromPath}`,
    );
  });

  return jestMap;
}

function resolveJestAliases() {
  const { paths } = getCompilerOptions();

  return pathsToModuleNameMapper(paths, '<rootDir>');
}

function resolveStorybookAliases() {
  const aliases = resolveAliases({
    isWebpack: true,
    isExternal: true,
    rootPath: '..',
  });

  if (!aliases) {
    throw new Error(
      'ModulePathAliases: Could not resolve paths from tsconfig.json or jsconfig.json.',
    );
  }

  return aliases;
}

/**
 * @return {Array<{ find: RegExp | string; replacement: string }>}
 */
function resolveVitestAliases() {
  const { paths } = getCompilerOptions();

  return Object.entries(pathsToModuleNameMapper(paths, '')).map(
    ([key, value]) => {
      const pathValue = Array.isArray(value) ? value[0] : value;

      return {
        find: new RegExp(key),
        replacement: path.resolve('.', pathValue),
      };
    },
  );
}

module.exports = {
  resolveJestAliases,
  resolveVitestAliases,
  resolveStorybookAliases,
};
