// nordic-babel-alias-resolver/src/loadModuleResolverPlugin.js
const resolveAliases = require('./resolveAliases');
const resolveSubpathImports = require('./resolveSubpathImports');

/**
 * @param {boolean} isClient - Whether the plugin is being used in a client context
 * @returns {Array<[string, import('../types').ResolverOptions]>}
 */
function loadModuleResolverPlugin(isClient) {
  const aliases = resolveAliases({
    isWebpack: false,
    isExternal: false,
    isBabelClient: isClient,
  });

  const subpathImports = resolveSubpathImports({
    isBabelClient: isClient,
  });

  if (!aliases && !subpathImports) {
    return [];
  }

  /** @type {import('../types').ResolverOptions} */
  const pluginOptions = {
    // @ts-expect-error - real return type
    root: aliases?.root || ['.'],
    // @ts-expect-error - isWebpack false guarantees aliases has { root, alias }
    alias: aliases?.alias,
  };

  if (subpathImports) {
    pluginOptions.subpathImportsResolver = (filename) =>
      resolveSubpathImports({ isBabelClient: isClient, filename });
  }

  return [['babel-plugin-nordic-module-resolver', pluginOptions]];
}

module.exports = loadModuleResolverPlugin;
