// @nordic/core-context/src/server/device/utils/web/network-state.ts
import { getReq } from '@meli/server-context';

type NetworkState = {
  isConnectionExpensive: boolean;
  isConnectionSlow: boolean;
  effectiveType?: string;
};

const SLOW_EFFECTIVE_TYPES = new Set(['slow-2g', '2g', '3g']);

/**
 * Returns an initial value of the network state.
 * The final network state value is obtained from [@nordic/network](https://github.com/melisource/fury_nordic-webkit/tree/master/packages/nordic-network).
 */
export function getNetworkState(): NetworkState {
  const {
    ect: effectiveType,
    downlink: downlinkStr,
    saveData,
  } = getReq().device;
  const downlink = typeof downlinkStr === 'string' ? +downlinkStr : downlinkStr;
  const isConnectionExpensive = saveData === 'on';
  const isConnectionSlow =
    typeof downlink === 'number' && downlink > 0 && downlink < 2;

  if (!effectiveType) {
    return { isConnectionExpensive, isConnectionSlow };
  }

  return {
    effectiveType,
    isConnectionExpensive,
    isConnectionSlow:
      isConnectionSlow || SLOW_EFFECTIVE_TYPES.has(effectiveType),
  };
}
