// node-melitk-otel/lib/events/sampler/alwaysRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlwaysRecordSampler = void 0;
exports.NewAlwaysRecordSampler = NewAlwaysRecordSampler;
const sdk_trace_base_1 = require("@opentelemetry/sdk-trace-base");
/**
 * AlwaysRecordSampler that decides whether a span should be sampled or not based on Events activation.
 * @implements {Sampler}
 * @class
 */
class AlwaysRecordSampler {
    rootSampler;
    pingSampler;
    activator;
    telemetry;
    constructor(rootSampler, pingSampler, activator, telemetry) {
        this.rootSampler = rootSampler;
        this.pingSampler = pingSampler;
        this.activator = activator;
        this.telemetry = telemetry;
    }
    /**
     * ShouldSample will decide whether a span should be exported based on the activator result.
     */
    shouldSample(context, traceId, spanName, spanKind, attributes, links) {
        // We must check if the span is a ping route before checking the activator.
        // If the span is a ping route, we should not record it regardless of the activator result.
        const pingSamplerResult = this.pingSampler.shouldSample(context, traceId, spanName, spanKind, attributes, links);
        if (pingSamplerResult.decision === sdk_trace_base_1.SamplingDecision.NOT_RECORD) {
            this.telemetry.incrementStartedSpansCounter(pingSamplerResult);
            return pingSamplerResult;
        }
        let res = this.rootSampler.shouldSample(context, traceId, spanName, spanKind, attributes, links);
        if (res.decision === sdk_trace_base_1.SamplingDecision.NOT_RECORD &&
            this.activator.isActive()) {
            res = { ...res, decision: sdk_trace_base_1.SamplingDecision.RECORD };
        }
        this.telemetry.incrementStartedSpansCounter(res);
        return res;
    }
    toString() {
        return `AlwaysRecordSampler(${this.rootSampler.toString()})`;
    }
}
exports.AlwaysRecordSampler = AlwaysRecordSampler;
function NewAlwaysRecordSampler(rootSampler, pingSampler, activator, telemetry) {
    return new AlwaysRecordSampler(rootSampler, pingSampler, activator, telemetry);
}
