// node-melitk-otel/lib/events/sampler/telemetry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_trace_base_1 = require("@opentelemetry/sdk-trace-base");
const metric_1 = require("../common/metric");
const config_1 = require("../config/config");
const samplingDecisionToStringMap = {
    [sdk_trace_base_1.SamplingDecision.NOT_RECORD]: 'drop',
    [sdk_trace_base_1.SamplingDecision.RECORD]: 'record_only',
    [sdk_trace_base_1.SamplingDecision.RECORD_AND_SAMPLED]: 'record_and_sample',
};
function getSamplingDecisionString(decision) {
    return samplingDecisionToStringMap[decision] || 'unknown';
}
/**
 * Represents the SamplerTelemetry class which extends O11yMetrics.
 * This class provides metrics related to the Sampler's behavior.
 * @class
 * @extends {O11yMetrics}
 */
class SamplerTelemetry extends metric_1.O11yMetrics {
    /**
     * Represents the metric `started.spans`.
     * This metric represents the number of spans that were started by the Sampler.
     * @type {Counter}
     * @private
     * @readonly
     */
    _startedSpansCounter;
    /**
     * Creates an instance of SamplerTelemetry.
     */
    constructor() {
        super();
        this._startedSpansCounter = this.newCounter('started.spans', {
            description: 'This metric represents the number of spans that were started by the Sampler.',
            unit: '1',
        });
    }
    /**
     * Increments the metric `started.spans` by 1 and adds attributes for sampling decision, technology, and sdkVersion.
     * @param samplingResult The sampling result containing the decision.
     */
    incrementStartedSpansCounter = (samplingResult) => {
        const attributes = {
            sampling_decision: getSamplingDecisionString(samplingResult.decision),
            ...config_1.metricDefaultAttributes,
        };
        this._startedSpansCounter.add(1, attributes);
    };
}
exports.default = SamplerTelemetry;
