// @andes/floating-menu/build/cjs/index.development.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var _extends = require('@babel/runtime/helpers/extends');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var React = require('react');
var PropTypes = require('prop-types');
var classNames = require('classnames');
var context = require('@andes/context');
var themes = require('@andes/themes');
var popper = require('@andes/popper');
var withTracker = require('@andes/with-tracker');
var keyboardCodes = require('@andes/common/build/constants/keyboardCodes');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var includes = require('@andes/common/build/utils/includes');
var usePreventScroll = require('@andes/common/build/hooks/usePreventScroll');
var focus = require('@react-aria/focus');
var visuallyHidden = require('@andes/visually-hidden');
var warning = require('@andes/common/build/utils/warning');
var Close16 = require('@andes/icons/Close16');
var Search16 = require('@andes/icons/Search16');
var Close20 = require('@andes/icons/Close20');
var Search20 = require('@andes/icons/Search20');
var checkbox = require('@andes/checkbox');
var listbox = require('@andes/listbox');
var list = require('@andes/list');
var metrics$1 = require('@andes/common/build/metrics');
var groupNodeGenerator = require('@andes/list/build/generators/groupNodeGenerator');
var itemNodeGenerator = require('@andes/list/build/generators/itemNodeGenerator');
function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var _objectWithoutProperties__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutProperties);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var includes__default = /*#__PURE__*/_interopDefaultLegacy(includes);
var usePreventScroll__default = /*#__PURE__*/_interopDefaultLegacy(usePreventScroll);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var Close16__default = /*#__PURE__*/_interopDefaultLegacy(Close16);
var Search16__default = /*#__PURE__*/_interopDefaultLegacy(Search16);
var Close20__default = /*#__PURE__*/_interopDefaultLegacy(Close20);
var Search20__default = /*#__PURE__*/_interopDefaultLegacy(Search20);
var groupNodeGenerator__default = /*#__PURE__*/_interopDefaultLegacy(groupNodeGenerator);
var itemNodeGenerator__default = /*#__PURE__*/_interopDefaultLegacy(itemNodeGenerator);
const ITEM_DISPLAY_NAME = 'FloatingMenuItem';
const GROUP_DISPLAY_NAME = 'FloatingMenuGroup';
const SUPPORTED_GROUP_TYPES = ['FloatingMenuGroup', 'DropdownGroup'];
const TYPE_MENU = 'menu';
const TYPE_SELECT = 'select';
const SEARCH_BOX_SIZE = 52;
const LIST_ITEM_SIZE = 48;
const listHoverClass = 'andes-floating-menu-item-hover';

/** @typedef {import('../../index').SearchConfigProps | boolean | undefined} searchConfig */

/**
 * @typedef {Object} componentSpecificProps
 * @prop {string} listId
 * @prop {string} id
 * @prop {searchConfig} searchConfig
 */

/**
 * @typedef {Object} componentInternalProps
 * @prop {import('@andes/popper').PopperProps['offset']} [offset]
 * @prop {import('@andes/popper').PopperProps['onMount']} [onMount]
 * @prop {import('@andes/popper').PopperProps['popperOptions']} [popperOptions]
 * @prop {import('react').CSSProperties} [style]
 */

/**
 * @param {import('react').RefObject<HTMLButtonElement | null>} triggerRef
 * @param {import('react').RefObject<HTMLUListElement | null>} menuRef
 * @param {import('react').RefObject<HTMLDivElement | null>} searchBoxRef
 * @param {import('../../index').FloatingMenuProps
 * & componentSpecificProps & componentInternalProps} componentProps
 */
const useFloatingMenu = (triggerRef, menuRef, searchBoxRef, componentProps) => {
  const allowUnselect = componentProps.allowUnselect,
    bottomContent = componentProps.bottomContent,
    children = componentProps.children,
    defaultOpen = componentProps.defaultOpen,
    defaultValue = componentProps.defaultValue,
    listId = componentProps.listId,
    menuMaxHeight = componentProps.menuMaxHeight,
    multiple = componentProps.multiple,
    offset = componentProps.offset,
    onChange = componentProps.onChange,
    onClose = componentProps.onClose,
    onOpen = componentProps.onOpen,
    open = componentProps.open,
    popperOptions = componentProps.popperOptions,
    searchConfig = componentProps.searchConfig,
    style = componentProps.style,
    topContent = componentProps.topContent,
    type = componentProps.type,
    value = componentProps.value,
    width = componentProps.width;
  const isGroup = React.useMemo(() => {
    var _firstOption$type;
    if (!children) return false;
    const firstOption = React.Children.toArray(children)[0];
    // @ts-expect-error: We add displayName property
    return includes__default["default"](SUPPORTED_GROUP_TYPES, firstOption === null || firstOption === void 0 || (_firstOption$type = firstOption.type) === null || _firstOption$type === void 0 ? void 0 : _firstOption$type.displayName);
  }, [children]);
  const _useState = React.useState(value || defaultValue),
    _useState2 = _slicedToArray__default["default"](_useState, 2),
    currentValues = _useState2[0],
    setCurrentValues = _useState2[1];
  const _useState3 = React.useState(false),
    _useState4 = _slicedToArray__default["default"](_useState3, 2),
    isMenuVisible = _useState4[0],
    setIsMenuVisible = _useState4[1];
  const _useState5 = React.useState(/** @type {string | undefined} */undefined),
    _useState6 = _slicedToArray__default["default"](_useState5, 2),
    searchItemFocused = _useState6[0],
    setSearchItemFocused = _useState6[1];
  const _useState7 = React.useState(typeof searchConfig === 'object' && (searchConfig === null || searchConfig === void 0 ? void 0 : searchConfig.defaultValue)),
    _useState8 = _slicedToArray__default["default"](_useState7, 2),
    searchText = _useState8[0],
    setSearchText = _useState8[1];
  usePreventScroll__default["default"]({
    isDisabled: type === TYPE_MENU || type === TYPE_SELECT && !isMenuVisible
  });

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const isUncontrolled = React.useMemo(() => value === undefined, []);
  const hasTopSlot = typeof topContent !== 'undefined';
  const hasBottomSlot = typeof bottomContent !== 'undefined';
  const isSlotted = hasTopSlot || hasBottomSlot || !!searchConfig;
  const role = type === 'select' ? 'combobox' : undefined;
  React.useEffect(() => {
    if (!isUncontrolled) {
      setCurrentValues(value);
    }
  }, [value, isUncontrolled]);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const isOpenUncontrolled = React.useMemo(() => open === undefined, []);
  const isEffectiveOpen = isOpenUncontrolled ? isMenuVisible : open !== null && open !== void 0 ? open : false;
  const ariaHasPopupValue = isSlotted || type === 'menu' && !multiple ? 'dialog' : 'listbox';

  /** @param {HTMLUListElement} menu */
  const getItemToFocus = menu => {
    const optionsList = Array.prototype.slice.call(menu.querySelectorAll('.andes-list__item'));
    if (optionsList.length) {
      let focusableOption;

      // Search first menu item with button or link when type menu
      if (type === TYPE_MENU) {
        focusableOption = optionsList[0].querySelector('.andes-list__item-action, .andes-list__item-anchor, .andes-list__item-actionable');
        if (focusableOption) return focusableOption;
      }

      // Search last navigated option when type select
      focusableOption = optionsList.find(option => option.tabIndex === 0);

      // Search selected option
      if (!focusableOption) {
        focusableOption = optionsList.find(option => option.className.indexOf('andes-list__item--selected') !== -1);
      }

      // default to first option
      if (!focusableOption) {
        var _optionsList = _slicedToArray__default["default"](optionsList, 1);
        focusableOption = _optionsList[0];
      }
      return focusableOption;
    }
    return null;
  };

  /**
   * @param {import('react').KeyboardEvent | import('react').MouseEvent} event
   * @param {string | string[]} [values]
   */
  const handleSelectionChange = (event, values) => {
    const shouldAvoidUnselect = !allowUnselect && !multiple && (values === undefined || !(values !== null && values !== void 0 && values.length)) && currentValues !== undefined;
    if (shouldAvoidUnselect) return;
    if (isUncontrolled) setCurrentValues(values);
    let newValue = values;
    if (!multiple && Array.isArray(values)) {
      var _values = _slicedToArray__default["default"](values, 1);
      newValue = _values[0];
    }
    if (searchConfig) setSearchItemFocused(state => !Array.isArray(newValue) ? newValue : state);
    if (typeof onChange === 'function') onChange(event, newValue);
  };
  const setPopoverFocus = () => {
    if (searchConfig) {
      var _searchBoxRef$current;
      /* The focus is applied to the label element instead of the input
      due to the keybord focus styles that need to be added */
      const label = searchBoxRef === null || searchBoxRef === void 0 || (_searchBoxRef$current = searchBoxRef.current) === null || _searchBoxRef$current === void 0 ? void 0 : _searchBoxRef$current.querySelector('label');
      setTimeout(() => label === null || label === void 0 ? void 0 : label.focus(), 0);
      return;
    }
    const menu = menuRef.current;
    if (menu) {
      const itemToFocus = getItemToFocus(menu);
      if (itemToFocus) {
        itemToFocus.focus();
        menu.scrollTop = itemToFocus.offsetTop - LIST_ITEM_SIZE;
      }
    }
  };

  /** @param {string} text */
  const onSearchChangeText = text => setSearchText(text);
  const fitsSearchValue = React.useCallback((/** @type {import('react').ReactElement<import('../../index').FloatingMenuItemProps>} */item) => {
    if (!searchText) return true;
    const optionText = item.props.title;
    if (typeof searchConfig === 'object' && typeof searchConfig.filterCondition === 'function') {
      return searchConfig.filterCondition(searchText, optionText, item);
    }
    return optionText && includes__default["default"](optionText.toLowerCase(), searchText.toLowerCase());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchText, typeof searchConfig === 'object' && (searchConfig === null || searchConfig === void 0 ? void 0 : searchConfig.filterCondition)]);
  const options = React.useMemo(() => {
    if (!children) return null;
    if (isGroup) {
      let index = 0;

      /** @type {import('react').ReactElement<import('../../index').FloatingMenuGroup>[] | null} */
      const groups = [];
      React.Children.forEach(children, (/** @type {React.ReactElement} */group, groupIndex) => {
        /** @type {import('react').ReactComponentElement<import('../../index').FloatingMenuGroup>[] | null} */
        const newChildren = [];
        React.Children.forEach(group === null || group === void 0 ? void 0 : group.props.children, item => {
          if (fitsSearchValue(item)) {
            newChildren.push(/*#__PURE__*/React.cloneElement(item, {
              index,
              groupIndex
            }));
          }
          index += 1;
        });
        if (newChildren.length > 0) {
          groups.push(/*#__PURE__*/React.cloneElement(group, {
            children: newChildren
          }));
        }
      });
      return groups;
    }

    /** @type {import('react').ReactElement<import('../../index').FloatingMenuItemProps>[]} */
    const newChildren = [];
    React.Children.forEach(children, (/** @type {React.ReactElement} */child, index) => {
      if (fitsSearchValue(child)) {
        newChildren.push(/*#__PURE__*/React.cloneElement(child, {
          index
        }));
      }
    });
    return newChildren;
  }, [children, isGroup, fitsSearchValue]);

  /**
   * @param {HTMLCollection | []} list
   * @param {number} itemSelected
   */
  const hoverElement = (list, itemSelected) => {
    /**
     * @param {HTMLCollection | HTMLUListElement[]} items
     * @param {number} i
     * @param {number} [index]
     */
    const addRemoveFocus = (items, i, index) => {
      if (items[i].classList.contains(listHoverClass)) {
        items[i].classList.remove(listHoverClass);
      }
      if (itemSelected === (index === undefined ? i : index)) {
        items[i].classList.add(listHoverClass);
      }
    };
    if (isGroup) {
      let index = 0;
      for (let j = 0; j < list.length; j += 1) {
        const groupItems = list[j].children[1].children;
        for (let i = 0; i < groupItems.length; i += 1) {
          addRemoveFocus(groupItems, i, index);
          index += 1;
        }
      }
      return;
    }
    for (let i = 0; i < list.length; i += 1) {
      addRemoveFocus(list, i);
    }
  };

  /** @param {import('react').KeyboardEvent} [event] */
  const onSearchFocus = event => {
    setTimeout(() => {
      var _searchBoxRef$current2;
      return (_searchBoxRef$current2 = searchBoxRef.current) === null || _searchBoxRef$current2 === void 0 ? void 0 : _searchBoxRef$current2.focus();
    }, 0);
    const list = isGroup
    // We use ts-ignore because VSCode throws an error that conflicts with the TS linter
    // @ts-ignore: There are issues present with Array.reduce. See: https://github.com/microsoft/TypeScript/issues/44063
    ? options === null || options === void 0 ? void 0 : options.reduce((/** @type {import('react').ReactElement[]} */previousValue, /** @type {import('react').ReactElement} */currentValue) => [...previousValue, ...currentValue.props.children], []) : options || [];
    if (!list) return;
    const listLength = list.length;
    setSearchItemFocused(state => {
      var _menuRef$current, _list$focusIndex;
      let index = list.findIndex((/** @type {{ props: Partial<{ value: string }> }} */_ref) => {
        let props = _ref.props;
        return props.value === state;
      });
      switch (event === null || event === void 0 ? void 0 : event.code) {
        case keyboardCodes.ARROW_DOWN:
          index += 1;
          break;
        case keyboardCodes.ARROW_UP:
          if (index === -1 || index === 0) {
            index = listLength - 1;
            break;
          }
          index -= 1;
          break;
        default:
          index = 0;
          break;
      }
      const focusIndex = index % listLength;
      hoverElement((menuRef === null || menuRef === void 0 || (_menuRef$current = menuRef.current) === null || _menuRef$current === void 0 ? void 0 : _menuRef$current.children) || [], focusIndex);
      return (_list$focusIndex = list[focusIndex]) === null || _list$focusIndex === void 0 || (_list$focusIndex = _list$focusIndex.props) === null || _list$focusIndex === void 0 ? void 0 : _list$focusIndex.value;
    });
  };
  const getFocusedItem = () => searchItemFocused;
  React.useEffect(() => {
    /* Manage custom scroll for search */
    if (isMenuVisible && searchConfig && searchItemFocused !== undefined) {
      const menu = menuRef.current;
      if (menu) {
        var _document;
        const itemToScrollTop = (_document = document) === null || _document === void 0 ? void 0 : _document.getElementById(`${listId}-option-${searchItemFocused}`);

        // If menuMaxHeight is undefined, calculations will result in NaN
        if (itemToScrollTop && menuMaxHeight) {
          const listHeight = menuMaxHeight - SEARCH_BOX_SIZE;
          const itemsToTop = Math.floor(listHeight / LIST_ITEM_SIZE) - 1;
          menu.scrollTop = itemToScrollTop.offsetTop - itemsToTop * LIST_ITEM_SIZE;
        }
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchConfig, menuRef, searchItemFocused, listId, isMenuVisible]);

  /** @param {import('react').KeyboardEvent | import('react').MouseEvent} event */
  const onSearchSelectionChange = event => {
    if (!searchItemFocused) return;
    if ((multiple || allowUnselect) && Array.isArray(currentValues)) {
      if ((currentValues || []).indexOf(searchItemFocused) > -1) {
        handleSelectionChange(event, currentValues.filter(item => item !== searchItemFocused));
        return;
      }
    }
    if (multiple) {
      handleSelectionChange(event, [...(currentValues || []), searchItemFocused]);
      return;
    }
    handleSelectionChange(event, [searchItemFocused]);
  };
  const resetSearchValues = () => {
    var _searchBoxRef$current3;
    const label = searchBoxRef === null || searchBoxRef === void 0 || (_searchBoxRef$current3 = searchBoxRef.current) === null || _searchBoxRef$current3 === void 0 ? void 0 : _searchBoxRef$current3.querySelector('label');
    if (label !== null && label !== void 0 && label.control) {
      /** @type {HTMLElement & { value?: string }} */
      const labelControl = label.control;
      labelControl.value = '';
    }
    setSearchText('');
    if (typeof searchConfig === 'object' && typeof searchConfig.onChange === 'function') searchConfig.onChange('');
  };

  /** @param {import('react').MouseEvent<HTMLButtonElement> | import('react').KeyboardEvent<HTMLButtonElement>} event */
  const openMenu = event => {
    setIsMenuVisible(true);
    if (!hasTopSlot) {
      setTimeout(setPopoverFocus, 0);
    }
    if (typeof onOpen === 'function') onOpen(event);
  };

  /**
   * @param {import('tippy.js').Instance | null} [_]
   * @param {boolean} [preventFocus]
   */
  const closeMenu = (_, preventFocus) => {
    const trigger = triggerRef.current;
    setIsMenuVisible(false);
    if (searchConfig) resetSearchValues();
    if (trigger && !preventFocus) setTimeout(() => trigger.focus(), 0);
    if (typeof onClose === 'function') onClose();
  };

  /** @param {import('tippy.js').Instance | null} _ */
  const handleClickOutside = _ => {
    closeMenu(_, true);
  };

  /** @param {import('react').MouseEvent<HTMLButtonElement>} event */
  const toggleMenu = event => {
    if (isOpenUncontrolled) {
      if (isMenuVisible) {
        closeMenu();
      } else {
        openMenu(event);
      }
      return;
    }
    if (open) {
      if (typeof onClose === 'function') onClose();
    } else if (typeof onOpen === 'function') {
      onOpen(event);
    }
  };
  React.useEffect(() => {
    if (defaultOpen) setTimeout(openMenu, 0);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  React.useEffect(() => {
    if (!isOpenUncontrolled) {
      if (open) {
        setTimeout(openMenu, 0);
      } else if (isMenuVisible) {
        closeMenu();
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [open]);
  const getMenuMaxHeightRelatedStyles = () => {
    if (!menuMaxHeight) return {};
    return {
      maxHeight: menuMaxHeight,
      overflowY: (/** @type {'auto'} */'auto')
    };
  };

  /** @param {import('react').KeyboardEvent<HTMLButtonElement>} event */
  const handleOpenPopup = event => {
    event.preventDefault();
    if (isOpenUncontrolled) {
      openMenu(event);
      return;
    }
    if (typeof onOpen === 'function') onOpen(event);
  };
  const handleClosePopup = () => {
    if (isOpenUncontrolled) {
      closeMenu();
      return;
    }
    if (typeof onClose === 'function') onClose();
  };

  /** @param {import('react').KeyboardEvent<HTMLButtonElement>} event */
  const handleTriggerKeyDown = event => {
    switch (event.code) {
      case keyboardCodes.ESCAPE:
        handleClosePopup();
        break;
      case keyboardCodes.ARROW_DOWN:
      case keyboardCodes.ARROW_UP:
        if (type === 'select') handleOpenPopup(event);
        break;
      case keyboardCodes.ENTER:
      case keyboardCodes.SPACE:
        handleOpenPopup(event);
        break;
    }
  };
  const getValuesArray = () => {
    if (Array.isArray(currentValues)) return currentValues;
    if (typeof currentValues === 'string' || typeof currentValues === 'number') return [currentValues];
    return [];
  };
  const getAriaControls = () => {
    if (type !== 'select') return undefined;
    if (!isEffectiveOpen) return undefined;
    if (role && isSlotted) {
      return `${componentProps === null || componentProps === void 0 ? void 0 : componentProps.id}-dialog`;
    }
    return `${componentProps === null || componentProps === void 0 ? void 0 : componentProps.id}-list`;
  };
  const menuProps = {
    visible: isMenuVisible,
    onSelectionChange: handleSelectionChange,
    options,
    selectedKeys: getValuesArray(),
    style: _objectSpread__default["default"](_objectSpread__default["default"]({}, getMenuMaxHeightRelatedStyles()), {}, {
      width
    }, style),
    popperOptions,
    plugins: [{
      name: 'hideOnEsc',
      defaultValue: true,
      /** @param {{ popper: HTMLDivElement }} _ */
      fn(_ref2) {
        let popper = _ref2.popper;
        function localOnKeyDown(/** @type {KeyboardEvent} */event) {
          if (event.key === keyboardCodes.ESCAPE) {
            handleClosePopup();
          }
        }
        return {
          onShow() {
            popper.addEventListener('keydown', localOnKeyDown);
          },
          onHide() {
            popper.removeEventListener('keydown', localOnKeyDown);
          }
        };
      }
    }],
    offset,
    onClickOutside: isOpenUncontrolled ? handleClickOutside : onClose,
    closeMenu: isOpenUncontrolled ? closeMenu : onClose,
    onSearchFocus: searchConfig ? onSearchFocus : undefined,
    onSearchSelectionChange: searchConfig ? onSearchSelectionChange : undefined,
    onSearchChangeText: searchConfig ? onSearchChangeText : undefined,
    getFocusedItem: searchConfig ? getFocusedItem : undefined,
    onMount: searchConfig ? componentProps.onMount : undefined
  };
  const triggerProps = {
    role,
    'aria-expanded': isMenuVisible,
    'aria-controls': getAriaControls(),
    'aria-haspopup': ariaHasPopupValue,
    onClick: toggleMenu,
    onKeyDown: handleTriggerKeyDown
  };
  return {
    menuProps,
    triggerProps
  };
};
var useFloatingMenu$1 = useFloatingMenu;
const namespace = 'andes-floating-menu';
const namespaceMenuMultiple = `${namespace}--multiple`;
const formNamespace = 'andes-form-control';
const namespaceMenuSearch = `${namespace}--search`;
const WARNING_DIALOG_SR_LABEL = `<FloatingMenu> needs some plain text content for the dialog element for accessibility reasons.
Please add a 'dialogSrLabel' prop.`;
const WARNING_SEARCHBOX_DISABLED = `If you are using the 'disabled' property to temporarily block the SearchBox, it is recommended to use loading instead.
The disabled state will be removed in future versions.`;
var enTranslations = {
  "andes-floating-menu\u0004No hay resultados": [null, "No results"],
  "andes-floating-menu\u0004Abrir buscador": [null, "Open search tool"],
  "andes-floating-menu\u0004Buscar": [null, "Search"],
  "andes-floating-menu\u0004Borrar": [null, "Delete search"],
  "andes-floating-menu\u0004Cerrar men├║": [null, "Close menu"]
};
var esArTranslations = {
  "andes-floating-menu\u0004No hay resultados": [null, "No hay resultados"],
  "andes-floating-menu\u0004Abrir buscador": [null, "Abrir buscador"],
  "andes-floating-menu\u0004Buscar": [null, "Buscar"],
  "andes-floating-menu\u0004Borrar": [null, "Borrar"],
  "andes-floating-menu\u0004Cerrar men├║": [null, "Cerrar men├║"]
};
var ptBrTranslations = {
  "andes-floating-menu\u0004No hay resultados": [null, "Sem resultados"],
  "andes-floating-menu\u0004Abrir buscador": [null, "Abrir mecanismo de busca"],
  "andes-floating-menu\u0004Buscar": [null, "Procurar"],
  "andes-floating-menu\u0004Borrar": [null, "Excluir"],
  "andes-floating-menu\u0004Cerrar men├║": [null, "Fechar menu"]
};
var zhCNTranslations = {
  "andes-floating-menu\u0004No hay resultados": [null, "µŚĀń╗ōµ×£"],
  "andes-floating-menu\u0004Abrir buscador": [null, "µēōÕ╝ĆµÉ£ń┤óÕĘźÕģĘ"],
  "andes-floating-menu\u0004Buscar": [null, "µÉ£ń┤ó"],
  "andes-floating-menu\u0004Borrar": [null, "ÕłĀķÖżµÉ£ń┤ó"],
  "andes-floating-menu\u0004Cerrar men├║": [null, "Õģ│ķŚŁĶÅ£ÕŹĢ"]
};

/**
 * This types do not pick from @andes/search-box to avoid circular dependencies
 * @typedef {Object} SearchboxProps
 * @prop {string} [className]
 * @prop {string} [defaultValue]
 * @prop {React.InputHTMLAttributes<HTMLInputElement>} [inputProps]
 * @prop {(event: React.ChangeEvent<HTMLInputElement>) => void} [onChange]
 * @prop {(event: React.KeyboardEvent<HTMLButtonElement> | React.MouseEvent<HTMLButtonElement>) => void} [onClear]
 * @prop {(event: React.KeyboardEvent<HTMLInputElement>) => void} [onKeyDown]
 * @prop {string} [placeholder]
 * @prop {'large' | 'small'} [size]
 * @prop {string} [srLabel]
 * @prop {string} [value]
 * @prop {React.RefObject<HTMLDivElement | null> | ((refElement: HTMLDivElement | null) => void)} [ref]
 */

/** @type {React.FunctionComponent<SearchboxProps>} */
const SearchBox = _ref => {
  let className = _ref.className,
    defaultValue = _ref.defaultValue,
    _ref$inputProps = _ref.inputProps,
    inputProps = _ref$inputProps === void 0 ? {} : _ref$inputProps,
    onChange = _ref.onChange,
    onClear = _ref.onClear,
    onKeyDown = _ref.onKeyDown,
    placeholder = _ref.placeholder,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'small' : _ref$size,
    srLabel = _ref.srLabel,
    value = _ref.value,
    ref = _ref.ref;
  const _useAndesContext = context.useAndesContext('@andes/floating-menu', {
      en: enTranslations,
      'es-AR': esArTranslations,
      'pt-BR': ptBrTranslations,
      'zh-CN': zhCNTranslations
    }),
    device = _useAndesContext.device,
    i18n = _useAndesContext.i18n;
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;

  /** @type {React.RefObject<HTMLInputElement | null>} */
  const inputRef = React.useRef(null);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const isControlled = React.useMemo(() => typeof value !== 'undefined', []);
  const _useState = React.useState(false),
    _useState2 = _slicedToArray__default["default"](_useState, 2),
    focused = _useState2[0],
    setFocused = _useState2[1];
  const _useState3 = React.useState(false),
    _useState4 = _slicedToArray__default["default"](_useState3, 2),
    showClear = _useState4[0],
    setShowClear = _useState4[1];
  const firstRender = React.useRef(true);
  const isWebMobileIOS = (device === null || device === void 0 ? void 0 : device.osName) === 'ios';
  warning__default["default"](!(inputProps !== null && inputProps !== void 0 && inputProps.disabled), WARNING_SEARCHBOX_DISABLED);
  React.useEffect(() => {
    if (!isControlled && defaultValue) {
      setShowClear(true);
    }
    firstRender.current = false;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  React.useEffect(() => {
    if (isControlled) {
      const showClearButton = !!value && (value === null || value === void 0 ? void 0 : value.length) > 0;
      setShowClear(showClearButton);
    }
  }, [isControlled, value, setShowClear]);
  React.useEffect(() => {
    if (focused) {
      var _inputRef$current;
      (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 || _inputRef$current.focus();
    }
  }, [focused]);
  const SearchIcon = () => {
    if (size === 'small') {
      return /*#__PURE__*/React__default["default"].createElement(Search16__default["default"], {
        color: "currentColor"
      });
    }
    return /*#__PURE__*/React__default["default"].createElement(Search20__default["default"], {
      color: "currentColor"
    });
  };
  const CloseIcon = () => {
    if (size === 'small') {
      return /*#__PURE__*/React__default["default"].createElement(Close16__default["default"], {
        color: "currentColor"
      });
    }
    return /*#__PURE__*/React__default["default"].createElement(Close20__default["default"], {
      color: "currentColor"
    });
  };

  /** @param {React.ChangeEvent<HTMLInputElement>} event */
  const handleChange = event => {
    var _inputRef$current2, _inputRef$current3;
    const showClearButton = !!(inputRef !== null && inputRef !== void 0 && (_inputRef$current2 = inputRef.current) !== null && _inputRef$current2 !== void 0 && _inputRef$current2.value) && (inputRef === null || inputRef === void 0 || (_inputRef$current3 = inputRef.current) === null || _inputRef$current3 === void 0 ? void 0 : _inputRef$current3.value.length) > 0;
    setShowClear(showClearButton);
    if (typeof onChange === 'function') onChange(event);
  };

  /** @param {React.KeyboardEvent<HTMLButtonElement> | React.MouseEvent<HTMLButtonElement>} event */
  const handleClear = event => {
    var _inputRef$current4;
    event.preventDefault();
    if (!isControlled && inputRef.current) inputRef.current.value = '';

    // @ts-expect-error: We add value
    event.target.value = '';
    setShowClear(false);
    if (typeof onClear === 'function') onClear(event);
    // @ts-expect-error: We modify event
    if (typeof onChange === 'function') onChange(event);
    (_inputRef$current4 = inputRef.current) === null || _inputRef$current4 === void 0 || _inputRef$current4.focus();
  };

  /** @param {React.FocusEvent<HTMLInputElement>} event */
  const handleBlur = event => {
    const onBlur = inputProps.onBlur;
    setFocused(false);
    if (typeof onBlur === 'function') onBlur(event);
  };

  /** @param {React.FocusEvent<HTMLInputElement>} event */
  const handleFocus = event => {
    const onFocus = inputProps.onFocus;
    setFocused(true);
    if (typeof onFocus === 'function') onFocus(event);
  };
  const dismissClassNames = classNames__default["default"](`${formNamespace}__close-icon`, {
    [`${formNamespace}__close-icon-visible`]: showClear
  });
  const wrapperClassNames = classNames__default["default"](namespaceMenuSearch, `${namespaceMenuSearch}--size-${size}`, className, {
    [`${namespaceMenuSearch}--web-mobile-ios`]: isWebMobileIOS,
    [`${namespaceMenuSearch}--focused`]: focused
  });
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    ref: ref,
    className: getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('searchbox')
  }, themeHash), /*#__PURE__*/React__default["default"].createElement(focus.FocusRing, {
    within: true,
    isTextInput: true,
    focusRingClass: "focus-visible"
  }, /*#__PURE__*/React__default["default"].createElement("div", {
    className: wrapperClassNames,
    "data-andes-searchbox": true,
    "data-andes-searchbox-variant": "simple",
    "data-andes-searchbox-size": size
  }, /*#__PURE__*/React__default["default"].createElement("label", null, /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHidden, null, srLabel || i18n.pgettext(namespace, 'Buscar')), /*#__PURE__*/React__default["default"].createElement("input", _extends__default["default"]({}, inputProps, {
    className: `${formNamespace}__field`,
    value: value,
    defaultValue: defaultValue,
    placeholder: placeholder || `${i18n.pgettext(namespace, 'Buscar')}`,
    onChange: handleChange,
    onBlur: handleBlur,
    onFocus: handleFocus,
    ref: inputRef,
    onKeyDown: onKeyDown,
    "data-andes-searchbox-input": true
  }))), /*#__PURE__*/React__default["default"].createElement("button", {
    className: dismissClassNames,
    onClick: handleClear,
    onMouseDown: handleClear,
    "aria-label": `${i18n.pgettext(namespace, 'Borrar')}`,
    type: "button",
    "data-andes-searchbox-close": true
  }, /*#__PURE__*/React__default["default"].createElement(CloseIcon, null)), /*#__PURE__*/React__default["default"].createElement("div", {
    className: `${formNamespace}__search-icon`,
    "data-andes-searchbox-search": true
  }, /*#__PURE__*/React__default["default"].createElement(SearchIcon, null)))));
};
SearchBox.propTypes = {
  className: PropTypes__default["default"].string,
  defaultValue: PropTypes__default["default"].string,
  inputProps: PropTypes__default["default"].shape({}),
  onChange: PropTypes__default["default"].func,
  onClear: PropTypes__default["default"].func,
  onKeyDown: PropTypes__default["default"].func,
  placeholder: PropTypes__default["default"].string,
  size: PropTypes__default["default"].oneOf(['small', 'large']),
  srLabel: PropTypes__default["default"].string,
  value: PropTypes__default["default"].string
};
SearchBox.displayName = 'SearchBox';
var SearchBox$1 = SearchBox;

/**
 * @typedef {Object} FloatingMenuListDialogContentProps
 * @prop {React.ReactNode} children
 * @prop {boolean} visible
 * @prop {string} id
 * @prop {string} [dialogSrLabel]
 * @prop {(() => void) | undefined} [closeMenu] - Function to close the menu
 */

/** @type {React.FunctionComponent<FloatingMenuListDialogContentProps>} */
// eslint-disable-next-line react/prop-types
const FloatingMenuListDialogContent = _ref => {
  let visible = _ref.visible,
    children = _ref.children,
    id = _ref.id,
    dialogSrLabel = _ref.dialogSrLabel,
    closeMenu = _ref.closeMenu;
  warning__default["default"](dialogSrLabel, WARNING_DIALOG_SR_LABEL);

  // Detect Android mobile to avoid issues with Talkback when focus is trapped
  const _useAndesContext = context.useAndesContext('@andes/floating-menu', {
      en: enTranslations,
      'es-AR': esArTranslations,
      'pt-BR': ptBrTranslations,
      'zh-CN': zhCNTranslations
    }),
    device = _useAndesContext.device,
    i18n = _useAndesContext.i18n;
  const shouldDisableFocusTrap = (device === null || device === void 0 ? void 0 : device.osName) === 'android' && (device === null || device === void 0 ? void 0 : device.mobile);

  // Accessible close button as first element inside the focus trap for Android mobile
  const hiddenCloseButton = /*#__PURE__*/React__default["default"].createElement(visuallyHidden.VisuallyHiddenFocusable, {
    onClick: closeMenu,
    type: "button"
  }, i18n.pgettext(namespace, 'Cerrar men├║'));
  return /*#__PURE__*/React__default["default"].createElement("div", {
    role: "dialog",
    "aria-modal": "true",
    "aria-label": dialogSrLabel,
    id: `${id}-dialog`
  }, /*#__PURE__*/React__default["default"].createElement(focus.FocusScope, {
    contain: !shouldDisableFocusTrap && visible,
    restoreFocus: true,
    autoFocus: true
  }, shouldDisableFocusTrap && hiddenCloseButton, children));
};
FloatingMenuListDialogContent.propTypes = {
  children: PropTypes__default["default"].node.isRequired,
  closeMenu: PropTypes__default["default"].func,
  dialogSrLabel: PropTypes__default["default"].string,
  id: PropTypes__default["default"].string.isRequired,
  visible: PropTypes__default["default"].bool.isRequired
};

/**
 * @typedef {Object} FloatingMenuListSlottedProps
 * @prop {React.ReactNode} children
 * @prop {React.ReactNode} [topContent]
 * @prop {React.ReactNode} [bottomContent]
 * @prop {boolean} visible
 * @prop {boolean} [dialog]
 * @prop {string} id
 * @prop {string} [dialogSrLabel]
 * @prop {(() => void) | undefined} [closeMenu] - Function to close the menu
 */

/** @type {React.FunctionComponent<FloatingMenuListSlottedProps>} */
const FloatingMenuListSlotted = _ref2 => {
  let children = _ref2.children,
    visible = _ref2.visible,
    topContent = _ref2.topContent,
    bottomContent = _ref2.bottomContent,
    dialog = _ref2.dialog,
    dialogSrLabel = _ref2.dialogSrLabel,
    id = _ref2.id,
    closeMenu = _ref2.closeMenu;
  const hasTopSlot = typeof topContent !== 'undefined';
  const hasBottomSlot = typeof bottomContent !== 'undefined';
  const isSlotted = hasTopSlot || hasBottomSlot;
  const content = /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, topContent, children, bottomContent);
  if (dialog === false) {
    return content;
  }
  if (dialog || isSlotted) {
    if (hasTopSlot) {
      // we need to render it conditionally so focusScope can focus the first element
      return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, visible && /*#__PURE__*/React__default["default"].createElement(FloatingMenuListDialogContent, {
        visible: visible,
        id: id,
        dialogSrLabel: dialogSrLabel,
        closeMenu: closeMenu
      }, content));
    }
    return /*#__PURE__*/React__default["default"].createElement(FloatingMenuListDialogContent, {
      visible: visible,
      id: id,
      dialogSrLabel: dialogSrLabel,
      closeMenu: closeMenu
    }, content);
  }
  return content;
};
var FloatingMenuListSlotted$1 = FloatingMenuListSlotted;
FloatingMenuListSlotted.propTypes = {
  bottomContent: PropTypes__default["default"].node,
  children: PropTypes__default["default"].node.isRequired,
  closeMenu: PropTypes__default["default"].func,
  dialog: PropTypes__default["default"].bool,
  dialogSrLabel: PropTypes__default["default"].string,
  id: PropTypes__default["default"].string,
  topContent: PropTypes__default["default"].node,
  visible: PropTypes__default["default"].bool.isRequired
};

/**
 * @typedef {Object} FloatingMenuSearchProps
 * @prop {import('../..').SearchConfigProps['defaultValue']} [defaultValue]
 * @prop {() => (string | undefined)} [getFocusedItem]
 * @prop {(event: React.KeyboardEvent) => void} [handleKeyDown]
 * @prop {React.ReactNode} list
 * @prop {string} listId
 * @prop {(event: React.ChangeEvent<HTMLInputElement>) => void} [onChange]
 * @prop {import('../..').SearchConfigProps['placeholder']} [placeholder]
 * @prop {boolean} visible
 * @prop {string} id
 * @prop {string} [dialogSrLabel]
 */

/**
 * @type {React.FunctionComponent<
 * FloatingMenuSearchProps & import('@andes/common').WithRef<HTMLDivElement>
 * >}
 */
const FloatingMenuSearch = _ref => {
  let defaultValue = _ref.defaultValue,
    getFocusedItem = _ref.getFocusedItem,
    handleKeyDown = _ref.handleKeyDown,
    list = _ref.list,
    listId = _ref.listId,
    onChange = _ref.onChange,
    placeholder = _ref.placeholder,
    visible = _ref.visible,
    dialogSrLabel = _ref.dialogSrLabel,
    id = _ref.id,
    ref = _ref.ref;
  const inputProps = {
    role: 'combobox',
    'aria-label': placeholder,
    'aria-autocomplete': 'list',
    'aria-controls': listId,
    // @ts-expect-error: In a search context, getFocusedItem will always be defined
    'aria-activedescendant': getFocusedItem() ? `${listId}-option-${getFocusedItem()}` : undefined,
    'aria-expanded': true
  };
  return /*#__PURE__*/React__default["default"].createElement(FloatingMenuListSlotted$1, {
    id: id,
    dialogSrLabel: dialogSrLabel,
    visible: visible,
    topContent: /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${namespace}__search-box`
    }, /*#__PURE__*/React__default["default"].createElement(SearchBox$1, {
      ref: ref,
      defaultValue: defaultValue,
      onChange: onChange,
      placeholder: placeholder,
      onKeyDown: handleKeyDown
      // @ts-expect-error: JSDoc does not recognize 'list' as string in 'aria-autocomplete'
      ,

      inputProps: inputProps
    }))
  }, list);
};
FloatingMenuSearch.propTypes = {
  defaultValue: PropTypes__default["default"].string,
  dialogSrLabel: PropTypes__default["default"].string,
  getFocusedItem: PropTypes__default["default"].func,
  handleKeyDown: PropTypes__default["default"].func,
  id: PropTypes__default["default"].string,
  list: PropTypes__default["default"].node.isRequired,
  listId: PropTypes__default["default"].string.isRequired,
  onChange: PropTypes__default["default"].func,
  placeholder: PropTypes__default["default"].string,
  visible: PropTypes__default["default"].bool.isRequired
};
FloatingMenuSearch.displayName = 'FloatingMenuSearch';
var FloatingMenuSearch$1 = FloatingMenuSearch;
const _excluded$3 = ["children", "className", "defaultValue", "onSelect", "value", "ref"];

/**
 * @typedef {Object} FloatingMenuMultipleProps
 * @prop {React.ReactElement | React.ReactElement[]} children
 * @prop {string} [className]
 * @prop {string[]} [defaultValue]
 * @prop {string} id
 * @prop {(
 * event: React.KeyboardEvent | React.MouseEvent,
 * values: string[],
 * itemDetails?: { value?: string
 * }) => void} [onSelect]
 * @prop {string} srLabel
 * @prop {string[]} [value]
 * @prop {React.RefObject<HTMLUListElement | null> | ((refElement: HTMLUListElement | null) => void)} [ref]
 */

/**
 * @typedef {Object} FloatingMenuMultipleInternalProps
 * @prop {boolean} [shouldUseVirtualFocus]
 */

/** @type {React.FunctionComponent<FloatingMenuMultipleProps & FloatingMenuMultipleInternalProps>} */
const FloatingMenuMultiple = _ref => {
  let children = _ref.children,
    className = _ref.className,
    defaultValue = _ref.defaultValue,
    onSelect = _ref.onSelect,
    value = _ref.value,
    ref = _ref.ref,
    listProps = _objectWithoutProperties__default["default"](_ref, _excluded$3);
  const _useState = React.useState(value || defaultValue),
    _useState2 = _slicedToArray__default["default"](_useState, 2),
    currentValues = _useState2[0],
    setCurrentValues = _useState2[1];
  const isUncontrolled = React.useMemo(() => value === undefined, []);
  React.useEffect(() => {
    if (!isUncontrolled) {
      setCurrentValues(value);
    }
  }, [value]);

  /**
   * @param {React.KeyboardEvent | React.MouseEvent} event
   * @param {string[]} selectedValues
   * @param {{ value?: string }} itemDetails
   */
  const customOnSelect = (event, selectedValues, itemDetails) => {
    if (isUncontrolled) setCurrentValues(selectedValues);
    if (typeof onSelect === 'function') onSelect(event, selectedValues, itemDetails);
  };
  const getValuesArray = () => Array.isArray(currentValues) ? currentValues : [];
  const stateProps = {};
  if (!isUncontrolled) {
    stateProps.selectedKeys = getValuesArray();
  } else if (defaultValue) {
    stateProps.defaultSelectedKeys = getValuesArray();
  }
  const floatingMenuMultipleClassNames = classNames__default["default"](namespaceMenuMultiple, className);
  return /*#__PURE__*/React__default["default"].createElement(listbox.ListBox, _extends__default["default"]({
    className: floatingMenuMultipleClassNames,
    selectionComponent: /*#__PURE__*/React__default["default"].createElement(checkbox.Checkbox, {
      srHidden: true
    }),
    selectionMode: "multiple",
    ref: ref,
    onSelect: customOnSelect
  }, listProps, stateProps), children);
};
FloatingMenuMultiple.propTypes = {
  children: PropTypes__default["default"].oneOfType([PropTypes__default["default"].element, PropTypes__default["default"].arrayOf(PropTypes__default["default"].element.isRequired)]).isRequired,
  className: PropTypes__default["default"].string,
  defaultValue: PropTypes__default["default"].arrayOf(PropTypes__default["default"].string.isRequired),
  id: PropTypes__default["default"].string.isRequired,
  onSelect: PropTypes__default["default"].func,
  value: PropTypes__default["default"].arrayOf(PropTypes__default["default"].string.isRequired)
};
FloatingMenuMultiple.displayName = 'FloatingMenuMultiple';
var FloatingMenuMultiple$1 = FloatingMenuMultiple;

/**
 * @typedef {Object} FloatingMenuListProps
 * @prop {(() => void) | undefined} [closeMenu] - Function to close the menu
 * @prop {import('@andes/common').BasicI18nProps} i18n
 * @prop {string} id
 * @prop {boolean} [multiple]
 * @prop {(event: React.KeyboardEvent) => void} onKeyDown
 * @prop {(
 * event: React.KeyboardEvent | React.MouseEvent,
 * values: string | string[],
 * itemDetails?: { value?: string }
 * ) => void} onSelect
 * @prop {React.ReactElement[] | null} options
 * @prop {boolean} [search]
 * @prop {string[]} [selectedKeys]
 * @prop {string} srLabel
 * @prop {'menu' | 'select'} [type]
 * @prop {boolean} visible
 * @prop {React.ReactNode} [topContent]
 * @prop {React.ReactNode} [bottomContent]
 * @prop {string} [dialogSrLabel]
 */

/** @type {React.FunctionComponent<FloatingMenuListProps & import('@andes/common').WithRef<HTMLUListElement>>} */
const FloatingMenuList = _ref => {
  let closeMenu = _ref.closeMenu,
    bottomContent = _ref.bottomContent,
    i18n = _ref.i18n,
    id = _ref.id,
    _ref$multiple = _ref.multiple,
    multiple = _ref$multiple === void 0 ? false : _ref$multiple,
    onKeyDown = _ref.onKeyDown,
    onSelect = _ref.onSelect,
    options = _ref.options,
    _ref$search = _ref.search,
    search = _ref$search === void 0 ? false : _ref$search,
    selectedKeys = _ref.selectedKeys,
    srLabel = _ref.srLabel,
    topContent = _ref.topContent,
    _ref$type = _ref.type,
    type = _ref$type === void 0 ? 'menu' : _ref$type,
    visible = _ref.visible,
    dialogSrLabel = _ref.dialogSrLabel,
    ref = _ref.ref;
  if (!options || !options.length) {
    return /*#__PURE__*/React__default["default"].createElement(FloatingMenuListSlotted$1, {
      visible: visible,
      topContent: topContent,
      bottomContent: bottomContent,
      dialog: !search,
      id: id,
      dialogSrLabel: dialogSrLabel
    }, /*#__PURE__*/React__default["default"].createElement("div", {
      className: `${namespace}__no-results`,
      "data-andes-floating-menu-no-result": true
    }, i18n.pgettext(namespace, 'No hay resultados')));
  }
  const commonProps = {
    id,
    ref,
    className: namespace,
    onKeyDown,
    srLabel
  };
  if (multiple) {
    return /*#__PURE__*/React__default["default"].createElement(FloatingMenuListSlotted$1, {
      visible: visible,
      topContent: topContent,
      bottomContent: bottomContent,
      dialog: !search,
      id: id,
      dialogSrLabel: dialogSrLabel
    }, /*#__PURE__*/React__default["default"].createElement(FloatingMenuMultiple$1, _extends__default["default"]({}, commonProps, {
      onSelect: onSelect,
      value: selectedKeys,
      shouldUseVirtualFocus: !!search
    }), options));
  }
  if (type === 'select' || search) {
    const isSlotted = typeof topContent !== 'undefined' || typeof bottomContent !== 'undefined';
    return /*#__PURE__*/React__default["default"].createElement(FloatingMenuListSlotted$1, {
      visible: visible,
      topContent: topContent,
      bottomContent: bottomContent,
      dialog: isSlotted && !search,
      id: id,
      dialogSrLabel: dialogSrLabel
    }, /*#__PURE__*/React__default["default"].createElement(listbox.ListBox, _extends__default["default"]({}, commonProps, {
      onSelect: onSelect,
      selectionMode: multiple ? 'multiple' : 'single',
      selectedKeys: selectedKeys,
      shouldUseVirtualFocus: !!search
    }), options));
  }

  /**
   * @param {(evt: React.MouseEvent, value: string | string[]) => void} childOnSelect
   * @param {React.MouseEvent} event
   * @param {string | string[]} value
   */
  const handleMenuClick = (childOnSelect, event, value) => {
    if (typeof childOnSelect === 'function') childOnSelect(event, value);
    if (typeof onSelect === 'function') onSelect(event, value);
  };

  /** @param {React.ReactElement} item */
  const getItem = item => /*#__PURE__*/React.cloneElement(item, {
    /**
     * @param {React.MouseEvent} evt
     * @param {string | string[]} value
     */
    onClick: (evt, value) => handleMenuClick(item.props.onClick, evt, value),
    tabIndex: undefined
  });
  return /*#__PURE__*/React__default["default"].createElement(FloatingMenuListSlotted$1, {
    visible: visible,
    topContent: topContent,
    bottomContent: bottomContent,
    dialog: true,
    id: id,
    dialogSrLabel: dialogSrLabel,
    closeMenu: closeMenu
  }, /*#__PURE__*/React__default["default"].createElement(list.List, commonProps, React__default["default"].Children.map(options, child => {
    var _child$type;
    // @ts-expect-error: we add displayName
    const isGroup = (child === null || child === void 0 || (_child$type = child.type) === null || _child$type === void 0 ? void 0 : _child$type.displayName) === GROUP_DISPLAY_NAME;
    if (isGroup) {
      const items = React__default["default"].Children.map(child.props.children, getItem);
      return /*#__PURE__*/React.cloneElement(child, {
        children: items
      });
    }
    return getItem(child);
  })));
};
FloatingMenuList.propTypes = {
  closeMenu: PropTypes__default["default"].func,
  bottomContent: PropTypes__default["default"].node,
  dialogSrLabel: PropTypes__default["default"].string,
  i18n: PropTypes__default["default"].shape({
    pgettext: PropTypes__default["default"].func.isRequired
  }).isRequired,
  id: PropTypes__default["default"].string.isRequired,
  multiple: PropTypes__default["default"].bool,
  onKeyDown: PropTypes__default["default"].func.isRequired,
  onSelect: PropTypes__default["default"].func.isRequired,
  options: PropTypes__default["default"].arrayOf(PropTypes__default["default"].element.isRequired).isRequired,
  search: PropTypes__default["default"].bool,
  selectedKeys: PropTypes__default["default"].arrayOf(PropTypes__default["default"].string.isRequired),
  srLabel: PropTypes__default["default"].string.isRequired,
  type: PropTypes__default["default"].oneOf(['menu', 'select']),
  topContent: PropTypes__default["default"].node,
  visible: PropTypes__default["default"].bool.isRequired
};
FloatingMenuList.displayName = 'FloatingMenuList';
var FloatingMenuList$1 = FloatingMenuList;

/**
 * @typedef {Object} ComputeSearchProps
 * @prop {import('../..').SearchConfigProps | boolean} search
 */

/** @param {ComputeSearchProps} computeSearchProps */
const computeSearchProps = _ref => {
  let searchProps = _ref.search;
  if (!searchProps) return undefined;
  if (typeof searchProps === 'boolean') {
    return {
      search: true
    };
  }
  return _objectSpread__default["default"](_objectSpread__default["default"]({}, searchProps), {}, {
    search: true
  });
};
var computeSearchProps$1 = computeSearchProps;
var _metrics$sendSlots;
const METRIC_COMPONENT = 'Floating-Menu';

/**
 * @param {Partial<import('../..').FloatingMenuProps>} componentProps
 */
const isSlotted = _ref => {
  let topContent = _ref.topContent,
    bottomContent = _ref.bottomContent,
    search = _ref.search;
  return typeof topContent !== 'undefined' && !search || typeof bottomContent !== 'undefined';
};
const metrics = metrics$1.senderMetrics(metrics$1.trackCustomization);
const FloatingMenuMetrics = {
  SSR: [metrics === null || metrics === void 0 || (_metrics$sendSlots = metrics.sendSlots) === null || _metrics$sendSlots === void 0 ? void 0 : _metrics$sendSlots.call(metrics, METRIC_COMPONENT, isSlotted)]
};
const _excluded$2 = ["ref"],
  _excluded2 = ["closeMenu", "getFocusedItem", "onSearchChangeText", "onSearchFocus", "onSearchSelectionChange", "onSelectionChange", "options", "selectedKeys"];

/**
 * @typedef {Object} FloatingMenuInternalProps
 * @prop {() => HTMLElement} [appendTo]
*/

/** @type {React.FunctionComponent<import('../../index').FloatingMenuProps & FloatingMenuInternalProps>} */
const FloatingMenu = _ref => {
  let ref = _ref.ref,
    props = _objectWithoutProperties__default["default"](_ref, _excluded$2);
  const _props$allowUnselect = props.allowUnselect,
    allowUnselect = _props$allowUnselect === void 0 ? false : _props$allowUnselect,
    appendTo = props.appendTo,
    bottomContent = props.bottomContent,
    className = props.className,
    _props$defaultOpen = props.defaultOpen,
    defaultOpen = _props$defaultOpen === void 0 ? false : _props$defaultOpen,
    id = props.id,
    menuMaxHeight = props.menuMaxHeight,
    _props$multiple = props.multiple,
    multiple = _props$multiple === void 0 ? false : _props$multiple,
    _props$placement = props.placement,
    placement = _props$placement === void 0 ? 'bottom' : _props$placement,
    _props$search = props.search,
    search = _props$search === void 0 ? false : _props$search,
    srLabel = props.srLabel,
    topContent = props.topContent,
    trigger = props.trigger,
    _props$type = props.type,
    type = _props$type === void 0 ? 'menu' : _props$type,
    _props$width = props.width,
    width = _props$width === void 0 ? 280 : _props$width,
    dialogSrLabel = props.dialogSrLabel;
  const componentId = context.useId(id);
  const _useThemeContext = themes.useThemeContext(),
    getThemedComponentClasses = _useThemeContext.getThemedComponentClasses,
    themeHash = _useThemeContext.themeHash;

  // TODO: Analyze if we can find a better solution for this
  // Since popper renders the list conditionally, we need to register the list component on SSR
  getThemedComponentClasses === null || getThemedComponentClasses === void 0 || getThemedComponentClasses('list');
  const searchConfig = computeSearchProps$1({
    search
  });
  const _useAndesContext = context.useAndesContext('@andes/floating-menu', {
      en: enTranslations,
      'es-AR': esArTranslations,
      'pt-BR': ptBrTranslations,
      'zh-CN': zhCNTranslations
    }),
    i18n = _useAndesContext.i18n;
  const listId = `${componentId}-list`;
  const triggerRef = React.useRef(null);
  /** @type {React.RefObject<HTMLUListElement | null>} */
  const menuRef = React.useRef(null);
  const searchBoxRef = React.useRef(null);
  const hasSearched = React.useRef(false);
  const _useFloatingMenu = useFloatingMenu$1(triggerRef, menuRef, searchBoxRef, _objectSpread__default["default"](_objectSpread__default["default"]({}, props), {}, {
      allowUnselect,
      defaultOpen,
      multiple,
      placement,
      type,
      width,
      listId,
      id: componentId,
      searchConfig
    })),
    menuProps = _useFloatingMenu.menuProps,
    triggerProps = _useFloatingMenu.triggerProps;
  const closeMenu = menuProps.closeMenu,
    getFocusedItem = menuProps.getFocusedItem,
    onSearchChangeText = menuProps.onSearchChangeText,
    onSearchFocus = menuProps.onSearchFocus,
    onSearchSelectionChange = menuProps.onSearchSelectionChange,
    onSelectionChange = menuProps.onSelectionChange,
    options = menuProps.options,
    selectedKeys = menuProps.selectedKeys,
    otherMenuProps = _objectWithoutProperties__default["default"](menuProps, _excluded2);
  React.useEffect(() => {
    if (hasSearched.current) {
      hasSearched.current = false;
      if (typeof onSearchFocus === 'function') onSearchFocus();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [hasSearched.current]);
  React.useEffect(() => {
    const menu = menuRef.current;
    if (menu && searchConfig) {
      requestAnimationFrame(() => {
        menu.style.maxHeight = menuMaxHeight ? `${menuMaxHeight - SEARCH_BOX_SIZE}px` : '';
        menu.style.overflowY = 'auto';
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [menuRef.current, menuMaxHeight, searchConfig]);

  /**
   * @param {React.KeyboardEvent | React.MouseEvent} event
   * @param {string | string[]} [values]
   */
  const handleSelectionChange = (event, values) => {
    if (values) {
      onSelectionChange(event, values);
    } else if (typeof onSearchSelectionChange === 'function') {
      onSearchSelectionChange(event);
    }
    if (!multiple && typeof closeMenu === 'function') closeMenu();
  };

  /** @param {React.KeyboardEvent} event */
  const handleKeyDown = event => {
    // Simulates native select behaviour on type select
    switch (event.code) {
      case keyboardCodes.TAB:
        if (type !== 'menu') event.preventDefault();
        break;
      case keyboardCodes.ARROW_UP:
      case keyboardCodes.ARROW_DOWN:
        // @ts-expect-error: In a search environment, this is defined
        if (searchConfig) onSearchFocus(event);
        break;
      case keyboardCodes.ENTER:
        if (searchConfig) handleSelectionChange(event);
        break;
    }
  };

  /** @param {React.ChangeEvent<HTMLInputElement>} event */
  const handleSearchChange = event => {
    const value = event.target.value;
    hasSearched.current = true;
    if (typeof onSearchChangeText === 'function') onSearchChangeText(value);
    if (typeof (searchConfig === null || searchConfig === void 0 ? void 0 : searchConfig.onChange) === 'function') searchConfig.onChange(value);
  };
  const getContent = () => {
    const list = /*#__PURE__*/React__default["default"].createElement(FloatingMenuList$1, {
      closeMenu: closeMenu,
      id: listId,
      multiple: multiple,
      onKeyDown: handleKeyDown,
      onSelect: handleSelectionChange,
      options: options,
      ref: menuRef,
      selectedKeys: selectedKeys,
      type: type,
      search: searchConfig === null || searchConfig === void 0 ? void 0 : searchConfig.search,
      srLabel: srLabel,
      visible: menuProps.visible,
      i18n: i18n,
      topContent: !searchConfig ? topContent : undefined,
      bottomContent: bottomContent,
      dialogSrLabel: dialogSrLabel
    });
    if (!searchConfig) return list;
    return /*#__PURE__*/React__default["default"].createElement(FloatingMenuSearch$1, _extends__default["default"]({
      getFocusedItem: getFocusedItem,
      handleKeyDown: handleKeyDown,
      list: list,
      listId: listId,
      ref: searchBoxRef,
      visible: menuProps.visible,
      id: componentId
    }, searchConfig, {
      onChange: handleSearchChange,
      dialogSrLabel: dialogSrLabel
    }));
  };
  const containerClassNames = classNames__default["default"](namespace, className, {
    [`${namespace}--multiple`]: multiple
  });
  const contentClassName = classNames__default["default"](`${namespace}__popper-content`, getThemedComponentClasses === null || getThemedComponentClasses === void 0 ? void 0 : getThemedComponentClasses('floatingMenu'));
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: containerClassNames,
    ref: ref
  }, /*#__PURE__*/React__default["default"].createElement(popper.Popper, _extends__default["default"]({
    appendTo: appendTo,
    content: getContent(),
    interactive: true,
    placement: placement,
    id: `${componentId}-popper`,
    focusScope: true
    // @ts-expect-error: internal props
    ,

    classNames: {
      content: contentClassName
    },
    attrs: {
      root: themeHash,
      content: _objectSpread__default["default"]({
        'data-andes-floating-menu': true
      }, themeHash)
    }
  }, otherMenuProps), /*#__PURE__*/React.cloneElement(/** @type {React.ReactElement} */trigger, _objectSpread__default["default"]({
    ref: triggerRef
  }, triggerProps))));
};
FloatingMenu.displayName = 'FloatingMenu';
FloatingMenu.propTypes = {
  allowUnselect: PropTypes__default["default"].bool,
  appendTo: PropTypes__default["default"].func,
  bottomContent: PropTypes__default["default"].node,
  children: PropTypes__default["default"].oneOfType([PropTypes__default["default"].element.isRequired, PropTypes__default["default"].arrayOf(PropTypes__default["default"].element.isRequired)]).isRequired,
  className: PropTypes__default["default"].string,
  defaultOpen: PropTypes__default["default"].bool,
  defaultValue: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string.isRequired, PropTypes__default["default"].arrayOf(PropTypes__default["default"].string.isRequired)]),
  dialogSrLabel: PropTypes__default["default"].string,
  id: PropTypes__default["default"].string,
  menuMaxHeight: PropTypes__default["default"].number,
  multiple: PropTypes__default["default"].bool,
  onChange: PropTypes__default["default"].func,
  onClose: PropTypes__default["default"].func,
  onOpen: PropTypes__default["default"].func,
  open: PropTypes__default["default"].bool,
  placement: PropTypes__default["default"].oneOf(['bottom', 'bottom-start', 'bottom-end', 'top', 'top-start', 'top-end', 'left', 'left-start', 'left-end', 'right', 'right-start', 'right-end']),
  search: PropTypes__default["default"].oneOfType([PropTypes__default["default"].bool, PropTypes__default["default"].shape({
    defaultValue: PropTypes__default["default"].string,
    filterCondition: PropTypes__default["default"].func,
    onChange: PropTypes__default["default"].func,
    placeholder: PropTypes__default["default"].string
  })]),
  srLabel: PropTypes__default["default"].string.isRequired,
  topContent: PropTypes__default["default"].node,
  trigger: PropTypes__default["default"].element.isRequired,
  type: PropTypes__default["default"].oneOf(['menu', 'select']),
  value: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string.isRequired, PropTypes__default["default"].arrayOf(PropTypes__default["default"].string.isRequired)]),
  width: PropTypes__default["default"].oneOfType([PropTypes__default["default"].number, PropTypes__default["default"].string])
};
var FloatingMenu$1 = withTracker.withTracker(FloatingMenu, FloatingMenuMetrics);
const _excluded$1 = ["children"];

/**
 * @type {typeof import('../../index').FloatingMenuGroup
 * & { getCollectionNode: (props: typeof import('../../index').FloatingMenuGroup) => Generator }}
 */
const FloatingMenuGroup = _ref => {
  let children = _ref.children,
    props = _objectWithoutProperties__default["default"](_ref, _excluded$1);
  return /*#__PURE__*/React__default["default"].createElement(list.ListGroup, props, children);
};
FloatingMenuGroup.getCollectionNode = groupNodeGenerator__default["default"];
FloatingMenuGroup.displayName = GROUP_DISPLAY_NAME;
FloatingMenuGroup.propTypes = {
  children: PropTypes__default["default"].oneOfType([PropTypes__default["default"].element, PropTypes__default["default"].arrayOf(PropTypes__default["default"].element.isRequired)]).isRequired
};
var FloatingMenuGroup$1 = FloatingMenuGroup;
const _excluded = ["children", "description", "disabled", "groupIndex", "href", "index", "onClick", "rightContent", "selected", "title", "value"];

/** @typedef {import('../../index').FloatingMenuItemProps} FloatingMenuItemPublicProps */

/**
 * @typedef {Object} FloatingMenuItemInternalProps
 * @prop {number} [index]
 * @prop {number} [groupIndex]
 */

/** @typedef {FloatingMenuItemPublicProps & FloatingMenuItemInternalProps} FloatingMenuItemMergedProps */

/**
 * @type {React.FunctionComponent<FloatingMenuItemMergedProps>
 * & { getCollectionNode: (props: FloatingMenuItemMergedProps) => Generator }}
 */
const FloatingMenuItem = _ref => {
  let children = _ref.children,
    description = _ref.description,
    _ref$disabled = _ref.disabled,
    disabled = _ref$disabled === void 0 ? false : _ref$disabled,
    groupIndex = _ref.groupIndex,
    href = _ref.href,
    index = _ref.index,
    onClick = _ref.onClick,
    rightContent = _ref.rightContent,
    _ref$selected = _ref.selected,
    selected = _ref$selected === void 0 ? false : _ref$selected,
    title = _ref.title,
    value = _ref.value,
    props = _objectWithoutProperties__default["default"](_ref, _excluded);
  const itemProps = _objectSpread__default["default"]({
    description,
    disabled,
    groupIndex,
    index,
    rightContent,
    selected,
    title,
    value
  }, props);

  /** @type {(e: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>) => void} */
  const onClickHandler = e => {
    if (!disabled) {
      if (typeof onClick === 'function') {
        onClick(e, value);
      }
    }
  };
  if (href || onClick) {
    return /*#__PURE__*/React__default["default"].createElement(list.ListActionItem, _extends__default["default"]({}, itemProps, {
      href: href,
      onClick: onClickHandler
      // @ts-expect-error: internal props
      ,

      legacyBehaviorSlot: true,
      chevron: false
    }), children);
  }
  return /*#__PURE__*/React__default["default"].createElement(list.ListItem, _extends__default["default"]({}, itemProps, {
    onClick: onClickHandler
  }), children);
};
FloatingMenuItem.getCollectionNode = itemNodeGenerator__default["default"];
FloatingMenuItem.displayName = ITEM_DISPLAY_NAME;
FloatingMenuItem.propTypes = {
  children: PropTypes__default["default"].node,
  description: PropTypes__default["default"].node,
  disabled: PropTypes__default["default"].bool,
  groupIndex: PropTypes__default["default"].number,
  href: PropTypes__default["default"].string,
  index: PropTypes__default["default"].number,
  onClick: PropTypes__default["default"].func,
  rightContent: PropTypes__default["default"].node,
  selected: PropTypes__default["default"].bool,
  title: PropTypes__default["default"].string,
  value: PropTypes__default["default"].string.isRequired
};
var FloatingMenuItem$1 = FloatingMenuItem;
exports.FloatingMenu = FloatingMenu$1;
exports.FloatingMenuGroup = FloatingMenuGroup$1;
exports.FloatingMenuItem = FloatingMenuItem$1;
